// @docker/trident/src/components/account-menu.tsx
import type {PropsWithChildren, ReactNode} from 'react';

import {Avatar, AvatarFallback, AvatarImage} from '@/components/ui/avatar';
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuLabel,
  DropdownMenuSeparator,
  DropdownMenuTrigger,
} from '@/components/ui/dropdown-menu';
import {cn} from '@/lib/utils';

export type AccountMenuUser = {
  username: string;
  avatarUrl?: string | null;
} | null;

type AccountMenuProps = PropsWithChildren<{
  user: AccountMenuUser;
  chips?: ReactNode[];
  className?: string;
}>;

/**
 * User account avatar trigger + dropdown menu.
 */
export const AccountMenu = ({
  user,
  chips,
  children,
  className,
}: AccountMenuProps) => {
  const initials = user?.username
    ? user.username.slice(0, 2).toUpperCase()
    : '?';

  return (
    <DropdownMenu>
      <DropdownMenuTrigger asChild>
        <button
          aria-label={user?.username ?? 'Account menu'}
          className={cn(
            'rounded-full ring-offset-2 ring-offset-primary outline-none',
            'focus-visible:ring-2 focus-visible:ring-primary-foreground',
            className
          )}
        >
          <Avatar className="h-8 w-8">
            {user?.avatarUrl && (
              <AvatarImage src={user.avatarUrl} alt={user.username} />
            )}
            <AvatarFallback className="text-xs">{initials}</AvatarFallback>
          </Avatar>
        </button>
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end" className="min-w-[14rem]">
        {user && (
          <>
            <DropdownMenuLabel className="flex flex-col gap-dds-100">
              <span className="font-heading">{user.username}</span>
              {chips && chips.length > 0 && (
                <div className="flex flex-wrap gap-dds-100">{chips}</div>
              )}
            </DropdownMenuLabel>
            <DropdownMenuSeparator />
          </>
        )}
        {children}
      </DropdownMenuContent>
    </DropdownMenu>
  );
};
