// @docker/trident/src/components/agent-mark.tsx
import {cn} from '@/lib/utils';

// Docker agent mark geometry — three horizontal cuts with consistent 5.1-unit spacing.
// CIRCLE_RADIUS and CUT_HEIGHT are part of the Docker visual identity spec.
const CIRCLE_RADIUS = 9.5;
const CUT_HEIGHT = 2.5;
const CUT_Y_POSITIONS = [3.6, 8.7, 13.8] as const;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type AgentMarkState = 'default' | 'active' | 'resolved' | 'muted';

/**
 * Valid sizes for the agent mark. Below 16px uses mask-based rendering.
 * At 16px+ the mask rendering serves until full Dash vector paths are supplied.
 */
type AgentMarkSize = 8 | 10 | 12 | 16 | 20 | 24 | 48;

type AgentMarkProps = React.ComponentProps<'span'> & {
  state?: AgentMarkState;
  size?: AgentMarkSize;
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const STATE_CLASSES: Record<AgentMarkState, string> = {
  default: 'text-primary',
  active: 'text-primary animate-agent-breathing',
  resolved: 'text-success',
  muted: 'text-muted-foreground',
};

// Unique mask ID per-instance would require useId but this is a purely visual
// SVG; mask collision only matters when multiple marks appear on the same page
// in the same document. Using a stable ID keeps SSR output deterministic.
const MASK_ID = 'dds-agent-mark-mask';

// ---------------------------------------------------------------------------
// Component
// ---------------------------------------------------------------------------

/**
 * The Docker agent mark — a circle with three horizontal cuts rendered via
 * SVG mask. Communicates agent state (default, active, resolved, muted).
 *
 * @see §10 Agent Identity in the Trident Visual Language spec
 *
 * @wcag 1.1.1 — carries role="img" and aria-label; presentational by default.
 */
function AgentMark({
  state = 'default',
  size = 20,
  className,
  ...props
}: AgentMarkProps) {
  return (
    <span
      data-slot="agent-mark"
      data-state={state}
      className={cn('inline-flex shrink-0', STATE_CLASSES[state], className)}
      {...props}
    >
      <svg
        width={size}
        height={size}
        viewBox="0 0 20 20"
        fill="none"
        aria-label="Agent mark"
        role="img"
      >
        <defs>
          <mask id={MASK_ID}>
            <circle cx="10" cy="10" r={CIRCLE_RADIUS} fill="white" />
            {CUT_Y_POSITIONS.map(y => (
              <rect
                key={y}
                x="0"
                y={y}
                width="20"
                height={CUT_HEIGHT}
                fill="black"
              />
            ))}
          </mask>
        </defs>
        <circle
          cx="10"
          cy="10"
          r={CIRCLE_RADIUS}
          fill="currentColor"
          mask={`url(#${MASK_ID})`}
        />
      </svg>
    </span>
  );
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {AgentMark};
export type {AgentMarkProps, AgentMarkSize, AgentMarkState};
