// @docker/trident/src/components/agent-activity-line.tsx
'use client';

import {cn} from '@/lib/utils';

import {ChevronRightIcon} from 'lucide-react';
import * as React from 'react';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type AgentActivityLineState = 'running' | 'done' | 'error';

type AgentActivityLineProps = React.ComponentProps<'div'> & {
  /** Past-tense verb describing the action (e.g. "Read", "Created", "Ran"). */
  verb: string;
  /** File path, search query, or resource acted upon — rendered in monospace. */
  target: string;
  /** Outcome of the action (e.g. "PASS", "+145 lines") — colored by result. */
  result?: string;
  /**
   * When true, result is treated as an error state. Defaults to false.
   * @deprecated Use `state="error"` instead. Kept for backward compatibility.
   */
  isError?: boolean;
  /** Optional icon node rendered before the verb (e.g. a tool-specific Lucide icon). */
  icon?: React.ReactNode;
  /** Controls icon color and verb tense display. Defaults to "done". */
  state?: AgentActivityLineState;
  /**
   * Duration in milliseconds. Shown as "1.2s" or "340ms" when state is "done".
   * Only rendered when `state === "done"`.
   */
  duration?: number;
  /**
   * When true (or when children are provided), the row becomes expandable.
   * A ChevronRight rotates 90° on expand. Expanded content renders below
   * with a vertical connector line centered under the icon column.
   */
  expandable?: boolean;
  /** Expanded content slot. Implies expandable when provided. */
  children?: React.ReactNode;
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function formatDuration(ms: number): string {
  if (ms >= 60_000) {
    return `${Math.floor(ms / 60_000)}m ${Math.round((ms % 60_000) / 1000)}s`;
  }
  if (ms >= 1000) {
    return `${(ms / 1000).toFixed(1)}s`;
  }
  return `${Math.round(ms)}ms`;
}

const STATE_ICON_CLASSES: Record<AgentActivityLineState, string> = {
  running: 'text-muted-foreground',
  done: 'text-muted-foreground',
  error: 'text-destructive',
};

// ---------------------------------------------------------------------------
// Component
// ---------------------------------------------------------------------------

/**
 * A compact past-tense row showing a single agent action.
 * Verb in label weight, target in monospace, optional result in semantic color.
 *
 * Supports an expandable slot for detailed output, with a chevron indicator
 * and a vertical connector line when expanded.
 *
 * @see §11 Agent Patterns — Activity line in the Trident Visual Language spec
 *
 * @wcag 2.1.1 Keyboard — expandable rows use native button element
 * @wcag 4.1.2 Name, Role, Value — button carries aria-expanded
 */
function AgentActivityLine({
  verb,
  target,
  result,
  isError = false,
  icon,
  state = 'done',
  duration,
  expandable: expandableProp,
  children,
  className,
  ...props
}: AgentActivityLineProps) {
  const [open, setOpen] = React.useState(false);

  const isExpandable = expandableProp ?? Boolean(children);
  const resolvedState: AgentActivityLineState = isError ? 'error' : state;
  const iconColorClass = STATE_ICON_CLASSES[resolvedState];
  const resultColorClass =
    resolvedState === 'error' ? 'text-destructive' : 'text-success';

  // Shared row content rendered inside either a <button> or <div>
  const rowContent = (
    <>
      {icon !== undefined && (
        <span
          data-slot="activity-icon"
          aria-hidden="true"
          className={cn(
            'flex size-4 shrink-0 items-center justify-center',
            iconColorClass
          )}
        >
          {icon}
        </span>
      )}

      <span className="shrink-0 font-[520] text-muted-foreground">{verb}</span>

      <span className="min-w-0 flex-1 truncate font-mono text-foreground">
        {target}
      </span>

      {result !== undefined && (
        <span
          data-slot="activity-result"
          className={cn('shrink-0 font-mono', resultColorClass)}
        >
          {result}
        </span>
      )}

      {resolvedState === 'done' && duration !== undefined && (
        <span
          data-slot="activity-duration"
          className="shrink-0 font-mono text-[11px] text-muted-foreground"
        >
          {formatDuration(duration)}
        </span>
      )}

      {isExpandable && (
        <ChevronRightIcon
          aria-hidden="true"
          className={cn(
            'size-3 shrink-0 text-muted-foreground transition-transform duration-150',
            open && 'rotate-90'
          )}
        />
      )}
    </>
  );

  return (
    <div
      data-slot="activity-line"
      data-state={resolvedState}
      className={cn('flex flex-col', className)}
      {...props}
    >
      {/* Main row: native <button> when expandable, plain <div> when static */}
      {isExpandable ? (
        <button
          type="button"
          aria-expanded={open}
          onClick={() => setOpen(v => !v)}
          className="flex h-7 w-full items-center gap-2 rounded-sm text-left text-xs hover:bg-hover-overlay focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none"
        >
          {rowContent}
        </button>
      ) : (
        <div className="flex h-7 items-center gap-2 text-xs">{rowContent}</div>
      )}

      {/* Expanded content */}
      {isExpandable && open && children && (
        <div
          data-slot="activity-expanded"
          className={cn('flex gap-2', icon !== undefined ? 'pl-6' : 'pl-0')}
        >
          {/* Vertical connector line centered under the icon column */}
          {icon !== undefined && (
            <div
              aria-hidden="true"
              className="mr-1 ml-[-1.25rem] w-px shrink-0 self-stretch bg-border"
            />
          )}
          <div className="min-w-0 flex-1 py-1 pb-2 text-xs text-muted-foreground">
            {children}
          </div>
        </div>
      )}
    </div>
  );
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {AgentActivityLine};
export type {AgentActivityLineProps, AgentActivityLineState};
