// @docker/trident/src/components/agent-progress-summary.tsx
import {cn} from '@/lib/utils';

import {CheckCircle2Icon} from 'lucide-react';
import * as React from 'react';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type AgentProgressSummaryItem = {
  /** Human-readable label for the action type (e.g. "read", "inspected"). */
  label: string;
  /** Number of times this action occurred. */
  count: number;
};

type AgentProgressSummaryProps = React.ComponentProps<'div'> & {
  /** Total elapsed time in milliseconds. */
  duration: number;
  /**
   * Breakdown of actions performed.
   * Each item renders as "N label" separated by · dots.
   */
  items: AgentProgressSummaryItem[];
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function formatDuration(ms: number): string {
  if (ms >= 1000) {
    return `${(ms / 1000).toFixed(1)}s`;
  }
  return `${Math.round(ms)}ms`;
}

// ---------------------------------------------------------------------------
// Component
// ---------------------------------------------------------------------------

/**
 * A compact one-line summary of a completed agent operation.
 * Renders a success icon, total duration, and a dot-separated list of
 * action counts (e.g. "1.2s · 3 read · 2 inspected").
 *
 * @see §11 Agent Patterns — Activity line in the Trident Visual Language spec
 */
function AgentProgressSummary({
  duration,
  items,
  className,
  ...props
}: AgentProgressSummaryProps) {
  return (
    <div
      data-slot="progress-summary"
      className={cn(
        'flex items-center gap-1.5 font-mono text-[11px] text-muted-foreground',
        className
      )}
      {...props}
    >
      <CheckCircle2Icon
        aria-hidden="true"
        className="size-3 shrink-0 text-success"
      />

      <span>{formatDuration(duration)}</span>

      {items.map(item => (
        <React.Fragment key={item.label}>
          <span aria-hidden="true">·</span>
          <span>
            {item.count} {item.label}
          </span>
        </React.Fragment>
      ))}
    </div>
  );
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {AgentProgressSummary};
export type {AgentProgressSummaryItem, AgentProgressSummaryProps};
