// @docker/trident/src/components/agent-approval-card.tsx
import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import {ShieldAlertIcon} from 'lucide-react';
import * as React from 'react';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type AgentApprovalCardProps = React.ComponentProps<'div'> & {
  /** The name of the tool requesting approval, rendered in monospace. */
  toolName: string;
  /** Human-readable summary of the tool arguments. */
  argsPreview?: string;
  /**
   * When true, uses a stronger destructive border to signal write/delete risk.
   * @default false
   */
  destructive?: boolean;
  /** Called when the user approves the tool call. */
  onApprove?: () => void;
  /** Called when the user rejects the tool call. */
  onReject?: () => void;
  /** Optional slot for custom content between the args preview and action row. */
  children?: React.ReactNode;
};

// ---------------------------------------------------------------------------
// Component
// ---------------------------------------------------------------------------

/**
 * A warning-toned bordered card for tool confirmation prompts.
 * Renders a header with the tool name, an optional args preview,
 * and Reject / Approve action buttons.
 *
 * Use `destructive` for write or delete operations that carry higher risk.
 *
 * @see §11 Agent Patterns — Pending item in the Trident Visual Language spec
 *
 * @wcag 1.4.3 — warning and destructive colors meet contrast requirements via DDS tokens
 * @wcag 4.1.2 — action buttons carry accessible labels via text content
 */
function AgentApprovalCard({
  toolName,
  argsPreview,
  destructive = false,
  onApprove,
  onReject,
  children,
  className,
  ...props
}: AgentApprovalCardProps) {
  return (
    <div
      data-slot="approval-card"
      data-destructive={destructive || undefined}
      className={cn(
        'rounded-dds-radius-sm flex flex-col overflow-hidden border bg-warning-wash',
        destructive ? 'border-destructive/40' : 'border-warning-subtle',
        className
      )}
      {...props}
    >
      {/* Header row */}
      <div className="flex items-center gap-1.5 px-3 py-2">
        <ShieldAlertIcon
          aria-hidden="true"
          className={cn(
            'size-3.5 shrink-0',
            destructive ? 'text-destructive' : 'text-warning'
          )}
        />
        <span className="min-w-0 flex-1 truncate font-mono text-xs font-medium text-foreground">
          {toolName}
        </span>
        <span className="shrink-0 text-[11px] text-muted-foreground">
          Needs approval
        </span>
      </div>

      {/* Args preview */}
      {argsPreview && (
        <div className="px-3 pb-2">
          <p className="font-mono text-[11px] text-muted-foreground">
            {argsPreview}
          </p>
        </div>
      )}

      {/* Custom content slot */}
      {children && (
        <div className="px-3 pb-2 text-xs text-foreground">{children}</div>
      )}

      {/* Action row */}
      <div className="flex items-center justify-between border-t border-border px-3 py-1.5">
        <Button
          variant="ghost"
          size="xs"
          onClick={onReject}
          className="text-muted-foreground hover:text-foreground"
        >
          Reject
        </Button>
        <Button variant="default" size="xs" onClick={onApprove}>
          Approve
        </Button>
      </div>
    </div>
  );
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {AgentApprovalCard};
export type {AgentApprovalCardProps};
