// @docker/trident/src/components/agent-composer.tsx
'use client';

import {cn} from '@/lib/utils';

import {ArrowUpIcon} from 'lucide-react';
import * as React from 'react';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type AgentComposerProps = React.ComponentProps<'div'> & {
  /** Placeholder text shown when the textarea is empty. */
  placeholder?: string;
  /** When true, disables the textarea and send button. */
  disabled?: boolean;
  /** Called with the trimmed text when the user submits. */
  onSubmit: (text: string) => void;
  /**
   * Slot for controls rendered in the bottom-left of the composer
   * (e.g. toggle buttons, attachment pickers).
   */
  children?: React.ReactNode;
};

// ---------------------------------------------------------------------------
// Component
// ---------------------------------------------------------------------------

/**
 * A two-row input container for composing agent messages.
 *
 * - Top row: auto-resizing textarea (min 36px, max 160px)
 * - Bottom row: children slot on left, send button on right
 * - Enter submits; Shift+Enter inserts a newline
 * - Send button appears only when the textarea has content
 *
 * The textarea intentionally suppresses the global :focus-visible outline.
 * The cursor position is a sufficient focus indicator for text inputs per
 * WCAG 2.4.7 (Focus Visible). The outer container provides structural context.
 *
 * @wcag 2.1.1 Keyboard — Enter submits; Shift+Enter inserts newline
 * @wcag 2.4.7 Focus Visible — cursor is a valid focus indicator for text inputs
 * @wcag 4.1.2 Name, Role, Value — send button carries aria-label
 */
function AgentComposer({
  placeholder = 'Message…',
  disabled = false,
  onSubmit,
  children,
  className,
  ...props
}: AgentComposerProps) {
  const [value, setValue] = React.useState('');
  const textareaRef = React.useRef<HTMLTextAreaElement>(null);

  const hasContent = value.trim().length > 0;

  function handleKeyDown(e: React.KeyboardEvent<HTMLTextAreaElement>) {
    if (e.key === 'Enter' && !e.shiftKey) {
      e.preventDefault();
      submit();
    }
  }

  function submit() {
    const trimmed = value.trim();
    if (!trimmed || disabled) return;
    onSubmit(trimmed);
    setValue('');
    // Reset textarea height after clearing
    if (textareaRef.current) {
      textareaRef.current.style.height = 'auto';
    }
  }

  function handleInput(e: React.ChangeEvent<HTMLTextAreaElement>) {
    setValue(e.target.value);
    // Auto-resize: reset then grow to scrollHeight, capped at 160px
    const el = e.target;
    el.style.height = 'auto';
    el.style.height = `${Math.min(el.scrollHeight, 160)}px`;
  }

  return (
    <div
      data-slot="agent-composer"
      className={cn(
        'flex flex-col rounded-xl border border-border bg-popover',
        disabled && 'opacity-50',
        className
      )}
      {...props}
    >
      {/* Textarea row */}
      <textarea
        ref={textareaRef}
        data-slot="composer-textarea"
        value={value}
        onChange={handleInput}
        onKeyDown={handleKeyDown}
        placeholder={placeholder}
        disabled={disabled}
        rows={1}
        aria-label={placeholder}
        className={cn(
          'w-full resize-none bg-transparent px-3 pt-2.5 pb-1 text-sm text-foreground placeholder:text-muted-foreground disabled:cursor-not-allowed',
          // min-h for single line, max-h enforced via JS
          'min-h-[36px]'
        )}
        style={{
          outline: 'none',
          overflowY: 'auto',
        }}
      />

      {/* Bottom row */}
      <div className="flex items-center justify-between px-2 pb-2">
        {/* Left slot for controls */}
        <div className="flex items-center gap-1">{children}</div>

        {/* Send button */}
        <button
          type="button"
          data-slot="composer-send"
          aria-label="Send message"
          disabled={disabled || !hasContent}
          onClick={submit}
          className={cn(
            'flex size-7 shrink-0 items-center justify-center rounded-full bg-primary text-primary-foreground transition-opacity duration-150',
            'focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none',
            'disabled:pointer-events-none',
            hasContent ? 'opacity-100' : 'pointer-events-none opacity-0'
          )}
        >
          <ArrowUpIcon aria-hidden="true" className="size-3.5" />
        </button>
      </div>
    </div>
  );
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {AgentComposer};
export type {AgentComposerProps};
