// @docker/trident/src/components/app-header.tsx
import type {AsChild} from '@/lib/types';
import type {PropsWithChildren, ReactNode} from 'react';

import {Input} from '@/components/ui/input';
import {SidebarTrigger} from '@/components/ui/sidebar';
import {cn} from '@/lib/utils';

import {Slot} from 'radix-ui';

// ---------------------------------------------------------------------------
// AppHeaderLeftNav
// ---------------------------------------------------------------------------

type AppHeaderLeftNavProps = PropsWithChildren<{
  className?: string;
}>;

type AppHeaderLinkProps = React.ComponentProps<'a'> &
  AsChild & {
    isActive?: boolean;
  };

type AppHeaderMetaProps = React.ComponentProps<'span'>;

type AppHeaderPillProps = PropsWithChildren<
  React.ComponentProps<'span'> & AsChild
>;

type AppHeaderBrandProps = {
  icon?: ReactNode;
  title: ReactNode;
  meta?: ReactNode;
  className?: string;
};

type AppHeaderSearchFieldProps = React.ComponentProps<typeof Input>;

const APP_HEADER_INTERACTIVE =
  'outline-hidden transition-colors focus-visible:ring-3 focus-visible:ring-ring';

/**
 * Navigation links displayed inside AppHeader on the left side.
 */
const AppHeaderLeftNav = ({children, className}: AppHeaderLeftNavProps) => {
  return (
    <nav
      aria-label="Primary navigation"
      className={cn('flex items-center gap-dds-150', className)}
    >
      {children}
    </nav>
  );
};

const AppHeaderLink = ({
  asChild = false,
  isActive = false,
  className,
  ...props
}: AppHeaderLinkProps) => {
  const Comp = asChild ? Slot.Root : 'a';

  return (
    <Comp
      className={cn(
        APP_HEADER_INTERACTIVE,
        'inline-flex items-center rounded-[calc(var(--radius)+1px)] px-dds-200 py-dds-100 whitespace-nowrap',
        'text-label leading-label font-label tracking-label text-header-foreground-muted',
        'hover:bg-primary-foreground-wash hover:text-header-foreground',
        isActive && 'bg-primary-foreground-wash text-header-foreground',
        className
      )}
      {...props}
    />
  );
};

const AppHeaderPill = ({
  asChild = false,
  className,
  children,
  ...props
}: AppHeaderPillProps) => {
  const Comp = asChild ? Slot.Root : 'span';

  return (
    <Comp
      className={cn(
        APP_HEADER_INTERACTIVE,
        'inline-flex items-center gap-dds-150 rounded-full bg-primary-foreground-wash px-dds-200 py-dds-100 whitespace-nowrap text-header-foreground',
        className
      )}
      {...props}
    >
      {children}
    </Comp>
  );
};

const AppHeaderMeta = ({className, ...props}: AppHeaderMetaProps) => {
  return (
    <span
      className={cn(
        'whitespace-nowrap text-header-foreground-muted',
        className
      )}
      {...props}
    />
  );
};

const AppHeaderBrand = ({
  icon,
  title,
  meta,
  className,
}: AppHeaderBrandProps) => {
  return (
    <div className={cn('flex min-w-0 items-center gap-dds-200', className)}>
      {icon ? (
        <span className="flex shrink-0 items-center justify-center text-header-foreground">
          {icon}
        </span>
      ) : null}
      <div className="grid min-w-0 gap-[2px] leading-none">
        <span className="truncate text-label leading-label font-label tracking-label text-header-foreground">
          {title}
        </span>
        {meta ? (
          <AppHeaderMeta className="truncate text-micro leading-snug tracking-caption">
            {meta}
          </AppHeaderMeta>
        ) : null}
      </div>
    </div>
  );
};

const AppHeaderSearchField = ({
  className,
  ...props
}: AppHeaderSearchFieldProps) => {
  return (
    <Input
      className={cn('h-8 w-48 shadow-none lg:w-56', className)}
      {...props}
    />
  );
};

// ---------------------------------------------------------------------------
// AppHeader
// ---------------------------------------------------------------------------

type AppHeaderProps = {
  logo: ReactNode;
  leftNav?: ReactNode;
  rightNav: ReactNode;
  search?: ReactNode;
  className?: string;
};

/**
 * Sticky application header for Docker apps.
 * Height controlled by `--app-header-height` (3.5rem).
 */
const AppHeader = ({
  logo,
  leftNav,
  rightNav,
  search,
  className,
}: AppHeaderProps) => {
  return (
    <header
      className={cn(
        'sticky top-0 h-[var(--app-header-height)]',
        'z-sticky',
        'bg-gradient-to-r from-header-from to-header-to text-header-foreground',
        'flex items-center gap-dds-300 px-dds-400',
        'border-b border-accent',
        className
      )}
    >
      <div className="flex min-w-0 flex-1 items-center gap-dds-200">
        <SidebarTrigger className="shrink-0 text-header-foreground hover:bg-primary-foreground-wash" />
        <div className="shrink-0">{logo}</div>
        {leftNav && (
          <div className="flex min-w-0 items-center gap-dds-200">{leftNav}</div>
        )}
      </div>
      <div className="flex min-w-0 items-center gap-dds-200">
        {search && <div className="min-w-0">{search}</div>}
        <div className="shrink-0">{rightNav}</div>
      </div>
    </header>
  );
};

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {
  AppHeader,
  AppHeaderBrand,
  AppHeaderLeftNav,
  AppHeaderLink,
  AppHeaderMeta,
  AppHeaderPill,
  AppHeaderSearchField,
};
export type {
  AppHeaderLeftNavProps,
  AppHeaderBrandProps,
  AppHeaderLinkProps,
  AppHeaderMetaProps,
  AppHeaderPillProps,
  AppHeaderProps,
  AppHeaderSearchFieldProps,
};
