// @docker/trident/src/components/app-footer.tsx
import type {PropsWithChildren, ReactElement} from 'react';

import {Separator} from '@/components/ui/separator';
import {cn} from '@/lib/utils';

import {Rss} from 'lucide-react';
import {useId} from 'react';

type AppFooterProps = PropsWithChildren<{
  showLinks?: boolean;
  enableCookieButton?: boolean;
  className?: string;
}>;

/**
 * Docker application footer.
 * Always renders the legal bar (copyright, ToS, Privacy, Legal) and social links.
 * When showLinks=true (default), also renders Docker logo + FooterLinkColumn children.
 */
export const AppFooter = ({
  children,
  showLinks = true,
  enableCookieButton = false,
  className,
}: AppFooterProps) => {
  const year = new Intl.DateTimeFormat('en-US', {year: 'numeric'}).format(
    Date.now()
  );

  return (
    <footer
      className={cn(
        'shrink-0 bg-sidebar py-dds-600 text-sidebar-foreground',
        className
      )}
    >
      <div className="mx-auto max-w-screen-2xl px-dds-600">
        {showLinks && (
          <>
            <div className="mb-dds-600 flex flex-wrap items-start gap-dds-800">
              <span className="text-lg font-bold text-foreground">Docker</span>
              {children}
              <FooterCompanyColumn />
            </div>
            <Separator className="mb-dds-600 bg-sidebar-border" />
          </>
        )}
        <div className="flex flex-col items-start justify-between gap-dds-400 md:flex-row md:items-center">
          <div className="flex flex-wrap gap-x-dds-400 gap-y-dds-100 text-sm text-muted-foreground">
            <span className="whitespace-nowrap">
              &copy; {year} Docker, Inc. All rights reserved.
            </span>
            <span aria-hidden="true">|</span>
            <a
              href="https://www.docker.com/legal/docker-terms-service"
              target="_blank"
              rel="noopener noreferrer"
              className="underline underline-offset-4 hover:text-foreground"
            >
              Terms of Service
            </a>
            <span aria-hidden="true">|</span>
            <a
              href="https://www.docker.com/legal/docker-subscription-service-agreement"
              target="_blank"
              rel="noopener noreferrer"
              className="underline underline-offset-4 hover:text-foreground"
            >
              Subscription Service Agreement
            </a>
            <span aria-hidden="true">|</span>
            <a
              href="https://www.docker.com/legal/privacy"
              target="_blank"
              rel="noopener noreferrer"
              className="underline underline-offset-4 hover:text-foreground"
            >
              Privacy
            </a>
            <span aria-hidden="true">|</span>
            <a
              href="https://www.docker.com/legal"
              target="_blank"
              rel="noopener noreferrer"
              className="underline underline-offset-4 hover:text-foreground"
            >
              Legal
            </a>
          </div>
          <SocialLinks />
        </div>
        {enableCookieButton && (
          <button
            type="button"
            id="ot-sdk-btn"
            className="chromatic-ignore ot-sdk-show-settings mt-dds-400 text-sm text-muted-foreground underline underline-offset-4"
          >
            Cookies Settings
          </button>
        )}
      </div>
    </footer>
  );
};

type FooterLinkColumnProps = PropsWithChildren<{
  heading: string;
  className?: string;
}>;

/**
 * A labelled column of links for use inside AppFooter.
 */
export const FooterLinkColumn = ({
  heading,
  children,
  className,
}: FooterLinkColumnProps) => {
  const uid = useId();
  const id = `footer-${uid}-heading`;
  return (
    <section
      aria-labelledby={id}
      className={cn('flex min-w-[8rem] flex-col gap-dds-200', className)}
    >
      <h2 id={id} className="text-sm font-heading text-foreground">
        {heading}
      </h2>
      <div className="flex flex-col gap-dds-100 text-sm text-muted-foreground">
        {children}
      </div>
    </section>
  );
};

const COMPANY_LINKS = [
  {label: 'About Us', href: 'https://www.docker.com/company'},
  {label: 'Blog', href: 'https://www.docker.com/blog/'},
  {label: 'Careers', href: 'https://www.docker.com/careers'},
  {label: 'Contact Us', href: 'https://www.docker.com/company/contact'},
  {label: 'Newsroom', href: 'https://www.docker.com/company/newsroom'},
  {label: 'System Status', href: 'https://www.dockerstatus.com'},
] as const;

const FooterCompanyColumn = () => (
  <FooterLinkColumn heading="Company">
    {COMPANY_LINKS.map(({label, href}) => (
      <a
        key={href}
        href={href}
        target="_blank"
        rel="noopener noreferrer"
        className="hover:text-foreground"
      >
        {label}
      </a>
    ))}
  </FooterLinkColumn>
);

const LinkedInIcon = () => (
  <svg
    aria-hidden="true"
    viewBox="0 0 24 24"
    fill="currentColor"
    className="size-4"
  >
    <path d="M16 8a6 6 0 0 1 6 6v7h-4v-7a2 2 0 0 0-2-2 2 2 0 0 0-2 2v7h-4v-7a6 6 0 0 1 6-6z" />
    <rect width="4" height="12" x="2" y="9" />
    <circle cx="4" cy="4" r="2" />
  </svg>
);

const XIcon = () => (
  <svg
    aria-hidden="true"
    viewBox="0 0 24 24"
    fill="currentColor"
    className="size-4"
  >
    <path d="M18.244 2.25h3.308l-7.227 8.26 8.502 11.24H16.17l-4.714-6.231-5.401 6.231H2.744l7.73-8.835L1.254 2.25H8.08l4.713 6.231zm-1.161 17.52h1.833L7.084 4.126H5.117z" />
  </svg>
);

const YouTubeIcon = () => (
  <svg
    aria-hidden="true"
    viewBox="0 0 24 24"
    fill="currentColor"
    className="size-4"
  >
    <path d="M22.54 6.42a2.78 2.78 0 0 0-1.95-1.96C18.88 4 12 4 12 4s-6.88 0-8.59.46a2.78 2.78 0 0 0-1.95 1.96A29 29 0 0 0 1 12a29 29 0 0 0 .46 5.58 2.78 2.78 0 0 0 1.95 1.95C5.12 20 12 20 12 20s6.88 0 8.59-.47a2.78 2.78 0 0 0 1.95-1.95A29 29 0 0 0 23 12a29 29 0 0 0-.46-5.58z" />
    <polygon points="9.75 15.02 15.5 12 9.75 8.98 9.75 15.02" />
  </svg>
);

const SOCIAL_LINKS: Array<{
  label: string;
  href: string;
  Icon: () => ReactElement;
}> = [
  {
    label: 'Visit our LinkedIn page',
    href: 'https://www.linkedin.com/company/docker',
    Icon: LinkedInIcon,
  },
  {label: 'Visit our X page', href: 'https://x.com/docker', Icon: XIcon},
  {
    label: 'Visit our YouTube page',
    href: 'https://www.youtube.com/user/dockerrun',
    Icon: YouTubeIcon,
  },
  {
    label: 'Visit our RSS feed',
    href: 'https://www.docker.com/feed/',
    Icon: () => <Rss aria-hidden="true" className="size-4" />,
  },
];

const SocialLinks = () => (
  <div className="flex gap-dds-100">
    {SOCIAL_LINKS.map(({label, href, Icon}) => (
      <a
        key={href}
        href={href}
        target="_blank"
        rel="noopener noreferrer"
        aria-label={label}
        className="rounded p-dds-200 text-muted-foreground transition-colors hover:text-foreground"
      >
        <Icon />
      </a>
    ))}
  </div>
);
