// @docker/trident/src/components/blockquote.tsx
// blockquote.tsx

import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';

const blockquoteVariants = cva('my-2 border-l-4 px-4 py-3 text-sm', {
  variants: {
    variant: {
      note: 'border-l-blockquote-note bg-blockquote-note-bg',
      tip: 'border-l-blockquote-tip bg-blockquote-tip-bg',
      important: 'border-l-blockquote-important bg-blockquote-important-bg',
      warning: 'border-l-blockquote-warning bg-blockquote-warning-bg',
      caution: 'border-l-blockquote-caution bg-blockquote-caution-bg',
    },
  },
  defaultVariants: {
    variant: 'note',
  },
});

const titleColorMap = {
  note: 'text-blockquote-note-title',
  tip: 'text-blockquote-tip-title',
  important: 'text-blockquote-important-title',
  warning: 'text-blockquote-warning-title',
  caution: 'text-blockquote-caution-title',
} as const;

type BlockquoteProps = React.ComponentProps<'blockquote'> & {
  /**
   * The visual variant of the blockquote.
   * @default "note"
   */
  variant?: VariantProps<typeof blockquoteVariants>['variant'];
  /** Optional title displayed above the content. */
  title?: React.ReactNode;
};

/**
 * Callout / admonition block with a colored left border for notes, tips,
 * warnings and other supplementary information.
 *
 * ### Component Props
 *
 * | Prop       | Type                                                    | Default  |
 * | :--------- | :------------------------------------------------------ | :------- |
 * | _variant_  | `"note" \| "tip" \| "important" \| "warning" \| "caution"` | `"note"` |
 * | _title_    | `ReactNode`                                             | -        |
 * | _children_ | `ReactNode`                                             | -        |
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.4.1 Use of Color — variants use both a colored border and a title label to convey meaning.
 * @wcag 1.4.3 Contrast (Minimum) — title and body text meet 4.5:1 contrast via DDS semantic tokens.
 * @wcag 4.1.2 Name, Role, Value — renders a semantic `<blockquote>` element.
 */
function Blockquote({
  variant = 'note',
  title,
  className,
  children,
  ...props
}: BlockquoteProps) {
  return (
    <blockquote
      data-slot="blockquote"
      data-variant={variant}
      className={cn(blockquoteVariants({variant}), className)}
      {...props}
    >
      {title && (
        <p
          className={cn(
            'mb-1 text-sm font-heading',
            titleColorMap[variant ?? 'note']
          )}
        >
          {title}
        </p>
      )}
      <div className="text-sm text-foreground">{children}</div>
    </blockquote>
  );
}

export {Blockquote, blockquoteVariants};
export type {BlockquoteProps};
