// @docker/trident/src/components/code-diff-viewer.tsx
'use client';

import {
  type DiffHunk,
  type DiffLine,
  formatLineNumber,
  gutterChar,
  gutterCharClasses,
  lineClasses,
  parseDiff,
  type SplitRow,
  toSplitRows,
} from '@/components/code-diff-viewer-utils';
import {Button} from '@/components/ui/button';
import {ButtonGroup} from '@/components/ui/button-group';
import {cn} from '@/lib/utils';

import {AlignLeftIcon, ChevronDownIcon, Columns2Icon} from 'lucide-react';
import {useCallback, useState} from 'react';

// ─── Public Types ────────────────────────────────────────────────────────────

export type DiffViewerProps = {
  /** Original / before content */
  original: string;
  /** Modified / after content */
  modified: string;
  /** Syntax context label shown in header: "dockerfile" | "yaml" | "json" | "bash" */
  language?: string;
  /** Left-panel header label. @default "Before" */
  originalLabel?: string;
  /** Right-panel header label. @default "After" */
  modifiedLabel?: string;
  /** Layout mode. @default "split" */
  mode?: 'split' | 'unified';
  /** Show line numbers in gutter. @default true */
  showLineNumbers?: boolean;
  /** Unchanged context lines shown around each change. @default 3 */
  expandContext?: number;
  /** Optional filename shown in the header bar. */
  filename?: string;
  className?: string;
};

// ─── Shared line-number gutter cell ─────────────────────────────────────────

const LineNumberCell = ({
  value,
  show,
}: {
  value: string | undefined;
  show: boolean;
}) => {
  if (!show) return null;
  return (
    <span className="min-w-[3ch] pr-3 text-right font-mono text-xs text-muted-foreground select-none">
      {value ?? ''}
    </span>
  );
};

// ─── Single unified-mode row ─────────────────────────────────────────────────

const UnifiedRow = ({
  line,
  showLineNumbers,
}: {
  line: DiffLine;
  showLineNumbers: boolean;
}) => {
  const char = gutterChar[line.type] ?? ' ';
  const charCls = gutterCharClasses[line.type] ?? '';

  return (
    <div className={cn('flex min-w-0', lineClasses[line.type] ?? '')}>
      <div className="flex shrink-0 items-center gap-1 border-r border-border px-2 select-none">
        <LineNumberCell
          value={formatLineNumber(line.oldLineNumber) || undefined}
          show={showLineNumbers}
        />
        <LineNumberCell
          value={formatLineNumber(line.newLineNumber) || undefined}
          show={showLineNumbers}
        />
        <span className={cn('w-3 font-mono text-xs', charCls)}>{char}</span>
      </div>
      <pre className="min-w-0 flex-1 overflow-x-auto px-3 py-0 font-mono text-sm leading-5">
        <code>{line.content}</code>
      </pre>
    </div>
  );
};

// ─── Single split-mode row pair ──────────────────────────────────────────────

const SplitRowPair = ({
  row,
  showLineNumbers,
}: {
  row: SplitRow;
  showLineNumbers: boolean;
}) => {
  const renderSide = (line: DiffLine | null, side: 'left' | 'right') => {
    const type = line?.type ?? 'context';
    const char = line ? (gutterChar[type] ?? ' ') : ' ';
    const charCls = line ? (gutterCharClasses[type] ?? '') : '';
    const lineNumValue =
      side === 'left'
        ? formatLineNumber(line?.oldLineNumber) || undefined
        : formatLineNumber(line?.newLineNumber) || undefined;

    return (
      // tabIndex={0} is required: WCAG scrollable-region-focusable mandates
      // keyboard access for overflow-x-auto regions
      <div
        tabIndex={0}
        className={cn(
          'flex min-w-0 flex-1 overflow-x-auto',
          line ? (lineClasses[type] ?? '') : 'bg-background'
        )}
      >
        <div className="flex shrink-0 items-center gap-1 border-r border-border px-2 select-none">
          <LineNumberCell value={lineNumValue} show={showLineNumbers} />
          <span className={cn('w-3 font-mono text-xs', charCls)}>{char}</span>
        </div>
        <pre className="min-w-0 flex-1 px-3 py-0 font-mono text-sm leading-5">
          <code>{line?.content ?? ''}</code>
        </pre>
      </div>
    );
  };

  return (
    <div className="flex divide-x divide-border">
      {renderSide(row.left, 'left')}
      {renderSide(row.right, 'right')}
    </div>
  );
};

// ─── Hunk expander button ────────────────────────────────────────────────────

const ExpanderRow = ({
  hiddenCount,
  onExpand,
}: {
  hiddenCount: number;
  onExpand: () => void;
}) => (
  <div className="flex items-center justify-center border-y border-border bg-diff-hunk-bg py-0.5">
    <Button
      variant="ghost"
      size="xs"
      className="gap-1 text-muted-foreground hover:text-foreground"
      onClick={onExpand}
    >
      <ChevronDownIcon className="size-3" />
      Show {hiddenCount} more line{hiddenCount !== 1 ? 's' : ''}
    </Button>
  </div>
);

// ─── Hunk renderer (handles context-collapse internally) ─────────────────────

const HunkBlock = ({
  hunk,
  mode,
  showLineNumbers,
  expandContext,
}: {
  hunk: DiffHunk;
  mode: 'split' | 'unified';
  showLineNumbers: boolean;
  expandContext: number;
}) => {
  // Track which "between-change" context runs have been expanded
  const [expandedRuns, setExpandedRuns] = useState<Set<number>>(new Set());

  // Split the hunk lines into segments: change blocks and context runs
  type Segment =
    | {kind: 'change'; lines: DiffLine[]}
    | {kind: 'context-run'; lines: DiffLine[]; runIndex: number};

  const segments: Segment[] = [];
  let runIndex = 0;
  let i = 0;
  const lines = hunk.lines;

  while (i < lines.length) {
    if (lines[i].type === 'context') {
      const start = i;
      while (i < lines.length && lines[i].type === 'context') i++;
      segments.push({
        kind: 'context-run',
        lines: lines.slice(start, i),
        runIndex: runIndex++,
      });
    } else {
      const start = i;
      while (i < lines.length && lines[i].type !== 'context') i++;
      segments.push({kind: 'change', lines: lines.slice(start, i)});
    }
  }

  const handleExpand = useCallback((ri: number) => {
    setExpandedRuns(prev => {
      const next = new Set(prev);
      next.add(ri);
      return next;
    });
  }, []);

  const renderLines = (diffLines: DiffLine[]) => {
    if (mode === 'unified') {
      return diffLines.map(line => {
        // Derive a stable key from line numbers: one side is always defined
        const key = `u-${line.type}-${line.oldLineNumber ?? 'x'}-${line.newLineNumber ?? 'x'}`;
        return (
          <UnifiedRow key={key} line={line} showLineNumbers={showLineNumbers} />
        );
      });
    }
    // Split mode: pair removed/added lines
    const rows = toSplitRows(diffLines);
    return rows.map(row => {
      const leftKey = row.left
        ? `${row.left.type}-${row.left.oldLineNumber}`
        : 'null';
      const rightKey = row.right
        ? `${row.right.type}-${row.right.newLineNumber}`
        : 'null';
      const key = `s-${leftKey}-${rightKey}`;
      return (
        <SplitRowPair key={key} row={row} showLineNumbers={showLineNumbers} />
      );
    });
  };

  return (
    <>
      {segments.map((seg, si) => {
        // Stable key: encode segment position and identity
        const segKey =
          seg.kind === 'change' ? `change-${si}` : `ctx-${seg.runIndex}`;

        if (seg.kind === 'change') {
          return (
            <div key={segKey} data-testid="diff-change-block">
              {renderLines(seg.lines)}
            </div>
          );
        }

        // Context run — collapse if it's an interior run (not first/last)
        // and has more than 2 * expandContext lines
        const isFirst = si === 0;
        const isLast = si === segments.length - 1;
        const isInterior = !isFirst && !isLast;
        const runLines = seg.lines;
        const isExpanded = expandedRuns.has(seg.runIndex);

        if (isInterior && runLines.length > expandContext * 2 && !isExpanded) {
          // Show last `expandContext` of previous change and first `expandContext` of next
          const head = runLines.slice(0, expandContext);
          const tail = runLines.slice(runLines.length - expandContext);
          const hidden = runLines.length - expandContext * 2;

          return (
            <div key={segKey} data-testid="diff-context-collapsed">
              {renderLines(head)}
              <ExpanderRow
                hiddenCount={hidden}
                onExpand={() => handleExpand(seg.runIndex)}
              />
              {renderLines(tail)}
            </div>
          );
        }

        // For leading/trailing context, also trim if longer than expandContext
        if (
          (isFirst || isLast) &&
          runLines.length > expandContext &&
          !isExpanded
        ) {
          const visible = isFirst
            ? runLines.slice(runLines.length - expandContext)
            : runLines.slice(0, expandContext);
          const hidden = runLines.length - expandContext;

          return (
            <div key={segKey} data-testid="diff-context-edge">
              {isFirst && (
                <ExpanderRow
                  hiddenCount={hidden}
                  onExpand={() => handleExpand(seg.runIndex)}
                />
              )}
              {renderLines(visible)}
              {isLast && (
                <ExpanderRow
                  hiddenCount={hidden}
                  onExpand={() => handleExpand(seg.runIndex)}
                />
              )}
            </div>
          );
        }

        return (
          <div key={segKey} data-testid="diff-context-run">
            {renderLines(runLines)}
          </div>
        );
      })}
    </>
  );
};

// ─── Main component ──────────────────────────────────────────────────────────

const CodeDiffViewer = ({
  original,
  modified,
  language,
  originalLabel = 'Before',
  modifiedLabel = 'After',
  mode: modeProp = 'split',
  showLineNumbers = true,
  expandContext = 3,
  filename,
  className,
}: DiffViewerProps) => {
  const [mode, setMode] = useState<'split' | 'unified'>(modeProp);
  const parsed = parseDiff(original, modified);

  // ── No-changes state ──
  if (!parsed.hasChanges) {
    return (
      <div
        data-slot="code-diff-viewer"
        className={cn(
          'overflow-hidden rounded border border-border bg-muted font-mono text-sm',
          className
        )}
      >
        <div className="flex items-center justify-center py-8 text-muted-foreground">
          No changes
        </div>
      </div>
    );
  }

  // ── Header ──
  const headerLeft = filename ? (
    <span className="truncate font-mono text-xs text-foreground">
      {filename}
    </span>
  ) : (
    <span className="flex items-center gap-2 text-xs text-muted-foreground">
      <span>{originalLabel}</span>
      <span aria-hidden>→</span>
      <span>{modifiedLabel}</span>
      {language && (
        <span className="rounded bg-muted px-1.5 py-0.5 text-xs capitalize">
          {language}
        </span>
      )}
    </span>
  );

  // ── Split-mode column headers ──
  const splitColumnHeaders =
    mode === 'split' ? (
      <div className="flex divide-x divide-border border-b border-border bg-muted text-xs text-muted-foreground">
        <div className="flex-1 px-3 py-1">{originalLabel}</div>
        <div className="flex-1 px-3 py-1">{modifiedLabel}</div>
      </div>
    ) : null;

  return (
    <section
      data-slot="code-diff-viewer"
      aria-label={filename ? `Code diff: ${filename}` : 'Code diff'}
      className={cn(
        'overflow-hidden rounded border border-border bg-background font-mono text-sm',
        className
      )}
    >
      {/* Top header bar */}
      <div className="flex items-center justify-between gap-2 border-b border-border bg-muted px-3 py-2">
        {headerLeft}
        <ButtonGroup>
          <Button
            variant="outline"
            size="xs"
            aria-pressed={mode === 'split'}
            data-testid="mode-split"
            onClick={() => setMode('split')}
            className={cn(mode === 'split' && 'bg-background shadow-none')}
          >
            <Columns2Icon />
            Split
          </Button>
          <Button
            variant="outline"
            size="xs"
            aria-pressed={mode === 'unified'}
            data-testid="mode-unified"
            onClick={() => setMode('unified')}
            className={cn(mode === 'unified' && 'bg-background shadow-none')}
          >
            <AlignLeftIcon />
            Unified
          </Button>
        </ButtonGroup>
      </div>

      {/* Split column sub-headers */}
      {splitColumnHeaders}

      {/* Diff body */}
      <div className="divide-y divide-border" data-testid="diff-body">
        {parsed.hunks.map(hunk => (
          <HunkBlock
            key={hunk.header}
            hunk={hunk}
            mode={mode}
            showLineNumbers={showLineNumbers}
            expandContext={expandContext}
          />
        ))}
      </div>
    </section>
  );
};

export {CodeDiffViewer};
