// @docker/trident/src/components/code-diff-viewer-utils.ts
import {structuredPatch} from 'diff';

/** @internal */
// Types for structured diff data
export type LineType = 'added' | 'removed' | 'context' | 'hunk-header';

export type DiffLine = {
  type: LineType;
  content: string;
  /** Line number in original file (undefined for added lines) */
  oldLineNumber?: number;
  /** Line number in modified file (undefined for removed lines) */
  newLineNumber?: number;
};

export type DiffHunk = {
  header: string;
  lines: DiffLine[];
};

export type ParsedDiff = {
  hunks: DiffHunk[];
  hasChanges: boolean;
};

/**
 * Parse two text strings into a single flat list of DiffLines using the
 * `diff` package with maximum context (all lines), so the component can
 * apply its own visual collapsing.
 *
 * Using `Number.MAX_SAFE_INTEGER` as the context value produces one
 * giant hunk containing every line of both files, giving the component
 * full control over which context lines are shown or collapsed.
 */
export function parseDiff(original: string, modified: string): ParsedDiff {
  // Request the full file as context so we see every line.
  const patch = structuredPatch(
    'original',
    'modified',
    original,
    modified,
    '',
    '',
    {context: Number.MAX_SAFE_INTEGER}
  );

  if (patch.hunks.length === 0) {
    return {hunks: [], hasChanges: false};
  }

  // Check whether there are any actual changes (not just context lines)
  const hasChanges = patch.hunks.some(h =>
    h.lines.some(l => l[0] === '+' || l[0] === '-')
  );

  if (!hasChanges) {
    return {hunks: [], hasChanges: false};
  }

  const hunks: DiffHunk[] = patch.hunks.map(hunk => {
    const header = `@@ -${hunk.oldStart},${hunk.oldLines} +${hunk.newStart},${hunk.newLines} @@`;
    const lines: DiffLine[] = [];

    let oldLineNum = hunk.oldStart;
    let newLineNum = hunk.newStart;

    for (const rawLine of hunk.lines) {
      // Skip the "No newline at end of file" marker
      if (rawLine.startsWith('\\ ')) continue;

      const prefix = rawLine[0];
      const content = rawLine.slice(1);

      if (prefix === '+') {
        lines.push({type: 'added', content, newLineNumber: newLineNum++});
      } else if (prefix === '-') {
        lines.push({type: 'removed', content, oldLineNumber: oldLineNum++});
      } else {
        // context line (space)
        lines.push({
          type: 'context',
          content,
          oldLineNumber: oldLineNum++,
          newLineNumber: newLineNum++,
        });
      }
    }

    return {header, lines};
  });

  return {hunks, hasChanges: true};
}

export type SplitRow = {
  left: DiffLine | null;
  right: DiffLine | null;
};

/**
 * Convert a flat list of DiffLines (from one hunk) into paired rows
 * for side-by-side split view. Removed lines pair with added lines;
 * context lines pair with themselves.
 */
export function toSplitRows(lines: DiffLine[]): SplitRow[] {
  const rows: SplitRow[] = [];
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];

    if (line.type === 'context') {
      rows.push({left: line, right: line});
      i++;
      continue;
    }

    // Collect a block of removed lines followed by a block of added lines
    const removed: DiffLine[] = [];
    const added: DiffLine[] = [];

    while (i < lines.length && lines[i].type === 'removed') {
      removed.push(lines[i++]);
    }
    while (i < lines.length && lines[i].type === 'added') {
      added.push(lines[i++]);
    }

    const maxLen = Math.max(removed.length, added.length);
    for (let j = 0; j < maxLen; j++) {
      rows.push({
        left: removed[j] ?? null,
        right: added[j] ?? null,
      });
    }
  }

  return rows;
}

/** Format a line number for display, right-aligned in a fixed char width. */
export function formatLineNumber(n: number | undefined): string {
  return n !== undefined ? String(n) : '';
}

export const lineClasses: Record<string, string> = {
  added: 'bg-diff-added-bg',
  removed: 'bg-diff-removed-bg',
  context: 'bg-background',
  'hunk-header': 'bg-diff-hunk-bg',
};
export const gutterCharClasses: Record<string, string> = {
  added: 'text-diff-added-text',
  removed: 'text-diff-removed-text',
  context: 'text-transparent select-none',
};
export const gutterChar: Record<string, string> = {
  added: '+',
  removed: '-',
  context: ' ',
};
