// @docker/trident/src/components/context-switcher.tsx
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
} from '@/components/ui/command';
import {Popover, PopoverContent, PopoverTrigger} from '@/components/ui/popover';
import {cn} from '@/lib/utils';

import {ChevronsUpDown} from 'lucide-react';
import {useState} from 'react';

export type ContextItem = {
  id: string;
  label: string;
  type: 'user' | 'org';
  avatarUrl?: string;
};

type ContextSwitcherProps = {
  items: ContextItem[];
  selected: ContextItem | undefined;
  onSelect: (item: ContextItem) => void;
  placeholder?: string;
  className?: string;
};

/**
 * Organisation / namespace context switcher.
 * Shows a searchable Command popover listing all available contexts.
 */
export const ContextSwitcher = ({
  items,
  selected,
  onSelect,
  placeholder = 'Select context…',
  className,
}: ContextSwitcherProps) => {
  const [open, setOpen] = useState(false);

  return (
    <Popover open={open} onOpenChange={setOpen}>
      <PopoverTrigger asChild>
        <button
          aria-label={selected?.label ?? placeholder}
          aria-expanded={open}
          aria-haspopup="listbox"
          className={cn(
            'flex w-full items-center gap-dds-200 px-dds-300 py-dds-200',
            'rounded text-sm font-label',
            'hover:bg-accent hover:text-accent-foreground',
            'focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none',
            'transition-colors',
            className
          )}
        >
          <span className="flex-1 truncate text-left">
            {selected?.label ?? placeholder}
          </span>
          <ChevronsUpDown className="h-4 w-4 shrink-0 text-muted-foreground" />
        </button>
      </PopoverTrigger>
      <PopoverContent
        aria-label="Select context"
        className="w-[var(--dds-width-context-switcher)] p-0"
        align="start"
      >
        <Command label="Search contexts">
          <CommandInput placeholder="Search…" />
          <CommandList>
            <CommandEmpty>No context found.</CommandEmpty>
            <CommandGroup>
              {items.map(item => (
                <CommandItem
                  key={item.id}
                  value={item.id}
                  onSelect={() => {
                    onSelect(item);
                    setOpen(false);
                  }}
                  className="gap-dds-200"
                  data-checked={selected?.id === item.id}
                >
                  <span className="truncate">{item.label}</span>
                </CommandItem>
              ))}
            </CommandGroup>
          </CommandList>
        </Command>
      </PopoverContent>
    </Popover>
  );
};
