// @docker/trident/src/components/conversation-layout.tsx
import {cn} from '@/lib/utils';

type ConversationLayoutProps = {
  width?: number;
  className?: string;
  children: React.ReactNode;
};

/**
 * Permanent conversation column with three pinned zones: session picker
 * header, scrollable message area, and pinned composer input.
 *
 * The v0/Cursor/Claude Desktop three-zone layout. Fills its parent height.
 * Width defaults to 380px; pass `width` to override.
 *
 * @example
 * <ConversationLayout width={380}>
 *   <ConversationLayout.Header>
 *     <SessionPicker ... />
 *   </ConversationLayout.Header>
 *   <ConversationLayout.Messages>
 *     {messages.map(m => <MessageBubble key={m.id} msg={m} />)}
 *   </ConversationLayout.Messages>
 *   <ConversationLayout.Input>
 *     <Composer onSubmit={handleSubmit} />
 *   </ConversationLayout.Input>
 * </ConversationLayout>
 */
function ConversationLayout({
  width = 380,
  className,
  children,
}: ConversationLayoutProps) {
  return (
    <div
      data-slot="conversation-layout"
      style={{width}}
      className={cn(
        'flex shrink-0 flex-col overflow-hidden border-r border-border bg-card',
        className
      )}
    >
      {children}
    </div>
  );
}

function ConversationLayoutHeader({
  className,
  ...props
}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="conversation-layout-header"
      className={cn(
        'flex h-10 shrink-0 items-center border-b border-border-subtle px-4',
        className
      )}
      {...props}
    />
  );
}

function ConversationLayoutMessages({
  className,
  ...props
}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="conversation-layout-messages"
      className={cn('flex min-h-0 flex-1 flex-col overflow-y-auto', className)}
      {...props}
    />
  );
}

function ConversationLayoutInput({
  className,
  ...props
}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="conversation-layout-input"
      className={cn('shrink-0 border-t border-border-subtle', className)}
      {...props}
    />
  );
}

const ConversationLayoutCompound = Object.assign(ConversationLayout, {
  Header: ConversationLayoutHeader,
  Messages: ConversationLayoutMessages,
  Input: ConversationLayoutInput,
});

export {ConversationLayoutCompound as ConversationLayout};
export type {ConversationLayoutProps};
