// @docker/trident/src/components/data-grid-edit.tsx
'use client';

import {Input} from '@/components/ui/input';
import {TableCell} from '@/components/ui/table';
import {cn} from '@/lib/utils';

import {type Cell, flexRender} from '@tanstack/react-table';
import * as React from 'react';

import {TEXT_ALIGN_CLASS, type ColumnMeta} from './data-grid-types';
import {unknownToStr} from './data-grid-utils';

// ─── EditableCell ─────────────────────────────────────────────────────────────

type EditableCellProps = {
  /** The current cell value, pre-filled into the input. */
  value: string;
  /** Called with the new string value when the user commits (Enter or blur). */
  onCommit: (newValue: string) => void;
  /** Called when the user cancels (Escape). */
  onCancel: () => void;
};

/**
 * Inline edit input for DataGrid editable columns.
 *
 * - Enter → commits, Escape → cancels, blur → commits.
 * - Auto-focuses on mount.
 *
 * @wcag 2.1.1 — keyboard accessible: Enter commits, Escape cancels.
 * @wcag 4.1.2 — aria-label describes the editable field purpose.
 */
const EditableCell = ({value, onCommit, onCancel}: EditableCellProps) => {
  const [draft, setDraft] = React.useState(value);
  const committed = React.useRef(false);

  const commit = React.useCallback(
    (val: string) => {
      if (committed.current) return;
      committed.current = true;
      onCommit(val);
    },
    [onCommit]
  );

  const handleKeyDown = React.useCallback(
    (e: React.KeyboardEvent<HTMLInputElement>) => {
      if (e.key === 'Enter') {
        e.preventDefault();
        commit(draft);
      } else if (e.key === 'Escape') {
        e.preventDefault();
        committed.current = true;
        onCancel();
      }
    },
    [draft, commit, onCancel]
  );

  const inputRef = React.useRef<HTMLInputElement>(null);
  React.useEffect(() => {
    inputRef.current?.focus();
  }, []);

  return (
    <Input
      ref={inputRef}
      size="sm"
      value={draft}
      aria-label="Edit cell value"
      onChange={e => setDraft(e.target.value)}
      onKeyDown={handleKeyDown}
      onBlur={() => commit(draft)}
      onClick={e => e.stopPropagation()}
    />
  );
};

// ─── DataGridCell ─────────────────────────────────────────────────────────────

type EditingCell = {rowId: string; field: string} | null;

type DataGridCellProps<TRow extends object> = {
  cell: Cell<TRow, unknown>;
  cClass: string;
  rowId: string;
  editingCell: EditingCell;
  onStartEdit: (next: EditingCell) => void;
  onCommit: (value: string) => void;
  onCancel: () => void;
};

/**
 * Renders a single DataGrid body cell, handling both static and inline-edit modes.
 * Reads `editable` from `cell.column.columnDef.meta` (set by `toColumnDef`).
 * Extracted here to keep data-grid.tsx under the 450-line limit.
 *
 * @wcag 2.1.1, 2.1.3 — editable cells expose `role="gridcell"`, enter the tab
 *   order (`tabIndex=0`), and activate edit mode on Enter / F2 / Space so the
 *   primary interaction is reachable without a pointing device.
 */
const DataGridCell = <TRow extends object>({
  cell,
  cClass,
  rowId,
  editingCell,
  onStartEdit,
  onCommit,
  onCancel,
}: DataGridCellProps<TRow>) => {
  const cMeta = cell.column.columnDef.meta as ColumnMeta | undefined;
  const cAlign = cMeta?.align ?? 'left';
  const isPinned = cell.column.getIsPinned();
  const isEditable = cMeta?.editable === true;
  const isEditing =
    isEditable &&
    editingCell?.rowId === rowId &&
    editingCell?.field === cell.column.id;
  // Editable cells must be reachable by keyboard (WCAG 2.1.1, 2.1.3). When the
  // cell is in edit mode, the inner Input owns focus and the cell itself is
  // removed from the tab order to avoid a double tabstop.
  const canStartEdit = isEditable && !isEditing;
  const handleKeyDown = canStartEdit
    ? (e: React.KeyboardEvent<HTMLTableCellElement>) => {
        // Enter and F2 match the WAI-ARIA Grid pattern; Space matches native
        // activation semantics for button-like affordances.
        if (e.key === 'Enter' || e.key === 'F2' || e.key === ' ') {
          e.preventDefault();
          e.stopPropagation();
          onStartEdit({rowId, field: cell.column.id});
        }
      }
    : undefined;
  return (
    <TableCell
      key={cell.id}
      role="gridcell"
      tabIndex={canStartEdit ? 0 : -1}
      data-align={cAlign}
      className={cn(
        cClass,
        TEXT_ALIGN_CLASS[cAlign],
        isPinned && 'bg-background',
        canStartEdit &&
          'cursor-text focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none focus-visible:ring-inset'
      )}
      style={{
        position: isPinned ? 'sticky' : undefined,
        left:
          isPinned === 'left' ? `${cell.column.getStart('left')}px` : undefined,
        right:
          isPinned === 'right'
            ? `${cell.column.getAfter('right')}px`
            : undefined,
      }}
      onClick={
        canStartEdit
          ? e => {
              e.stopPropagation();
              onStartEdit({rowId, field: cell.column.id});
            }
          : undefined
      }
      onKeyDown={handleKeyDown}
    >
      {isEditing ? (
        <EditableCell
          value={unknownToStr(cell.getValue())}
          onCommit={onCommit}
          onCancel={onCancel}
        />
      ) : (
        flexRender(cell.column.columnDef.cell, cell.getContext())
      )}
    </TableCell>
  );
};

/** @internal */
export {DataGridCell, EditableCell};
export type {DataGridCellProps, EditableCellProps, EditingCell};
