// @docker/trident/src/components/data-grid-tree.ts
/**
 * Tree-data preprocessing for DataGrid.
 *
 * Converts a flat row array into a display list with indentation metadata,
 * driven by `getTreeDataPath`. The DataGrid renders this list directly
 * instead of relying on TanStack Table's built-in tree features.
 *
 * @module data-grid-tree
 * @internal
 */

// ─── Types ───────────────────────────────────────────────────────────────────

/** Metadata attached to each row in the flattened tree display list. */
export type TreeRowMeta = {
  /** Zero-based depth level for indentation. */
  indent: number;
  /** Whether this row has child rows. */
  hasChildren: boolean;
  /** Whether this row is currently expanded (only meaningful when `hasChildren`). */
  isExpanded: boolean;
  /** Unique path key (segments joined with `/`). Used as the expand/collapse key. */
  pathKey: string;
};

/** A row in the flattened tree display list, pairing the original row with tree metadata. */
export type TreeDisplayRow<TRow> = {
  row: TRow;
  meta: TreeRowMeta;
};

// ─── Internal tree node ──────────────────────────────────────────────────────

type TreeNode<TRow> = {
  pathKey: string;
  segments: string[];
  row: TRow;
  children: TreeNode<TRow>[];
};

// ─── Build tree ──────────────────────────────────────────────────────────────

/**
 * Build a forest (list of root nodes) from flat rows + path accessor.
 *
 * Uses a two-pass algorithm so rows can arrive in any order — parents do not
 * need to precede their children in the input array.
 */
function buildForest<TRow>(
  rows: TRow[],
  getTreeDataPath: (row: TRow) => string[]
): TreeNode<TRow>[] {
  const nodeMap = new Map<string, TreeNode<TRow>>();

  // Pass 1: register every node (no parent assignment yet).
  for (const row of rows) {
    const segments = getTreeDataPath(row);
    const pathKey = segments.join('/');
    nodeMap.set(pathKey, {pathKey, segments, row, children: []});
  }

  // Pass 2: assign children to parents; collect roots.
  const roots: TreeNode<TRow>[] = [];
  for (const row of rows) {
    const segments = getTreeDataPath(row);
    const pathKey = segments.join('/');
    const node = nodeMap.get(pathKey);
    if (!node) continue;
    if (segments.length <= 1) {
      roots.push(node);
    } else {
      const parentKey = segments.slice(0, -1).join('/');
      const parent = nodeMap.get(parentKey);
      if (parent) {
        parent.children.push(node);
      } else {
        // Orphan (no parent row in dataset) — promote to root.
        roots.push(node);
      }
    }
  }

  return roots;
}

// ─── Flatten tree ────────────────────────────────────────────────────────────

/**
 * Flatten the forest into a display list, respecting which paths are expanded.
 * Only children of expanded parents appear in the output.
 */
function flattenForest<TRow>(
  nodes: TreeNode<TRow>[],
  expandedPaths: ReadonlySet<string>,
  indent: number
): TreeDisplayRow<TRow>[] {
  const result: TreeDisplayRow<TRow>[] = [];

  for (const node of nodes) {
    const hasChildren = node.children.length > 0;
    const isExpanded = hasChildren && expandedPaths.has(node.pathKey);

    result.push({
      row: node.row,
      meta: {
        indent,
        hasChildren,
        isExpanded,
        pathKey: node.pathKey,
      },
    });

    if (isExpanded) {
      const children = flattenForest(node.children, expandedPaths, indent + 1);
      result.push(...children);
    }
  }

  return result;
}

// ─── Public API ──────────────────────────────────────────────────────────────

/**
 * Pre-process rows into a flat display list for tree rendering.
 *
 * @param rows          - Original flat row array
 * @param getTreeDataPath - Returns the hierarchy path for each row
 * @param expandedPaths - Set of currently expanded path keys
 * @returns Flattened display list with tree metadata
 */
export function buildTreeDisplayList<TRow>(
  rows: TRow[],
  getTreeDataPath: (row: TRow) => string[],
  expandedPaths: ReadonlySet<string>
): TreeDisplayRow<TRow>[] {
  const forest = buildForest(rows, getTreeDataPath);
  return flattenForest(forest, expandedPaths, 0);
}

/**
 * Toggle a path key in an expanded set, returning a new Set.
 */
export function toggleTreePath(
  expandedPaths: Set<string>,
  pathKey: string
): Set<string> {
  const next = new Set(expandedPaths);
  if (next.has(pathKey)) {
    next.delete(pathKey);
  } else {
    next.add(pathKey);
  }
  return next;
}
