// @docker/trident/src/components/data-grid-types.ts
import type * as React from 'react';

// ─── Sort model ───────────────────────────────────────────────────────────────

/** @internal */
export type SortItem = {field: string; sort: 'asc' | 'desc' | null};
export type SortModel = SortItem[];

// ─── Filter model ─────────────────────────────────────────────────────────────

export type FilterOperator =
  | 'contains'
  | 'equals'
  | 'startsWith'
  | 'endsWith'
  | 'greaterThan'
  | 'lessThan'
  | 'isEmpty'
  | 'isNotEmpty';

export type ColumnFilter = {
  columnId: string;
  operator: FilterOperator;
  /** Stringified filter value. Not needed for isEmpty / isNotEmpty. */
  value?: string;
};

/** An ordered list of active column filters (AND semantics). */
export type FilterModel = ColumnFilter[];

// ─── CSV / API ────────────────────────────────────────────────────────────────

export type CsvExportOptions = {
  /** Downloaded file name without extension. @default "export" */
  fileName?: string;
  /** Field delimiter. @default "," */
  delimiter?: string;
  /** Prepend UTF-8 BOM for Excel compatibility. @default false */
  utf8WithBom?: boolean;
};

export type DataGridApi = {
  exportDataAsCsv: (options?: CsvExportOptions) => void;
};

// ─── Column meta (passed through TanStack ColumnDef.meta) ────────────────────

export type ColumnMeta = {
  align?: 'left' | 'center' | 'right';
  headerAlign?: 'left' | 'center' | 'right';
  disableColumnMenu?: boolean;
  resizable?: boolean;
  pin?: 'left' | 'right';
  /** Whether this column renders a filter button. Set by toColumnDef. */
  filterable?: boolean;
  /** Operators available in the filter popover for this column. */
  filterOperators?: FilterOperator[];
  /** Whether this column supports inline cell editing. Set by toColumnDef. */
  editable?: boolean;
};

// ─── Column definition ────────────────────────────────────────────────────────

/**
 * Column definition for DataGrid. Note: DataGrid uses its own column
 * definition type with `field`/`headerName` instead of TanStack Table's
 * `accessorKey`/`header`. Use `DataTable` for direct TanStack Table column
 * definitions.
 */
export type DataGridColumn<TRow extends object = object> = {
  /** Row field key. */
  field: string;
  /** Column header label. Falls back to `field`. */
  headerName?: string;
  /** Fixed width in px. */
  width?: number;
  /** Minimum width in px. */
  minWidth?: number;
  /** Maximum width in px (resizable columns). */
  maxWidth?: number;
  /** @default true */
  sortable?: boolean;
  /**
   * Opt-in per-column filtering. When `true` a filter icon button appears in
   * the column header.
   * @default false
   */
  filterable?: boolean;
  /**
   * Restrict which operators are shown in the filter popover for this column.
   * Defaults to string operators for text fields, numeric operators for number fields.
   */
  filterOperators?: FilterOperator[];
  renderCell?: (params: {row: TRow; value: unknown}) => React.ReactNode;
  renderHeader?: () => React.ReactNode;
  valueFormatter?: (value: unknown) => string;
  valueGetter?: (row: TRow) => unknown;
  align?: 'left' | 'center' | 'right';
  headerAlign?: 'left' | 'center' | 'right';
  /** @default false */
  hide?: boolean;
  /** Disable sort icon for this column only. */
  disableColumnMenu?: boolean;
  /** Allow column resize. Overrides `disableColumnResize` prop. */
  resizable?: boolean;
  /** Pin this column to an edge. */
  pin?: 'left' | 'right';
  /** Enable inline cell editing for this column. Commit via `processRowUpdate`. */
  editable?: boolean;
};

// ─── Slots ────────────────────────────────────────────────────────────────────

export type DataGridSlots = {
  toolbar?: React.ComponentType;
  noRowsOverlay?: React.ComponentType;
  loadingOverlay?: React.ComponentType;
};

// ─── Slot props ───────────────────────────────────────────────────────────────

export type DataGridToolbarSlotProps = {
  onExportCsv?: () => void;
  exportCsvLabel?: string;
  filterComponent?: React.ReactNode;
};

export type DataGridNoRowsOverlaySlotProps = {
  message?: string;
};

export type DataGridLoadingOverlaySlotProps = {
  message?: string;
};

export type DataGridSlotProps = {
  toolbar?: DataGridToolbarSlotProps;
  noRowsOverlay?: DataGridNoRowsOverlaySlotProps;
  loadingOverlay?: DataGridLoadingOverlaySlotProps;
};

// ─── Props ────────────────────────────────────────────────────────────────────

export type DataGridProps<TRow extends object = object> = {
  rows: TRow[];
  columns: DataGridColumn<TRow>[];
  loading?: boolean;
  pagination?: boolean;
  /** @default 25 */
  pageSize?: number;
  rowsPerPageOptions?: number[];
  rowCount?: number;
  onRowClick?: (params: {row: TRow; id: string}) => void;
  checkboxSelection?: boolean;
  disableRowSelectionOnClick?: boolean;
  /** @default "client" */
  sortingMode?: 'client' | 'server';
  onSortModelChange?: (model: SortModel) => void;
  /**
   * Controlled filter model. When provided, the DataGrid applies client-side
   * filtering before passing rows to TanStack Table.
   */
  filterModel?: FilterModel;
  /** Called whenever the user applies or clears a column filter. */
  onFilterModelChange?: (model: FilterModel) => void;
  /**
   * Disable column filter UI entirely (no filter icons in headers).
   * @default false
   */
  disableColumnFilter?: boolean;
  getRowId?: (row: TRow) => string;
  slots?: DataGridSlots;
  slotProps?: DataGridSlotProps;
  /** @default "default" */
  density?: 'compact' | 'default' | 'spacious';
  /** Disable sort icons on all columns. */
  disableColumnMenu?: boolean;
  /** Disable column resizing globally. */
  disableColumnResize?: boolean;
  /** Disable column drag-and-drop reordering. */
  disableColumnReorder?: boolean;
  /** Pin columns by field name. Merged with column-level `pin` prop. */
  pinnedColumns?: {left?: string[]; right?: string[]};
  /** Fields to group rows by. Groups start collapsed. */
  rowGroupingModel?: string[];
  /** Called when the grouping model changes. */
  onRowGroupingModelChange?: (model: string[]) => void;
  /**
   * Controlled column visibility map. Keys are column `field` values; `true` = visible.
   * Falls back to the static `column.hide` property when not provided.
   */
  columnVisibilityModel?: Record<string, boolean>;
  /** Called whenever the user or code toggles column visibility. */
  onColumnVisibilityModelChange?: (model: Record<string, boolean>) => void;
  /**
   * Called when the user commits an inline cell edit. Return the updated row
   * (or a Promise that resolves to it). Throw or reject to abort the update.
   */
  processRowUpdate?: (newRow: TRow, oldRow: TRow) => TRow | Promise<TRow>;
  /** Called when `processRowUpdate` throws or its Promise rejects. */
  onProcessRowUpdateError?: (error: unknown) => void;
  /** Expose imperative grid API (e.g. `exportDataAsCsv`). */
  apiRef?: React.RefObject<DataGridApi | null>;
  hideFooter?: boolean;
  autoHeight?: boolean;
  className?: string;
  /**
   * Accessible name for the scrollable grid region (`role="region"`).
   * Provide a concise label that identifies the data being shown, e.g.
   * `"Container images"`. Falls back to `"Data grid"` when omitted.
   * @wcag 4.1.2 — landmark regions must have an accessible name
   */
  'aria-label'?: string;
  /**
   * ID of an element whose text content labels this grid region.
   * Use when a visible heading already describes the grid.
   * @wcag 4.1.2 — landmark regions must have an accessible name
   */
  'aria-labelledby'?: string;
  /**
   * Called when row checkbox selection changes. Receives the full array of
   * currently selected row objects (not IDs).
   *
   * Requires `checkboxSelection={true}`.
   *
   * ```tsx
   * <DataGrid
   *   checkboxSelection
   *   onRowSelectionChange={(rows) => setSelected(rows)}
   * />
   * ```
   */
  onRowSelectionChange?: (selectedRows: TRow[]) => void;
  /**
   * Enable hierarchical (tree) row display. Requires `getTreeDataPath`.
   * When true, rows are preprocessed into a parent/child hierarchy and the
   * first data column renders expand/collapse toggles with indentation.
   * The `<table>` element receives `role="treegrid"` for ARIA compliance.
   */
  treeData?: boolean;
  /**
   * Returns the hierarchy path for a row. Each element is a level label.
   * Example: `['Documents', 'Photos', 'vacation.jpg']` places a row three
   * levels deep. Required when `treeData` is `true`.
   */
  getTreeDataPath?: (row: TRow) => string[];
};

// ─── Layout constants ─────────────────────────────────────────────────────────

export const HEADER_CLASS = {
  compact: 'h-dds-800 px-dds-200 py-dds-0',
  default: 'h-dds-1000 px-dds-200',
  spacious: 'h-dds-1200 px-dds-300',
} as const;

export const CELL_CLASS = {
  compact: 'py-dds-050 px-dds-200',
  default: 'py-dds-150 px-dds-200',
  spacious: 'py-dds-300 px-dds-300',
} as const;

export const TEXT_ALIGN_CLASS = {
  left: 'text-left',
  center: 'text-center',
  right: 'text-right',
} as const;
