// @docker/trident/src/components/data-grid-header.tsx
'use client';

import type * as React from 'react';

import {
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from '@/components/ui/table';
import {cn} from '@/lib/utils';

import {
  type ColumnOrderState,
  flexRender,
  type Row,
  type Table,
} from '@tanstack/react-table';
import {
  ChevronDown,
  ChevronRight,
  ChevronsUpDown,
  ChevronUp,
} from 'lucide-react';

import {DataGridFilterButton} from './data-grid-filter';
import {
  type ColumnFilter,
  type ColumnMeta,
  type FilterModel,
  TEXT_ALIGN_CLASS,
} from './data-grid-types';

// ─── DataGridHeader ───────────────────────────────────────────────────────────

type DataGridHeaderProps<TRow extends object> = {
  table: Table<TRow>;
  hClass: string;
  disableColumnMenu: boolean;
  disableColumnResize: boolean;
  disableColumnReorder: boolean;
  disableColumnFilter: boolean;
  dragId: React.MutableRefObject<string | null>;
  setColumnOrder: React.Dispatch<React.SetStateAction<ColumnOrderState>>;
  filterModel: FilterModel;
  onFilterChange: (filter: ColumnFilter | null, columnId: string) => void;
};

function DataGridHeader<TRow extends object>({
  table,
  hClass,
  disableColumnMenu,
  disableColumnResize,
  disableColumnReorder,
  disableColumnFilter,
  dragId,
  setColumnOrder,
  filterModel,
  onFilterChange,
}: DataGridHeaderProps<TRow>) {
  return (
    <TableHeader>
      {table.getHeaderGroups().map(hg => (
        <TableRow key={hg.id}>
          {hg.headers.map(header => {
            const meta = header.column.columnDef.meta as ColumnMeta | undefined;
            const canSort =
              header.column.getCanSort() &&
              !disableColumnMenu &&
              !meta?.disableColumnMenu;
            const sortDir = header.column.getIsSorted();
            const hAlign = meta?.headerAlign ?? 'left';
            const isPinned = header.column.getIsPinned();
            const canResize =
              !disableColumnResize &&
              meta?.resizable !== false &&
              header.column.id !== '__select__';
            const isDraggable =
              !disableColumnReorder && header.column.id !== '__select__';
            const colFilterable =
              !disableColumnFilter &&
              meta?.filterable === true &&
              header.column.id !== '__select__';

            const sortHandler = canSort
              ? header.column.getToggleSortingHandler()
              : undefined;
            const headerLabel = header.isPlaceholder
              ? null
              : flexRender(header.column.columnDef.header, header.getContext());
            const sortIndicator = canSort ? (
              <span className="text-muted-foreground" aria-hidden="true">
                {sortDir === 'asc' ? (
                  <ChevronUp className="size-3.5" />
                ) : sortDir === 'desc' ? (
                  <ChevronDown className="size-3.5" />
                ) : (
                  <ChevronsUpDown className="size-3.5 opacity-40" />
                )}
              </span>
            ) : null;
            return (
              <TableHead
                key={header.id}
                scope="col"
                data-field={header.column.id}
                data-align={hAlign}
                data-sortable={canSort ? 'true' : undefined}
                data-pinned={isPinned || undefined}
                aria-sort={
                  canSort
                    ? sortDir === 'asc'
                      ? 'ascending'
                      : sortDir === 'desc'
                        ? 'descending'
                        : 'none'
                    : undefined
                }
                draggable={isDraggable || undefined}
                className={cn(
                  hClass,
                  TEXT_ALIGN_CLASS[hAlign],
                  isPinned && 'bg-background',
                  'relative'
                )}
                style={{
                  width: header.getSize(),
                  position: isPinned ? 'sticky' : undefined,
                  left:
                    isPinned === 'left'
                      ? `${header.column.getStart('left')}px`
                      : undefined,
                  right:
                    isPinned === 'right'
                      ? `${header.column.getAfter('right')}px`
                      : undefined,
                  zIndex: isPinned ? 1 : undefined,
                }}
                onDragStart={
                  isDraggable
                    ? () => {
                        dragId.current = header.column.id;
                      }
                    : undefined
                }
                onDragOver={isDraggable ? e => e.preventDefault() : undefined}
                onDrop={
                  isDraggable
                    ? () => {
                        const from = dragId.current;
                        const to = header.column.id;
                        if (!from || from === to) return;
                        setColumnOrder(prev => {
                          const next = [...prev];
                          const fi = next.indexOf(from);
                          const ti = next.indexOf(to);
                          if (fi < 0 || ti < 0) return prev;
                          next.splice(fi, 1);
                          next.splice(ti, 0, from);
                          return next;
                        });
                        dragId.current = null;
                      }
                    : undefined
                }
              >
                <span className="flex items-center gap-1">
                  {canSort ? (
                    <button
                      type="button"
                      data-slot="column-sort"
                      onClick={sortHandler}
                      className={cn(
                        'flex flex-1 cursor-pointer items-center gap-1 rounded border-0 bg-transparent p-0 [color:inherit] select-none [font:inherit]',
                        TEXT_ALIGN_CLASS[hAlign],
                        'focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none'
                      )}
                    >
                      {headerLabel}
                      {sortIndicator}
                    </button>
                  ) : (
                    <>
                      {headerLabel}
                      {sortIndicator}
                    </>
                  )}
                  {colFilterable && (
                    <DataGridFilterButton
                      columnId={header.column.id}
                      filterModel={filterModel}
                      onFilterChange={f => onFilterChange(f, header.column.id)}
                      operators={meta?.filterOperators}
                    />
                  )}
                </span>
                {canResize && (
                  <div
                    data-slot="column-resize-handle"
                    onMouseDown={header.getResizeHandler()}
                    onTouchStart={header.getResizeHandler()}
                    className={cn(
                      'absolute top-0 right-0 h-full w-1 cursor-col-resize touch-none select-none',
                      'bg-border opacity-0 hover:bg-primary hover:opacity-100',
                      header.column.getIsResizing() && 'bg-primary opacity-100'
                    )}
                  />
                )}
              </TableHead>
            );
          })}
        </TableRow>
      ))}
    </TableHeader>
  );
}

// ─── DataGridGroupRow ─────────────────────────────────────────────────────────

type DataGridGroupRowProps<TRow extends object> = {
  row: Row<TRow>;
  cClass: string;
};

function DataGridGroupRow<TRow extends object>({
  row,
  cClass,
}: DataGridGroupRowProps<TRow>) {
  return (
    <TableRow
      data-slot="data-grid-group-row"
      className="bg-hover-overlay font-label"
    >
      {row.getVisibleCells().map(cell => {
        if (cell.getIsGrouped()) {
          return (
            <TableCell key={cell.id} className={cn(cClass, 'font-label')}>
              <span className="flex items-center gap-1">
                <button
                  type="button"
                  data-slot="group-expand"
                  onClick={() => row.toggleExpanded()}
                  aria-label={
                    row.getIsExpanded() ? 'Collapse group' : 'Expand group'
                  }
                  className="flex items-center text-muted-foreground"
                >
                  {row.getIsExpanded() ? (
                    <ChevronDown className="size-4" />
                  ) : (
                    <ChevronRight className="size-4" />
                  )}
                </button>
                {String(cell.getValue())}
                <span className="text-xs text-muted-foreground">
                  ({row.subRows.length})
                </span>
              </span>
            </TableCell>
          );
        }
        if (cell.getIsAggregated()) {
          return (
            <TableCell key={cell.id} className={cClass}>
              {flexRender(
                cell.column.columnDef.aggregatedCell ??
                  cell.column.columnDef.cell,
                cell.getContext()
              )}
            </TableCell>
          );
        }
        return <TableCell key={cell.id} />;
      })}
    </TableRow>
  );
}

/** @internal */
export {DataGridGroupRow, DataGridHeader};
