// @docker/trident/src/components/data-grid.tsx
'use client';

import {Checkbox} from '@/components/ui/checkbox';
import {Table, TableBody, TableCell, TableRow} from '@/components/ui/table';
import {cn} from '@/lib/utils';

import {
  type ColumnDef,
  type ColumnOrderState,
  type ColumnSizingState,
  type ExpandedState,
  type GroupingState,
  getCoreRowModel,
  getExpandedRowModel,
  getFilteredRowModel,
  getGroupedRowModel,
  getPaginationRowModel,
  getSortedRowModel,
  type PaginationState,
  type RowSelectionState,
  type SortingState,
  useReactTable,
  type VisibilityState,
} from '@tanstack/react-table';
import * as React from 'react';

import {DataGridCell, type EditingCell} from './data-grid-edit';
import {applyFilters} from './data-grid-filter';
import {DataGridFooter} from './data-grid-footer';
import {DataGridGroupRow, DataGridHeader} from './data-grid-header';
import {
  buildTreeDisplayList,
  toggleTreePath,
  type TreeRowMeta,
} from './data-grid-tree';
import {DataGridTreeRow} from './data-grid-tree-row';
import {
  CELL_CLASS,
  type ColumnFilter,
  type DataGridProps,
  type FilterModel,
  HEADER_CLASS,
} from './data-grid-types';
import {
  buildColumnPinning,
  exportToCsv,
  toColumnDef,
  unknownToStr,
} from './data-grid-utils';

/**
 * Full-featured data grid built on TanStack Table v8 with DDS styling.
 *
 * @wcag 1.3.1 — semantic `<table>` with `<th>` column headers
 * @wcag 2.1.1 — sort, checkboxes, pagination, expand buttons are keyboard-accessible
 * @wcag 4.1.2 — sort buttons have aria-label; checkboxes have aria-label
 */
function DataGrid<TRow extends object = object>({
  rows,
  columns,
  loading = false,
  pagination = false,
  pageSize = 25,
  rowsPerPageOptions,
  rowCount,
  onRowClick,
  checkboxSelection = false,
  disableRowSelectionOnClick = false,
  sortingMode = 'client',
  onSortModelChange,
  getRowId,
  slots,
  slotProps,
  density = 'default',
  disableColumnMenu = false,
  disableColumnResize = false,
  disableColumnReorder = false,
  pinnedColumns,
  rowGroupingModel,
  onRowGroupingModelChange,
  columnVisibilityModel,
  onColumnVisibilityModelChange,
  processRowUpdate,
  onProcessRowUpdateError,
  apiRef,
  hideFooter = false,
  autoHeight = false,
  className,
  filterModel: filterModelProp,
  onFilterModelChange,
  disableColumnFilter = false,
  treeData = false,
  getTreeDataPath,
  'aria-label': ariaLabel,
  'aria-labelledby': ariaLabelledby,
  onRowSelectionChange,
}: DataGridProps<TRow>) {
  const [editingCell, setEditingCell] = React.useState<EditingCell>(null);
  const [sorting, setSorting] = React.useState<SortingState>([]);
  const [paginationState, setPaginationState] = React.useState<PaginationState>(
    {pageIndex: 0, pageSize}
  );
  const [rowSelection, setRowSelection] = React.useState<RowSelectionState>({});
  const [columnSizing, setColumnSizing] = React.useState<ColumnSizingState>({});
  const [grouping, setGrouping] = React.useState<GroupingState>(
    rowGroupingModel ?? []
  );
  const [expanded, setExpanded] = React.useState<ExpandedState>({});
  const [treeExpandedPaths, setTreeExpandedPaths] = React.useState<Set<string>>(
    () => new Set<string>()
  );
  const [internalFilterModel, setInternalFilterModel] =
    React.useState<FilterModel>([]);
  const filterModel: FilterModel = filterModelProp ?? internalFilterModel;
  const handleFilterChange = React.useCallback(
    (filter: ColumnFilter | null, columnId: string) => {
      const next: FilterModel = filter
        ? [...filterModel.filter(f => f.columnId !== columnId), filter]
        : filterModel.filter(f => f.columnId !== columnId);
      if (filterModelProp === undefined) setInternalFilterModel(next);
      onFilterModelChange?.(next);
    },
    [filterModel, filterModelProp, onFilterModelChange]
  );
  const [columnVisibility, setColumnVisibility] =
    React.useState<VisibilityState>(() => {
      if (columnVisibilityModel !== undefined) return columnVisibilityModel;
      return Object.fromEntries(
        columns.filter(c => c.hide).map(c => [c.field, false])
      );
    });
  React.useEffect(() => {
    setPaginationState(prev => ({...prev, pageSize}));
  }, [pageSize]);
  React.useEffect(() => {
    if (rowGroupingModel !== undefined) setGrouping(rowGroupingModel);
  }, [rowGroupingModel]);
  const [columnOrder, setColumnOrder] = React.useState<ColumnOrderState>(() => {
    const base = columns.map(c => c.field);
    return checkboxSelection ? ['__select__', ...base] : base;
  });

  const columnPinning = React.useMemo(
    () => buildColumnPinning(columns, pinnedColumns),
    [columns, pinnedColumns]
  );
  const dragId = React.useRef<string | null>(null);
  const columnDefs = React.useMemo<ColumnDef<TRow>[]>(() => {
    const selectCol: ColumnDef<TRow> = {
      id: '__select__',
      header: ({table}) => (
        <Checkbox
          checked={
            table.getIsAllRowsSelected()
              ? true
              : table.getIsSomeRowsSelected()
                ? 'indeterminate'
                : false
          }
          onCheckedChange={v => table.toggleAllRowsSelected(v === true)}
          aria-label="Select all rows"
        />
      ),
      cell: ({row}) => (
        <Checkbox
          checked={row.getIsSelected()}
          onCheckedChange={v => row.toggleSelected(v === true)}
          aria-label={`Select row ${row.id}`}
          onClick={e => e.stopPropagation()}
        />
      ),
      enableSorting: false,
      size: 40,
    };
    const dataCols = columns.map(toColumnDef);
    return checkboxSelection ? [selectCol, ...dataCols] : dataCols;
  }, [columns, checkboxSelection]);
  const rowIdFn = React.useCallback(
    (row: TRow, index: number): string => {
      if (getRowId) return getRowId(row);
      const id = unknownToStr(Reflect.get(row, 'id'));
      return id !== '' ? id : String(index);
    },
    [getRowId]
  );
  const filteredRows = React.useMemo(
    () => applyFilters(rows, filterModel, columns),
    [rows, filterModel, columns]
  );
  const treeDisplayList = React.useMemo(
    () =>
      treeData && getTreeDataPath
        ? buildTreeDisplayList(filteredRows, getTreeDataPath, treeExpandedPaths)
        : null,
    [treeData, getTreeDataPath, filteredRows, treeExpandedPaths]
  );
  const tableRows = treeDisplayList
    ? treeDisplayList.map(item => item.row)
    : filteredRows;
  const table = useReactTable<TRow>({
    data: tableRows,
    columns: columnDefs,
    state: {
      sorting,
      pagination: paginationState,
      rowSelection,
      columnSizing,
      columnOrder,
      columnPinning,
      columnVisibility: columnVisibilityModel ?? columnVisibility,
      grouping,
      expanded,
    },
    columnResizeMode: 'onChange',
    onSortingChange: updater => {
      const next = typeof updater === 'function' ? updater(sorting) : updater;
      setSorting(next);
      onSortModelChange?.(
        next.map(s => ({field: s.id, sort: s.desc ? 'desc' : 'asc'}))
      );
    },
    onColumnVisibilityChange: updater => {
      const current = columnVisibilityModel ?? columnVisibility;
      const next = typeof updater === 'function' ? updater(current) : updater;
      setColumnVisibility(next);
      onColumnVisibilityModelChange?.(next);
    },
    onPaginationChange: setPaginationState,
    onRowSelectionChange: setRowSelection,
    onColumnSizingChange: setColumnSizing,
    onColumnOrderChange: setColumnOrder,
    onGroupingChange: updater => {
      const next = typeof updater === 'function' ? updater(grouping) : updater;
      setGrouping(next);
      onRowGroupingModelChange?.(next);
    },
    onExpandedChange: setExpanded,
    getCoreRowModel: getCoreRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    getGroupedRowModel: grouping.length ? getGroupedRowModel() : undefined,
    getSortedRowModel: getSortedRowModel(),
    getExpandedRowModel: grouping.length ? getExpandedRowModel() : undefined,
    ...(pagination && {getPaginationRowModel: getPaginationRowModel()}),
    getRowId: rowIdFn,
    enableRowSelection: checkboxSelection,
    manualSorting: sortingMode === 'server',
    groupedColumnMode: false,
  });
  // Fire onRowSelectionChange whenever the selection state changes (skip initial render).
  const isFirstRender = React.useRef(true);
  React.useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
      return;
    }
    if (!onRowSelectionChange) return;
    const selected = rows.filter((row, i) =>
      Boolean(rowSelection[rowIdFn(row, i)])
    );
    onRowSelectionChange(selected);
  }, [onRowSelectionChange, rowIdFn, rowSelection, rows]);

  // WCAG 4.1.2 — role="region" must have an accessible name.
  React.useEffect(() => {
    if (!ariaLabel && !ariaLabelledby) {
      console.warn(
        '[DataGrid] Missing accessible name. ' +
          'Add aria-label="Description of this grid" or aria-labelledby="heading-element-id" ' +
          'to comply with WCAG 4.1.2 (role="region" must have an accessible name).'
      );
    }
  }, [ariaLabel, ariaLabelledby]);

  React.useImperativeHandle(
    apiRef,
    () => ({
      exportDataAsCsv: opts => exportToCsv(columns, rows, opts ?? {}),
    }),
    [columns, rows]
  );
  const commitEdit = React.useCallback(
    async (rowOriginal: TRow, field: string, value: string) => {
      setEditingCell(null);
      if (processRowUpdate) {
        try {
          await processRowUpdate(
            {...rowOriginal, [field]: value} as TRow,
            rowOriginal
          );
        } catch (err) {
          onProcessRowUpdateError?.(err);
        }
      }
    },
    [processRowUpdate, onProcessRowUpdateError]
  );
  // Key by object identity so pagination/sorting don't break the lookup.
  const treeMetaMap = React.useMemo(() => {
    if (!treeDisplayList) return null;
    const map = new Map<TRow, TreeRowMeta>();
    for (const item of treeDisplayList) map.set(item.row, item.meta);
    return map;
  }, [treeDisplayList]);
  const handleTreeToggle = React.useCallback((pathKey: string) => {
    setTreeExpandedPaths(prev => toggleTreePath(prev, pathKey));
  }, []);
  const visibleRows = table.getRowModel().rows;
  const hClass = HEADER_CLASS[density];
  const cClass = CELL_CLASS[density];
  const Toolbar = slots?.toolbar as React.ElementType | undefined;
  const NoRowsOverlay = slots?.noRowsOverlay as React.ElementType | undefined;
  const LoadingOverlay = slots?.loadingOverlay as React.ElementType | undefined;
  const {pageIndex, pageSize: currentPageSize} = table.getState().pagination;
  const totalRows = rowCount ?? rows.length;
  const firstRow = totalRows === 0 ? 0 : pageIndex * currentPageSize + 1;
  const lastRow = Math.min((pageIndex + 1) * currentPageSize, totalRows);

  return (
    <div
      data-slot="data-grid"
      data-density={density}
      role="region"
      aria-label={ariaLabelledby ? undefined : (ariaLabel ?? 'Data grid')}
      aria-labelledby={ariaLabelledby}
      aria-busy={loading || undefined}
      className={cn(
        'relative flex flex-col overflow-hidden rounded border border-border bg-background text-foreground',
        autoHeight ? 'h-auto' : 'h-full',
        className
      )}
    >
      {Toolbar && <Toolbar {...(slotProps?.toolbar ?? {})} />}

      {loading && (
        <div
          data-slot="data-grid-loading"
          className="absolute inset-x-0 top-0 z-raised h-dds-050 overflow-hidden bg-muted"
        >
          {LoadingOverlay ? (
            <LoadingOverlay {...(slotProps?.loadingOverlay ?? {})} />
          ) : (
            <div className="h-full w-1/3 animate-pulse bg-primary" />
          )}
        </div>
      )}

      <div className="flex-1 overflow-auto">
        <Table {...(treeData ? {role: 'treegrid'} : {})}>
          <DataGridHeader
            table={table}
            hClass={hClass}
            disableColumnMenu={disableColumnMenu}
            disableColumnResize={disableColumnResize}
            disableColumnReorder={disableColumnReorder}
            disableColumnFilter={disableColumnFilter}
            dragId={dragId}
            setColumnOrder={setColumnOrder}
            filterModel={filterModel}
            onFilterChange={handleFilterChange}
          />

          <TableBody>
            {visibleRows.length > 0
              ? visibleRows.map(row => {
                  if (row.getIsGrouped()) {
                    return (
                      <DataGridGroupRow
                        key={row.id}
                        row={row}
                        cClass={cClass}
                      />
                    );
                  }
                  const treeMeta = treeMetaMap?.get(row.original);
                  if (treeMeta) {
                    return (
                      <DataGridTreeRow
                        key={row.id}
                        row={row}
                        treeMeta={treeMeta}
                        cClass={cClass}
                        editingCell={editingCell}
                        onStartEdit={setEditingCell}
                        onCommitEdit={commitEdit}
                        onCancelEdit={() => setEditingCell(null)}
                        onToggle={handleTreeToggle}
                        onRowClick={onRowClick}
                        disableRowSelectionOnClick={disableRowSelectionOnClick}
                      />
                    );
                  }
                  const rowInteractive = Boolean(onRowClick);
                  const handleRowActivate = onRowClick
                    ? () => onRowClick({row: row.original, id: row.id})
                    : undefined;
                  return (
                    <TableRow
                      key={row.id}
                      data-slot="data-grid-row"
                      data-state={row.getIsSelected() ? 'selected' : undefined}
                      tabIndex={rowInteractive ? 0 : undefined}
                      aria-label={rowInteractive ? `Row ${row.id}` : undefined}
                      className={cn(
                        rowInteractive &&
                          !disableRowSelectionOnClick &&
                          'cursor-pointer',
                        rowInteractive &&
                          'focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none'
                      )}
                      onClick={handleRowActivate}
                      onKeyDown={
                        rowInteractive && handleRowActivate
                          ? e => {
                              if (e.target !== e.currentTarget) return;
                              if (e.key === 'Enter') {
                                e.preventDefault();
                                handleRowActivate();
                              } else if (e.key === ' ') {
                                e.preventDefault();
                                handleRowActivate();
                              }
                            }
                          : undefined
                      }
                    >
                      {row.getVisibleCells().map(cell => (
                        <DataGridCell
                          key={cell.id}
                          cell={cell}
                          cClass={cClass}
                          rowId={row.id}
                          editingCell={editingCell}
                          onStartEdit={setEditingCell}
                          onCommit={v =>
                            commitEdit(row.original, cell.column.id, v)
                          }
                          onCancel={() => setEditingCell(null)}
                        />
                      ))}
                    </TableRow>
                  );
                })
              : !loading && (
                  <TableRow>
                    <TableCell
                      colSpan={columnDefs.length || 1}
                      className="h-dds-2400 text-center"
                    >
                      <div data-slot="data-grid-no-rows">
                        {NoRowsOverlay ? (
                          <NoRowsOverlay
                            {...(slotProps?.noRowsOverlay ?? {})}
                          />
                        ) : (
                          <span className="text-sm text-muted-foreground">
                            No rows
                          </span>
                        )}
                      </div>
                    </TableCell>
                  </TableRow>
                )}
          </TableBody>
        </Table>
      </div>

      {!hideFooter && (
        <DataGridFooter
          table={table}
          pagination={pagination}
          rowsPerPageOptions={rowsPerPageOptions}
          totalRows={totalRows}
          firstRow={firstRow}
          lastRow={lastRow}
          currentPageSize={currentPageSize}
          setPaginationState={setPaginationState}
        />
      )}
    </div>
  );
}

export {DataGrid};
