// @docker/trident/src/components/data-grid-filter.tsx
'use client';

import type {
  ColumnFilter,
  DataGridColumn,
  FilterModel,
  FilterOperator,
} from './data-grid-types';

import {Badge} from '@/components/ui/badge';
import {Button} from '@/components/ui/button';
import {Input} from '@/components/ui/input';
import {Popover, PopoverContent, PopoverTrigger} from '@/components/ui/popover';
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from '@/components/ui/select';
import {cn} from '@/lib/utils';

import {FilterIcon} from 'lucide-react';
import * as React from 'react';

// ─── Operator metadata ─────────────────────────────────────────────────────────

const OPERATOR_LABELS: Record<FilterOperator, string> = {
  contains: 'Contains',
  equals: 'Equals',
  startsWith: 'Starts with',
  endsWith: 'Ends with',
  greaterThan: 'Greater than',
  lessThan: 'Less than',
  isEmpty: 'Is empty',
  isNotEmpty: 'Is not empty',
};

const STRING_OPERATORS: FilterOperator[] = [
  'contains',
  'equals',
  'startsWith',
  'endsWith',
  'isEmpty',
  'isNotEmpty',
];

const NUMBER_OPERATORS: FilterOperator[] = [
  'equals',
  'greaterThan',
  'lessThan',
  'isEmpty',
  'isNotEmpty',
];

/** Operators that do not require a value input */
const NO_VALUE_OPERATORS = new Set<FilterOperator>(['isEmpty', 'isNotEmpty']);

/** Type predicate — confirms a raw string is a valid FilterOperator */
const ALL_OPERATORS = new Set<string>([
  'contains',
  'equals',
  'startsWith',
  'endsWith',
  'greaterThan',
  'lessThan',
  'isEmpty',
  'isNotEmpty',
]);

function isFilterOperator(v: string): v is FilterOperator {
  return ALL_OPERATORS.has(v);
}

// ─── applyFilters ─────────────────────────────────────────────────────────────

/**
 * Applies a FilterModel to an array of rows client-side.
 * Uses `accessorKey` / `field` to resolve the cell value per column.
 */
function applyFilters<T extends object>(
  rows: T[],
  filters: FilterModel,
  columns: DataGridColumn<T>[]
): T[] {
  if (filters.length === 0) return rows;

  return rows.filter(row => {
    for (const filter of filters) {
      const col = columns.find(c => c.field === filter.columnId);
      if (!col) continue;

      const raw = col.valueGetter
        ? col.valueGetter(row)
        : Reflect.get(row, filter.columnId);

      if (!evaluateFilter(raw, filter)) return false;
    }
    return true;
  });
}

function toStr(val: unknown): string {
  if (val == null) return '';
  if (typeof val === 'object') return JSON.stringify(val);
  if (
    typeof val === 'string' ||
    typeof val === 'number' ||
    typeof val === 'boolean'
  ) {
    return String(val);
  }
  return JSON.stringify(val);
}

function evaluateFilter(raw: unknown, filter: ColumnFilter): boolean {
  const {operator, value} = filter;

  if (operator === 'isEmpty') {
    return raw == null || toStr(raw).trim() === '';
  }
  if (operator === 'isNotEmpty') {
    return raw != null && toStr(raw).trim() !== '';
  }

  // All remaining operators need a value
  if (value === undefined || value === '') return true;

  const cellStr = toStr(raw ?? '').toLowerCase();
  const filterStr = value.toLowerCase();

  switch (operator) {
    case 'contains':
      return cellStr.includes(filterStr);
    case 'equals':
      return cellStr === filterStr;
    case 'startsWith':
      return cellStr.startsWith(filterStr);
    case 'endsWith':
      return cellStr.endsWith(filterStr);
    case 'greaterThan': {
      const cellNum = Number(raw);
      const filterNum = Number(value);
      return (
        !Number.isNaN(cellNum) &&
        !Number.isNaN(filterNum) &&
        cellNum > filterNum
      );
    }
    case 'lessThan': {
      const cellNum = Number(raw);
      const filterNum = Number(value);
      return (
        !Number.isNaN(cellNum) &&
        !Number.isNaN(filterNum) &&
        cellNum < filterNum
      );
    }
    default:
      return true;
  }
}

// ─── DataGridFilterButton ─────────────────────────────────────────────────────

type DataGridFilterButtonProps = {
  columnId: string;
  filterModel: FilterModel;
  onFilterChange: (filter: ColumnFilter | null) => void;
  operators?: FilterOperator[];
};

const DataGridFilterButton = ({
  columnId,
  filterModel,
  onFilterChange,
  operators = STRING_OPERATORS,
}: DataGridFilterButtonProps) => {
  const activeFilter = filterModel.find(f => f.columnId === columnId);

  const [open, setOpen] = React.useState(false);
  const [operator, setOperator] = React.useState<FilterOperator>(
    activeFilter?.operator ?? operators[0] ?? 'contains'
  );
  const [value, setValue] = React.useState(activeFilter?.value ?? '');

  // Sync local state when the active filter changes externally
  React.useEffect(() => {
    if (activeFilter) {
      setOperator(activeFilter.operator);
      setValue(activeFilter.value ?? '');
    } else {
      setOperator(operators[0] ?? 'contains');
      setValue('');
    }
  }, [activeFilter, operators]);

  const isActive = activeFilter !== undefined;
  const needsValue = !NO_VALUE_OPERATORS.has(operator);

  const handleApply = () => {
    onFilterChange({
      columnId,
      operator,
      value: needsValue ? value : undefined,
    });
    setOpen(false);
  };

  const handleClear = () => {
    onFilterChange(null);
    setValue('');
    setOperator(operators[0] ?? 'contains');
    setOpen(false);
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') handleApply();
  };

  return (
    <Popover open={open} onOpenChange={setOpen}>
      <PopoverTrigger asChild>
        <button
          type="button"
          data-slot="data-grid-filter-button"
          aria-label={`Filter ${columnId} column`}
          aria-pressed={isActive}
          onClick={e => e.stopPropagation()}
          className={cn(
            'relative inline-flex items-center justify-center rounded p-0.5 transition-colors',
            'text-muted-foreground hover:bg-muted hover:text-foreground',
            'focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none',
            isActive && 'text-primary'
          )}
        >
          <FilterIcon
            className={cn(
              'size-3.5',
              isActive && 'fill-primary stroke-primary'
            )}
            aria-hidden="true"
          />
          {isActive && (
            <Badge
              variant="default"
              size="sm"
              className="absolute -top-1.5 -right-1.5 size-3.5 min-w-0 px-0 text-2xs leading-none"
              aria-hidden="true"
            >
              1
            </Badge>
          )}
        </button>
      </PopoverTrigger>

      <PopoverContent
        align="start"
        sideOffset={6}
        className="w-64"
        aria-label="Column filter"
        onClick={e => e.stopPropagation()}
      >
        <div className="flex flex-col gap-dds-200">
          <p className="text-xs font-label tracking-wide text-muted-foreground uppercase">
            Filter
          </p>

          {/* Operator selector */}
          <Select
            value={operator}
            onValueChange={v => {
              if (isFilterOperator(v)) setOperator(v);
            }}
          >
            <SelectTrigger
              size="sm"
              className="w-full"
              aria-label="Filter operator"
            >
              <SelectValue />
            </SelectTrigger>
            <SelectContent position="popper" sideOffset={4}>
              {operators.map(op => (
                <SelectItem key={op} value={op}>
                  {OPERATOR_LABELS[op]}
                </SelectItem>
              ))}
            </SelectContent>
          </Select>

          {/* Value input — hidden for isEmpty / isNotEmpty */}
          {needsValue && (
            <Input
              size="sm"
              placeholder="Filter value…"
              value={value}
              onChange={e => setValue(e.target.value)}
              onKeyDown={handleKeyDown}
              aria-label="Filter value"
            />
          )}

          {/* Actions */}
          <div className="flex items-center justify-end gap-dds-100">
            <Button
              variant="ghost"
              size="sm"
              onClick={handleClear}
              disabled={!isActive}
            >
              Clear
            </Button>
            <Button size="sm" onClick={handleApply}>
              Apply
            </Button>
          </div>
        </div>
      </PopoverContent>
    </Popover>
  );
};

/** @internal */
export {
  DataGridFilterButton,
  applyFilters,
  STRING_OPERATORS,
  NUMBER_OPERATORS,
  OPERATOR_LABELS,
};
export type {DataGridFilterButtonProps};
