// @docker/trident/src/components/data-grid-footer.tsx
'use client';

import type {PaginationState, Table} from '@tanstack/react-table';
import type * as React from 'react';

import {Button} from '@/components/ui/button';

import {ChevronLeft, ChevronRight} from 'lucide-react';

// ─── DataGridFooter ───────────────────────────────────────────────────────────

type DataGridFooterProps<TRow extends object> = {
  table: Table<TRow>;
  pagination: boolean;
  rowsPerPageOptions?: number[];
  totalRows: number;
  firstRow: number;
  lastRow: number;
  currentPageSize: number;
  setPaginationState: React.Dispatch<React.SetStateAction<PaginationState>>;
};

function DataGridFooter<TRow extends object>({
  table,
  pagination,
  rowsPerPageOptions,
  totalRows,
  firstRow,
  lastRow,
  currentPageSize,
  setPaginationState,
}: DataGridFooterProps<TRow>) {
  return (
    <nav
      data-slot="data-grid-footer"
      aria-label="Table pagination"
      className="flex items-center justify-between border-t border-border px-dds-400 py-dds-200 text-sm text-muted-foreground"
    >
      <div className="flex items-center gap-dds-300">
        <span aria-live="polite" aria-atomic="true">
          {totalRows === 0
            ? 'No rows'
            : pagination
              ? `${firstRow}–${lastRow} of ${totalRows}`
              : `${totalRows} row${totalRows !== 1 ? 's' : ''}`}
        </span>

        {pagination && rowsPerPageOptions && rowsPerPageOptions.length > 0 && (
          <div className="flex items-center gap-dds-150">
            <label
              htmlFor="data-grid-page-size"
              className="text-xs text-muted-foreground"
            >
              Rows per page:
            </label>
            <select
              id="data-grid-page-size"
              value={currentPageSize}
              onChange={e =>
                setPaginationState({
                  pageIndex: 0,
                  pageSize: Number(e.target.value),
                })
              }
              className="h-7 rounded border border-input bg-transparent px-dds-100 text-sm outline-none focus:ring-1 focus:ring-ring"
            >
              {rowsPerPageOptions.map(s => (
                <option key={s} value={s}>
                  {s}
                </option>
              ))}
            </select>
          </div>
        )}
      </div>

      {pagination && (
        <div className="flex items-center gap-dds-100">
          <Button
            variant="ghost"
            size="sm"
            aria-label="Previous page"
            disabled={!table.getCanPreviousPage()}
            onClick={() => table.previousPage()}
          >
            <ChevronLeft aria-hidden="true" className="size-4" />
            <span className="sr-only">Prev</span>
          </Button>
          <Button
            variant="ghost"
            size="sm"
            aria-label="Next page"
            disabled={!table.getCanNextPage()}
            onClick={() => table.nextPage()}
          >
            <ChevronRight aria-hidden="true" className="size-4" />
            <span className="sr-only">Next</span>
          </Button>
        </div>
      )}
    </nav>
  );
}

/** @internal */
export {DataGridFooter};
