// @docker/trident/src/components/data-grid-tree-row.tsx
'use client';

import type {EditingCell} from './data-grid-edit';
import type {TreeRowMeta} from './data-grid-tree';

import {TableCell, TableRow} from '@/components/ui/table';
import {cn} from '@/lib/utils';

import {type Cell, type Row, flexRender} from '@tanstack/react-table';
import {ChevronDown, ChevronRight} from 'lucide-react';
import * as React from 'react';

import {EditableCell} from './data-grid-edit';
import {TEXT_ALIGN_CLASS, type ColumnMeta} from './data-grid-types';
import {unknownToStr} from './data-grid-utils';

// ─── Constants ───────────────────────────────────────────────────────────────

/** Pixels of indentation per tree level. */
const INDENT_PX = 20;

// ─── Tree cell renderer ─────────────────────────────────────────────────────

type TreeCellProps<TRow extends object> = {
  cell: Cell<TRow, unknown>;
  cClass: string;
  isFirstDataCell: boolean;
  treeMeta: TreeRowMeta;
  rowId: string;
  editingCell: EditingCell;
  onStartEdit: (next: EditingCell) => void;
  onCommit: (value: string) => void;
  onCancel: () => void;
  onToggle: (pathKey: string) => void;
};

const TreeCell = <TRow extends object>({
  cell,
  cClass,
  isFirstDataCell,
  treeMeta,
  rowId,
  editingCell,
  onStartEdit,
  onCommit,
  onCancel,
  onToggle,
}: TreeCellProps<TRow>) => {
  const cMeta = cell.column.columnDef.meta as ColumnMeta | undefined;
  const cAlign = cMeta?.align ?? 'left';
  const isPinned = cell.column.getIsPinned();
  const isEditable = cMeta?.editable === true;
  const isEditing =
    isEditable &&
    editingCell?.rowId === rowId &&
    editingCell?.field === cell.column.id;

  const handleKeyDown = React.useCallback(
    (e: React.KeyboardEvent<HTMLButtonElement>) => {
      if (e.key === 'Enter' || e.key === ' ') {
        e.preventDefault();
        onToggle(treeMeta.pathKey);
      }
    },
    [onToggle, treeMeta.pathKey]
  );

  const cellContent = isEditing ? (
    <EditableCell
      value={unknownToStr(cell.getValue())}
      onCommit={onCommit}
      onCancel={onCancel}
    />
  ) : (
    flexRender(cell.column.columnDef.cell, cell.getContext())
  );

  return (
    <TableCell
      data-align={cAlign}
      className={cn(
        cClass,
        TEXT_ALIGN_CLASS[cAlign],
        isPinned && 'bg-background',
        isEditable && !isEditing && 'cursor-text'
      )}
      style={{
        position: isPinned ? 'sticky' : undefined,
        left:
          isPinned === 'left' ? `${cell.column.getStart('left')}px` : undefined,
        right:
          isPinned === 'right'
            ? `${cell.column.getAfter('right')}px`
            : undefined,
      }}
      onClick={
        isEditable && !isEditing
          ? e => {
              e.stopPropagation();
              onStartEdit({rowId, field: cell.column.id});
            }
          : undefined
      }
    >
      {isFirstDataCell ? (
        <span
          className="flex items-center"
          style={{paddingLeft: `${treeMeta.indent * INDENT_PX}px`}}
        >
          {treeMeta.hasChildren ? (
            <button
              type="button"
              data-slot="tree-expand"
              aria-expanded={treeMeta.isExpanded}
              aria-label={treeMeta.isExpanded ? 'Collapse row' : 'Expand row'}
              onClick={e => {
                e.stopPropagation();
                onToggle(treeMeta.pathKey);
              }}
              onKeyDown={handleKeyDown}
              className="mr-1 flex shrink-0 items-center rounded text-muted-foreground hover:text-foreground focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none"
            >
              {treeMeta.isExpanded ? (
                <ChevronDown className="size-4" />
              ) : (
                <ChevronRight className="size-4" />
              )}
            </button>
          ) : (
            <span
              className="mr-1 inline-block size-4 shrink-0"
              aria-hidden="true"
            />
          )}
          {cellContent}
        </span>
      ) : (
        cellContent
      )}
    </TableCell>
  );
};

// ─── DataGridTreeRow ─────────────────────────────────────────────────────────

type DataGridTreeRowProps<TRow extends object> = {
  row: Row<TRow>;
  treeMeta: TreeRowMeta;
  cClass: string;
  editingCell: EditingCell;
  onStartEdit: (next: EditingCell) => void;
  onCommitEdit: (rowOriginal: TRow, field: string, value: string) => void;
  onCancelEdit: () => void;
  onToggle: (pathKey: string) => void;
  onRowClick?: (params: {row: TRow; id: string}) => void;
  disableRowSelectionOnClick?: boolean;
};

/**
 * Renders a single tree data row with indentation and expand/collapse toggle.
 *
 * @wcag 1.3.1 — aria-level conveys depth in the treegrid
 * @wcag 4.1.2 — aria-expanded on parent rows, toggle buttons have aria-label
 * @wcag 2.1.1 — Enter/Space toggles expand on focused row
 */
const DataGridTreeRow = <TRow extends object>({
  row,
  treeMeta,
  cClass,
  editingCell,
  onStartEdit,
  onCommitEdit,
  onCancelEdit,
  onToggle,
  onRowClick,
  disableRowSelectionOnClick,
}: DataGridTreeRowProps<TRow>) => {
  const cells = row.getVisibleCells();
  // The first non-select column is where the tree indent + toggle renders
  const firstDataCellIndex = cells.findIndex(c => c.column.id !== '__select__');
  const rowInteractive = Boolean(onRowClick);
  const handleRowActivate = onRowClick
    ? () => onRowClick({row: row.original, id: row.id})
    : undefined;

  return (
    <TableRow
      data-slot="data-grid-row"
      data-state={row.getIsSelected() ? 'selected' : undefined}
      aria-level={treeMeta.indent + 1}
      aria-expanded={treeMeta.hasChildren ? treeMeta.isExpanded : undefined}
      tabIndex={rowInteractive ? 0 : undefined}
      aria-label={rowInteractive ? `Row ${row.id}` : undefined}
      className={cn(
        rowInteractive && !disableRowSelectionOnClick && 'cursor-pointer',
        rowInteractive &&
          'focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none'
      )}
      onClick={handleRowActivate}
      onKeyDown={
        rowInteractive && handleRowActivate
          ? e => {
              if (e.target !== e.currentTarget) return;
              if (e.key === 'Enter') {
                e.preventDefault();
                handleRowActivate();
              } else if (e.key === ' ') {
                e.preventDefault();
                handleRowActivate();
              }
            }
          : undefined
      }
    >
      {cells.map((cell, cellIndex) => (
        <TreeCell
          key={cell.id}
          cell={cell}
          cClass={cClass}
          isFirstDataCell={cellIndex === firstDataCellIndex}
          treeMeta={treeMeta}
          rowId={row.id}
          editingCell={editingCell}
          onStartEdit={onStartEdit}
          onCommit={v => onCommitEdit(row.original, cell.column.id, v)}
          onCancel={onCancelEdit}
          onToggle={onToggle}
        />
      ))}
    </TableRow>
  );
};

/** @internal */
export {DataGridTreeRow};
