// @docker/trident/src/components/data-grid-utils.ts
import type {
  ColumnMeta,
  CsvExportOptions,
  DataGridColumn,
} from './data-grid-types';
import type {ColumnDef, ColumnPinningState} from '@tanstack/react-table';

// ─── Column pinning ───────────────────────────────────────────────────────────

function dedup(arr: string[]): string[] {
  return arr.filter((v, i, a) => a.indexOf(v) === i);
}

/** @internal */
/** Merge column-level pin props with the pinnedColumns prop into a TanStack ColumnPinningState. */
export function buildColumnPinning<TRow extends object>(
  columns: DataGridColumn<TRow>[],
  pinnedColumns?: {left?: string[]; right?: string[]}
): ColumnPinningState {
  const left = dedup([
    ...columns.filter(c => c.pin === 'left').map(c => c.field),
    ...(pinnedColumns?.left ?? []),
  ]);
  const right = dedup([
    ...columns.filter(c => c.pin === 'right').map(c => c.field),
    ...(pinnedColumns?.right ?? []),
  ]);
  return {
    left: left.length ? left : undefined,
    right: right.length ? right : undefined,
  };
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

/** Convert an unknown cell/field value to a display string without hitting [object Object]. */
export function unknownToStr(v: unknown): string {
  if (v == null) return '';
  if (typeof v === 'string') return v;
  if (typeof v === 'number' || typeof v === 'boolean' || typeof v === 'bigint')
    return String(v);
  return JSON.stringify(v);
}

// ─── Column definition builder ────────────────────────────────────────────────

export function toColumnDef<TRow extends object>(
  col: DataGridColumn<TRow>
): ColumnDef<TRow> {
  return {
    id: col.field,
    accessorFn: col.valueGetter
      ? row => col.valueGetter?.(row)
      : row => Reflect.get(row, col.field),
    header: col.renderHeader
      ? () => col.renderHeader?.()
      : (col.headerName ?? col.field),
    cell: col.renderCell
      ? ({row, getValue}) =>
          col.renderCell?.({row: row.original, value: getValue()})
      : col.valueFormatter
        ? ({getValue}) => unknownToStr(col.valueFormatter?.(getValue()))
        : ({getValue}) => unknownToStr(getValue()),
    enableSorting: col.sortable !== false,
    sortDescFirst: false, // MUI parity: first click always sorts ascending
    size: col.width,
    minSize: col.minWidth,
    maxSize: col.maxWidth,
    meta: {
      align: col.align,
      headerAlign: col.headerAlign ?? col.align,
      disableColumnMenu: col.disableColumnMenu,
      resizable: col.resizable,
      pin: col.pin,
      filterable: col.filterable,
      filterOperators: col.filterOperators,
      editable: col.editable,
    } satisfies ColumnMeta,
  };
}

// ─── CSV export ───────────────────────────────────────────────────────────────

export function exportToCsv<TRow extends object>(
  columns: DataGridColumn<TRow>[],
  rows: TRow[],
  options: CsvExportOptions = {}
): void {
  const {fileName = 'export', delimiter = ',', utf8WithBom = false} = options;

  const visible = columns.filter(c => !c.hide);
  const header = visible.map(c => c.headerName ?? c.field);

  const data = rows.map(row =>
    visible.map(col => {
      const raw = col.valueGetter
        ? col.valueGetter(row)
        : Reflect.get(row, col.field);
      const formatted = col.valueFormatter ? col.valueFormatter(raw) : raw;
      return unknownToStr(formatted);
    })
  );

  const csv = [header, ...data]
    .map(line =>
      line.map(v => `"${String(v).replace(/"/g, '""')}"`).join(delimiter)
    )
    .join('\n');

  const bom = utf8WithBom ? '\uFEFF' : '';
  const blob = new Blob([bom + csv], {type: 'text/csv;charset=utf-8;'});
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = `${fileName}.csv`;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}
