// @docker/trident/src/components/date-time-picker.tsx
'use client';

import {Calendar} from '@/components/ui/calendar';
import {Label} from '@/components/ui/label';
import {cn} from '@/lib/utils';

import * as React from 'react';

// ─── Time Input ─────────────────────────────────────────────────────────────

type TimeInputProps = {
  value: {hours: number; minutes: number};
  onChange: (time: {hours: number; minutes: number}) => void;
  label?: string;
  className?: string;
};

function TimeInput({value, onChange, label, className}: TimeInputProps) {
  const hoursStr = String(value.hours).padStart(2, '0');
  const minutesStr = String(value.minutes).padStart(2, '0');

  const handleHoursChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const h = Math.min(
      23,
      Math.max(0, Number.parseInt(e.target.value, 10) || 0)
    );
    onChange({...value, hours: h});
  };

  const handleMinutesChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const m = Math.min(
      59,
      Math.max(0, Number.parseInt(e.target.value, 10) || 0)
    );
    onChange({...value, minutes: m});
  };

  return (
    <div
      data-slot="time-input"
      className={cn('flex items-center gap-1.5', className)}
    >
      {label && (
        <Label className="text-xs text-muted-foreground">{label}</Label>
      )}
      <div className="flex items-center gap-0.5">
        <input
          type="number"
          min={0}
          max={23}
          value={hoursStr}
          onChange={handleHoursChange}
          aria-label={label ? `${label} hours` : 'Hours'}
          className="h-8 w-10 rounded border border-input bg-background px-1 text-center text-sm tabular-nums outline-none focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring"
        />
        <span className="text-sm text-muted-foreground">:</span>
        <input
          type="number"
          min={0}
          max={59}
          value={minutesStr}
          onChange={handleMinutesChange}
          aria-label={label ? `${label} minutes` : 'Minutes'}
          className="h-8 w-10 rounded border border-input bg-background px-1 text-center text-sm tabular-nums outline-none focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring"
        />
      </div>
    </div>
  );
}

// ─── DateTimePicker ─────────────────────────────────────────────────────────

type DateTimePickerProps = {
  /** The selected date and time. */
  value?: Date;
  /** Called when the date or time changes. */
  onChange?: (date: Date | undefined) => void;
  /** Minimum selectable date. */
  minDate?: Date;
  /** Maximum selectable date. */
  maxDate?: Date;
  className?: string;
};

/**
 * A date and time picker combining Calendar with time inputs.
 * Replaces `@mui/x-date-pickers-pro/DateTimeRangePicker` for single datetime selection.
 *
 * ### Component Props
 *
 * | Prop        | Type                              | Default |
 * | :---------- | :-------------------------------- | :------ |
 * | _value_     | `Date`                            | -       |
 * | _onChange_   | `(date: Date \| undefined) => void` | -    |
 * | _minDate_   | `Date`                            | -       |
 * | _maxDate_   | `Date`                            | -       |
 *
 * ### Example Usage
 *
 * ```tsx
 * <DateTimePicker
 *   value={date}
 *   onChange={setDate}
 * />
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — time inputs have associated labels.
 * @wcag 2.1.1 Keyboard — Calendar and time inputs are fully keyboard navigable.
 */
function DateTimePicker({
  value,
  onChange,
  minDate,
  maxDate,
  className,
}: DateTimePickerProps) {
  const time = React.useMemo(
    () => ({
      hours: value?.getHours() ?? 0,
      minutes: value?.getMinutes() ?? 0,
    }),
    [value]
  );

  const handleDateSelect = (date: Date | undefined) => {
    if (!date) {
      onChange?.(undefined);
      return;
    }
    const merged = new Date(date);
    merged.setHours(time.hours, time.minutes, 0, 0);
    onChange?.(merged);
  };

  const handleTimeChange = (newTime: {hours: number; minutes: number}) => {
    const base = value ?? new Date();
    const merged = new Date(base);
    merged.setHours(newTime.hours, newTime.minutes, 0, 0);
    onChange?.(merged);
  };

  return (
    <div data-slot="date-time-picker" className={cn('space-y-3', className)}>
      <Calendar
        mode="single"
        selected={value}
        onSelect={handleDateSelect}
        disabled={
          minDate || maxDate
            ? date => {
                if (minDate && date < minDate) return true;
                if (maxDate && date > maxDate) return true;
                return false;
              }
            : undefined
        }
      />
      <div className="flex justify-center border-t border-border px-3 pt-3">
        <TimeInput value={time} onChange={handleTimeChange} label="Time" />
      </div>
    </div>
  );
}

// ─── DateTimeRangePicker ────────────────────────────────────────────────────

type DateTimeRange = {
  from?: Date;
  to?: Date;
};

type DateTimeRangePickerProps = {
  /** The selected date/time range. */
  value?: DateTimeRange;
  /** Called when the range changes. */
  onChange?: (range: DateTimeRange) => void;
  className?: string;
};

/**
 * A date-time range picker with two calendars and time inputs.
 * Replaces `@mui/x-date-pickers-pro/DateTimeRangePicker`.
 *
 * ### Component Props
 *
 * | Prop       | Type                                        | Default |
 * | :--------- | :------------------------------------------ | :------ |
 * | _value_    | `{ from?: Date; to?: Date }`                | -       |
 * | _onChange_  | `(range: DateTimeRange) => void`            | -       |
 *
 * ### Example Usage
 *
 * ```tsx
 * <DateTimeRangePicker
 *   value={range}
 *   onChange={setRange}
 * />
 * ```
 */
function DateTimeRangePicker({
  value,
  onChange,
  className,
}: DateTimeRangePickerProps) {
  const fromTime = React.useMemo(
    () => ({
      hours: value?.from?.getHours() ?? 0,
      minutes: value?.from?.getMinutes() ?? 0,
    }),
    [value?.from]
  );

  const toTime = React.useMemo(
    () => ({
      hours: value?.to?.getHours() ?? 23,
      minutes: value?.to?.getMinutes() ?? 59,
    }),
    [value?.to]
  );

  const handleRangeSelect = (range: {from?: Date; to?: Date} | undefined) => {
    if (!range) {
      onChange?.({});
      return;
    }
    const from = range.from ? new Date(range.from) : undefined;
    const to = range.to ? new Date(range.to) : undefined;
    if (from) from.setHours(fromTime.hours, fromTime.minutes, 0, 0);
    if (to) to.setHours(toTime.hours, toTime.minutes, 0, 0);
    onChange?.({from, to});
  };

  const handleFromTimeChange = (time: {hours: number; minutes: number}) => {
    const from = value?.from ? new Date(value.from) : new Date();
    from.setHours(time.hours, time.minutes, 0, 0);
    onChange?.({...value, from});
  };

  const handleToTimeChange = (time: {hours: number; minutes: number}) => {
    const to = value?.to ? new Date(value.to) : new Date();
    to.setHours(time.hours, time.minutes, 0, 0);
    onChange?.({...value, to});
  };

  return (
    <div
      data-slot="date-time-range-picker"
      className={cn('space-y-3', className)}
    >
      <Calendar
        mode="range"
        selected={
          value?.from && value?.to
            ? {from: value.from, to: value.to}
            : undefined
        }
        onSelect={handleRangeSelect}
        numberOfMonths={2}
      />
      <div className="flex justify-center gap-6 border-t border-border px-3 pt-3">
        <TimeInput
          value={fromTime}
          onChange={handleFromTimeChange}
          label="Start"
        />
        <TimeInput value={toTime} onChange={handleToTimeChange} label="End" />
      </div>
    </div>
  );
}

export {DateTimePicker, DateTimeRangePicker, TimeInput};
export type {
  DateTimePickerProps,
  DateTimeRangePickerProps,
  DateTimeRange,
  TimeInputProps,
};
