// @docker/trident/src/components/date-range-picker.tsx
'use client';

import type {DateRange} from 'react-day-picker';

import {Button} from '@/components/ui/button';
import {Calendar} from '@/components/ui/calendar';
import {
  Popover,
  PopoverAnchor,
  PopoverContent,
  PopoverTrigger,
} from '@/components/ui/popover';
import {cn} from '@/lib/utils';

import {
  endOfMonth,
  format,
  isValid,
  parse,
  startOfMonth,
  startOfToday,
  subDays,
} from 'date-fns';
import {CalendarIcon} from 'lucide-react';
import * as React from 'react';

// ─── Types ───────────────────────────────────────────────────────────────────

type DateRangePreset = {
  label: string;
  range: DateRange;
};

type DateRangePickerProps = {
  /** The controlled selected range. */
  value?: DateRange;
  /** Called when the user commits a selection (clicks Apply or selects a preset). */
  onValueChange?: (range: DateRange | undefined) => void;
  /** Shortcut presets shown in the sidebar. Pass [] to hide the sidebar entirely. */
  presets?: DateRangePreset[];
  /** Placeholder text shown when no range is selected. */
  placeholder?: string;
  /** Disables the trigger and prevents interaction. */
  disabled?: boolean;
  /** Earliest selectable date (inclusive). */
  minDate?: Date;
  /** Latest selectable date (inclusive). */
  maxDate?: Date;
  /** Disable all dates after today (inclusive). */
  disableFuture?: boolean;
  /** Disable all dates before today (inclusive). */
  disablePast?: boolean;
  /** Callback to individually disable arbitrary dates. Return true to disable. */
  shouldDisableDate?: (date: Date) => boolean;
  /**
   * Number of calendar months to show side-by-side.
   * @default 2
   */
  numberOfMonths?: number;
  /**
   * Controlled open state. When provided the caller is responsible for
   * opening and closing the popover.
   */
  open?: boolean;
  /** Called when the popover opens. */
  onOpen?: () => void;
  /** Called when the popover closes. */
  onClose?: () => void;
  /**
   * Trigger style.
   * - `"button"` — a single button showing the formatted range (default).
   * - `"input"` — two editable text inputs for start and end date, matching the
   *   `@mui/x-date-pickers-pro/DateRangePicker` UX.
   * @default "button"
   */
  variant?: 'button' | 'input';
  className?: string;
};

// ─── Constants ───────────────────────────────────────────────────────────────

const INPUT_FORMAT = 'MM/dd/yyyy';
const INPUT_PLACEHOLDER = 'MM/DD/YYYY';
const PARSE_FORMATS = [
  'MM/dd/yyyy',
  'M/d/yyyy',
  'MMM d, yyyy',
  'yyyy-MM-dd',
  'MM/dd/yy',
];

function tryParseDate(str: string): Date | null {
  const trimmed = str.trim();
  if (!trimmed) return null;
  const ref = new Date();
  for (const fmt of PARSE_FORMATS) {
    const parsedDate = parse(trimmed, fmt, ref);
    if (isValid(parsedDate)) return parsedDate;
  }
  return null;
}

// ─── Default presets ─────────────────────────────────────────────────────────

const getDefaultPresets = (): DateRangePreset[] => {
  const now = new Date();
  return [
    {label: 'Last 7 days', range: {from: subDays(now, 6), to: now}},
    {label: 'Last 30 days', range: {from: subDays(now, 29), to: now}},
    {label: 'Last 90 days', range: {from: subDays(now, 89), to: now}},
    {
      label: 'This month',
      range: {from: startOfMonth(now), to: endOfMonth(now)},
    },
  ];
};

const DEFAULT_PRESETS: DateRangePreset[] = getDefaultPresets();

// ─── DateRangePicker ─────────────────────────────────────────────────────────

/**
 * A popover-based date range picker for selecting a start and end date,
 * with optional preset shortcuts and an Apply/Clear footer.
 *
 * Replaces `@mui/x-date-pickers-pro/DateRangePicker` in hub-ui and app-ui.
 * For datetime range selection use `DateTimeRangePicker` instead.
 *
 * ### Component Props
 *
 * | Prop                | Type                                      | Default              |
 * | :------------------ | :---------------------------------------- | :------------------- |
 * | _value_             | `DateRange`                               | -                    |
 * | _onValueChange_     | `(range: DateRange \| undefined) => void` | -                    |
 * | _presets_           | `DateRangePreset[]`                       | `DEFAULT_PRESETS`    |
 * | _placeholder_       | `string`                                  | `"Pick a date range"` |
 * | _disabled_          | `boolean`                                 | `false`              |
 * | _minDate_           | `Date`                                    | -                    |
 * | _maxDate_           | `Date`                                    | -                    |
 * | _disableFuture_     | `boolean`                                 | `false`              |
 * | _disablePast_       | `boolean`                                 | `false`              |
 * | _shouldDisableDate_ | `(date: Date) => boolean`                 | -                    |
 * | _numberOfMonths_    | `number`                                  | `2`                  |
 * | _open_              | `boolean`                                 | -                    |
 * | _onOpen_            | `() => void`                              | -                    |
 * | _onClose_           | `() => void`                              | -                    |
 * | _variant_           | `"button" \| "input"`                     | `"button"`           |
 *
 * ### Example Usage
 *
 * ```tsx
 * import type {DateRange} from '@/components/date-range-picker';
 * import {DateRangePicker} from '@/components/date-range-picker';
 *
 * const [range, setRange] = React.useState<DateRange | undefined>();
 *
 * // Button trigger (default)
 * <DateRangePicker value={range} onValueChange={setRange} />
 *
 * // Text input trigger (MUI-style)
 * <DateRangePicker value={range} onValueChange={setRange} variant="input" />
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — trigger has aria-label matching the displayed range or placeholder.
 * @wcag 2.1.1 Keyboard — trigger opens on Enter/Space; Escape discards draft and closes.
 * @wcag 2.4.3 Focus Order — presets → calendar grid → Clear → Apply.
 * @wcag 2.4.7 Focus Visible — focus rings via existing Button and CalendarDayButton classes.
 * @wcag 4.1.2 Name, Role, Value — aria-disabled on disabled trigger; popover content has aria-label.
 */
function DateRangePicker({
  value,
  onValueChange,
  presets = DEFAULT_PRESETS,
  placeholder = 'Pick a date range',
  disabled = false,
  minDate,
  maxDate,
  disableFuture = false,
  disablePast = false,
  shouldDisableDate,
  numberOfMonths = 2,
  open: controlledOpen,
  onOpen,
  onClose,
  variant = 'button',
  className,
}: DateRangePickerProps) {
  const [internalOpen, setInternalOpen] = React.useState(false);
  const isControlled = controlledOpen !== undefined;
  const open = isControlled ? controlledOpen : internalOpen;

  const [draft, setDraft] = React.useState<DateRange | undefined>(value);
  const [fromStr, setFromStr] = React.useState(
    value?.from ? format(value.from, INPUT_FORMAT) : ''
  );
  const [toStr, setToStr] = React.useState(
    value?.to ? format(value.to, INPUT_FORMAT) : ''
  );

  // Sync draft to committed value on open; sync input strings when value changes.
  React.useEffect(() => {
    if (open) setDraft(value);
    else {
      setFromStr(value?.from ? format(value.from, INPUT_FORMAT) : '');
      setToStr(value?.to ? format(value.to, INPUT_FORMAT) : '');
    }
  }, [open, value]);

  // ── Helpers ────────────────────────────────────────────────────────────────

  const commitClose = (
    next: DateRange | undefined,
    nextFrom: string,
    nextTo: string
  ) => {
    onValueChange?.(next);
    setFromStr(nextFrom);
    setToStr(nextTo);
    if (!isControlled) setInternalOpen(false);
    onClose?.();
  };

  const handleOpenChange = React.useCallback(
    (next: boolean) => {
      if (!isControlled) setInternalOpen(next);
      if (next && !open) onOpen?.();
      if (!next && open) onClose?.();
    },
    [isControlled, open, onOpen, onClose]
  );

  // ── Calendar / footer handlers ─────────────────────────────────────────────

  const handleApply = () =>
    commitClose(
      draft,
      draft?.from ? format(draft.from, INPUT_FORMAT) : '',
      draft?.to ? format(draft.to, INPUT_FORMAT) : ''
    );

  const handlePreset = (range: DateRange) =>
    commitClose(
      range,
      range.from ? format(range.from, INPUT_FORMAT) : '',
      range.to ? format(range.to, INPUT_FORMAT) : ''
    );

  const handleClear = () => commitClose(undefined, '', '');

  const handleRangeSelect = (range: DateRange | undefined) => {
    setDraft(range);
    setFromStr(range?.from ? format(range.from, INPUT_FORMAT) : '');
    setToStr(range?.to ? format(range.to, INPUT_FORMAT) : '');
  };

  // ── Input variant handlers ──────────────────────────────────────────────────

  const handleFromBlur = () => {
    const parsed = tryParseDate(fromStr);
    if (parsed) {
      setDraft(prev => ({...prev, from: parsed}));
    } else {
      setFromStr(draft?.from ? format(draft.from, INPUT_FORMAT) : '');
    }
  };

  const handleToBlur = () => {
    const parsed = tryParseDate(toStr);
    if (parsed) {
      setDraft(prev => ({from: prev?.from, to: parsed}));
    } else {
      setToStr(draft?.to ? format(draft.to, INPUT_FORMAT) : '');
    }
  };

  // ── Labels ──────────────────────────────────────────────────────────────────

  const triggerLabel = value?.from
    ? value.to
      ? `${format(value.from, 'MMM d, yyyy')} – ${format(value.to, 'MMM d, yyyy')}`
      : format(value.from, 'MMM d, yyyy')
    : placeholder;

  const calendarDisabled = React.useCallback(
    (date: Date) => {
      if (minDate && date < minDate) return true;
      if (maxDate && date > maxDate) return true;
      if (disableFuture && date > new Date()) return true;
      if (disablePast && date < startOfToday()) return true;
      if (shouldDisableDate?.(date)) return true;
      return false;
    },
    [minDate, maxDate, disableFuture, disablePast, shouldDisableDate]
  );

  // ── Popover content (shared between variants) ───────────────────────────────

  const popoverContent = (
    <PopoverContent
      align="start"
      className="w-auto p-0"
      aria-label="Date range picker"
    >
      <div data-slot="date-range-picker" className="flex flex-col md:flex-row">
        {presets.length > 0 && (
          <div
            data-slot="date-range-presets"
            className="flex flex-row flex-wrap gap-1 border-b border-border p-3 md:flex-col md:border-r md:border-b-0"
          >
            {presets.map(preset => (
              <Button
                key={preset.label}
                variant="ghost"
                size="sm"
                onClick={() => handlePreset(preset.range)}
                className="justify-start text-sm"
              >
                {preset.label}
              </Button>
            ))}
          </div>
        )}
        <div data-slot="date-range-calendars" className="p-3">
          <Calendar
            mode="range"
            selected={draft}
            onSelect={handleRangeSelect}
            numberOfMonths={numberOfMonths}
            disabled={calendarDisabled}
          />
          <div className="flex items-center justify-end gap-2 border-t border-border pt-3">
            <Button variant="ghost" size="sm" onClick={handleClear}>
              Clear
            </Button>
            <Button size="sm" onClick={handleApply}>
              Apply
            </Button>
          </div>
        </div>
      </div>
    </PopoverContent>
  );

  // ── Render ──────────────────────────────────────────────────────────────────

  if (variant === 'input') {
    return (
      <Popover open={open} onOpenChange={handleOpenChange}>
        <PopoverAnchor asChild>
          <div
            data-slot="date-range-input"
            className={cn(
              'flex h-9 items-center gap-1 rounded border border-input bg-transparent px-2.5 text-sm shadow-raised transition-[border-color,box-shadow] focus-within:border-ring focus-within:ring-3 focus-within:ring-ring',
              disabled && 'pointer-events-none cursor-not-allowed opacity-50',
              className
            )}
          >
            <CalendarIcon
              className="size-4 shrink-0 text-muted-foreground"
              aria-hidden="true"
            />
            <input
              type="text"
              className="w-[5.5rem] min-w-0 bg-transparent outline-none placeholder:text-muted-foreground"
              value={fromStr}
              onChange={e => setFromStr(e.target.value)}
              onFocus={() => !open && handleOpenChange(true)}
              onBlur={handleFromBlur}
              placeholder={INPUT_PLACEHOLDER}
              aria-label="Start date"
              disabled={disabled}
            />
            <span className="text-muted-foreground" aria-hidden="true">
              –
            </span>
            <input
              type="text"
              className="w-[5.5rem] min-w-0 bg-transparent outline-none placeholder:text-muted-foreground"
              value={toStr}
              onChange={e => setToStr(e.target.value)}
              onFocus={() => !open && handleOpenChange(true)}
              onBlur={handleToBlur}
              placeholder={INPUT_PLACEHOLDER}
              aria-label="End date"
              disabled={disabled}
            />
          </div>
        </PopoverAnchor>
        {popoverContent}
      </Popover>
    );
  }

  return (
    <Popover open={open} onOpenChange={handleOpenChange}>
      <PopoverTrigger asChild>
        <Button
          variant="outline"
          disabled={disabled}
          aria-label={triggerLabel}
          className={cn(
            'justify-start gap-2 font-normal',
            !value?.from && 'text-muted-foreground',
            className
          )}
        >
          <CalendarIcon className="size-4 shrink-0" aria-hidden="true" />
          <span>{triggerLabel}</span>
        </Button>
      </PopoverTrigger>
      {popoverContent}
    </Popover>
  );
}

export {DateRangePicker, DEFAULT_PRESETS, getDefaultPresets};
export type {DateRange, DateRangePickerProps, DateRangePreset};
