// @docker/trident/src/components/digest.tsx
// digest.tsx

import type {AsChild} from '@/lib/types';
import type {ReactNode} from 'react';

import {CopyToClipboardButton} from '@/components/copy-to-clipboard-button';
import {Tooltip, TooltipContent, TooltipTrigger} from '@/components/ui/tooltip';
import {cn} from '@/lib/utils';

import {Slot} from 'radix-ui';

type DigestProps = {
  /** Full 64-character hex digest, without "sha256:" prefix */
  fullDigest: string;
  /** Truncate to maxLength characters. @default true */
  abbrev?: boolean;
  /** Max characters to show when abbreviated. @default 12 */
  maxLength?: number;
  /** Show copy-to-clipboard button. @default true */
  copyable?: boolean;
  tooltipPlacement?: 'top' | 'bottom' | 'left' | 'right';
  className?: string;
  /** Child element to render when `asChild` is true. */
  children?: ReactNode;
} & AsChild;

function Digest({
  fullDigest,
  abbrev = true,
  maxLength = 12,
  copyable = true,
  tooltipPlacement = 'bottom',
  className,
  asChild = false,
  children,
}: DigestProps) {
  const isAbbreviated = abbrev && fullDigest.length > maxLength;
  const displayText = isAbbreviated
    ? `${fullDigest.substring(0, maxLength)}\u2026`
    : fullDigest;

  // When abbreviated, the visible text is a truncated prefix. To preserve
  // keyboard + screen-reader access to the full digest (WCAG 2.1.1, 1.3.1,
  // 1.4.13) we render the trigger as a native `<button>` named with the full
  // digest via `aria-label`. Radix's TooltipTrigger then activates on focus
  // and can be dismissed with Escape (1.4.13). The inner text is marked
  // `aria-hidden` so AT does not announce the partial hash in addition to the
  // full name.
  const digestText = isAbbreviated ? (
    <button
      type="button"
      data-slot="digest-text"
      aria-label={fullDigest}
      className={cn(
        'cursor-default rounded-xs border-0 bg-transparent p-0 font-mono text-sm text-inherit outline-none',
        'focus-visible:ring-3 focus-visible:ring-ring'
      )}
    >
      <span aria-hidden="true">{displayText}</span>
    </button>
  ) : (
    <span data-slot="digest-text" className="font-mono text-sm">
      {displayText}
    </span>
  );

  if (asChild) {
    return (
      <Slot.Root
        data-slot="digest"
        className={cn('inline-flex items-center gap-1', className)}
      >
        {children}
      </Slot.Root>
    );
  }

  return (
    <span
      data-slot="digest"
      className={cn('inline-flex items-center gap-1', className)}
    >
      {isAbbreviated ? (
        <Tooltip>
          <TooltipTrigger asChild>{digestText}</TooltipTrigger>
          <TooltipContent side={tooltipPlacement} className="font-mono text-xs">
            {fullDigest}
          </TooltipContent>
        </Tooltip>
      ) : (
        digestText
      )}
      {copyable && (
        <CopyToClipboardButton
          text={fullDigest}
          size="icon-xs"
          className="size-5 text-muted-foreground hover:text-foreground"
        />
      )}
    </span>
  );
}

export {Digest};
export type {DigestProps};
