// @docker/trident/src/components/docker-sidebar.tsx
'use client';

import type {PropsWithChildren, ReactNode} from 'react';

import {
  Collapsible,
  CollapsibleContent,
  CollapsibleTrigger,
} from '@/components/ui/collapsible';
import {
  Sidebar,
  SidebarContent,
  SidebarFooter,
  SidebarHeader,
  SidebarMenuButton,
  SidebarMenuItem,
  SidebarMenuSub,
  SidebarMenuSubButton,
  SidebarMenuSubItem,
  SidebarProvider,
  SidebarRail,
  useSidebar,
} from '@/components/ui/sidebar';
import {cn} from '@/lib/utils';

import {ChevronRight} from 'lucide-react';
import * as React from 'react';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const COOKIE_NAME = 'sidebar:state';
const COOKIE_MAX_AGE = 60 * 60 * 24 * 7; // 7 days in seconds

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function readCookie(name: string): string | null {
  if (typeof document === 'undefined') return null;
  const match = document.cookie
    .split('; ')
    .find(row => row.startsWith(`${name}=`));
  return match ? decodeURIComponent(match.split('=')[1] ?? '') : null;
}

function writeCookie(name: string, value: string, maxAge: number): void {
  if (typeof document === 'undefined') return;
  document.cookie = `${name}=${encodeURIComponent(value)}; path=/; max-age=${maxAge}; SameSite=Lax`;
}

// ---------------------------------------------------------------------------
// DockerSidebar
// ---------------------------------------------------------------------------

type DockerSidebarProps = {
  /** The sidebar content, rendered as a persistent side panel. */
  sidebar?: ReactNode;
  /** The main page content, rendered alongside the sidebar. */
  children?: ReactNode;
  defaultOpen?: boolean;
  keyboardShortcut?: string | null; // null disables shortcut, default 'b'
  persistState?: boolean; // default true
};

type DockerSidebarInnerProps = {
  sidebar?: ReactNode;
  children?: ReactNode;
  keyboardShortcut: string | null;
  persistState: boolean;
};

const DockerSidebarInner = ({
  sidebar,
  children,
  keyboardShortcut,
  persistState,
}: DockerSidebarInnerProps) => {
  const {open, setOpen} = useSidebar();

  React.useEffect(() => {
    if (!persistState) return;
    writeCookie(COOKIE_NAME, String(open), COOKIE_MAX_AGE);
  }, [open, persistState]);

  React.useEffect(() => {
    if (!keyboardShortcut) return;

    const handleKeyDown = (event: KeyboardEvent) => {
      const isModifier = event.metaKey || event.ctrlKey;
      if (
        isModifier &&
        event.key.toLowerCase() === keyboardShortcut.toLowerCase()
      ) {
        event.preventDefault();
        setOpen(prev => !prev);
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    return () => document.removeEventListener('keydown', handleKeyDown);
  }, [keyboardShortcut, setOpen]);

  return (
    <>
      {sidebar}
      {children}
    </>
  );
};

/**
 * A thin wrapper around `SidebarProvider` that adds Docker-specific opinions:
 * - Cmd+B / Ctrl+B keyboard shortcut to toggle the sidebar
 * - Cookie persistence of the open/closed state across page loads
 */
const DockerSidebar = ({
  defaultOpen = true,
  keyboardShortcut = 'b',
  persistState = true,
  sidebar,
  children,
}: DockerSidebarProps) => {
  const resolvedDefaultOpen = React.useMemo(() => {
    if (!persistState) return defaultOpen;
    const cookie = readCookie(COOKIE_NAME);
    if (cookie === 'true') return true;
    if (cookie === 'false') return false;
    return defaultOpen;
  }, [defaultOpen, persistState]);

  return (
    <SidebarProvider defaultOpen={resolvedDefaultOpen}>
      <DockerSidebarInner
        sidebar={sidebar}
        keyboardShortcut={keyboardShortcut}
        persistState={persistState}
      >
        {children}
      </DockerSidebarInner>
    </SidebarProvider>
  );
};

// ---------------------------------------------------------------------------
// AppSidebar
// ---------------------------------------------------------------------------

type AppSidebarProps = PropsWithChildren<{
  header?: ReactNode;
  footer?: ReactNode;
  className?: string;
  variant?: 'sidebar' | 'floating' | 'inset';
}>;

/**
 * Docker app sidebar shell.
 * Persistent left panel, collapsible to icon-only via SidebarRail.
 * Must be a descendant of DockerSidebar (which provides SidebarProvider context).
 * Usage:
 * ```tsx
 * <DockerSidebar sidebar={<AppSidebar>...</AppSidebar>}>
 *   <main>page content</main>
 * </DockerSidebar>
 * ```
 */
const AppSidebar = ({
  children,
  header,
  footer,
  className,
  variant,
}: AppSidebarProps) => {
  return (
    <Sidebar collapsible="icon" variant={variant} className={className}>
      {header && <SidebarHeader>{header}</SidebarHeader>}
      <SidebarContent>{children}</SidebarContent>
      {footer && <SidebarFooter>{footer}</SidebarFooter>}
      <SidebarRail />
    </Sidebar>
  );
};

// ---------------------------------------------------------------------------
// AppSidebarItem / AppSidebarSubItem
// ---------------------------------------------------------------------------

type AppSidebarItemProps = PropsWithChildren<{
  icon?: ReactNode;
  detail?: ReactNode;
  label: string;
  href?: string;
  isActive?: boolean;
  className?: string;
}>;

/**
 * Navigation item for use inside AppSidebar.
 * Leaf: pass href for a link, omit for a button.
 * Parent: pass children (AppSidebarSubItem instances) for a collapsible group.
 */
const AppSidebarItem = ({
  icon,
  label,
  detail,
  href,
  isActive,
  children,
  className,
}: AppSidebarItemProps) => {
  const hasChildren = Boolean(children);
  const buttonSize = detail ? 'lg' : 'default';
  const itemContent = (
    <>
      {icon}
      {detail ? (
        <div className="grid min-w-0 gap-dds-050 leading-tight group-data-[collapsible=icon]:hidden">
          <span className="truncate text-label leading-label font-label tracking-label">
            {label}
          </span>
          <span className="truncate text-micro leading-snug tracking-caption text-sidebar-foreground-muted">
            {detail}
          </span>
        </div>
      ) : (
        label && <span>{label}</span>
      )}
    </>
  );

  if (hasChildren) {
    return (
      <Collapsible asChild defaultOpen={isActive} className="group/collapsible">
        <SidebarMenuItem>
          <CollapsibleTrigger asChild>
            <SidebarMenuButton
              tooltip={label}
              isActive={isActive}
              size={buttonSize}
              className={cn(className)}
            >
              {itemContent}
              <ChevronRight
                className={cn(
                  'ml-auto h-4 w-4 transition-transform duration-200',
                  'group-data-[state=open]/collapsible:rotate-90'
                )}
              />
            </SidebarMenuButton>
          </CollapsibleTrigger>
          <CollapsibleContent>
            <SidebarMenuSub>{children}</SidebarMenuSub>
          </CollapsibleContent>
        </SidebarMenuItem>
      </Collapsible>
    );
  }

  return (
    <SidebarMenuItem>
      <SidebarMenuButton
        asChild={Boolean(href)}
        tooltip={label}
        isActive={isActive}
        size={buttonSize}
        className={cn(className)}
        data-active={isActive ? 'true' : undefined}
      >
        {href ? (
          <a href={href} aria-label={label}>
            {itemContent}
          </a>
        ) : (
          itemContent
        )}
      </SidebarMenuButton>
    </SidebarMenuItem>
  );
};

/**
 * Sub-item for nested navigation inside AppSidebarItem parent.
 */
const AppSidebarSubItem = ({
  label,
  href,
  isActive,
  className,
}: Omit<AppSidebarItemProps, 'icon' | 'children'>) => {
  return (
    <SidebarMenuSubItem>
      <SidebarMenuSubButton
        asChild={Boolean(href)}
        isActive={isActive}
        className={cn(className)}
        data-active={isActive ? 'true' : undefined}
      >
        {href ? <a href={href}>{label}</a> : <span>{label}</span>}
      </SidebarMenuSubButton>
    </SidebarMenuSubItem>
  );
};

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {AppSidebar, AppSidebarItem, AppSidebarSubItem, DockerSidebar};
export type {AppSidebarItemProps, AppSidebarProps, DockerSidebarProps};
