// @docker/trident/src/components/error-state.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {OctagonX} from 'lucide-react';

type ErrorStateProps = {
  title: React.ReactNode;
  content?: React.ReactNode;
  actions?: React.ReactNode;
  /** Custom icon. Pass null to hide. @default OctagonX */
  icon?: React.ComponentType<React.SVGProps<SVGSVGElement>> | null;
  className?: string;
};

function ErrorState({
  title,
  content,
  actions,
  icon: Icon = OctagonX,
  className,
}: ErrorStateProps) {
  return (
    <div
      data-slot="error-state"
      className={cn(
        'flex flex-col items-center justify-center gap-4 py-16 text-center',
        className
      )}
    >
      {Icon !== null && (
        <div className="flex size-14 items-center justify-center rounded-full bg-destructive-subtle">
          <Icon className="size-7 text-destructive" aria-hidden="true" />
        </div>
      )}
      <div className="space-y-1">
        <p className="text-sm font-label text-foreground">{title}</p>
        {content && (
          <p className="max-w-sm text-sm text-muted-foreground">{content}</p>
        )}
      </div>
      {actions && <div className="flex gap-2">{actions}</div>}
    </div>
  );
}

export {ErrorState};
export type {ErrorStateProps};
