// @docker/trident/src/components/external-link.tsx
// external-link.tsx
'use client';

import type * as React from 'react';

import {Link} from '@/components/link';
import {Tooltip, TooltipContent, TooltipTrigger} from '@/components/ui/tooltip';
import {cn} from '@/lib/utils';

function ExternalIcon({className}: {className?: string}) {
  return (
    <svg
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
      className={cn('size-3', className)}
      aria-hidden="true"
    >
      <path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6" />
      <polyline points="15 3 21 3 21 9" />
      <line x1="10" y1="14" x2="21" y2="3" />
    </svg>
  );
}

type ExternalLinkProps = Omit<
  React.ComponentProps<typeof Link>,
  'target' | 'rel'
> & {
  /** Show external link icon. @default true */
  showIcon?: boolean;
  /** Tooltip text shown on hover. If omitted and `title` is set, `title` is used. */
  tooltip?: string;
  /** Custom tooltip content. Falls back to `href` when provided without `tooltip`. */
  title?: string;
  tooltipPlacement?: 'top' | 'bottom' | 'left' | 'right';
};

function ExternalLink({
  className,
  children,
  showIcon = true,
  tooltip,
  title,
  tooltipPlacement = 'top',
  href,
  onAuxClick,
  onClick,
  ...props
}: ExternalLinkProps) {
  // Resolve tooltip content: explicit `tooltip` wins, then `title`, then fall back to `href`
  const tooltipContent = tooltip || title || (href ?? undefined);

  const link = (
    <Link
      data-slot="external-link"
      href={href}
      target="_blank"
      rel="noopener noreferrer"
      className={cn('inline-flex items-baseline gap-0.5', className)}
      onAuxClick={event => {
        if (onAuxClick) {
          onAuxClick(event);
          return;
        }
        // Middle-click (button === 1) triggers onClick for consistent navigation handling
        if (event.button === 1) {
          onClick?.(event);
        }
      }}
      onClick={onClick}
      {...props}
    >
      {children}
      {/* WCAG 2.4.4 / axe best-practice: inform screen readers that the link opens in a new tab */}
      <span className="sr-only">(opens in new tab)</span>
      {(showIcon || !children) && (
        <ExternalIcon className="relative top-px shrink-0" />
      )}
    </Link>
  );

  if (!tooltipContent) return link;

  return (
    <Tooltip>
      <TooltipTrigger asChild>{link}</TooltipTrigger>
      <TooltipContent side={tooltipPlacement}>{tooltipContent}</TooltipContent>
    </Tooltip>
  );
}

export {ExternalLink};
export type {ExternalLinkProps};
