// @docker/trident/src/components/fab.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';

const fabVariants = cva(
  'inline-flex shrink-0 items-center justify-center rounded-full font-label shadow-floating ring-1 ring-border transition-all ease-default outline-none select-none focus-visible:ring-3 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0',
  {
    variants: {
      variant: {
        default: 'bg-primary text-primary-foreground hover:bg-accent-hover',
        secondary:
          'bg-secondary text-secondary-foreground hover:bg-secondary-hover',
        destructive:
          'bg-destructive text-destructive-foreground hover:bg-destructive-hover focus-visible:ring-destructive/50',
        outline:
          'border border-border bg-card hover:bg-muted hover:text-foreground',
      },
      size: {
        sm: "size-10 [&_svg:not([class*='size-'])]:size-4",
        default: "size-14 [&_svg:not([class*='size-'])]:size-6",
        lg: "size-16 [&_svg:not([class*='size-'])]:size-7",
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

type FabProps = React.ComponentProps<'button'> &
  VariantProps<typeof fabVariants>;

/**
 * A floating action button for primary screen actions. Replaces `@mui/material/Fab`.
 *
 * Renders a round, elevated button. Place it in a fixed corner via parent CSS
 * (e.g., `fixed bottom-6 right-6`). Pass an icon as children and provide an
 * `aria-label` for accessibility.
 *
 * ### Component Props
 *
 * | Prop      | Type                                                      | Default     |
 * | :-------- | :-------------------------------------------------------- | :---------- |
 * | _variant_ | `"default" \| "secondary" \| "destructive" \| "outline"`  | `"default"` |
 * | _size_    | `"sm" \| "default" \| "lg"`                               | `"default"` |
 *
 * ### Example Usage
 *
 * ```tsx
 * <div className="fixed bottom-6 right-6">
 *   <Fab aria-label="Add item">
 *     <PlusIcon />
 *   </Fab>
 * </div>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.4.3 Contrast (Minimum) — uses DDS semantic tokens with 4.5:1+ contrast.
 * @wcag 2.1.1 Keyboard — native `<button>` element, focusable and activated by Enter/Space.
 * @wcag 4.1.2 Name, Role, Value — requires `aria-label` since content is typically icon-only.
 */
function Fab({className, variant, size, ...props}: FabProps) {
  return (
    <button
      data-slot="fab"
      className={cn(fabVariants({variant, size, className}))}
      {...props}
    />
  );
}

export {Fab, fabVariants};
export type {FabProps};
