// @docker/trident/src/components/file-upload.tsx
'use client';

import {
  fileMatchesAccept,
  formatBytes,
  isImageFile,
} from '@/components/file-upload-utils';
import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {
  FileIcon,
  ImageIcon,
  Loader2Icon,
  UploadCloudIcon,
  XIcon,
} from 'lucide-react';
import * as React from 'react';

// ---------------------------------------------------------------------------
// CVA variants
// ---------------------------------------------------------------------------

const dropZoneVariants = cva(
  [
    'relative flex w-full cursor-pointer flex-col items-center justify-center rounded border-2 border-dashed',
    'transition-colors duration-150',
    'focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none',
    'aria-disabled:pointer-events-none aria-disabled:cursor-not-allowed aria-disabled:opacity-50',
  ],
  {
    variants: {
      variant: {
        default: 'border-border bg-hover-overlay hover:bg-hover-overlay',
        compact: 'border-border bg-hover-overlay hover:bg-hover-overlay',
      },
      size: {
        sm: 'gap-1 px-4 py-4 text-sm',
        default: 'gap-2 px-6 py-8 text-sm',
        lg: 'gap-3 px-8 py-12 text-base',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

// ---------------------------------------------------------------------------
// FilePreviewItem (internal)
// ---------------------------------------------------------------------------

type FilePreviewItemProps = {
  file: File;
  showThumbnail: boolean;
  onRemove: () => void;
  disabled?: boolean;
};

const FilePreviewItem = ({
  file,
  showThumbnail,
  onRemove,
  disabled,
}: FilePreviewItemProps) => {
  const [objectUrl, setObjectUrl] = React.useState<string | null>(null);

  React.useEffect(() => {
    if (!showThumbnail || !isImageFile(file)) return;
    const url = URL.createObjectURL(file);
    setObjectUrl(url);
    return () => URL.revokeObjectURL(url);
  }, [file, showThumbnail]);

  return (
    <li
      data-slot="file-upload-item"
      className="flex items-center gap-3 rounded border border-border bg-background px-3 py-2 text-sm"
    >
      {showThumbnail && objectUrl ? (
        <img
          src={objectUrl}
          alt={file.name}
          className="size-8 shrink-0 rounded object-cover"
        />
      ) : (
        <span className="flex size-8 shrink-0 items-center justify-center rounded bg-muted text-muted-foreground">
          {isImageFile(file) ? (
            <ImageIcon className="size-4" />
          ) : (
            <FileIcon className="size-4" />
          )}
        </span>
      )}

      <div className="min-w-0 flex-1">
        <p className="truncate font-label text-foreground">{file.name}</p>
        <p className="text-xs text-muted-foreground">
          {formatBytes(file.size)}
        </p>
      </div>

      <button
        type="button"
        onClick={onRemove}
        disabled={disabled}
        aria-label={`Remove ${file.name}`}
        className={cn(
          'flex size-6 shrink-0 items-center justify-center rounded text-muted-foreground',
          'hover:bg-muted hover:text-foreground',
          'focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none',
          'disabled:pointer-events-none disabled:opacity-50',
          'transition-colors'
        )}
      >
        <XIcon className="size-3.5" />
      </button>
    </li>
  );
};

// ---------------------------------------------------------------------------
// FileUpload
// ---------------------------------------------------------------------------

type FileUploadProps = {
  /** Accepted file types, e.g. "image/*" or ".pdf,.csv" */
  accept?: string;
  /** Maximum file size in bytes. Default: 5 MB */
  maxSize?: number;
  /** Maximum number of files. Default: 1 */
  maxFiles?: number;
  /** Controlled file list */
  value?: File[];
  /** Called when the file list changes */
  onFilesChange?: (files: File[]) => void;
  /** Called when a validation error occurs */
  onError?: (error: string) => void;
  /** Show spinner overlay — use while upload is in progress */
  uploading?: boolean;
  disabled?: boolean;
  className?: string;
  /** Override the drop-zone headline */
  label?: React.ReactNode;
  /** Override the description beneath the headline */
  description?: React.ReactNode;
  variant?: VariantProps<typeof dropZoneVariants>['variant'];
  size?: VariantProps<typeof dropZoneVariants>['size'];
};

/**
 * Drag-and-drop file upload zone with file list, validation, and image previews.
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.4.3 Contrast — labels and descriptions use DDS muted/foreground tokens.
 * @wcag 2.1.1 Keyboard — drop zone is focusable; Enter/Space open the file picker.
 * @wcag 2.4.7 Focus Visible — focus ring via `focus-visible:ring-3`.
 * @wcag 4.1.2 Name, Role, Value — `role="button"` with `aria-label`; `aria-disabled` on disabled state.
 */
const FileUpload = ({
  accept,
  maxSize = 5 * 1024 * 1024,
  maxFiles = 1,
  value,
  onFilesChange,
  onError,
  uploading = false,
  disabled = false,
  className,
  label,
  description,
  variant = 'default',
  size = 'default',
}: FileUploadProps) => {
  const multiple = maxFiles > 1;
  const inputRef = React.useRef<HTMLInputElement>(null);
  const [dragOver, setDragOver] = React.useState(false);
  const [internalFiles, setInternalFiles] = React.useState<File[]>([]);
  const [error, setError] = React.useState<string | null>(null);

  const files = value ?? internalFiles;
  const showThumbnails = Boolean(accept?.includes('image'));

  const validateAndAdd = React.useCallback(
    (incoming: File[]) => {
      const emitError = (msg: string) => {
        setError(msg);
        onError?.(msg);
      };

      if (accept) {
        const invalid = incoming.find(f => !fileMatchesAccept(f, accept));
        if (invalid) {
          emitError(`Invalid file type: ${invalid.name}`);
          return;
        }
      }

      const tooBig = incoming.find(f => f.size > maxSize);
      if (tooBig) {
        emitError(
          `File too large: ${tooBig.name} (${formatBytes(tooBig.size)} — max ${formatBytes(maxSize)})`
        );
        return;
      }

      const combined = [...files, ...incoming].slice(0, maxFiles);
      if (files.length + incoming.length > maxFiles) {
        emitError(`Too many files — max ${maxFiles}`);
      } else {
        setError(null);
      }

      if (!value) setInternalFiles(combined);
      onFilesChange?.(combined);
    },
    [accept, maxSize, maxFiles, files, value, onFilesChange, onError]
  );

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const picked = Array.from(e.target.files ?? []);
    if (picked.length > 0) validateAndAdd(picked);
    e.target.value = '';
  };

  const openPicker = () => {
    if (disabled || uploading) return;
    inputRef.current?.click();
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLButtonElement>) => {
    if (e.key === 'Enter' || e.key === ' ') {
      e.preventDefault();
      openPicker();
    }
  };

  const handleDragOver = (e: React.DragEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (!disabled && !uploading) setDragOver(true);
  };

  const handleDragLeave = (e: React.DragEvent<HTMLButtonElement>) => {
    e.preventDefault();
    setDragOver(false);
  };

  const handleDrop = (e: React.DragEvent<HTMLButtonElement>) => {
    e.preventDefault();
    setDragOver(false);
    if (disabled || uploading) return;
    const dropped = Array.from(e.dataTransfer.files);
    if (dropped.length > 0) validateAndAdd(dropped);
  };

  const removeFile = (index: number) => {
    const next = files.filter((_, i) => i !== index);
    if (!value) setInternalFiles(next);
    onFilesChange?.(next);
    if (next.length === 0) setError(null);
  };

  const defaultLabel =
    variant === 'compact'
      ? 'Click or drag to upload'
      : 'Drop files here or click to upload';

  const defaultDescription = (() => {
    const parts: string[] = [];
    if (accept) parts.push(accept.replace(/,/g, ', '));
    if (maxSize) parts.push(`max ${formatBytes(maxSize)}`);
    if (maxFiles > 1) parts.push(`up to ${maxFiles} files`);
    return parts.join(' · ');
  })();

  const isDisabled = disabled || uploading;
  const iconSize =
    size === 'sm' ? 'size-6' : size === 'lg' ? 'size-10' : 'size-8';

  return (
    <div
      data-slot="file-upload"
      className={cn('flex flex-col gap-3', className)}
    >
      <input
        ref={inputRef}
        type="file"
        accept={accept}
        multiple={multiple}
        disabled={isDisabled}
        onChange={handleInputChange}
        aria-hidden="true"
        aria-label="File upload"
        tabIndex={-1}
        className="sr-only"
        data-slot="file-upload-input"
      />

      <button
        type="button"
        aria-label="File upload drop zone"
        aria-disabled={isDisabled}
        disabled={isDisabled}
        onKeyDown={handleKeyDown}
        onClick={openPicker}
        onDragOver={handleDragOver}
        onDragLeave={handleDragLeave}
        onDrop={handleDrop}
        className={cn(
          dropZoneVariants({variant, size}),
          dragOver && 'border-primary bg-accent-subtle',
          error &&
            'border-destructive bg-destructive-wash focus-visible:ring-destructive'
        )}
      >
        {uploading && (
          <output
            aria-label="Uploading"
            className="absolute inset-0 flex items-center justify-center rounded bg-overlay-frosted backdrop-blur-sm"
          >
            <Loader2Icon className="size-6 animate-spin text-muted-foreground" />
          </output>
        )}

        <UploadCloudIcon
          className={cn(iconSize, 'text-muted-foreground')}
          aria-hidden="true"
        />

        <p className="font-label text-foreground">{label ?? defaultLabel}</p>

        {(description ?? defaultDescription) ? (
          <p className="text-center text-xs text-muted-foreground">
            {description ?? defaultDescription}
          </p>
        ) : null}
      </button>

      {error && (
        <p
          role="alert"
          data-slot="file-upload-error"
          className="text-sm text-destructive"
        >
          {error}
        </p>
      )}

      {files.length > 0 && (
        <ul
          data-slot="file-upload-list"
          aria-label="Selected files"
          className="flex flex-col gap-2"
        >
          {files.map((file, i) => (
            <FilePreviewItem
              key={`${file.name}-${file.size}`}
              file={file}
              showThumbnail={showThumbnails}
              onRemove={() => removeFile(i)}
              disabled={isDisabled}
            />
          ))}
        </ul>
      )}
    </div>
  );
};

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {FileUpload, dropZoneVariants, formatBytes};
export type {FileUploadProps};
