// @docker/trident/src/components/freshness-indicator.tsx
import type * as React from 'react';

// freshness-indicator.tsx
import {RelativeTime} from '@/components/relative-time';
import {cn} from '@/lib/utils';

type FreshnessBand = 'fresh' | 'aging' | 'stale';

/**
 * Infers a freshness band from a date.
 * - fresh  — < 7 days
 * - aging  — 7–29 days (exclusive upper bound: diffDays < 30)
 * - stale  — ≥ 30 days
 */
function inferBand(date: Date | string): FreshnessBand {
  const d = date instanceof Date ? date : new Date(date);
  const diffDays = (Date.now() - d.getTime()) / (1000 * 60 * 60 * 24);
  if (diffDays < 7) return 'fresh';
  if (diffDays < 30) return 'aging';
  return 'stale';
}

const BAND_COLOR: Record<FreshnessBand, string> = {
  fresh: 'text-success',
  aging: 'text-warning',
  stale: 'text-destructive',
};

type FreshnessIndicatorProps = React.ComponentProps<'span'> & {
  /** ISO date string or Date of last update. */
  date: Date | string;
  /** Override the inferred band. */
  band?: FreshnessBand;
};

function FreshnessIndicator({
  date,
  band,
  className,
  ...props
}: FreshnessIndicatorProps) {
  const resolvedBand = band ?? inferBand(date);

  return (
    <span
      data-slot="freshness-indicator"
      className={cn(
        'inline-flex items-center gap-1 text-xs',
        BAND_COLOR[resolvedBand],
        className
      )}
      {...props}
    >
      <RelativeTime date={date} className="text-xs text-inherit" />
    </span>
  );
}

export {FreshnessIndicator, inferBand};
export type {FreshnessBand, FreshnessIndicatorProps};
