// @docker/trident/src/components/health-grade.tsx
// health-grade.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

/** Scout health grade A-F color map using semantic design tokens */
const GRADE_CLASSES: Record<
  string,
  {bg: string; text: string; border: string}
> = {
  A: {
    bg: 'bg-grade-a-bg',
    text: 'text-grade-a-text',
    border: 'border-grade-a-border',
  },
  B: {
    bg: 'bg-grade-b-bg',
    text: 'text-grade-b-text',
    border: 'border-grade-b-border',
  },
  C: {
    bg: 'bg-grade-c-bg',
    text: 'text-grade-c-text',
    border: 'border-grade-c-border',
  },
  D: {
    bg: 'bg-grade-d-bg',
    text: 'text-grade-d-text',
    border: 'border-grade-d-border',
  },
  F: {
    bg: 'bg-grade-f-bg',
    text: 'text-grade-f-text',
    border: 'border-grade-f-border',
  },
};

type HealthGradeProps = React.ComponentProps<'span'> & {
  grade: 'A' | 'B' | 'C' | 'D' | 'F';
  /**
   * Visual display variant:
   * - `'badge'` (default) — 28×28 letter badge
   * - `'badge-sm'` — 20×20 smaller letter badge
   * - `'icon'` — colored shield icon, no letter
   */
  display?: 'badge' | 'badge-sm' | 'icon';
};

function HealthGrade({
  grade,
  display = 'badge',
  className,
  ...props
}: HealthGradeProps) {
  const gradeClasses = GRADE_CLASSES[grade] ?? GRADE_CLASSES['F'];
  const bg = gradeClasses?.bg ?? '';
  const text = gradeClasses?.text ?? '';
  const border = gradeClasses?.border ?? '';

  if (display === 'icon') {
    return (
      <span
        data-slot="health-grade"
        role="img"
        aria-label={`Scout health grade ${grade}`}
        className={cn(
          'inline-flex items-center justify-center rounded-full p-0.5',
          bg,
          className
        )}
        {...props}
      >
        <svg
          aria-hidden="true"
          className={cn('h-4 w-4', text)}
          viewBox="0 0 20 20"
          fill="currentColor"
        >
          <path
            fillRule="evenodd"
            d="M10 1a.75.75 0 0 1 .494.187l6.25 5.5A.75.75 0 0 1 17 7.25v3.5C17 14.104 13.855 17.5 10 17.5S3 14.104 3 10.75v-3.5a.75.75 0 0 1 .256-.563l6.25-5.5A.75.75 0 0 1 10 1Z"
            clipRule="evenodd"
          />
        </svg>
      </span>
    );
  }

  const sizeClass =
    display === 'badge-sm' ? 'h-5 w-5 text-xs' : 'h-7 w-7 text-sm';

  return (
    <span
      data-slot="health-grade"
      role="img"
      aria-label={`Scout health grade ${grade}`}
      className={cn(
        'inline-flex items-center justify-center rounded-full border font-bold tabular-nums select-none',
        sizeClass,
        bg,
        text,
        border,
        className
      )}
      {...props}
    >
      {grade}
    </span>
  );
}

export {HealthGrade};
export type {HealthGradeProps};
