// @docker/trident/src/components/inline-segmented-toggle.tsx
import {cn} from '@/lib/utils';

type InlineSegmentedToggleOption<T extends string> = {
  value: T;
  label: string;
};

type InlineSegmentedToggleProps<T extends string> = {
  options: InlineSegmentedToggleOption<T>[];
  value: T;
  onChange: (value: T) => void;
  className?: string;
};

/**
 * Compact inline toggle for mutually exclusive options. Designed to live
 * inside dense node cards — always visible, no hover required. Uses
 * radiogroup semantics so assistive tech understands the single-selection
 * constraint.
 *
 * @example
 * <InlineSegmentedToggle
 *   options={[{value: 'local', label: 'local'}, {value: 'remote', label: 'remote'}]}
 *   value={mode}
 *   onChange={setMode}
 * />
 */
function InlineSegmentedToggle<T extends string>({
  options,
  value,
  onChange,
  className,
}: InlineSegmentedToggleProps<T>) {
  return (
    <div
      role="radiogroup"
      className={cn(
        'flex items-center gap-px rounded-[3px] border border-border-subtle bg-hover-overlay p-[2px]',
        className
      )}
    >
      {options.map(opt => (
        <InlineSegmentedToggleItem
          key={opt.value}
          active={opt.value === value}
          onClick={() => onChange(opt.value)}
        >
          {opt.label}
        </InlineSegmentedToggleItem>
      ))}
    </div>
  );
}

function InlineSegmentedToggleItem({
  active,
  onClick,
  children,
}: {
  active: boolean;
  onClick: () => void;
  children: React.ReactNode;
}) {
  return (
    <button
      type="button"
      role="radio"
      aria-checked={active}
      onClick={e => {
        e.stopPropagation();
        onClick();
      }}
      className={cn(
        'rounded-[2px] px-1.5 py-[1px] text-[10px] font-label tracking-[0.04em] uppercase transition-colors',
        active
          ? 'bg-primary text-primary-foreground'
          : 'text-muted-foreground hover:bg-hover-overlay hover:text-foreground'
      )}
    >
      {children}
    </button>
  );
}

export {InlineSegmentedToggle};
export type {InlineSegmentedToggleOption, InlineSegmentedToggleProps};
