// @docker/trident/src/components/inspector-panel.tsx
// inspector-panel.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {X} from 'lucide-react';

type Props = {
  open: boolean;
  onClose: () => void;
  title?: string;
  width?: number;
  children: React.ReactNode;
  className?: string;
};

function InspectorPanel({
  open,
  onClose,
  title,
  width = 320,
  children,
  className,
}: Props) {
  return (
    <aside
      data-slot="inspector-panel"
      className={cn(
        'absolute top-0 right-0 z-overlay flex h-full flex-col overflow-hidden',
        'bg-[var(--card)] transition-transform duration-normal ease-[var(--dds-ease-out)]',
        open ? 'translate-x-0' : 'pointer-events-none translate-x-full',
        className
      )}
      style={{
        width,
        boxShadow: open
          ? '-8px 0 24px oklch(0 0 0 / 0.25), -1px 0 0 oklch(0 0 0 / 0.1)'
          : 'none',
      }}
      aria-hidden={!open}
      inert={!open}
    >
      {title !== undefined && (
        <header
          data-slot="inspector-panel-header"
          className="flex h-9 shrink-0 items-center justify-between border-b border-[var(--border)] px-3.5"
        >
          <span className="min-w-0 truncate font-mono text-[13px] font-semibold tracking-[-0.011em] text-[var(--foreground)]">
            {title}
          </span>
          <button
            type="button"
            onClick={onClose}
            aria-label="Close"
            className="flex size-5 shrink-0 items-center justify-center rounded-[var(--dds-radius-sm)] text-muted-foreground transition-colors hover:bg-muted hover:text-foreground focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none"
          >
            <X className="size-3" />
          </button>
        </header>
      )}
      <div
        data-slot="inspector-panel-content"
        className="flex-1 overflow-y-auto"
      >
        {children}
      </div>
    </aside>
  );
}

export {InspectorPanel};
export type {Props as InspectorPanelProps};
