// @docker/trident/src/components/link.tsx
// link.tsx

import type {AsChild} from '@/lib/types';
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {Slot} from 'radix-ui';

const linkVariants = cva(
  'underline underline-offset-4 transition-colors ease-default outline-none focus-visible:ring-3 focus-visible:ring-ring high-contrast:focus-visible:ring-ring',
  {
    variants: {
      variant: {
        default:
          'text-primary hover:text-link-hover high-contrast:text-link-hover high-contrast:hover:text-white',
        muted:
          'text-muted-foreground hover:text-foreground high-contrast:text-foreground',
        destructive:
          'text-destructive hover:text-destructive-link-hover high-contrast:text-text-error high-contrast:hover:text-white',
      },
    },
    defaultVariants: {variant: 'default'},
  }
);

type LinkProps = React.ComponentProps<'a'> &
  AsChild &
  VariantProps<typeof linkVariants>;

function Link({className, variant, asChild = false, ...props}: LinkProps) {
  const Comp = asChild ? Slot.Root : 'a';
  return (
    <Comp
      data-slot="link"
      className={cn(linkVariants({variant}), className)}
      {...props}
    />
  );
}

export {Link, linkVariants};
export type {LinkProps};
