// @docker/trident/src/components/log-viewer-ansi.ts
/**
 * Minimal ANSI escape code parser for log viewer output.
 *
 * Handles:
 *   - Colors 30-37 (standard foreground)
 *   - Colors 90-97 (bright/high-intensity foreground)
 *   - Bold (1)
 *   - Reset (0)
 *
 * Returns an array of text segments, each with an optional DDS token class.
 * @internal
 */

export type AnsiSegment = {
  text: string;
  className?: string;
};

/**
 * Maps ANSI color codes to DDS semantic token classes.
 *
 * All mode variants (light/dark/high-contrast) are handled by the token layer —
 * each single class automatically resolves to the correct OKLCH value for the
 * active surface mode. No inline dark:/high-contrast: variants needed here.
 */
const ANSI_COLOR_MAP: Record<number, string> = {
  // Standard colors (ANSI 30-37)
  30: 'text-terminal-black',
  31: 'text-terminal-red',
  32: 'text-terminal-green',
  33: 'text-terminal-yellow',
  34: 'text-terminal-blue',
  35: 'text-terminal-magenta',
  36: 'text-terminal-cyan',
  37: 'text-terminal-white',
  // Bright colors (ANSI 90-97)
  90: 'text-terminal-bright-black',
  91: 'text-terminal-bright-red',
  92: 'text-terminal-bright-green',
  93: 'text-terminal-bright-yellow',
  94: 'text-terminal-bright-blue',
  95: 'text-terminal-bright-magenta',
  96: 'text-terminal-bright-cyan',
  97: 'text-terminal-bright-white',
};

const ANSI_ESCAPE_RE = /\u001b\[([0-9;]*)m/g;

function resolveAnsiClass(codes: number[]): string | undefined {
  let cls: string | undefined;
  for (const code of codes) {
    if (code === 0) {
      // Reset
      cls = undefined;
    } else if (code === 1) {
      // Bold
      cls = cls ? `${cls} font-bold` : 'font-bold';
    } else if (ANSI_COLOR_MAP[code]) {
      const color = ANSI_COLOR_MAP[code];
      if (cls?.includes('font-bold')) {
        cls = `${color} font-bold`;
      } else {
        cls = color;
      }
    }
  }
  return cls;
}

/**
 * Parse a string containing ANSI escape sequences into styled segments.
 * Returns segments with `text` content and optional Tailwind `className`.
 */
export function parseAnsi(input: string): AnsiSegment[] {
  const segments: AnsiSegment[] = [];
  let currentClass: string | undefined;
  let lastIndex = 0;

  ANSI_ESCAPE_RE.lastIndex = 0;
  let match: RegExpExecArray | null;

  match = ANSI_ESCAPE_RE.exec(input);
  while (match !== null) {
    const [fullMatch, rawCodes] = match;
    const matchStart = match.index;

    if (matchStart > lastIndex) {
      segments.push({
        text: input.slice(lastIndex, matchStart),
        className: currentClass,
      });
    }

    const codes = (rawCodes ?? '')
      .split(';')
      .map(Number)
      .filter(n => !Number.isNaN(n));

    currentClass = resolveAnsiClass(codes.length > 0 ? codes : [0]);
    lastIndex = matchStart + fullMatch.length;

    match = ANSI_ESCAPE_RE.exec(input);
  }

  if (lastIndex < input.length) {
    segments.push({
      text: input.slice(lastIndex),
      className: currentClass,
    });
  }

  if (segments.length === 0) {
    segments.push({text: input});
  }

  return segments;
}
