// @docker/trident/src/components/log-viewer.tsx
'use client';

import type * as React from 'react';

import {parseAnsi} from '@/components/log-viewer-ansi';
import {Search} from '@/components/search';
import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import {useVirtualizer} from '@tanstack/react-virtual';
import {ArrowDownIcon, PauseIcon, PlayIcon} from 'lucide-react';
import {useCallback, useEffect, useMemo, useRef, useState} from 'react';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type LogLevel = 'info' | 'warn' | 'error' | 'debug';
type LogStream = 'stdout' | 'stderr';

type LogEntry = {
  id: string | number;
  timestamp?: Date | string;
  level?: LogLevel;
  stream?: LogStream;
  message: string;
};

type LogFilterFn = (entry: LogEntry, search: string) => boolean;

type LogViewerProps = {
  entries: LogEntry[];
  /** Container height. Default 400. */
  height?: number | string;
  /** Auto-scroll to newest entry when true. Default false. */
  follow?: boolean;
  /** Called when follow state should change (user pauses or resumes). */
  onFollowChange?: (follow: boolean) => void;
  /** Show ISO timestamp column. Default true. */
  showTimestamps?: boolean;
  /** Show 1-based line numbers. Default false. */
  showLineNumbers?: boolean;
  /** Controlled search value. Pair with `onSearchChange`. */
  searchValue?: string;
  /** Called when the toolbar search input changes. */
  onSearchChange?: (value: string) => void;
  /**
   * Predicate applied to every entry against the active search value. Default
   * is a case-insensitive substring match on `entry.message`. Return `true`
   * for every entry to disable filtering and only highlight matches.
   */
  filterFn?: LogFilterFn;
  className?: string;
  /** Font family override (defaults to `font-mono`). */
  fontFamily?: string;
};

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const LEVEL_CLASS: Record<LogLevel, string> = {
  error: 'text-destructive',
  warn: 'text-warning',
  debug: 'text-muted-foreground',
  info: 'text-foreground',
};

const AT_BOTTOM_THRESHOLD = 8; // px tolerance for "is scrolled to bottom"
const ESTIMATED_ROW_HEIGHT = 24; // px
const OVERSCAN = 10;

const defaultFilterFn: LogFilterFn = (entry, search) =>
  entry.message.toLowerCase().includes(search.toLowerCase());

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function formatTimestamp(ts: Date | string): string {
  const date = ts instanceof Date ? ts : new Date(ts);
  return date.toISOString();
}

type TextPart = {text: string; match: boolean};

function splitBySearch(text: string, lowerSearch: string): TextPart[] {
  if (!lowerSearch) return [{text, match: false}];
  const parts: TextPart[] = [];
  const lowerText = text.toLowerCase();
  let cursor = 0;

  while (cursor < text.length) {
    const idx = lowerText.indexOf(lowerSearch, cursor);
    if (idx === -1) {
      parts.push({text: text.slice(cursor), match: false});
      break;
    }
    if (idx > cursor) {
      parts.push({text: text.slice(cursor, idx), match: false});
    }
    parts.push({
      text: text.slice(idx, idx + lowerSearch.length),
      match: true,
    });
    cursor = idx + lowerSearch.length;
  }
  return parts;
}

// ---------------------------------------------------------------------------
// Sub-component: rendered message with ANSI colors + search highlighting
// ---------------------------------------------------------------------------

type MessageProps = {message: string; search: string};

const LogMessage = ({message, search}: MessageProps) => {
  const segments = parseAnsi(message);
  const lowerSearch = search.toLowerCase();

  return (
    <>
      {segments.map((seg, segIdx) => {
        if (!search) {
          return (
            <span key={segIdx} className={seg.className}>
              {seg.text}
            </span>
          );
        }
        const parts = splitBySearch(seg.text, lowerSearch);
        return (
          <span key={segIdx} className={seg.className}>
            {parts.map((part, partIdx) =>
              part.match ? (
                <mark
                  key={partIdx}
                  className="rounded-xs bg-log-warn-bg text-alert-warning-fg"
                >
                  {part.text}
                </mark>
              ) : (
                <span key={partIdx}>{part.text}</span>
              )
            )}
          </span>
        );
      })}
    </>
  );
};

// ---------------------------------------------------------------------------
// Sub-component: a single log row
// ---------------------------------------------------------------------------

type LogRowProps = {
  entry: LogEntry;
  index: number;
  showTimestamps: boolean;
  showLineNumbers: boolean;
  searchValue: string;
};

const LogRow = ({
  entry,
  index,
  showTimestamps,
  showLineNumbers,
  searchValue,
}: LogRowProps) => {
  const isStderr = entry.stream === 'stderr';
  const levelClass = entry.level ? LEVEL_CLASS[entry.level] : 'text-foreground';

  return (
    <div
      data-slot="log-row"
      className={cn(
        'flex min-h-(--dds-space-600) w-full items-baseline gap-2 px-3 py-0.5',
        'font-mono text-xs',
        isStderr && 'bg-destructive-wash'
      )}
    >
      {showLineNumbers && (
        <span
          aria-hidden="true"
          className="w-10 shrink-0 text-right text-muted-foreground tabular-nums select-none"
        >
          {index + 1}
        </span>
      )}
      {showTimestamps && entry.timestamp && (
        <span className="shrink-0 text-muted-foreground tabular-nums">
          <time dateTime={formatTimestamp(entry.timestamp)}>
            {formatTimestamp(entry.timestamp)}
          </time>
        </span>
      )}
      {entry.level && (
        <span className={cn('w-12 shrink-0 font-bold uppercase', levelClass)}>
          {entry.level}
        </span>
      )}
      <span className={cn('flex-1 break-all whitespace-pre-wrap', levelClass)}>
        <LogMessage message={entry.message} search={searchValue} />
      </span>
    </div>
  );
};

// ---------------------------------------------------------------------------
// Main component
// ---------------------------------------------------------------------------

function LogViewer({
  entries,
  height = 400,
  follow = false,
  onFollowChange,
  showTimestamps = true,
  showLineNumbers = false,
  searchValue,
  onSearchChange,
  filterFn = defaultFilterFn,
  className,
  fontFamily,
}: LogViewerProps) {
  const scrollRef = useRef<HTMLDivElement>(null);

  // Search state — controlled if `searchValue` is provided, otherwise managed
  // internally so consumers get filtering for free.
  const [internalSearch, setInternalSearch] = useState('');
  const isSearchControlled = searchValue !== undefined;
  const activeSearch = isSearchControlled ? searchValue : internalSearch;

  const handleSearchChange = useCallback(
    (next: string) => {
      if (!isSearchControlled) setInternalSearch(next);
      onSearchChange?.(next);
    },
    [isSearchControlled, onSearchChange]
  );

  // Filter entries against the active search value. We keep the original
  // index alongside the entry so consumers (and line numbers) can reference
  // the unfiltered position.
  const filteredEntries = useMemo(() => {
    if (activeSearch === '') {
      return entries.map((entry, originalIndex) => ({entry, originalIndex}));
    }
    const result: {entry: LogEntry; originalIndex: number}[] = [];
    for (let i = 0; i < entries.length; i += 1) {
      const entry = entries[i];
      if (filterFn(entry, activeSearch)) {
        result.push({entry, originalIndex: i});
      }
    }
    return result;
  }, [entries, activeSearch, filterFn]);

  const visibleCount = filteredEntries.length;
  const prevCountRef = useRef(visibleCount);

  const virtualizer = useVirtualizer({
    count: visibleCount,
    getScrollElement: () => scrollRef.current,
    estimateSize: () => ESTIMATED_ROW_HEIGHT,
    overscan: OVERSCAN,
    getItemKey: index => filteredEntries[index]?.entry.id ?? index,
  });

  // Auto-scroll to bottom when following and new (visible) entries arrive
  useEffect(() => {
    if (!follow) return;
    if (visibleCount === prevCountRef.current) return;
    prevCountRef.current = visibleCount;
    if (visibleCount > 0) {
      virtualizer.scrollToIndex(visibleCount - 1, {align: 'end'});
    }
  }, [follow, visibleCount, virtualizer]);

  // Sync count and scroll when follow toggles on
  useEffect(() => {
    if (follow) {
      prevCountRef.current = visibleCount;
      if (visibleCount > 0) {
        virtualizer.scrollToIndex(visibleCount - 1, {align: 'end'});
      }
    }
  }, [follow, visibleCount, virtualizer]);

  const handleScroll = useCallback(
    (e: React.UIEvent<HTMLDivElement>) => {
      if (!follow) return;
      const el = e.currentTarget;
      const atBottom =
        Math.abs(el.scrollHeight - el.scrollTop - el.clientHeight) <
        AT_BOTTOM_THRESHOLD;
      if (!atBottom) {
        onFollowChange?.(false);
      }
    },
    [follow, onFollowChange]
  );

  const handleFollowToggle = useCallback(() => {
    const next = !follow;
    onFollowChange?.(next);
    if (next && visibleCount > 0) {
      virtualizer.scrollToIndex(visibleCount - 1, {align: 'end'});
    }
  }, [follow, onFollowChange, visibleCount, virtualizer]);

  const outerStyle: React.CSSProperties = {
    height: typeof height === 'number' ? `${height}px` : height,
    ...(fontFamily ? {fontFamily} : {}),
  };

  const totalSize = virtualizer.getTotalSize();
  const virtualItems = virtualizer.getVirtualItems();

  const isEmpty = entries.length === 0;
  const hasNoMatches = !isEmpty && visibleCount === 0;

  return (
    <div
      data-slot="log-viewer"
      className={cn(
        'relative flex flex-col overflow-hidden rounded border border-border bg-background',
        className
      )}
      style={outerStyle}
    >
      {/* Toolbar */}
      <div
        role="toolbar"
        aria-label="Log viewer controls"
        data-slot="log-viewer-toolbar"
        className="flex shrink-0 items-center gap-2 border-b border-border bg-hover-overlay px-2 py-1.5"
      >
        <Search
          className="h-7 flex-1 text-xs"
          placeholder="Search logs..."
          value={activeSearch}
          onChange={handleSearchChange}
          aria-label="Search log entries"
        />
        <Button
          variant="ghost"
          size="icon-sm"
          onClick={handleFollowToggle}
          aria-label={follow ? 'Pause log follow' : 'Follow latest logs'}
          aria-pressed={follow}
          title={follow ? 'Pause follow' : 'Follow latest logs'}
        >
          {follow ? (
            <PauseIcon className="size-3.5" />
          ) : (
            <PlayIcon className="size-3.5" />
          )}
        </Button>
      </div>

      {/* Virtualized scroll area */}
      <div className="relative min-h-0 flex-1">
        {isEmpty ? (
          <div
            role="status"
            aria-live="polite"
            className="flex h-full items-center justify-center font-mono text-xs text-muted-foreground"
          >
            No log entries
          </div>
        ) : hasNoMatches ? (
          <div
            role="status"
            aria-live="polite"
            className="flex h-full items-center justify-center font-mono text-xs text-muted-foreground"
          >
            No log entries match &ldquo;{activeSearch}&rdquo;
          </div>
        ) : (
          <div
            role="region"
            aria-label="Log output"
            aria-live="polite"
            aria-relevant="additions"
            ref={scrollRef}
            className="size-full overflow-y-auto"
            onScroll={handleScroll}
            tabIndex={0}
          >
            <ol
              aria-label="Log entries"
              style={{position: 'relative', height: `${totalSize}px`}}
            >
              {virtualItems.map(virtualRow => {
                const visible = filteredEntries[virtualRow.index];
                if (!visible) return null;
                const {entry, originalIndex} = visible;
                return (
                  <li
                    key={virtualRow.key}
                    data-index={virtualRow.index}
                    ref={virtualizer.measureElement}
                    style={{
                      position: 'absolute',
                      top: 0,
                      left: 0,
                      width: '100%',
                      transform: `translateY(${virtualRow.start}px)`,
                    }}
                  >
                    <LogRow
                      entry={entry}
                      index={originalIndex}
                      showTimestamps={showTimestamps}
                      showLineNumbers={showLineNumbers}
                      searchValue={activeSearch}
                    />
                  </li>
                );
              })}
            </ol>
          </div>
        )}

        {/* "Resume" pill — shown when user has paused follow */}
        {!follow && !isEmpty && !hasNoMatches && (
          <button
            type="button"
            onClick={handleFollowToggle}
            className={cn(
              'absolute right-5 bottom-3 z-raised flex items-center gap-1.5',
              'rounded-full border border-border bg-background px-3 py-1',
              'text-xs font-label shadow-raised transition-colors hover:bg-muted'
            )}
            aria-label="Resume scrolling to latest logs"
          >
            <ArrowDownIcon className="size-3" aria-hidden="true" />
            Resume
          </button>
        )}
      </div>
    </div>
  );
}

export {LogViewer, LogRow};
export type {LogEntry, LogViewerProps, LogLevel, LogStream, LogFilterFn};
