// @docker/trident/src/components/markdown.tsx
import type {AnchorHTMLAttributes} from 'react';

import {cn} from '@/lib/utils';

import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';

type MarkdownProps = {
  /** Markdown string to render. */
  content: string;
  /** Extra classes merged onto the rendered wrapper. */
  className?: string;
};

/**
 * Detect external URLs — anything absolute with an http(s) scheme or
 * protocol-relative (`//example.com`) is treated as external.
 */
function isExternalHref(href: string | undefined): boolean {
  if (!href) return false;
  return /^(https?:)?\/\//i.test(href);
}

function MarkdownLink({
  href,
  children,
  ...props
}: AnchorHTMLAttributes<HTMLAnchorElement>) {
  const external = isExternalHref(href);
  return (
    <a
      href={href}
      target={external ? '_blank' : undefined}
      rel={external ? 'noopener noreferrer' : undefined}
      {...props}
    >
      {children}
    </a>
  );
}

function Markdown({content, className}: MarkdownProps) {
  return (
    <div
      data-slot="markdown"
      className={cn(
        // Layout
        'max-w-none text-sm text-foreground',
        // Headings — real descendant selectors, not dead prose modifiers
        '[&_h1]:mt-dds-600 [&_h1]:mb-dds-400 [&_h1]:text-2xl [&_h1]:font-heading [&_h1]:tracking-heading [&_h1]:text-foreground',
        '[&_h2]:mt-dds-500 [&_h2]:mb-dds-300 [&_h2]:text-xl [&_h2]:font-heading [&_h2]:tracking-heading [&_h2]:text-foreground',
        '[&_h3]:mt-dds-400 [&_h3]:mb-dds-200 [&_h3]:text-lg [&_h3]:font-heading [&_h3]:tracking-heading [&_h3]:text-foreground',
        '[&_h4]:mt-dds-400 [&_h4]:mb-dds-200 [&_h4]:text-base [&_h4]:font-heading [&_h4]:tracking-heading [&_h4]:text-foreground',
        '[&_h5]:mt-dds-300 [&_h5]:mb-dds-100 [&_h5]:text-sm [&_h5]:font-heading [&_h5]:tracking-heading [&_h5]:text-foreground',
        '[&_h6]:mt-dds-300 [&_h6]:mb-dds-100 [&_h6]:text-sm [&_h6]:font-heading [&_h6]:tracking-heading [&_h6]:text-muted-foreground',
        // First heading has no top margin
        '[&_>_:first-child]:mt-0',
        // Paragraphs + inline text
        '[&_p]:my-dds-300 [&_p]:leading-reading [&_p]:text-foreground',
        '[&_strong]:font-medium [&_strong]:text-foreground',
        '[&_em]:italic',
        // Links — WCAG 1.4.1 / 2.4.4: color + underline + focus ring
        '[&_a]:text-primary [&_a]:underline [&_a]:decoration-1 [&_a]:underline-offset-2',
        '[&_a:hover]:text-primary [&_a:hover]:decoration-2',
        '[&_a:focus-visible]:rounded-sm [&_a:focus-visible]:ring-3 [&_a:focus-visible]:ring-ring [&_a:focus-visible]:outline-none',
        // Lists
        '[&_ul]:my-dds-300 [&_ul]:list-disc [&_ul]:pl-dds-500',
        '[&_ol]:my-dds-300 [&_ol]:list-decimal [&_ol]:pl-dds-500',
        '[&_li]:my-dds-100 [&_li]:leading-reading',
        // Inline code
        '[&_code]:rounded [&_code]:bg-muted [&_code]:px-1.5 [&_code]:py-0.5 [&_code]:font-mono [&_code]:text-[0.9em] [&_code]:text-foreground',
        // Fenced code blocks — strip inline code styling, apply block styling
        '[&_pre]:my-dds-400 [&_pre]:overflow-x-auto [&_pre]:rounded [&_pre]:bg-muted [&_pre]:p-dds-400 [&_pre]:font-mono [&_pre]:text-sm [&_pre]:text-foreground',
        '[&_pre_code]:block [&_pre_code]:bg-transparent [&_pre_code]:p-0',
        // Blockquote
        '[&_blockquote]:my-dds-400 [&_blockquote]:border-l-2 [&_blockquote]:border-border [&_blockquote]:pl-dds-400 [&_blockquote]:text-muted-foreground [&_blockquote]:italic',
        // Horizontal rule
        '[&_hr]:my-dds-500 [&_hr]:border-border',
        // Tables (GFM)
        '[&_table]:my-dds-400 [&_table]:w-full [&_table]:border-collapse [&_table]:text-sm',
        '[&_thead]:border-b [&_thead]:border-border',
        '[&_th]:px-dds-300 [&_th]:py-dds-200 [&_th]:text-left [&_th]:font-medium [&_th]:text-foreground',
        '[&_td]:border-b [&_td]:border-border [&_td]:px-dds-300 [&_td]:py-dds-200 [&_td]:text-foreground',
        '[&_tr:last-child_td]:border-b-0',
        // Images
        '[&_img]:my-dds-400 [&_img]:max-w-full [&_img]:rounded',
        className
      )}
    >
      <ReactMarkdown
        remarkPlugins={[remarkGfm]}
        components={{
          a: MarkdownLink,
        }}
      >
        {content}
      </ReactMarkdown>
    </div>
  );
}

export {Markdown};
export type {MarkdownProps};
