// @docker/trident/src/components/nav-rail.tsx
import {cn} from '@/lib/utils';

type NavRailItem = {
  id: string;
  icon: React.ElementType;
  label: string;
};

type NavRailProps = {
  items: NavRailItem[];
  activeId?: string;
  onSelect: (id: string) => void;
  /** Optional bottom-anchored action (e.g. Settings button). */
  footer?: React.ReactNode;
  className?: string;
};

/**
 * Vertical icon rail for primary app navigation. 48px wide, icon-only with
 * tooltip titles. Active item gets primary-wash fill; inactive items are
 * muted with hover state.
 *
 * Decoupled from any store — accepts items + activeId as props so it can be
 * embedded in Storybook and used in any shell context.
 *
 * @example
 * <NavRail
 *   items={[{id: 'app', icon: Layers, label: 'Applications'}]}
 *   activeId="app"
 *   onSelect={setActiveTab}
 *   footer={<NavRailSettingsButton onClick={openSettings} />}
 * />
 */
function NavRail({items, activeId, onSelect, footer, className}: NavRailProps) {
  return (
    <nav
      className={cn(
        'flex w-12 shrink-0 flex-col items-center gap-1 border-r border-border bg-card py-2',
        className
      )}
      style={{boxShadow: 'inset 0 1px 0 var(--highlight)'}}
    >
      {/* Docker whale mark at top */}
      <div className="mb-2 flex size-8 items-center justify-center">
        <NavRailWhaleIcon className="size-5 text-muted-foreground" />
      </div>

      {/* Primary nav items */}
      {items.map(item => (
        <button
          key={item.id}
          type="button"
          onClick={() => onSelect(item.id)}
          title={item.label}
          aria-label={item.label}
          aria-current={item.id === activeId ? 'page' : undefined}
          className={cn(
            'flex size-8 items-center justify-center rounded-[var(--dds-radius-default)] transition-colors',
            item.id === activeId
              ? 'bg-primary-wash text-primary'
              : 'text-muted-foreground hover:bg-hover-overlay hover:text-foreground'
          )}
        >
          <item.icon className="size-4" aria-hidden />
        </button>
      ))}

      <div className="flex-1" />

      {footer}
    </nav>
  );
}

/**
 * Docker whale SVG mark for the nav rail header. Matches the Docker visual
 * identity spec path data.
 */
function NavRailWhaleIcon({className}: {className?: string}) {
  return (
    <svg
      viewBox="0 0 24 24"
      fill="currentColor"
      className={className}
      xmlns="http://www.w3.org/2000/svg"
      aria-label="Docker"
      role="img"
    >
      <path d="M13.983 11.078h2.119a.186.186 0 00.186-.185V9.006a.186.186 0 00-.186-.186h-2.119a.185.185 0 00-.185.185v1.888c0 .102.083.185.185.185m-2.954-5.43h2.118a.186.186 0 00.186-.186V3.574a.186.186 0 00-.186-.185h-2.118a.185.185 0 00-.185.185v1.888c0 .102.082.185.185.185m0 2.716h2.118a.187.187 0 00.186-.186V6.29a.186.186 0 00-.186-.185h-2.118a.185.185 0 00-.185.185v1.887c0 .102.082.185.185.186m-2.93 0h2.12a.186.186 0 00.184-.186V6.29a.185.185 0 00-.185-.185H8.1a.185.185 0 00-.185.185v1.887c0 .102.083.185.185.186m-2.964 0h2.119a.186.186 0 00.185-.186V6.29a.185.185 0 00-.185-.185H5.136a.186.186 0 00-.186.185v1.887c0 .102.084.185.186.186m5.893 2.715h2.118a.186.186 0 00.186-.185V9.006a.186.186 0 00-.186-.186h-2.118a.185.185 0 00-.185.185v1.888c0 .102.082.185.185.185m-2.93 0h2.12a.185.185 0 00.184-.185V9.006a.185.185 0 00-.184-.186h-2.12a.185.185 0 00-.184.185v1.888c0 .102.083.185.185.185m-2.964 0h2.119a.185.185 0 00.185-.185V9.006a.185.185 0 00-.184-.186h-2.12a.186.186 0 00-.186.185v1.888c0 .102.084.185.186.185m-2.92 0h2.12a.186.186 0 00.184-.185V9.006a.185.185 0 00-.184-.186h-2.12a.185.185 0 00-.185.185v1.888c0 .102.083.185.185.185M23.763 9.89c-.065-.051-.672-.51-1.954-.51-.338.001-.676.03-1.01.087-.248-1.7-1.653-2.53-1.716-2.566l-.344-.199-.226.327c-.284.438-.49.922-.612 1.43-.23.97-.09 1.882.403 2.661-.595.332-1.55.413-1.744.42H.751a.751.751 0 00-.75.748 11.376 11.376 0 00.692 4.062c.545 1.428 1.355 2.48 2.41 3.124 1.18.723 3.1 1.137 5.275 1.137.983.003 1.963-.086 2.93-.266a12.248 12.248 0 003.823-1.389c.98-.567 1.86-1.288 2.61-2.136 1.252-1.418 1.998-2.997 2.553-4.4h.221c1.372 0 2.215-.549 2.68-1.009.309-.293.55-.65.707-1.046l.098-.288Z" />
    </svg>
  );
}

export {NavRail, NavRailWhaleIcon};
export type {NavRailItem, NavRailProps};
