// @docker/trident/src/components/node-popover.tsx
import {cn} from '@/lib/utils';

import {createPortal} from 'react-dom';

const NODE_POPOVER_WIDTH = 220;
const NODE_POPOVER_GAP = 8;

type AnchorRect = {
  top: number;
  left: number;
  width: number;
  height: number;
};

type BadgeTone = 'ok' | 'err' | 'warn' | 'loading';

type NodePopoverProps = {
  anchorRect: AnchorRect;
  onClose?: () => void;
  children: React.ReactNode;
  'aria-label'?: string;
};

/**
 * Portal-based floating detail card anchored to a graph node's viewport
 * coordinates. Renders via createPortal above the canvas transform layer.
 *
 * Radix Popover cannot be used here — it expects a trigger element in the
 * DOM, which breaks when the trigger moves via CSS transform (pan/zoom canvas).
 * Instead, position: fixed + anchorRect gives reliable placement.
 *
 * Floats to the right of the node by default; falls back to left if the
 * right side would overflow the viewport.
 *
 * @example
 * <NodePopover
 *   anchorRect={{ top: 200, left: 500, width: 140, height: 88 }}
 *   onClose={() => clearSelection()}
 * >
 *   <NodePopover.Header>
 *     <NodePopover.Title>app-ui</NodePopover.Title>
 *     <NodePopover.Badge tone="ok">up</NodePopover.Badge>
 *   </NodePopover.Header>
 *   <NodePopover.Section>
 *     <NodePopover.Metric label="cpu" value="6.2%" />
 *     <NodePopover.Metric label="mem" value="235 MB" />
 *   </NodePopover.Section>
 *   <NodePopover.Section label="calls">
 *     <NodePopover.Link>billing-orchestrator-api</NodePopover.Link>
 *   </NodePopover.Section>
 *   <NodePopover.Actions>
 *     <Button size="sm" variant="ghost">Logs</Button>
 *   </NodePopover.Actions>
 * </NodePopover>
 */
function NodePopover({
  anchorRect,
  onClose,
  children,
  'aria-label': ariaLabel,
}: NodePopoverProps) {
  const viewportWidth =
    typeof window !== 'undefined' ? window.innerWidth : 1440;
  const fitsRight =
    anchorRect.left + anchorRect.width + NODE_POPOVER_GAP + NODE_POPOVER_WIDTH <
    viewportWidth;

  const style: React.CSSProperties = fitsRight
    ? {
        position: 'fixed',
        top: anchorRect.top,
        left: anchorRect.left + anchorRect.width + NODE_POPOVER_GAP,
        width: NODE_POPOVER_WIDTH,
        zIndex: 50,
      }
    : {
        position: 'fixed',
        top: anchorRect.top,
        left: anchorRect.left - NODE_POPOVER_WIDTH - NODE_POPOVER_GAP,
        width: NODE_POPOVER_WIDTH,
        zIndex: 50,
      };

  return createPortal(
    <div
      role="dialog"
      aria-modal="false"
      aria-label={ariaLabel ?? 'Node details'}
      style={{...style, boxShadow: 'var(--dds-material-floating-shadow)'}}
      className="rounded-[var(--dds-radius-default)] border border-border bg-popover p-3"
      onClick={e => e.stopPropagation()}
      onKeyDown={e => {
        if (e.key === 'Escape') onClose?.();
      }}
    >
      {children}
    </div>,
    document.body
  );
}

/* ── Sub-components ────────────────────────────────────────────────────────── */

function NodePopoverHeader({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="node-popover-header"
      className={cn('flex items-center gap-2', className)}
      {...props}
    />
  );
}

function NodePopoverTitle({className, ...props}: React.ComponentProps<'span'>) {
  return (
    <span
      data-slot="node-popover-title"
      className={cn(
        'text-[13px] font-label tracking-[-0.008em] text-foreground',
        className
      )}
      {...props}
    />
  );
}

const BADGE_TONE_CLASS: Record<BadgeTone, string> = {
  ok: 'text-success',
  err: 'text-destructive',
  warn: 'text-warning',
  loading: 'text-primary',
};

function NodePopoverBadge({
  tone = 'loading',
  className,
  ...props
}: React.ComponentProps<'span'> & {tone?: BadgeTone}) {
  return (
    <span
      data-slot="node-popover-badge"
      className={cn('font-mono text-[12px]', BADGE_TONE_CLASS[tone], className)}
      {...props}
    />
  );
}

function NodePopoverSection({
  label,
  className,
  children,
  ...props
}: React.ComponentProps<'div'> & {label?: string}) {
  return (
    <div
      data-slot="node-popover-section"
      className={cn('mt-2.5 border-t border-border-subtle pt-2', className)}
      {...props}
    >
      {label && (
        <span className="text-[10px] font-label tracking-[0.06em] text-muted-foreground uppercase">
          {label}
        </span>
      )}
      <div className={cn('flex flex-col gap-0.5', label && 'mt-1')}>
        {children}
      </div>
    </div>
  );
}

function NodePopoverMetric({label, value}: {label: string; value: string}) {
  return (
    <div className="flex items-center gap-3 font-mono text-[12px] text-muted-foreground">
      <span className="text-muted-foreground">{label}</span>
      <span>{value}</span>
    </div>
  );
}

function NodePopoverLink({className, ...props}: React.ComponentProps<'span'>) {
  return (
    <span
      data-slot="node-popover-link"
      className={cn('font-mono text-[12px] text-muted-foreground', className)}
      {...props}
    />
  );
}

function NodePopoverActions({
  className,
  ...props
}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="node-popover-actions"
      className={cn(
        'mt-2.5 flex items-center gap-1.5 border-t border-border-subtle pt-2',
        className
      )}
      {...props}
    />
  );
}

/* ── Compound export ───────────────────────────────────────────────────────── */

const NodePopoverCompound = Object.assign(NodePopover, {
  Header: NodePopoverHeader,
  Title: NodePopoverTitle,
  Badge: NodePopoverBadge,
  Section: NodePopoverSection,
  Metric: NodePopoverMetric,
  Link: NodePopoverLink,
  Actions: NodePopoverActions,
});

export {NodePopoverCompound as NodePopover};
export type {NodePopoverProps, AnchorRect, BadgeTone};
