// @docker/trident/src/components/notification-snackbar.tsx
'use client';

import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import {X} from 'lucide-react';
import * as React from 'react';

type NotificationSnackbarProps = {
  /** Controls visibility */
  open: boolean;
  /** Icon rendered on the left */
  icon: React.ReactNode;
  /** Main notification text (max 3 lines) */
  message: string;
  /** Optional bold title above message (max 1 line) */
  messageTitle?: string;
  /** If true, never auto-dismiss. @default false */
  persist?: boolean;
  /** If true, hide the dismiss ✕ button. @default false */
  hideDismissAction?: boolean;
  /** Optional primary action button */
  action?: {text: string; fn: () => void; label?: string};
  /** Called when the notification should close */
  onClose: () => void;
  /**
   * Controls the ARIA live region urgency.
   * - `"polite"` → `role="status"` + `aria-live="polite"` (informational, non-interrupting)
   * - `"assertive"` → `role="alert"` + `aria-live="assertive"` (critical, interrupts AT)
   * @default "assertive"
   */
  urgency?: 'polite' | 'assertive';
  className?: string;
};

const AUTO_DISMISS_MS = 10_000;

function NotificationSnackbar({
  open,
  icon,
  message,
  messageTitle,
  persist = false,
  hideDismissAction = false,
  action,
  onClose,
  urgency = 'assertive',
  className,
}: NotificationSnackbarProps) {
  React.useEffect(() => {
    if (!open || persist) return;
    const timer = setTimeout(onClose, AUTO_DISMISS_MS);
    return () => clearTimeout(timer);
  }, [open, persist, onClose]);

  if (!open) return null;

  const role = urgency === 'polite' ? 'status' : 'alert';
  const ariaLive = urgency === 'polite' ? 'polite' : 'assertive';

  return (
    <div
      role={role}
      aria-live={ariaLive}
      aria-atomic="true"
      data-slot="notification-snackbar"
      className={cn(
        'flex w-full max-w-sm items-start gap-3 rounded border border-border bg-popover p-4 shadow-floating',
        className
      )}
    >
      <div aria-hidden="true" className="mt-0.5 shrink-0 text-foreground">
        {icon}
      </div>
      <div className="min-w-0 flex-1">
        {messageTitle ? (
          <p className="line-clamp-1 text-sm font-heading text-foreground">
            {messageTitle}
          </p>
        ) : null}
        <p className="line-clamp-3 text-sm text-muted-foreground">{message}</p>
        {action ? (
          <Button
            variant="link"
            size="sm"
            aria-label={action.label ?? action.text}
            className="mt-1 h-auto p-0 text-sm"
            onClick={action.fn}
          >
            {action.text}
          </Button>
        ) : null}
      </div>
      {hideDismissAction ? null : (
        <Button
          type="button"
          variant="ghost"
          size="icon"
          aria-label="Dismiss notification"
          className="h-6 w-6 shrink-0"
          onClick={onClose}
        >
          <X aria-hidden="true" className="h-3.5 w-3.5" />
        </Button>
      )}
    </div>
  );
}

export {NotificationSnackbar};
export type {NotificationSnackbarProps};
