// @docker/trident/src/components/provenance-indicator.tsx
// provenance-indicator.tsx

import type * as React from 'react';

import {cn} from '@/lib/utils';

import {Check, HelpCircle, X} from 'lucide-react';

type ProvenanceIndicatorStatus = 'signed' | 'unsigned' | 'unknown';

type ProvenanceIndicatorProps = React.ComponentProps<'span'> & {
  status: ProvenanceIndicatorStatus;
};

const STATUS_CONFIG: Record<
  ProvenanceIndicatorStatus,
  {Icon: React.ElementType; label: string; colorClass: string}
> = {
  signed: {
    Icon: Check,
    label: 'Provenance signed',
    colorClass: 'text-success',
  },
  unsigned: {
    Icon: X,
    label: 'Provenance unsigned',
    colorClass: 'text-muted-foreground',
  },
  unknown: {
    Icon: HelpCircle,
    label: 'Provenance unknown',
    colorClass: 'text-muted-foreground',
  },
};

function ProvenanceIndicator({
  status,
  className,
  ...props
}: ProvenanceIndicatorProps) {
  const {Icon, label, colorClass} = STATUS_CONFIG[status];

  return (
    <span
      data-slot="provenance-indicator"
      role="img"
      aria-label={label}
      className={cn(
        'inline-flex items-center gap-1 font-mono text-xs',
        colorClass,
        className
      )}
      {...props}
    >
      <Icon className="size-3" aria-hidden />
      Provenance
    </span>
  );
}

export {ProvenanceIndicator};
export type {ProvenanceIndicatorProps, ProvenanceIndicatorStatus};
