// @docker/trident/src/components/rating.tsx
'use client';

import type * as React from 'react';

import {cn} from '@/lib/utils';

import {Star} from 'lucide-react';
import {useCallback, useId, useMemo, useRef, useState} from 'react';

type RatingSize = 'sm' | 'default' | 'lg';

type RatingProps = {
  /**
   * The current rating value.
   *
   * @default 0
   */
  value?: number;
  /**
   * Callback fired when the rating value changes.
   * Not called when `readOnly` is true.
   */
  onChange?: (value: number) => void;
  /**
   * The maximum number of stars.
   *
   * @default 5
   */
  max?: number;
  /**
   * Whether the rating is read-only (non-interactive).
   *
   * @default false
   */
  readOnly?: boolean;
  /**
   * The size of the star icons.
   *
   * @default "default"
   */
  size?: RatingSize;
  /**
   * The minimum increment of the rating value.
   * Set to `0.5` to allow half-star ratings.
   *
   * @default 1
   */
  precision?: 0.5 | 1;
  /**
   * Additional class names to apply to the container.
   */
  className?: string;
  /**
   * Accessible label for the rating group.
   *
   * @default "Rating"
   */
  'aria-label'?: string;
};

const sizeClasses: Record<RatingSize, string> = {
  sm: 'size-4',
  default: 'size-5',
  lg: 'size-7',
};

const gapClasses: Record<RatingSize, string> = {
  sm: 'gap-0.5',
  default: 'gap-1',
  lg: 'gap-1.5',
};

/**
 * Displays an interactive or read-only star rating.
 *
 * Replaces `@mui/material/Rating` with a token-driven, accessible implementation
 * built on DDS semantic colors and lucide-react icons.
 *
 * ### Component Props
 *
 * | Prop        | Type                     | Default     |
 * | :---------- | :----------------------- | :---------- |
 * | _value_     | `number`                 | `0`         |
 * | _onChange_   | `(value: number) => void` | -          |
 * | _max_       | `number`                 | `5`         |
 * | _readOnly_  | `boolean`                | `false`     |
 * | _size_      | `"sm" \| "default" \| "lg"` | `"default"` |
 * | _precision_ | `0.5 \| 1`               | `1`         |
 *
 * ### Example Usage
 *
 * **Interactive**
 * ```tsx
 * const [value, setValue] = useState(3);
 * <Rating value={value} onChange={setValue} />
 * ```
 *
 * **Read-only with half stars**
 * ```tsx
 * <Rating value={3.5} precision={0.5} readOnly />
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.4.3 Contrast (Minimum) — star fill colors use DDS `warning` and `muted-foreground` tokens meeting 4.5:1.
 * @wcag 2.1.1 Keyboard — arrow keys change value; tab moves focus to/from the component.
 * @wcag 2.4.7 Focus Visible — focus ring shown via `focus-visible:ring-3`.
 * @wcag 4.1.2 Name, Role, Value — uses `role="slider"` with `aria-valuenow`, `aria-valuemin`, `aria-valuemax`, and `aria-label`.
 */
function Rating({
  value = 0,
  onChange,
  max = 5,
  readOnly = false,
  size = 'default',
  precision = 1,
  className,
  'aria-label': ariaLabel = 'Rating',
}: RatingProps) {
  const [hoverValue, setHoverValue] = useState<number | null>(null);
  const containerRef = useRef<HTMLDivElement>(null);
  const id = useId();

  const displayValue = hoverValue ?? value;

  const roundToStep = useCallback(
    (v: number): number => {
      const clamped = Math.max(0, Math.min(v, max));
      return Math.round(clamped / precision) * precision;
    },
    [max, precision]
  );

  const resolveStarFromEvent = useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      const target =
        event.target instanceof Element
          ? event.target.closest('[data-star-index]')
          : null;
      if (!target) return null;
      const starIndex = Number(target.getAttribute('data-star-index'));
      if (precision === 0.5) {
        const rect = target.getBoundingClientRect();
        const isLeftHalf = event.clientX - rect.left < rect.width / 2;
        return isLeftHalf ? starIndex + 0.5 : starIndex + 1;
      }
      return starIndex + 1;
    },
    [precision]
  );

  const handleClick = useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      if (readOnly) return;
      const newValue = resolveStarFromEvent(event);
      if (newValue !== null) onChange?.(newValue === value ? 0 : newValue);
    },
    [readOnly, resolveStarFromEvent, onChange, value]
  );

  const handleMouseMove = useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      if (readOnly) return;
      const newValue = resolveStarFromEvent(event);
      if (newValue !== null) setHoverValue(newValue);
    },
    [readOnly, resolveStarFromEvent]
  );

  const handleMouseLeave = useCallback(() => {
    if (!readOnly) {
      setHoverValue(null);
    }
  }, [readOnly]);

  const handleKeyDown = useCallback(
    (event: React.KeyboardEvent<HTMLDivElement>) => {
      if (readOnly) return;

      let newValue = value;

      switch (event.key) {
        case 'ArrowRight':
        case 'ArrowUp':
          event.preventDefault();
          newValue = roundToStep(Math.min(value + precision, max));
          break;
        case 'ArrowLeft':
        case 'ArrowDown':
          event.preventDefault();
          newValue = roundToStep(Math.max(value - precision, 0));
          break;
        case 'Home':
          event.preventDefault();
          newValue = 0;
          break;
        case 'End':
          event.preventDefault();
          newValue = max;
          break;
        default:
          return;
      }

      onChange?.(newValue);
    },
    [readOnly, value, precision, max, roundToStep, onChange]
  );

  const getStarFill = (starIndex: number): 'full' | 'half' | 'empty' => {
    const starPosition = starIndex + 1;
    if (displayValue >= starPosition) return 'full';
    if (displayValue >= starPosition - 0.5 && precision === 0.5) return 'half';
    return 'empty';
  };

  const interactive = !readOnly;
  const starValues = useMemo(
    () => Array.from({length: max}, (_, i) => i),
    [max]
  );

  return (
    <div
      ref={containerRef}
      data-slot="rating"
      {...(readOnly
        ? {
            role: 'img' as const,
            'aria-label': `${ariaLabel}: ${value} out of ${max} stars`,
            tabIndex: -1,
          }
        : {
            role: 'slider' as const,
            'aria-label': ariaLabel,
            'aria-valuemin': 0,
            'aria-valuemax': max,
            'aria-valuenow': value,
            'aria-valuetext': `${value} out of ${max} stars`,
            tabIndex: 0,
            onClick: handleClick,
            onMouseMove: handleMouseMove,
            onMouseLeave: handleMouseLeave,
            onKeyDown: handleKeyDown,
          })}
      className={cn(
        'inline-flex items-center',
        gapClasses[size],
        interactive &&
          'cursor-pointer rounded outline-none focus-visible:ring-3 focus-visible:ring-ring',
        readOnly && 'pointer-events-none',
        className
      )}
    >
      {starValues.map(i => (
        <RatingStar
          key={`${id}-star-${i}`}
          index={i}
          fill={getStarFill(i)}
          size={size}
        />
      ))}
    </div>
  );
}

type RatingStarProps = {
  index: number;
  fill: 'full' | 'half' | 'empty';
  size: RatingSize;
};

function RatingStar({index, fill, size}: RatingStarProps) {
  const iconClass = sizeClasses[size];

  if (fill === 'half') {
    return (
      <span
        data-star-index={index}
        role="presentation"
        className="relative inline-flex shrink-0"
      >
        {/* Background empty star */}
        <Star
          className={cn(iconClass, 'text-muted-foreground')}
          fill="none"
          strokeWidth={1.5}
        />
        {/* Foreground half-filled star using clip-path */}
        <Star
          className={cn(iconClass, 'absolute inset-0 text-warning')}
          fill="currentColor"
          strokeWidth={1.5}
          style={{clipPath: 'inset(0 50% 0 0)'}}
        />
      </span>
    );
  }

  return (
    <span
      data-star-index={index}
      role="presentation"
      className="inline-flex shrink-0"
    >
      <Star
        className={cn(
          iconClass,
          fill === 'full' ? 'text-warning' : 'text-muted-foreground'
        )}
        fill={fill === 'full' ? 'currentColor' : 'none'}
        strokeWidth={1.5}
      />
    </span>
  );
}

export {Rating};
export type {RatingProps, RatingSize};
