// @docker/trident/src/components/repo-card.tsx
// repo-card.tsx
import type * as React from 'react';

import {HealthGrade} from '@/components/health-grade';
import {TrustBadge, type TrustBadgeProps} from '@/components/trust-badge';
import {Badge} from '@/components/ui/badge';
import {cn} from '@/lib/utils';

type RepoCardProps = {
  /** Namespace + name shown as the card title */
  namespace: string;
  name: string;
  /** Short description, clamped to 2 lines */
  description?: string;
  /** Trust tier for the TrustBadge */
  tier?: TrustBadgeProps['tier'];
  /** Formatted pull count string e.g. "1.8B pulls" */
  pullCount?: string;
  /** Scout health grade — omit to hide */
  healthGrade?: 'A' | 'B' | 'C' | 'D' | 'F';
  /** Optional slot for VulnerabilityChips or other security signals */
  securitySlot?: React.ReactNode;
  /** Click handler or href — wrap in <a> externally if needed */
  onClick?: () => void;
  className?: string;
  /** URL for image thumbnail — shown instead of initials placeholder */
  imageUrl?: string;
  /** Alt text for the image thumbnail */
  imageAlt?: string;
  /** Star/download count string e.g. "1.2k" */
  starCount?: string;
  /** Card variant badge — 'hardened' | 'marketplace' */
  variant?: 'hardened' | 'marketplace';
};

function RepoCard({
  namespace,
  name,
  description,
  tier = 'community',
  pullCount,
  healthGrade,
  securitySlot,
  onClick,
  className,
  imageUrl,
  imageAlt: _imageAlt,
  starCount,
  variant,
}: RepoCardProps) {
  const baseClassName = cn(
    'flex w-full items-start gap-3 rounded border border-border bg-card p-4',
    className
  );

  const body = (
    <>
      {/* Icon slot — decorative; accessible name comes from the visible text */}
      <div
        data-slot="repo-card-icon"
        aria-hidden="true"
        className="flex h-10 w-10 shrink-0 items-center justify-center overflow-hidden rounded border border-border bg-accent text-sm font-bold text-accent-foreground uppercase"
      >
        {imageUrl ? (
          <img src={imageUrl} alt="" className="h-full w-full object-cover" />
        ) : (
          name.slice(0, 2)
        )}
      </div>

      <div className="min-w-0 flex-1">
        {/* Name + trust badge */}
        <div className="flex flex-wrap items-center gap-2">
          <span
            data-slot="repo-card-name"
            className="text-sm leading-tight font-heading text-foreground"
          >
            {namespace}/{name}
          </span>
          <TrustBadge tier={tier} />
        </div>

        {/* Description */}
        {description ? (
          <p
            data-slot="repo-card-description"
            className="mt-0.5 line-clamp-2 text-xs leading-relaxed text-muted-foreground high-contrast:text-foreground"
          >
            {description}
          </p>
        ) : null}

        {/* Variant / star row */}
        {(variant ?? starCount) ? (
          <div className="mt-1.5 flex items-center gap-2">
            {variant ? (
              <Badge variant="secondary" size="sm">
                {variant === 'hardened' ? 'Hardened' : 'Marketplace'}
              </Badge>
            ) : null}
            {starCount ? (
              <span
                role="img"
                aria-label={`${starCount} stars`}
                className="flex items-center gap-1 text-xs text-muted-foreground high-contrast:text-foreground"
              >
                <svg
                  aria-hidden="true"
                  className="h-3 w-3 fill-current"
                  viewBox="0 0 16 16"
                >
                  <path d="M8 .25a.75.75 0 0 1 .673.418l1.882 3.815 4.21.612a.75.75 0 0 1 .416 1.279l-3.046 2.97.719 4.192a.75.75 0 0 1-1.088.791L8 12.347l-3.766 1.98a.75.75 0 0 1-1.088-.79l.72-4.194L.818 6.374a.75.75 0 0 1 .416-1.28l4.21-.611L7.327.668A.75.75 0 0 1 8 .25Z" />
                </svg>
                <span aria-hidden="true">{starCount}</span>
              </span>
            ) : null}
          </div>
        ) : null}

        {/* Signals row */}
        <div
          data-slot="repo-card-signals"
          className="mt-2 flex flex-wrap items-center gap-3"
        >
          {pullCount ? (
            <span className="text-xs text-muted-foreground tabular-nums high-contrast:text-foreground">
              {pullCount}
            </span>
          ) : null}
          {healthGrade ? (
            <HealthGrade grade={healthGrade} className="h-6 w-6 text-xs" />
          ) : null}
          {securitySlot}
        </div>
      </div>
    </>
  );

  if (onClick) {
    return (
      <button
        type="button"
        data-slot="repo-card"
        aria-label={`${namespace}/${name} repository`}
        className={cn(
          baseClassName,
          'cursor-pointer text-left transition-all ease-default hover:border-dds-blue-outlined-border hover:shadow-raised'
        )}
        onClick={onClick}
      >
        {body}
      </button>
    );
  }

  return (
    <div data-slot="repo-card" className={baseClassName}>
      {body}
    </div>
  );
}

export {RepoCard};
export type {RepoCardProps};
