// @docker/trident/src/components/sbom-indicator.tsx
// sbom-indicator.tsx

import type * as React from 'react';

import {cn} from '@/lib/utils';

import {Check, Minus, X} from 'lucide-react';

type SbomIndicatorStatus = 'present' | 'absent' | 'partial';

type SbomIndicatorProps = React.ComponentProps<'span'> & {
  status: SbomIndicatorStatus;
};

const STATUS_CONFIG: Record<
  SbomIndicatorStatus,
  {Icon: React.ElementType; label: string; colorClass: string}
> = {
  present: {Icon: Check, label: 'SBOM present', colorClass: 'text-success'},
  absent: {
    Icon: X,
    label: 'SBOM absent',
    colorClass: 'text-muted-foreground',
  },
  partial: {Icon: Minus, label: 'SBOM partial', colorClass: 'text-warning'},
};

function SbomIndicator({status, className, ...props}: SbomIndicatorProps) {
  const {Icon, label, colorClass} = STATUS_CONFIG[status];

  return (
    <span
      data-slot="sbom-indicator"
      role="img"
      aria-label={label}
      className={cn(
        'inline-flex items-center gap-1 font-mono text-xs',
        colorClass,
        className
      )}
      {...props}
    >
      <Icon className="size-3" aria-hidden />
      SBOM
    </span>
  );
}

export {SbomIndicator};
export type {SbomIndicatorProps, SbomIndicatorStatus};
