// @docker/trident/src/components/selection-card.tsx
import {Badge} from '@/components/ui/badge';
import {cn} from '@/lib/utils';

import {CircleCheck} from 'lucide-react';

type SelectionCardProps = {
  selected: boolean;
  onSelect: () => void;
  title: string;
  description?: string;
  badge?: string;
  disabled?: boolean;
  className?: string;
};

const SelectionCard = ({
  selected,
  onSelect,
  title,
  description,
  badge,
  disabled = false,
  className,
}: SelectionCardProps) => (
  <label
    className={cn(
      'group relative flex w-full cursor-pointer flex-col gap-1 rounded border p-4 text-left transition-all has-[:focus-visible]:ring-2 has-[:focus-visible]:ring-ring has-[:focus-visible]:ring-offset-2',
      selected
        ? 'border-primary bg-accent-wash ring-2 ring-accent-subtle'
        : 'border-border hover:border-primary hover:bg-hover-overlay',
      disabled && 'pointer-events-none opacity-50',
      className
    )}
  >
    <input
      type="radio"
      checked={selected}
      onChange={onSelect}
      onKeyDown={e => {
        if (e.key === 'Enter' && !disabled) {
          e.preventDefault();
          onSelect();
        }
      }}
      disabled={disabled}
      className="sr-only"
    />
    <div className="flex items-center gap-2">
      <span
        className={cn(
          'flex size-4 shrink-0 items-center justify-center rounded-full border',
          selected
            ? 'border-primary bg-primary text-primary-foreground'
            : 'border-input'
        )}
      >
        {selected && <CircleCheck className="size-3" aria-hidden />}
      </span>
      <span className="text-sm font-label">{title}</span>
      {badge && (
        <Badge variant="secondary" size="sm">
          {badge}
        </Badge>
      )}
    </div>
    {description && (
      <p className="pl-6 text-xs text-muted-foreground">{description}</p>
    )}
  </label>
);

export {SelectionCard};
export type {SelectionCardProps};
