// @docker/trident/src/components/session-picker.tsx
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
} from '@/components/ui/command';
import {Popover, PopoverContent, PopoverTrigger} from '@/components/ui/popover';
import {cn} from '@/lib/utils';

import {Check, ChevronDown, Plus} from 'lucide-react';
import {useState} from 'react';

type Session = {
  id: string;
  title: string;
  updatedAt?: string;
  active?: boolean;
};

type SessionPickerProps = {
  sessions: Session[];
  activeId?: string;
  onSelect: (id: string) => void;
  onNew?: () => void;
  className?: string;
};

/**
 * Conversation session switcher for the ConversationLayout header. Shows
 * the active session title; clicking opens a Popover+Command list of all
 * sessions with timestamps. Includes a "New conversation" item at the bottom.
 *
 * Pair with ConversationLayout.Header:
 *
 * @example
 * <ConversationLayout.Header>
 *   <SessionPicker
 *     sessions={sessions}
 *     activeId={activeSessionId}
 *     onSelect={(id) => setActive(id)}
 *     onNew={() => createSession()}
 *   />
 * </ConversationLayout.Header>
 */
function SessionPicker({
  sessions,
  activeId,
  onSelect,
  onNew,
  className,
}: SessionPickerProps) {
  const [open, setOpen] = useState(false);

  const activeSession = sessions.find(s => s.id === activeId);
  const title = activeSession?.title?.trim() || 'New conversation';

  return (
    <Popover open={open} onOpenChange={setOpen}>
      <PopoverTrigger asChild>
        <button
          type="button"
          aria-expanded={open}
          aria-haspopup="listbox"
          aria-label={`Current session: ${title}. Click to switch.`}
          className={cn(
            'flex min-w-0 items-center gap-2 rounded-[var(--dds-radius-sm)] px-1 py-1 transition-colors',
            'hover:bg-hover-overlay',
            'focus-visible:ring-2 focus-visible:ring-primary-subtle focus-visible:outline-none',
            className
          )}
        >
          <span className="truncate text-[13px] font-label tracking-[-0.008em] text-foreground">
            {title}
          </span>
          <ChevronDown
            className="size-3 shrink-0 text-muted-foreground"
            aria-hidden
          />
        </button>
      </PopoverTrigger>
      <PopoverContent
        className="w-72 p-0"
        align="start"
        sideOffset={4}
        aria-label="Sessions"
      >
        <Command>
          <CommandInput placeholder="Search sessions..." />
          <CommandList>
            <CommandEmpty className="py-4 text-center text-[13px] text-muted-foreground">
              No sessions.
            </CommandEmpty>
            <CommandGroup>
              {sessions.map(session => (
                <CommandItem
                  key={session.id}
                  value={session.id}
                  onSelect={() => {
                    onSelect(session.id);
                    setOpen(false);
                  }}
                  className="flex items-center gap-2"
                >
                  <span className="min-w-0 flex-1 truncate text-[13px]">
                    {session.title?.trim() || 'New conversation'}
                  </span>
                  {session.updatedAt && (
                    <span className="shrink-0 font-mono text-[11px] text-muted-foreground">
                      {session.updatedAt}
                    </span>
                  )}
                  {session.id === activeId && (
                    <Check
                      className="size-3 shrink-0 text-primary"
                      aria-hidden
                    />
                  )}
                </CommandItem>
              ))}
            </CommandGroup>
            {onNew && (
              <>
                <div className="mx-2 border-t border-border-subtle" />
                <CommandGroup>
                  <CommandItem
                    onSelect={() => {
                      onNew();
                      setOpen(false);
                    }}
                    className="gap-2 text-primary"
                  >
                    <Plus className="size-3.5 shrink-0" aria-hidden />
                    <span className="text-[13px]">New conversation</span>
                  </CommandItem>
                </CommandGroup>
              </>
            )}
          </CommandList>
        </Command>
      </PopoverContent>
    </Popover>
  );
}

export {SessionPicker};
export type {SessionPickerProps, Session};
