// @docker/trident/src/components/speed-dial.tsx
'use client';

import type {
  SpeedDialDirection,
  SpeedDialPosition,
} from '@/components/speed-dial-context';

import {SpeedDialContext, useSpeedDial} from '@/components/speed-dial-context';
import {Button} from '@/components/ui/button';
import {Tooltip, TooltipContent, TooltipTrigger} from '@/components/ui/tooltip';
import {cn} from '@/lib/utils';

import {Plus} from 'lucide-react';
import * as React from 'react';

/* -------------------------------------------------------------------------------------------------
 * SpeedDial
 * ---------------------------------------------------------------------------------------------- */

type SpeedDialProps = React.ComponentProps<'div'> & {
  /**
   * Fixed position on the viewport.
   *
   * @default "bottom-right"
   */
  position?: SpeedDialPosition;
  /**
   * Direction actions expand from the trigger.
   *
   * @default "up"
   */
  direction?: SpeedDialDirection;
  /**
   * Controlled open state.
   */
  open?: boolean;
  /**
   * Callback when the open state changes.
   */
  onOpenChange?: (open: boolean) => void;
};

/**
 * A floating action button that expands to reveal a set of related actions.
 *
 * ### Component Props
 *
 * | Prop            | Type                                                              | Default          |
 * | :-------------- | :---------------------------------------------------------------- | :--------------- |
 * | _position_      | `"bottom-right" \| "bottom-left" \| "top-right" \| "top-left"`   | `"bottom-right"` |
 * | _direction_     | `"up" \| "down"`                                                 | `"up"`           |
 * | _open_          | `boolean`                                                        | -                |
 * | _onOpenChange_  | `(open: boolean) => void`                                        | -                |
 *
 * ### Example Usage
 *
 * ```tsx
 * <SpeedDial position="bottom-right" direction="up">
 *   <SpeedDialTrigger />
 *   <SpeedDialAction icon={<CopyIcon />} label="Copy" onClick={handleCopy} />
 *   <SpeedDialAction icon={<SaveIcon />} label="Save" onClick={handleSave} />
 * </SpeedDial>
 * ```
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 2.1.1 Keyboard — Enter/Space toggles the menu; Escape closes; Arrow keys navigate actions.
 * @wcag 2.4.7 Focus Visible — focus ring on trigger and actions via `focus-visible:ring-3`.
 * @wcag 4.1.2 Name, Role, Value — trigger has `aria-expanded`; actions are labeled via `aria-label`.
 */
function SpeedDial({
  className,
  position = 'bottom-right',
  direction = 'up',
  open: controlledOpen,
  onOpenChange,
  children,
  ...props
}: SpeedDialProps) {
  const [internalOpen, setInternalOpen] = React.useState(false);
  const isControlled = controlledOpen !== undefined;
  const open = isControlled ? controlledOpen : internalOpen;

  const setOpen = React.useCallback(
    (value: React.SetStateAction<boolean>) => {
      const next = typeof value === 'function' ? value(open) : value;
      if (!isControlled) {
        setInternalOpen(next);
      }
      onOpenChange?.(next);
    },
    [isControlled, open, onOpenChange]
  );

  const [actionCount, setActionCount] = React.useState(0);
  const [focusedIndex, setFocusedIndex] = React.useState(-1);
  const actionRefs = React.useRef<(HTMLButtonElement | null)[]>([]);
  const triggerRef = React.useRef<HTMLButtonElement | null>(null);

  const contextValue = React.useMemo(
    () => ({
      open,
      setOpen,
      position,
      direction,
      actionCount,
      setActionCount,
      focusedIndex,
      setFocusedIndex,
      actionRefs,
      triggerRef,
    }),
    [open, setOpen, position, direction, actionCount, focusedIndex]
  );

  // Close on outside click
  const containerRef = React.useRef<HTMLDivElement | null>(null);

  React.useEffect(() => {
    if (!open) return;

    const handlePointerDown = (e: PointerEvent) => {
      if (
        containerRef.current &&
        e.target instanceof Node &&
        !containerRef.current.contains(e.target)
      ) {
        setOpen(false);
      }
    };

    document.addEventListener('pointerdown', handlePointerDown);
    return () => document.removeEventListener('pointerdown', handlePointerDown);
  }, [open, setOpen]);

  // Close on Escape (global)
  React.useEffect(() => {
    if (!open) return;

    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        setOpen(false);
        triggerRef.current?.focus();
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    return () => document.removeEventListener('keydown', handleKeyDown);
  }, [open, setOpen]);

  const positionClasses: Record<SpeedDialPosition, string> = {
    'bottom-right': 'fixed right-6 bottom-6',
    'bottom-left': 'fixed bottom-6 left-6',
    'top-right': 'fixed top-6 right-6',
    'top-left': 'fixed top-6 left-6',
  };

  const directionClasses: Record<SpeedDialDirection, string> = {
    up: 'flex-col-reverse',
    down: 'flex-col',
  };

  return (
    <SpeedDialContext.Provider value={contextValue}>
      <div
        ref={containerRef}
        data-slot="speed-dial"
        data-state={open ? 'open' : 'closed'}
        data-position={position}
        data-direction={direction}
        role="group"
        className={cn(
          'z-toast flex items-center gap-3',
          positionClasses[position],
          directionClasses[direction],
          className
        )}
        {...props}
      >
        {children}
      </div>
    </SpeedDialContext.Provider>
  );
}

/* -------------------------------------------------------------------------------------------------
 * SpeedDialTrigger
 * ---------------------------------------------------------------------------------------------- */

type SpeedDialTriggerProps = Omit<
  React.ComponentProps<typeof Button>,
  'size' | 'variant'
> & {
  /**
   * Custom icon to display. Defaults to a Plus icon that rotates to X when open.
   */
  icon?: React.ReactNode;
};

/**
 * The floating action button that toggles the SpeedDial menu open and closed.
 *
 * ### Component Props
 *
 * | Prop   | Type              | Default           |
 * | :----- | :---------------- | :---------------- |
 * | _icon_ | `React.ReactNode` | `<Plus />` icon   |
 *
 * ### Example Usage
 *
 * ```tsx
 * <SpeedDialTrigger />
 * <SpeedDialTrigger icon={<PencilIcon />} />
 * ```
 */
function SpeedDialTrigger({className, icon, ...props}: SpeedDialTriggerProps) {
  const {open, setOpen, actionRefs, setFocusedIndex, actionCount, triggerRef} =
    useSpeedDial();

  const handleClick = () => {
    setOpen(prev => !prev);
    setFocusedIndex(-1);
  };

  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === 'ArrowUp' || e.key === 'ArrowDown') {
      e.preventDefault();
      if (!open) {
        setOpen(true);
      }
      // Focus first action
      const nextIndex = e.key === 'ArrowUp' ? actionCount - 1 : 0;
      setFocusedIndex(nextIndex);
      actionRefs.current[nextIndex]?.focus();
    }
  };

  return (
    <Button
      ref={triggerRef}
      data-slot="speed-dial-trigger"
      variant="default"
      size="icon-lg"
      aria-expanded={open}
      aria-haspopup="true"
      aria-label={open ? 'Close actions' : 'Open actions'}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      className={cn(
        'size-14 rounded-full shadow-floating ring-1 ring-border transition-transform duration-normal ease-default focus-visible:ring-3 focus-visible:ring-ring',
        open && 'rotate-45',
        className
      )}
      {...props}
    >
      {icon ?? <Plus className="size-6" />}
    </Button>
  );
}

/* -------------------------------------------------------------------------------------------------
 * SpeedDialAction
 * ---------------------------------------------------------------------------------------------- */

type SpeedDialActionProps = Omit<
  React.ComponentProps<typeof Button>,
  'size' | 'variant' | 'children'
> & {
  /**
   * The icon displayed inside the action button.
   */
  icon: React.ReactNode;
  /**
   * Optional tooltip label shown beside the action.
   */
  label?: string;
  /**
   * The position index of this action. Auto-assigned when omitted.
   */
  index?: number;
};

/**
 * An individual action button within the SpeedDial menu.
 *
 * ### Component Props
 *
 * | Prop    | Type              | Default |
 * | :------ | :---------------- | :------ |
 * | _icon_  | `React.ReactNode` | -       |
 * | _label_ | `string`          | -       |
 *
 * ### Example Usage
 *
 * ```tsx
 * <SpeedDialAction icon={<CopyIcon />} label="Copy" onClick={handleCopy} />
 * ```
 */
function SpeedDialAction({
  className,
  icon,
  label,
  index: indexProp,
  ...props
}: SpeedDialActionProps) {
  const {
    open,
    setOpen,
    position,
    direction,
    actionRefs,
    focusedIndex,
    setFocusedIndex,
    actionCount,
    setActionCount,
    triggerRef,
  } = useSpeedDial();

  // Auto-register action index
  const indexRef = React.useRef<number>(-1);

  React.useEffect(() => {
    if (indexProp !== undefined) {
      indexRef.current = indexProp;
      // Ensure actionCount reflects explicitly-indexed actions too
      setActionCount(prev => Math.max(prev, indexProp + 1));
    } else {
      // Count-based auto indexing
      setActionCount(prev => {
        indexRef.current = prev;
        return prev + 1;
      });
    }
    return () => {
      if (indexProp === undefined) {
        setActionCount(prev => Math.max(0, prev - 1));
      }
    };
  }, [indexProp, setActionCount]);

  const currentIndex = indexProp ?? indexRef.current;

  const registerRef = React.useCallback(
    (el: HTMLButtonElement | null) => {
      actionRefs.current[currentIndex] = el;
    },
    [currentIndex, actionRefs]
  );

  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === 'ArrowUp') {
      e.preventDefault();
      const prev = focusedIndex - 1;
      if (prev >= 0) {
        setFocusedIndex(prev);
        actionRefs.current[prev]?.focus();
      } else {
        triggerRef.current?.focus();
        setFocusedIndex(-1);
      }
    } else if (e.key === 'ArrowDown') {
      e.preventDefault();
      const next = focusedIndex + 1;
      if (next < actionCount) {
        setFocusedIndex(next);
        actionRefs.current[next]?.focus();
      } else {
        triggerRef.current?.focus();
        setFocusedIndex(-1);
      }
    } else if (e.key === 'Enter' || e.key === ' ') {
      // Let the click handler fire, then close
      requestAnimationFrame(() => {
        setOpen(false);
        triggerRef.current?.focus();
      });
    }
  };

  // Determine tooltip side based on position
  const tooltipSide = position.includes('right') ? 'left' : 'right';

  // Stagger delay for animation (in ms)
  const staggerDelay = currentIndex * 50;

  // Animation direction: actions slide in from the opposite direction they expand
  const slideFrom =
    direction === 'up' ? 'translate-y-2 opacity-0' : '-translate-y-2 opacity-0';

  const actionButton = (
    <Button
      ref={registerRef}
      data-slot="speed-dial-action"
      variant="outline"
      size="icon"
      tabIndex={open ? 0 : -1}
      aria-label={label}
      onClick={e => {
        props.onClick?.(e);
        setOpen(false);
        triggerRef.current?.focus();
      }}
      onKeyDown={handleKeyDown}
      className={cn(
        'size-11 rounded-full bg-popover shadow-floating ring-1 ring-border transition-all ease-out',
        open
          ? 'translate-y-0 scale-100 opacity-100'
          : `pointer-events-none ${slideFrom} scale-75`,
        className
      )}
      style={{
        transitionDelay: open ? `${staggerDelay}ms` : '0ms',
        transitionDuration: 'var(--dds-duration-normal)',
      }}
      {...props}
    >
      {icon}
    </Button>
  );

  if (!label) {
    return actionButton;
  }

  return (
    <Tooltip open={open && focusedIndex === currentIndex ? true : undefined}>
      <TooltipTrigger asChild>{actionButton}</TooltipTrigger>
      <TooltipContent side={tooltipSide} sideOffset={8}>
        {label}
      </TooltipContent>
    </Tooltip>
  );
}

export {SpeedDial, SpeedDialTrigger, SpeedDialAction};
