// @docker/trident/src/components/speed-dial-context.tsx
'use client';

import * as React from 'react';

// ─── Types ───────────────────────────────────────────────────────────────────

type SpeedDialPosition =
  | 'bottom-right'
  | 'bottom-left'
  | 'top-right'
  | 'top-left';
type SpeedDialDirection = 'up' | 'down';

type SpeedDialContextValue = {
  open: boolean;
  setOpen: React.Dispatch<React.SetStateAction<boolean>>;
  position: SpeedDialPosition;
  direction: SpeedDialDirection;
  actionCount: number;
  setActionCount: React.Dispatch<React.SetStateAction<number>>;
  focusedIndex: number;
  setFocusedIndex: React.Dispatch<React.SetStateAction<number>>;
  actionRefs: React.MutableRefObject<(HTMLButtonElement | null)[]>;
  triggerRef: React.RefObject<HTMLButtonElement | null>;
};

// ─── Context ─────────────────────────────────────────────────────────────────

const SpeedDialContext = React.createContext<SpeedDialContextValue | null>(
  null
);

function useSpeedDial() {
  const context = React.useContext(SpeedDialContext);
  if (!context) {
    throw new Error(
      'SpeedDial compound components must be used within <SpeedDial>.'
    );
  }
  return context;
}

/** @internal */
export {SpeedDialContext, useSpeedDial};
export type {SpeedDialContextValue, SpeedDialDirection, SpeedDialPosition};
