// @docker/trident/src/components/status-indicator.tsx
import type {AsChild} from '@/lib/types';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {
  AlertTriangle,
  Circle,
  CircleCheck,
  Info,
  Minus,
  OctagonX,
  X,
} from 'lucide-react';
import {Slot} from 'radix-ui';
import * as React from 'react';

// ---------------------------------------------------------------------------
// StatusIcons
// ---------------------------------------------------------------------------

type StatusIconProps = React.SVGProps<SVGSVGElement> & {
  className?: string;
};

function CreatedIcon({className, ...props}: StatusIconProps) {
  return (
    <Circle
      className={cn('size-4 text-muted-foreground opacity-40', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

function RunningIcon({className, ...props}: StatusIconProps) {
  return (
    <CircleCheck
      className={cn('size-4 text-success', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

function PausedIcon({className, ...props}: StatusIconProps) {
  return (
    <Minus
      className={cn('size-4 text-muted-foreground', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

function ExitedIcon({className, ...props}: StatusIconProps) {
  return (
    <Circle
      className={cn('size-4 text-muted-foreground opacity-20', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

function DeadIcon({className, ...props}: StatusIconProps) {
  return (
    <X
      className={cn('size-4 text-muted-foreground', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

function CompletedIcon({className, ...props}: StatusIconProps) {
  return (
    <CircleCheck
      className={cn('size-4 text-success', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

function MixedIcon({className, ...props}: StatusIconProps) {
  return (
    <Info
      className={cn('size-4 text-primary', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

function ErrorIcon({className, ...props}: StatusIconProps) {
  return (
    <OctagonX
      className={cn('size-4 text-destructive', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

function WarningIcon({className, ...props}: StatusIconProps) {
  return (
    <AlertTriangle
      className={cn('size-4 text-warning', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

function OtherIcon({className, ...props}: StatusIconProps) {
  return (
    <AlertTriangle
      className={cn('size-4 text-warning', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    />
  );
}

// Inline SVG arc spinner — no matching icon in @docker/icons
function RestartingIcon({className, ...props}: StatusIconProps) {
  return (
    <svg
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
      className={cn('size-4 animate-spin text-primary', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    >
      {!props['aria-label'] && <title>Restarting</title>}
      <path d="M21 12a9 9 0 1 1-6.219-8.56" />
    </svg>
  );
}

function RemovingIcon({className, ...props}: StatusIconProps) {
  return (
    <svg
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
      className={cn('size-4 animate-spin text-destructive', className)}
      aria-hidden={!props['aria-label']}
      {...props}
    >
      {!props['aria-label'] && <title>Removing</title>}
      <path d="M21 12a9 9 0 1 1-6.219-8.56" />
    </svg>
  );
}

// Aliases matching old DDS export names
const CreatingIcon = RestartingIcon;
const StartingIcon = RestartingIcon;

// ---------------------------------------------------------------------------
// StatusIndicator
// ---------------------------------------------------------------------------

/**
 * A small status dot indicating the operational state of a resource.
 * Common in Docker dashboards for containers, images, and services.
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.1.1 Non-text Content — dot carries `role="img"` and `aria-label` describing the state.
 * @wcag 1.4.1 Use of Color — status label is available via `showLabel` to avoid color-only signaling.
 */

type StatusValue = 'running' | 'stopped' | 'warning' | 'error' | 'starting';

const STATUS_LABELS: Record<StatusValue, string> = {
  running: 'Running',
  stopped: 'Stopped',
  warning: 'Warning',
  error: 'Error',
  starting: 'Starting',
};

const dotVariants = cva('shrink-0 rounded-full', {
  variants: {
    status: {
      running: 'bg-status-running',
      stopped: 'bg-status-stopped',
      warning: 'bg-status-paused',
      error: 'bg-status-error',
      starting: 'animate-pulse bg-primary',
    },
    size: {
      sm: 'size-2',
      md: 'size-2.5',
      lg: 'size-3',
    },
  },
  defaultVariants: {
    status: 'stopped',
    size: 'md',
  },
});

/**
 * Status states that receive a glow in dark mode.
 * Stopped and starting do not glow — stopped is inert, starting pulses.
 * High-contrast mode: glow removed, 2px outline replaces it.
 *
 * Two-layer glow: tight core (4px blur, 70% opacity) + soft spread (12px blur, 25% opacity).
 * Uses the dark-mode token values directly since glow is dark-mode only.
 */
const GLOW_STATES = new Set<StatusValue>(['running', 'warning', 'error']);

/* Glow values use dark-mode status token OKLCH values directly:
 * running: oklch(0.720 0.160 155) — dark --status-running
 * warning: oklch(0.750 0.150 75)  — dark --status-paused
 * error:   oklch(0.720 0.180 25)  — dark --status-error
 * Applied via Tailwind dark:shadow-[...] arbitrary values in the dot element.
 */

type StatusIndicatorProps = React.ComponentProps<'span'> &
  VariantProps<typeof dotVariants> &
  AsChild & {
    status: StatusValue;
    /** Display a text label next to the dot. Defaults to `false`. */
    showLabel?: boolean;
    /** Override the default label text derived from `status`. */
    label?: string;
    /**
     * Add a subtle glow around active/alive states in dark mode.
     * Defaults to `true`. Glow is removed in light mode (unnecessary)
     * and high-contrast mode (replaced by outline).
     */
    glow?: boolean;
  };

function StatusIndicator({
  status,
  size,
  showLabel = false,
  label,
  glow = true,
  className,
  asChild = false,
  children,
  ...props
}: StatusIndicatorProps) {
  const resolvedLabel = label ?? STATUS_LABELS[status];
  const ariaLabel = `Status: ${resolvedLabel.toLowerCase()}`;
  const shouldGlow = glow && GLOW_STATES.has(status);

  const sharedProps = {
    'data-slot': 'status-indicator',
    'data-status': status,
    className: cn('inline-flex items-center gap-1.5', className),
    ...props,
  };

  const dot = (
    <span
      role="img"
      aria-label={ariaLabel}
      data-glow={shouldGlow ? '' : undefined}
      className={cn(
        dotVariants({status, size}),
        shouldGlow &&
          status === 'running' &&
          'dark:shadow-[var(--dds-shadow-glow-running)]',
        shouldGlow &&
          status === 'warning' &&
          'dark:shadow-[var(--dds-shadow-glow-warning)]',
        shouldGlow &&
          status === 'error' &&
          'dark:shadow-[var(--dds-shadow-glow-error)]',
        shouldGlow &&
          'high-contrast:shadow-none high-contrast:outline-2 high-contrast:outline-offset-1 high-contrast:outline-current'
      )}
    />
  );

  const labelEl = showLabel ? (
    <span aria-hidden="true" className="text-sm leading-none text-foreground">
      {resolvedLabel}
    </span>
  ) : null;

  if (asChild) {
    const child = React.Children.only(children);
    if (!React.isValidElement(child)) return null;
    return (
      <Slot.Root {...sharedProps}>
        {React.cloneElement(child, {}, dot, labelEl)}
      </Slot.Root>
    );
  }

  return (
    <span {...sharedProps}>
      {dot}
      {labelEl}
    </span>
  );
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {
  CompletedIcon,
  CreatedIcon,
  CreatingIcon,
  DeadIcon,
  ErrorIcon,
  ExitedIcon,
  MixedIcon,
  OtherIcon,
  PausedIcon,
  RemovingIcon,
  RestartingIcon,
  RunningIcon,
  StartingIcon,
  StatusIndicator,
  WarningIcon,
  dotVariants,
};
export type {StatusIconProps, StatusIndicatorProps, StatusValue};
