// @docker/trident/src/components/stepper.tsx
import {StepIndicator} from '@/components/step-indicator';
import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import * as React from 'react';

// ─── Types ────────────────────────────────────────────────────────────────────

type StepConfig = {
  /** Unique identifier for the step. */
  id: string;
  /** Display title shown in the step header. */
  title: string;
  /** Optional subtitle shown below the title. */
  description?: string;
  /** Optional steps can be skipped. */
  optional?: boolean;
  /**
   * Validation called before advancing past this step.
   * Return `true` to allow advancement, `false` to block silently,
   * or a string to block and show an error message.
   */
  validate?: () => boolean | string | Promise<boolean | string>;
};

type StepperContextValue = {
  steps: StepConfig[];
  currentStep: number;
  totalSteps: number;
  isFirst: boolean;
  isLast: boolean;
  canGoBack: boolean;
  validationError: string | null;
  isValidating: boolean;
  goNext: () => Promise<void>;
  goBack: () => void;
  goTo: (step: number) => void;
};

// ─── Context ──────────────────────────────────────────────────────────────────

const StepperContext = React.createContext<StepperContextValue | null>(null);

/**
 * Access the Stepper's state and navigation controls from any child component.
 *
 * Must be used inside a `<Stepper>` component.
 */
const useStepper = (): StepperContextValue => {
  const ctx = React.useContext(StepperContext);
  if (!ctx) {
    throw new Error('useStepper must be used inside a <Stepper> component.');
  }
  return ctx;
};

// ─── Stepper (Root / Provider) ────────────────────────────────────────────────

type StepperProps = {
  /** Ordered list of step definitions. */
  steps: StepConfig[];
  /**
   * Controlled current step (0-indexed).
   * If omitted the component is uncontrolled.
   */
  currentStep?: number;
  /** Called when the step changes (controlled mode). */
  onStepChange?: (step: number) => void;
  /** Called when the user advances past the final step. */
  onComplete?: () => void;
  className?: string;
  children?: React.ReactNode;
};

/**
 * Multi-step wizard orchestrator.
 *
 * Wraps content in a context that manages step state, validation, and
 * navigation. Compose with `StepperContent` and `StepperNavigation`.
 *
 * ### Example Usage
 *
 * ```tsx
 * <Stepper steps={steps} onComplete={() => setDone(true)}>
 *   <StepperContent step={0}><StepOneForm /></StepperContent>
 *   <StepperContent step={1}><StepTwoForm /></StepperContent>
 *   <StepperNavigation />
 * </Stepper>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — step progress communicated via StepIndicator role="img" labels.
 * @wcag 2.1.1 Keyboard — Enter in StepperNavigation triggers goNext.
 * @wcag 2.4.3 Focus Order — focus moves to new step content on step change.
 * @wcag 4.1.3 Status Messages — aria-live region announces step changes.
 */
const Stepper = ({
  steps,
  currentStep: controlledStep,
  onStepChange,
  onComplete,
  className,
  children,
}: StepperProps) => {
  const isControlled = controlledStep !== undefined;
  const [internalStep, setInternalStep] = React.useState(0);
  const [validationError, setValidationError] = React.useState<string | null>(
    null
  );
  const [isValidating, setIsValidating] = React.useState(false);

  const currentStep = isControlled ? controlledStep : internalStep;
  const totalSteps = steps.length;
  const isFirst = currentStep === 0;
  const isLast = currentStep === totalSteps - 1;

  const setStep = React.useCallback(
    (next: number) => {
      if (isControlled) {
        onStepChange?.(next);
      } else {
        setInternalStep(next);
      }
    },
    [isControlled, onStepChange]
  );

  const goNext = React.useCallback(async () => {
    const stepConfig = steps[currentStep];
    setValidationError(null);

    if (stepConfig?.validate) {
      setIsValidating(true);
      let result: boolean | string;
      try {
        result = await stepConfig.validate();
      } finally {
        setIsValidating(false);
      }

      if (result === false) {
        return;
      }
      if (typeof result === 'string') {
        setValidationError(result);
        return;
      }
    }

    if (isLast) {
      onComplete?.();
      return;
    }

    setStep(currentStep + 1);
  }, [currentStep, isLast, onComplete, setStep, steps]);

  const goBack = React.useCallback(() => {
    if (isFirst) return;
    setValidationError(null);
    setStep(currentStep - 1);
  }, [currentStep, isFirst, setStep]);

  const goTo = React.useCallback(
    (step: number) => {
      if (step < 0 || step >= totalSteps) return;
      setValidationError(null);
      setStep(step);
    },
    [setStep, totalSteps]
  );

  const contextValue = React.useMemo<StepperContextValue>(
    () => ({
      steps,
      currentStep,
      totalSteps,
      isFirst,
      isLast,
      canGoBack: !isFirst,
      validationError,
      isValidating,
      goNext,
      goBack,
      goTo,
    }),
    [
      steps,
      currentStep,
      totalSteps,
      isFirst,
      isLast,
      validationError,
      isValidating,
      goNext,
      goBack,
      goTo,
    ]
  );

  return (
    <StepperContext.Provider value={contextValue}>
      <div data-slot="stepper" className={cn('flex flex-col gap-6', className)}>
        {/* Progress indicator */}
        <StepIndicator
          steps={totalSteps}
          currentStep={currentStep + 1}
          data-slot="stepper-indicator"
        />

        {/* Step titles */}
        <StepperHeader />

        {/* aria-live region announces step changes to screen readers */}
        <div
          aria-live="polite"
          aria-atomic="true"
          className="sr-only"
          data-slot="stepper-announcer"
        >
          {`Step ${currentStep + 1} of ${totalSteps}: ${steps[currentStep]?.title ?? ''}`}
        </div>

        {children}
      </div>
    </StepperContext.Provider>
  );
};

// ─── StepperHeader (internal) ─────────────────────────────────────────────────

const StepperHeader = () => {
  const {steps, currentStep} = useStepper();
  const step = steps[currentStep];
  if (!step) return null;

  return (
    <div
      data-slot="stepper-header"
      aria-current="step"
      className="flex flex-col gap-1"
    >
      <div className="flex items-center gap-2">
        <h2 className="text-base font-heading text-foreground">{step.title}</h2>
        {step.optional && (
          <span className="text-xs text-muted-foreground">(Optional)</span>
        )}
      </div>
      {step.description && (
        <p className="text-sm text-muted-foreground">{step.description}</p>
      )}
    </div>
  );
};

// ─── StepperContent ───────────────────────────────────────────────────────────

type StepperContentProps = {
  /** The 0-indexed step this content belongs to. */
  step: number;
  children?: React.ReactNode;
  className?: string;
};

/**
 * Renders its children only when the Stepper's current step matches `step`.
 *
 * ```tsx
 * <StepperContent step={0}>
 *   <form>...</form>
 * </StepperContent>
 * ```
 */
const StepperContent = ({step, children, className}: StepperContentProps) => {
  const {currentStep} = useStepper();
  const contentRef = React.useRef<HTMLDivElement>(null);

  // Move focus to the content region when this step becomes active
  React.useEffect(() => {
    if (currentStep === step && contentRef.current) {
      // Focus the region so keyboard users land in the right place
      contentRef.current.focus({preventScroll: false});
    }
  }, [currentStep, step]);

  if (currentStep !== step) return null;

  return (
    <div
      ref={contentRef}
      data-slot="stepper-content"
      data-step={step}
      tabIndex={-1}
      className={cn('outline-none', className)}
    >
      {children}
    </div>
  );
};

// ─── StepperNavigation ────────────────────────────────────────────────────────

type StepperNavigationProps = {
  /** Label for the Next button. @default "Next" */
  nextLabel?: string;
  /** Label for the Back button. @default "Back" */
  backLabel?: string;
  /** Label for the Complete button on the last step. @default "Complete" */
  completeLabel?: string;
  /**
   * Show a spinner on the Next/Complete button while async validation runs.
   * Can be overridden by a consumer; the internal async state also controls it.
   */
  nextLoading?: boolean;
  className?: string;
};

/**
 * Pre-built Back / Next / Complete navigation row for a Stepper.
 *
 * Pressing Enter while focused inside this element triggers `goNext`.
 *
 * ```tsx
 * <StepperNavigation nextLabel="Continue" completeLabel="Submit" />
 * ```
 */
const StepperNavigation = ({
  nextLabel = 'Next',
  backLabel = 'Back',
  completeLabel = 'Complete',
  nextLoading,
  className,
}: StepperNavigationProps) => {
  const {
    isFirst,
    isLast,
    canGoBack,
    goNext,
    goBack,
    isValidating,
    validationError,
  } = useStepper();

  const loading = nextLoading ?? isValidating;

  return (
    <div
      data-slot="stepper-navigation"
      className={cn('flex flex-col gap-2', className)}
    >
      <div className="flex items-center gap-2">
        {canGoBack && (
          <Button
            variant="outline"
            onClick={goBack}
            disabled={isFirst}
            data-slot="stepper-back"
          >
            {backLabel}
          </Button>
        )}
        <Button
          onClick={() => void goNext()}
          loading={loading}
          data-slot="stepper-next"
        >
          {isLast ? completeLabel : nextLabel}
        </Button>
      </div>

      {validationError && (
        <p
          role="alert"
          data-slot="stepper-error"
          className="text-sm text-destructive"
        >
          {validationError}
        </p>
      )}
    </div>
  );
};

// ─── Exports ──────────────────────────────────────────────────────────────────

export {Stepper, StepperContent, StepperNavigation, useStepper};
export type {
  StepConfig,
  StepperContentProps,
  StepperNavigationProps,
  StepperProps,
};
