// @docker/trident/src/components/success-state.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {CircleCheck} from 'lucide-react';

type SuccessStateProps = {
  title: React.ReactNode;
  content?: React.ReactNode;
  actions?: React.ReactNode;
  /** Custom icon. Pass null to hide. @default CircleCheck */
  icon?: React.ComponentType<React.SVGProps<SVGSVGElement>> | null;
  className?: string;
};

/**
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.4.1 Use of Color — success state uses both a green icon and descriptive text.
 * @wcag 1.4.3 Contrast (Minimum) — text meets 4.5:1 contrast via DDS semantic tokens.
 * @wcag 4.1.2 Name, Role, Value — decorative icon is hidden via aria-hidden.
 */
function SuccessState({
  title,
  content,
  actions,
  icon: Icon = CircleCheck,
  className,
}: SuccessStateProps) {
  return (
    <div
      data-slot="success-state"
      className={cn(
        'flex flex-col items-center justify-center gap-4 py-16 text-center',
        className
      )}
    >
      {Icon !== null && (
        <div className="flex size-14 items-center justify-center rounded-full bg-success-wash">
          <Icon className="size-7 text-success" aria-hidden="true" />
        </div>
      )}
      <div className="space-y-1">
        <p className="text-sm font-label text-foreground">{title}</p>
        {content && (
          <p className="max-w-sm text-sm text-muted-foreground">{content}</p>
        )}
      </div>
      {actions && <div className="flex gap-2">{actions}</div>}
    </div>
  );
}

export {SuccessState};
export type {SuccessStateProps};
