// @docker/trident/src/components/tag.tsx
// tag.tsx

import type {AsChild} from '@/lib/types';

import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {X} from 'lucide-react';
import {Slot} from 'radix-ui';
import * as React from 'react';

const tagVariants = cva(
  'inline-flex items-center gap-1 rounded border select-none data-disabled:pointer-events-none data-disabled:opacity-60',
  {
    variants: {
      variant: {
        default: 'border-border bg-muted text-foreground',
        success: 'border-transparent bg-tag-success-bg text-tag-success-fg',
        warning: 'border-transparent bg-tag-warning-bg text-tag-warning-fg',
        destructive:
          'border-transparent bg-tag-destructive-bg text-tag-destructive-fg',
        info: 'border-transparent bg-tag-info-bg text-tag-info-fg',
      },
      size: {
        sm: 'h-5 px-1.5 text-2xs font-label',
        md: 'h-6 px-2 text-xs font-label',
        lg: 'h-7 px-2.5 text-sm font-label',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'md',
    },
  }
);

/**
 * A compact label element with an optional remove button.
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 4.1.2 Name, Role, Value — remove button includes a computed `aria-label` (e.g. "Remove {value}").
 * @wcag 2.1.1 Keyboard — remove button is a focusable `<button>` element, activated via Enter/Space.
 * @wcag 1.4.1 Use of Color — semantic variants use both color and context to convey meaning.
 * @wcag 1.4.3 Contrast (Minimum) — tag text meets 4.5:1 contrast via DDS tag-specific semantic tokens.
 */
type TagProps = React.ComponentProps<'span'> &
  AsChild & {
    /** The visual style of the tag. @default "default" */
    variant?: VariantProps<typeof tagVariants>['variant'];
    /** The size of the tag. @default "md" */
    size?: VariantProps<typeof tagVariants>['size'];
    /** Called when the user clicks the remove button. Omit to hide remove button. */
    onRemove?: () => void;
    disabled?: boolean;
  };

const Tag = ({
  className,
  children,
  variant = 'default',
  size = 'md',
  onRemove,
  disabled,
  asChild = false,
  ...props
}: TagProps) => {
  const sharedProps = {
    'data-slot': 'tag',
    'data-disabled': disabled || undefined,
    className: cn(tagVariants({variant, size}), className),
    ...props,
  };

  const removeButton = onRemove ? (
    <Button
      variant="ghost"
      size="icon-xs"
      onClick={onRemove}
      disabled={disabled}
      className="-mr-1 size-4 rounded opacity-60 hover:opacity-100"
      aria-label={`Remove ${typeof children === 'string' ? children : 'tag'}`}
    >
      <X className="size-2.5" aria-hidden="true" />
    </Button>
  ) : null;

  if (asChild) {
    const child = React.Children.only(children);
    if (!React.isValidElement<{children?: React.ReactNode}>(child)) return null;
    return (
      <Slot.Root {...sharedProps}>
        {React.cloneElement(child, {}, child.props.children, removeButton)}
      </Slot.Root>
    );
  }

  return (
    <span {...sharedProps}>
      {children}
      {removeButton}
    </span>
  );
};

export {Tag, tagVariants};
export type {TagProps};
