// @docker/trident/src/components/text-with-tooltip.tsx
// text-with-tooltip.tsx

import type * as React from 'react';

import {Tooltip, TooltipContent, TooltipTrigger} from '@/components/ui/tooltip';
import {cn} from '@/lib/utils';

type TextWithTooltipProps = {
  /** Content to render inside the tooltip */
  tooltip: React.ReactNode;
  /** Where the tooltip appears. @default "top" */
  tooltipPlacement?: 'top' | 'bottom' | 'left' | 'right';
  children: React.ReactNode;
  className?: string;
};

function TextWithTooltip({
  tooltip,
  tooltipPlacement = 'top',
  children,
  className,
}: TextWithTooltipProps) {
  return (
    <Tooltip>
      <TooltipTrigger asChild>
        <span
          data-slot="text-with-tooltip"
          // Radix's `asChild` slot forwards handlers but does not inject
          // `tabindex` onto elements that aren't inherently focusable. Opt the
          // span into the tab order so keyboard users can reveal the tooltip
          // (WCAG 2.1.1 Keyboard, 1.4.13 Content on Hover or Focus).
          tabIndex={0}
          className={cn(
            'cursor-help rounded-sm underline decoration-muted-foreground-decoration decoration-dashed underline-offset-2 outline-none hover:text-foreground focus-visible:ring-2 focus-visible:ring-ring',
            className
          )}
        >
          {children}
        </span>
      </TooltipTrigger>
      <TooltipContent side={tooltipPlacement}>{tooltip}</TooltipContent>
    </Tooltip>
  );
}

export {TextWithTooltip};
export type {TextWithTooltipProps};
