// @docker/trident/src/components/themed-picture.tsx
import {cn} from '@/lib/utils';

type ThemedPictureProps = {
  /** Image src for light mode */
  lightSrc: string;
  /** Image src for dark mode */
  darkSrc: string;
  alt: string;
  width: number;
  height: number;
  className?: string;
};

/**
 * Renders the appropriate image for the current theme (light/dark).
 * Responds to the `.dark` CSS class on an ancestor element, matching
 * Trident's class-based dark mode — not the OS `prefers-color-scheme` media query.
 */
function ThemedPicture({
  lightSrc,
  darkSrc,
  alt,
  width,
  height,
  className,
}: ThemedPictureProps) {
  return (
    <>
      {/* eslint-disable-next-line @next/next/no-img-element */}
      <img
        src={lightSrc}
        alt={alt}
        width={width}
        height={height}
        className={cn('dark:hidden', className)}
        loading="lazy"
        decoding="async"
      />
      {/* eslint-disable-next-line @next/next/no-img-element */}
      <img
        src={darkSrc}
        alt={alt}
        width={width}
        height={height}
        className={cn('hidden dark:block', className)}
        loading="lazy"
        decoding="async"
      />
    </>
  );
}

export {ThemedPicture};
export type {ThemedPictureProps};
