// @docker/trident/src/components/timeline.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';

/**
 * A vertical timeline that visually represents a sequence of events or steps.
 *
 * ### Composable Sub-components
 *
 * | Component             | Description                                      |
 * | :-------------------- | :----------------------------------------------- |
 * | `Timeline`            | Root container for timeline items                 |
 * | `TimelineItem`        | A single entry in the timeline                    |
 * | `TimelineSeparator`   | Draws the dot and the connecting line             |
 * | `TimelineDot`         | A circular indicator for each timeline entry      |
 * | `TimelineContent`     | The content area beside the separator             |
 *
 * ### Example Usage
 *
 * ```tsx
 * <Timeline>
 *   <TimelineItem>
 *     <TimelineSeparator>
 *       <TimelineDot />
 *     </TimelineSeparator>
 *     <TimelineContent>First event</TimelineContent>
 *   </TimelineItem>
 *   <TimelineItem>
 *     <TimelineSeparator>
 *       <TimelineDot variant="filled" />
 *     </TimelineSeparator>
 *     <TimelineContent>Last event</TimelineContent>
 *   </TimelineItem>
 * </Timeline>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — uses an ordered list (`ol`) to convey sequential structure.
 * @wcag 1.3.2 Meaningful Sequence — DOM order matches visual presentation order.
 * @wcag 1.4.1 Use of Color — dot variants use border/fill in addition to color to convey state.
 * @wcag 1.4.3 Contrast (Minimum) — all visual indicators meet 4.5:1 contrast via DDS semantic tokens.
 */
function Timeline({className, ...props}: React.ComponentProps<'ol'>) {
  return (
    <ol
      data-slot="timeline"
      className={cn('flex flex-col', className)}
      {...props}
    />
  );
}

/**
 * A single entry in the timeline, rendered as a list item with a separator and content side by side.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function TimelineItem({className, ...props}: React.ComponentProps<'li'>) {
  return (
    <li
      data-slot="timeline-item"
      className={cn('flex gap-3', className)}
      {...props}
    />
  );
}

/**
 * Renders the TimelineDot and the vertical line connecting adjacent items.
 *
 * The connecting line is drawn via `::before` and `::after` pseudo-elements
 * so the dot stays vertically centered with its content. The leading segment
 * is hidden on the first item and the trailing segment on the last, so a
 * single-item timeline renders as a dot only.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function TimelineSeparator({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="timeline-separator"
      className={cn(
        'flex flex-col items-center',
        'before:mb-1 before:w-px before:flex-1 before:bg-border before:content-[""]',
        'after:mt-1 after:w-px after:flex-1 after:bg-border after:content-[""]',
        '[li:first-child_&]:before:invisible [li:first-child_&]:before:mb-0',
        '[li:last-child_&]:after:invisible [li:last-child_&]:after:mt-0',
        className
      )}
      {...props}
    />
  );
}

const timelineDotVariants = cva(
  'flex shrink-0 items-center justify-center rounded-full [&>svg]:size-3.5',
  {
    variants: {
      variant: {
        default: 'size-3 bg-border',
        outline: 'size-3 border-2 border-border bg-transparent',
        filled: 'size-3 bg-primary',
      },
      size: {
        default: '',
        icon: 'size-7',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

type TimelineDotProps = React.ComponentProps<'span'> & {
  /**
   * The visual style of the dot.
   * @default "default"
   * @example
   * <TimelineDot variant="filled" />
   */
  variant?: VariantProps<typeof timelineDotVariants>['variant'];
};

/**
 * A circular indicator for each timeline entry. Supports `default`, `outline`, and `filled` variants.
 * Pass children (e.g., icons) to render them inside the dot.
 *
 * ### Component Props
 *
 * | Prop      | Type                                      | Default     |
 * | :-------- | :---------------------------------------- | :---------- |
 * | _variant_ | `"default" \| "outline" \| "filled"`      | `"default"` |
 *
 * ### Example Usage
 *
 * ```tsx
 * <TimelineDot />
 * <TimelineDot variant="outline" />
 * <TimelineDot variant="filled" />
 * <TimelineDot variant="filled" className="bg-success">
 *   <CheckCircle />
 * </TimelineDot>
 * ```
 */
function TimelineDot({
  className,
  variant = 'default',
  children,
  ...props
}: TimelineDotProps) {
  const hasChildren = children != null;
  return (
    <span
      data-slot="timeline-dot"
      data-variant={variant}
      className={cn(
        timelineDotVariants({
          variant,
          size: hasChildren ? 'icon' : 'default',
        }),
        className
      )}
      {...props}
    >
      {children}
    </span>
  );
}

/**
 * The content area beside the timeline separator. Typically contains text, timestamps, or any custom content.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function TimelineContent({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="timeline-content"
      className={cn(
        'py-1 text-sm [li:first-child_&]:pt-0 [li:last-child_&]:pb-0',
        className
      )}
      {...props}
    />
  );
}

export {
  Timeline,
  TimelineItem,
  TimelineSeparator,
  TimelineDot,
  TimelineContent,
  timelineDotVariants,
};
export type {TimelineDotProps};
