// @docker/trident/src/components/topology-node.tsx
import type * as React from 'react';

import {InlineSegmentedToggle} from '@/components/inline-segmented-toggle';
import {cn} from '@/lib/utils';

import {ChevronRight, Cloud, Home, KeyRound} from 'lucide-react';

export const TOPOLOGY_NODE_W = 140;
export const TOPOLOGY_NODE_H = 88;

type TopoMode = 'local' | 'remote';
type BadgeTone = 'ok' | 'err' | 'warn' | 'loading';

type TopologyNodeProps = {
  label: string;
  mode: TopoMode;
  endpoint?: string;
  badge?: string;
  badgeTone?: BadgeTone;
  selected?: boolean;
  requestsPerSec?: number;
  latencyMs?: number;
  secretCount?: number;
  hasDrilldown?: boolean;
  extra?: string;
  onClick?: () => void;
  onDoubleClick?: () => void;
  onToggleMode?: (next: TopoMode) => void;
  className?: string;
};

const MODE_TOGGLE_OPTIONS: {value: TopoMode; label: string}[] = [
  {value: 'local', label: 'local'},
  {value: 'remote', label: 'remote'},
];

const BADGE_DOT_CLASS: Record<BadgeTone, string> = {
  ok: 'bg-success',
  err: 'bg-destructive',
  warn: 'bg-warning',
  loading: 'bg-primary animate-pulse',
};

const BADGE_TEXT_CLASS: Record<BadgeTone, string> = {
  ok: 'text-success',
  err: 'font-semibold text-destructive',
  warn: 'text-warning',
  loading: 'text-primary',
};

function formatRate(n: number): string {
  if (n >= 1000) return `${(n / 1000).toFixed(n >= 10_000 ? 0 : 1)}k`;
  return String(n);
}

/**
 * 140x88px interactive card for dependency graph nodes. Renders inside a
 * pan/zoom SVG canvas via a foreignObject. All nodes share the same surface
 * material; local vs remote is differentiated by the Home/Cloud icon.
 *
 * The mode toggle is rendered outside the card button so buttons never nest.
 *
 * @example
 * <TopologyNode
 *   label="app-ui"
 *   mode="local"
 *   endpoint="localhost:3000"
 *   badge="up"
 *   badgeTone="ok"
 *   selected={false}
 *   cpu="6.2%"
 *   memory="235 MB"
 *   onClick={() => select(node)}
 *   onToggleMode={(next) => setMode(next)}
 * />
 */
function TopologyNode({
  label,
  mode,
  endpoint,
  badge,
  badgeTone = 'loading',
  selected = false,
  requestsPerSec,
  latencyMs,
  secretCount,
  hasDrilldown = false,
  extra,
  onClick,
  onDoubleClick,
  onToggleMode,
  className,
}: TopologyNodeProps) {
  const isRemote = mode === 'remote';

  const trafficParts: string[] = [];
  if (typeof requestsPerSec === 'number')
    trafficParts.push(`${formatRate(requestsPerSec)} req/s`);
  if (isRemote && typeof latencyMs === 'number')
    trafficParts.push(`${latencyMs}ms`);
  const trafficLine = trafficParts.join(' · ');

  // Decorative shadows (inset highlight + selected glow) are rendered on a
  // `::before` pseudo-layer so the button's own `box-shadow` is free for
  // Tailwind's `ring-*` utilities to produce a visible focus indicator.
  // See docs/audits/issues/topology-node/p0-focus-visible-ring-invisible-all-three-themes.md
  const decorativeShadow = selected
    ? '0 0 12px color-mix(in oklch, var(--primary) 15%, transparent), inset 0 1px 0 var(--highlight)'
    : 'inset 0 1px 0 var(--highlight)';

  return (
    <div className={cn('flex w-full flex-col', className)}>
      <button
        type="button"
        onClick={onClick}
        onDoubleClick={onDoubleClick}
        aria-pressed={selected}
        style={
          {
            height: onToggleMode ? undefined : TOPOLOGY_NODE_H,
            '--topology-node-shadow': decorativeShadow,
          } as React.CSSProperties
        }
        className={cn(
          'relative flex w-full cursor-pointer flex-col border border-border bg-popover pt-1.5 pr-2.5 pb-1.5 pl-2.5 text-left transition-all duration-100',
          // Decorative inset highlight + selected glow on a pseudo-layer so
          // focus-visible ring (compiled to box-shadow by Tailwind) is not
          // clobbered.
          'before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:shadow-[var(--topology-node-shadow)] before:content-[""]',
          onToggleMode
            ? 'rounded-t-[var(--dds-radius-default)]'
            : 'rounded-[var(--dds-radius-default)]',
          'hover:-translate-y-px hover:border-muted-foreground',
          'focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none',
          selected && 'border-primary'
        )}
      >
        {/* Row 1: mode icon + label + meta icons */}
        <div className="flex items-start gap-1">
          {isRemote ? (
            <Cloud
              aria-label="remote"
              className="mt-px mr-1 size-3 shrink-0 text-muted-foreground"
            />
          ) : (
            <Home
              aria-label="local"
              className="mt-px mr-1 size-3 shrink-0 text-primary"
            />
          )}
          <span className="min-w-0 flex-1 truncate text-[14px] leading-[1.3] font-label tracking-[-0.011em] text-foreground">
            {label}
          </span>
          {secretCount !== null &&
            secretCount !== undefined &&
            secretCount > 0 && (
              <span
                className="mt-px flex size-3 shrink-0 items-center justify-center text-muted-foreground"
                title={`${secretCount} secret${secretCount === 1 ? '' : 's'}`}
              >
                <KeyRound
                  aria-label={`${secretCount} secret${secretCount === 1 ? '' : 's'}`}
                  className="size-3"
                />
              </span>
            )}
          {hasDrilldown && (
            <ChevronRight
              aria-label="drill down"
              className="mt-px size-3 shrink-0 text-primary"
            />
          )}
        </div>

        {/* Row 2: endpoint */}
        {endpoint && (
          <span className="block truncate font-mono text-[12px] leading-[1.4] text-muted-foreground">
            {endpoint}
          </span>
        )}

        {/* Row 3: traffic line */}
        {trafficLine && (
          <span className="block truncate font-mono text-[11px] leading-[1.35] text-muted-foreground">
            {trafficLine}
          </span>
        )}

        {/* Row 4: status dot + badge */}
        {badge && (
          <div className="mt-1 flex items-center gap-1 font-mono text-[12px]">
            <span
              aria-hidden
              className={cn(
                'size-[6px] shrink-0 rounded-full',
                BADGE_DOT_CLASS[badgeTone]
              )}
            />
            <span className={cn('truncate', BADGE_TEXT_CLASS[badgeTone])}>
              {badge}
            </span>
          </div>
        )}

        {/* Row 5: Gordon / extra activity */}
        {extra && (
          <span className="mt-0.5 block font-mono text-[10px] leading-[1.3] text-primary">
            {extra}
          </span>
        )}
      </button>
      {onToggleMode && (
        <div className="rounded-b-[var(--dds-radius-default)] border border-t-0 border-border bg-popover px-2.5 pb-1.5">
          <InlineSegmentedToggle
            options={MODE_TOGGLE_OPTIONS}
            value={mode}
            onChange={onToggleMode}
          />
        </div>
      )}
    </div>
  );
}

export {TopologyNode};
export type {TopologyNodeProps, TopoMode, BadgeTone};
