// @docker/trident/src/components/transfer-list-panel.tsx
'use client';

import type {TransferItem} from '@/components/transfer-list';
import type * as React from 'react';

import {Search} from '@/components/search';
import {Checkbox} from '@/components/ui/checkbox';
import {ScrollArea} from '@/components/ui/scroll-area';
import {cn} from '@/lib/utils';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type TransferListPanelProps = {
  label: string;
  description?: string;
  items: TransferItem[];
  checkedIds: ReadonlySet<string>;
  onCheckedChange: (id: string, checked: boolean) => void;
  onSelectAll: (checked: boolean) => void;
  searchable: boolean;
  searchPlaceholder: string;
  searchQuery: string;
  onSearchChange: (value: string) => void;
  renderItem?: (
    item: TransferItem,
    side: 'available' | 'selected'
  ) => React.ReactNode;
  side: 'available' | 'selected';
  disabled?: boolean;
  filteredItems: TransferItem[];
};

// ---------------------------------------------------------------------------
// TransferListPanel
// ---------------------------------------------------------------------------

const TransferListPanel = ({
  label,
  description,
  items,
  checkedIds,
  onCheckedChange,
  onSelectAll,
  searchable,
  searchPlaceholder,
  searchQuery,
  onSearchChange,
  renderItem,
  side,
  disabled,
  filteredItems,
}: TransferListPanelProps) => {
  const totalCount = items.length;
  const checkedCount = filteredItems.filter(i => checkedIds.has(i.id)).length;
  const allChecked =
    filteredItems.length > 0 && checkedCount === filteredItems.length;
  const someChecked = checkedCount > 0 && !allChecked;
  const selectAllState: boolean | 'indeterminate' = someChecked
    ? 'indeterminate'
    : allChecked;

  const listboxId = `transfer-list-${side}`;

  return (
    <div
      className="flex min-w-0 flex-1 flex-col rounded border border-border bg-background"
      data-slot="transfer-panel"
      data-side={side}
    >
      {/* Panel header */}
      <div className="border-b border-border px-3 py-2">
        <div className="flex items-baseline gap-1.5">
          <span className="text-sm font-label text-foreground">{label}</span>
          <span className="text-xs text-muted-foreground">({totalCount})</span>
        </div>
        {description && (
          <p className="mt-0.5 text-xs text-muted-foreground">{description}</p>
        )}
      </div>

      {/* Search */}
      {searchable && (
        <div className="border-b border-border px-3 py-2">
          <Search
            value={searchQuery}
            onChange={onSearchChange}
            placeholder={searchPlaceholder}
            disabled={disabled}
            className="w-full"
            aria-label={`Search ${label}`}
          />
        </div>
      )}

      {/* Select-all row */}
      <div className="flex items-center gap-2 border-b border-border px-3 py-1.5">
        <Checkbox
          id={`${listboxId}-select-all`}
          checked={selectAllState}
          onCheckedChange={val => onSelectAll(val === true)}
          disabled={disabled || filteredItems.length === 0}
          aria-label={`Select all ${label} items`}
        />
        <label
          htmlFor={`${listboxId}-select-all`}
          className={cn(
            'cursor-pointer text-xs text-muted-foreground select-none',
            (disabled || filteredItems.length === 0) &&
              'cursor-not-allowed opacity-50'
          )}
        >
          {checkedCount > 0
            ? `${checkedCount} of ${filteredItems.length} selected`
            : 'Select all'}
        </label>
      </div>

      {/* Scrollable item list */}
      <ScrollArea className="h-64">
        <div
          id={listboxId}
          role="listbox"
          aria-label={label}
          aria-multiselectable="true"
          className="flex flex-col py-1"
        >
          {filteredItems.length === 0 ? (
            <div className="px-3 py-4 text-center text-xs text-muted-foreground">
              {searchQuery ? 'No results match your search.' : 'No items.'}
            </div>
          ) : (
            filteredItems.map(item => {
              const isChecked = checkedIds.has(item.id);
              const isDisabled = disabled || item.disabled;

              return (
                <div
                  key={item.id}
                  role="option"
                  aria-selected={isChecked}
                  aria-disabled={isDisabled}
                  className={cn(
                    'flex cursor-pointer items-start gap-2.5 px-3 py-2 transition-colors',
                    'hover:bg-muted',
                    isChecked && 'bg-selected-overlay',
                    isDisabled && 'cursor-not-allowed opacity-50'
                  )}
                  onClick={() => {
                    if (!isDisabled) {
                      onCheckedChange(item.id, !isChecked);
                    }
                  }}
                  onKeyDown={e => {
                    if (!isDisabled && (e.key === ' ' || e.key === 'Enter')) {
                      e.preventDefault();
                      onCheckedChange(item.id, !isChecked);
                    }
                  }}
                  tabIndex={isDisabled ? -1 : 0}
                >
                  {/* Visual-only checkbox indicator — the parent div[role="option"]
                      owns all interactivity. A Radix Checkbox renders as <button>
                      which triggers axe nested-interactive inside a focusable option. */}
                  <div
                    aria-hidden="true"
                    className={cn(
                      'mt-0.5 size-4 shrink-0 rounded border border-input transition-colors',
                      isChecked && 'border-primary bg-primary',
                      isDisabled && 'opacity-50'
                    )}
                  >
                    {isChecked && (
                      <svg
                        viewBox="0 0 12 12"
                        fill="none"
                        className="size-full p-px text-primary-foreground"
                        aria-hidden="true"
                      >
                        <polyline
                          points="1.5,6 4.5,9.5 10.5,2.5"
                          stroke="currentColor"
                          strokeWidth="1.5"
                          strokeLinecap="round"
                          strokeLinejoin="round"
                        />
                      </svg>
                    )}
                  </div>
                  <div className="min-w-0 flex-1">
                    {renderItem ? (
                      renderItem(item, side)
                    ) : (
                      <>
                        <span className="block truncate text-sm text-foreground">
                          {item.label}
                        </span>
                        {item.description && (
                          <span className="block truncate text-xs text-muted-foreground">
                            {item.description}
                          </span>
                        )}
                      </>
                    )}
                  </div>
                </div>
              );
            })
          )}
        </div>
      </ScrollArea>
    </div>
  );
};

/** @internal */
export {TransferListPanel};
export type {TransferListPanelProps};
