// @docker/trident/src/components/transfer-list.tsx
'use client';

import {TransferListPanel} from '@/components/transfer-list-panel';
import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import {
  ChevronLeft,
  ChevronRight,
  ChevronsLeft,
  ChevronsRight,
} from 'lucide-react';
import * as React from 'react';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type TransferItem = {
  id: string;
  label: string;
  description?: string;
  disabled?: boolean;
};

type TransferListProps = {
  // Data
  available: TransferItem[];
  selected: TransferItem[];
  onChange: (selected: TransferItem[]) => void;

  // Labels
  availableLabel?: string;
  selectedLabel?: string;
  availableDescription?: string;
  selectedDescription?: string;

  // Search
  searchable?: boolean;
  searchPlaceholder?: string;

  // Transfer controls
  showMoveAll?: boolean;

  // Custom rendering
  renderItem?: (
    item: TransferItem,
    side: 'available' | 'selected'
  ) => React.ReactNode;

  disabled?: boolean;
  className?: string;
};

// ---------------------------------------------------------------------------
// TransferList
// ---------------------------------------------------------------------------

const TransferList = ({
  available,
  selected,
  onChange,
  availableLabel = 'Available',
  selectedLabel = 'Selected',
  availableDescription,
  selectedDescription,
  searchable = true,
  searchPlaceholder = 'Search…',
  showMoveAll = true,
  renderItem,
  disabled = false,
  className,
}: TransferListProps) => {
  // Checked ids on each side (checked = staged for transfer, not "on the selected side")
  const [availableChecked, setAvailableChecked] = React.useState<Set<string>>(
    new Set()
  );
  const [selectedChecked, setSelectedChecked] = React.useState<Set<string>>(
    new Set()
  );

  // Independent search state per side
  const [availableSearch, setAvailableSearch] = React.useState('');
  const [selectedSearch, setSelectedSearch] = React.useState('');

  // Filtered views — search only affects display, not the underlying data
  const filteredAvailable = React.useMemo(
    () =>
      available.filter(i =>
        i.label.toLowerCase().includes(availableSearch.toLowerCase())
      ),
    [available, availableSearch]
  );

  const filteredSelected = React.useMemo(
    () =>
      selected.filter(i =>
        i.label.toLowerCase().includes(selectedSearch.toLowerCase())
      ),
    [selected, selectedSearch]
  );

  // Derived counts for button disabled state
  const availableCheckedCount = filteredAvailable.filter(i =>
    availableChecked.has(i.id)
  ).length;
  const selectedCheckedCount = filteredSelected.filter(i =>
    selectedChecked.has(i.id)
  ).length;

  const selectedIds = React.useMemo(
    () => new Set(selected.map(i => i.id)),
    [selected]
  );

  // ----- Transfer handlers -----

  const moveCheckedToSelected = () => {
    const toMove = filteredAvailable.filter(
      i => availableChecked.has(i.id) && !i.disabled
    );
    if (toMove.length === 0) return;
    const toMoveIds = new Set(toMove.map(i => i.id));
    onChange([...selected, ...toMove]);
    setAvailableChecked(
      new Set([...availableChecked].filter(id => !toMoveIds.has(id)))
    );
  };

  const moveAllToSelected = () => {
    const toMove = filteredAvailable.filter(
      i => !i.disabled && !selectedIds.has(i.id)
    );
    if (toMove.length === 0) return;
    const toMoveIds = new Set(toMove.map(i => i.id));
    onChange([...selected, ...toMove]);
    setAvailableChecked(
      new Set([...availableChecked].filter(id => !toMoveIds.has(id)))
    );
  };

  const moveCheckedToAvailable = () => {
    const toRemoveIds = new Set(
      filteredSelected
        .filter(i => selectedChecked.has(i.id) && !i.disabled)
        .map(i => i.id)
    );
    if (toRemoveIds.size === 0) return;
    onChange(selected.filter(i => !toRemoveIds.has(i.id)));
    setSelectedChecked(
      new Set([...selectedChecked].filter(id => !toRemoveIds.has(id)))
    );
  };

  const moveAllToAvailable = () => {
    const toRemoveIds = new Set(
      filteredSelected.filter(i => !i.disabled).map(i => i.id)
    );
    if (toRemoveIds.size === 0) return;
    onChange(selected.filter(i => !toRemoveIds.has(i.id)));
    setSelectedChecked(
      new Set([...selectedChecked].filter(id => !toRemoveIds.has(id)))
    );
  };

  // ----- Select-all handlers -----

  const handleAvailableSelectAll = (checked: boolean) => {
    setAvailableChecked(
      checked
        ? new Set(filteredAvailable.filter(i => !i.disabled).map(i => i.id))
        : new Set()
    );
  };

  const handleSelectedSelectAll = (checked: boolean) => {
    setSelectedChecked(
      checked
        ? new Set(filteredSelected.filter(i => !i.disabled).map(i => i.id))
        : new Set()
    );
  };

  // ----- Per-item check handlers -----

  const handleAvailableCheck = (id: string, checked: boolean) => {
    setAvailableChecked(prev => {
      const next = new Set(prev);
      if (checked) next.add(id);
      else next.delete(id);
      return next;
    });
  };

  const handleSelectedCheck = (id: string, checked: boolean) => {
    setSelectedChecked(prev => {
      const next = new Set(prev);
      if (checked) next.add(id);
      else next.delete(id);
      return next;
    });
  };

  // ----- Button disabled states -----

  const moveRightDisabled = disabled || availableCheckedCount === 0;
  const moveAllRightDisabled =
    disabled ||
    filteredAvailable.filter(i => !i.disabled && !selectedIds.has(i.id))
      .length === 0;
  const moveLeftDisabled = disabled || selectedCheckedCount === 0;
  const moveAllLeftDisabled =
    disabled || filteredSelected.filter(i => !i.disabled).length === 0;

  return (
    <div
      data-slot="transfer-list"
      className={cn('flex items-stretch gap-3', className)}
    >
      {/* Available panel */}
      <TransferListPanel
        label={availableLabel}
        description={availableDescription}
        items={available}
        checkedIds={availableChecked}
        onCheckedChange={handleAvailableCheck}
        onSelectAll={handleAvailableSelectAll}
        searchable={searchable}
        searchPlaceholder={searchPlaceholder}
        searchQuery={availableSearch}
        onSearchChange={setAvailableSearch}
        renderItem={renderItem}
        side="available"
        disabled={disabled}
        filteredItems={filteredAvailable}
      />

      {/* Transfer controls */}
      <div className="flex shrink-0 flex-col items-center justify-center gap-1.5">
        {showMoveAll && (
          <Button
            variant="outline"
            size="icon-sm"
            onClick={moveAllToSelected}
            disabled={moveAllRightDisabled}
            aria-label={`Move all ${availableLabel} items to ${selectedLabel}`}
            title="Move all →"
          >
            <ChevronsRight className="size-4" aria-hidden="true" />
          </Button>
        )}

        <Button
          variant="outline"
          size="icon-sm"
          onClick={moveCheckedToSelected}
          disabled={moveRightDisabled}
          aria-label={`Move checked ${availableLabel} items to ${selectedLabel}`}
          title="Move selected →"
          onKeyDown={e => {
            if (e.key === 'Enter') moveCheckedToSelected();
          }}
        >
          <ChevronRight className="size-4" aria-hidden="true" />
        </Button>

        <Button
          variant="outline"
          size="icon-sm"
          onClick={moveCheckedToAvailable}
          disabled={moveLeftDisabled}
          aria-label={`Move checked ${selectedLabel} items back to ${availableLabel}`}
          title="← Move selected"
          onKeyDown={e => {
            if (e.key === 'Enter') moveCheckedToAvailable();
          }}
        >
          <ChevronLeft className="size-4" aria-hidden="true" />
        </Button>

        {showMoveAll && (
          <Button
            variant="outline"
            size="icon-sm"
            onClick={moveAllToAvailable}
            disabled={moveAllLeftDisabled}
            aria-label={`Move all ${selectedLabel} items back to ${availableLabel}`}
            title="← Move all"
          >
            <ChevronsLeft className="size-4" aria-hidden="true" />
          </Button>
        )}
      </div>

      {/* Selected panel */}
      <TransferListPanel
        label={selectedLabel}
        description={selectedDescription}
        items={selected}
        checkedIds={selectedChecked}
        onCheckedChange={handleSelectedCheck}
        onSelectAll={handleSelectedSelectAll}
        searchable={searchable}
        searchPlaceholder={searchPlaceholder}
        searchQuery={selectedSearch}
        onSearchChange={setSelectedSearch}
        renderItem={renderItem}
        side="selected"
        disabled={disabled}
        filteredItems={filteredSelected}
      />
    </div>
  );
};

export {TransferList};
export type {TransferItem, TransferListProps};
