// @docker/trident/src/components/transparency-label.tsx
// transparency-label.tsx
import type * as React from 'react';

import {FreshnessIndicator} from '@/components/freshness-indicator';
import {cn} from '@/lib/utils';

// ---------------------------------------------------------------------------
// Internal Row helper
// ---------------------------------------------------------------------------

type RowProps = {
  label: string;
  children: React.ReactNode;
  className?: string;
};

function Row({label, children, className}: RowProps) {
  return (
    <div
      className={cn('flex items-start justify-between gap-4 py-1.5', className)}
    >
      <span className="shrink-0 text-xs text-muted-foreground">{label}</span>
      <span className="min-w-0 text-right text-xs text-foreground">
        {children}
      </span>
    </div>
  );
}

// ---------------------------------------------------------------------------
// Section divider
// ---------------------------------------------------------------------------

function SectionHeading({children}: {children: React.ReactNode}) {
  return (
    <p className="mt-3 mb-1 text-2xs font-heading tracking-wide text-muted-foreground uppercase first:mt-0">
      {children}
    </p>
  );
}

// ---------------------------------------------------------------------------
// TransparencyLabel types
// ---------------------------------------------------------------------------

type VulnerabilityCounts = {
  critical?: number;
  high?: number;
  medium?: number;
  low?: number;
};

type TransparencyLabelProps = {
  // --- Identity ---
  /** Full image reference, e.g. "docker.io/library/nginx" */
  image: string;
  /** Image tag, e.g. "1.27-alpine" */
  tag: string;
  /** Compressed image size, e.g. "45.3 MB" */
  compressedSize?: string;

  // --- Freshness ---
  /** ISO date string or Date of last push. */
  pushedAt?: Date | string;

  // --- Security ---
  /** Total CVE count (all severities). */
  cveTotal?: number;
  /** Per-severity breakdown — omit fields to hide them. */
  vulnerabilities?: VulnerabilityCounts;
  /** Whether the image passes Docker Scout policy. */
  policyPassed?: boolean;

  // --- Provenance ---
  /** OCI image digest, e.g. "sha256:abc123…" */
  digest?: string;
  /** SLSA provenance level, e.g. "SLSA Level 2". */
  slsaLevel?: string;
  /** Whether the image has a valid Sigstore signature. */
  signed?: boolean;

  className?: string;
};

/**
 * A structured metadata panel for a Docker image — analogous to a nutrition
 * label — covering identity, freshness, security signals, and provenance.
 *
 * ### Layout
 *
 * Four labeled sections stacked vertically, each using the internal `<Row>`
 * helper for consistent label/value alignment.
 */
function TransparencyLabel({
  image,
  tag,
  compressedSize,
  pushedAt,
  cveTotal,
  vulnerabilities,
  policyPassed,
  digest,
  slsaLevel,
  signed,
  className,
}: TransparencyLabelProps) {
  return (
    <div
      data-slot="transparency-label"
      className={cn(
        'w-72 rounded-lg border border-border bg-card p-4 text-xs shadow-raised',
        className
      )}
    >
      {/* ------------------------------------------------------------------ */}
      {/* Identity                                                            */}
      {/* ------------------------------------------------------------------ */}
      <SectionHeading>Identity</SectionHeading>
      <div className="divide-y divide-border-subtle">
        <Row label="Image">
          <span className="truncate font-mono text-2xs">{image}</span>
        </Row>
        <Row label="Tag">{tag}</Row>
        {compressedSize !== undefined && (
          <Row label="Size">{compressedSize}</Row>
        )}
      </div>

      {/* ------------------------------------------------------------------ */}
      {/* Freshness                                                           */}
      {/* ------------------------------------------------------------------ */}
      <SectionHeading>Freshness</SectionHeading>
      <div className="divide-y divide-border-subtle">
        <Row label="Last pushed">
          {pushedAt !== undefined ? (
            <FreshnessIndicator date={pushedAt} />
          ) : (
            <span className="text-muted-foreground">Unknown</span>
          )}
        </Row>
      </div>

      {/* ------------------------------------------------------------------ */}
      {/* Security                                                            */}
      {/* ------------------------------------------------------------------ */}
      <SectionHeading>Security</SectionHeading>
      <div className="divide-y divide-border-subtle">
        {cveTotal !== undefined && (
          <Row label="CVEs">{cveTotal === 0 ? 'None found' : cveTotal}</Row>
        )}
        {vulnerabilities?.critical !== undefined && (
          <Row label="Critical">{vulnerabilities.critical}</Row>
        )}
        {vulnerabilities?.high !== undefined && (
          <Row label="High">{vulnerabilities.high}</Row>
        )}
        {vulnerabilities?.medium !== undefined && (
          <Row label="Medium">{vulnerabilities.medium}</Row>
        )}
        {vulnerabilities?.low !== undefined && (
          <Row label="Low">{vulnerabilities.low}</Row>
        )}
        {policyPassed !== undefined && (
          <Row label="Policy">
            {policyPassed ? (
              <span className="text-success">Passed</span>
            ) : (
              <span className="text-destructive">Failed</span>
            )}
          </Row>
        )}
      </div>

      {/* ------------------------------------------------------------------ */}
      {/* Provenance                                                          */}
      {/* ------------------------------------------------------------------ */}
      <SectionHeading>Provenance</SectionHeading>
      <div className="divide-y divide-border-subtle">
        {digest !== undefined && (
          <Row label="Digest">
            <span className="truncate font-mono text-2xs">
              {digest.length > 20 ? `${digest.slice(0, 20)}…` : digest}
            </span>
          </Row>
        )}
        {slsaLevel !== undefined && <Row label="SLSA">{slsaLevel}</Row>}
        {signed !== undefined && (
          <Row label="Signed">{signed ? 'Yes' : 'No'}</Row>
        )}
      </div>
    </div>
  );
}

export {TransparencyLabel};
export type {TransparencyLabelProps, VulnerabilityCounts};
