// @docker/trident/src/components/tree-view.tsx
'use client';

import type {
  SelectEventMeta,
  TreeNode,
  TreeViewNodeProps,
  TreeViewProps,
} from './tree-view-types';

import {cn} from '@/lib/utils';

import {ChevronRight, File, Folder} from 'lucide-react';
import * as React from 'react';

import {computeMultiSelection, getVisibleIds} from './tree-view-selection';

// ─── TreeViewNode (internal) ──────────────────────────────────────────────────

const TreeViewNode = <T = Record<string, unknown>>({
  node,
  depth,
  expandedIds,
  selectedIds,
  focusedId,
  onToggle,
  onSelect,
  onKeyDown,
  renderNode,
  setRef,
}: TreeViewNodeProps<T>) => {
  const hasChildren = Boolean(node.children && node.children.length > 0);
  const isExpanded = expandedIds.has(node.id);
  const isSelected = selectedIds.has(node.id);
  const isFocused = focusedId === node.id;

  const handleClick = (e: React.MouseEvent) => {
    e.stopPropagation();
    if (hasChildren) {
      onToggle(node.id);
    }
    onSelect(node.id, node, {
      ctrlKey: e.ctrlKey || e.metaKey,
      shiftKey: e.shiftKey,
    });
  };

  const handleKeyDown = (e: React.KeyboardEvent) => {
    e.stopPropagation();
    onKeyDown(e, node.id, node, hasChildren, isExpanded);
  };

  const indentStyle =
    depth > 0
      ? {paddingInlineStart: `calc(${depth} * var(--dds-space-500))`}
      : undefined;

  return (
    <li
      role="treeitem"
      aria-expanded={hasChildren ? isExpanded : undefined}
      aria-selected={isSelected}
      aria-level={depth + 1}
      ref={el => setRef(node.id, el)}
      tabIndex={isFocused ? 0 : -1}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      data-slot="tree-view-node"
      data-node-id={node.id}
      data-selected={isSelected ? 'true' : undefined}
      className="group outline-none"
    >
      <div
        style={indentStyle}
        className={cn(
          'flex cursor-pointer items-center gap-1.5 rounded px-2 py-1 text-sm transition-colors select-none',
          'hover:bg-hover-overlay',
          'group-focus-visible:ring-2 group-focus-visible:ring-ring group-focus-visible:ring-offset-1',
          isSelected && 'bg-muted text-foreground',
          !isSelected && 'text-foreground'
        )}
      >
        <span
          aria-hidden="true"
          className={cn(
            'flex size-4 shrink-0 items-center justify-center text-muted-foreground transition-transform duration-150 ease-in-out',
            hasChildren && isExpanded && 'rotate-90',
            !hasChildren && 'pointer-events-none opacity-0'
          )}
        >
          <ChevronRight className="size-3.5" />
        </span>

        <span
          aria-hidden="true"
          className="flex shrink-0 items-center text-muted-foreground"
        >
          {hasChildren ? (
            <Folder className="size-4" />
          ) : (
            <File className="size-4" />
          )}
        </span>

        {renderNode ? (
          renderNode(node, depth)
        ) : (
          <span className="truncate">{node.label}</span>
        )}
      </div>

      {hasChildren && isExpanded && (
        <ul role="group">
          {(node.children ?? []).map(child => (
            <TreeViewNode<T>
              key={child.id}
              node={child}
              depth={depth + 1}
              expandedIds={expandedIds}
              selectedIds={selectedIds}
              focusedId={focusedId}
              onToggle={onToggle}
              onSelect={onSelect}
              onKeyDown={onKeyDown}
              renderNode={renderNode}
              setRef={setRef}
            />
          ))}
        </ul>
      )}
    </li>
  );
};

// ─── TreeView ─────────────────────────────────────────────────────────────────

/**
 * Hierarchical tree component for SBOM dependency trees, file trees, and
 * other nested data. Supports keyboard navigation, controlled/uncontrolled
 * expand state, single or multi selection, and custom node rendering.
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — role="tree" on root, role="treeitem" on each item.
 * @wcag 2.1.1 Keyboard — ArrowUp/Down navigates, ArrowRight expands, ArrowLeft collapses, Enter selects.
 * @wcag 2.4.7 Focus Visible — focused item uses focus-visible:ring-2 outline.
 * @wcag 4.1.2 Name, Role, Value — aria-expanded, aria-selected, aria-level, aria-multiselectable present.
 */
const TreeView = <T = Record<string, unknown>>(props: TreeViewProps<T>) => {
  const {
    nodes,
    renderNode,
    defaultExpanded = [],
    expanded: controlledExpanded,
    onExpandedChange,
    className,
  } = props;

  const isMultiSelect = props.multiSelect === true;
  const isControlledExpand = controlledExpanded !== undefined;

  const [internalExpanded, setInternalExpanded] = React.useState<Set<string>>(
    () => new Set(defaultExpanded)
  );

  const expandedIds = React.useMemo<Set<string>>(
    () => (isControlledExpand ? new Set(controlledExpanded) : internalExpanded),
    [isControlledExpand, controlledExpanded, internalExpanded]
  );

  const toggleExpanded = React.useCallback(
    (id: string) => {
      const next = new Set(expandedIds);
      if (next.has(id)) {
        next.delete(id);
      } else {
        next.add(id);
      }
      if (!isControlledExpand) {
        setInternalExpanded(next);
      }
      onExpandedChange?.(Array.from(next));
    },
    [expandedIds, isControlledExpand, onExpandedChange]
  );

  const controlledSingleProp = !isMultiSelect
    ? (props as {selected?: string}).selected
    : undefined;
  const isControlledSingle = controlledSingleProp !== undefined;
  const [internalSingleSelected, setInternalSingleSelected] = React.useState<
    string | undefined
  >(undefined);

  const controlledMultiProp = isMultiSelect
    ? (props as {selected?: string[]}).selected
    : undefined;
  const isControlledMulti = controlledMultiProp !== undefined;
  const [internalMultiSelected, setInternalMultiSelected] = React.useState<
    Set<string>
  >(new Set());

  const lastSelectedRef = React.useRef<string | undefined>(undefined);

  const selectedIds = React.useMemo<Set<string>>(() => {
    if (isMultiSelect) {
      return isControlledMulti && controlledMultiProp
        ? new Set(controlledMultiProp)
        : internalMultiSelected;
    }
    const single = isControlledSingle
      ? controlledSingleProp
      : internalSingleSelected;
    return single ? new Set([single]) : new Set<string>();
  }, [
    isMultiSelect,
    isControlledMulti,
    isControlledSingle,
    controlledMultiProp,
    controlledSingleProp,
    internalMultiSelected,
    internalSingleSelected,
  ]);

  const [focusedId, setFocusedId] = React.useState<string | undefined>(
    () => nodes[0]?.id
  );

  const nodeRefs = React.useRef<Map<string, HTMLLIElement>>(new Map());

  const setRef = React.useCallback((id: string, el: HTMLLIElement | null) => {
    if (el) {
      nodeRefs.current.set(id, el);
    } else {
      nodeRefs.current.delete(id);
    }
  }, []);

  const handleSelect = React.useCallback(
    (id: string, node: TreeNode<T>, meta: SelectEventMeta) => {
      setFocusedId(id);

      if (isMultiSelect) {
        const visibleIds = getVisibleIds(nodes, expandedIds);
        const next = computeMultiSelection({
          clickedId: id,
          currentSelection: selectedIds,
          lastSelectedId: lastSelectedRef.current,
          visibleIds,
          ctrlKey: meta.ctrlKey,
          shiftKey: meta.shiftKey,
        });
        lastSelectedRef.current = id;

        if (!isControlledMulti) {
          setInternalMultiSelected(next);
        }
        (props as {onSelect?: (ids: string[]) => void}).onSelect?.(
          Array.from(next)
        );
      } else {
        lastSelectedRef.current = id;
        if (!isControlledSingle) {
          setInternalSingleSelected(id);
        }
        (
          props as {onSelect?: (id: string, node: TreeNode<T>) => void}
        ).onSelect?.(id, node);
      }
    },
    [
      isMultiSelect,
      isControlledMulti,
      isControlledSingle,
      nodes,
      expandedIds,
      selectedIds,
      props,
    ]
  );

  const handleKeyDown = React.useCallback(
    (
      e: React.KeyboardEvent,
      id: string,
      node: TreeNode<T>,
      hasChildren: boolean,
      isExpanded: boolean
    ) => {
      const visibleIds = getVisibleIds(nodes, expandedIds);
      const currentIndex = visibleIds.indexOf(id);

      switch (e.key) {
        case 'ArrowDown': {
          e.preventDefault();
          const nextId = visibleIds[currentIndex + 1];
          if (nextId) {
            setFocusedId(nextId);
            nodeRefs.current.get(nextId)?.focus();
          }
          break;
        }
        case 'ArrowUp': {
          e.preventDefault();
          const prevId = visibleIds[currentIndex - 1];
          if (prevId) {
            setFocusedId(prevId);
            nodeRefs.current.get(prevId)?.focus();
          }
          break;
        }
        case 'ArrowRight': {
          e.preventDefault();
          if (hasChildren && !isExpanded) {
            toggleExpanded(id);
          } else if (hasChildren && isExpanded) {
            const firstChildId = visibleIds[currentIndex + 1];
            if (firstChildId) {
              setFocusedId(firstChildId);
              nodeRefs.current.get(firstChildId)?.focus();
            }
          }
          break;
        }
        case 'ArrowLeft': {
          e.preventDefault();
          if (hasChildren && isExpanded) {
            toggleExpanded(id);
          } else {
            const currentEl = nodeRefs.current.get(id);
            if (currentEl) {
              const currentLevel = parseInt(
                currentEl
                  .closest('[role=treeitem]')
                  ?.getAttribute('aria-level') ?? '1',
                10
              );
              for (let i = currentIndex - 1; i >= 0; i--) {
                const candidateId = visibleIds[i];
                const candidateEl = nodeRefs.current.get(candidateId);
                const candidateLevel = parseInt(
                  candidateEl
                    ?.closest('[role=treeitem]')
                    ?.getAttribute('aria-level') ?? '1',
                  10
                );
                if (candidateLevel < currentLevel) {
                  setFocusedId(candidateId);
                  candidateEl?.focus();
                  break;
                }
              }
            }
          }
          break;
        }
        case 'Enter':
        case ' ': {
          e.preventDefault();
          handleSelect(id, node, {ctrlKey: false, shiftKey: false});
          break;
        }
        case 'Home': {
          e.preventDefault();
          const firstId = visibleIds[0];
          if (firstId) {
            setFocusedId(firstId);
            nodeRefs.current.get(firstId)?.focus();
          }
          break;
        }
        case 'End': {
          e.preventDefault();
          const lastId = visibleIds[visibleIds.length - 1];
          if (lastId) {
            setFocusedId(lastId);
            nodeRefs.current.get(lastId)?.focus();
          }
          break;
        }
      }
    },
    [nodes, expandedIds, toggleExpanded, handleSelect]
  );

  React.useEffect(() => {
    if (!focusedId) {
      setFocusedId(nodes[0]?.id);
    }
  }, [nodes, focusedId]);

  return (
    <ul
      role="tree"
      data-slot="tree-view"
      aria-multiselectable={isMultiSelect ? 'true' : undefined}
      className={cn('flex flex-col gap-0.5 p-1', className)}
    >
      {nodes.map(node => (
        <TreeViewNode<T>
          key={node.id}
          node={node}
          depth={0}
          expandedIds={expandedIds}
          selectedIds={selectedIds}
          focusedId={focusedId}
          onToggle={toggleExpanded}
          onSelect={handleSelect}
          onKeyDown={handleKeyDown}
          renderNode={renderNode}
          setRef={setRef}
        />
      ))}
    </ul>
  );
};

// ─── Exports ──────────────────────────────────────────────────────────────────

export {TreeView};
export type {TreeNode, TreeViewProps} from './tree-view-types';
