// @docker/trident/src/components/tree-view-selection.ts
// ─── Multi-select utilities for TreeView ─────────────────────────────────────

/**
 * Computes the next multi-select state given a click event on `clickedId`.
 *
 * - Plain click:    select only `clickedId` (clear others)
 * - Ctrl+click:     toggle `clickedId` without touching others
 * - Shift+click:    range-select between `lastSelectedId` and `clickedId`
 *                   in the flat visible order; adds to current selection
 */
function computeMultiSelection(opts: {
  clickedId: string;
  currentSelection: Set<string>;
  lastSelectedId: string | undefined;
  visibleIds: string[];
  ctrlKey: boolean;
  shiftKey: boolean;
}): Set<string> {
  const {
    clickedId,
    currentSelection,
    lastSelectedId,
    visibleIds,
    ctrlKey,
    shiftKey,
  } = opts;

  if (shiftKey && lastSelectedId) {
    // Range select: from lastSelectedId to clickedId (inclusive), preserving
    // any existing selection outside the range.
    const anchorIndex = visibleIds.indexOf(lastSelectedId);
    const targetIndex = visibleIds.indexOf(clickedId);

    if (anchorIndex === -1 || targetIndex === -1) {
      // Fallback: just add the clicked item
      return new Set([...currentSelection, clickedId]);
    }

    const lo = Math.min(anchorIndex, targetIndex);
    const hi = Math.max(anchorIndex, targetIndex);
    const rangeIds = visibleIds.slice(lo, hi + 1);

    return new Set([...currentSelection, ...rangeIds]);
  }

  if (ctrlKey) {
    // Toggle the clicked item
    const next = new Set(currentSelection);
    if (next.has(clickedId)) {
      next.delete(clickedId);
    } else {
      next.add(clickedId);
    }
    return next;
  }

  // Plain click — clear and select only this item
  return new Set([clickedId]);
}

/**
 * Collect all node IDs in depth-first order for visible nodes only
 * (i.e. nodes whose ancestors are all expanded).
 */
function getVisibleIds<T extends {id: string; children?: T[]}>(
  nodes: T[],
  expandedIds: Set<string>
): string[] {
  const ids: string[] = [];
  for (const node of nodes) {
    ids.push(node.id);
    if (expandedIds.has(node.id) && node.children?.length) {
      ids.push(...getVisibleIds(node.children, expandedIds));
    }
  }
  return ids;
}

/** @internal */
export {computeMultiSelection, getVisibleIds};
