// @docker/trident/src/components/trust-badge.tsx
import type {AsChild} from '@/lib/types';
import type * as React from 'react';

import {cn} from '@/lib/utils';

// trust-badge.tsx
import {cva, type VariantProps} from 'class-variance-authority';
import {Slot} from 'radix-ui';

const trustBadgeVariants = cva(
  'inline-flex items-center rounded border px-1.5 py-0.5 text-xs leading-none font-label select-none',
  {
    variants: {
      tier: {
        /** Docker Official Image — highest trust, Docker-maintained */
        official:
          'border-trust-official-border bg-trust-official-bg text-trust-official-text',
        /** Docker Hardened Image — distroless, security-hardened */
        hardened:
          'border-trust-hardened-border bg-trust-hardened-bg text-trust-hardened-text',
        /** Verified Publisher — vetted commercial/OSS publisher */
        verified:
          'border-trust-verified-border bg-trust-verified-bg text-trust-verified-text',
        /** Community — unvetted, public */
        community:
          'border-trust-community-border bg-trust-community-bg text-trust-community-text',
      },
    },
    defaultVariants: {
      tier: 'community',
    },
  }
);

const TIER_LABELS: Record<NonNullable<TrustBadgeProps['tier']>, string> = {
  official: 'Docker Official',
  hardened: 'Docker Hardened',
  verified: 'Verified Publisher',
  community: 'Community',
};

type TrustBadgeProps = React.ComponentProps<'span'> &
  VariantProps<typeof trustBadgeVariants> &
  AsChild;

function TrustBadge({
  tier,
  className,
  asChild = false,
  children,
  ...props
}: TrustBadgeProps) {
  const Comp = asChild ? Slot.Root : 'span';
  return (
    <Comp
      data-slot="trust-badge"
      className={cn(trustBadgeVariants({tier}), className)}
      {...props}
    >
      {asChild ? children : TIER_LABELS[tier ?? 'community']}
    </Comp>
  );
}

export {TrustBadge, trustBadgeVariants};
export type {TrustBadgeProps};
