// @docker/trident/src/components/virtual-list.tsx
import type * as React from 'react';

import {Skeleton} from '@/components/ui/skeleton';
import {cn} from '@/lib/utils';

import {useVirtualizer} from '@tanstack/react-virtual';
import {useRef} from 'react';

/**
 * Props for VirtualList. All native `<div>` attributes (role, aria-*, tabIndex,
 * id, etc.) are accepted and forwarded to the outer scroll container, giving
 * consumers full control over landmark semantics.
 *
 * By default, no `role` or `aria-label` is applied — pass them explicitly when
 * the list needs to be a named landmark (e.g. role="region" aria-label="…").
 */
type VirtualListProps<T> = {
  /** The array of items to render. */
  items: T[];
  /** Estimated row height in px for initial layout. Default 48. */
  estimateSize?: number;
  /** Extra rows rendered above/below the visible window. Default 5. */
  overscan?: number;
  /** Container height. Default 400. */
  height?: number | string;
  /** Renders a single item row. */
  renderItem: (item: T, index: number) => React.ReactNode;
  /** Stable key for each item. Defaults to index. */
  getItemKey?: (item: T, index: number) => string | number;
  /** Shown when items.length === 0 and loading is false. */
  emptyState?: React.ReactNode;
  /** Shown as a slot above the list rows. */
  header?: React.ReactNode;
  /** Shown as a slot below the list rows. */
  footer?: React.ReactNode;
  /** When true, renders skeleton placeholder rows instead of items. */
  loading?: boolean;
  /** Number of skeleton rows to show while loading. Default 6. */
  loadingRows?: number;
} & Omit<React.ComponentProps<'div'>, 'children'>;

const DEFAULT_ESTIMATE_SIZE = 48;
const DEFAULT_OVERSCAN = 5;
const DEFAULT_HEIGHT = 400;
const DEFAULT_LOADING_ROWS = 6;

function VirtualList<T>({
  items,
  estimateSize = DEFAULT_ESTIMATE_SIZE,
  overscan = DEFAULT_OVERSCAN,
  height = DEFAULT_HEIGHT,
  renderItem,
  getItemKey,
  className,
  emptyState,
  header,
  footer,
  loading = false,
  loadingRows = DEFAULT_LOADING_ROWS,
  style,
  ...divProps
}: VirtualListProps<T>) {
  const scrollRef = useRef<HTMLDivElement>(null);

  const virtualizer = useVirtualizer({
    count: items.length,
    getScrollElement: () => scrollRef.current,
    estimateSize: () => estimateSize,
    overscan,
    getItemKey: getItemKey
      ? index => getItemKey(items[index], index)
      : undefined,
  });

  const containerStyle: React.CSSProperties = {
    height: typeof height === 'number' ? `${height}px` : height,
    ...style,
  };

  if (loading) {
    return (
      <div
        data-slot="virtual-list"
        tabIndex={0}
        {...divProps}
        className={cn('overflow-y-auto', className)}
        style={containerStyle}
      >
        {header}
        <div className="flex flex-col gap-2 p-2">
          {Array.from({length: loadingRows}).map((_, i) => (
            <Skeleton key={i} className="h-10 w-full rounded" />
          ))}
        </div>
        {footer}
      </div>
    );
  }

  if (items.length === 0) {
    return (
      <div
        data-slot="virtual-list"
        tabIndex={0}
        {...divProps}
        className={cn('overflow-y-auto', className)}
        style={containerStyle}
      >
        {header}
        <div className="flex h-full items-center justify-center text-sm text-muted-foreground">
          {emptyState ?? 'No items'}
        </div>
        {footer}
      </div>
    );
  }

  const virtualItems = virtualizer.getVirtualItems();
  const totalSize = virtualizer.getTotalSize();

  return (
    <div
      data-slot="virtual-list"
      tabIndex={0}
      {...divProps}
      ref={scrollRef}
      className={cn('overflow-y-auto', className)}
      style={containerStyle}
    >
      {header}
      {/* Inner container: must be exactly the total virtual height */}
      <div style={{position: 'relative', height: `${totalSize}px`}}>
        {virtualItems.map(virtualRow => (
          <div
            key={virtualRow.key}
            data-index={virtualRow.index}
            ref={virtualizer.measureElement}
            style={{
              position: 'absolute',
              top: 0,
              left: 0,
              width: '100%',
              transform: `translateY(${virtualRow.start}px)`,
            }}
          >
            {renderItem(items[virtualRow.index], virtualRow.index)}
          </div>
        ))}
      </div>
      {footer}
    </div>
  );
}

export {VirtualList};
export type {VirtualListProps};
