// @docker/trident/src/components/ui/alert.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';

const alertVariants = cva(
  "group/alert relative grid w-full gap-0.5 rounded border px-4 py-3 text-left text-sm has-data-[slot=alert-action]:relative has-data-[slot=alert-action]:pr-18 has-[>svg]:grid-cols-[auto_1fr] has-[>svg]:gap-x-2.5 *:[svg]:row-span-2 *:[svg]:translate-y-0.5 *:[svg]:text-current *:[svg:not([class*='size-'])]:size-4",
  {
    variants: {
      variant: {
        default: 'bg-card text-card-foreground',
        destructive:
          'bg-alert-error-bg text-alert-error-fg *:data-[slot=alert-description]:text-alert-error-fg *:[svg]:text-alert-error-icon',
        success:
          'bg-alert-success-bg text-alert-success-fg *:data-[slot=alert-description]:text-alert-success-fg *:[svg]:text-alert-success-icon',
        warning:
          'bg-alert-warning-bg text-alert-warning-fg *:data-[slot=alert-description]:text-alert-warning-fg *:[svg]:text-alert-warning-icon',
        info: 'bg-alert-info-bg text-alert-info-fg *:data-[slot=alert-description]:text-alert-info-fg *:[svg]:text-alert-info-icon',
      },
    },
    defaultVariants: {
      variant: 'default',
    },
  }
);

/**
 * Displays a callout for important messages, warnings, or contextual information.
 *
 * ### Component Props
 *
 * | Prop       | Type                         | Default     |
 * | :--------- | :--------------------------- | :---------- |
 * | _variant_  | `"default" \| "destructive" \| "success" \| "warning" \| "info"` | `"default"` |
 *
 *
 * ### Example Usage
 *
 * **Default**
 * ```tsx
 * <Alert>
 *   <AlertTitle>Heads up!</AlertTitle>
 *   <AlertDescription>
 *     You can add components to your app using the cli.
 *   </AlertDescription>
 * </Alert>
 * ```
 *
 * **Destructive**
 * ```tsx
 * <Alert variant="destructive">
 *   <AlertTitle>Error</AlertTitle>
 *   <AlertDescription>Something went wrong.</AlertDescription>
 * </Alert>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 4.1.2 Name, Role, Value — uses `role="alert"` to convey importance to assistive technology.
 * @wcag 1.4.3 Contrast (Minimum) — text colors use DDS semantic tokens meeting 4.5:1 contrast.
 */
function Alert({
  className,
  variant,
  ...props
}: React.ComponentProps<'div'> & VariantProps<typeof alertVariants>) {
  return (
    <div
      data-slot="alert"
      role="alert"
      className={cn(alertVariants({variant}), className)}
      {...props}
    />
  );
}

/**
 * The title of an alert message. Renders as a styled `div`.
 *
 * ### Component Props
 *
 * This component extends the HTML `div` element.
 * It does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertTitle>Heads up!</AlertTitle>
 * ```
 */
function AlertTitle({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="alert-title"
      className={cn(
        'leading-heading font-heading tracking-heading group-has-[>svg]/alert:col-start-2 [&_a]:underline [&_a]:underline-offset-3 [&_a]:hover:text-foreground',
        className
      )}
      {...props}
    />
  );
}

/**
 * The description text of an alert message. Renders as a styled `div`.
 *
 * ### Component Props
 *
 * This component extends the HTML `div` element.
 * It does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDescription>
 *   You can add components to your app using the cli.
 * </AlertDescription>
 * ```
 */
function AlertDescription({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="alert-description"
      className={cn(
        'text-sm text-balance text-muted-foreground md:text-pretty [&_a]:underline [&_a]:underline-offset-3 [&_a]:hover:text-foreground [&_p:not(:last-child)]:mb-4',
        className
      )}
      {...props}
    />
  );
}

/**
 * A container for an action element positioned at the top-right of an alert.
 *
 * ### Component Props
 *
 * This component extends the HTML `div` element.
 * It does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Alert>
 *   <AlertTitle>New update available</AlertTitle>
 *   <AlertAction>
 *     <Button variant="outline" size="sm">Update</Button>
 *   </AlertAction>
 * </Alert>
 * ```
 */
function AlertAction({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="alert-action"
      className={cn('absolute top-2.5 right-3', className)}
      {...props}
    />
  );
}

export {Alert, AlertTitle, AlertDescription, AlertAction};
