// @docker/trident/src/components/ui/alert-dialog.tsx
import type * as React from 'react';

import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import {AlertDialog as AlertDialogPrimitive} from 'radix-ui';

/**
 * A modal dialog that interrupts the user with important content and expects a response.
 *
 * ### Component Props
 *
 * See [AlertDialog props](https://www.radix-ui.com/primitives/docs/components/alert-dialog) documentation for full details.
 *
 * | Prop             | Type                            | Default |
 * | :--------------- | :------------------------------ | :------ |
 * | _defaultOpen_    | `boolean`                       | -       |
 * | _open_           | `boolean`                       | -       |
 * | _onOpenChange_   | `(open: boolean) => void`       | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialog>
 *   <AlertDialogTrigger>Open</AlertDialogTrigger>
 *   <AlertDialogContent>
 *     <AlertDialogHeader>
 *       <AlertDialogTitle>Are you sure?</AlertDialogTitle>
 *       <AlertDialogDescription>
 *         This action cannot be undone.
 *       </AlertDialogDescription>
 *     </AlertDialogHeader>
 *     <AlertDialogFooter>
 *       <AlertDialogCancel>Cancel</AlertDialogCancel>
 *       <AlertDialogAction>Continue</AlertDialogAction>
 *     </AlertDialogFooter>
 *   </AlertDialogContent>
 * </AlertDialog>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 2.1.2 No Keyboard Trap — Escape key closes the dialog; focus returns to trigger.
 * @wcag 2.4.3 Focus Order — focus is trapped within the dialog while open via Radix AlertDialog primitive.
 * @wcag 4.1.2 Name, Role, Value — `alertdialog` role with `aria-labelledby` / `aria-describedby` via Radix primitives.
 */
function AlertDialog({
  ...props
}: React.ComponentProps<typeof AlertDialogPrimitive.Root>) {
  return <AlertDialogPrimitive.Root data-slot="alert-dialog" {...props} />;
}

/**
 * A button that opens the alert dialog.
 *
 * ### Component Props
 *
 * See [AlertDialog props](https://www.radix-ui.com/primitives/docs/components/alert-dialog) documentation for full details.
 *
 * | Prop       | Type      | Default |
 * | :--------- | :-------- | :------ |
 * | _asChild_  | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogTrigger>Open Dialog</AlertDialogTrigger>
 * ```
 */
function AlertDialogTrigger({
  ...props
}: React.ComponentProps<typeof AlertDialogPrimitive.Trigger>) {
  return (
    <AlertDialogPrimitive.Trigger data-slot="alert-dialog-trigger" {...props} />
  );
}

/**
 * Portals the overlay and content parts into the `body`.
 *
 * ### Component Props
 *
 * See [AlertDialog props](https://www.radix-ui.com/primitives/docs/components/alert-dialog) documentation for full details.
 *
 * | Prop          | Type          | Default         |
 * | :------------ | :------------ | :-------------- |
 * | _forceMount_  | `boolean`     | -               |
 * | _container_   | `HTMLElement`  | `document.body` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogPortal>
 *   <AlertDialogOverlay />
 *   <AlertDialogContent>...</AlertDialogContent>
 * </AlertDialogPortal>
 * ```
 */
function AlertDialogPortal({
  ...props
}: React.ComponentProps<typeof AlertDialogPrimitive.Portal>) {
  return (
    <AlertDialogPrimitive.Portal data-slot="alert-dialog-portal" {...props} />
  );
}

/**
 * A layer that covers the inert portion of the view when the dialog is open.
 *
 * ### Component Props
 *
 * See [AlertDialog props](https://www.radix-ui.com/primitives/docs/components/alert-dialog) documentation for full details.
 *
 * | Prop          | Type      | Default |
 * | :------------ | :-------- | :------ |
 * | _asChild_     | `boolean` | `false` |
 * | _forceMount_  | `boolean` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogOverlay />
 * ```
 */
function AlertDialogOverlay({
  className,
  ...props
}: React.ComponentProps<typeof AlertDialogPrimitive.Overlay>) {
  return (
    <AlertDialogPrimitive.Overlay
      data-slot="alert-dialog-overlay"
      className={cn(
        'fixed inset-0 z-modal bg-dds-overlay-backdrop duration-fast ease-out supports-backdrop-filter:backdrop-blur-xs data-open:animate-in data-open:fade-in-0 data-closed:animate-out data-closed:fade-out-0',
        className
      )}
      {...props}
    />
  );
}

/**
 * Contains the content to be rendered when the alert dialog is open.
 *
 * ### Component Props
 *
 * See [AlertDialog props](https://www.radix-ui.com/primitives/docs/components/alert-dialog) documentation for full details.
 *
 * | Prop                | Type                             | Default     |
 * | :------------------ | :------------------------------- | :---------- |
 * | _asChild_           | `boolean`                        | `false`     |
 * | _forceMount_        | `boolean`                        | -           |
 * | _onOpenAutoFocus_   | `(event: Event) => void`         | -           |
 * | _onCloseAutoFocus_  | `(event: Event) => void`         | -           |
 * | _onEscapeKeyDown_   | `(event: KeyboardEvent) => void` | -           |
 * | _size_              | `"default" \| "sm"`              | `"default"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogContent>
 *   <AlertDialogHeader>
 *     <AlertDialogTitle>Title</AlertDialogTitle>
 *     <AlertDialogDescription>Description</AlertDialogDescription>
 *   </AlertDialogHeader>
 *   <AlertDialogFooter>
 *     <AlertDialogCancel>Cancel</AlertDialogCancel>
 *     <AlertDialogAction>Continue</AlertDialogAction>
 *   </AlertDialogFooter>
 * </AlertDialogContent>
 * ```
 */
function AlertDialogContent({
  className,
  size = 'default',
  ...props
}: React.ComponentProps<typeof AlertDialogPrimitive.Content> & {
  size?: 'default' | 'sm';
}) {
  return (
    <AlertDialogPortal>
      <AlertDialogOverlay />
      <AlertDialogPrimitive.Content
        data-slot="alert-dialog-content"
        data-size={size}
        className={cn(
          'group/alert-dialog-content fixed top-1/2 left-1/2 z-modal grid w-full -translate-x-1/2 -translate-y-1/2 gap-6 rounded-lg bg-background p-6 shadow-modal ring-1 ring-border duration-fast ease-out outline-none data-[size=default]:max-w-xs data-[size=sm]:max-w-xs data-[size=default]:sm:max-w-lg data-open:animate-in data-open:fade-in-0 data-open:zoom-in-95 data-closed:animate-out data-closed:fade-out-0 data-closed:zoom-out-95',
          className
        )}
        {...props}
      />
    </AlertDialogPortal>
  );
}

/**
 * A container for the title and description of the alert dialog.
 *
 * ### Component Props
 *
 * This is a custom layout component that extends the HTML `div` element.
 * It does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogHeader>
 *   <AlertDialogTitle>Are you sure?</AlertDialogTitle>
 *   <AlertDialogDescription>This action cannot be undone.</AlertDialogDescription>
 * </AlertDialogHeader>
 * ```
 */
function AlertDialogHeader({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="alert-dialog-header"
      className={cn(
        'grid grid-rows-[auto_1fr] place-items-center gap-1.5 text-center has-data-[slot=alert-dialog-media]:grid-rows-[auto_auto_1fr] has-data-[slot=alert-dialog-media]:gap-x-6 sm:group-data-[size=default]/alert-dialog-content:place-items-start sm:group-data-[size=default]/alert-dialog-content:text-left sm:group-data-[size=default]/alert-dialog-content:has-data-[slot=alert-dialog-media]:grid-rows-[auto_1fr]',
        className
      )}
      {...props}
    />
  );
}

/**
 * A container for the action and cancel buttons of the alert dialog.
 *
 * ### Component Props
 *
 * This is a custom layout component that extends the HTML `div` element.
 * It does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogFooter>
 *   <AlertDialogCancel>Cancel</AlertDialogCancel>
 *   <AlertDialogAction>Continue</AlertDialogAction>
 * </AlertDialogFooter>
 * ```
 */
function AlertDialogFooter({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="alert-dialog-footer"
      className={cn(
        'flex flex-col-reverse gap-2 group-data-[size=sm]/alert-dialog-content:grid group-data-[size=sm]/alert-dialog-content:grid-cols-2 sm:flex-row sm:justify-end',
        className
      )}
      {...props}
    />
  );
}

/**
 * An optional media container for displaying an icon or image in the alert dialog header.
 *
 * ### Component Props
 *
 * This is a custom layout component that extends the HTML `div` element.
 * It does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogHeader>
 *   <AlertDialogMedia>
 *     <WarningIcon />
 *   </AlertDialogMedia>
 *   <AlertDialogTitle>Are you sure?</AlertDialogTitle>
 * </AlertDialogHeader>
 * ```
 */
function AlertDialogMedia({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="alert-dialog-media"
      className={cn(
        "mb-2 inline-flex size-16 items-center justify-center rounded bg-muted sm:group-data-[size=default]/alert-dialog-content:row-span-2 *:[svg:not([class*='size-'])]:size-8",
        className
      )}
      {...props}
    />
  );
}

/**
 * An accessible name to be announced when the alert dialog is opened.
 *
 * ### Component Props
 *
 * See [AlertDialog props](https://www.radix-ui.com/primitives/docs/components/alert-dialog) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogTitle>Are you absolutely sure?</AlertDialogTitle>
 * ```
 */
function AlertDialogTitle({
  className,
  ...props
}: React.ComponentProps<typeof AlertDialogPrimitive.Title>) {
  return (
    <AlertDialogPrimitive.Title
      data-slot="alert-dialog-title"
      className={cn(
        'text-lg leading-heading font-heading tracking-heading sm:group-data-[size=default]/alert-dialog-content:group-has-data-[slot=alert-dialog-media]/alert-dialog-content:col-start-2',
        className
      )}
      {...props}
    />
  );
}

/**
 * An accessible description to be announced when the alert dialog is opened.
 *
 * ### Component Props
 *
 * See [AlertDialog props](https://www.radix-ui.com/primitives/docs/components/alert-dialog) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogDescription>
 *   This action cannot be undone. This will permanently delete your account.
 * </AlertDialogDescription>
 * ```
 */
function AlertDialogDescription({
  className,
  ...props
}: React.ComponentProps<typeof AlertDialogPrimitive.Description>) {
  return (
    <AlertDialogPrimitive.Description
      data-slot="alert-dialog-description"
      className={cn(
        'text-sm text-balance text-muted-foreground md:text-pretty *:[a]:underline *:[a]:underline-offset-3 *:[a]:hover:text-foreground',
        className
      )}
      {...props}
    />
  );
}

/**
 * A button that confirms the alert dialog action and closes the dialog.
 *
 * ### Component Props
 *
 * See [AlertDialog props](https://www.radix-ui.com/primitives/docs/components/alert-dialog) documentation for full details.
 *
 * | Prop       | Type                                                                            | Default     |
 * | :--------- | :------------------------------------------------------------------------------ | :---------- |
 * | _asChild_  | `boolean`                                                                       | `false`     |
 * | _variant_  | `"default" \| "outline" \| "secondary" \| "ghost" \| "destructive" \| "link"`   | `"default"` |
 * | _size_     | `"default" \| "xs" \| "sm" \| "lg" \| "icon" \| "icon-xs" \| "icon-sm" \| "icon-lg"` | `"default"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogAction>Continue</AlertDialogAction>
 * ```
 */
function AlertDialogAction({
  className,
  variant = 'default',
  size = 'default',
  ...props
}: React.ComponentProps<typeof AlertDialogPrimitive.Action> &
  Pick<React.ComponentProps<typeof Button>, 'variant' | 'size'>) {
  return (
    <Button variant={variant} size={size} asChild>
      <AlertDialogPrimitive.Action
        data-slot="alert-dialog-action"
        className={cn(className)}
        {...props}
      />
    </Button>
  );
}

/**
 * A button that cancels the alert dialog action and closes the dialog.
 *
 * ### Component Props
 *
 * See [AlertDialog props](https://www.radix-ui.com/primitives/docs/components/alert-dialog) documentation for full details.
 *
 * | Prop       | Type                                                                            | Default     |
 * | :--------- | :------------------------------------------------------------------------------ | :---------- |
 * | _asChild_  | `boolean`                                                                       | `false`     |
 * | _variant_  | `"default" \| "outline" \| "secondary" \| "ghost" \| "destructive" \| "link"`   | `"outline"` |
 * | _size_     | `"default" \| "xs" \| "sm" \| "lg" \| "icon" \| "icon-xs" \| "icon-sm" \| "icon-lg"` | `"default"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AlertDialogCancel>Cancel</AlertDialogCancel>
 * ```
 */
function AlertDialogCancel({
  className,
  variant = 'outline',
  size = 'default',
  ...props
}: React.ComponentProps<typeof AlertDialogPrimitive.Cancel> &
  Pick<React.ComponentProps<typeof Button>, 'variant' | 'size'>) {
  return (
    <Button variant={variant} size={size} asChild>
      <AlertDialogPrimitive.Cancel
        data-slot="alert-dialog-cancel"
        className={cn(className)}
        {...props}
      />
    </Button>
  );
}

export {
  AlertDialog,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogContent,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogMedia,
  AlertDialogOverlay,
  AlertDialogPortal,
  AlertDialogTitle,
  AlertDialogTrigger,
};
