// @docker/trident/src/components/ui/avatar.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {Avatar as AvatarPrimitive} from 'radix-ui';

/**
 * An image element with a fallback for representing the user.
 *
 * ### Component Props
 *
 * See [Avatar props](https://www.radix-ui.com/primitives/docs/components/avatar) documentation for full details.
 *
 * | Prop      | Type                         | Default     |
 * | :-------- | :--------------------------- | :---------- |
 * | _asChild_ | `boolean`                    | `false`     |
 * | _size_    | `"default" \| "sm" \| "lg"` | `"default"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Avatar>
 *   <AvatarImage src="/avatar.png" alt="User" />
 *   <AvatarFallback>JD</AvatarFallback>
 * </Avatar>
 * ```
 */
function Avatar({
  className,
  size = 'default',
  ...props
}: React.ComponentProps<typeof AvatarPrimitive.Root> & {
  size?: 'default' | 'sm' | 'lg';
}) {
  return (
    <AvatarPrimitive.Root
      data-slot="avatar"
      data-size={size}
      className={cn(
        'group/avatar relative flex size-8 shrink-0 rounded-full select-none after:absolute after:inset-0 after:rounded-full after:border after:border-border after:mix-blend-darken data-[size=lg]:size-10 data-[size=sm]:size-6 dark:after:mix-blend-lighten',
        className
      )}
      {...props}
    />
  );
}

/**
 * The image to render inside the avatar. It will only render when loaded.
 *
 * ### Component Props
 *
 * See [Avatar props](https://www.radix-ui.com/primitives/docs/components/avatar) documentation for full details.
 *
 * | Prop                     | Type                                                      | Default |
 * | :----------------------- | :-------------------------------------------------------- | :------ |
 * | _asChild_                | `boolean`                                                 | `false` |
 * | _onLoadingStatusChange_  | `(status: "idle" \| "loading" \| "loaded" \| "error") => void` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AvatarImage src="/avatar.png" alt="User" />
 * ```
 */
function AvatarImage({
  className,
  ...props
}: React.ComponentProps<typeof AvatarPrimitive.Image>) {
  return (
    <AvatarPrimitive.Image
      data-slot="avatar-image"
      className={cn(
        'aspect-square size-full rounded-full object-cover',
        className
      )}
      {...props}
    />
  );
}

/**
 * An element that renders when the image has not loaded. This can be initials or an icon.
 *
 * ### Component Props
 *
 * See [Avatar props](https://www.radix-ui.com/primitives/docs/components/avatar) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 * | _delayMs_   | `number`  | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AvatarFallback>JD</AvatarFallback>
 * ```
 */
function AvatarFallback({
  className,
  ...props
}: React.ComponentProps<typeof AvatarPrimitive.Fallback>) {
  return (
    <AvatarPrimitive.Fallback
      data-slot="avatar-fallback"
      className={cn(
        'flex size-full items-center justify-center rounded-full bg-muted text-sm text-muted-foreground group-data-[size=sm]/avatar:text-xs',
        className
      )}
      {...props}
    />
  );
}

/**
 * A small indicator displayed at the bottom-right corner of an avatar, typically for status or notification dots.
 *
 * ### Component Props
 *
 * This component extends the HTML `span` element.
 * It does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Avatar>
 *   <AvatarImage src="/avatar.png" alt="User" />
 *   <AvatarFallback>JD</AvatarFallback>
 *   <AvatarBadge />
 * </Avatar>
 * ```
 */
function AvatarBadge({className, ...props}: React.ComponentProps<'span'>) {
  return (
    <span
      data-slot="avatar-badge"
      className={cn(
        'absolute right-0 bottom-0 z-raised inline-flex items-center justify-center rounded-full bg-primary text-primary-foreground bg-blend-color ring-2 ring-background select-none',
        'group-data-[size=sm]/avatar:size-2 group-data-[size=sm]/avatar:[&>svg]:hidden',
        'group-data-[size=default]/avatar:size-2.5 group-data-[size=default]/avatar:[&>svg]:size-2',
        'group-data-[size=lg]/avatar:size-3 group-data-[size=lg]/avatar:[&>svg]:size-2',
        className
      )}
      {...props}
    />
  );
}

/**
 * A container that displays multiple avatars in a stacked, overlapping layout.
 *
 * ### Component Props
 *
 * This component extends the HTML `div` element.
 * It does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AvatarGroup>
 *   <Avatar>
 *     <AvatarImage src="/user1.png" alt="User 1" />
 *     <AvatarFallback>U1</AvatarFallback>
 *   </Avatar>
 *   <Avatar>
 *     <AvatarImage src="/user2.png" alt="User 2" />
 *     <AvatarFallback>U2</AvatarFallback>
 *   </Avatar>
 *   <AvatarGroupCount>+3</AvatarGroupCount>
 * </AvatarGroup>
 * ```
 */
function AvatarGroup({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="avatar-group"
      className={cn(
        'group/avatar-group flex -space-x-2 *:data-[slot=avatar]:ring-2 *:data-[slot=avatar]:ring-background',
        className
      )}
      {...props}
    />
  );
}

/**
 * A count indicator used within an `AvatarGroup` to show the number of additional avatars not displayed.
 *
 * ### Component Props
 *
 * This component extends the HTML `div` element.
 * It does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <AvatarGroupCount>+3</AvatarGroupCount>
 * ```
 */
function AvatarGroupCount({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="avatar-group-count"
      className={cn(
        'relative flex size-8 shrink-0 items-center justify-center rounded-full bg-muted text-sm text-muted-foreground ring-2 ring-background group-has-data-[size=lg]/avatar-group:size-10 group-has-data-[size=sm]/avatar-group:size-6 [&>svg]:size-4 group-has-data-[size=lg]/avatar-group:[&>svg]:size-5 group-has-data-[size=sm]/avatar-group:[&>svg]:size-3',
        className
      )}
      {...props}
    />
  );
}

export {
  Avatar,
  AvatarImage,
  AvatarFallback,
  AvatarGroup,
  AvatarGroupCount,
  AvatarBadge,
};
