// @docker/trident/src/components/ui/badge.tsx
import type {AsChild} from '@/lib/types';
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {Slot} from 'radix-ui';

const badgeVariants = cva(
  "group/badge inline-flex w-fit shrink-0 items-center justify-center gap-1 overflow-hidden rounded-full border border-transparent py-0.5 font-label tracking-label whitespace-nowrap transition-all ease-default outline-none focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring has-data-[icon=inline-end]:pr-1.5 has-data-[icon=inline-start]:pl-1.5 aria-invalid:border-destructive aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 [&>svg]:pointer-events-none [&>svg:not([class*='size-'])]:size-3",
  {
    variants: {
      variant: {
        default: 'bg-primary text-primary-foreground [a]:hover:bg-accent-hover',
        secondary:
          'bg-secondary text-secondary-foreground [a]:hover:bg-secondary-hover',
        premium:
          'bg-premium text-premium-foreground focus-visible:border-premium focus-visible:ring-premium [a]:hover:bg-premium-hover',
        destructive:
          'bg-destructive text-destructive-foreground focus-visible:border-destructive focus-visible:ring-destructive [a]:hover:bg-destructive-hover',
        success:
          'bg-success text-success-foreground focus-visible:border-success focus-visible:ring-success [a]:hover:bg-success-hover',
        warning:
          'bg-warning text-warning-foreground focus-visible:border-warning focus-visible:ring-warning [a]:hover:bg-warning-hover',
        outline:
          'border-border text-foreground [a]:hover:bg-muted [a]:hover:text-muted-foreground',
        ghost:
          'text-foreground hover:bg-muted hover:text-muted-foreground dark:hover:bg-hover-overlay',
        link: 'text-primary underline-offset-4 hover:underline',
      },
      size: {
        default: 'h-5 px-2 text-xs',
        sm: 'h-4 px-1.5 text-2xs',
        lg: 'h-6 px-3 text-sm',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

type BadgeProps = React.ComponentProps<'span'> &
  AsChild & {
    /**
     * The visual style of the badge.
     * @default "default"
     * @example
     * <Badge variant="secondary">Beta</Badge>
     */
    variant?: VariantProps<typeof badgeVariants>['variant'];
    /**
     * The size of the badge.
     * @default "default"
     * @example
     * <Badge size="sm">New</Badge>
     */
    size?: VariantProps<typeof badgeVariants>['size'];
  };

/**
 * Displays a badge or a component that looks like a badge.
 *
 * ### Component Props
 *
 * See [Badge props](https://ui.shadcn.com/docs/components/radix/badge) documentation for full details.
 *
 * | Prop      | Type                                                                                               | Default     |
 * | :-------- | :------------------------------------------------------------------------------------------------- | :---------- |
 * | _variant_ | `"default" \| "secondary" \| "premium" \| "destructive" \| "success" \| "warning" \| "outline" \| "ghost" \| "link"` | `"default"` |
 * | _size_    | `"default" \| "sm" \| "lg"`                                                  | `"default"` |
 * | _asChild_ | `boolean`                                                                    | `false`     |
 *
 *
 * ### Example Usage
 *
 * **Default**
 * ```tsx
 * <Badge>Default</Badge>
 * ```
 *
 * **Secondary**
 * ```tsx
 * <Badge variant="secondary">Beta</Badge>
 * ```
 *
 * **Destructive**
 * ```tsx
 * <Badge variant="destructive">Error</Badge>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.4.1 Use of Color — badge variants use both color and text/icon to convey meaning.
 * @wcag 1.4.3 Contrast (Minimum) — badge text meets 4.5:1 contrast via DDS semantic tokens.
 * @wcag 4.1.2 Name, Role, Value — badges are presentational; wrap in a `<span aria-label>` if the badge is the only indicator.
 */
function Badge({
  className,
  variant = 'default',
  size = 'default',
  asChild = false,
  ...props
}: BadgeProps) {
  const Comp = asChild ? Slot.Root : 'span';

  return (
    <Comp
      data-slot="badge"
      data-variant={variant}
      className={cn(badgeVariants({variant, size}), className)}
      {...props}
    />
  );
}

export {Badge, badgeVariants};
