// @docker/trident/src/components/ui/button.tsx
import type {AsChild} from '@/lib/types';
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {Loader2Icon} from 'lucide-react';
import {Slot} from 'radix-ui';

const buttonVariants = cva(
  "group/button inline-flex shrink-0 items-center justify-center rounded border border-transparent bg-clip-padding text-sm font-label tracking-label whitespace-nowrap transition-all ease-default outline-none select-none focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 aria-invalid:border-destructive aria-invalid:ring-3 aria-invalid:ring-destructive/20 dark:aria-invalid:border-destructive/50 dark:aria-invalid:ring-destructive/40 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
  {
    variants: {
      variant: {
        default:
          'dds-shimmer bg-primary text-primary-foreground dds-shimmer-glow hover:bg-accent-hover',
        outline:
          'border-border bg-background shadow-raised hover:bg-muted hover:text-foreground aria-expanded:bg-muted aria-expanded:text-foreground',
        secondary:
          'bg-secondary text-secondary-foreground hover:bg-secondary-hover aria-expanded:bg-secondary aria-expanded:text-secondary-foreground',
        ghost:
          'hover:bg-muted hover:text-foreground aria-expanded:bg-muted aria-expanded:text-foreground dark:hover:bg-hover-overlay',
        destructive:
          'dds-shimmer bg-destructive text-destructive-foreground dds-shimmer-glow-destructive hover:bg-destructive-hover focus-visible:border-destructive-hover focus-visible:ring-destructive/50',
        link: 'text-primary underline-offset-4 hover:underline',
      },
      size: {
        default:
          'h-9 gap-1.5 px-2.5 in-data-[slot=button-group]:rounded has-data-[icon=inline-end]:pr-2 has-data-[icon=inline-start]:pl-2',
        xs: "h-6 gap-1 rounded px-2 text-xs in-data-[slot=button-group]:rounded has-data-[icon=inline-end]:pr-1.5 has-data-[icon=inline-start]:pl-1.5 [&_svg:not([class*='size-'])]:size-3",
        sm: 'h-8 gap-1 rounded px-2.5 in-data-[slot=button-group]:rounded has-data-[icon=inline-end]:pr-1.5 has-data-[icon=inline-start]:pl-1.5',
        lg: 'h-10 gap-1.5 px-2.5 has-data-[icon=inline-end]:pr-3 has-data-[icon=inline-start]:pl-3',
        icon: 'size-9',
        'icon-xs':
          "relative size-6 rounded after:absolute after:-inset-1 in-data-[slot=button-group]:rounded [&_svg:not([class*='size-'])]:size-3",
        'icon-sm': 'size-8 rounded in-data-[slot=button-group]:rounded',
        'icon-lg': 'size-10',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

type ButtonProps = React.ComponentProps<'button'> &
  AsChild & {
    /**
     * The style of button.
     *
     * @default "default"
     */
    variant?: VariantProps<typeof buttonVariants>['variant'];
    /**
     * The size of the button.
     *
     * Use `icon`, `icon-xs`, `icon-sm`, or `icon-lg` when the button only
     * contains an icon. These sizes create a square button and scale the icon.
     *
     * @default "default"
     *
     * @example
     * <Button size="icon" aria-label="Search">
     *   <SearchIcon />
     * </Button>
     */
    size?: VariantProps<typeof buttonVariants>['size'];
    /**
     * Shows a spinner and disables the button. Replaces `@mui/lab/LoadingButton`.
     *
     * @default false
     *
     * @example
     * <Button loading>Saving…</Button>
     */
    loading?: boolean;
  };

/**
 * Displays a button or a component that looks like a button.
 *
 * ### Component Props
 *
 * See [Button props](https://ui.shadcn.com/docs/components/radix/button) documentation for full details.
 *
 * | Prop      | Type                                                                          | Default     |
 * | :-------- | :---------------------------------------------------------------------------- | :---------- |
 * | _variant_ | `"default" \| "outline" \| "ghost" \| "destructive" \| "secondary" \| "link"` | `"default"` |
 * | _size_    | `"default" \| "xs" \| "sm" \| "lg" \| "icon" \| "icon-xs" \| "icon-sm" \| "icon-lg"` | `"default"` |
 * | _loading_ | `boolean`                                                                     | `false`     |
 * | _asChild_ | `boolean`                                                                     | `false`     |
 *
 *
 * ### Example Usage
 *
 * **Default**
 * ```tsx
 * <Button>Default</Button>
 * ```
 *
 * **Outline + Small**
 * ```tsx
 * <Button variant="outline" size="sm">Click me</Button>
 * ```
 *
 * **Destructive**
 * ```tsx
 * <Button variant="destructive">Delete</Button>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.4.3 Contrast (Minimum) — button text colors meet 4.5:1 against their backgrounds via DDS semantic tokens.
 * @wcag 2.1.1 Keyboard — button is natively focusable and activated by Enter/Space.
 * @wcag 2.4.7 Focus Visible — focus ring uses `focus-visible:ring-3` visible in both light and dark mode.
 * @wcag 4.1.2 Name, Role, Value — uses native `<button>` element; accessible name comes from text content or `aria-label`.
 */
function Button({
  className,
  variant = 'default',
  size = 'default',
  asChild = false,
  loading = false,
  disabled,
  children,
  ...props
}: ButtonProps) {
  const Comp = asChild ? Slot.Root : 'button';

  return (
    <Comp
      data-slot="button"
      data-variant={variant}
      data-size={size}
      disabled={disabled || loading}
      aria-busy={loading || undefined}
      className={cn(buttonVariants({variant, size, className}))}
      {...props}
    >
      {loading ? (
        <>
          <Loader2Icon className="animate-spin" data-slot="button-spinner" />
          <span className="inline-flex items-center gap-1.5">{children}</span>
        </>
      ) : (
        children
      )}
    </Comp>
  );
}

export {Button, buttonVariants};
export type {ButtonProps};
