// @docker/trident/src/components/ui/button-group.tsx
import type {AsChild} from '@/lib/types';

import {Separator} from '@/components/ui/separator';
import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {Slot} from 'radix-ui';

const buttonGroupVariants = cva(
  "flex w-fit items-stretch has-[>[data-slot=button-group]]:gap-2 [&>*]:focus-visible:relative [&>*]:focus-visible:z-raised has-[select[aria-hidden=true]:last-child]:[&>[data-slot=select-trigger]:last-of-type]:rounded-r [&>[data-slot=select-trigger]:not([class*='w-'])]:w-fit [&>input]:flex-1",
  {
    variants: {
      orientation: {
        horizontal:
          '[&>*:not(:first-child)]:rounded-l-none [&>*:not(:first-child)]:border-l-0 [&>*:not(:last-child)]:rounded-r-none [&>[data-slot]:not(:has(~[data-slot]))]:rounded-r!',
        vertical:
          'flex-col [&>*:not(:first-child)]:rounded-t-none [&>*:not(:first-child)]:border-t-0 [&>*:not(:last-child)]:rounded-b-none [&>[data-slot]:not(:has(~[data-slot]))]:rounded-b!',
      },
    },
    defaultVariants: {
      orientation: 'horizontal',
    },
  }
);

/**
 * A container that groups related buttons together with connected borders and spacing.
 *
 * ### Component Props
 *
 * | Prop            | Type                            | Default        |
 * | :-------------- | :------------------------------ | :------------- |
 * | _orientation_   | `"horizontal" \| "vertical"`    | `"horizontal"` |
 *
 *
 * ### Example Usage
 *
 * **Horizontal**
 * ```tsx
 * <ButtonGroup>
 *   <Button variant="outline">Left</Button>
 *   <Button variant="outline">Center</Button>
 *   <Button variant="outline">Right</Button>
 * </ButtonGroup>
 * ```
 *
 * **Vertical**
 * ```tsx
 * <ButtonGroup orientation="vertical">
 *   <Button variant="outline">Top</Button>
 *   <Button variant="outline">Bottom</Button>
 * </ButtonGroup>
 * ```
 */
function ButtonGroup({
  className,
  orientation,
  ...props
}: React.ComponentProps<'fieldset'> &
  VariantProps<typeof buttonGroupVariants>) {
  return (
    <fieldset
      data-slot="button-group"
      data-orientation={orientation}
      className={cn(buttonGroupVariants({orientation}), className)}
      {...props}
    />
  );
}

/**
 * A styled text or label element within a button group, typically used for static labels or addons.
 *
 * ### Component Props
 *
 * | Prop       | Type      | Default |
 * | :--------- | :-------- | :------ |
 * | _asChild_  | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <ButtonGroup>
 *   <ButtonGroupText>Label</ButtonGroupText>
 *   <Button variant="outline">Action</Button>
 * </ButtonGroup>
 * ```
 */
function ButtonGroupText({
  className,
  asChild = false,
  ...props
}: React.ComponentProps<'div'> & AsChild) {
  const Comp = asChild ? Slot.Root : 'div';

  return (
    <Comp
      className={cn(
        "flex items-center gap-2 rounded border bg-muted px-2.5 text-sm font-label shadow-raised [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4",
        className
      )}
      {...props}
    />
  );
}

/**
 * A visual separator line between items in a button group.
 *
 * ### Component Props
 *
 * | Prop           | Type                           | Default      |
 * | :------------- | :----------------------------- | :----------- |
 * | _orientation_  | `"horizontal" \| "vertical"`   | `"vertical"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <ButtonGroup>
 *   <Button variant="outline">Left</Button>
 *   <ButtonGroupSeparator />
 *   <Button variant="outline">Right</Button>
 * </ButtonGroup>
 * ```
 */
function ButtonGroupSeparator({
  className,
  orientation = 'vertical',
  ...props
}: React.ComponentProps<typeof Separator>) {
  return (
    <Separator
      data-slot="button-group-separator"
      orientation={orientation}
      className={cn(
        'relative self-stretch bg-input data-[orientation=horizontal]:mx-px data-[orientation=horizontal]:w-auto data-[orientation=vertical]:my-px data-[orientation=vertical]:h-auto',
        className
      )}
      {...props}
    />
  );
}

export {
  ButtonGroup,
  ButtonGroupSeparator,
  ButtonGroupText,
  buttonGroupVariants,
};
