// @docker/trident/src/components/ui/card.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

/**
 * A container that groups related content and actions, providing a consistent visual structure.
 *
 * ### Component Props
 *
 * | Prop   | Type                  | Default     |
 * | :----- | :-------------------- | :---------- |
 * | _size_ | `"default" \| "sm"`   | `"default"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Card>
 *   <CardHeader>
 *     <CardTitle>Card Title</CardTitle>
 *     <CardDescription>Card Description</CardDescription>
 *   </CardHeader>
 *   <CardContent>
 *     <p>Card Content</p>
 *   </CardContent>
 *   <CardFooter>
 *     <p>Card Footer</p>
 *   </CardFooter>
 * </Card>
 * ```
 */
function Card({
  className,
  size = 'default',
  ...props
}: React.ComponentProps<'div'> & {size?: 'default' | 'sm'}) {
  return (
    <div
      data-slot="card"
      data-size={size}
      className={cn(
        'group/card @container/card flex flex-col gap-6 overflow-hidden rounded bg-card py-6 text-sm text-card-foreground shadow-raised ring-1 ring-border has-[>img:first-child]:pt-0 data-[size=sm]:gap-4 data-[size=sm]:py-4 *:[img:first-child]:rounded-t *:[img:last-child]:rounded-b',
        className
      )}
      {...props}
    />
  );
}

/**
 * The header section of a Card, typically containing a title and description.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <CardHeader>
 *   <CardTitle>Title</CardTitle>
 *   <CardDescription>Description</CardDescription>
 *   <CardAction>
 *     <Button variant="outline">Action</Button>
 *   </CardAction>
 * </CardHeader>
 * ```
 */
function CardHeader({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="card-header"
      className={cn(
        'group/card-header @container/card-header grid auto-rows-min items-start gap-1 rounded-t px-6 group-data-[size=sm]/card:px-4 has-data-[slot=card-action]:grid-cols-[1fr_auto] has-data-[slot=card-description]:grid-rows-[auto_auto] @max-[300px]/card:px-3 [.border-b]:pb-6 group-data-[size=sm]/card:[.border-b]:pb-4',
        className
      )}
      {...props}
    />
  );
}

/**
 * The title element within a CardHeader.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CardTitle({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="card-title"
      className={cn(
        'text-base leading-heading font-heading tracking-heading group-data-[size=sm]/card:text-sm @max-[300px]/card:text-sm',
        className
      )}
      {...props}
    />
  );
}

/**
 * A secondary text element within a CardHeader, used to provide additional context.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CardDescription({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="card-description"
      className={cn(
        'text-sm leading-body tracking-body text-muted-foreground',
        className
      )}
      {...props}
    />
  );
}

/**
 * An action slot within a CardHeader, positioned to the right of the title and description.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CardAction({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="card-action"
      className={cn(
        'col-start-2 row-span-2 row-start-1 self-start justify-self-end',
        className
      )}
      {...props}
    />
  );
}

/**
 * The main content area of a Card.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CardContent({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="card-content"
      className={cn(
        'px-6 group-data-[size=sm]/card:px-4 @max-[300px]/card:px-3',
        className
      )}
      {...props}
    />
  );
}

/**
 * The footer section of a Card, typically used for actions or supplementary information.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CardFooter({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="card-footer"
      className={cn(
        'flex items-center rounded-b px-6 group-data-[size=sm]/card:px-4 @max-[300px]/card:px-3 [.border-t]:pt-6 group-data-[size=sm]/card:[.border-t]:pt-4',
        className
      )}
      {...props}
    />
  );
}

// ---------------------------------------------------------------------------
// CardStack
// ---------------------------------------------------------------------------

function BackCard({depth}: {depth: 1 | 2}) {
  return (
    <Card
      aria-hidden="true"
      className={cn(
        'shadow-subtle pointer-events-none absolute h-full w-full border border-dashed border-border bg-muted opacity-60 ring-0',
        depth === 1 && '-top-dds-300 -left-dds-300',
        depth === 2 && '-top-dds-600 -left-dds-600'
      )}
    />
  );
}

/**
 * Wraps a Card with 1 or 2 background cards offset behind it, creating a
 * stacked depth effect. Wrap in a sized container to control width.
 *
 * ### Component Props
 *
 * | Prop              | Type       | Default |
 * | :---------------- | :--------- | :------ |
 * | _stackDepth_      | `1 \| 2`   | `1`     |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <CardStack stackDepth={2}>
 *   <CardContent>Content</CardContent>
 * </CardStack>
 * ```
 */
function CardStack({
  stackDepth = 1,
  className,
  children,
  ...props
}: React.ComponentProps<'div'> & {stackDepth?: 1 | 2}) {
  // Render deepest first so it paints behind closer back cards.
  const backCards: React.ReactNode[] = [];
  for (let depth = stackDepth; depth >= 1; depth--) {
    backCards.push(<BackCard key={depth} depth={depth} />);
  }

  return (
    <div
      data-slot="card-stack"
      className={cn('relative', className)}
      {...props}
    >
      {backCards}
      <Card className="relative">{children}</Card>
    </div>
  );
}

export {
  Card,
  CardHeader,
  CardFooter,
  CardTitle,
  CardAction,
  CardDescription,
  CardContent,
  CardStack,
};
