// @docker/trident/src/components/ui/combobox.tsx
'use client';

import {Button} from '@/components/ui/button';
import {
  InputGroup,
  InputGroupAddon,
  InputGroupButton,
  InputGroupInput,
} from '@/components/ui/input-group';
import {cn} from '@/lib/utils';

import {Combobox as ComboboxPrimitive} from '@base-ui/react';
import {Check, ChevronDown, Loader2, X} from 'lucide-react';
import * as React from 'react';

/**
 * An autocomplete input with a filterable listbox, allowing users to search and select from a list of options.
 *
 * For multi-select, use the `ComboboxChips` composition. See the
 * `MultiSelectWithChips` story.
 *
 * ### Component Props
 *
 * See [Base UI Combobox](https://base-ui.com/react/components/combobox) documentation for full details.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Combobox>
 *   <ComboboxInput placeholder="Search..." />
 *   <ComboboxContent>
 *     <ComboboxList>
 *       <ComboboxItem value="option-1">Option 1</ComboboxItem>
 *       <ComboboxItem value="option-2">Option 2</ComboboxItem>
 *     </ComboboxList>
 *     <ComboboxEmpty>No results found.</ComboboxEmpty>
 *   </ComboboxContent>
 * </Combobox>
 * ```
 *
 * > **Important:** Place `ComboboxEmpty` as a sibling of `ComboboxList` (inside
 * > `ComboboxContent`, outside `ComboboxList`). Placing it inside the list renders
 * > `role="status"` as a child of `role="listbox"`, which violates ARIA required
 * > children (WCAG 4.1.2).
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 2.1.1 Keyboard — input is natively focusable; arrow keys navigate the listbox; Enter selects; Escape closes.
 * @wcag 4.1.2 Name, Role, Value — `combobox` role with `aria-expanded`, `aria-activedescendant`, and `listbox` via Base UI primitives.
 * @wcag 2.4.7 Focus Visible — input focus ring via `focus-visible:ring-3` on the input group; highlighted items via `data-highlighted:bg-accent`.
 */
const Combobox = ComboboxPrimitive.Root;

/**
 * Displays the currently selected value of the Combobox.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ComboboxValue({...props}: ComboboxPrimitive.Value.Props) {
  return <ComboboxPrimitive.Value data-slot="combobox-value" {...props} />;
}

/**
 * A button that toggles the Combobox popup open or closed, displaying a chevron icon.
 *
 * ### Component Props
 *
 * | Prop           | Type     | Default  |
 * | :------------- | :------- | :------- |
 * | _aria-label_   | `string` | `"Open"` |
 *
 * Pass a custom `aria-label` to provide a more descriptive accessible name
 * (e.g. `aria-label="Open country list"`). The default `"Open"` is used when
 * no `aria-label` is provided (WCAG 4.1.2).
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _loading_ | `boolean` | `false` |
 */
function ComboboxTrigger({
  className,
  children,
  loading = false,
  ...props
}: ComboboxPrimitive.Trigger.Props & {loading?: boolean}) {
  return (
    <ComboboxPrimitive.Trigger
      data-slot="combobox-trigger"
      aria-label="Open"
      disabled={props.disabled || loading}
      aria-busy={loading || undefined}
      className={cn("[&_svg:not([class*='size-'])]:size-4", className)}
      {...props}
    >
      {children}
      {loading ? (
        <Loader2
          data-slot="combobox-spinner"
          aria-hidden="true"
          className="pointer-events-none size-4 animate-spin text-muted-foreground"
        />
      ) : (
        <ChevronDown
          data-slot="combobox-chevron"
          aria-hidden="true"
          className="pointer-events-none size-4 text-muted-foreground"
        />
      )}
    </ComboboxPrimitive.Trigger>
  );
}

function ComboboxClear({className, ...props}: ComboboxPrimitive.Clear.Props) {
  return (
    <ComboboxPrimitive.Clear
      data-slot="combobox-clear"
      render={<InputGroupButton variant="ghost" size="icon-xs" />}
      className={cn(className)}
      {...props}
    >
      <X className="pointer-events-none" />
    </ComboboxPrimitive.Clear>
  );
}

/**
 * The text input field for the Combobox, with optional trigger and clear buttons.
 *
 * ### Component Props
 *
 * | Prop           | Type      | Default |
 * | :------------- | :-------- | :------ |
 * | _showTrigger_  | `boolean` | `true`  |
 * | _showClear_    | `boolean` | `false` |
 * | _disabled_     | `boolean` | `false` |
 * | _loading_      | `boolean` | `false` |
 */
function ComboboxInput({
  className,
  children,
  disabled = false,
  showTrigger = true,
  showClear = false,
  loading = false,
  ...props
}: ComboboxPrimitive.Input.Props & {
  showTrigger?: boolean;
  showClear?: boolean;
  loading?: boolean;
}) {
  return (
    <InputGroup className={cn('w-auto', className)}>
      <ComboboxPrimitive.Input
        render={<InputGroupInput disabled={disabled || loading} />}
        {...props}
      />
      <InputGroupAddon align="inline-end">
        {showTrigger && (
          <InputGroupButton
            size="icon-xs"
            variant="ghost"
            asChild
            data-slot="input-group-button"
            className="rounded-md group-has-data-[slot=combobox-clear]/input-group:hidden data-pressed:bg-transparent"
            disabled={disabled || loading}
          >
            <ComboboxTrigger loading={loading} />
          </InputGroupButton>
        )}
        {showClear && <ComboboxClear disabled={disabled || loading} />}
      </InputGroupAddon>
      {children}
    </InputGroup>
  );
}

/**
 * The dropdown popup container for the Combobox, rendered through a portal with positioning options.
 *
 * ### Component Props
 *
 * | Prop           | Type                                          | Default    |
 * | :------------- | :-------------------------------------------- | :--------- |
 * | _side_         | `"top" \| "right" \| "bottom" \| "left"`      | `"bottom"` |
 * | _sideOffset_   | `number`                                      | `6`        |
 * | _align_        | `"start" \| "center" \| "end"`                | `"start"`  |
 * | _alignOffset_  | `number`                                      | `0`        |
 * | _anchor_       | `HTMLElement \| null`                          | -          |
 */
function ComboboxContent({
  className,
  side = 'bottom',
  sideOffset = 6,
  align = 'start',
  alignOffset = 0,
  anchor,
  ...props
}: ComboboxPrimitive.Popup.Props &
  Pick<
    ComboboxPrimitive.Positioner.Props,
    'side' | 'align' | 'sideOffset' | 'alignOffset' | 'anchor'
  >) {
  return (
    <ComboboxPrimitive.Portal>
      <ComboboxPrimitive.Positioner
        side={side}
        sideOffset={sideOffset}
        align={align}
        alignOffset={alignOffset}
        anchor={anchor}
        className="isolate z-floating"
      >
        <ComboboxPrimitive.Popup
          data-slot="combobox-content"
          data-chips={!!anchor}
          className={cn(
            'group/combobox-content relative max-h-72 w-(--anchor-width) max-w-(--available-width) min-w-36 origin-(--transform-origin) overflow-hidden rounded bg-popover text-popover-foreground shadow-floating ring-1 ring-border duration-fast ease-out data-[chips=true]:min-w-(--anchor-width) data-[side=bottom]:slide-in-from-top-2 data-[side=inline-end]:slide-in-from-left-2 data-[side=inline-start]:slide-in-from-right-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 *:data-[slot=input-group]:m-1 *:data-[slot=input-group]:mb-0 *:data-[slot=input-group]:h-8 *:data-[slot=input-group]:border-input *:data-[slot=input-group]:bg-input-bg *:data-[slot=input-group]:shadow-none data-open:animate-in data-open:fade-in-0 data-open:zoom-in-95 data-closed:animate-out data-closed:fade-out-0 data-closed:zoom-out-95',
            className
          )}
          {...props}
        />
      </ComboboxPrimitive.Positioner>
    </ComboboxPrimitive.Portal>
  );
}

/**
 * The scrollable list container within the Combobox popup that holds ComboboxItem elements.
 *
 * ### Component Props
 *
 * | Prop               | Type     | Default | Description                                      |
 * | :----------------- | :------- | :------ | :----------------------------------------------- |
 * | _aria-label_       | `string` | -       | Accessible name for the listbox (WCAG 4.1.2).    |
 * | _aria-labelledby_  | `string` | -       | ID(s) of element(s) that label the listbox.      |
 *
 * One of `aria-label` or `aria-labelledby` **must** be provided to satisfy
 * WCAG 2.2 SC 4.1.2 (Name, Role, Value). The listbox role requires an
 * accessible name — pass `aria-label="Options"` or a context-specific label
 * (e.g. `aria-label="Available fruits"`).
 */
function ComboboxList({
  className,
  'aria-label': ariaLabel = 'Options',
  ...props
}: ComboboxPrimitive.List.Props) {
  return (
    <ComboboxPrimitive.List
      data-slot="combobox-list"
      aria-label={ariaLabel}
      className={cn(
        'no-scrollbar max-h-[min(calc(--spacing(72)---spacing(9)),calc(var(--available-height)---spacing(9)))] scroll-py-1 overflow-y-auto overscroll-contain p-1 data-empty:p-0',
        className
      )}
      {...props}
    />
  );
}

/**
 * An individual selectable option within the ComboboxList, displaying a check indicator when selected.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ComboboxItem({
  className,
  children,
  ...props
}: ComboboxPrimitive.Item.Props) {
  return (
    <ComboboxPrimitive.Item
      data-slot="combobox-item"
      className={cn(
        "relative flex w-full cursor-default items-center gap-2 rounded py-1.5 pr-8 pl-2 text-sm outline-hidden select-none data-highlighted:bg-accent data-highlighted:text-accent-foreground not-data-[variant=destructive]:data-highlighted:**:text-accent-foreground data-disabled:pointer-events-none data-disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        className
      )}
      {...props}
    >
      {children}
      <ComboboxPrimitive.ItemIndicator
        render={
          <span className="pointer-events-none absolute right-2 flex size-4 items-center justify-center" />
        }
      >
        <Check className="pointer-events-none" />
      </ComboboxPrimitive.ItemIndicator>
    </ComboboxPrimitive.Item>
  );
}

/**
 * A grouping container for related ComboboxItem elements, optionally with a ComboboxLabel.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ComboboxGroup({className, ...props}: ComboboxPrimitive.Group.Props) {
  return (
    <ComboboxPrimitive.Group
      data-slot="combobox-group"
      className={cn(className)}
      {...props}
    />
  );
}

/**
 * A label displayed at the top of a ComboboxGroup to describe the group of options.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ComboboxLabel({
  className,
  ...props
}: ComboboxPrimitive.GroupLabel.Props) {
  return (
    <ComboboxPrimitive.GroupLabel
      data-slot="combobox-label"
      className={cn('px-2 py-1.5 text-xs text-muted-foreground', className)}
      {...props}
    />
  );
}

/**
 * A wrapper for providing a custom collection of items to the Combobox.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ComboboxCollection({...props}: ComboboxPrimitive.Collection.Props) {
  return (
    <ComboboxPrimitive.Collection data-slot="combobox-collection" {...props} />
  );
}

/**
 * A message displayed when no Combobox items match the current input value.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ComboboxEmpty({className, ...props}: ComboboxPrimitive.Empty.Props) {
  return (
    <ComboboxPrimitive.Empty
      data-slot="combobox-empty"
      className={cn(
        'hidden w-full justify-center py-2 text-center text-sm text-muted-foreground group-data-empty/combobox-content:flex',
        className
      )}
      {...props}
    />
  );
}

/**
 * A visual separator between groups or items within the Combobox popup.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ComboboxSeparator({
  className,
  ...props
}: ComboboxPrimitive.Separator.Props) {
  return (
    <ComboboxPrimitive.Separator
      data-slot="combobox-separator"
      className={cn('-mx-1 my-1 h-px bg-border', className)}
      {...props}
    />
  );
}

/**
 * A container for displaying selected values as removable chips in a multi-select Combobox.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ComboboxChips({
  className,
  ...props
}: React.ComponentPropsWithRef<typeof ComboboxPrimitive.Chips> &
  ComboboxPrimitive.Chips.Props) {
  return (
    <ComboboxPrimitive.Chips
      data-slot="combobox-chips"
      className={cn(
        'flex min-h-9 flex-wrap items-center gap-1.5 rounded border border-input bg-transparent bg-clip-padding px-2.5 py-1.5 text-sm shadow-raised transition-[color,box-shadow] ease-default focus-within:border-ring focus-within:ring-3 focus-within:ring-ring has-aria-invalid:border-destructive has-aria-invalid:ring-3 has-aria-invalid:ring-destructive/20 has-data-[slot=combobox-chip]:px-1.5 dark:bg-input-bg dark:has-aria-invalid:border-destructive/50 dark:has-aria-invalid:ring-destructive/40',
        className
      )}
      {...props}
    />
  );
}

/**
 * An individual chip representing a selected value, with an optional remove button.
 *
 * ### Component Props
 *
 * | Prop          | Type      | Default |
 * | :------------ | :-------- | :------ |
 * | _showRemove_  | `boolean` | `true`  |
 */
function ComboboxChip({
  className,
  children,
  showRemove = true,
  ...props
}: ComboboxPrimitive.Chip.Props & {
  showRemove?: boolean;
}) {
  return (
    <ComboboxPrimitive.Chip
      data-slot="combobox-chip"
      className={cn(
        'flex h-[calc(--spacing(5.5))] w-fit items-center justify-center gap-1 rounded bg-muted px-1.5 text-xs font-label whitespace-nowrap text-foreground has-disabled:pointer-events-none has-disabled:cursor-not-allowed has-disabled:opacity-50 has-data-[slot=combobox-chip-remove]:pr-0',
        className
      )}
      {...props}
    >
      {children}
      {showRemove && (
        <ComboboxPrimitive.ChipRemove
          aria-label="Remove"
          render={<Button variant="ghost" size="icon-xs" />}
          className="-ml-1 opacity-50 hover:opacity-100"
          data-slot="combobox-chip-remove"
        >
          <X aria-hidden className="pointer-events-none" />
        </ComboboxPrimitive.ChipRemove>
      )}
    </ComboboxPrimitive.Chip>
  );
}

/**
 * A text input designed to be used inline within the ComboboxChips container for multi-select filtering.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ComboboxChipsInput({
  className,
  ...props
}: ComboboxPrimitive.Input.Props) {
  return (
    <ComboboxPrimitive.Input
      data-slot="combobox-chip-input"
      className={cn('min-w-16 flex-1 outline-none', className)}
      {...props}
    />
  );
}

/**
 * A hook that returns a ref to be used as the anchor element for the ComboboxContent positioning.
 *
 * ### Example Usage
 *
 * ```tsx
 * const anchor = useComboboxAnchor();
 * <ComboboxChips ref={anchor}>...</ComboboxChips>
 * <ComboboxContent anchor={anchor.current}>...</ComboboxContent>
 * ```
 */
function useComboboxAnchor() {
  return React.useRef<HTMLDivElement | null>(null);
}

export {
  Combobox,
  ComboboxInput,
  ComboboxContent,
  ComboboxContent as ComboboxPopup, // alias for Base UI engineers expecting *Popup naming
  ComboboxList,
  ComboboxItem,
  ComboboxGroup,
  ComboboxLabel,
  ComboboxCollection,
  ComboboxEmpty,
  ComboboxSeparator,
  ComboboxChips,
  ComboboxChip,
  ComboboxChipsInput,
  ComboboxTrigger,
  ComboboxValue,
  useComboboxAnchor,
};
