// @docker/trident/src/components/ui/command.tsx
import type * as React from 'react';

import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
} from '@/components/ui/dialog';
import {InputGroup, InputGroupAddon} from '@/components/ui/input-group';
import {cn} from '@/lib/utils';

import {Command as CommandPrimitive} from 'cmdk';
import {Check, Search} from 'lucide-react';

/**
 * A fast, composable command menu built on top of cmdk for searching and selecting from a list of actions.
 *
 * ### Component Props
 *
 * See [cmdk](https://cmdk.paco.me) documentation for full details.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Command>
 *   <CommandInput placeholder="Type a command or search..." />
 *   <CommandList>
 *     <CommandEmpty>No results found.</CommandEmpty>
 *     <CommandGroup heading="Suggestions">
 *       <CommandItem>Calendar</CommandItem>
 *       <CommandItem>Search</CommandItem>
 *     </CommandGroup>
 *   </CommandList>
 * </Command>
 * ```
 */
function Command({
  className,
  ...props
}: React.ComponentProps<typeof CommandPrimitive>) {
  return (
    <CommandPrimitive
      data-slot="command"
      className={cn(
        'flex size-full flex-col overflow-hidden rounded-lg bg-popover p-1 text-popover-foreground',
        className
      )}
      {...props}
    />
  );
}

/**
 * A Dialog-based wrapper for the Command component, typically triggered by a keyboard shortcut.
 *
 * ### Component Props
 *
 * | Prop               | Type      | Default                          |
 * | :----------------- | :-------- | :------------------------------- |
 * | _title_            | `string`  | `"Command Palette"`              |
 * | _description_      | `string`  | `"Search for a command to run..."` |
 * | _showCloseButton_  | `boolean` | `false`                          |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <CommandDialog open={open} onOpenChange={setOpen}>
 *   <Command>
 *     <CommandInput placeholder="Type a command or search..." />
 *     <CommandList>
 *       <CommandEmpty>No results found.</CommandEmpty>
 *       <CommandGroup heading="Suggestions">
 *         <CommandItem>Calendar</CommandItem>
 *       </CommandGroup>
 *     </CommandList>
 *   </Command>
 * </CommandDialog>
 * ```
 */
function CommandDialog({
  title = 'Command Palette',
  description = 'Search for a command to run...',
  children,
  className,
  showCloseButton = false,
  ...props
}: React.ComponentProps<typeof Dialog> & {
  title?: string;
  description?: string;
  className?: string;
  showCloseButton?: boolean;
}) {
  return (
    <Dialog {...props}>
      <DialogHeader className="sr-only">
        <DialogTitle>{title}</DialogTitle>
        <DialogDescription>{description}</DialogDescription>
      </DialogHeader>
      <DialogContent
        className={cn('top-1/3 translate-y-0 overflow-hidden p-0', className)}
        showCloseButton={showCloseButton}
      >
        {children}
      </DialogContent>
    </Dialog>
  );
}

/**
 * The search input field for the Command component, with a built-in search icon.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CommandInput({
  className,
  ...props
}: React.ComponentProps<typeof CommandPrimitive.Input>) {
  return (
    <div data-slot="command-input-wrapper" className="p-1 pb-0">
      <InputGroup className="h-8 bg-input-bg shadow-none">
        <CommandPrimitive.Input
          data-slot="input-group-control"
          className={cn(
            'w-full text-sm outline-none disabled:cursor-not-allowed disabled:opacity-50',
            className
          )}
          {...props}
        />
        <InputGroupAddon>
          <Search className="size-4 shrink-0 opacity-50" />
        </InputGroupAddon>
      </InputGroup>
    </div>
  );
}

/**
 * The scrollable list container for Command items and groups.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CommandList({
  className,
  ...props
}: React.ComponentProps<typeof CommandPrimitive.List>) {
  return (
    <CommandPrimitive.List
      data-slot="command-list"
      className={cn(
        'no-scrollbar max-h-72 scroll-py-1 overflow-x-hidden overflow-y-auto outline-none',
        className
      )}
      {...props}
    />
  );
}

/**
 * A message displayed when no Command items match the current search query.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CommandEmpty({
  className,
  ...props
}: React.ComponentProps<typeof CommandPrimitive.Empty>) {
  return (
    <CommandPrimitive.Empty
      data-slot="command-empty"
      className={cn('py-6 text-center text-sm', className)}
      {...props}
    />
  );
}

/**
 * A grouping container for related CommandItem elements, with an optional heading.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CommandGroup({
  className,
  ...props
}: React.ComponentProps<typeof CommandPrimitive.Group>) {
  return (
    <CommandPrimitive.Group
      data-slot="command-group"
      className={cn(
        'overflow-hidden p-1 text-foreground **:[[cmdk-group-heading]]:px-2 **:[[cmdk-group-heading]]:py-1.5 **:[[cmdk-group-heading]]:text-xs **:[[cmdk-group-heading]]:font-label **:[[cmdk-group-heading]]:text-muted-foreground',
        className
      )}
      {...props}
    />
  );
}

/**
 * A visual separator between groups or items within the Command list.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CommandSeparator({
  className,
  ...props
}: React.ComponentProps<typeof CommandPrimitive.Separator>) {
  return (
    <CommandPrimitive.Separator
      data-slot="command-separator"
      className={cn('-mx-1 h-px w-auto bg-border', className)}
      {...props}
    />
  );
}

/**
 * An individual selectable action or option within the Command list, displaying a check indicator when selected.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CommandItem({
  className,
  children,
  ...props
}: React.ComponentProps<typeof CommandPrimitive.Item>) {
  return (
    <CommandPrimitive.Item
      data-slot="command-item"
      className={cn(
        "group/command-item relative flex cursor-default items-center gap-2 rounded px-2 py-1.5 text-sm outline-hidden select-none in-data-[slot=dialog-content]:rounded-lg data-[disabled=true]:pointer-events-none data-[disabled=true]:opacity-50 data-selected:bg-muted data-selected:text-foreground [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 data-selected:**:[svg]:text-foreground",
        className
      )}
      {...props}
    >
      {children}
      <Check className="ml-auto opacity-0 group-has-data-[slot=command-shortcut]/command-item:hidden group-data-[checked=true]/command-item:opacity-100" />
    </CommandPrimitive.Item>
  );
}

/**
 * A styled span for displaying keyboard shortcut hints within a CommandItem.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function CommandShortcut({className, ...props}: React.ComponentProps<'span'>) {
  return (
    <span
      data-slot="command-shortcut"
      className={cn(
        'ml-auto text-xs tracking-widest text-muted-foreground group-data-selected/command-item:text-foreground',
        className
      )}
      {...props}
    />
  );
}

export {
  Command,
  CommandDialog,
  CommandInput,
  CommandList,
  CommandEmpty,
  CommandGroup,
  CommandItem,
  CommandShortcut,
  CommandSeparator,
};
