// @docker/trident/src/components/ui/context-menu.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {Check, ChevronRight} from 'lucide-react';
import {ContextMenu as ContextMenuPrimitive} from 'radix-ui';

/**
 * Displays a menu located at the pointer, triggered by a right-click or a long press.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop            | Type                           | Default |
 * | :-------------- | :----------------------------- | :------ |
 * | _dir_           | `"ltr" \| "rtl"`               | -       |
 * | _onOpenChange_  | `(open: boolean) => void`      | -       |
 * | _modal_         | `boolean`                      | `true`  |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <ContextMenu>
 *   <ContextMenuTrigger>Right click here</ContextMenuTrigger>
 *   <ContextMenuContent>
 *     <ContextMenuItem>Profile</ContextMenuItem>
 *     <ContextMenuItem>Billing</ContextMenuItem>
 *     <ContextMenuSeparator />
 *     <ContextMenuItem>Logout</ContextMenuItem>
 *   </ContextMenuContent>
 * </ContextMenu>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 2.1.1 Keyboard — arrow keys navigate items; Escape closes the menu.
 * @wcag 4.1.2 Name, Role, Value — `menu` and `menuitem` roles via Radix ContextMenu primitives.
 */
function ContextMenu({
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.Root>) {
  return <ContextMenuPrimitive.Root data-slot="context-menu" {...props} />;
}

/**
 * The element that triggers the ContextMenu on right-click or long press. Wraps the target area.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 * | _disabled_  | `boolean` | `false` |
 */
function ContextMenuTrigger({
  className,
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.Trigger>) {
  return (
    <ContextMenuPrimitive.Trigger
      data-slot="context-menu-trigger"
      className={cn('select-none', className)}
      {...props}
    />
  );
}

/**
 * A grouping container for related ContextMenuItem elements.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 */
function ContextMenuGroup({
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.Group>) {
  return (
    <ContextMenuPrimitive.Group data-slot="context-menu-group" {...props} />
  );
}

/**
 * A portal container that renders ContextMenu content outside the DOM hierarchy.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop           | Type          | Default         |
 * | :------------- | :------------ | :-------------- |
 * | _forceMount_   | `boolean`     | -               |
 * | _container_    | `HTMLElement`  | `document.body` |
 */
function ContextMenuPortal({
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.Portal>) {
  return (
    <ContextMenuPrimitive.Portal data-slot="context-menu-portal" {...props} />
  );
}

/**
 * A container for a nested submenu within the ContextMenu.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop            | Type                           | Default |
 * | :-------------- | :----------------------------- | :------ |
 * | _defaultOpen_   | `boolean`                      | -       |
 * | _open_          | `boolean`                      | -       |
 * | _onOpenChange_  | `(open: boolean) => void`      | -       |
 */
function ContextMenuSub({
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.Sub>) {
  return <ContextMenuPrimitive.Sub data-slot="context-menu-sub" {...props} />;
}

/**
 * A group of radio items within the ContextMenu where only one can be selected at a time.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop              | Type                         | Default |
 * | :---------------- | :--------------------------- | :------ |
 * | _asChild_         | `boolean`                    | `false` |
 * | _value_           | `string`                     | -       |
 * | _onValueChange_   | `(value: string) => void`    | -       |
 */
function ContextMenuRadioGroup({
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.RadioGroup>) {
  return (
    <ContextMenuPrimitive.RadioGroup
      data-slot="context-menu-radio-group"
      {...props}
    />
  );
}

/**
 * The popup content area of the ContextMenu, rendered through a portal.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop                    | Type                                     | Default       |
 * | :---------------------- | :--------------------------------------- | :------------ |
 * | _asChild_               | `boolean`                                | `false`       |
 * | _loop_                  | `boolean`                                | `false`       |
 * | _forceMount_            | `boolean`                                | -             |
 * | _alignOffset_           | `number`                                 | `0`           |
 * | _avoidCollisions_       | `boolean`                                | `true`        |
 * | _collisionPadding_      | `number \| Padding`                      | `0`           |
 * | _sticky_                | `"partial" \| "always"`                  | `"partial"`   |
 * | _hideWhenDetached_      | `boolean`                                | `false`       |
 * | _onCloseAutoFocus_      | `(event: Event) => void`                 | -             |
 * | _onEscapeKeyDown_       | `(event: KeyboardEvent) => void`         | -             |
 * | _onPointerDownOutside_  | `(event: PointerDownOutsideEvent) => void` | -           |
 * | _onFocusOutside_        | `(event: FocusOutsideEvent) => void`     | -             |
 * | _onInteractOutside_     | `(event: Event) => void`                 | -             |
 */
function ContextMenuContent({
  className,
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.Content> & {
  side?: 'top' | 'right' | 'bottom' | 'left';
}) {
  return (
    <ContextMenuPrimitive.Portal>
      <ContextMenuPrimitive.Content
        data-slot="context-menu-content"
        className={cn(
          'z-floating max-h-(--radix-context-menu-content-available-height) min-w-36 origin-(--radix-context-menu-content-transform-origin) overflow-x-hidden overflow-y-auto rounded bg-popover p-1 text-popover-foreground shadow-floating ring-1 ring-border duration-fast ease-out data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 data-open:animate-in data-open:fade-in-0 data-open:zoom-in-95 data-closed:animate-out data-closed:fade-out-0 data-closed:zoom-out-95',
          className
        )}
        {...props}
      />
    </ContextMenuPrimitive.Portal>
  );
}

/**
 * An individual selectable action within the ContextMenu.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop          | Type                           | Default     |
 * | :------------ | :----------------------------- | :---------- |
 * | _asChild_     | `boolean`                      | `false`     |
 * | _disabled_    | `boolean`                      | -           |
 * | _onSelect_    | `(event: Event) => void`       | -           |
 * | _textValue_   | `string`                       | -           |
 * | _inset_       | `boolean`                      | -           |
 * | _variant_     | `"default" \| "destructive"`   | `"default"` |
 */
function ContextMenuItem({
  className,
  inset,
  variant = 'default',
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.Item> & {
  inset?: boolean;
  variant?: 'default' | 'destructive';
}) {
  return (
    <ContextMenuPrimitive.Item
      data-slot="context-menu-item"
      data-inset={inset}
      data-variant={variant}
      className={cn(
        "group/context-menu-item relative flex cursor-default items-center gap-2 rounded px-2 py-1.5 text-sm outline-hidden select-none focus:bg-accent focus:text-accent-foreground focus:ring-1 focus:ring-ring focus:ring-inset data-inset:pl-8 data-[variant=destructive]:text-destructive data-[variant=destructive]:focus:bg-destructive-subtle data-[variant=destructive]:focus:text-destructive data-[variant=destructive]:focus:ring-destructive data-disabled:pointer-events-none data-disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 focus:*:[svg]:text-accent-foreground data-[variant=destructive]:*:[svg]:text-destructive",
        className
      )}
      {...props}
    />
  );
}

/**
 * The item within the ContextMenu that opens a submenu, displaying a chevron indicator.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop          | Type      | Default |
 * | :------------ | :-------- | :------ |
 * | _asChild_     | `boolean` | `false` |
 * | _disabled_    | `boolean` | -       |
 * | _textValue_   | `string`  | -       |
 * | _inset_       | `boolean` | -       |
 */
function ContextMenuSubTrigger({
  className,
  inset,
  children,
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.SubTrigger> & {
  inset?: boolean;
}) {
  return (
    <ContextMenuPrimitive.SubTrigger
      data-slot="context-menu-sub-trigger"
      data-inset={inset}
      className={cn(
        "flex cursor-default items-center rounded px-2 py-1.5 text-sm outline-hidden select-none focus:bg-accent focus:text-accent-foreground focus:ring-1 focus:ring-ring focus:ring-inset data-inset:pl-8 data-open:bg-accent data-open:text-accent-foreground data-open:ring-1 data-open:ring-ring data-open:ring-inset [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        className
      )}
      {...props}
    >
      {children}
      <ChevronRight className="ml-auto" />
    </ContextMenuPrimitive.SubTrigger>
  );
}

/**
 * The popup content area for a ContextMenu submenu.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop                    | Type                                     | Default       |
 * | :---------------------- | :--------------------------------------- | :------------ |
 * | _asChild_               | `boolean`                                | `false`       |
 * | _loop_                  | `boolean`                                | `false`       |
 * | _forceMount_            | `boolean`                                | -             |
 * | _sideOffset_            | `number`                                 | `0`           |
 * | _alignOffset_           | `number`                                 | `0`           |
 * | _avoidCollisions_       | `boolean`                                | `true`        |
 * | _collisionPadding_      | `number \| Padding`                      | `0`           |
 * | _sticky_                | `"partial" \| "always"`                  | `"partial"`   |
 * | _hideWhenDetached_      | `boolean`                                | `false`       |
 * | _onEscapeKeyDown_       | `(event: KeyboardEvent) => void`         | -             |
 * | _onPointerDownOutside_  | `(event: PointerDownOutsideEvent) => void` | -           |
 * | _onFocusOutside_        | `(event: FocusOutsideEvent) => void`     | -             |
 * | _onInteractOutside_     | `(event: Event) => void`                 | -             |
 */
function ContextMenuSubContent({
  className,
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.SubContent>) {
  return (
    <ContextMenuPrimitive.SubContent
      data-slot="context-menu-sub-content"
      className={cn(
        'z-floating min-w-32 origin-(--radix-context-menu-content-transform-origin) overflow-hidden rounded border bg-popover p-1 text-popover-foreground shadow-floating duration-fast ease-out data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 data-open:animate-in data-open:fade-in-0 data-open:zoom-in-95 data-closed:animate-out data-closed:fade-out-0 data-closed:zoom-out-95',
        className
      )}
      {...props}
    />
  );
}

/**
 * A menu item that can be toggled between checked and unchecked states.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop                | Type                                          | Default |
 * | :------------------ | :-------------------------------------------- | :------ |
 * | _asChild_           | `boolean`                                     | `false` |
 * | _checked_           | `boolean \| "indeterminate"`                  | -       |
 * | _onCheckedChange_   | `(checked: boolean) => void`                  | -       |
 * | _disabled_          | `boolean`                                     | -       |
 * | _onSelect_          | `(event: Event) => void`                      | -       |
 * | _textValue_         | `string`                                      | -       |
 */
function ContextMenuCheckboxItem({
  className,
  children,
  checked,
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.CheckboxItem>) {
  return (
    <ContextMenuPrimitive.CheckboxItem
      data-slot="context-menu-checkbox-item"
      className={cn(
        "relative flex cursor-default items-center gap-2 rounded py-1.5 pr-8 pl-2 text-sm outline-hidden select-none focus:bg-accent focus:text-accent-foreground focus:ring-1 focus:ring-ring focus:ring-inset data-disabled:pointer-events-none data-disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        className
      )}
      checked={checked}
      {...props}
    >
      <span className="pointer-events-none absolute right-2">
        <ContextMenuPrimitive.ItemIndicator>
          <Check />
        </ContextMenuPrimitive.ItemIndicator>
      </span>
      {children}
    </ContextMenuPrimitive.CheckboxItem>
  );
}

/**
 * A radio item within a ContextMenuRadioGroup, where only one item can be selected at a time.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop                 | Type                           | Default |
 * | :------------------- | :----------------------------- | :------ |
 * | _asChild_            | `boolean`                      | `false` |
 * | _value_ *(required)* | `string`                       | -       |
 * | _disabled_           | `boolean`                      | -       |
 * | _onSelect_           | `(event: Event) => void`       | -       |
 * | _textValue_          | `string`                       | -       |
 */
function ContextMenuRadioItem({
  className,
  children,
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.RadioItem>) {
  return (
    <ContextMenuPrimitive.RadioItem
      data-slot="context-menu-radio-item"
      className={cn(
        "relative flex cursor-default items-center gap-2 rounded py-1.5 pr-8 pl-2 text-sm outline-hidden select-none focus:bg-accent focus:text-accent-foreground focus:ring-1 focus:ring-ring focus:ring-inset data-disabled:pointer-events-none data-disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        className
      )}
      {...props}
    >
      <span className="pointer-events-none absolute right-2">
        <ContextMenuPrimitive.ItemIndicator>
          <Check />
        </ContextMenuPrimitive.ItemIndicator>
      </span>
      {children}
    </ContextMenuPrimitive.RadioItem>
  );
}

/**
 * A non-interactive label within the ContextMenu, used to describe a group of items.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 * | _inset_     | `boolean` | -       |
 */
function ContextMenuLabel({
  className,
  inset,
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.Label> & {
  inset?: boolean;
}) {
  return (
    <ContextMenuPrimitive.Label
      data-slot="context-menu-label"
      data-inset={inset}
      className={cn(
        'px-2 py-1.5 text-xs font-label text-muted-foreground data-inset:pl-8',
        className
      )}
      {...props}
    />
  );
}

/**
 * A visual separator between items or groups within the ContextMenu.
 *
 * ### Component Props
 *
 * See [Context Menu props](https://www.radix-ui.com/primitives/docs/components/context-menu) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 */
function ContextMenuSeparator({
  className,
  ...props
}: React.ComponentProps<typeof ContextMenuPrimitive.Separator>) {
  return (
    <ContextMenuPrimitive.Separator
      data-slot="context-menu-separator"
      className={cn('-mx-1 my-1 h-px bg-border', className)}
      {...props}
    />
  );
}

/**
 * A styled span for displaying keyboard shortcut hints within a ContextMenuItem.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function ContextMenuShortcut({
  className,
  ...props
}: React.ComponentProps<'span'>) {
  return (
    <span
      data-slot="context-menu-shortcut"
      className={cn(
        'ml-auto text-xs tracking-widest text-muted-foreground group-focus/context-menu-item:text-accent-foreground',
        className
      )}
      {...props}
    />
  );
}

export {
  ContextMenu,
  ContextMenuTrigger,
  ContextMenuContent,
  ContextMenuItem,
  ContextMenuCheckboxItem,
  ContextMenuRadioItem,
  ContextMenuLabel,
  ContextMenuSeparator,
  ContextMenuShortcut,
  ContextMenuGroup,
  ContextMenuPortal,
  ContextMenuSub,
  ContextMenuSubContent,
  ContextMenuSubTrigger,
  ContextMenuRadioGroup,
};
