// @docker/trident/src/components/ui/data-table.tsx
'use client';

import {Button} from '@/components/ui/button';
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from '@/components/ui/table';
import {cn} from '@/lib/utils';

import {
  type ColumnDef,
  type ColumnFiltersState,
  flexRender,
  getCoreRowModel,
  getFilteredRowModel,
  getPaginationRowModel,
  getSortedRowModel,
  type Header,
  type SortingState,
  type TableOptions,
  useReactTable,
  type VisibilityState,
} from '@tanstack/react-table';
import {
  ArrowDown,
  ArrowUp,
  ChevronLeft,
  ChevronRight,
  ChevronsUpDown,
} from 'lucide-react';
import {useState} from 'react';

// ─── Column Header ──────────────────────────────────────────────────────────

type DataTableColumnHeaderProps<TData, TValue> = {
  header: Header<TData, TValue>;
  title: string;
  className?: string;
};

/**
 * Sortable column header for DataTable. Replaces `@mui/material/TableSortLabel`.
 *
 * Renders the column title with a sort-direction indicator when the column is
 * sortable. Clicking toggles between ascending → descending → no sort.
 *
 * @wcag 1.4.3 Contrast (Minimum) — sort icons use foreground color via currentColor.
 * @wcag 2.1.1 Keyboard — uses native `<button>`, focusable and activatable by Enter/Space.
 * @wcag 4.1.2 Name, Role, Value — aria-sort on the parent `<th>` communicates sort state.
 */
function DataTableColumnHeader<TData, TValue>({
  header,
  title,
  className,
}: DataTableColumnHeaderProps<TData, TValue>) {
  if (!header.column.getCanSort()) {
    return <span className={cn(className)}>{title}</span>;
  }

  const sorted = header.column.getIsSorted();

  return (
    <Button
      variant="ghost"
      size="sm"
      className={cn('-ml-2 h-8 gap-1 font-label', className)}
      onClick={() => header.column.toggleSorting()}
    >
      {title}
      {sorted === 'asc' ? (
        <ArrowUp className="size-3.5" />
      ) : sorted === 'desc' ? (
        <ArrowDown className="size-3.5" />
      ) : (
        <ChevronsUpDown className="size-3.5 opacity-50" />
      )}
    </Button>
  );
}

// ─── Pagination ─────────────────────────────────────────────────────────────

type DataTablePaginationProps<TData = unknown> = {
  table: ReturnType<typeof useReactTable<TData>>;
  className?: string;
};

/**
 * Pagination controls for DataTable. Replaces `@mui/material/TablePagination`.
 *
 * Shows row count, current page range, and previous/next buttons.
 */
function DataTablePagination<TData>({
  table,
  className,
}: DataTablePaginationProps<TData>) {
  const pageIndex = table.getState().pagination.pageIndex;
  const pageSize = table.getState().pagination.pageSize;
  const totalRows = table.getFilteredRowModel().rows.length;
  const start = pageIndex * pageSize + 1;
  const end = Math.min((pageIndex + 1) * pageSize, totalRows);

  return (
    <div
      data-slot="data-table-pagination"
      className={cn('flex items-center justify-between px-2 py-4', className)}
    >
      <p className="text-sm text-muted-foreground">
        {totalRows > 0
          ? `${start}–${end} of ${totalRows} row${totalRows === 1 ? '' : 's'}`
          : 'No results'}
      </p>
      <div className="flex items-center gap-2">
        <Button
          variant="outline"
          size="sm"
          onClick={() => table.previousPage()}
          disabled={!table.getCanPreviousPage()}
        >
          <ChevronLeft data-icon="inline-start" />
          Previous
        </Button>
        <Button
          variant="outline"
          size="sm"
          onClick={() => table.nextPage()}
          disabled={!table.getCanNextPage()}
        >
          Next
          <ChevronRight data-icon="inline-end" />
        </Button>
      </div>
    </div>
  );
}

// ─── DataTable ──────────────────────────────────────────────────────────────

type DataTableProps<TData, TValue> = {
  /** TanStack column definitions. */
  columns: ColumnDef<TData, TValue>[];
  /** Row data. */
  data: TData[];
  /** Rows per page. Set to `0` to disable pagination. @default 10 */
  pageSize?: number;
  /** Content shown when no rows match. */
  noResultsMessage?: string;
  /** Extra props forwarded to `useReactTable`. */
  tableOptions?: Partial<
    Omit<TableOptions<TData>, 'data' | 'columns' | 'getCoreRowModel'>
  >;
  className?: string;
};

/**
 * A fully-featured data table built on TanStack Table and DDS Table primitives.
 *
 * Replaces `@mui/x-data-grid-pro/DataGridPro` for the common case of a
 * sortable, filterable, paginated table. For advanced features (column
 * reordering, virtualisation, tree data) consumers should use TanStack Table
 * directly with the DDS Table primitives.
 *
 * ### Component Props
 *
 * | Prop                  | Type                            | Default        |
 * | :-------------------- | :------------------------------ | :------------- |
 * | _columns_             | `ColumnDef<TData, TValue>[]`    | **required**   |
 * | _data_                | `TData[]`                       | **required**   |
 * | _pageSize_            | `number`                        | `10`           |
 * | _noResultsMessage_    | `string`                        | `"No results"` |
 * | _tableOptions_        | `Partial<TableOptions<TData>>`  | -              |
 *
 * ### Example Usage
 *
 * ```tsx
 * const columns: ColumnDef<Payment>[] = [
 *   { accessorKey: "status", header: "Status" },
 *   { accessorKey: "amount", header: ({ column }) => (
 *     <DataTableColumnHeader header={...} title="Amount" />
 *   )},
 * ];
 *
 * <DataTable columns={columns} data={payments} />
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — renders semantic `<table>`, `<thead>`, `<tbody>`, `<th>`, `<td>`.
 * @wcag 2.1.1 Keyboard — sort buttons and pagination are keyboard-accessible.
 * @wcag 4.1.2 Name, Role, Value — sortable columns expose aria-sort on `<th>`.
 */
function DataTable<TData, TValue>({
  columns,
  data,
  pageSize = 10,
  noResultsMessage = 'No results',
  tableOptions,
  className,
}: DataTableProps<TData, TValue>) {
  const [sorting, setSorting] = useState<SortingState>([]);
  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>([]);
  const [columnVisibility, setColumnVisibility] = useState<VisibilityState>({});

  const table = useReactTable({
    data,
    columns,
    getCoreRowModel: getCoreRowModel(),
    ...(pageSize > 0 && {getPaginationRowModel: getPaginationRowModel()}),
    getSortedRowModel: getSortedRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    onSortingChange: setSorting,
    onColumnFiltersChange: setColumnFilters,
    onColumnVisibilityChange: setColumnVisibility,
    state: {
      sorting,
      columnFilters,
      columnVisibility,
      ...(pageSize > 0 && {pagination: {pageIndex: 0, pageSize}}),
      ...tableOptions?.state,
    },
    ...tableOptions,
  });

  return (
    <div
      data-slot="data-table"
      className={cn('@container/data-table', className)}
    >
      <div className="@max-[400px]/data-table:overflow-x-auto">
        <Table>
          <TableHeader>
            {table.getHeaderGroups().map(headerGroup => (
              <TableRow key={headerGroup.id}>
                {headerGroup.headers.map(header => (
                  <TableHead
                    key={header.id}
                    aria-sort={
                      header.column.getIsSorted() === 'asc'
                        ? 'ascending'
                        : header.column.getIsSorted() === 'desc'
                          ? 'descending'
                          : header.column.getCanSort()
                            ? 'none'
                            : undefined
                    }
                    className="@max-[400px]/data-table:px-2"
                  >
                    {header.isPlaceholder
                      ? null
                      : flexRender(
                          header.column.columnDef.header,
                          header.getContext()
                        )}
                  </TableHead>
                ))}
              </TableRow>
            ))}
          </TableHeader>
          <TableBody>
            {table.getRowModel().rows.length > 0 ? (
              table.getRowModel().rows.map(row => (
                <TableRow
                  key={row.id}
                  data-state={row.getIsSelected() ? 'selected' : undefined}
                >
                  {row.getVisibleCells().map(cell => (
                    <TableCell
                      key={cell.id}
                      className="@max-[400px]/data-table:px-2 @max-[400px]/data-table:py-2 @max-[400px]/data-table:text-xs"
                    >
                      {flexRender(
                        cell.column.columnDef.cell,
                        cell.getContext()
                      )}
                    </TableCell>
                  ))}
                </TableRow>
              ))
            ) : (
              <TableRow>
                <TableCell
                  colSpan={columns.length}
                  className="h-24 text-center text-muted-foreground"
                >
                  {noResultsMessage}
                </TableCell>
              </TableRow>
            )}
          </TableBody>
        </Table>
      </div>
      {pageSize > 0 && <DataTablePagination table={table} />}
    </div>
  );
}

export {DataTable, DataTableColumnHeader, DataTablePagination};
export type {
  DataTableProps,
  DataTableColumnHeaderProps,
  DataTablePaginationProps,
};
