// @docker/trident/src/components/ui/date-picker.tsx
'use client';

import type {DateRange} from 'react-day-picker';

import {Button} from '@/components/ui/button';
import {Calendar} from '@/components/ui/calendar';
import {Popover, PopoverContent, PopoverTrigger} from '@/components/ui/popover';
import {cn} from '@/lib/utils';

import {format} from 'date-fns';
import {CalendarIcon} from 'lucide-react';
import * as React from 'react';

type SharedCalendarProps = Omit<
  React.ComponentProps<typeof Calendar>,
  'mode' | 'selected' | 'onSelect'
>;

// ---------------------------------------------------------------------------
// DatePicker (single)
// ---------------------------------------------------------------------------

type DatePickerProps = {
  value?: Date;
  onChange?: (date: Date | undefined) => void;
  placeholder?: string;
  disabled?: boolean;
  className?: string;
  calendarProps?: SharedCalendarProps;
};

/**
 * A single-date picker composed from Calendar + Popover.
 *
 * ### Example Usage
 *
 * ```tsx
 * const [date, setDate] = React.useState<Date | undefined>();
 *
 * <DatePicker value={date} onChange={setDate} placeholder="Pick a date" />
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — selected date is exposed via button label.
 * @wcag 2.1.1 Keyboard — calendar grid is keyboard navigable via arrow keys.
 * @wcag 4.1.2 Name, Role, Value — trigger button announces selected date.
 */
function DatePicker({
  value,
  onChange,
  placeholder = 'Pick a date',
  disabled,
  className,
  calendarProps,
}: DatePickerProps) {
  return (
    <Popover>
      <PopoverTrigger asChild>
        <Button
          variant="outline"
          data-empty={!value}
          disabled={disabled}
          className={cn(
            'w-[var(--dds-width-date-picker-sm)] justify-start text-left font-normal data-[empty=true]:text-muted-foreground',
            className
          )}
        >
          <CalendarIcon />
          {value ? format(value, 'PPP') : <span>{placeholder}</span>}
        </Button>
      </PopoverTrigger>
      <PopoverContent className="w-auto p-0" align="start">
        <Calendar
          {...calendarProps}
          mode="single"
          selected={value}
          onSelect={onChange}
          initialFocus={calendarProps?.initialFocus ?? true}
        />
      </PopoverContent>
    </Popover>
  );
}

// ---------------------------------------------------------------------------
// DateRangePicker
// ---------------------------------------------------------------------------

type DateRangePickerProps = {
  value?: DateRange;
  onChange?: (range: DateRange | undefined) => void;
  placeholder?: string;
  disabled?: boolean;
  numberOfMonths?: number;
  className?: string;
  calendarProps?: SharedCalendarProps;
};

/**
 * A date-range picker composed from Calendar + Popover.
 * Selects a start and end date.
 *
 * ### Example Usage
 *
 * ```tsx
 * const [range, setRange] = React.useState<DateRange | undefined>();
 *
 * <DateRangePicker value={range} onChange={setRange} />
 * ```
 */
function DateRangePicker({
  value,
  onChange,
  placeholder = 'Pick a date range',
  disabled,
  numberOfMonths = 2,
  className,
  calendarProps,
}: DateRangePickerProps) {
  const label = React.useMemo(() => {
    if (!value?.from) return placeholder;
    if (!value.to) return format(value.from, 'LLL dd, y');
    return `${format(value.from, 'LLL dd, y')} - ${format(value.to, 'LLL dd, y')}`;
  }, [value, placeholder]);

  return (
    <Popover>
      <PopoverTrigger asChild>
        <Button
          variant="outline"
          data-empty={!value?.from}
          disabled={disabled}
          className={cn(
            'w-[var(--dds-width-date-picker-md)] justify-start text-left font-normal data-[empty=true]:text-muted-foreground',
            className
          )}
        >
          <CalendarIcon />
          <span>{label}</span>
        </Button>
      </PopoverTrigger>
      <PopoverContent className="w-auto p-0" align="start">
        <Calendar
          {...calendarProps}
          mode="range"
          selected={value}
          onSelect={onChange}
          numberOfMonths={calendarProps?.numberOfMonths ?? numberOfMonths}
          initialFocus={calendarProps?.initialFocus ?? true}
        />
      </PopoverContent>
    </Popover>
  );
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {DatePicker, DateRangePicker};
export type {DatePickerProps, DateRangePickerProps};
