// @docker/trident/src/components/ui/dialog.tsx
'use client';

import type * as React from 'react';

import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import {X} from 'lucide-react';
import {Dialog as DialogPrimitive} from 'radix-ui';

/**
 * A window overlaid on either the primary window or another dialog window, rendering the content underneath inert.
 *
 * ### Component Props
 *
 * See [Dialog props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop            | Type                           | Default |
 * | :-------------- | :----------------------------- | :------ |
 * | _defaultOpen_   | `boolean`                      | -       |
 * | _open_          | `boolean`                      | -       |
 * | _onOpenChange_  | `(open: boolean) => void`      | -       |
 * | _modal_         | `boolean`                      | `true`  |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Dialog>
 *   <DialogTrigger>Open</DialogTrigger>
 *   <DialogContent>
 *     <DialogHeader>
 *       <DialogTitle>Are you sure?</DialogTitle>
 *       <DialogDescription>
 *         This action cannot be undone.
 *       </DialogDescription>
 *     </DialogHeader>
 *     <DialogFooter>
 *       <Button>Confirm</Button>
 *     </DialogFooter>
 *   </DialogContent>
 * </Dialog>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — role="dialog" with aria-labelledby pointing to DialogTitle.
 * @wcag 2.1.2 No Keyboard Trap — Radix UI constrains focus within the dialog while open; Escape key exits.
 * @wcag 2.4.3 Focus Order — focus moves to dialog on open and returns to trigger on close.
 * @wcag 4.1.2 Name, Role, Value — dialog role and accessible name via DialogTitle (required).
 */
function Dialog({...props}: React.ComponentProps<typeof DialogPrimitive.Root>) {
  return <DialogPrimitive.Root data-slot="dialog" {...props} />;
}

/**
 * The button or element that opens the Dialog when clicked.
 *
 * ### Component Props
 *
 * See [Dialog props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 */
function DialogTrigger({
  ...props
}: React.ComponentProps<typeof DialogPrimitive.Trigger>) {
  return <DialogPrimitive.Trigger data-slot="dialog-trigger" {...props} />;
}

/**
 * A portal container that renders Dialog content outside the DOM hierarchy.
 *
 * ### Component Props
 *
 * See [Dialog props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop           | Type          | Default         |
 * | :------------- | :------------ | :-------------- |
 * | _forceMount_   | `boolean`     | -               |
 * | _container_    | `HTMLElement`  | `document.body` |
 */
function DialogPortal({
  ...props
}: React.ComponentProps<typeof DialogPrimitive.Portal>) {
  return <DialogPrimitive.Portal data-slot="dialog-portal" {...props} />;
}

/**
 * A button that closes the Dialog when clicked.
 *
 * ### Component Props
 *
 * See [Dialog props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 */
function DialogClose({
  ...props
}: React.ComponentProps<typeof DialogPrimitive.Close>) {
  return <DialogPrimitive.Close data-slot="dialog-close" {...props} />;
}

/**
 * A translucent backdrop layer rendered behind the Dialog content.
 *
 * ### Component Props
 *
 * See [Dialog props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop           | Type      | Default |
 * | :------------- | :-------- | :------ |
 * | _asChild_      | `boolean` | `false` |
 * | _forceMount_   | `boolean` | -       |
 */
function DialogOverlay({
  className,
  ...props
}: React.ComponentProps<typeof DialogPrimitive.Overlay>) {
  return (
    <DialogPrimitive.Overlay
      data-slot="dialog-overlay"
      className={cn(
        'fixed inset-0 isolate z-modal bg-overlay-scrim duration-fast ease-out supports-backdrop-filter:backdrop-blur-xs data-open:animate-in data-open:fade-in-0 data-closed:animate-out data-closed:fade-out-0',
        className
      )}
      {...props}
    />
  );
}

/**
 * The main content container for the Dialog, rendered with an overlay and optional close button.
 *
 * ### Component Props
 *
 * See [Dialog props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop                    | Type                                     | Default |
 * | :---------------------- | :--------------------------------------- | :------ |
 * | _asChild_               | `boolean`                                | `false` |
 * | _forceMount_            | `boolean`                                | -       |
 * | _onOpenAutoFocus_       | `(event: Event) => void`                 | -       |
 * | _onCloseAutoFocus_      | `(event: Event) => void`                 | -       |
 * | _onEscapeKeyDown_       | `(event: KeyboardEvent) => void`         | -       |
 * | _onPointerDownOutside_  | `(event: PointerDownOutsideEvent) => void` | -     |
 * | _onInteractOutside_     | `(event: Event) => void`                 | -       |
 * | _showCloseButton_       | `boolean`                                | `true`  |
 */
function DialogContent({
  className,
  children,
  showCloseButton = true,
  ...props
}: React.ComponentProps<typeof DialogPrimitive.Content> & {
  showCloseButton?: boolean;
}) {
  return (
    <DialogPortal>
      <DialogOverlay />
      <DialogPrimitive.Content
        data-slot="dialog-content"
        className={cn(
          'fixed top-1/2 left-1/2 z-modal grid w-full max-w-[calc(100%-2rem)] -translate-x-1/2 -translate-y-1/2 gap-6 rounded-lg bg-popover p-6 text-sm shadow-modal ring-1 ring-border duration-fast ease-out outline-none sm:max-w-md data-open:animate-in data-open:fade-in-0 data-open:zoom-in-95 data-closed:animate-out data-closed:fade-out-0 data-closed:zoom-out-95',
          className
        )}
        {...props}
      >
        {children}
        {showCloseButton && (
          <DialogPrimitive.Close data-slot="dialog-close" asChild>
            <Button
              variant="ghost"
              className="absolute top-4 right-4"
              size="icon-sm"
            >
              <X />
              <span className="sr-only">Close</span>
            </Button>
          </DialogPrimitive.Close>
        )}
      </DialogPrimitive.Content>
    </DialogPortal>
  );
}

/**
 * The header section of a Dialog, typically containing a title and description.
 *
 * ### Component Props
 *
 * This component does not have any component-specific props.
 */
function DialogHeader({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="dialog-header"
      className={cn('flex flex-col gap-2', className)}
      {...props}
    />
  );
}

/**
 * The footer section of a Dialog, typically containing action buttons.
 *
 * ### Component Props
 *
 * | Prop               | Type      | Default |
 * | :----------------- | :-------- | :------ |
 * | _showCloseButton_  | `boolean` | `false` |
 */
function DialogFooter({
  className,
  showCloseButton = false,
  children,
  ...props
}: React.ComponentProps<'div'> & {
  showCloseButton?: boolean;
}) {
  return (
    <div
      data-slot="dialog-footer"
      className={cn(
        'flex flex-col-reverse gap-2 sm:flex-row sm:justify-end',
        className
      )}
      {...props}
    >
      {children}
      {showCloseButton && (
        <DialogPrimitive.Close asChild>
          <Button variant="outline">Close</Button>
        </DialogPrimitive.Close>
      )}
    </div>
  );
}

/**
 * The title element within a DialogHeader, used for accessibility and visual labeling.
 *
 * ### Component Props
 *
 * See [Dialog props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 */
function DialogTitle({
  className,
  ...props
}: React.ComponentProps<typeof DialogPrimitive.Title>) {
  return (
    <DialogPrimitive.Title
      data-slot="dialog-title"
      className={cn('leading-heading font-heading tracking-heading', className)}
      {...props}
    />
  );
}

/**
 * A secondary text element within a DialogHeader, providing additional context or instructions.
 *
 * ### Component Props
 *
 * See [Dialog props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 */
function DialogDescription({
  className,
  ...props
}: React.ComponentProps<typeof DialogPrimitive.Description>) {
  return (
    <DialogPrimitive.Description
      data-slot="dialog-description"
      className={cn(
        'text-sm text-muted-foreground *:[a]:underline *:[a]:underline-offset-3 *:[a]:hover:text-foreground',
        className
      )}
      {...props}
    />
  );
}

export {
  Dialog,
  DialogClose,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogOverlay,
  DialogPortal,
  DialogTitle,
  DialogTrigger,
};
