// @docker/trident/src/components/ui/drawer.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {Drawer as DrawerPrimitive} from 'vaul';

const drawerContentVariants = cva('', {
  variants: {
    size: {
      default: '',
      sm: 'data-[vaul-drawer-direction=left]:sm:max-w-xs data-[vaul-drawer-direction=right]:sm:max-w-xs',
      lg: 'data-[vaul-drawer-direction=left]:sm:max-w-2xl data-[vaul-drawer-direction=right]:sm:max-w-2xl',
      full: 'data-[vaul-drawer-direction=left]:w-screen data-[vaul-drawer-direction=right]:w-screen data-[vaul-drawer-direction=left]:sm:max-w-none data-[vaul-drawer-direction=right]:sm:max-w-none',
    },
  },
  defaultVariants: {
    size: 'default',
  },
});

/**
 * A draggable drawer component built on top of Vaul. Supports multiple directions and snap points.
 *
 * ### Component Props
 *
 * See [Vaul Drawer props](https://vaul.emilkowal.ski/api) documentation for full details.
 *
 * | Prop                 | Type                                          | Default    |
 * | :------------------- | :-------------------------------------------- | :--------- |
 * | _defaultOpen_        | `boolean`                                     | -          |
 * | _open_               | `boolean`                                     | -          |
 * | _onOpenChange_       | `(open: boolean) => void`                     | -          |
 * | _modal_              | `boolean`                                     | `true`     |
 * | _direction_          | `"top" \| "right" \| "bottom" \| "left"`      | `"bottom"` |
 * | _dismissible_        | `boolean`                                     | `true`     |
 * | _handleOnly_         | `boolean`                                     | `false`    |
 * | _snapPoints_         | `(number \| string)[]`                        | -          |
 * | _repositionInputs_   | `boolean`                                     | `true`     |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Drawer>
 *   <DrawerTrigger>Open Drawer</DrawerTrigger>
 *   <DrawerContent>
 *     <DrawerHeader>
 *       <DrawerTitle>Drawer Title</DrawerTitle>
 *       <DrawerDescription>Drawer description text.</DrawerDescription>
 *     </DrawerHeader>
 *     <div className="p-4">Drawer body content.</div>
 *     <DrawerFooter>
 *       <DrawerClose>Close</DrawerClose>
 *     </DrawerFooter>
 *   </DrawerContent>
 * </Drawer>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 2.1.2 No Keyboard Trap — Escape key closes the drawer; focus returns to trigger.
 * @wcag 2.4.3 Focus Order — focus is trapped within the drawer while open via Vaul primitive.
 * @wcag 4.1.2 Name, Role, Value — `dialog` role with `aria-labelledby` / `aria-describedby` via Vaul primitives.
 */
function Drawer({...props}: React.ComponentProps<typeof DrawerPrimitive.Root>) {
  return <DrawerPrimitive.Root data-slot="drawer" {...props} />;
}

/**
 * A button that opens the drawer.
 *
 * ### Component Props
 *
 * See [Vaul Drawer props](https://vaul.emilkowal.ski/api) documentation for full details.
 *
 * | Prop       | Type      | Default |
 * | :--------- | :-------- | :------ |
 * | _asChild_  | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <DrawerTrigger>Open Drawer</DrawerTrigger>
 * ```
 */
function DrawerTrigger({
  ...props
}: React.ComponentProps<typeof DrawerPrimitive.Trigger>) {
  return <DrawerPrimitive.Trigger data-slot="drawer-trigger" {...props} />;
}

/**
 * Renders the drawer overlay and content into a portal, appended to `document.body`.
 *
 * ### Example Usage
 *
 * ```tsx
 * <DrawerPortal>
 *   <DrawerOverlay />
 *   <DrawerContent>...</DrawerContent>
 * </DrawerPortal>
 * ```
 */
function DrawerPortal({
  ...props
}: React.ComponentProps<typeof DrawerPrimitive.Portal>) {
  return <DrawerPrimitive.Portal data-slot="drawer-portal" {...props} />;
}

/**
 * A button that closes the drawer.
 *
 * ### Component Props
 *
 * See [Vaul Drawer props](https://vaul.emilkowal.ski/api) documentation for full details.
 *
 * | Prop       | Type      | Default |
 * | :--------- | :-------- | :------ |
 * | _asChild_  | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <DrawerClose>Close</DrawerClose>
 * ```
 */
function DrawerClose({
  ...props
}: React.ComponentProps<typeof DrawerPrimitive.Close>) {
  return <DrawerPrimitive.Close data-slot="drawer-close" {...props} />;
}

/**
 * A layer that covers the inert portion of the view when the drawer is open.
 *
 * ### Component Props
 *
 * See [Vaul Drawer props](https://vaul.emilkowal.ski/api) documentation for full details.
 *
 * | Prop       | Type      | Default |
 * | :--------- | :-------- | :------ |
 * | _asChild_  | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <DrawerOverlay />
 * ```
 */
function DrawerOverlay({
  className,
  ...props
}: React.ComponentProps<typeof DrawerPrimitive.Overlay>) {
  return (
    <DrawerPrimitive.Overlay
      data-slot="drawer-overlay"
      className={cn(
        'fixed inset-0 z-modal bg-dds-overlay-backdrop supports-backdrop-filter:backdrop-blur-xs data-open:animate-in data-open:fade-in-0 data-closed:animate-out data-closed:fade-out-0',
        className
      )}
      {...props}
    />
  );
}

/**
 * Contains the content of the drawer. Automatically wraps itself in a `DrawerPortal` and `DrawerOverlay`.
 *
 * ### Component Props
 *
 * See [Vaul Drawer props](https://vaul.emilkowal.ski/api) documentation for full details.
 *
 * | Prop       | Type      | Default |
 * | :--------- | :-------- | :------ |
 * | _asChild_  | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <DrawerContent>
 *   <DrawerHeader>
 *     <DrawerTitle>Title</DrawerTitle>
 *   </DrawerHeader>
 *   <div className="p-4">Content here</div>
 * </DrawerContent>
 * ```
 */
function DrawerContent({
  className,
  children,
  size,
  ...props
}: React.ComponentProps<typeof DrawerPrimitive.Content> &
  VariantProps<typeof drawerContentVariants>) {
  return (
    <DrawerPortal data-slot="drawer-portal">
      <DrawerOverlay />
      <DrawerPrimitive.Content
        data-slot="drawer-content"
        className={cn(
          'group/drawer-content fixed z-modal flex h-auto flex-col bg-background text-sm shadow-modal data-[vaul-drawer-direction=bottom]:inset-x-0 data-[vaul-drawer-direction=bottom]:bottom-0 data-[vaul-drawer-direction=bottom]:mt-24 data-[vaul-drawer-direction=bottom]:max-h-[80vh] data-[vaul-drawer-direction=bottom]:rounded-t-xl data-[vaul-drawer-direction=bottom]:border-t data-[vaul-drawer-direction=bottom]:border-border data-[vaul-drawer-direction=left]:inset-y-0 data-[vaul-drawer-direction=left]:left-0 data-[vaul-drawer-direction=left]:w-3/4 data-[vaul-drawer-direction=left]:rounded-r-xl data-[vaul-drawer-direction=left]:border-r data-[vaul-drawer-direction=right]:inset-y-0 data-[vaul-drawer-direction=right]:right-0 data-[vaul-drawer-direction=right]:w-3/4 data-[vaul-drawer-direction=right]:rounded-l-xl data-[vaul-drawer-direction=right]:border-l data-[vaul-drawer-direction=top]:inset-x-0 data-[vaul-drawer-direction=top]:top-0 data-[vaul-drawer-direction=top]:mb-24 data-[vaul-drawer-direction=top]:max-h-[80vh] data-[vaul-drawer-direction=top]:rounded-b-xl data-[vaul-drawer-direction=top]:border-b data-[vaul-drawer-direction=top]:border-border data-[vaul-drawer-direction=left]:sm:max-w-sm data-[vaul-drawer-direction=right]:sm:max-w-sm',
          drawerContentVariants({size}),
          className
        )}
        {...props}
      >
        <div className="mx-auto mt-4 hidden h-1.5 w-25 shrink-0 rounded-full bg-muted group-data-[vaul-drawer-direction=bottom]/drawer-content:block" />
        {children}
      </DrawerPrimitive.Content>
    </DrawerPortal>
  );
}

/**
 * A container for the drawer's heading elements, typically wrapping `DrawerTitle` and `DrawerDescription`.
 *
 * ### Example Usage
 *
 * ```tsx
 * <DrawerHeader>
 *   <DrawerTitle>Title</DrawerTitle>
 *   <DrawerDescription>Description text.</DrawerDescription>
 * </DrawerHeader>
 * ```
 */
function DrawerHeader({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="drawer-header"
      className={cn(
        'flex flex-col gap-0.5 p-4 group-data-[vaul-drawer-direction=bottom]/drawer-content:text-center group-data-[vaul-drawer-direction=top]/drawer-content:text-center md:gap-1.5 md:text-left',
        className
      )}
      {...props}
    />
  );
}

/**
 * A container for the drawer's footer actions, typically wrapping buttons or a `DrawerClose`.
 *
 * ### Example Usage
 *
 * ```tsx
 * <DrawerFooter>
 *   <Button>Submit</Button>
 *   <DrawerClose>Cancel</DrawerClose>
 * </DrawerFooter>
 * ```
 */
function DrawerFooter({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="drawer-footer"
      className={cn('mt-auto flex flex-col gap-2 p-4', className)}
      {...props}
    />
  );
}

/**
 * An accessible title to be announced when the drawer is opened.
 *
 * ### Component Props
 *
 * See [Vaul Drawer props](https://vaul.emilkowal.ski/api) documentation for full details.
 *
 * | Prop       | Type      | Default |
 * | :--------- | :-------- | :------ |
 * | _asChild_  | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <DrawerTitle>Edit Profile</DrawerTitle>
 * ```
 */
function DrawerTitle({
  className,
  ...props
}: React.ComponentProps<typeof DrawerPrimitive.Title>) {
  return (
    <DrawerPrimitive.Title
      data-slot="drawer-title"
      className={cn(
        'leading-heading font-heading tracking-heading text-foreground',
        className
      )}
      {...props}
    />
  );
}

/**
 * An accessible description to be announced when the drawer is opened.
 *
 * ### Component Props
 *
 * See [Vaul Drawer props](https://vaul.emilkowal.ski/api) documentation for full details.
 *
 * | Prop       | Type      | Default |
 * | :--------- | :-------- | :------ |
 * | _asChild_  | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <DrawerDescription>
 *   Make changes to your profile here.
 * </DrawerDescription>
 * ```
 */
function DrawerDescription({
  className,
  ...props
}: React.ComponentProps<typeof DrawerPrimitive.Description>) {
  return (
    <DrawerPrimitive.Description
      data-slot="drawer-description"
      className={cn('text-sm text-muted-foreground', className)}
      {...props}
    />
  );
}

export {
  Drawer,
  DrawerPortal,
  DrawerOverlay,
  DrawerTrigger,
  DrawerClose,
  DrawerContent,
  DrawerHeader,
  DrawerFooter,
  DrawerTitle,
  DrawerDescription,
};
