// @docker/trident/src/components/ui/empty.tsx
import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';

/**
 * A placeholder component for empty states, used to communicate that a section has no content yet.
 *
 * ### Example Usage
 *
 * ```tsx
 * <Empty>
 *   <EmptyHeader>
 *     <EmptyMedia variant="icon">
 *       <InboxIcon />
 *     </EmptyMedia>
 *     <EmptyTitle>No messages</EmptyTitle>
 *     <EmptyDescription>You have no messages in your inbox.</EmptyDescription>
 *   </EmptyHeader>
 *   <EmptyContent>
 *     <Button>Compose</Button>
 *   </EmptyContent>
 * </Empty>
 * ```
 */
function Empty({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="empty"
      className={cn(
        'flex w-full min-w-0 flex-1 flex-col items-center justify-center gap-4 rounded border-dashed p-12 text-center text-balance',
        className
      )}
      {...props}
    />
  );
}

/**
 * A container for the heading area of an empty state, typically wrapping `EmptyMedia`, `EmptyTitle`, and `EmptyDescription`.
 *
 * ### Example Usage
 *
 * ```tsx
 * <EmptyHeader>
 *   <EmptyTitle>No results</EmptyTitle>
 *   <EmptyDescription>Try adjusting your search.</EmptyDescription>
 * </EmptyHeader>
 * ```
 */
function EmptyHeader({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="empty-header"
      className={cn('flex max-w-sm flex-col items-center gap-2', className)}
      {...props}
    />
  );
}

const emptyMediaVariants = cva(
  'mb-2 flex shrink-0 items-center justify-center [&_svg]:pointer-events-none [&_svg]:shrink-0',
  {
    variants: {
      variant: {
        default: 'bg-transparent',
        icon: "flex size-10 shrink-0 items-center justify-center rounded bg-muted text-foreground [&_svg:not([class*='size-'])]:size-6",
      },
    },
    defaultVariants: {
      variant: 'default',
    },
  }
);

/**
 * A media container for an icon or illustration within an empty state.
 *
 * ### Component Props
 *
 * | Prop       | Type                     | Default     |
 * | :--------- | :----------------------- | :---------- |
 * | _variant_  | `"default" \| "icon"`    | `"default"` |
 *
 *
 * ### Example Usage
 *
 * **Default**
 * ```tsx
 * <EmptyMedia>
 *   <LargeIllustration />
 * </EmptyMedia>
 * ```
 *
 * **Icon**
 * ```tsx
 * <EmptyMedia variant="icon">
 *   <InboxIcon />
 * </EmptyMedia>
 * ```
 */
function EmptyMedia({
  className,
  variant = 'default',
  ...props
}: React.ComponentProps<'div'> & VariantProps<typeof emptyMediaVariants>) {
  return (
    <div
      data-slot="empty-icon"
      data-variant={variant}
      className={cn(emptyMediaVariants({variant, className}))}
      {...props}
    />
  );
}

/**
 * The title text for an empty state.
 *
 * ### Example Usage
 *
 * ```tsx
 * <EmptyTitle>No items found</EmptyTitle>
 * ```
 */
function EmptyTitle({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="empty-title"
      className={cn(
        'text-lg leading-heading font-heading tracking-heading',
        className
      )}
      {...props}
    />
  );
}

/**
 * A supporting description for an empty state.
 *
 * ### Example Usage
 *
 * ```tsx
 * <EmptyDescription>
 *   There are no items to display. Try creating a new one.
 * </EmptyDescription>
 * ```
 */
function EmptyDescription({className, ...props}: React.ComponentProps<'p'>) {
  return (
    <div
      data-slot="empty-description"
      className={cn(
        'text-sm/relaxed text-muted-foreground [&>a]:underline [&>a]:underline-offset-4 [&>a:hover]:text-primary',
        className
      )}
      {...props}
    />
  );
}

/**
 * A container for actionable content within an empty state, such as buttons or links.
 *
 * ### Example Usage
 *
 * ```tsx
 * <EmptyContent>
 *   <Button>Create Item</Button>
 * </EmptyContent>
 * ```
 */
function EmptyContent({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="empty-content"
      className={cn(
        'flex w-full max-w-sm min-w-0 flex-col items-center gap-4 text-sm text-balance',
        className
      )}
      {...props}
    />
  );
}

export {
  Empty,
  EmptyHeader,
  EmptyTitle,
  EmptyDescription,
  EmptyContent,
  EmptyMedia,
};
