// @docker/trident/src/components/ui/field.tsx
'use client';

import {Label} from '@/components/ui/label';
import {Separator} from '@/components/ui/separator';
import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {useMemo} from 'react';

/**
 * A semantic `<fieldset>` wrapper that groups related form fields together.
 *
 * ### Example Usage
 *
 * ```tsx
 * <FieldSet>
 *   <FieldLegend>Personal Information</FieldLegend>
 *   <FieldGroup>
 *     <Field>
 *       <FieldLabel>Name</FieldLabel>
 *       <Input placeholder="Enter your name" />
 *     </Field>
 *   </FieldGroup>
 * </FieldSet>
 * ```
 */
function FieldSet({className, ...props}: React.ComponentProps<'fieldset'>) {
  return (
    <fieldset
      data-slot="field-set"
      className={cn(
        'flex flex-col gap-6 has-[>[data-slot=checkbox-group]]:gap-3 has-[>[data-slot=radio-group]]:gap-3',
        className
      )}
      {...props}
    />
  );
}

/**
 * A `<legend>` element for a `FieldSet`, providing a heading for the group of fields.
 *
 * ### Component Props
 *
 * | Prop       | Type                     | Default    |
 * | :--------- | :----------------------- | :--------- |
 * | _variant_  | `"legend" \| "label"`    | `"legend"` |
 *
 *
 * ### Example Usage
 *
 * **Legend**
 * ```tsx
 * <FieldLegend>Personal Information</FieldLegend>
 * ```
 *
 * **Label**
 * ```tsx
 * <FieldLegend variant="label">Account Settings</FieldLegend>
 * ```
 */
function FieldLegend({
  className,
  variant = 'legend',
  ...props
}: React.ComponentProps<'legend'> & {variant?: 'legend' | 'label'}) {
  return (
    <legend
      data-slot="field-legend"
      data-variant={variant}
      className={cn(
        'mb-3 font-label tracking-label data-[variant=label]:text-sm data-[variant=legend]:text-base',
        className
      )}
      {...props}
    />
  );
}

/**
 * A container that groups multiple `Field` components with consistent spacing.
 *
 * ### Example Usage
 *
 * ```tsx
 * <FieldGroup>
 *   <Field>
 *     <FieldLabel>First Name</FieldLabel>
 *     <Input />
 *   </Field>
 *   <Field>
 *     <FieldLabel>Last Name</FieldLabel>
 *     <Input />
 *   </Field>
 * </FieldGroup>
 * ```
 */
function FieldGroup({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="field-group"
      className={cn(
        'group/field-group @container/field-group flex w-full flex-col gap-7 data-[slot=checkbox-group]:gap-3 *:data-[slot=field-group]:gap-4',
        className
      )}
      {...props}
    />
  );
}

const fieldVariants = cva(
  'group/field flex w-full gap-3 data-[invalid=true]:text-destructive',
  {
    variants: {
      orientation: {
        vertical: 'flex-col [&>*]:w-full [&>.sr-only]:w-auto',
        horizontal:
          'flex-row items-center has-[>[data-slot=field-content]]:items-start [&>[data-slot=field-label]]:flex-auto has-[>[data-slot=field-content]]:[&>[role=checkbox],[role=radio]]:mt-px',
        responsive:
          'flex-col @md/field-group:flex-row @md/field-group:items-center @md/field-group:has-[>[data-slot=field-content]]:items-start [&>*]:w-full @md/field-group:[&>*]:w-auto [&>.sr-only]:w-auto @md/field-group:[&>[data-slot=field-label]]:flex-auto @md/field-group:has-[>[data-slot=field-content]]:[&>[role=checkbox],[role=radio]]:mt-px',
      },
    },
    defaultVariants: {
      orientation: 'vertical',
    },
  }
);

/**
 * A form field wrapper that arranges a label, input, description, and error message with proper layout.
 *
 * ### Component Props
 *
 * | Prop            | Type                                          | Default      |
 * | :-------------- | :-------------------------------------------- | :----------- |
 * | _orientation_   | `"vertical" \| "horizontal" \| "responsive"`  | `"vertical"` |
 *
 *
 * ### Example Usage
 *
 * **Vertical**
 * ```tsx
 * <Field>
 *   <FieldLabel>Email</FieldLabel>
 *   <FieldContent>
 *     <Input type="email" placeholder="you@example.com" />
 *     <FieldDescription>We will never share your email.</FieldDescription>
 *   </FieldContent>
 * </Field>
 * ```
 *
 * **Horizontal**
 * ```tsx
 * <Field orientation="horizontal">
 *   <FieldLabel>Name</FieldLabel>
 *   <FieldContent>
 *     <Input placeholder="Enter your name" />
 *   </FieldContent>
 * </Field>
 * ```
 *
 * **Responsive**
 * ```tsx
 * <Field orientation="responsive">
 *   <FieldLabel>Bio</FieldLabel>
 *   <FieldContent>
 *     <Textarea placeholder="Tell us about yourself" />
 *   </FieldContent>
 * </Field>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — FieldLabel associates with input via htmlFor/id.
 * @wcag 3.3.1 Error Identification — FieldError uses role="alert" to announce errors to screen readers.
 * @wcag 3.3.2 Labels or Instructions — FieldLabel and FieldDescription provide context for the input.
 * @wcag 4.1.3 Status Messages — error messages announced via role="alert" without receiving focus.
 */
function Field({
  className,
  orientation = 'vertical',
  ...props
}: React.ComponentProps<'fieldset'> & VariantProps<typeof fieldVariants>) {
  return (
    <fieldset
      data-slot="field"
      data-orientation={orientation}
      className={cn(fieldVariants({orientation}), className)}
      {...props}
    />
  );
}

/**
 * A container for a field's input, description, and error message content.
 *
 * ### Example Usage
 *
 * ```tsx
 * <FieldContent>
 *   <Input />
 *   <FieldDescription>Helper text goes here.</FieldDescription>
 * </FieldContent>
 * ```
 */
function FieldContent({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="field-content"
      className={cn(
        'group/field-content flex flex-1 flex-col gap-1 leading-snug',
        className
      )}
      {...props}
    />
  );
}

/**
 * A label element for a form field. Wraps the `Label` component with field-specific styling.
 *
 * ### Example Usage
 *
 * ```tsx
 * <FieldLabel htmlFor="name">Full Name</FieldLabel>
 * ```
 */
function FieldLabel({className, ...props}: React.ComponentProps<typeof Label>) {
  return (
    <Label
      data-slot="field-label"
      className={cn(
        'group/field-label peer/field-label flex w-fit gap-2 leading-snug group-data-[disabled=true]/field:opacity-50 has-data-checked:border-accent has-data-checked:bg-accent-wash has-[>[data-slot=field]]:rounded has-[>[data-slot=field]]:border *:data-[slot=field]:p-3',
        'has-[>[data-slot=field]]:w-full has-[>[data-slot=field]]:flex-col',
        className
      )}
      {...props}
    />
  );
}

/**
 * A non-label title element for a form field, useful when a `<label>` is not needed.
 *
 * ### Example Usage
 *
 * ```tsx
 * <FieldTitle>Notification Preferences</FieldTitle>
 * ```
 */
function FieldTitle({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="field-label"
      className={cn(
        'flex w-fit items-center gap-2 text-label leading-label font-label tracking-label group-data-[disabled=true]/field:opacity-50',
        className
      )}
      {...props}
    />
  );
}

/**
 * A helper text element that provides additional context or instructions for a form field.
 *
 * ### Example Usage
 *
 * ```tsx
 * <FieldDescription>
 *   Must be at least 8 characters long.
 * </FieldDescription>
 * ```
 */
function FieldDescription({className, ...props}: React.ComponentProps<'p'>) {
  return (
    <p
      data-slot="field-description"
      className={cn(
        'text-left text-sm leading-normal font-normal text-muted-foreground group-has-data-[orientation=horizontal]/field:text-balance [[data-variant=legend]+&]:-mt-1.5',
        'last:mt-0 nth-last-2:-mt-1',
        '[&>a]:underline [&>a]:underline-offset-4 [&>a:hover]:text-primary',
        className
      )}
      {...props}
    />
  );
}

/**
 * A visual separator between fields, optionally displaying text content in the center.
 *
 * ### Example Usage
 *
 * ```tsx
 * <Field>...</Field>
 * <FieldSeparator>or</FieldSeparator>
 * <Field>...</Field>
 * ```
 */
function FieldSeparator({
  children,
  className,
  ...props
}: React.ComponentProps<'div'> & {
  children?: React.ReactNode;
}) {
  return (
    <div
      data-slot="field-separator"
      data-content={!!children}
      className={cn(
        'relative -my-2 h-5 text-sm group-data-[variant=outline]/field-group:-mb-2',
        className
      )}
      {...props}
    >
      <Separator className="absolute inset-0 top-1/2" />
      {children && (
        <span
          className="relative mx-auto block w-fit bg-background px-2 text-muted-foreground"
          data-slot="field-separator-content"
        >
          {children}
        </span>
      )}
    </div>
  );
}

/**
 * Displays validation error messages for a form field. Supports both direct children and an `errors` array.
 *
 * ### Component Props
 *
 * | Prop      | Type                                        | Default |
 * | :-------- | :------------------------------------------ | :------ |
 * | _errors_  | `Array<{ message?: string } \| undefined>`  | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <FieldError>This field is required.</FieldError>
 *
 * <FieldError errors={[{ message: "Too short" }, { message: "Invalid format" }]} />
 * ```
 *
 * ```tsx
 * <Field>
 *   <FieldLabel htmlFor="email">Email</FieldLabel>
 *   <Input id="email" aria-describedby="email-error" aria-invalid={hasError} />
 *   <FieldError id="email-error" errors={errors} />
 * </Field>
 * ```
 *
 * Pass `id` and set `aria-describedby={id}` on the associated input for screen reader error announcement.
 */
function FieldError({
  className,
  children,
  errors,
  ...props
}: React.ComponentProps<'div'> & {
  errors?: Array<{message?: string} | undefined>;
}) {
  const content = useMemo(() => {
    if (children) {
      return children;
    }

    if (!errors?.length) {
      return null;
    }

    const uniqueErrors = [
      ...new Map(errors.map(error => [error?.message, error])).values(),
    ];

    if (uniqueErrors?.length === 1) {
      return uniqueErrors[0]?.message;
    }

    return (
      <ul className="ml-4 flex list-disc flex-col gap-1">
        {uniqueErrors.map(
          error =>
            error?.message && <li key={error.message}>{error.message}</li>
        )}
      </ul>
    );
  }, [children, errors]);

  if (!content) {
    return null;
  }

  return (
    <div
      role="alert"
      data-slot="field-error"
      className={cn('text-sm font-normal text-destructive', className)}
      {...props}
    >
      {content}
    </div>
  );
}

export {
  Field,
  FieldLabel,
  FieldDescription,
  FieldError,
  FieldGroup,
  FieldLegend,
  FieldSeparator,
  FieldSet,
  FieldContent,
  FieldTitle,
};
