// @docker/trident/src/components/ui/field-counter.tsx
'use client';

import type {ChangeEvent} from 'react';

import {cn} from '@/lib/utils';

import {useCallback, useState} from 'react';

// ─── Types ────────────────────────────────────────────────────────────────────

type FieldCounterProps = {
  /**
   * Current character count or the input value to measure.
   * If a string is passed, `.length` is used. If a number is passed, it is used directly.
   */
  value?: string | number;
  /** Maximum character count. */
  max: number;
  /**
   * Percentage of max at which the warning color applies.
   * @default 80
   */
  warnAt?: number;
  /**
   * Percentage of max at which the error color applies.
   * @default 100
   */
  errorAt?: number;
  className?: string;
  /** For `aria-describedby` linking on the associated input. */
  id?: string;
};

type UseFieldCounterReturn = {
  value: string;
  count: number;
  inputProps: {
    value: string;
    onChange: (e: ChangeEvent<HTMLInputElement>) => void;
    maxLength: number;
  };
  counterProps: {
    value: number;
    max: number;
  };
};

// ─── Component ────────────────────────────────────────────────────────────────

/**
 * Displays a character count indicator in the format "X / MAX" for use alongside
 * Input or Textarea elements in the Field system.
 *
 * ### Component Props
 *
 * | Prop       | Type                  | Default |
 * | :--------- | :-------------------- | :------ |
 * | _value_    | `string \| number`    | -       |
 * | _max_      | `number`              | -       |
 * | _warnAt_   | `number`              | `80`    |
 * | _errorAt_  | `number`              | `100`   |
 *
 * ### Example Usage
 *
 * ```tsx
 * <Field>
 *   <FieldLabel htmlFor="org-name">Organization name</FieldLabel>
 *   <FieldContent>
 *     <Input
 *       id="org-name"
 *       value={value}
 *       onChange={(e) => setValue(e.target.value)}
 *       maxLength={50}
 *       aria-describedby="org-name-counter"
 *     />
 *     <div className="flex items-center justify-between">
 *       <FieldDescription>Must be unique across Docker Hub</FieldDescription>
 *       <FieldCounter id="org-name-counter" value={value} max={50} />
 *     </div>
 *   </FieldContent>
 * </Field>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 4.1.3 Status Messages — uses `role="status"` and `aria-live="polite"` so screen readers
 *   announce count changes without moving focus.
 * @wcag 1.4.3 Contrast (Minimum) — warning state uses `text-warning` (≥4.5:1 on background in all modes).
 * @wcag 1.3.1 Info and Relationships — hidden descriptive text conveys meaning beyond color alone.
 */
const FieldCounter = ({
  value,
  max,
  warnAt = 80,
  errorAt = 100,
  className,
  id,
}: FieldCounterProps) => {
  const count = typeof value === 'string' ? value.length : (value ?? 0);
  const pct = max > 0 ? (count / max) * 100 : 0;

  const isError = pct >= errorAt || count >= max;
  const isWarn = !isError && pct >= warnAt;

  const colorClass = isError
    ? 'text-destructive high-contrast:text-tag-destructive-fg'
    : isWarn
      ? 'text-warning'
      : 'text-muted-foreground high-contrast:text-foreground';

  return (
    <output
      id={id}
      aria-live="polite"
      data-slot="field-counter"
      className={cn(
        'inline-flex items-center justify-end text-sm tabular-nums',
        colorClass,
        className
      )}
    >
      {/* Visible display */}
      <span aria-hidden="true">
        {count} / {max}
      </span>
      {/* Screen reader text */}
      <span className="sr-only">
        {count} of {max} characters used
      </span>
    </output>
  );
};

// ─── Hook ─────────────────────────────────────────────────────────────────────

/**
 * Convenience hook that wires up a controlled input with a FieldCounter.
 *
 * ### Example Usage
 *
 * ```tsx
 * const { inputProps, counterProps } = useFieldCounter(50);
 *
 * <Input {...inputProps} />
 * <FieldCounter {...counterProps} />
 * ```
 */
const useFieldCounter = (max: number): UseFieldCounterReturn => {
  const [value, setValue] = useState('');

  const onChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    setValue(e.target.value);
  }, []);

  return {
    value,
    count: value.length,
    inputProps: {
      value,
      onChange,
      maxLength: max,
    },
    counterProps: {
      value: value.length,
      max,
    },
  };
};

export {FieldCounter, useFieldCounter};
export type {FieldCounterProps, UseFieldCounterReturn};
