// @docker/trident/src/components/ui/form.tsx
'use client';

import {Label} from '@/components/ui/label';
import {cn} from '@/lib/utils';

import {Slot} from 'radix-ui';
import * as React from 'react';
import {
  Controller,
  type ControllerProps,
  type FieldPath,
  type FieldValues,
  FormProvider,
  useFormContext,
} from 'react-hook-form';

// ---------------------------------------------------------------------------
// Form (re-export of FormProvider)
// ---------------------------------------------------------------------------

/**
 * Wraps `react-hook-form`'s `FormProvider`. Place this at the root of your
 * form, passing the `useForm()` return value as props.
 *
 * ### Example Usage
 *
 * ```tsx
 * const form = useForm<FormValues>();
 *
 * <Form {...form}>
 *   <form onSubmit={form.handleSubmit(onSubmit)}>
 *     <FormField
 *       control={form.control}
 *       name="email"
 *       render={({field}) => (
 *         <FormItem>
 *           <FormLabel>Email</FormLabel>
 *           <FormControl><Input {...field} /></FormControl>
 *           <FormMessage />
 *         </FormItem>
 *       )}
 *     />
 *   </form>
 * </Form>
 * ```
 */
const Form = FormProvider;

// ---------------------------------------------------------------------------
// Internal context
// ---------------------------------------------------------------------------

type FormFieldContextValue<
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> = {name: TName};

const FormFieldContext = React.createContext<FormFieldContextValue | null>(
  null
);

type FormItemContextValue = {id: string};

const FormItemContext = React.createContext<FormItemContextValue | null>(null);

// ---------------------------------------------------------------------------
// useFormField
// ---------------------------------------------------------------------------

/**
 * Returns the current field's state, IDs for aria wiring, and react-hook-form
 * field state. Must be called inside a `<FormField>`.
 */
function useFormField() {
  const fieldContext = React.useContext(FormFieldContext);
  const itemContext = React.useContext(FormItemContext);
  const {getFieldState, formState} = useFormContext();

  if (!fieldContext) {
    throw new Error('useFormField must be used inside <FormField>');
  }

  const fieldState = getFieldState(fieldContext.name, formState);
  const {id} = itemContext ?? {id: ''};

  return {
    id,
    name: fieldContext.name,
    formItemId: `${id}-form-item`,
    formDescriptionId: `${id}-form-item-description`,
    formMessageId: `${id}-form-item-message`,
    ...fieldState,
  };
}

// ---------------------------------------------------------------------------
// FormField
// ---------------------------------------------------------------------------

/**
 * Connects a form control to `react-hook-form` via `Controller`. Pass a
 * `render` prop that receives `{field, fieldState, formState}`.
 *
 * ### Example Usage
 *
 * ```tsx
 * <FormField
 *   control={form.control}
 *   name="username"
 *   render={({field}) => (
 *     <FormItem>
 *       <FormLabel>Username</FormLabel>
 *       <FormControl><Input {...field} /></FormControl>
 *       <FormMessage />
 *     </FormItem>
 *   )}
 * />
 * ```
 */
function FormField<
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
>(props: ControllerProps<TFieldValues, TName>) {
  const value = React.useMemo(() => ({name: props.name}), [props.name]);
  return (
    <FormFieldContext.Provider value={value}>
      <Controller {...props} />
    </FormFieldContext.Provider>
  );
}

// ---------------------------------------------------------------------------
// FormItem
// ---------------------------------------------------------------------------

/**
 * Wraps a single field with a stable `id` context used by `FormLabel`,
 * `FormControl`, `FormDescription`, and `FormMessage` for aria wiring.
 */
function FormItem({className, ...props}: React.ComponentProps<'div'>) {
  const id = React.useId();
  const value = React.useMemo(() => ({id}), [id]);

  return (
    <FormItemContext.Provider value={value}>
      <div
        data-slot="form-item"
        className={cn('flex flex-col gap-dds-100', className)}
        {...props}
      />
    </FormItemContext.Provider>
  );
}

// ---------------------------------------------------------------------------
// FormLabel
// ---------------------------------------------------------------------------

/**
 * A `Label` pre-wired to the sibling `FormControl` via `htmlFor`.
 * Turns destructive-colored when the field has a validation error.
 */
function FormLabel({className, ...props}: React.ComponentProps<typeof Label>) {
  const {error, formItemId} = useFormField();

  return (
    <Label
      data-slot="form-label"
      htmlFor={formItemId}
      className={cn(error && 'text-destructive', className)}
      {...props}
    />
  );
}

// ---------------------------------------------------------------------------
// FormControl
// ---------------------------------------------------------------------------

/**
 * A `Slot` that injects `id`, `aria-describedby`, and `aria-invalid` onto
 * the wrapped control so screen readers associate it with its label,
 * description, and error message automatically.
 */
function FormControl({...props}: React.ComponentProps<typeof Slot.Root>) {
  const {error, formItemId, formDescriptionId, formMessageId} = useFormField();

  return (
    <Slot.Root
      data-slot="form-control"
      id={formItemId}
      aria-describedby={
        error ? `${formDescriptionId} ${formMessageId}` : formDescriptionId
      }
      aria-invalid={Boolean(error)}
      {...props}
    />
  );
}

// ---------------------------------------------------------------------------
// FormDescription
// ---------------------------------------------------------------------------

/**
 * Helper text displayed below the control. Wired to `aria-describedby` on
 * the `FormControl`.
 */
function FormDescription({className, ...props}: React.ComponentProps<'p'>) {
  const {formDescriptionId} = useFormField();

  return (
    <p
      data-slot="form-description"
      id={formDescriptionId}
      className={cn(
        'text-xs text-muted-foreground high-contrast:text-foreground',
        className
      )}
      {...props}
    />
  );
}

// ---------------------------------------------------------------------------
// FormMessage
// ---------------------------------------------------------------------------

/**
 * Renders the field's validation error. Falls back to `children` when there
 * is no error. Renders nothing when both are absent.
 */
function FormMessage({
  className,
  children,
  ...props
}: React.ComponentProps<'p'>) {
  const {error, formMessageId} = useFormField();
  const body = error ? String(error.message ?? '') : children;

  if (!body) return null;

  return (
    <p
      data-slot="form-message"
      id={formMessageId}
      className={cn('text-xs font-label text-destructive', className)}
      {...props}
    >
      {body}
    </p>
  );
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
  useFormField,
};
