// @docker/trident/src/components/ui/input-otp.tsx
'use client';

import {cn} from '@/lib/utils';

import {OTPInput, OTPInputContext} from 'input-otp';
import {Minus} from 'lucide-react';
import * as React from 'react';

/**
 * A one-time password input component. Accessible, unstyled, and customizable.
 *
 * ### Component Props
 *
 * See [input-otp](https://input-otp.rodz.dev) documentation for full details.
 *
 * | Prop                  | Type      | Default |
 * | :-------------------- | :-------- | :------ |
 * | _maxLength_ *(required)* | `number`  | -       |
 * | _containerClassName_  | `string`  | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <InputOTP maxLength={6}>
 *   <InputOTPGroup>
 *     <InputOTPSlot index={0} />
 *     <InputOTPSlot index={1} />
 *     <InputOTPSlot index={2} />
 *   </InputOTPGroup>
 *   <InputOTPSeparator />
 *   <InputOTPGroup>
 *     <InputOTPSlot index={3} />
 *     <InputOTPSlot index={4} />
 *     <InputOTPSlot index={5} />
 *   </InputOTPGroup>
 * </InputOTP>
 * ```
 */
function InputOTP({
  className,
  containerClassName,
  ...props
}: React.ComponentProps<typeof OTPInput> & {
  containerClassName?: string;
}) {
  return (
    <OTPInput
      data-slot="input-otp"
      containerClassName={cn(
        'cn-input-otp flex items-center has-disabled:opacity-50',
        containerClassName
      )}
      spellCheck={false}
      className={cn('disabled:cursor-not-allowed', className)}
      {...props}
    />
  );
}

/**
 * A container that groups multiple `InputOTPSlot` components together visually.
 *
 * ### Example Usage
 *
 * ```tsx
 * <InputOTPGroup>
 *   <InputOTPSlot index={0} />
 *   <InputOTPSlot index={1} />
 *   <InputOTPSlot index={2} />
 * </InputOTPGroup>
 * ```
 */
function InputOTPGroup({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="input-otp-group"
      className={cn(
        'flex items-center rounded has-aria-invalid:border-destructive has-aria-invalid:ring-3 has-aria-invalid:ring-destructive',
        className
      )}
      {...props}
    />
  );
}

/**
 * A single character slot within the OTP input. Displays the entered character and a caret when active.
 *
 * ### Component Props
 *
 * | Prop                 | Type     | Default |
 * | :------------------- | :------- | :------ |
 * | _index_ *(required)* | `number` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <InputOTPSlot index={0} />
 * ```
 */
function InputOTPSlot({
  index,
  className,
  ...props
}: React.ComponentProps<'div'> & {
  index: number;
}) {
  const inputOTPContext = React.useContext(OTPInputContext);
  const {char, hasFakeCaret, isActive} = inputOTPContext?.slots[index] ?? {};

  return (
    <div
      data-slot="input-otp-slot"
      data-active={isActive}
      className={cn(
        'relative flex size-9 items-center justify-center border-y border-r border-input text-sm shadow-raised transition-all ease-default outline-none first:rounded-l first:border-l last:rounded-r aria-invalid:border-destructive data-[active=true]:z-raised data-[active=true]:border-ring data-[active=true]:ring-3 data-[active=true]:ring-ring data-[active=true]:aria-invalid:border-destructive data-[active=true]:aria-invalid:ring-destructive dark:bg-input-bg',
        className
      )}
      {...props}
    >
      {char}
      {hasFakeCaret && (
        <div className="pointer-events-none absolute inset-0 flex items-center justify-center">
          {/* duration-1000 matches native OS caret blink rate; intentionally not a DDS motion token */}
          <div className="h-4 w-px animate-caret-blink bg-foreground duration-1000" />
        </div>
      )}
    </div>
  );
}

/**
 * A visual separator (dash) placed between `InputOTPGroup` components.
 *
 * ### Example Usage
 *
 * ```tsx
 * <InputOTPGroup>...</InputOTPGroup>
 * <InputOTPSeparator />
 * <InputOTPGroup>...</InputOTPGroup>
 * ```
 */
function InputOTPSeparator({
  ...props
}: React.ComponentProps<'div'> & {children?: React.ReactNode}) {
  return (
    <div
      data-slot="input-otp-separator"
      className="flex items-center [&_svg:not([class*='size-'])]:size-4"
      aria-hidden="true"
      {...props}
    >
      <Minus />
    </div>
  );
}

export {InputOTP, InputOTPGroup, InputOTPSlot, InputOTPSeparator};
