// @docker/trident/src/components/ui/input.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';

const inputVariants = cva(
  'w-full min-w-0 rounded border border-input bg-input-bg text-base text-foreground transition-[color,box-shadow] ease-default outline-none file:inline-flex file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:border-destructive aria-invalid:ring-3 aria-invalid:ring-destructive-subtle md:text-sm high-contrast:placeholder:text-foreground',
  {
    variants: {
      size: {
        sm: 'h-8 px-2 py-0.5 text-sm file:h-6',
        default: 'h-9 px-2.5 py-1 file:h-7',
        lg: 'h-10 px-3 py-1.5 file:h-8',
      },
    },
    defaultVariants: {
      size: 'default',
    },
  }
);

type InputProps = Omit<React.ComponentProps<'input'>, 'size'> & {
  /**
   * The size of the input.
   *
   * @default "default"
   */
  size?: VariantProps<typeof inputVariants>['size'];
};

/**
 * A styled text input element for forms.
 *
 * ### Example Usage
 *
 * ```tsx
 * <Input type="email" placeholder="you@example.com" />
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.4.3 Contrast (Minimum) — input text and placeholder meet contrast requirements via DDS tokens.
 * @wcag 2.1.1 Keyboard — natively focusable; value editable via keyboard.
 * @wcag 2.4.7 Focus Visible — focus ring applied via `focus-visible:ring-3`.
 * @wcag 4.1.2 Name, Role, Value — use with `FieldLabel` (htmlFor) or `aria-label` for accessible name.
 */
function Input({className, type, size, ref, ...props}: InputProps) {
  return (
    <input
      ref={ref}
      type={type}
      data-slot="input"
      className={cn(inputVariants({size}), className)}
      {...props}
    />
  );
}

export {Input, inputVariants};
export type {InputProps};
