// @docker/trident/src/components/ui/item.tsx
import type {AsChild} from '@/lib/types';
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {Slot} from 'radix-ui';

/**
 * A container that groups multiple `Item` components into a vertical list layout.
 *
 * Renders as `<ul role="list">` so grouped items are announced as a list.
 *
 * ### Component Props
 *
 * Accepts all standard `<ul>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <ItemGroup>
 *   <Item>
 *     <ItemContent>
 *       <ItemTitle>First item</ItemTitle>
 *     </ItemContent>
 *   </Item>
 *   <ItemSeparator />
 *   <Item>
 *     <ItemContent>
 *       <ItemTitle>Second item</ItemTitle>
 *     </ItemContent>
 *   </Item>
 * </ItemGroup>
 * ```
 */
function ItemGroup({className, ...props}: React.ComponentProps<'ul'>) {
  return (
    <ul
      data-slot="item-group"
      role="list"
      className={cn(
        'group/item-group flex w-full list-none flex-col gap-4 has-data-[size=sm]:gap-2.5 has-data-[size=xs]:gap-2',
        className
      )}
      {...props}
    />
  );
}

/**
 * A horizontal separator for visually dividing items within an `ItemGroup`.
 *
 * Renders as an `<li aria-hidden="true">` so it is a valid child of
 * `ItemGroup`'s `<ul>` (satisfying HTML's content model and WCAG 1.3.1)
 * while being hidden from assistive technologies as a purely decorative
 * divider.
 *
 * ### Component Props
 *
 * Accepts all standard `<li>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <ItemGroup>
 *   <Item>...</Item>
 *   <ItemSeparator />
 *   <Item>...</Item>
 * </ItemGroup>
 * ```
 */
function ItemSeparator({className, ...props}: React.ComponentProps<'li'>) {
  return (
    <li
      data-slot="item-separator"
      aria-hidden="true"
      className={cn('my-2 h-px w-full shrink-0 list-none bg-border', className)}
      {...props}
    />
  );
}

const itemVariants = cva(
  'group/item flex w-full flex-wrap items-center rounded border text-sm transition-colors duration-fast ease-default outline-none focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring [a]:transition-colors [a]:hover:bg-muted',
  {
    variants: {
      variant: {
        default: 'border-transparent',
        outline: 'border-border',
        muted: 'border-transparent bg-selected-overlay',
      },
      size: {
        default: 'gap-3.5 px-4 py-3.5',
        sm: 'gap-2.5 px-3 py-2.5',
        xs: 'gap-2 px-2.5 py-2 [[data-slot=dropdown-menu-content]_&]:p-0',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

/**
 * A versatile list item component with variant and size options for displaying structured content.
 *
 * Renders as `<li>` by default so it composes cleanly inside `ItemGroup`'s
 * `<ul>` container without violating list semantics. When used standalone
 * (outside an `ItemGroup`), use `asChild` to render as a different element.
 *
 * ### Component Props
 *
 * Accepts all standard `<li>` HTML element props, plus:
 *
 * | Prop      | Type                                | Default     |
 * | :-------- | :---------------------------------- | :---------- |
 * | _variant_ | `"default" \| "outline" \| "muted"` | `"default"` |
 * | _size_    | `"default" \| "sm" \| "xs"`         | `"default"` |
 * | _asChild_ | `boolean`                           | `false`     |
 *
 *
 * ### Example Usage
 *
 * **Default**
 * ```tsx
 * <Item>
 *   <ItemContent>
 *     <ItemTitle>Default item</ItemTitle>
 *     <ItemDescription>No border, standard padding</ItemDescription>
 *   </ItemContent>
 * </Item>
 * ```
 *
 * **Outline + Small**
 * ```tsx
 * <Item variant="outline" size="sm">
 *   <ItemMedia variant="icon">
 *     <FileIcon />
 *   </ItemMedia>
 *   <ItemContent>
 *     <ItemTitle>Document</ItemTitle>
 *     <ItemDescription>A brief description</ItemDescription>
 *   </ItemContent>
 * </Item>
 * ```
 *
 * **Muted + Extra Small**
 * ```tsx
 * <Item variant="muted" size="xs">
 *   <ItemContent>
 *     <ItemTitle>Compact muted item</ItemTitle>
 *   </ItemContent>
 * </Item>
 * ```
 */
function Item({
  className,
  variant = 'default',
  size = 'default',
  asChild = false,
  ...props
}: React.ComponentProps<'li'> & VariantProps<typeof itemVariants> & AsChild) {
  const Comp = asChild ? Slot.Root : 'li';
  return (
    <Comp
      data-slot="item"
      data-variant={variant}
      data-size={size}
      className={cn(itemVariants({variant, size, className}))}
      {...props}
    />
  );
}

const itemMediaVariants = cva(
  'flex shrink-0 items-center justify-center gap-2 group-has-[[data-slot=item-description]]/item:translate-y-0.5 group-has-[[data-slot=item-description]]/item:self-start [&_svg]:pointer-events-none',
  {
    variants: {
      variant: {
        default: 'bg-transparent',
        icon: "[&_svg:not([class*='size-'])]:size-4",
        image:
          'size-10 overflow-hidden rounded group-data-[size=sm]/item:size-8 group-data-[size=xs]/item:size-6 [&_img]:size-full [&_img]:object-cover',
      },
    },
    defaultVariants: {
      variant: 'default',
    },
  }
);

/**
 * A media container within an `Item` for displaying icons or images.
 *
 * ### Component Props
 *
 * Accepts all standard `<div>` HTML element props, plus:
 *
 * | Prop      | Type                                | Default     |
 * | :-------- | :---------------------------------- | :---------- |
 * | _variant_ | `"default" \| "icon" \| "image"`    | `"default"` |
 *
 *
 * ### Example Usage
 *
 * **Icon**
 * ```tsx
 * <ItemMedia variant="icon">
 *   <FileIcon />
 * </ItemMedia>
 * ```
 *
 * **Image**
 * ```tsx
 * <ItemMedia variant="image">
 *   <img src="/avatar.jpg" alt="User" />
 * </ItemMedia>
 * ```
 */
function ItemMedia({
  className,
  variant = 'default',
  ...props
}: React.ComponentProps<'div'> & VariantProps<typeof itemMediaVariants>) {
  return (
    <div
      data-slot="item-media"
      data-variant={variant}
      className={cn(itemMediaVariants({variant, className}))}
      {...props}
    />
  );
}

/**
 * A content container within an `Item` for holding title, description, and other text content.
 *
 * ### Component Props
 *
 * Accepts all standard `<div>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <ItemContent>
 *   <ItemTitle>Title</ItemTitle>
 *   <ItemDescription>Description text</ItemDescription>
 * </ItemContent>
 * ```
 */
function ItemContent({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="item-content"
      className={cn(
        'flex flex-1 flex-col gap-1 group-data-[size=xs]/item:gap-0 [&+[data-slot=item-content]]:flex-none',
        className
      )}
      {...props}
    />
  );
}

/**
 * Renders the title text within an `ItemContent` container.
 *
 * ### Component Props
 *
 * Accepts all standard `<div>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <ItemTitle>Document name</ItemTitle>
 * ```
 */
function ItemTitle({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="item-title"
      className={cn(
        'line-clamp-1 flex w-fit items-center gap-2 text-sm leading-snug font-label underline-offset-4',
        className
      )}
      {...props}
    />
  );
}

/**
 * Renders secondary description text within an `ItemContent` container.
 *
 * ### Component Props
 *
 * Accepts all standard `<p>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <ItemDescription>A brief description of the item</ItemDescription>
 * ```
 */
function ItemDescription({className, ...props}: React.ComponentProps<'p'>) {
  return (
    <p
      data-slot="item-description"
      className={cn(
        'line-clamp-2 text-left text-sm leading-normal font-normal text-muted-foreground group-data-[size=xs]/item:text-xs [&>a]:underline [&>a]:underline-offset-4 [&>a:hover]:text-primary',
        className
      )}
      {...props}
    />
  );
}

/**
 * A container for action buttons or controls positioned at the end of an `Item`.
 *
 * ### Component Props
 *
 * Accepts all standard `<div>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <ItemActions>
 *   <Button size="sm">Edit</Button>
 *   <Button size="sm" variant="ghost">Delete</Button>
 * </ItemActions>
 * ```
 */
function ItemActions({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="item-actions"
      className={cn('flex items-center gap-2', className)}
      {...props}
    />
  );
}

/**
 * A full-width header row within an `Item` that spans the entire width and aligns content horizontally.
 *
 * ### Component Props
 *
 * Accepts all standard `<div>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Item>
 *   <ItemHeader>
 *     <ItemTitle>Title</ItemTitle>
 *     <Badge>New</Badge>
 *   </ItemHeader>
 * </Item>
 * ```
 */
function ItemHeader({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="item-header"
      className={cn(
        'flex basis-full items-center justify-between gap-2',
        className
      )}
      {...props}
    />
  );
}

/**
 * A full-width footer row within an `Item` that spans the entire width and aligns content horizontally.
 *
 * ### Component Props
 *
 * Accepts all standard `<div>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Item>
 *   <ItemContent>
 *     <ItemTitle>Title</ItemTitle>
 *   </ItemContent>
 *   <ItemFooter>
 *     <span>Updated 2 hours ago</span>
 *   </ItemFooter>
 * </Item>
 * ```
 */
function ItemFooter({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="item-footer"
      className={cn(
        'flex basis-full items-center justify-between gap-2',
        className
      )}
      {...props}
    />
  );
}

export {
  Item,
  ItemMedia,
  ItemContent,
  ItemActions,
  ItemGroup,
  ItemSeparator,
  ItemTitle,
  ItemDescription,
  ItemHeader,
  ItemFooter,
};
