// @docker/trident/src/components/ui/list.tsx
import type {AsChild} from '@/lib/types';

import {cn} from '@/lib/utils';

import {Slot} from 'radix-ui';
import * as React from 'react';

type ListProps = React.ComponentProps<'ul'> & {
  /** Renders as an ordered list (`<ol>`) when true */
  ordered?: boolean;
};

function List({ordered = false, className, children, ...props}: ListProps) {
  const Tag = (ordered ? 'ol' : 'ul') as React.ElementType;
  return (
    <Tag
      data-slot="list"
      className={cn(
        'space-y-1 text-sm text-foreground',
        ordered ? 'list-decimal pl-5' : 'list-disc pl-5',
        className
      )}
      {...props}
    >
      {children}
    </Tag>
  );
}

type ListItemProps = React.ComponentProps<'li'>;

function ListItem({className, ...props}: ListItemProps) {
  return (
    <li
      data-slot="list-item"
      className={cn('leading-relaxed', className)}
      {...props}
    />
  );
}

type ListItemButtonProps = React.ComponentProps<'button'> &
  AsChild & {
    /** Marks the item as selected — sets aria-pressed and data-selected */
    selected?: boolean;
  };

/**
 * A clickable list row that replaces MUI's `ListItemButton`.
 *
 * Renders as a `<li>` containing a `<button>` by default. Use `asChild` to
 * compose with a router `<Link>` or `<a>` element.
 *
 * ### Component Props
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _selected_  | `boolean` | `false` |
 * | _disabled_  | `boolean` | `false` |
 * | _asChild_   | `boolean` | `false` |
 *
 * ### Example Usage
 *
 * **Basic**
 * ```tsx
 * <List>
 *   <ListItemButton selected>Images</ListItemButton>
 *   <ListItemButton>Containers</ListItemButton>
 * </List>
 * ```
 *
 * **With router link (asChild)**
 * ```tsx
 * <List>
 *   <ListItemButton asChild>
 *     <a href="/images">Images</a>
 *   </ListItemButton>
 * </List>
 * ```
 */
function ListItemButton({
  className,
  selected = false,
  disabled = false,
  asChild = false,
  ...props
}: ListItemButtonProps) {
  const Comp = asChild ? Slot.Root : 'button';

  return (
    <li data-slot="list-item-button" className="list-none">
      <Comp
        data-slot="list-item-button-inner"
        data-selected={selected ? 'true' : undefined}
        {...(!asChild && {'aria-pressed': selected})}
        disabled={disabled}
        className={cn(
          'flex w-full cursor-pointer items-center gap-2 rounded px-3 py-2 text-sm text-foreground transition-colors',
          'hover:bg-muted hover:text-foreground',
          'focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none',
          'disabled:pointer-events-none disabled:opacity-50',
          'data-[selected=true]:bg-accent data-[selected=true]:text-accent-foreground',
          className
        )}
        {...props}
      />
    </li>
  );
}

export {List, ListItem, ListItemButton};
export type {ListProps, ListItemProps, ListItemButtonProps};
