// @docker/trident/src/components/ui/menubar.tsx
'use client';

import type * as React from 'react';

import {cn} from '@/lib/utils';

import {Check, ChevronRight, CircleIcon} from 'lucide-react';
import {Menubar as MenubarPrimitive} from 'radix-ui';

/**
 * A visually persistent menu common in desktop applications that provides quick access to a consistent set of commands.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop            | Type                         | Default |
 * | :-------------- | :--------------------------- | :------ |
 * | _asChild_       | `boolean`                    | `false` |
 * | _defaultValue_  | `string`                     | -       |
 * | _value_         | `string`                     | -       |
 * | _onValueChange_ | `(value: string) => void`    | -       |
 * | _dir_           | `"ltr" \| "rtl"`             | -       |
 * | _loop_          | `boolean`                    | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Menubar>
 *   <MenubarMenu>
 *     <MenubarTrigger>File</MenubarTrigger>
 *     <MenubarContent>
 *       <MenubarItem>New Tab</MenubarItem>
 *       <MenubarItem>New Window</MenubarItem>
 *       <MenubarSeparator />
 *       <MenubarItem>Print</MenubarItem>
 *     </MenubarContent>
 *   </MenubarMenu>
 * </Menubar>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 2.1.1 Keyboard — arrow keys navigate between menus and items; Enter/Space activates; Escape closes.
 * @wcag 2.4.7 Focus Visible — focused items render an inset focus ring (`focus:ring-2 focus:ring-ring focus:ring-inset`) in addition to `bg-accent`, guaranteeing a visible 3:1 indicator against `bg-popover` in all themes (dark `--accent` and `--popover` otherwise collide at ~1:1).
 * @wcag 4.1.2 Name, Role, Value — `menubar`, `menu`, and `menuitem` roles via Radix Menubar primitives.
 */
function Menubar({
  className,
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Root>) {
  return (
    <MenubarPrimitive.Root
      data-slot="menubar"
      className={cn(
        'flex h-9 items-center gap-1 rounded border bg-background p-1 shadow-raised',
        className
      )}
      {...props}
    />
  );
}

/**
 * A top-level menu within the menubar that contains a trigger and content.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop    | Type     | Default |
 * | :------ | :------- | :------ |
 * | _value_ | `string` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarMenu>
 *   <MenubarTrigger>File</MenubarTrigger>
 *   <MenubarContent>
 *     <MenubarItem>New Tab</MenubarItem>
 *   </MenubarContent>
 * </MenubarMenu>
 * ```
 */
function MenubarMenu({
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Menu>) {
  return <MenubarPrimitive.Menu data-slot="menubar-menu" {...props} />;
}

/**
 * Groups multiple menubar items together.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarGroup>
 *   <MenubarItem>Cut</MenubarItem>
 *   <MenubarItem>Copy</MenubarItem>
 *   <MenubarItem>Paste</MenubarItem>
 * </MenubarGroup>
 * ```
 */
function MenubarGroup({
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Group>) {
  return <MenubarPrimitive.Group data-slot="menubar-group" {...props} />;
}

/**
 * Portals the menubar content into the document body.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop         | Type          | Default         |
 * | :----------- | :------------ | :-------------- |
 * | _forceMount_ | `boolean`     | -               |
 * | _container_  | `HTMLElement` | `document.body` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarPortal>
 *   <MenubarContent>...</MenubarContent>
 * </MenubarPortal>
 * ```
 */
function MenubarPortal({
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Portal>) {
  return <MenubarPrimitive.Portal data-slot="menubar-portal" {...props} />;
}

/**
 * Groups multiple radio items together, ensuring only one can be selected at a time.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop            | Type                      | Default |
 * | :-------------- | :------------------------ | :------ |
 * | _asChild_       | `boolean`                 | `false` |
 * | _value_         | `string`                  | -       |
 * | _onValueChange_ | `(value: string) => void` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarRadioGroup value="andy">
 *   <MenubarRadioItem value="andy">Andy</MenubarRadioItem>
 *   <MenubarRadioItem value="benoit">Benoit</MenubarRadioItem>
 * </MenubarRadioGroup>
 * ```
 */
function MenubarRadioGroup({
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.RadioGroup>) {
  return (
    <MenubarPrimitive.RadioGroup data-slot="menubar-radio-group" {...props} />
  );
}

/**
 * The button that toggles a menubar menu open or closed.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarTrigger>File</MenubarTrigger>
 * ```
 */
function MenubarTrigger({
  className,
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Trigger>) {
  return (
    <MenubarPrimitive.Trigger
      data-slot="menubar-trigger"
      className={cn(
        'flex items-center rounded px-2 py-1 text-sm font-label outline-hidden select-none hover:bg-muted focus:bg-accent focus:text-accent-foreground focus-visible:ring-3 focus-visible:ring-ring aria-expanded:bg-muted',
        className
      )}
      {...props}
    />
  );
}

/**
 * The dropdown content displayed when a menubar menu is open.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop                   | Type                                     | Default     |
 * | :--------------------- | :--------------------------------------- | :---------- |
 * | _asChild_              | `boolean`                                | `false`     |
 * | _loop_                 | `boolean`                                | `false`     |
 * | _side_                 | `"top" \| "right" \| "bottom" \| "left"` | `"bottom"`  |
 * | _sideOffset_           | `number`                                 | `0`         |
 * | _align_                | `"start" \| "center" \| "end"`           | `"center"`  |
 * | _alignOffset_          | `number`                                 | `0`         |
 * | _avoidCollisions_      | `boolean`                                | `true`      |
 * | _collisionPadding_     | `number \| Padding`                      | `0`         |
 * | _sticky_               | `"partial" \| "always"`                  | `"partial"` |
 * | _hideWhenDetached_     | `boolean`                                | `false`     |
 * | _onCloseAutoFocus_     | `(event: Event) => void`                 | -           |
 * | _onEscapeKeyDown_      | `(event: KeyboardEvent) => void`         | -           |
 * | _onPointerDownOutside_ | `(event: PointerDownOutsideEvent) => void` | -         |
 * | _onInteractOutside_    | `(event: Event) => void`                 | -           |
 * | _forceMount_           | `boolean`                                | -           |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarContent>
 *   <MenubarItem>New Tab</MenubarItem>
 *   <MenubarSeparator />
 *   <MenubarItem>Print</MenubarItem>
 * </MenubarContent>
 * ```
 */
function MenubarContent({
  className,
  align = 'start',
  alignOffset = -4,
  sideOffset = 8,
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Content>) {
  return (
    <MenubarPortal>
      <MenubarPrimitive.Content
        data-slot="menubar-content"
        align={align}
        alignOffset={alignOffset}
        sideOffset={sideOffset}
        className={cn(
          'z-floating min-w-36 origin-(--radix-menubar-content-transform-origin) overflow-hidden rounded bg-popover p-1 text-popover-foreground shadow-floating ring-1 ring-border duration-fast ease-out data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 data-open:animate-in data-open:fade-in-0 data-open:zoom-in-95',
          className
        )}
        {...props}
      />
    </MenubarPortal>
  );
}

/**
 * An interactive item within a menubar menu that can be selected.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop        | Type                            | Default     |
 * | :---------- | :------------------------------ | :---------- |
 * | _asChild_   | `boolean`                       | `false`     |
 * | _disabled_  | `boolean`                       | -           |
 * | _onSelect_  | `(event: Event) => void`        | -           |
 * | _textValue_ | `string`                        | -           |
 * | _inset_     | `boolean`                       | -           |
 * | _variant_   | `"default" \| "destructive"`    | `"default"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarItem>
 *   New Tab <MenubarShortcut>⌘T</MenubarShortcut>
 * </MenubarItem>
 * ```
 */
function MenubarItem({
  className,
  inset,
  variant = 'default',
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Item> & {
  inset?: boolean;
  variant?: 'default' | 'destructive';
}) {
  return (
    <MenubarPrimitive.Item
      data-slot="menubar-item"
      data-inset={inset}
      data-variant={variant}
      className={cn(
        "group/menubar-item relative flex cursor-default items-center gap-2 rounded px-2 py-1.5 text-sm outline-hidden select-none focus:bg-accent focus:text-accent-foreground focus:ring-2 focus:ring-ring focus:ring-inset not-data-[variant=destructive]:focus:**:text-accent-foreground data-inset:pl-8 data-[variant=destructive]:text-destructive data-[variant=destructive]:focus:bg-destructive-subtle data-[variant=destructive]:focus:text-destructive data-disabled:pointer-events-none data-disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 [&_svg:not([class*='text-'])]:text-muted-foreground data-[variant=destructive]:*:[svg]:text-destructive!",
        className
      )}
      {...props}
    />
  );
}

/**
 * A menubar item that can be toggled on or off, displaying a checkmark when checked.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop              | Type                                         | Default |
 * | :---------------- | :------------------------------------------- | :------ |
 * | _asChild_         | `boolean`                                    | `false` |
 * | _checked_         | `boolean \| "indeterminate"`                 | -       |
 * | _onCheckedChange_ | `(checked: boolean) => void`                 | -       |
 * | _disabled_        | `boolean`                                    | -       |
 * | _onSelect_        | `(event: Event) => void`                     | -       |
 * | _textValue_       | `string`                                     | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarCheckboxItem checked={showBookmarks} onCheckedChange={setShowBookmarks}>
 *   Show Bookmarks
 * </MenubarCheckboxItem>
 * ```
 */
function MenubarCheckboxItem({
  className,
  children,
  checked,
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.CheckboxItem>) {
  return (
    <MenubarPrimitive.CheckboxItem
      data-slot="menubar-checkbox-item"
      className={cn(
        'relative flex cursor-default items-center gap-2 rounded py-1.5 pr-2 pl-8 text-sm outline-hidden select-none focus:bg-accent focus:text-accent-foreground focus:ring-2 focus:ring-ring focus:ring-inset focus:**:text-accent-foreground data-disabled:pointer-events-none data-disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0',
        className
      )}
      checked={checked}
      {...props}
    >
      <span className="pointer-events-none absolute left-2 flex size-4 items-center justify-center [&_svg:not([class*='size-'])]:size-4">
        <MenubarPrimitive.ItemIndicator>
          <Check />
        </MenubarPrimitive.ItemIndicator>
      </span>
      {children}
    </MenubarPrimitive.CheckboxItem>
  );
}

/**
 * A radio-style menubar item used within a `MenubarRadioGroup`.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop                  | Type                     | Default |
 * | :-------------------- | :----------------------- | :------ |
 * | _asChild_             | `boolean`                | `false` |
 * | _value_ *(required)*  | `string`                 | -       |
 * | _disabled_            | `boolean`                | -       |
 * | _onSelect_            | `(event: Event) => void` | -       |
 * | _textValue_           | `string`                 | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarRadioGroup value="andy">
 *   <MenubarRadioItem value="andy">Andy</MenubarRadioItem>
 *   <MenubarRadioItem value="benoit">Benoit</MenubarRadioItem>
 * </MenubarRadioGroup>
 * ```
 */
function MenubarRadioItem({
  className,
  children,
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.RadioItem>) {
  return (
    <MenubarPrimitive.RadioItem
      data-slot="menubar-radio-item"
      className={cn(
        "relative flex cursor-default items-center gap-2 rounded py-1.5 pr-2 pl-8 text-sm outline-hidden select-none focus:bg-accent focus:text-accent-foreground focus:ring-2 focus:ring-ring focus:ring-inset focus:**:text-accent-foreground data-disabled:pointer-events-none data-disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        className
      )}
      {...props}
    >
      <span className="pointer-events-none absolute left-2 flex size-4 items-center justify-center [&_svg:not([class*='size-'])]:size-4">
        <MenubarPrimitive.ItemIndicator>
          <CircleIcon className="size-2 fill-current" />
        </MenubarPrimitive.ItemIndicator>
      </span>
      {children}
    </MenubarPrimitive.RadioItem>
  );
}

/**
 * A non-interactive label within a menubar menu used to title a section of items.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 * | _inset_   | `boolean` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarLabel>People</MenubarLabel>
 * ```
 */
function MenubarLabel({
  className,
  inset,
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Label> & {
  inset?: boolean;
}) {
  return (
    <MenubarPrimitive.Label
      data-slot="menubar-label"
      data-inset={inset}
      className={cn(
        'px-2 py-1.5 text-sm font-label data-inset:pl-8',
        className
      )}
      {...props}
    />
  );
}

/**
 * A visual separator between groups of menubar items.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarItem>Cut</MenubarItem>
 * <MenubarSeparator />
 * <MenubarItem>Paste</MenubarItem>
 * ```
 */
function MenubarSeparator({
  className,
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Separator>) {
  return (
    <MenubarPrimitive.Separator
      data-slot="menubar-separator"
      className={cn('-mx-1 my-1 h-px bg-border', className)}
      {...props}
    />
  );
}

/**
 * Displays a keyboard shortcut hint aligned to the right of a menubar item.
 *
 * ### Component Props
 *
 * Accepts all standard `<span>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarItem>
 *   New Tab <MenubarShortcut>⌘T</MenubarShortcut>
 * </MenubarItem>
 * ```
 */
function MenubarShortcut({className, ...props}: React.ComponentProps<'span'>) {
  return (
    <span
      data-slot="menubar-shortcut"
      className={cn(
        'ml-auto text-xs tracking-widest text-muted-foreground group-focus/menubar-item:text-accent-foreground',
        className
      )}
      {...props}
    />
  );
}

/**
 * A submenu container within a menubar menu that holds a sub-trigger and sub-content.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop            | Type                         | Default |
 * | :-------------- | :--------------------------- | :------ |
 * | _defaultOpen_   | `boolean`                    | -       |
 * | _open_          | `boolean`                    | -       |
 * | _onOpenChange_  | `(open: boolean) => void`    | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarSub>
 *   <MenubarSubTrigger>Share</MenubarSubTrigger>
 *   <MenubarSubContent>
 *     <MenubarItem>Email</MenubarItem>
 *     <MenubarItem>Messages</MenubarItem>
 *   </MenubarSubContent>
 * </MenubarSub>
 * ```
 */
function MenubarSub({
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.Sub>) {
  return <MenubarPrimitive.Sub data-slot="menubar-sub" {...props} />;
}

/**
 * The item that opens a submenu within a menubar, displaying a right-facing chevron.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop        | Type      | Default |
 * | :---------- | :-------- | :------ |
 * | _asChild_   | `boolean` | `false` |
 * | _disabled_  | `boolean` | -       |
 * | _textValue_ | `string`  | -       |
 * | _inset_     | `boolean` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarSubTrigger>Share</MenubarSubTrigger>
 * ```
 */
function MenubarSubTrigger({
  className,
  inset,
  children,
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.SubTrigger> & {
  inset?: boolean;
}) {
  return (
    <MenubarPrimitive.SubTrigger
      data-slot="menubar-sub-trigger"
      data-inset={inset}
      className={cn(
        "flex cursor-default items-center gap-2 rounded px-2 py-1.5 text-sm outline-hidden select-none focus:bg-accent focus:text-accent-foreground focus:ring-2 focus:ring-ring focus:ring-inset data-inset:pl-8 data-open:bg-accent data-open:text-accent-foreground [&_svg:not([class*='size-'])]:size-4 [&_svg:not([class*='text-'])]:text-muted-foreground",
        className
      )}
      {...props}
    >
      {children}
      <ChevronRight className="ml-auto size-4" />
    </MenubarPrimitive.SubTrigger>
  );
}

/**
 * The content displayed when a menubar submenu is open.
 *
 * ### Component Props
 *
 * See [Menubar props](https://www.radix-ui.com/primitives/docs/components/menubar) documentation for full details.
 *
 * | Prop                   | Type                                       | Default     |
 * | :--------------------- | :----------------------------------------- | :---------- |
 * | _asChild_              | `boolean`                                  | `false`     |
 * | _loop_                 | `boolean`                                  | `false`     |
 * | _sideOffset_           | `number`                                   | `0`         |
 * | _alignOffset_          | `number`                                   | `0`         |
 * | _avoidCollisions_      | `boolean`                                  | `true`      |
 * | _collisionPadding_     | `number \| Padding`                        | `0`         |
 * | _sticky_               | `"partial" \| "always"`                    | `"partial"` |
 * | _hideWhenDetached_     | `boolean`                                  | `false`     |
 * | _onEscapeKeyDown_      | `(event: KeyboardEvent) => void`           | -           |
 * | _onPointerDownOutside_ | `(event: PointerDownOutsideEvent) => void` | -           |
 * | _onInteractOutside_    | `(event: Event) => void`                   | -           |
 * | _forceMount_           | `boolean`                                  | -           |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <MenubarSubContent>
 *   <MenubarItem>Email</MenubarItem>
 *   <MenubarItem>Messages</MenubarItem>
 * </MenubarSubContent>
 * ```
 */
function MenubarSubContent({
  className,
  ...props
}: React.ComponentProps<typeof MenubarPrimitive.SubContent>) {
  return (
    <MenubarPrimitive.SubContent
      data-slot="menubar-sub-content"
      className={cn(
        'z-floating min-w-32 origin-(--radix-menubar-content-transform-origin) overflow-hidden rounded bg-popover p-1 text-popover-foreground shadow-floating ring-1 ring-border duration-fast ease-out data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 data-open:animate-in data-open:fade-in-0 data-open:zoom-in-95 data-closed:animate-out data-closed:fade-out-0 data-closed:zoom-out-95',
        className
      )}
      {...props}
    />
  );
}

export {
  Menubar,
  MenubarPortal,
  MenubarMenu,
  MenubarTrigger,
  MenubarContent,
  MenubarGroup,
  MenubarSeparator,
  MenubarLabel,
  MenubarItem,
  MenubarShortcut,
  MenubarCheckboxItem,
  MenubarRadioGroup,
  MenubarRadioItem,
  MenubarSub,
  MenubarSubTrigger,
  MenubarSubContent,
};
