// @docker/trident/src/components/ui/navigation-menu.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva} from 'class-variance-authority';
import {ChevronDown} from 'lucide-react';
import {NavigationMenu as NavigationMenuPrimitive} from 'radix-ui';

/**
 * A collection of links for navigating websites.
 *
 * ### Component Props
 *
 * See [Navigation Menu props](https://www.radix-ui.com/primitives/docs/components/navigation-menu) documentation for full details.
 *
 * | Prop                | Type                             | Default        |
 * | :------------------ | :------------------------------- | :------------- |
 * | _defaultValue_      | `string`                         | -              |
 * | _value_             | `string`                         | -              |
 * | _onValueChange_     | `(value: string) => void`        | -              |
 * | _delayDuration_     | `number`                         | `200`          |
 * | _skipDelayDuration_ | `number`                         | `300`          |
 * | _dir_               | `"ltr" \| "rtl"`                 | -              |
 * | _orientation_       | `"vertical" \| "horizontal"`     | `"horizontal"` |
 * | _viewport_          | `boolean`                        | `true`         |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <NavigationMenu>
 *   <NavigationMenuList>
 *     <NavigationMenuItem>
 *       <NavigationMenuTrigger>Getting Started</NavigationMenuTrigger>
 *       <NavigationMenuContent>
 *         <NavigationMenuLink href="/docs">Documentation</NavigationMenuLink>
 *       </NavigationMenuContent>
 *     </NavigationMenuItem>
 *   </NavigationMenuList>
 * </NavigationMenu>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 2.1.1 Keyboard — Tab and Arrow keys navigate menu items; Radix UI manages focus between triggers and content.
 * @wcag 2.4.3 Focus Order — focus follows a logical order through triggers and their expanded content panels.
 * @wcag 4.1.2 Name, Role, Value — Radix UI renders role="navigation" with appropriate aria-expanded on triggers.
 */
function NavigationMenu({
  className,
  children,
  viewport = true,
  ...props
}: React.ComponentProps<typeof NavigationMenuPrimitive.Root> & {
  viewport?: boolean;
}) {
  return (
    <div className="@container/nav">
      <NavigationMenuPrimitive.Root
        data-slot="navigation-menu"
        data-viewport={viewport}
        className={cn(
          'group/navigation-menu relative flex max-w-max flex-1 items-center justify-center',
          className
        )}
        {...props}
      >
        {children}
        {viewport && <NavigationMenuViewport />}
      </NavigationMenuPrimitive.Root>
    </div>
  );
}

/**
 * A list container for navigation menu items arranged horizontally.
 *
 * ### Component Props
 *
 * See [Navigation Menu props](https://www.radix-ui.com/primitives/docs/components/navigation-menu) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <NavigationMenuList>
 *   <NavigationMenuItem>...</NavigationMenuItem>
 * </NavigationMenuList>
 * ```
 */
function NavigationMenuList({
  className,
  ...props
}: React.ComponentProps<typeof NavigationMenuPrimitive.List>) {
  return (
    <NavigationMenuPrimitive.List
      data-slot="navigation-menu-list"
      className={cn(
        'group flex flex-1 list-none items-center justify-center gap-0 @[240px]/nav:flex-col @[240px]/nav:items-start @[360px]/nav:flex-wrap',
        className
      )}
      {...props}
    />
  );
}

/**
 * An individual item within the navigation menu list that can contain a trigger and content.
 *
 * ### Component Props
 *
 * See [Navigation Menu props](https://www.radix-ui.com/primitives/docs/components/navigation-menu) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 * | _value_   | `string`  | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <NavigationMenuItem>
 *   <NavigationMenuTrigger>Components</NavigationMenuTrigger>
 *   <NavigationMenuContent>...</NavigationMenuContent>
 * </NavigationMenuItem>
 * ```
 */
function NavigationMenuItem({
  className,
  ...props
}: React.ComponentProps<typeof NavigationMenuPrimitive.Item>) {
  return (
    <NavigationMenuPrimitive.Item
      data-slot="navigation-menu-item"
      className={cn('relative', className)}
      {...props}
    />
  );
}

/**
 * A cva style variant that can be applied to navigation menu links or triggers to match the default trigger styling.
 *
 * ### Example Usage
 *
 * ```tsx
 * <NavigationMenuLink className={navigationMenuTriggerStyle()}>
 *   Home
 * </NavigationMenuLink>
 * ```
 */
const navigationMenuTriggerStyle = cva(
  'group/navigation-menu-trigger inline-flex h-9 w-max items-center justify-center rounded bg-background px-4 py-2 text-sm font-label transition-all ease-default outline-none hover:bg-muted focus:bg-muted focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-popup-open:bg-selected-overlay data-popup-open:hover:bg-muted @[240px]/nav:w-full @[240px]/nav:justify-start @[360px]/nav:px-3 @[360px]/nav:text-xs data-open:bg-selected-overlay data-open:hover:bg-muted data-open:focus:bg-muted'
);

/**
 * The button that toggles a navigation menu item's content, with a chevron indicator.
 *
 * ### Component Props
 *
 * See [Navigation Menu props](https://www.radix-ui.com/primitives/docs/components/navigation-menu) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <NavigationMenuTrigger>Getting Started</NavigationMenuTrigger>
 * ```
 */
function NavigationMenuTrigger({
  className,
  children,
  ...props
}: React.ComponentProps<typeof NavigationMenuPrimitive.Trigger>) {
  return (
    <NavigationMenuPrimitive.Trigger
      data-slot="navigation-menu-trigger"
      className={cn(navigationMenuTriggerStyle(), 'group', className)}
      {...props}
    >
      {children}{' '}
      <ChevronDown
        className="relative top-px ml-1 size-3 transition duration-slow ease-default group-data-popup-open/navigation-menu-trigger:rotate-180 group-data-open/navigation-menu-trigger:rotate-180"
        aria-hidden="true"
      />
    </NavigationMenuPrimitive.Trigger>
  );
}

/**
 * The content displayed when a navigation menu item is active, with enter/exit animations.
 *
 * ### Component Props
 *
 * See [Navigation Menu props](https://www.radix-ui.com/primitives/docs/components/navigation-menu) documentation for full details.
 *
 * | Prop                   | Type                                       | Default |
 * | :--------------------- | :----------------------------------------- | :------ |
 * | _asChild_              | `boolean`                                  | `false` |
 * | _onEscapeKeyDown_      | `(event: KeyboardEvent) => void`           | -       |
 * | _onPointerDownOutside_ | `(event: PointerDownOutsideEvent) => void` | -       |
 * | _onInteractOutside_    | `(event: Event) => void`                   | -       |
 * | _forceMount_           | `boolean`                                  | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <NavigationMenuContent>
 *   <NavigationMenuLink href="/docs">Documentation</NavigationMenuLink>
 * </NavigationMenuContent>
 * ```
 */
function NavigationMenuContent({
  className,
  ...props
}: React.ComponentProps<typeof NavigationMenuPrimitive.Content>) {
  return (
    <NavigationMenuPrimitive.Content
      data-slot="navigation-menu-content"
      className={cn(
        'top-0 left-0 w-full p-2 pr-2.5 ease-[cubic-bezier(0.22,1,0.36,1)] group-data-[viewport=false]/navigation-menu:top-full group-data-[viewport=false]/navigation-menu:mt-1.5 group-data-[viewport=false]/navigation-menu:overflow-hidden group-data-[viewport=false]/navigation-menu:rounded group-data-[viewport=false]/navigation-menu:bg-popover group-data-[viewport=false]/navigation-menu:text-popover-foreground group-data-[viewport=false]/navigation-menu:shadow group-data-[viewport=false]/navigation-menu:ring-1 group-data-[viewport=false]/navigation-menu:ring-border group-data-[viewport=false]/navigation-menu:duration-slow data-[motion=from-end]:slide-in-from-right-52 data-[motion=from-start]:slide-in-from-left-52 data-[motion=to-end]:slide-out-to-right-52 data-[motion=to-start]:slide-out-to-left-52 data-[motion^=from-]:animate-in data-[motion^=from-]:fade-in data-[motion^=to-]:animate-out data-[motion^=to-]:fade-out **:data-[slot=navigation-menu-link]:focus:ring-0 **:data-[slot=navigation-menu-link]:focus:outline-none md:absolute md:w-auto group-data-[viewport=false]/navigation-menu:data-open:animate-in group-data-[viewport=false]/navigation-menu:data-open:fade-in-0 group-data-[viewport=false]/navigation-menu:data-open:zoom-in-95 group-data-[viewport=false]/navigation-menu:data-closed:animate-out group-data-[viewport=false]/navigation-menu:data-closed:fade-out-0 group-data-[viewport=false]/navigation-menu:data-closed:zoom-out-95',
        className
      )}
      {...props}
    />
  );
}

/**
 * A viewport element that renders the active navigation menu content with animated transitions.
 *
 * ### Component Props
 *
 * See [Navigation Menu props](https://www.radix-ui.com/primitives/docs/components/navigation-menu) documentation for full details.
 *
 * | Prop         | Type      | Default |
 * | :----------- | :-------- | :------ |
 * | _asChild_    | `boolean` | `false` |
 * | _forceMount_ | `boolean` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <NavigationMenuViewport />
 * ```
 */
function NavigationMenuViewport({
  className,
  ...props
}: React.ComponentProps<typeof NavigationMenuPrimitive.Viewport>) {
  return (
    <div
      className={cn(
        'absolute top-full left-0 isolate z-floating flex justify-center'
      )}
    >
      <NavigationMenuPrimitive.Viewport
        data-slot="navigation-menu-viewport"
        className={cn(
          'origin-top-center relative mt-1.5 h-(--radix-navigation-menu-viewport-height) w-full overflow-hidden rounded-lg bg-popover text-popover-foreground shadow ring-1 ring-border duration-fast ease-out md:w-(--radix-navigation-menu-viewport-width) data-open:animate-in data-open:zoom-in-90 data-closed:animate-out data-closed:zoom-out-95',
          className
        )}
        {...props}
      />
    </div>
  );
}

/**
 * A navigation link within the navigation menu with hover and focus states.
 *
 * ### Component Props
 *
 * See [Navigation Menu props](https://www.radix-ui.com/primitives/docs/components/navigation-menu) documentation for full details.
 *
 * | Prop       | Type                     | Default |
 * | :--------- | :----------------------- | :------ |
 * | _asChild_  | `boolean`                | `false` |
 * | _active_   | `boolean`                | `false` |
 * | _onSelect_ | `(event: Event) => void` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <NavigationMenuLink href="/docs">Documentation</NavigationMenuLink>
 * ```
 */
function NavigationMenuLink({
  className,
  ...props
}: React.ComponentProps<typeof NavigationMenuPrimitive.Link>) {
  return (
    <NavigationMenuPrimitive.Link
      data-slot="navigation-menu-link"
      className={cn(
        // Hover/focus use `accent-subtle` (accent-tinted overlay) instead of
        // `muted`. Inside a `NavigationMenuContent`/`Viewport`, `bg-muted` lands
        // on top of `bg-popover` and in dark mode the pair renders at 1.07:1
        // (WCAG 1.4.11 fail). `accent-subtle` is a translucent accent wash that
        // remains perceptible against every surface level (ground, elevated,
        // popover) in light, dark, and high-contrast.
        "flex items-center gap-1.5 rounded p-2 text-sm transition-all ease-default outline-none hover:bg-accent-subtle focus:bg-accent-subtle focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-1 data-[active=true]:bg-selected-overlay data-[active=true]:hover:bg-accent-subtle data-[active=true]:focus:bg-accent-subtle [&_svg:not([class*='size-'])]:size-4",
        className
      )}
      {...props}
    />
  );
}

/**
 * An optional indicator element that highlights the active navigation menu item.
 *
 * ### Component Props
 *
 * See [Navigation Menu props](https://www.radix-ui.com/primitives/docs/components/navigation-menu) documentation for full details.
 *
 * | Prop         | Type      | Default |
 * | :----------- | :-------- | :------ |
 * | _asChild_    | `boolean` | `false` |
 * | _forceMount_ | `boolean` | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <NavigationMenuIndicator />
 * ```
 */
function NavigationMenuIndicator({
  className,
  ...props
}: React.ComponentProps<typeof NavigationMenuPrimitive.Indicator>) {
  return (
    <NavigationMenuPrimitive.Indicator
      data-slot="navigation-menu-indicator"
      className={cn(
        'top-full z-1 flex h-1.5 items-end justify-center overflow-hidden data-[state=hidden]:animate-out data-[state=hidden]:fade-out data-[state=visible]:animate-in data-[state=visible]:fade-in',
        className
      )}
      {...props}
    >
      <div className="relative top-[60%] h-2 w-2 rotate-45 rounded-tl bg-border shadow-floating" />
    </NavigationMenuPrimitive.Indicator>
  );
}

export {
  NavigationMenu,
  NavigationMenuList,
  NavigationMenuItem,
  NavigationMenuContent,
  NavigationMenuTrigger,
  NavigationMenuLink,
  NavigationMenuIndicator,
  NavigationMenuViewport,
  navigationMenuTriggerStyle,
};
