// @docker/trident/src/components/ui/pagination.tsx
import type * as React from 'react';

import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import {ChevronLeft, ChevronRight, Ellipsis} from 'lucide-react';

/**
 * A navigation container for pagination controls with built-in accessibility attributes.
 *
 * ### Component Props
 *
 * Accepts all standard `<nav>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Pagination>
 *   <PaginationContent>
 *     <PaginationItem>
 *       <PaginationPrevious href="#" />
 *     </PaginationItem>
 *     <PaginationItem>
 *       <PaginationLink href="#">1</PaginationLink>
 *     </PaginationItem>
 *     <PaginationItem>
 *       <PaginationLink href="#" isActive>2</PaginationLink>
 *     </PaginationItem>
 *     <PaginationItem>
 *       <PaginationEllipsis />
 *     </PaginationItem>
 *     <PaginationItem>
 *       <PaginationNext href="#" />
 *     </PaginationItem>
 *   </PaginationContent>
 * </Pagination>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 2.1.1 Keyboard — page links are native anchors, natively focusable and activatable via keyboard.
 * @wcag 2.4.4 Link Purpose — PaginationPrevious and PaginationNext carry explicit aria-label; page links convey number via text content.
 * @wcag 4.1.2 Name, Role, Value — renders a `<nav>` element with aria-label="pagination" for landmark identification.
 */
function Pagination({className, ...props}: React.ComponentProps<'nav'>) {
  return (
    <nav
      aria-label="pagination"
      data-slot="pagination"
      className={cn('mx-auto flex w-full justify-center', className)}
      {...props}
    />
  );
}

/**
 * A list container for pagination items that arranges them horizontally.
 *
 * ### Component Props
 *
 * Accepts all standard `<ul>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <PaginationContent>
 *   <PaginationItem>
 *     <PaginationLink href="#">1</PaginationLink>
 *   </PaginationItem>
 * </PaginationContent>
 * ```
 */
function PaginationContent({className, ...props}: React.ComponentProps<'ul'>) {
  return (
    <ul
      data-slot="pagination-content"
      className={cn('flex items-center gap-1', className)}
      {...props}
    />
  );
}

/**
 * A list item wrapper for individual pagination elements.
 *
 * ### Component Props
 *
 * Accepts all standard `<li>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <PaginationItem>
 *   <PaginationLink href="#">1</PaginationLink>
 * </PaginationItem>
 * ```
 */
function PaginationItem({...props}: React.ComponentProps<'li'>) {
  return <li data-slot="pagination-item" {...props} />;
}

type PaginationLinkProps = {
  isActive?: boolean;
} & Pick<React.ComponentProps<typeof Button>, 'size'> &
  React.ComponentProps<'a'>;

/**
 * A pagination link rendered as a button-styled anchor element.
 *
 * ### Component Props
 *
 * Accepts all standard `<a>` HTML element props, plus:
 *
 * | Prop       | Type      | Default  |
 * | :--------- | :-------- | :------- |
 * | _isActive_ | `boolean` | -        |
 * | _size_     | `string`  | `"icon"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <PaginationLink href="#" isActive>2</PaginationLink>
 * ```
 */
function PaginationLink({
  className,
  isActive,
  size = 'icon',
  children,
  ...props
}: PaginationLinkProps) {
  return (
    <Button
      asChild
      variant={isActive ? 'outline' : 'ghost'}
      size={size}
      className={cn(className)}
    >
      <a
        aria-current={isActive ? 'page' : undefined}
        data-slot="pagination-link"
        data-active={isActive}
        {...props}
      >
        {children}
      </a>
    </Button>
  );
}

/**
 * A pagination link for navigating to the previous page, with a left chevron icon.
 *
 * ### Component Props
 *
 * Accepts all `PaginationLink` props, plus:
 *
 * | Prop   | Type     | Default      |
 * | :----- | :------- | :----------- |
 * | _text_ | `string` | `"Previous"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <PaginationPrevious href="#" />
 * ```
 */
function PaginationPrevious({
  className,
  text = 'Previous',
  ...props
}: React.ComponentProps<typeof PaginationLink> & {text?: string}) {
  return (
    <PaginationLink
      aria-label="Go to previous page"
      size="default"
      className={className}
      {...props}
    >
      <ChevronLeft data-icon="inline-start" />
      <span className="hidden sm:block">{text}</span>
    </PaginationLink>
  );
}

/**
 * A pagination link for navigating to the next page, with a right chevron icon.
 *
 * ### Component Props
 *
 * Accepts all `PaginationLink` props, plus:
 *
 * | Prop   | Type     | Default  |
 * | :----- | :------- | :------- |
 * | _text_ | `string` | `"Next"` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <PaginationNext href="#" />
 * ```
 */
function PaginationNext({
  className,
  text = 'Next',
  ...props
}: React.ComponentProps<typeof PaginationLink> & {text?: string}) {
  return (
    <PaginationLink
      aria-label="Go to next page"
      size="default"
      className={className}
      {...props}
    >
      <span className="hidden sm:block">{text}</span>
      <ChevronRight data-icon="inline-end" />
    </PaginationLink>
  );
}

/**
 * A visual ellipsis indicator representing hidden page numbers in a pagination sequence.
 *
 * ### Component Props
 *
 * Accepts all standard `<span>` HTML element props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <PaginationEllipsis />
 * ```
 */
function PaginationEllipsis({
  className,
  ...props
}: React.ComponentProps<'span'>) {
  return (
    <span
      aria-hidden
      data-slot="pagination-ellipsis"
      className={cn(
        "flex size-9 items-center justify-center [&_svg:not([class*='size-'])]:size-4",
        className
      )}
      {...props}
    >
      <Ellipsis />
      <span className="sr-only">More pages</span>
    </span>
  );
}

export {
  Pagination,
  PaginationContent,
  PaginationEllipsis,
  PaginationItem,
  PaginationLink,
  PaginationNext,
  PaginationPrevious,
};
