// @docker/trident/src/components/ui/sheet.tsx
import type * as React from 'react';

import {Button} from '@/components/ui/button';
import {cn} from '@/lib/utils';

import {X} from 'lucide-react';
import {Dialog as SheetPrimitive} from 'radix-ui';

/**
 * A panel that slides in from the edge of the screen, built on top of the Dialog primitive.
 *
 * ### Component Props
 *
 * See [Dialog props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop            | Type                       | Default |
 * | :-------------- | :------------------------- | :------ |
 * | _defaultOpen_   | `boolean`                  | -       |
 * | _open_          | `boolean`                  | -       |
 * | _onOpenChange_  | `(open: boolean) => void`  | -       |
 * | _modal_         | `boolean`                  | `true`  |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Sheet>
 *   <SheetTrigger>Open</SheetTrigger>
 *   <SheetContent>
 *     <SheetHeader>
 *       <SheetTitle>Sheet Title</SheetTitle>
 *       <SheetDescription>Sheet description text.</SheetDescription>
 *     </SheetHeader>
 *     <div>Sheet body content</div>
 *     <SheetFooter>
 *       <SheetClose asChild>
 *         <Button variant="outline">Cancel</Button>
 *       </SheetClose>
 *     </SheetFooter>
 *   </SheetContent>
 * </Sheet>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 2.1.2 No Keyboard Trap — Escape key closes the sheet; focus returns to trigger.
 * @wcag 2.4.3 Focus Order — focus is trapped within the sheet while open via Radix Dialog primitive.
 * @wcag 4.1.2 Name, Role, Value — `dialog` role with `aria-labelledby` / `aria-describedby` via Radix Dialog primitives; close button has `sr-only` label.
 */
function Sheet({...props}: React.ComponentProps<typeof SheetPrimitive.Root>) {
  return <SheetPrimitive.Root data-slot="sheet" {...props} />;
}

/**
 * A button that opens the sheet when clicked.
 *
 * ### Component Props
 *
 * See [Dialog Trigger props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SheetTrigger asChild>
 *   <Button>Open Sheet</Button>
 * </SheetTrigger>
 * ```
 */
function SheetTrigger({
  ...props
}: React.ComponentProps<typeof SheetPrimitive.Trigger>) {
  return <SheetPrimitive.Trigger data-slot="sheet-trigger" {...props} />;
}

/**
 * A button that closes the sheet when clicked.
 *
 * ### Component Props
 *
 * See [Dialog Close props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SheetClose asChild>
 *   <Button variant="outline">Close</Button>
 * </SheetClose>
 * ```
 */
function SheetClose({
  ...props
}: React.ComponentProps<typeof SheetPrimitive.Close>) {
  return <SheetPrimitive.Close data-slot="sheet-close" {...props} />;
}

function SheetPortal({
  ...props
}: React.ComponentProps<typeof SheetPrimitive.Portal>) {
  return <SheetPrimitive.Portal data-slot="sheet-portal" {...props} />;
}

function SheetOverlay({
  className,
  ...props
}: React.ComponentProps<typeof SheetPrimitive.Overlay>) {
  return (
    <SheetPrimitive.Overlay
      data-slot="sheet-overlay"
      className={cn(
        'fixed inset-0 z-overlay bg-dds-overlay-backdrop duration-fast ease-out data-ending-style:opacity-0 data-starting-style:opacity-0 supports-backdrop-filter:backdrop-blur-xs data-open:animate-in data-open:fade-in-0 data-closed:animate-out data-closed:fade-out-0',
        className
      )}
      {...props}
    />
  );
}

/**
 * The main content area of the sheet that slides in from an edge of the screen.
 *
 * ### Component Props
 *
 * See [Dialog Content props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop                    | Type                                          | Default   |
 * | :---------------------- | :-------------------------------------------- | :-------- |
 * | _asChild_               | `boolean`                                     | `false`   |
 * | _side_                  | `"top" \| "right" \| "bottom" \| "left"`      | `"right"` |
 * | _showCloseButton_       | `boolean`                                     | `true`    |
 * | _forceMount_            | `boolean`                                     | -         |
 * | _onOpenAutoFocus_       | `(event: Event) => void`                      | -         |
 * | _onCloseAutoFocus_      | `(event: Event) => void`                      | -         |
 * | _onEscapeKeyDown_       | `(event: KeyboardEvent) => void`              | -         |
 * | _onPointerDownOutside_  | `(event: PointerDownOutsideEvent) => void`    | -         |
 * | _onInteractOutside_     | `(event: Event) => void`                      | -         |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SheetContent side="left">
 *   <SheetHeader>
 *     <SheetTitle>Navigation</SheetTitle>
 *   </SheetHeader>
 *   <nav>Menu items here</nav>
 * </SheetContent>
 * ```
 */
function SheetContent({
  className,
  children,
  side = 'right',
  showCloseButton = true,
  ...props
}: React.ComponentProps<typeof SheetPrimitive.Content> & {
  side?: 'top' | 'right' | 'bottom' | 'left';
  showCloseButton?: boolean;
}) {
  return (
    <SheetPortal>
      <SheetOverlay />
      <SheetPrimitive.Content
        data-slot="sheet-content"
        data-side={side}
        className={cn(
          'fixed z-overlay flex flex-col gap-4 bg-popover bg-clip-padding text-sm shadow-overlay duration-normal ease-in-out data-[side=bottom]:inset-x-0 data-[side=bottom]:bottom-0 data-[side=bottom]:h-auto data-[side=bottom]:border-t data-[side=bottom]:border-border data-[side=left]:inset-y-0 data-[side=left]:left-0 data-[side=left]:h-full data-[side=left]:w-3/4 data-[side=left]:border-r data-[side=right]:inset-y-0 data-[side=right]:right-0 data-[side=right]:h-full data-[side=right]:w-3/4 data-[side=right]:border-l data-[side=top]:inset-x-0 data-[side=top]:top-0 data-[side=top]:h-auto data-[side=top]:border-b data-[side=top]:border-border data-[side=left]:sm:max-w-sm data-[side=right]:sm:max-w-sm data-open:animate-in data-open:fade-in-0 data-[side=bottom]:data-open:slide-in-from-bottom-10 data-[side=left]:data-open:slide-in-from-left-10 data-[side=right]:data-open:slide-in-from-right-10 data-[side=top]:data-open:slide-in-from-top-10 data-closed:animate-out data-closed:fade-out-0 data-[side=bottom]:data-closed:slide-out-to-bottom-10 data-[side=left]:data-closed:slide-out-to-left-10 data-[side=right]:data-closed:slide-out-to-right-10 data-[side=top]:data-closed:slide-out-to-top-10',
          className
        )}
        {...props}
      >
        {children}
        {showCloseButton && (
          <SheetPrimitive.Close data-slot="sheet-close" asChild>
            <Button
              variant="ghost"
              className="absolute top-4 right-4"
              size="icon-sm"
            >
              <X />
              <span className="sr-only">Close</span>
            </Button>
          </SheetPrimitive.Close>
        )}
      </SheetPrimitive.Content>
    </SheetPortal>
  );
}

/**
 * A container for the sheet's header section, typically containing SheetTitle and SheetDescription.
 *
 * ### Component Props
 *
 * Accepts all standard `div` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SheetHeader>
 *   <SheetTitle>Edit Profile</SheetTitle>
 *   <SheetDescription>Make changes to your profile here.</SheetDescription>
 * </SheetHeader>
 * ```
 */
function SheetHeader({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="sheet-header"
      className={cn('flex flex-col gap-1.5 p-4', className)}
      {...props}
    />
  );
}

/**
 * A container for the sheet's footer section, typically containing action buttons.
 *
 * ### Component Props
 *
 * Accepts all standard `div` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SheetFooter>
 *   <SheetClose asChild>
 *     <Button variant="outline">Cancel</Button>
 *   </SheetClose>
 *   <Button>Save</Button>
 * </SheetFooter>
 * ```
 */
function SheetFooter({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="sheet-footer"
      className={cn('mt-auto flex flex-col gap-2 p-4', className)}
      {...props}
    />
  );
}

/**
 * An accessible title for the sheet, announced by screen readers.
 *
 * ### Component Props
 *
 * See [Dialog Title props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SheetTitle>Sheet Title</SheetTitle>
 * ```
 */
function SheetTitle({
  className,
  ...props
}: React.ComponentProps<typeof SheetPrimitive.Title>) {
  return (
    <SheetPrimitive.Title
      data-slot="sheet-title"
      className={cn(
        'leading-heading font-heading tracking-heading text-foreground',
        className
      )}
      {...props}
    />
  );
}

/**
 * An accessible description for the sheet, announced by screen readers.
 *
 * ### Component Props
 *
 * See [Dialog Description props](https://www.radix-ui.com/primitives/docs/components/dialog) documentation for full details.
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SheetDescription>
 *   This action cannot be undone. This will permanently delete your account.
 * </SheetDescription>
 * ```
 */
function SheetDescription({
  className,
  ...props
}: React.ComponentProps<typeof SheetPrimitive.Description>) {
  return (
    <SheetPrimitive.Description
      data-slot="sheet-description"
      className={cn('text-sm text-muted-foreground', className)}
      {...props}
    />
  );
}

export {
  Sheet,
  SheetTrigger,
  SheetClose,
  SheetContent,
  SheetHeader,
  SheetFooter,
  SheetTitle,
  SheetDescription,
};
