// @docker/trident/src/components/ui/sidebar.tsx
'use client';

import type {AsChild} from '@/lib/types';

import {Button} from '@/components/ui/button';
import {Input} from '@/components/ui/input';
import {Separator} from '@/components/ui/separator';
import {
  Sheet,
  SheetContent,
  SheetDescription,
  SheetHeader,
  SheetTitle,
} from '@/components/ui/sheet';
import {Skeleton} from '@/components/ui/skeleton';
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
} from '@/components/ui/tooltip';
import {useIsMobile} from '@/hooks/use-mobile';
import {cn, cssVars} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {PanelLeft} from 'lucide-react';
import {Slot} from 'radix-ui';
import * as React from 'react';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

// Default widths mirror shadcn's sidebar contract. SidebarProvider writes
// these into CSS variables, so they must be literal fallback values rather
// than self-referential `var(--sidebar-width, …)` expressions.
// Values are sourced from DDS tokens: --dds-width-sidebar-{default,expanded,collapsed}.
const SIDEBAR_COOKIE_NAME = 'sidebar_state';
const SIDEBAR_COOKIE_MAX_AGE = 60 * 60 * 24 * 7;
const SIDEBAR_WIDTH = 'var(--dds-width-sidebar-default, 16rem)';
const SIDEBAR_WIDTH_MOBILE = 'var(--dds-width-sidebar-mobile, 18rem)';
const SIDEBAR_WIDTH_ICON = 'var(--dds-width-sidebar-collapsed, 3rem)';
const SIDEBAR_KEYBOARD_SHORTCUT = 'b';

// ---------------------------------------------------------------------------
// Context
// ---------------------------------------------------------------------------

type SidebarContextProps = {
  state: 'expanded' | 'collapsed';
  open: boolean;
  setOpen: (open: boolean | ((prev: boolean) => boolean)) => void;
  openMobile: boolean;
  setOpenMobile: (open: boolean) => void;
  isMobile: boolean;
  toggleSidebar: () => void;
};

const SidebarContext = React.createContext<SidebarContextProps | null>(null);

/**
 * A hook that provides access to the sidebar context, including its state and toggle functions.
 * Must be used within a SidebarProvider.
 *
 * ### Return Value
 *
 * | Property         | Type                          | Description                                  |
 * | :--------------- | :---------------------------- | :------------------------------------------- |
 * | _state_          | `"expanded" \| "collapsed"`   | Current sidebar state                        |
 * | _open_           | `boolean`                     | Whether the sidebar is open (desktop)        |
 * | _setOpen_        | `(open: boolean) => void`     | Set the sidebar open state (desktop)         |
 * | _openMobile_     | `boolean`                     | Whether the mobile sidebar sheet is open     |
 * | _setOpenMobile_  | `(open: boolean) => void`     | Set the mobile sidebar sheet open state      |
 * | _isMobile_       | `boolean`                     | Whether the viewport is below mobile breakpoint |
 * | _toggleSidebar_  | `() => void`                  | Toggle the sidebar open/closed               |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * function MyComponent() {
 *   const { state, toggleSidebar } = useSidebar();
 *   return <button onClick={toggleSidebar}>{state}</button>;
 * }
 * ```
 */
function useSidebar() {
  const context = React.useContext(SidebarContext);
  if (!context) {
    throw new Error('useSidebar must be used within a SidebarProvider.');
  }

  return context;
}

// ---------------------------------------------------------------------------
// SidebarProvider
// ---------------------------------------------------------------------------

/**
 * Provides sidebar context to child components. Wraps the application layout to enable sidebar state management.
 *
 * ### Component Props
 *
 * | Prop            | Type                       | Default |
 * | :-------------- | :------------------------- | :------ |
 * | _defaultOpen_   | `boolean`                  | `true`  |
 * | _open_          | `boolean`                  | -       |
 * | _onOpenChange_  | `(open: boolean) => void`  | -       |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarProvider>
 *   <Sidebar>
 *     <SidebarContent>...</SidebarContent>
 *   </Sidebar>
 *   <SidebarInset>
 *     <main>Page content</main>
 *   </SidebarInset>
 * </SidebarProvider>
 * ```
 */
function SidebarProvider({
  defaultOpen = true,
  open: openProp,
  onOpenChange: setOpenProp,
  className,
  style,
  children,
  ...props
}: React.ComponentProps<'div'> & {
  defaultOpen?: boolean;
  open?: boolean;
  onOpenChange?: (open: boolean) => void;
}) {
  const isMobile = useIsMobile();
  const [openMobile, setOpenMobile] = React.useState(false);

  // This is the internal state of the sidebar.
  // We use openProp and setOpenProp for control from outside the component.
  const [_open, _setOpen] = React.useState(defaultOpen);
  const open = openProp ?? _open;
  const setOpen = React.useCallback(
    (value: boolean | ((value: boolean) => boolean)) => {
      const openState = typeof value === 'function' ? value(open) : value;
      if (setOpenProp) {
        setOpenProp(openState);
      } else {
        _setOpen(openState);
      }

      // Persist the sidebar state in a cookie so it can be restored on the
      // next visit (and on the server in SSR setups).
      if (typeof document !== 'undefined') {
        document.cookie = `${SIDEBAR_COOKIE_NAME}=${openState}; path=/; max-age=${SIDEBAR_COOKIE_MAX_AGE}`;
      }
    },
    [setOpenProp, open]
  );

  // Helper to toggle the sidebar.
  const toggleSidebar = React.useCallback(() => {
    return isMobile ? setOpenMobile(prev => !prev) : setOpen(prev => !prev);
  }, [isMobile, setOpen, setOpenMobile]);

  // Adds a keyboard shortcut (Cmd/Ctrl+B) to toggle the sidebar.
  React.useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if (
        event.key === SIDEBAR_KEYBOARD_SHORTCUT &&
        (event.metaKey || event.ctrlKey)
      ) {
        event.preventDefault();
        toggleSidebar();
      }
    };

    window.addEventListener('keydown', handleKeyDown);
    return () => window.removeEventListener('keydown', handleKeyDown);
  }, [toggleSidebar]);

  // We add a state so that we can do data-state="expanded" or "collapsed".
  // This makes it easier to style the sidebar with Tailwind classes.
  const state = open ? 'expanded' : 'collapsed';

  const contextValue = React.useMemo<SidebarContextProps>(
    () => ({
      state,
      open,
      setOpen,
      isMobile,
      openMobile,
      setOpenMobile,
      toggleSidebar,
    }),
    [state, open, setOpen, isMobile, openMobile, setOpenMobile, toggleSidebar]
  );

  return (
    <SidebarContext.Provider value={contextValue}>
      <TooltipProvider delayDuration={0}>
        <div
          data-slot="sidebar-wrapper"
          style={cssVars({
            '--sidebar-width': SIDEBAR_WIDTH,
            '--sidebar-width-icon': SIDEBAR_WIDTH_ICON,
            ...style,
          })}
          className={cn(
            'group/sidebar-wrapper flex min-h-svh w-full has-data-[variant=inset]:bg-sidebar',
            className
          )}
          {...props}
        >
          {children}
        </div>
      </TooltipProvider>
    </SidebarContext.Provider>
  );
}

// ---------------------------------------------------------------------------
// Sidebar layout components
// ---------------------------------------------------------------------------

/**
 * The main sidebar container that supports collapsible states and multiple variants.
 *
 * ### Component Props
 *
 * | Prop            | Type                                    | Default       |
 * | :-------------- | :-------------------------------------- | :------------ |
 * | _side_          | `"left" \| "right"`                     | `"left"`      |
 * | _variant_       | `"sidebar" \| "floating" \| "inset"`    | `"sidebar"`   |
 * | _collapsible_   | `"offcanvas" \| "icon" \| "none"`       | `"offcanvas"` |
 *
 *
 * ### Example Usage
 *
 * **Sidebar**
 * ```tsx
 * <Sidebar>
 *   <SidebarHeader />
 *   <SidebarContent>
 *     <SidebarGroup>...</SidebarGroup>
 *   </SidebarContent>
 *   <SidebarFooter />
 * </Sidebar>
 * ```
 *
 * **Floating + Icon Collapsible**
 * ```tsx
 * <Sidebar variant="floating" collapsible="icon">
 *   <SidebarContent>...</SidebarContent>
 * </Sidebar>
 * ```
 *
 * **Inset + Right Side**
 * ```tsx
 * <Sidebar variant="inset" side="right">
 *   <SidebarContent>...</SidebarContent>
 * </Sidebar>
 * ```
 */
function Sidebar({
  side = 'left',
  variant = 'sidebar',
  collapsible = 'offcanvas',
  className,
  children,
  dir,
  ...props
}: React.ComponentProps<'div'> & {
  side?: 'left' | 'right';
  variant?: 'sidebar' | 'floating' | 'inset';
  collapsible?: 'offcanvas' | 'icon' | 'none';
}) {
  const {isMobile, state, openMobile, setOpenMobile} = useSidebar();

  if (collapsible === 'none') {
    return (
      <div
        data-slot="sidebar"
        className={cn(
          'flex h-full w-(--sidebar-width) flex-col bg-sidebar text-sidebar-foreground',
          className
        )}
        {...props}
      >
        {children}
      </div>
    );
  }

  if (isMobile) {
    return (
      <Sheet open={openMobile} onOpenChange={setOpenMobile} {...props}>
        <SheetContent
          dir={dir}
          data-sidebar="sidebar"
          data-slot="sidebar"
          data-mobile="true"
          className="w-(--sidebar-width) bg-sidebar p-0 text-sidebar-foreground [&>button]:hidden"
          style={cssVars({
            '--sidebar-width': SIDEBAR_WIDTH_MOBILE,
          })}
          side={side}
        >
          <SheetHeader className="sr-only">
            <SheetTitle>Sidebar</SheetTitle>
            <SheetDescription>Displays the mobile sidebar.</SheetDescription>
          </SheetHeader>
          <div className="flex h-full w-full flex-col">{children}</div>
        </SheetContent>
      </Sheet>
    );
  }

  return (
    <div
      className="group peer hidden text-sidebar-foreground md:block"
      data-state={state}
      data-collapsible={state === 'collapsed' ? collapsible : ''}
      data-variant={variant}
      data-side={side}
      data-slot="sidebar"
    >
      {/* This is what handles the sidebar gap on desktop */}
      <div
        data-slot="sidebar-gap"
        className={cn(
          'relative w-(--sidebar-width) bg-transparent transition-[width] duration-normal ease-in-out',
          'group-data-[collapsible=offcanvas]:w-0',
          'group-data-[side=right]:rotate-180',
          variant === 'floating' || variant === 'inset'
            ? 'group-data-[collapsible=icon]:w-[calc(var(--sidebar-width-icon)+(--spacing(4)))]'
            : 'group-data-[collapsible=icon]:w-(--sidebar-width-icon)'
        )}
      />
      <div
        data-slot="sidebar-container"
        data-side={side}
        className={cn(
          '@container/sidebar fixed inset-y-0 z-raised hidden h-svh w-(--sidebar-width) transition-[left,right,width] duration-normal ease-in-out data-[side=left]:left-0 data-[side=left]:group-data-[collapsible=offcanvas]:left-[calc(var(--sidebar-width)*-1)] data-[side=right]:right-0 data-[side=right]:group-data-[collapsible=offcanvas]:right-[calc(var(--sidebar-width)*-1)] md:flex',
          // Adjust the padding for floating and inset variants.
          variant === 'floating' || variant === 'inset'
            ? 'p-2 group-data-[collapsible=icon]:w-[calc(var(--sidebar-width-icon)+(--spacing(4))+var(--dds-border-offset))]'
            : 'group-data-[collapsible=icon]:w-(--sidebar-width-icon) group-data-[side=left]:border-r group-data-[side=right]:border-l',
          className
        )}
        {...props}
      >
        <div
          data-sidebar="sidebar"
          data-slot="sidebar-inner"
          className="flex size-full flex-col bg-sidebar group-data-[variant=floating]:rounded-lg group-data-[variant=floating]:shadow-raised group-data-[variant=floating]:ring-1 group-data-[variant=floating]:ring-sidebar-border"
        >
          {children}
        </div>
      </div>
    </div>
  );
}

/**
 * A button that toggles the sidebar open or closed. Renders a toggle icon with screen reader text.
 *
 * ### Component Props
 *
 * Accepts all Button component props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarTrigger />
 * ```
 */
function SidebarTrigger({
  className,
  onClick,
  ...props
}: React.ComponentProps<typeof Button>) {
  const {toggleSidebar} = useSidebar();

  return (
    <Button
      data-sidebar="trigger"
      data-slot="sidebar-trigger"
      variant="ghost"
      size="icon-sm"
      className={cn(className)}
      onClick={event => {
        onClick?.(event);
        toggleSidebar();
      }}
      {...props}
    >
      <PanelLeft />
      <span className="sr-only">Toggle Sidebar</span>
    </Button>
  );
}

/**
 * A thin rail at the sidebar edge that toggles the sidebar on click. Provides a subtle resize-like affordance.
 *
 * ### Component Props
 *
 * Accepts all standard `button` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Sidebar>
 *   <SidebarContent>...</SidebarContent>
 *   <SidebarRail />
 * </Sidebar>
 * ```
 */
function SidebarRail({className, ...props}: React.ComponentProps<'button'>) {
  const {toggleSidebar} = useSidebar();

  return (
    <button
      data-sidebar="rail"
      data-slot="sidebar-rail"
      aria-label="Toggle Sidebar"
      tabIndex={-1}
      onClick={toggleSidebar}
      title="Toggle Sidebar"
      className={cn(
        'absolute inset-y-0 z-floating hidden w-4 transition-all ease-in-out group-data-[side=left]:-right-4 group-data-[side=right]:left-0 after:absolute after:inset-y-0 after:start-1/2 after:w-0.5 hover:after:bg-sidebar-border sm:flex ltr:-translate-x-1/2 rtl:-translate-x-1/2',
        'in-data-[side=left]:cursor-w-resize in-data-[side=right]:cursor-e-resize',
        '[[data-side=left][data-state=collapsed]_&]:cursor-e-resize [[data-side=right][data-state=collapsed]_&]:cursor-w-resize',
        'group-data-[collapsible=offcanvas]:translate-x-0 group-data-[collapsible=offcanvas]:after:left-full hover:group-data-[collapsible=offcanvas]:bg-sidebar',
        '[[data-side=left][data-collapsible=offcanvas]_&]:-right-2',
        '[[data-side=right][data-collapsible=offcanvas]_&]:-left-2',
        className
      )}
      {...props}
    />
  );
}

/**
 * The main content area that sits alongside the sidebar. Renders as a `main` element.
 *
 * ### Component Props
 *
 * Accepts all standard `main` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarProvider>
 *   <Sidebar>...</Sidebar>
 *   <SidebarInset>
 *     <main>Page content</main>
 *   </SidebarInset>
 * </SidebarProvider>
 * ```
 */
function SidebarInset({className, ...props}: React.ComponentProps<'main'>) {
  return (
    <main
      data-slot="sidebar-inset"
      className={cn(
        'relative flex w-full flex-1 flex-col bg-background md:peer-data-[variant=inset]:m-2 md:peer-data-[variant=inset]:ml-0 md:peer-data-[variant=inset]:rounded-lg md:peer-data-[variant=inset]:shadow-raised md:peer-data-[variant=inset]:peer-data-[state=collapsed]:ml-2',
        className
      )}
      {...props}
    />
  );
}

/**
 * A styled input field for use within the sidebar, such as for search functionality.
 *
 * ### Component Props
 *
 * Accepts all Input component props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarHeader>
 *   <SidebarInput placeholder="Search..." />
 * </SidebarHeader>
 * ```
 */
function SidebarInput({
  className,
  ...props
}: React.ComponentProps<typeof Input>) {
  return (
    <Input
      data-slot="sidebar-input"
      data-sidebar="input"
      className={cn('h-8 w-full bg-input-bg shadow-none', className)}
      {...props}
    />
  );
}

/**
 * A container for the top section of the sidebar, typically containing branding or a search input.
 *
 * ### Component Props
 *
 * Accepts all standard `div` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarHeader>
 *   <h2>My App</h2>
 * </SidebarHeader>
 * ```
 */
function SidebarHeader({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="sidebar-header"
      data-sidebar="header"
      className={cn('flex flex-col gap-2 p-2', className)}
      {...props}
    />
  );
}

/**
 * A container for the bottom section of the sidebar, typically containing user info or settings.
 *
 * ### Component Props
 *
 * Accepts all standard `div` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarFooter>
 *   <SidebarMenu>
 *     <SidebarMenuItem>
 *       <SidebarMenuButton>Settings</SidebarMenuButton>
 *     </SidebarMenuItem>
 *   </SidebarMenu>
 * </SidebarFooter>
 * ```
 */
function SidebarFooter({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="sidebar-footer"
      data-sidebar="footer"
      className={cn('flex flex-col gap-2 p-2', className)}
      {...props}
    />
  );
}

/**
 * A scrollable container for the main sidebar content, placed between SidebarHeader and SidebarFooter.
 *
 * ### Component Props
 *
 * Accepts all standard `div` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarContent>
 *   <SidebarGroup>
 *     <SidebarMenu>...</SidebarMenu>
 *   </SidebarGroup>
 * </SidebarContent>
 * ```
 */
function SidebarContent({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="sidebar-content"
      data-sidebar="content"
      className={cn(
        'no-scrollbar flex min-h-0 flex-1 flex-col gap-2 overflow-auto group-data-[collapsible=icon]:overflow-hidden',
        className
      )}
      {...props}
    />
  );
}

// ---------------------------------------------------------------------------
// Sidebar group components
// ---------------------------------------------------------------------------

/**
 * A visual separator for dividing sections within the sidebar.
 *
 * ### Component Props
 *
 * Accepts all Separator component props.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarContent>
 *   <SidebarGroup>...</SidebarGroup>
 *   <SidebarSeparator />
 *   <SidebarGroup>...</SidebarGroup>
 * </SidebarContent>
 * ```
 */
function SidebarSeparator({
  className,
  ...props
}: React.ComponentProps<typeof Separator>) {
  return (
    <Separator
      data-slot="sidebar-separator"
      data-sidebar="separator"
      className={cn('mx-2 w-auto bg-sidebar-border', className)}
      {...props}
    />
  );
}

/**
 * A logical grouping of sidebar items, optionally with a label and action button.
 *
 * ### Component Props
 *
 * Accepts all standard `div` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarGroup>
 *   <SidebarGroupLabel>Navigation</SidebarGroupLabel>
 *   <SidebarGroupContent>
 *     <SidebarMenu>...</SidebarMenu>
 *   </SidebarGroupContent>
 * </SidebarGroup>
 * ```
 */
function SidebarGroup({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="sidebar-group"
      data-sidebar="group"
      className={cn('relative flex w-full min-w-0 flex-col p-2', className)}
      {...props}
    />
  );
}

/**
 * A label for a SidebarGroup, displayed as a small heading above the group content.
 *
 * ### Component Props
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarGroupLabel>Navigation</SidebarGroupLabel>
 * ```
 */
function SidebarGroupLabel({
  className,
  asChild = false,
  ...props
}: React.ComponentProps<'div'> & AsChild) {
  const Comp = asChild ? Slot.Root : 'div';

  return (
    <Comp
      data-slot="sidebar-group-label"
      data-sidebar="group-label"
      className={cn(
        'flex h-8 shrink-0 items-center rounded px-2 text-xs font-medium text-sidebar-foreground-muted ring-sidebar-ring outline-hidden transition-[margin,opacity] duration-normal ease-in-out group-data-[collapsible=icon]:-mt-8 group-data-[collapsible=icon]:opacity-0 focus-visible:ring-2 @max-[200px]/sidebar:hidden [&>svg]:size-4 [&>svg]:shrink-0',
        className
      )}
      {...props}
    />
  );
}

/**
 * An action button positioned at the top-right of a SidebarGroup, typically used for add or edit actions.
 *
 * ### Component Props
 *
 * | Prop      | Type      | Default |
 * | :-------- | :-------- | :------ |
 * | _asChild_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarGroup>
 *   <SidebarGroupLabel>Projects</SidebarGroupLabel>
 *   <SidebarGroupAction>
 *     <PlusIcon /> <span className="sr-only">Add Project</span>
 *   </SidebarGroupAction>
 *   <SidebarGroupContent>...</SidebarGroupContent>
 * </SidebarGroup>
 * ```
 */
function SidebarGroupAction({
  className,
  asChild = false,
  ...props
}: React.ComponentProps<'button'> & AsChild) {
  const Comp = asChild ? Slot.Root : 'button';

  return (
    <Comp
      data-slot="sidebar-group-action"
      data-sidebar="group-action"
      className={cn(
        'absolute top-3.5 right-3 flex aspect-square w-5 items-center justify-center rounded p-0 text-sidebar-foreground ring-sidebar-ring outline-hidden transition-transform ease-default group-data-[collapsible=icon]:hidden after:absolute after:-inset-2 hover:bg-hover-overlay hover:text-sidebar-accent-foreground focus-visible:ring-2 md:after:hidden [&>svg]:size-4 [&>svg]:shrink-0',
        className
      )}
      {...props}
    />
  );
}

/**
 * A wrapper for the content within a SidebarGroup, containing menus or other elements.
 *
 * ### Component Props
 *
 * Accepts all standard `div` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarGroupContent>
 *   <SidebarMenu>...</SidebarMenu>
 * </SidebarGroupContent>
 * ```
 */
function SidebarGroupContent({
  className,
  ...props
}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="sidebar-group-content"
      data-sidebar="group-content"
      className={cn('w-full text-sm', className)}
      {...props}
    />
  );
}

// ---------------------------------------------------------------------------
// Sidebar menu components
// ---------------------------------------------------------------------------

/**
 * A list container for sidebar menu items, rendered as an unordered list.
 *
 * ### Component Props
 *
 * Accepts all standard `ul` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarMenu>
 *   <SidebarMenuItem>
 *     <SidebarMenuButton>Dashboard</SidebarMenuButton>
 *   </SidebarMenuItem>
 *   <SidebarMenuItem>
 *     <SidebarMenuButton>Settings</SidebarMenuButton>
 *   </SidebarMenuItem>
 * </SidebarMenu>
 * ```
 */
function SidebarMenu({className, ...props}: React.ComponentProps<'ul'>) {
  return (
    <ul
      data-slot="sidebar-menu"
      data-sidebar="menu"
      className={cn('flex w-full min-w-0 flex-col gap-1', className)}
      {...props}
    />
  );
}

/**
 * A list item within a SidebarMenu, wrapping a SidebarMenuButton and optional actions.
 *
 * ### Component Props
 *
 * Accepts all standard `li` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarMenuItem>
 *   <SidebarMenuButton>Item Label</SidebarMenuButton>
 * </SidebarMenuItem>
 * ```
 */
function SidebarMenuItem({className, ...props}: React.ComponentProps<'li'>) {
  return (
    <li
      data-slot="sidebar-menu-item"
      data-sidebar="menu-item"
      className={cn('group/menu-item relative', className)}
      {...props}
    />
  );
}

const sidebarMenuButtonVariants = cva(
  'peer/menu-button group/menu-button flex w-full items-center gap-2 overflow-hidden rounded border-l-2 border-l-transparent p-2 text-left text-sm ring-sidebar-ring outline-hidden transition-[width,height,padding,border-color] ease-in-out group-has-data-[sidebar=menu-action]/menu-item:pr-8 group-data-[collapsible=icon]:size-8! group-data-[collapsible=icon]:p-2! hover:bg-hover-overlay focus-visible:ring-2 active:bg-active-overlay disabled:pointer-events-none disabled:opacity-50 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-open:hover:bg-hover-overlay data-active:border-l-primary data-active:bg-accent-subtle data-active:font-medium data-active:text-sidebar-accent-foreground [&_svg]:size-4 [&_svg]:shrink-0 [&_svg]:[stroke-width:1.5] [&>span:last-child]:truncate @max-[200px]/sidebar:[&>span:last-child]:hidden',
  {
    variants: {
      variant: {
        default: 'hover:bg-hover-overlay',
        outline:
          'bg-background shadow-[0_0_0_1px_var(--sidebar-border)] hover:bg-sidebar-accent hover:text-sidebar-accent-foreground hover:shadow-[0_0_0_1px_var(--sidebar-accent)]',
      },
      size: {
        default: 'h-8 text-sm',
        sm: 'h-7 text-xs',
        lg: 'h-12 text-sm group-data-[collapsible=icon]:p-0!',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

/**
 * An interactive button within a SidebarMenuItem. Supports variants, sizes, active state, and optional tooltip.
 *
 * ### Component Props
 *
 * | Prop        | Type                                               | Default     |
 * | :---------- | :------------------------------------------------- | :---------- |
 * | _asChild_   | `boolean`                                          | `false`     |
 * | _isActive_  | `boolean`                                          | `false`     |
 * | _variant_   | `"default" \| "outline"`                           | `"default"` |
 * | _size_      | `"default" \| "sm" \| "lg"`                        | `"default"` |
 * | _tooltip_   | `string \| React.ComponentProps<typeof TooltipContent>` | -      |
 *
 *
 * ### Example Usage
 *
 * **Default**
 * ```tsx
 * <SidebarMenuButton isActive tooltip="Dashboard">
 *   <HomeIcon />
 *   <span>Dashboard</span>
 * </SidebarMenuButton>
 * ```
 *
 * **Outline + Small**
 * ```tsx
 * <SidebarMenuButton variant="outline" size="sm">
 *   <SettingsIcon />
 *   <span>Settings</span>
 * </SidebarMenuButton>
 * ```
 *
 * **Large + asChild**
 * ```tsx
 * <SidebarMenuButton size="lg" asChild>
 *   <a href="/profile">
 *     <Avatar />
 *     <span>John Doe</span>
 *   </a>
 * </SidebarMenuButton>
 * ```
 */
function SidebarMenuButton({
  asChild = false,
  isActive = false,
  variant = 'default',
  size = 'default',
  tooltip,
  className,
  ...props
}: React.ComponentProps<'button'> &
  AsChild & {
    isActive?: boolean;
    tooltip?: string | React.ComponentProps<typeof TooltipContent>;
  } & VariantProps<typeof sidebarMenuButtonVariants>) {
  const Comp = asChild ? Slot.Root : 'button';
  const {isMobile, state} = useSidebar();

  const button = (
    <Comp
      data-slot="sidebar-menu-button"
      data-sidebar="menu-button"
      data-size={size}
      data-active={isActive}
      className={cn(sidebarMenuButtonVariants({variant, size}), className)}
      {...props}
    />
  );

  if (!tooltip) {
    return button;
  }

  if (typeof tooltip === 'string') {
    tooltip = {
      children: tooltip,
    };
  }

  return (
    <Tooltip>
      <TooltipTrigger asChild>{button}</TooltipTrigger>
      <TooltipContent
        side="right"
        align="center"
        hidden={state !== 'collapsed' || isMobile}
        {...tooltip}
      />
    </Tooltip>
  );
}

/**
 * An action button displayed alongside a SidebarMenuButton, such as a context menu trigger or delete button.
 *
 * ### Component Props
 *
 * | Prop          | Type      | Default |
 * | :------------ | :-------- | :------ |
 * | _asChild_     | `boolean` | `false` |
 * | _showOnHover_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarMenuItem>
 *   <SidebarMenuButton>Item</SidebarMenuButton>
 *   <SidebarMenuAction showOnHover>
 *     <MoreHorizontalIcon />
 *   </SidebarMenuAction>
 * </SidebarMenuItem>
 * ```
 */
function SidebarMenuAction({
  className,
  asChild = false,
  showOnHover = false,
  ...props
}: React.ComponentProps<'button'> &
  AsChild & {
    showOnHover?: boolean;
  }) {
  const Comp = asChild ? Slot.Root : 'button';

  return (
    <Comp
      data-slot="sidebar-menu-action"
      data-sidebar="menu-action"
      className={cn(
        'absolute top-1.5 right-1 flex aspect-square w-5 items-center justify-center rounded p-0 text-sidebar-foreground ring-sidebar-ring outline-hidden transition-transform ease-default group-data-[collapsible=icon]:hidden peer-hover/menu-button:text-sidebar-accent-foreground peer-data-[size=default]/menu-button:top-1.5 peer-data-[size=lg]/menu-button:top-2.5 peer-data-[size=sm]/menu-button:top-1 after:absolute after:-inset-2 hover:bg-hover-overlay hover:text-sidebar-accent-foreground focus-visible:ring-2 md:after:hidden [&>svg]:size-4 [&>svg]:shrink-0',
        showOnHover &&
          'group-focus-within/menu-item:opacity-100 group-hover/menu-item:opacity-100 peer-data-active/menu-button:text-sidebar-accent-foreground md:opacity-0 data-open:opacity-100',
        className
      )}
      {...props}
    />
  );
}

/**
 * A badge displayed alongside a SidebarMenuButton, used to show counts or status indicators.
 *
 * ### Component Props
 *
 * Accepts all standard `div` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarMenuItem>
 *   <SidebarMenuButton>Inbox</SidebarMenuButton>
 *   <SidebarMenuBadge>24</SidebarMenuBadge>
 * </SidebarMenuItem>
 * ```
 */
function SidebarMenuBadge({className, ...props}: React.ComponentProps<'div'>) {
  return (
    <div
      data-slot="sidebar-menu-badge"
      data-sidebar="menu-badge"
      className={cn(
        'pointer-events-none absolute right-1 flex h-5 min-w-5 items-center justify-center rounded px-1 text-xs font-medium text-sidebar-foreground tabular-nums select-none group-data-[collapsible=icon]:hidden peer-hover/menu-button:text-sidebar-accent-foreground peer-data-[size=default]/menu-button:top-1.5 peer-data-[size=lg]/menu-button:top-2.5 peer-data-[size=sm]/menu-button:top-1 peer-data-active/menu-button:text-sidebar-accent-foreground',
        className
      )}
      {...props}
    />
  );
}

/**
 * A loading skeleton placeholder for a sidebar menu item, with optional icon skeleton.
 *
 * ### Component Props
 *
 * | Prop       | Type      | Default |
 * | :--------- | :-------- | :------ |
 * | _showIcon_ | `boolean` | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarMenu>
 *   {Array.from({ length: 5 }).map((_, i) => (
 *     <SidebarMenuItem key={i}>
 *       <SidebarMenuSkeleton showIcon />
 *     </SidebarMenuItem>
 *   ))}
 * </SidebarMenu>
 * ```
 */
function SidebarMenuSkeleton({
  className,
  showIcon = false,
  ...props
}: React.ComponentProps<'div'> & {
  showIcon?: boolean;
}) {
  // Random width between 50 to 90%.
  const width = React.useMemo(() => {
    return `${Math.floor(Math.random() * 40) + 50}%`;
  }, []);

  return (
    <div
      data-slot="sidebar-menu-skeleton"
      data-sidebar="menu-skeleton"
      className={cn('flex h-8 items-center gap-2 rounded px-2', className)}
      {...props}
    >
      {showIcon && (
        <Skeleton
          className="size-4 rounded"
          data-sidebar="menu-skeleton-icon"
        />
      )}
      <Skeleton
        className="h-4 max-w-(--skeleton-width) flex-1"
        data-sidebar="menu-skeleton-text"
        style={cssVars({'--skeleton-width': width})}
      />
    </div>
  );
}

/**
 * A nested sub-menu list within a SidebarMenuItem, displayed with an indented border.
 *
 * ### Component Props
 *
 * Accepts all standard `ul` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarMenuItem>
 *   <SidebarMenuButton>Parent</SidebarMenuButton>
 *   <SidebarMenuSub>
 *     <SidebarMenuSubItem>
 *       <SidebarMenuSubButton>Child 1</SidebarMenuSubButton>
 *     </SidebarMenuSubItem>
 *     <SidebarMenuSubItem>
 *       <SidebarMenuSubButton>Child 2</SidebarMenuSubButton>
 *     </SidebarMenuSubItem>
 *   </SidebarMenuSub>
 * </SidebarMenuItem>
 * ```
 */
function SidebarMenuSub({className, ...props}: React.ComponentProps<'ul'>) {
  return (
    <ul
      data-slot="sidebar-menu-sub"
      data-sidebar="menu-sub"
      className={cn(
        'mx-3.5 flex min-w-0 translate-x-px flex-col gap-1 border-l border-sidebar-border px-2.5 py-0.5 group-data-[collapsible=icon]:hidden',
        className
      )}
      {...props}
    />
  );
}

/**
 * A list item within a SidebarMenuSub, wrapping a SidebarMenuSubButton.
 *
 * ### Component Props
 *
 * Accepts all standard `li` HTML attributes.
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarMenuSubItem>
 *   <SidebarMenuSubButton>Sub-item Label</SidebarMenuSubButton>
 * </SidebarMenuSubItem>
 * ```
 */
function SidebarMenuSubItem({className, ...props}: React.ComponentProps<'li'>) {
  return (
    <li
      data-slot="sidebar-menu-sub-item"
      data-sidebar="menu-sub-item"
      className={cn('group/menu-sub-item relative', className)}
      {...props}
    />
  );
}

/**
 * An interactive button for a sub-menu item within SidebarMenuSub. Supports active state and size variants.
 *
 * ### Component Props
 *
 * | Prop       | Type              | Default |
 * | :--------- | :---------------- | :------ |
 * | _asChild_  | `boolean`         | `false` |
 * | _size_     | `"sm" \| "md"`    | `"md"`  |
 * | _isActive_ | `boolean`         | `false` |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <SidebarMenuSubButton isActive>
 *   <span>Active Sub-item</span>
 * </SidebarMenuSubButton>
 * ```
 */
function SidebarMenuSubButton({
  asChild = false,
  size = 'md',
  isActive = false,
  className,
  ...props
}: React.ComponentProps<'a'> &
  AsChild & {
    size?: 'sm' | 'md';
    isActive?: boolean;
  }) {
  const Comp = asChild ? Slot.Root : 'a';

  return (
    <Comp
      data-slot="sidebar-menu-sub-button"
      data-sidebar="menu-sub-button"
      data-size={size}
      data-active={isActive}
      className={cn(
        'flex h-7 min-w-0 -translate-x-px items-center gap-2 overflow-hidden rounded px-2 text-sidebar-foreground ring-sidebar-ring outline-hidden group-data-[collapsible=icon]:hidden hover:bg-hover-overlay focus-visible:ring-2 active:bg-active-overlay disabled:pointer-events-none disabled:opacity-50 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-[size=md]:text-sm data-[size=sm]:text-xs data-active:bg-accent-subtle data-active:text-sidebar-accent-foreground [&_svg]:[stroke-width:1.5] [&>span:last-child]:truncate [&>svg]:size-4 [&>svg]:shrink-0',
        className
      )}
      {...props}
    />
  );
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {
  Sidebar,
  SidebarContent,
  SidebarFooter,
  SidebarGroup,
  SidebarGroupAction,
  SidebarGroupContent,
  SidebarGroupLabel,
  SidebarHeader,
  SidebarInput,
  SidebarInset,
  SidebarMenu,
  SidebarMenuAction,
  SidebarMenuBadge,
  SidebarMenuButton,
  SidebarMenuItem,
  SidebarMenuSkeleton,
  SidebarMenuSub,
  SidebarMenuSubButton,
  SidebarMenuSubItem,
  SidebarProvider,
  SidebarRail,
  SidebarSeparator,
  SidebarTrigger,
  SIDEBAR_WIDTH,
  SIDEBAR_WIDTH_ICON,
  useSidebar,
};

export type {SidebarContextProps};
