// @docker/trident/src/components/ui/slider.tsx
'use client';

import {cn} from '@/lib/utils';

import {Slider as SliderPrimitive} from 'radix-ui';
import * as React from 'react';

type MarkItem = {value: number; label?: string};

type SliderProps = React.ComponentProps<typeof SliderPrimitive.Root> & {
  /**
   * Show the current value as a label above/below each thumb.
   * @default false
   */
  showValue?: boolean;
  /**
   * Format the displayed value label.
   * @default String(value)
   */
  valueFormatter?: (value: number) => string;
  /**
   * Position of the value label relative to the thumb.
   * @default "top"
   */
  valuePosition?: 'top' | 'bottom';
  /**
   * Render step marks on the track.
   * - `true` — a dot at every step
   * - `MarkItem[]` — custom positions with optional text labels
   */
  marks?: boolean | MarkItem[];
  /**
   * Visual size of each mark dot.
   * @default "sm"
   */
  markSize?: 'sm' | 'default';
};

/**
 * An input where the user selects a value from within a given range.
 *
 * ### Component Props
 *
 * See [Slider props](https://www.radix-ui.com/primitives/docs/components/slider) documentation for full details.
 *
 * | Prop                    | Type                                                   | Default        |
 * | :---------------------- | :----------------------------------------------------- | :------------- |
 * | _asChild_               | `boolean`                                              | `false`        |
 * | _defaultValue_          | `number[]`                                             | -              |
 * | _value_                 | `number[]`                                             | -              |
 * | _onValueChange_         | `(value: number[]) => void`                            | -              |
 * | _onValueCommit_         | `(value: number[]) => void`                            | -              |
 * | _name_                  | `string`                                               | -              |
 * | _disabled_              | `boolean`                                              | `false`        |
 * | _orientation_           | `"horizontal" \| "vertical"`                           | `"horizontal"` |
 * | _dir_                   | `"ltr" \| "rtl"`                                       | -              |
 * | _inverted_              | `boolean`                                              | `false`        |
 * | _min_                   | `number`                                               | `0`            |
 * | _max_                   | `number`                                               | `100`          |
 * | _step_                  | `number`                                               | `1`            |
 * | _minStepsBetweenThumbs_ | `number`                                               | `0`            |
 * | _showValue_             | `boolean`                                              | `false`        |
 * | _valueFormatter_        | `(value: number) => string`                            | `String`       |
 * | _valuePosition_         | `"top" \| "bottom"`                                    | `"top"`        |
 * | _marks_                 | `boolean \| Array<{ value: number; label?: string }>` | -              |
 * | _markSize_              | `"sm" \| "default"`                                    | `"sm"`         |
 *
 *
 * ### Example Usage
 *
 * ```tsx
 * <Slider defaultValue={[50]} max={100} step={1} />
 *
 * // With value label
 * <Slider defaultValue={[4]} showValue valueFormatter={(v) => `${v} cores`} />
 *
 * // With step marks
 * <Slider defaultValue={[50]} marks />
 *
 * // With custom marks
 * <Slider defaultValue={[50]} marks={[{ value: 0, label: "0" }, { value: 100, label: "100" }]} />
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 4.1.2 Name, Role, Value — `slider` role via Radix Slider primitive; `aria-label` forwarded to thumbs.
 * @wcag 2.1.1 Keyboard — thumbs focusable and adjusted via arrow keys.
 * @wcag 2.4.7 Focus Visible — thumb focus indicator via `focus-visible:outline-2 focus-visible:outline-offset-2 outline-ring`. Uses CSS `outline` rather than `box-shadow`-based ring so the focus indicator is not extinguished when the composed `box-shadow` chain contains a `none`-valued token (e.g. `--dds-material-surface-shadow` in light/high-contrast).
 */
function Slider({
  className,
  defaultValue,
  value,
  min = 0,
  max = 100,
  step = 1,
  'aria-label': ariaLabel,
  showValue = false,
  valueFormatter = (v: number) => String(v),
  valuePosition = 'top',
  marks,
  markSize = 'sm',
  onValueChange,
  ...props
}: SliderProps) {
  const initialValues = React.useMemo(
    () =>
      Array.isArray(value)
        ? value
        : Array.isArray(defaultValue)
          ? defaultValue
          : [min, max],
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  // Track live values for the value label display
  const [liveValues, setLiveValues] = React.useState<number[]>(initialValues);

  // Sync controlled `value` prop into live state
  React.useEffect(() => {
    if (Array.isArray(value)) {
      setLiveValues(value);
    }
  }, [value]);

  const handleValueChange = React.useCallback(
    (next: number[]) => {
      setLiveValues(next);
      onValueChange?.(next);
    },
    [onValueChange]
  );

  // Compute mark items from the `marks` prop
  const markItems = React.useMemo<MarkItem[]>(() => {
    if (!marks) return [];
    if (Array.isArray(marks)) return marks;
    // marks === true: a dot at every step
    const items: MarkItem[] = [];
    const count = Math.round((max - min) / step);
    for (let i = 0; i <= count; i++) {
      items.push({value: min + i * step});
    }
    return items;
  }, [marks, min, max, step]);

  const pct = (v: number) => ((v - min) / (max - min)) * 100;

  return (
    <div
      className={cn(
        'relative w-full',
        marks && 'pb-5',
        showValue && valuePosition === 'top' && 'pt-7',
        showValue && valuePosition === 'bottom' && 'pb-7'
      )}
    >
      <SliderPrimitive.Root
        data-slot="slider"
        defaultValue={defaultValue}
        value={value}
        min={min}
        max={max}
        step={step}
        aria-label={ariaLabel}
        onValueChange={handleValueChange}
        className={cn(
          'relative flex w-full touch-none items-center select-none data-disabled:opacity-50 data-vertical:h-full data-vertical:min-h-40 data-vertical:w-auto data-vertical:flex-col',
          className
        )}
        {...props}
      >
        <SliderPrimitive.Track
          data-slot="slider-track"
          className="relative grow overflow-hidden rounded-full bg-muted data-horizontal:h-1.5 data-horizontal:w-full data-vertical:h-full data-vertical:w-1.5"
        >
          <SliderPrimitive.Range
            data-slot="slider-range"
            className="absolute bg-primary select-none data-horizontal:h-full data-vertical:w-full"
          />
        </SliderPrimitive.Track>

        {liveValues.map((_, idx) => (
          // eslint-disable-next-line react/no-array-index-key
          <SliderPrimitive.Thumb
            data-slot="slider-thumb"
            key={idx}
            aria-label={ariaLabel}
            className="block size-4 shrink-0 rounded-full border border-primary bg-background shadow-raised ring-ring outline-ring transition-[color,box-shadow,outline] ease-default select-none hover:ring-4 focus-visible:outline-2 focus-visible:outline-offset-2 disabled:pointer-events-none disabled:opacity-50"
          />
        ))}
      </SliderPrimitive.Root>

      {/* Value labels — rendered outside the Radix root so they don't interfere with thumb hit areas */}
      {showValue &&
        liveValues.map((thumbValue, idx) => (
          // eslint-disable-next-line react/no-array-index-key
          <div
            key={idx}
            data-slot="slider-value-label"
            aria-hidden="true"
            className={cn(
              'pointer-events-none absolute z-floating -translate-x-1/2 rounded bg-popover px-1.5 py-0.5 text-xs whitespace-nowrap text-popover-foreground shadow-floating',
              valuePosition === 'top' && 'top-0',
              valuePosition === 'bottom' && 'bottom-0'
            )}
            style={{left: `${pct(thumbValue)}%`}}
          >
            {valueFormatter(thumbValue)}
          </div>
        ))}

      {/* Mark dots and labels */}
      {markItems.length > 0 && (
        <div
          aria-hidden="true"
          data-slot="slider-marks"
          className="pointer-events-none absolute inset-x-0 bottom-0 flex"
        >
          {markItems.map(mark => (
            <div
              key={mark.value}
              className="absolute -translate-x-1/2"
              style={{left: `${pct(mark.value)}%`}}
            >
              <div
                data-slot="slider-mark"
                className={cn(
                  'mx-auto rounded-full bg-muted-foreground-track',
                  markSize === 'sm' ? 'size-1' : 'size-1.5'
                )}
              />
              {mark.label !== undefined && (
                <span
                  data-slot="slider-mark-label"
                  className="mt-1 block text-center text-2xs leading-none text-muted-foreground"
                >
                  {mark.label}
                </span>
              )}
            </div>
          ))}
        </div>
      )}
    </div>
  );
}

export {Slider};
export type {SliderProps, MarkItem};
