// @docker/trident/src/components/ui/sonner.tsx
import {cssVars} from '@/lib/utils';

import {
  AlertTriangle,
  CircleCheck,
  Info,
  Loader2Icon,
  OctagonX,
} from 'lucide-react';
import {useTheme} from 'next-themes';
import {useEffect, useState} from 'react';
import {Toaster as Sonner, type ToasterProps} from 'sonner';

const validThemes = ['dark', 'light', 'system'] as const;

// Module-level pub/sub so announceError() can drive any mounted ToastAnnouncer
type ErrorListener = (message: string) => void;
const errorListeners = new Set<ErrorListener>();

/**
 * Triggers an assertive (`role="alert"`) screen-reader announcement for urgent
 * error messages. Use this instead of `toast.error()` when the error must
 * interrupt the user immediately (WCAG 4.1.3 — Status Messages).
 *
 * Requires `<ToastAnnouncer />` to be mounted somewhere in the tree.
 * Calls that arrive while no announcer is mounted are silently dropped.
 *
 * ```tsx
 * announceError('Upload failed — disk full');
 * ```
 */
function announceError(message: string): void {
  for (const fn of errorListeners) fn(message);
}

/**
 * Renders a visually hidden `role="alert"` live region that receives urgent
 * error announcements from `announceError()`. Mount once near `<Toaster />`
 * in your application root.
 *
 * ```tsx
 * <Toaster />
 * <ToastAnnouncer />
 * ```
 */
function ToastAnnouncer() {
  const [message, setMessage] = useState('');

  useEffect(() => {
    const handler = (msg: string) => {
      // Clear first so repeated identical messages trigger a fresh announcement
      setMessage('');
      requestAnimationFrame(() => setMessage(msg));
    };
    errorListeners.add(handler);
    return () => {
      errorListeners.delete(handler);
    };
  }, []);

  return (
    <div
      role="alert"
      aria-live="assertive"
      aria-atomic="true"
      data-slot="toast-announcer"
      className="sr-only"
    >
      {message}
    </div>
  );
}

/**
 * A themed toast notification container powered by Sonner. Renders toast notifications
 * with custom icons and theme-aware styling.
 *
 * ### Component Props
 *
 * See [Sonner props](https://sonner.emilkowal.dev/) documentation for full details.
 *
 * | Prop             | Type                                  | Default    |
 * | :--------------- | :------------------------------------ | :--------- |
 * | _theme_          | `"light" \| "dark" \| "system"`       | `"system"` |
 * | _richColors_     | `boolean`                             | `false`    |
 * | _expand_         | `boolean`                             | `false`    |
 * | _duration_       | `number`                              | `4000`     |
 * | _position_       | `"top-left" \| "top-right" \| "bottom-left" \| "bottom-right" \| "top-center" \| "bottom-center"` | `"bottom-right"` |
 * | _closeButton_    | `boolean`                             | `false`    |
 * | _offset_         | `number \| string`                    | `"32px"`   |
 * | _dir_            | `"ltr" \| "rtl" \| "auto"`            | `"auto"`   |
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * Sonner renders a single `aria-live="polite"` container for all toast types
 * (neutral, success, info, warning). For urgent errors that must interrupt the
 * user immediately, call `announceError()` — it updates a `role="alert"` live
 * region provided by `<ToastAnnouncer />`.
 *
 * @wcag 4.1.3 Status Messages — polite: Sonner; assertive: `announceError()` +
 * `<ToastAnnouncer />`.
 *
 * ### Example Usage
 *
 * ```tsx
 * import { toast } from "sonner";
 * import { Toaster, ToastAnnouncer, announceError } from "@/components/ui/sonner";
 *
 * // In your app root:
 * <Toaster />
 * <ToastAnnouncer />
 *
 * // Polite announcements (neutral, success, info, warning):
 * toast("Event has been created");
 * toast.success("Success!");
 *
 * // Assertive announcement (urgent errors only):
 * announceError("Upload failed — disk full");
 * ```
 */
const Toaster = ({...props}: ToasterProps) => {
  const {theme = 'system'} = useTheme();
  const resolvedTheme = validThemes.find(t => t === theme) ?? 'system';

  return (
    <Sonner
      theme={resolvedTheme}
      className="toaster group"
      icons={{
        success: <CircleCheck className="size-4" />,
        info: <Info className="size-4" />,
        warning: <AlertTriangle className="size-4" />,
        error: <OctagonX className="size-4" />,
        loading: <Loader2Icon className="size-4 animate-spin" />,
      }}
      style={cssVars({
        '--normal-bg': 'var(--popover)',
        '--normal-text': 'var(--popover-foreground)',
        '--normal-border': 'var(--border)',
        '--border-radius': 'var(--radius)',
      })}
      toastOptions={{
        classNames: {
          toast: 'cn-toast',
        },
      }}
      {...props}
    />
  );
};

// Re-export toast so consumers never need a direct `sonner` dependency.
// pnpm strict isolation blocks transitive imports even when TypeScript resolves them.
export {toast} from 'sonner';
export {Toaster, ToastAnnouncer, announceError};
