// @docker/trident/src/components/ui/tabs.tsx
'use client';

import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {Tabs as TabsPrimitive} from 'radix-ui';

/**
 * A set of layered sections of content -- known as tab panels -- that are displayed one at a time.
 *
 * ### Component Props
 *
 * See [Tabs props](https://www.radix-ui.com/primitives/docs/components/tabs) documentation for full details.
 *
 * | Prop              | Type                          | Default          |
 * | :---------------- | :---------------------------- | :--------------- |
 * | _asChild_         | `boolean`                     | `false`          |
 * | _defaultValue_    | `string`                      | -                |
 * | _value_           | `string`                      | -                |
 * | _onValueChange_   | `(value: string) => void`     | -                |
 * | _orientation_     | `"horizontal" \| "vertical"`  | `"horizontal"`   |
 * | _dir_             | `"ltr" \| "rtl"`              | -                |
 * | _activationMode_  | `"automatic" \| "manual"`     | `"automatic"`    |
 *
 *
 * ### Example Usage
 *
 * **Default**
 * ```tsx
 * <Tabs defaultValue="tab1">
 *   <TabsList>
 *     <TabsTrigger value="tab1">Tab 1</TabsTrigger>
 *     <TabsTrigger value="tab2">Tab 2</TabsTrigger>
 *   </TabsList>
 *   <TabsContent value="tab1">Content for tab 1</TabsContent>
 *   <TabsContent value="tab2">Content for tab 2</TabsContent>
 * </Tabs>
 * ```
 *
 * **Line**
 * ```tsx
 * <Tabs defaultValue="tab1">
 *   <TabsList variant="line">
 *     <TabsTrigger value="tab1">Tab 1</TabsTrigger>
 *     <TabsTrigger value="tab2">Tab 2</TabsTrigger>
 *   </TabsList>
 *   <TabsContent value="tab1">Content for tab 1</TabsContent>
 *   <TabsContent value="tab2">Content for tab 2</TabsContent>
 * </Tabs>
 * ```
 *
 * **Vertical**
 * ```tsx
 * <Tabs defaultValue="tab1" orientation="vertical">
 *   <TabsList>
 *     <TabsTrigger value="tab1">Tab 1</TabsTrigger>
 *     <TabsTrigger value="tab2">Tab 2</TabsTrigger>
 *   </TabsList>
 *   <TabsContent value="tab1">Content for tab 1</TabsContent>
 *   <TabsContent value="tab2">Content for tab 2</TabsContent>
 * </Tabs>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 4.1.2 Name, Role, Value — `tablist`, `tab`, and `tabpanel` roles via Radix Tabs primitives.
 * @wcag 2.1.1 Keyboard — arrow keys move between tabs; Space/Enter activates.
 * @wcag 2.4.7 Focus Visible — focus ring on triggers via `focus-visible:ring-3`.
 */
function Tabs({
  className,
  orientation = 'horizontal',
  ...props
}: React.ComponentProps<typeof TabsPrimitive.Root>) {
  return (
    <TabsPrimitive.Root
      data-slot="tabs"
      orientation={orientation}
      className={cn(
        'group/tabs flex gap-2 data-[orientation=horizontal]:flex-col',
        className
      )}
      {...props}
    />
  );
}

const tabsListVariants = cva(
  'group/tabs-list inline-flex w-fit items-center justify-center rounded p-[var(--dds-spacing-tabs-inner)] text-muted-foreground group-data-[orientation=vertical]/tabs:h-fit group-data-[orientation=vertical]/tabs:flex-col group-data-horizontal/tabs:h-9 data-[variant=line]:rounded-none',
  {
    variants: {
      variant: {
        default: 'bg-muted',
        line: 'gap-1 bg-transparent',
      },
    },
    defaultVariants: {
      variant: 'default',
    },
  }
);

/**
 * The container for tab triggers, providing visual grouping and keyboard navigation.
 *
 * ### Component Props
 *
 * See [Tabs.List props](https://www.radix-ui.com/primitives/docs/components/tabs) documentation for full details.
 *
 * | Prop       | Type                       | Default     |
 * | :--------- | :------------------------- | :---------- |
 * | _asChild_  | `boolean`                  | `false`     |
 * | _loop_     | `boolean`                  | `true`      |
 * | _variant_  | `"default" \| "line"`      | `"default"` |
 */
function TabsList({
  className,
  variant = 'default',
  ...props
}: React.ComponentProps<typeof TabsPrimitive.List> &
  VariantProps<typeof tabsListVariants>) {
  return (
    <TabsPrimitive.List
      data-slot="tabs-list"
      data-variant={variant}
      className={cn(tabsListVariants({variant}), className)}
      {...props}
    />
  );
}

/**
 * A button that activates its associated TabsContent panel.
 *
 * ### Component Props
 *
 * See [Tabs.Trigger props](https://www.radix-ui.com/primitives/docs/components/tabs) documentation for full details.
 *
 * | Prop                  | Type       | Default |
 * | :-------------------- | :--------- | :------ |
 * | _asChild_             | `boolean`  | `false` |
 * | _value_ *(required)*  | `string`   | -       |
 * | _disabled_            | `boolean`  | `false` |
 */
function TabsTrigger({
  className,
  ...props
}: React.ComponentProps<typeof TabsPrimitive.Trigger>) {
  return (
    <TabsPrimitive.Trigger
      data-slot="tabs-trigger"
      className={cn(
        "relative inline-flex h-[calc(100%-var(--dds-height-tab-indicator-inset))] flex-1 items-center justify-center gap-1.5 rounded border border-transparent px-2 py-1 text-sm font-label tracking-label whitespace-nowrap text-muted-foreground transition-all ease-default group-data-[orientation=vertical]/tabs:w-full group-data-[orientation=vertical]/tabs:justify-start hover:text-foreground focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-1 focus-visible:outline-ring disabled:pointer-events-none disabled:opacity-50 group-data-[variant=default]/tabs-list:data-active:shadow-raised group-data-[variant=line]/tabs-list:data-active:shadow-none [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        'group-data-[variant=line]/tabs-list:bg-transparent group-data-[variant=line]/tabs-list:data-active:bg-transparent dark:group-data-[variant=line]/tabs-list:data-active:border-transparent dark:group-data-[variant=line]/tabs-list:data-active:bg-transparent',
        'data-active:bg-background data-active:text-foreground dark:data-active:border-input dark:data-active:bg-input-bg dark:data-active:text-foreground',
        'after:absolute after:bg-foreground after:opacity-0 after:transition-opacity group-data-[orientation=horizontal]/tabs:after:inset-x-0 group-data-[orientation=horizontal]/tabs:after:-bottom-1.25 group-data-[orientation=horizontal]/tabs:after:h-0.5 group-data-[orientation=vertical]/tabs:after:inset-y-0 group-data-[orientation=vertical]/tabs:after:-right-1 group-data-[orientation=vertical]/tabs:after:w-0.5 group-data-[variant=line]/tabs-list:data-active:after:opacity-100',
        className
      )}
      {...props}
    />
  );
}

/**
 * The content panel associated with a TabsTrigger, displayed when its trigger is active.
 *
 * ### Component Props
 *
 * See [Tabs.Content props](https://www.radix-ui.com/primitives/docs/components/tabs) documentation for full details.
 *
 * | Prop                 | Type       | Default |
 * | :------------------- | :--------- | :------ |
 * | _asChild_            | `boolean`  | `false` |
 * | _value_ *(required)* | `string`   | -       |
 * | _forceMount_         | `boolean`  | -       |
 */
function TabsContent({
  className,
  ...props
}: React.ComponentProps<typeof TabsPrimitive.Content>) {
  return (
    <TabsPrimitive.Content
      data-slot="tabs-content"
      className={cn('flex-1 text-sm outline-none', className)}
      {...props}
    />
  );
}

export {Tabs, TabsList, TabsTrigger, TabsContent, tabsListVariants};
