// @docker/trident/src/components/ui/text-field.tsx
'use client';

import {
  Field,
  FieldContent,
  FieldDescription,
  FieldError,
  FieldLabel,
} from '@/components/ui/field';
import {Input, type InputProps} from '@/components/ui/input';

import {useId} from 'react';

type TextFieldProps = InputProps & {
  /** The visible label rendered above the input. */
  label: React.ReactNode;
  /** Optional helper text rendered below the input. */
  description?: React.ReactNode;
  /** Shorthand for a single error message. Use `errors` for multiple. */
  error?: string;
  /** Array of validation errors — renders a list when more than one message is present. */
  errors?: Array<{message?: string} | undefined>;
  /** Layout orientation of the label/input pair. Defaults to `"vertical"`. */
  orientation?: 'vertical' | 'horizontal' | 'responsive';
};

/**
 * A single-line text field that composes `Field`, `FieldLabel`, `Input`,
 * `FieldDescription`, and `FieldError` into one ergonomic component.
 *
 * ### Example Usage
 *
 * **Basic**
 * ```tsx
 * <TextField label="Email" type="email" placeholder="you@example.com" />
 * ```
 *
 * **With description and error**
 * ```tsx
 * <TextField
 *   label="Username"
 *   description="Must be 3–20 characters."
 *   error="Username is already taken."
 * />
 * ```
 *
 * **Multiple validation errors**
 * ```tsx
 * <TextField
 *   label="Password"
 *   type="password"
 *   errors={[{message: 'Too short'}, {message: 'Must include a number'}]}
 * />
 * ```
 *
 * **Horizontal layout**
 * ```tsx
 * <TextField label="Display name" orientation="horizontal" />
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 1.3.1 Info and Relationships — label wired to input via htmlFor/id.
 * @wcag 3.3.1 Error Identification — errors rendered with role="alert".
 * @wcag 3.3.2 Labels or Instructions — description wired via aria-describedby.
 * @wcag 4.1.2 Name, Role, Value — aria-invalid set when errors are present.
 */
function TextField({
  id: idProp,
  label,
  description,
  error,
  errors,
  orientation,
  ...inputProps
}: TextFieldProps) {
  const autoId = useId();
  const id = idProp ?? autoId;

  const normalizedErrors = [
    ...(error ? [{message: error}] : []),
    ...(errors ?? []),
  ];
  const hasError = normalizedErrors.some(e => e?.message);

  const descriptionId = description ? `${id}-desc` : undefined;
  const errorId = hasError ? `${id}-error` : undefined;
  const describedBy =
    [descriptionId, errorId].filter(Boolean).join(' ') || undefined;

  return (
    <Field orientation={orientation}>
      <FieldLabel htmlFor={id}>{label}</FieldLabel>
      <FieldContent>
        <Input
          id={id}
          aria-describedby={describedBy}
          aria-invalid={hasError || undefined}
          {...inputProps}
        />
        {description && (
          <FieldDescription id={descriptionId}>{description}</FieldDescription>
        )}
        {hasError && <FieldError id={errorId} errors={normalizedErrors} />}
      </FieldContent>
    </Field>
  );
}

export {TextField};
export type {TextFieldProps};
