// @docker/trident/src/components/ui/theme-switcher.tsx
'use client';

import {Button} from '@/components/ui/button';
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger,
} from '@/components/ui/dropdown-menu';
import {useTheme} from '@/hooks/use-theme';
import {cn} from '@/lib/utils';

import {Monitor, Moon, Sun} from 'lucide-react';

type Theme = 'light' | 'dark';
type ThemeWithSystem = Theme | 'system';

type ThemeSwitcherBaseProps = {
  className?: string;
};

type ThemeSwitcherControlledProps = ThemeSwitcherBaseProps & {
  /** Omit to use next-themes ThemeProvider for state management. */
  systemPreference?: false;
  theme: Theme;
  onThemeChange: (theme: Theme) => void;
};

type ThemeSwitcherSystemProps = ThemeSwitcherBaseProps & {
  /**
   * When `true`, enables a third "System" mode in addition to light and dark.
   * Renders a dropdown with all three options.
   *
   * Requires `ThemeProvider` from `next-themes` at the app root:
   * ```tsx
   * import {ThemeProvider} from 'next-themes';
   *
   * <ThemeProvider attribute="class" defaultTheme="system" enableSystem>
   *   {children}
   * </ThemeProvider>
   * ```
   */
  systemPreference: true;
  theme?: never;
  onThemeChange?: never;
};

type ThemeSwitcherProps =
  | ThemeSwitcherControlledProps
  | ThemeSwitcherSystemProps;

const THEME_ICONS: Record<ThemeWithSystem, React.ReactNode> = {
  light: (
    <Sun
      data-testid="theme-switcher-icon"
      data-theme="light"
      className="h-4 w-4"
    />
  ),
  dark: (
    <Moon
      data-testid="theme-switcher-icon"
      data-theme="dark"
      className="h-4 w-4"
    />
  ),
  system: (
    <Monitor
      data-testid="theme-switcher-icon"
      data-theme="system"
      className="h-4 w-4"
    />
  ),
};

const THEME_LABELS: Record<ThemeWithSystem, string> = {
  light: 'Light',
  dark: 'Dark',
  system: 'System',
};

/**
 * A button that cycles between light and dark themes, or optionally shows a
 * dropdown with light, dark, and system options.
 *
 * ### Controlled (no next-themes)
 * Pass `theme` and `onThemeChange` to manage state yourself.
 *
 * ### Uncontrolled with system preference
 * Set `systemPreference={true}` to wire directly to `next-themes`.
 * Requires `ThemeProvider` from `next-themes` at the app root:
 * ```tsx
 * import {ThemeProvider} from 'next-themes';
 *
 * <ThemeProvider attribute="class" defaultTheme="system" enableSystem>
 *   {children}
 * </ThemeProvider>
 * ```
 */
const ThemeSwitcher = (props: ThemeSwitcherProps) => {
  if (props.systemPreference === true) {
    return <ThemeSwitcherSystem className={props.className} />;
  }

  const {theme, onThemeChange, className} = props;
  const isDark = theme === 'dark';

  return (
    <Button
      variant="ghost"
      size="icon"
      aria-label={isDark ? 'Switch to light mode' : 'Switch to dark mode'}
      onClick={() => onThemeChange(isDark ? 'light' : 'dark')}
      className={cn(className)}
    >
      {isDark ? (
        <Sun
          data-testid="theme-switcher-icon"
          data-theme="dark"
          className="h-4 w-4"
        />
      ) : (
        <Moon
          data-testid="theme-switcher-icon"
          data-theme="light"
          className="h-4 w-4"
        />
      )}
    </Button>
  );
};

/** Inner component for system-preference mode — requires next-themes context. */
const ThemeSwitcherSystem = ({className}: {className?: string}) => {
  const {theme, setTheme} = useTheme();

  const currentTheme: ThemeWithSystem =
    theme === 'light' || theme === 'dark' || theme === 'system'
      ? theme
      : 'system';

  return (
    <DropdownMenu>
      <DropdownMenuTrigger asChild>
        <Button
          variant="ghost"
          size="icon"
          aria-label={`Theme: ${THEME_LABELS[currentTheme]}. Click to change`}
          className={cn(className)}
        >
          {THEME_ICONS[currentTheme]}
        </Button>
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end">
        {(['light', 'dark', 'system'] as const).map(mode => (
          <DropdownMenuItem
            key={mode}
            onSelect={() => setTheme(mode)}
            aria-current={currentTheme === mode ? 'true' : undefined}
            className={cn(
              'flex items-center gap-2',
              currentTheme === mode && 'font-label'
            )}
          >
            {THEME_ICONS[mode]}
            {THEME_LABELS[mode]}
          </DropdownMenuItem>
        ))}
      </DropdownMenuContent>
    </DropdownMenu>
  );
};

export {ThemeSwitcher};
export type {ThemeSwitcherProps, Theme, ThemeWithSystem};
