// @docker/trident/src/components/ui/toggle-group.tsx
'use client';

import type {VariantProps} from 'class-variance-authority';

import {toggleVariants} from '@/components/ui/toggle';
import {cn, cssVars} from '@/lib/utils';

import {ToggleGroup as ToggleGroupPrimitive} from 'radix-ui';
import * as React from 'react';

const ToggleGroupContext = React.createContext<
  VariantProps<typeof toggleVariants> & {
    spacing?: number;
    orientation?: 'horizontal' | 'vertical';
  }
>({
  size: 'default',
  variant: 'default',
  spacing: 0,
  orientation: 'horizontal',
});

/**
 * A set of two-state buttons that can be toggled on or off.
 *
 * ### Component Props
 *
 * See [Toggle Group props](https://www.radix-ui.com/primitives/docs/components/toggle-group) documentation for full details.
 *
 * | Prop                | Type                              | Default          |
 * | :------------------ | :-------------------------------- | :--------------- |
 * | _asChild_           | `boolean`                         | `false`          |
 * | _type_ *(required)* | `"single" \| "multiple"`          | -                |
 * | _value_             | `string \| string[]`              | -                |
 * | _defaultValue_      | `string \| string[]`              | -                |
 * | _onValueChange_     | `(value: string \| string[]) => void` | -            |
 * | _disabled_          | `boolean`                         | `false`          |
 * | _rovingFocus_       | `boolean`                         | `true`           |
 * | _orientation_       | `"horizontal" \| "vertical"`      | `"horizontal"`   |
 * | _dir_               | `"ltr" \| "rtl"`                  | -                |
 * | _loop_              | `boolean`                         | `true`           |
 * | _variant_           | `"default" \| "outline"`          | `"default"`      |
 * | _size_              | `"default" \| "sm" \| "lg"`       | `"default"`      |
 * | _spacing_           | `number`                          | `0`              |
 *
 *
 * ### Example Usage
 *
 * **Default**
 * ```tsx
 * <ToggleGroup type="single" defaultValue="center">
 *   <ToggleGroupItem value="left" aria-label="Align left">
 *     <AlignLeft className="size-4" />
 *   </ToggleGroupItem>
 *   <ToggleGroupItem value="center" aria-label="Align center">
 *     <AlignCenter className="size-4" />
 *   </ToggleGroupItem>
 * </ToggleGroup>
 * ```
 *
 * **Outline**
 * ```tsx
 * <ToggleGroup type="single" variant="outline" defaultValue="center">
 *   <ToggleGroupItem value="left" aria-label="Align left">
 *     <AlignLeft className="size-4" />
 *   </ToggleGroupItem>
 *   <ToggleGroupItem value="center" aria-label="Align center">
 *     <AlignCenter className="size-4" />
 *   </ToggleGroupItem>
 * </ToggleGroup>
 * ```
 *
 * **Small + Spacing**
 * ```tsx
 * <ToggleGroup type="multiple" size="sm" spacing={2}>
 *   <ToggleGroupItem value="bold" aria-label="Bold">
 *     <Bold className="size-4" />
 *   </ToggleGroupItem>
 *   <ToggleGroupItem value="italic" aria-label="Italic">
 *     <Italic className="size-4" />
 *   </ToggleGroupItem>
 * </ToggleGroup>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 4.1.2 Name, Role, Value — `group` role with `aria-pressed` on items via Radix ToggleGroup primitive.
 * @wcag 2.1.1 Keyboard — roving focus with arrow keys between items; Enter/Space toggles.
 */
function ToggleGroup({
  className,
  variant,
  size,
  spacing = 0,
  orientation = 'horizontal',
  children,
  ...props
}: React.ComponentProps<typeof ToggleGroupPrimitive.Root> &
  VariantProps<typeof toggleVariants> & {
    spacing?: number;
    orientation?: 'horizontal' | 'vertical';
  }) {
  const contextValue = React.useMemo(
    () => ({variant, size, spacing, orientation}),
    [variant, size, spacing, orientation]
  );

  return (
    <ToggleGroupPrimitive.Root
      data-slot="toggle-group"
      data-variant={variant}
      data-size={size}
      data-spacing={spacing}
      data-orientation={orientation}
      style={cssVars({'--gap': spacing})}
      className={cn(
        'group/toggle-group flex w-fit flex-row items-center gap-[--spacing(var(--gap))] rounded data-[orientation=vertical]:flex-col data-[orientation=vertical]:items-stretch data-[spacing=0]:data-[variant=outline]:shadow-raised',
        className
      )}
      {...props}
    >
      <ToggleGroupContext.Provider value={contextValue}>
        {children}
      </ToggleGroupContext.Provider>
    </ToggleGroupPrimitive.Root>
  );
}

/**
 * An individual toggle button within a ToggleGroup.
 *
 * ### Component Props
 *
 * See [ToggleGroup.Item props](https://www.radix-ui.com/primitives/docs/components/toggle-group) documentation for full details.
 *
 * | Prop                  | Type                         | Default     |
 * | :-------------------- | :--------------------------- | :---------- |
 * | _asChild_             | `boolean`                    | `false`     |
 * | _value_ *(required)*  | `string`                     | -           |
 * | _disabled_            | `boolean`                    | -           |
 * | _variant_             | `"default" \| "outline"`     | `"default"` |
 * | _size_                | `"default" \| "sm" \| "lg"`  | `"default"` |
 */
function ToggleGroupItem({
  className,
  children,
  variant = 'default',
  size = 'default',
  ...props
}: React.ComponentProps<typeof ToggleGroupPrimitive.Item> &
  VariantProps<typeof toggleVariants>) {
  const context = React.useContext(ToggleGroupContext);

  return (
    <ToggleGroupPrimitive.Item
      data-slot="toggle-group-item"
      data-variant={context.variant || variant}
      data-size={context.size || size}
      data-spacing={context.spacing}
      className={cn(
        'shrink-0 group-data-[spacing=0]/toggle-group:rounded-none group-data-[spacing=0]/toggle-group:px-2 group-data-[spacing=0]/toggle-group:shadow-none focus:z-raised focus-visible:z-raised group-data-horizontal/toggle-group:data-[spacing=0]:first:rounded-l group-data-vertical/toggle-group:data-[spacing=0]:first:rounded-t group-data-horizontal/toggle-group:data-[spacing=0]:last:rounded-r group-data-vertical/toggle-group:data-[spacing=0]:last:rounded-b data-[state=on]:bg-muted group-data-horizontal/toggle-group:data-[spacing=0]:data-[variant=outline]:border-l-0 group-data-vertical/toggle-group:data-[spacing=0]:data-[variant=outline]:border-t-0 group-data-horizontal/toggle-group:data-[spacing=0]:data-[variant=outline]:first:border-l group-data-vertical/toggle-group:data-[spacing=0]:data-[variant=outline]:first:border-t group-data-vertical/toggle-group:data-[spacing=0]:data-[variant=outline]:first:border-border',
        toggleVariants({
          variant: context.variant || variant,
          size: context.size || size,
        }),
        className
      )}
      {...props}
    >
      {children}
    </ToggleGroupPrimitive.Item>
  );
}

export {ToggleGroup, ToggleGroupItem};
