// @docker/trident/src/components/ui/toggle.tsx
import type * as React from 'react';

import {cn} from '@/lib/utils';

import {cva, type VariantProps} from 'class-variance-authority';
import {Toggle as TogglePrimitive} from 'radix-ui';

const toggleVariants = cva(
  "group/toggle inline-flex items-center justify-center gap-1 rounded text-sm font-label whitespace-nowrap transition-[color,box-shadow] ease-default outline-none hover:bg-muted hover:text-foreground focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 aria-invalid:border-destructive aria-invalid:ring-destructive/20 aria-pressed:bg-muted dark:aria-invalid:ring-destructive/40 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
  {
    variants: {
      variant: {
        default: 'bg-transparent',
        outline:
          'border border-input bg-transparent shadow-raised hover:bg-muted',
      },
      size: {
        default: 'h-9 min-w-9 px-2',
        sm: 'h-8 min-w-8 px-1.5',
        lg: 'h-10 min-w-10 px-2.5',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

/**
 * A two-state button that can be either on or off.
 *
 * ### Component Props
 *
 * See [Toggle props](https://www.radix-ui.com/primitives/docs/components/toggle) documentation for full details.
 *
 * | Prop               | Type                              | Default     |
 * | :----------------- | :-------------------------------- | :---------- |
 * | _asChild_          | `boolean`                         | `false`     |
 * | _defaultPressed_   | `boolean`                         | -           |
 * | _pressed_          | `boolean`                         | -           |
 * | _onPressedChange_  | `(pressed: boolean) => void`      | -           |
 * | _disabled_         | `boolean`                         | -           |
 * | _variant_          | `"default" \| "outline"`          | `"default"` |
 * | _size_             | `"default" \| "sm" \| "lg"`       | `"default"` |
 *
 *
 * ### Example Usage
 *
 * **Default**
 * ```tsx
 * <Toggle aria-label="Toggle bold">
 *   <Bold className="size-4" />
 * </Toggle>
 * ```
 *
 * **Outline**
 * ```tsx
 * <Toggle variant="outline" pressed={isBold} onPressedChange={setIsBold}>
 *   <Bold className="size-4" />
 * </Toggle>
 * ```
 *
 * **Large**
 * ```tsx
 * <Toggle size="lg" aria-label="Toggle italic">
 *   <Italic className="size-4" />
 * </Toggle>
 * ```
 *
 * ### Accessibility (WCAG 2.2 AA)
 *
 * @wcag 4.1.2 Name, Role, Value ŌĆö `button` role with `aria-pressed` state via Radix Toggle primitive.
 * @wcag 2.1.1 Keyboard ŌĆö focusable and toggled via Enter/Space.
 * @wcag 2.4.7 Focus Visible ŌĆö focus ring applied via `focus-visible:ring-3`.
 */
function Toggle({
  className,
  variant = 'default',
  size = 'default',
  ...props
}: React.ComponentProps<typeof TogglePrimitive.Root> &
  VariantProps<typeof toggleVariants>) {
  return (
    <TogglePrimitive.Root
      data-slot="toggle"
      className={cn(toggleVariants({variant, size, className}))}
      {...props}
    />
  );
}

export {Toggle, toggleVariants};
