// @docker/trident/dist/components/code-block/code-block-server.js
// src/components/code-block/code-block-server.tsx
import {
  transformerNotationDiff,
  transformerNotationErrorLevel,
  transformerNotationFocus,
  transformerNotationHighlight,
  transformerNotationWordHighlight
} from "@shikijs/transformers";
import {
  codeToHtml
} from "shiki";
import { jsx } from "react/jsx-runtime";
var CodeBlockContent = async ({
  children,
  themes,
  language,
  syntaxHighlighting = true,
  ...props
}) => {
  if (!syntaxHighlighting) {
    return /* @__PURE__ */ jsx("div", { ...props, children: /* @__PURE__ */ jsx("pre", { children: /* @__PURE__ */ jsx("code", { children }) }) });
  }
  const html = await codeToHtml(children, {
    lang: language ?? "typescript",
    themes: themes ?? {
      light: "github-light",
      dark: "github-dark-default"
    },
    transformers: [
      transformerNotationDiff({
        matchAlgorithm: "v3"
      }),
      transformerNotationHighlight({
        matchAlgorithm: "v3"
      }),
      transformerNotationWordHighlight({
        matchAlgorithm: "v3"
      }),
      transformerNotationFocus({
        matchAlgorithm: "v3"
      }),
      transformerNotationErrorLevel({
        matchAlgorithm: "v3"
      })
    ]
  });
  return /* @__PURE__ */ jsx("div", { dangerouslySetInnerHTML: { __html: html }, ...props });
};
export {
  CodeBlockContent
};
