// @docker/trident/dist/chunk-MTSDKPQJ.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/rating.tsx
import { Star } from "lucide-react";
import { useCallback, useId, useMemo, useRef, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var sizeClasses = {
  sm: "size-4",
  default: "size-5",
  lg: "size-7"
};
var gapClasses = {
  sm: "gap-0.5",
  default: "gap-1",
  lg: "gap-1.5"
};
function Rating({
  value = 0,
  onChange,
  max = 5,
  readOnly = false,
  size = "default",
  precision = 1,
  className,
  "aria-label": ariaLabel = "Rating"
}) {
  const [hoverValue, setHoverValue] = useState(null);
  const containerRef = useRef(null);
  const id = useId();
  const displayValue = hoverValue ?? value;
  const roundToStep = useCallback(
    (v) => {
      const clamped = Math.max(0, Math.min(v, max));
      return Math.round(clamped / precision) * precision;
    },
    [max, precision]
  );
  const resolveStarFromEvent = useCallback(
    (event) => {
      const target = event.target instanceof Element ? event.target.closest("[data-star-index]") : null;
      if (!target) return null;
      const starIndex = Number(target.getAttribute("data-star-index"));
      if (precision === 0.5) {
        const rect = target.getBoundingClientRect();
        const isLeftHalf = event.clientX - rect.left < rect.width / 2;
        return isLeftHalf ? starIndex + 0.5 : starIndex + 1;
      }
      return starIndex + 1;
    },
    [precision]
  );
  const handleClick = useCallback(
    (event) => {
      if (readOnly) return;
      const newValue = resolveStarFromEvent(event);
      if (newValue !== null) onChange?.(newValue === value ? 0 : newValue);
    },
    [readOnly, resolveStarFromEvent, onChange, value]
  );
  const handleMouseMove = useCallback(
    (event) => {
      if (readOnly) return;
      const newValue = resolveStarFromEvent(event);
      if (newValue !== null) setHoverValue(newValue);
    },
    [readOnly, resolveStarFromEvent]
  );
  const handleMouseLeave = useCallback(() => {
    if (!readOnly) {
      setHoverValue(null);
    }
  }, [readOnly]);
  const handleKeyDown = useCallback(
    (event) => {
      if (readOnly) return;
      let newValue = value;
      switch (event.key) {
        case "ArrowRight":
        case "ArrowUp":
          event.preventDefault();
          newValue = roundToStep(Math.min(value + precision, max));
          break;
        case "ArrowLeft":
        case "ArrowDown":
          event.preventDefault();
          newValue = roundToStep(Math.max(value - precision, 0));
          break;
        case "Home":
          event.preventDefault();
          newValue = 0;
          break;
        case "End":
          event.preventDefault();
          newValue = max;
          break;
        default:
          return;
      }
      onChange?.(newValue);
    },
    [readOnly, value, precision, max, roundToStep, onChange]
  );
  const getStarFill = (starIndex) => {
    const starPosition = starIndex + 1;
    if (displayValue >= starPosition) return "full";
    if (displayValue >= starPosition - 0.5 && precision === 0.5) return "half";
    return "empty";
  };
  const interactive = !readOnly;
  const starValues = useMemo(
    () => Array.from({ length: max }, (_, i) => i),
    [max]
  );
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: containerRef,
      "data-slot": "rating",
      ...readOnly ? {
        role: "img",
        "aria-label": `${ariaLabel}: ${value} out of ${max} stars`,
        tabIndex: -1
      } : {
        role: "slider",
        "aria-label": ariaLabel,
        "aria-valuemin": 0,
        "aria-valuemax": max,
        "aria-valuenow": value,
        "aria-valuetext": `${value} out of ${max} stars`,
        tabIndex: 0,
        onClick: handleClick,
        onMouseMove: handleMouseMove,
        onMouseLeave: handleMouseLeave,
        onKeyDown: handleKeyDown
      },
      className: cn(
        "inline-flex items-center",
        gapClasses[size],
        interactive && "cursor-pointer rounded outline-none focus-visible:ring-3 focus-visible:ring-ring",
        readOnly && "pointer-events-none",
        className
      ),
      children: starValues.map((i) => /* @__PURE__ */ jsx(
        RatingStar,
        {
          index: i,
          fill: getStarFill(i),
          size
        },
        `${id}-star-${i}`
      ))
    }
  );
}
function RatingStar({ index, fill, size }) {
  const iconClass = sizeClasses[size];
  if (fill === "half") {
    return /* @__PURE__ */ jsxs(
      "span",
      {
        "data-star-index": index,
        role: "presentation",
        className: "relative inline-flex shrink-0",
        children: [
          /* @__PURE__ */ jsx(
            Star,
            {
              className: cn(iconClass, "text-muted-foreground"),
              fill: "none",
              strokeWidth: 1.5
            }
          ),
          /* @__PURE__ */ jsx(
            Star,
            {
              className: cn(iconClass, "absolute inset-0 text-warning"),
              fill: "currentColor",
              strokeWidth: 1.5,
              style: { clipPath: "inset(0 50% 0 0)" }
            }
          )
        ]
      }
    );
  }
  return /* @__PURE__ */ jsx(
    "span",
    {
      "data-star-index": index,
      role: "presentation",
      className: "inline-flex shrink-0",
      children: /* @__PURE__ */ jsx(
        Star,
        {
          className: cn(
            iconClass,
            fill === "full" ? "text-warning" : "text-muted-foreground"
          ),
          fill: fill === "full" ? "currentColor" : "none",
          strokeWidth: 1.5
        }
      )
    }
  );
}

export {
  Rating
};
