// @docker/trident/dist/chunk-JY5DMKQ2.js
import {
  ScrollArea
} from "./chunk-DIP2JZ4Q.js";
import {
  Search
} from "./chunk-TLIJZ6UL.js";
import {
  Checkbox
} from "./chunk-R5XEJKHJ.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/transfer-list-panel.tsx
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var TransferListPanel = ({
  label,
  description,
  items,
  checkedIds,
  onCheckedChange,
  onSelectAll,
  searchable,
  searchPlaceholder,
  searchQuery,
  onSearchChange,
  renderItem,
  side,
  disabled,
  filteredItems
}) => {
  const totalCount = items.length;
  const checkedCount = filteredItems.filter((i) => checkedIds.has(i.id)).length;
  const allChecked = filteredItems.length > 0 && checkedCount === filteredItems.length;
  const someChecked = checkedCount > 0 && !allChecked;
  const selectAllState = someChecked ? "indeterminate" : allChecked;
  const listboxId = `transfer-list-${side}`;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: "flex min-w-0 flex-1 flex-col rounded border border-border bg-background",
      "data-slot": "transfer-panel",
      "data-side": side,
      children: [
        /* @__PURE__ */ jsxs("div", { className: "border-b border-border px-3 py-2", children: [
          /* @__PURE__ */ jsxs("div", { className: "flex items-baseline gap-1.5", children: [
            /* @__PURE__ */ jsx("span", { className: "text-sm font-label text-foreground", children: label }),
            /* @__PURE__ */ jsxs("span", { className: "text-xs text-muted-foreground", children: [
              "(",
              totalCount,
              ")"
            ] })
          ] }),
          description && /* @__PURE__ */ jsx("p", { className: "mt-0.5 text-xs text-muted-foreground", children: description })
        ] }),
        searchable && /* @__PURE__ */ jsx("div", { className: "border-b border-border px-3 py-2", children: /* @__PURE__ */ jsx(
          Search,
          {
            value: searchQuery,
            onChange: onSearchChange,
            placeholder: searchPlaceholder,
            disabled,
            className: "w-full",
            "aria-label": `Search ${label}`
          }
        ) }),
        /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-2 border-b border-border px-3 py-1.5", children: [
          /* @__PURE__ */ jsx(
            Checkbox,
            {
              id: `${listboxId}-select-all`,
              checked: selectAllState,
              onCheckedChange: (val) => onSelectAll(val === true),
              disabled: disabled || filteredItems.length === 0,
              "aria-label": `Select all ${label} items`
            }
          ),
          /* @__PURE__ */ jsx(
            "label",
            {
              htmlFor: `${listboxId}-select-all`,
              className: cn(
                "cursor-pointer text-xs text-muted-foreground select-none",
                (disabled || filteredItems.length === 0) && "cursor-not-allowed opacity-50"
              ),
              children: checkedCount > 0 ? `${checkedCount} of ${filteredItems.length} selected` : "Select all"
            }
          )
        ] }),
        /* @__PURE__ */ jsx(ScrollArea, { className: "h-64", children: /* @__PURE__ */ jsx(
          "div",
          {
            id: listboxId,
            role: "listbox",
            "aria-label": label,
            "aria-multiselectable": "true",
            className: "flex flex-col py-1",
            children: filteredItems.length === 0 ? /* @__PURE__ */ jsx("div", { className: "px-3 py-4 text-center text-xs text-muted-foreground", children: searchQuery ? "No results match your search." : "No items." }) : filteredItems.map((item) => {
              const isChecked = checkedIds.has(item.id);
              const isDisabled = disabled || item.disabled;
              return /* @__PURE__ */ jsxs(
                "div",
                {
                  role: "option",
                  "aria-selected": isChecked,
                  "aria-disabled": isDisabled,
                  className: cn(
                    "flex cursor-pointer items-start gap-2.5 px-3 py-2 transition-colors",
                    "hover:bg-muted",
                    isChecked && "bg-selected-overlay",
                    isDisabled && "cursor-not-allowed opacity-50"
                  ),
                  onClick: () => {
                    if (!isDisabled) {
                      onCheckedChange(item.id, !isChecked);
                    }
                  },
                  onKeyDown: (e) => {
                    if (!isDisabled && (e.key === " " || e.key === "Enter")) {
                      e.preventDefault();
                      onCheckedChange(item.id, !isChecked);
                    }
                  },
                  tabIndex: isDisabled ? -1 : 0,
                  children: [
                    /* @__PURE__ */ jsx(
                      "div",
                      {
                        "aria-hidden": "true",
                        className: cn(
                          "mt-0.5 size-4 shrink-0 rounded border border-input transition-colors",
                          isChecked && "border-primary bg-primary",
                          isDisabled && "opacity-50"
                        ),
                        children: isChecked && /* @__PURE__ */ jsx(
                          "svg",
                          {
                            viewBox: "0 0 12 12",
                            fill: "none",
                            className: "size-full p-px text-primary-foreground",
                            "aria-hidden": "true",
                            children: /* @__PURE__ */ jsx(
                              "polyline",
                              {
                                points: "1.5,6 4.5,9.5 10.5,2.5",
                                stroke: "currentColor",
                                strokeWidth: "1.5",
                                strokeLinecap: "round",
                                strokeLinejoin: "round"
                              }
                            )
                          }
                        )
                      }
                    ),
                    /* @__PURE__ */ jsx("div", { className: "min-w-0 flex-1", children: renderItem ? renderItem(item, side) : /* @__PURE__ */ jsxs(Fragment, { children: [
                      /* @__PURE__ */ jsx("span", { className: "block truncate text-sm text-foreground", children: item.label }),
                      item.description && /* @__PURE__ */ jsx("span", { className: "block truncate text-xs text-muted-foreground", children: item.description })
                    ] }) })
                  ]
                },
                item.id
              );
            })
          }
        ) })
      ]
    }
  );
};

export {
  TransferListPanel
};
