// @docker/trident/dist/chunk-YYG5D2XZ.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/icon-button.tsx
import { cva } from "class-variance-authority";
import { Slot } from "radix-ui";
import { jsx } from "react/jsx-runtime";
var iconButtonVariants = cva(
  "relative inline-flex shrink-0 items-center justify-center rounded-sm transition-colors outline-none select-none disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0",
  {
    variants: {
      variant: {
        ghost: "text-muted-foreground hover:bg-muted hover:text-foreground focus-visible:ring-1 focus-visible:ring-ring",
        outline: "border border-border text-muted-foreground hover:bg-muted hover:text-foreground focus-visible:ring-1 focus-visible:ring-ring",
        destructive: "text-muted-foreground hover:bg-destructive/10 hover:text-destructive focus-visible:ring-1 focus-visible:ring-destructive"
      },
      size: {
        xs: "size-4 after:absolute after:-inset-1.5 [&_svg:not([class*='size-'])]:size-3",
        sm: "size-5 after:absolute after:-inset-1 [&_svg:not([class*='size-'])]:size-3.5",
        md: "size-6 after:absolute after:-inset-0.5 [&_svg:not([class*='size-'])]:size-4",
        lg: "size-8 [&_svg:not([class*='size-'])]:size-4"
      }
    },
    defaultVariants: {
      variant: "ghost",
      size: "sm"
    }
  }
);
function IconButton({
  className,
  variant = "ghost",
  size = "sm",
  asChild = false,
  ...props
}) {
  const Comp = asChild ? Slot.Root : "button";
  return /* @__PURE__ */ jsx(
    Comp,
    {
      type: "button",
      "data-slot": "icon-button",
      "data-variant": variant,
      "data-size": size,
      className: cn(iconButtonVariants({ variant, size, className })),
      ...props
    }
  );
}

export {
  iconButtonVariants,
  IconButton
};
