// @docker/trident/dist/chunk-JC6O5PBN.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/card.tsx
import { jsx, jsxs } from "react/jsx-runtime";
function Card({
  className,
  size = "default",
  ...props
}) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "card",
      "data-size": size,
      className: cn(
        "group/card @container/card flex flex-col gap-6 overflow-hidden rounded bg-card py-6 text-sm text-card-foreground shadow-raised ring-1 ring-border has-[>img:first-child]:pt-0 data-[size=sm]:gap-4 data-[size=sm]:py-4 *:[img:first-child]:rounded-t *:[img:last-child]:rounded-b",
        className
      ),
      ...props
    }
  );
}
function CardHeader({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "card-header",
      className: cn(
        "group/card-header @container/card-header grid auto-rows-min items-start gap-1 rounded-t px-6 group-data-[size=sm]/card:px-4 has-data-[slot=card-action]:grid-cols-[1fr_auto] has-data-[slot=card-description]:grid-rows-[auto_auto] @max-[300px]/card:px-3 [.border-b]:pb-6 group-data-[size=sm]/card:[.border-b]:pb-4",
        className
      ),
      ...props
    }
  );
}
function CardTitle({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "card-title",
      className: cn(
        "text-base leading-heading font-heading tracking-heading group-data-[size=sm]/card:text-sm @max-[300px]/card:text-sm",
        className
      ),
      ...props
    }
  );
}
function CardDescription({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "card-description",
      className: cn(
        "text-sm leading-body tracking-body text-muted-foreground",
        className
      ),
      ...props
    }
  );
}
function CardAction({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "card-action",
      className: cn(
        "col-start-2 row-span-2 row-start-1 self-start justify-self-end",
        className
      ),
      ...props
    }
  );
}
function CardContent({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "card-content",
      className: cn(
        "px-6 group-data-[size=sm]/card:px-4 @max-[300px]/card:px-3",
        className
      ),
      ...props
    }
  );
}
function CardFooter({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "card-footer",
      className: cn(
        "flex items-center rounded-b px-6 group-data-[size=sm]/card:px-4 @max-[300px]/card:px-3 [.border-t]:pt-6 group-data-[size=sm]/card:[.border-t]:pt-4",
        className
      ),
      ...props
    }
  );
}
function BackCard({ depth }) {
  return /* @__PURE__ */ jsx(
    Card,
    {
      "aria-hidden": "true",
      className: cn(
        "shadow-subtle pointer-events-none absolute h-full w-full border border-dashed border-border bg-muted opacity-60 ring-0",
        depth === 1 && "-top-dds-300 -left-dds-300",
        depth === 2 && "-top-dds-600 -left-dds-600"
      )
    }
  );
}
function CardStack({
  stackDepth = 1,
  className,
  children,
  ...props
}) {
  const backCards = [];
  for (let depth = stackDepth; depth >= 1; depth--) {
    backCards.push(/* @__PURE__ */ jsx(BackCard, { depth }, depth));
  }
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "card-stack",
      className: cn("relative", className),
      ...props,
      children: [
        backCards,
        /* @__PURE__ */ jsx(Card, { className: "relative", children })
      ]
    }
  );
}

export {
  Card,
  CardHeader,
  CardTitle,
  CardDescription,
  CardAction,
  CardContent,
  CardFooter,
  CardStack
};
