// @docker/trident/dist/chunk-KXARZUZR.js
import {
  DataGridFilterButton
} from "./chunk-BJJFXHWE.js";
import {
  TableCell,
  TableHead,
  TableHeader,
  TableRow
} from "./chunk-SKOSYJCU.js";
import {
  TEXT_ALIGN_CLASS
} from "./chunk-BBDVHE6E.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/data-grid-header.tsx
import {
  flexRender
} from "@tanstack/react-table";
import {
  ChevronDown,
  ChevronRight,
  ChevronsUpDown,
  ChevronUp
} from "lucide-react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function DataGridHeader({
  table,
  hClass,
  disableColumnMenu,
  disableColumnResize,
  disableColumnReorder,
  disableColumnFilter,
  dragId,
  setColumnOrder,
  filterModel,
  onFilterChange
}) {
  return /* @__PURE__ */ jsx(TableHeader, { children: table.getHeaderGroups().map((hg) => /* @__PURE__ */ jsx(TableRow, { children: hg.headers.map((header) => {
    const meta = header.column.columnDef.meta;
    const canSort = header.column.getCanSort() && !disableColumnMenu && !meta?.disableColumnMenu;
    const sortDir = header.column.getIsSorted();
    const hAlign = meta?.headerAlign ?? "left";
    const isPinned = header.column.getIsPinned();
    const canResize = !disableColumnResize && meta?.resizable !== false && header.column.id !== "__select__";
    const isDraggable = !disableColumnReorder && header.column.id !== "__select__";
    const colFilterable = !disableColumnFilter && meta?.filterable === true && header.column.id !== "__select__";
    const sortHandler = canSort ? header.column.getToggleSortingHandler() : void 0;
    const headerLabel = header.isPlaceholder ? null : flexRender(header.column.columnDef.header, header.getContext());
    const sortIndicator = canSort ? /* @__PURE__ */ jsx("span", { className: "text-muted-foreground", "aria-hidden": "true", children: sortDir === "asc" ? /* @__PURE__ */ jsx(ChevronUp, { className: "size-3.5" }) : sortDir === "desc" ? /* @__PURE__ */ jsx(ChevronDown, { className: "size-3.5" }) : /* @__PURE__ */ jsx(ChevronsUpDown, { className: "size-3.5 opacity-40" }) }) : null;
    return /* @__PURE__ */ jsxs(
      TableHead,
      {
        scope: "col",
        "data-field": header.column.id,
        "data-align": hAlign,
        "data-sortable": canSort ? "true" : void 0,
        "data-pinned": isPinned || void 0,
        "aria-sort": canSort ? sortDir === "asc" ? "ascending" : sortDir === "desc" ? "descending" : "none" : void 0,
        draggable: isDraggable || void 0,
        className: cn(
          hClass,
          TEXT_ALIGN_CLASS[hAlign],
          isPinned && "bg-background",
          "relative"
        ),
        style: {
          width: header.getSize(),
          position: isPinned ? "sticky" : void 0,
          left: isPinned === "left" ? `${header.column.getStart("left")}px` : void 0,
          right: isPinned === "right" ? `${header.column.getAfter("right")}px` : void 0,
          zIndex: isPinned ? 1 : void 0
        },
        onDragStart: isDraggable ? () => {
          dragId.current = header.column.id;
        } : void 0,
        onDragOver: isDraggable ? (e) => e.preventDefault() : void 0,
        onDrop: isDraggable ? () => {
          const from = dragId.current;
          const to = header.column.id;
          if (!from || from === to) return;
          setColumnOrder((prev) => {
            const next = [...prev];
            const fi = next.indexOf(from);
            const ti = next.indexOf(to);
            if (fi < 0 || ti < 0) return prev;
            next.splice(fi, 1);
            next.splice(ti, 0, from);
            return next;
          });
          dragId.current = null;
        } : void 0,
        children: [
          /* @__PURE__ */ jsxs("span", { className: "flex items-center gap-1", children: [
            canSort ? /* @__PURE__ */ jsxs(
              "button",
              {
                type: "button",
                "data-slot": "column-sort",
                onClick: sortHandler,
                className: cn(
                  "flex flex-1 cursor-pointer items-center gap-1 rounded border-0 bg-transparent p-0 [color:inherit] select-none [font:inherit]",
                  TEXT_ALIGN_CLASS[hAlign],
                  "focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none"
                ),
                children: [
                  headerLabel,
                  sortIndicator
                ]
              }
            ) : /* @__PURE__ */ jsxs(Fragment, { children: [
              headerLabel,
              sortIndicator
            ] }),
            colFilterable && /* @__PURE__ */ jsx(
              DataGridFilterButton,
              {
                columnId: header.column.id,
                filterModel,
                onFilterChange: (f) => onFilterChange(f, header.column.id),
                operators: meta?.filterOperators
              }
            )
          ] }),
          canResize && /* @__PURE__ */ jsx(
            "div",
            {
              "data-slot": "column-resize-handle",
              onMouseDown: header.getResizeHandler(),
              onTouchStart: header.getResizeHandler(),
              className: cn(
                "absolute top-0 right-0 h-full w-1 cursor-col-resize touch-none select-none",
                "bg-border opacity-0 hover:bg-primary hover:opacity-100",
                header.column.getIsResizing() && "bg-primary opacity-100"
              )
            }
          )
        ]
      },
      header.id
    );
  }) }, hg.id)) });
}
function DataGridGroupRow({
  row,
  cClass
}) {
  return /* @__PURE__ */ jsx(
    TableRow,
    {
      "data-slot": "data-grid-group-row",
      className: "bg-hover-overlay font-label",
      children: row.getVisibleCells().map((cell) => {
        if (cell.getIsGrouped()) {
          return /* @__PURE__ */ jsx(TableCell, { className: cn(cClass, "font-label"), children: /* @__PURE__ */ jsxs("span", { className: "flex items-center gap-1", children: [
            /* @__PURE__ */ jsx(
              "button",
              {
                type: "button",
                "data-slot": "group-expand",
                onClick: () => row.toggleExpanded(),
                "aria-label": row.getIsExpanded() ? "Collapse group" : "Expand group",
                className: "flex items-center text-muted-foreground",
                children: row.getIsExpanded() ? /* @__PURE__ */ jsx(ChevronDown, { className: "size-4" }) : /* @__PURE__ */ jsx(ChevronRight, { className: "size-4" })
              }
            ),
            String(cell.getValue()),
            /* @__PURE__ */ jsxs("span", { className: "text-xs text-muted-foreground", children: [
              "(",
              row.subRows.length,
              ")"
            ] })
          ] }) }, cell.id);
        }
        if (cell.getIsAggregated()) {
          return /* @__PURE__ */ jsx(TableCell, { className: cClass, children: flexRender(
            cell.column.columnDef.aggregatedCell ?? cell.column.columnDef.cell,
            cell.getContext()
          ) }, cell.id);
        }
        return /* @__PURE__ */ jsx(TableCell, {}, cell.id);
      })
    }
  );
}

export {
  DataGridHeader,
  DataGridGroupRow
};
