// @docker/trident/dist/chunk-7JUAA2Z6.js
import {
  CopyToClipboardButton
} from "./chunk-LJSD3DSH.js";
import {
  Tooltip,
  TooltipContent,
  TooltipTrigger
} from "./chunk-HVYJXDDW.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/digest.tsx
import { Slot } from "radix-ui";
import { jsx, jsxs } from "react/jsx-runtime";
function Digest({
  fullDigest,
  abbrev = true,
  maxLength = 12,
  copyable = true,
  tooltipPlacement = "bottom",
  className,
  asChild = false,
  children
}) {
  const isAbbreviated = abbrev && fullDigest.length > maxLength;
  const displayText = isAbbreviated ? `${fullDigest.substring(0, maxLength)}\u2026` : fullDigest;
  const digestText = isAbbreviated ? /* @__PURE__ */ jsx(
    "button",
    {
      type: "button",
      "data-slot": "digest-text",
      "aria-label": fullDigest,
      className: cn(
        "cursor-default rounded-xs border-0 bg-transparent p-0 font-mono text-sm text-inherit outline-none",
        "focus-visible:ring-3 focus-visible:ring-ring"
      ),
      children: /* @__PURE__ */ jsx("span", { "aria-hidden": "true", children: displayText })
    }
  ) : /* @__PURE__ */ jsx("span", { "data-slot": "digest-text", className: "font-mono text-sm", children: displayText });
  if (asChild) {
    return /* @__PURE__ */ jsx(
      Slot.Root,
      {
        "data-slot": "digest",
        className: cn("inline-flex items-center gap-1", className),
        children
      }
    );
  }
  return /* @__PURE__ */ jsxs(
    "span",
    {
      "data-slot": "digest",
      className: cn("inline-flex items-center gap-1", className),
      children: [
        isAbbreviated ? /* @__PURE__ */ jsxs(Tooltip, { children: [
          /* @__PURE__ */ jsx(TooltipTrigger, { asChild: true, children: digestText }),
          /* @__PURE__ */ jsx(TooltipContent, { side: tooltipPlacement, className: "font-mono text-xs", children: fullDigest })
        ] }) : digestText,
        copyable && /* @__PURE__ */ jsx(
          CopyToClipboardButton,
          {
            text: fullDigest,
            size: "icon-xs",
            className: "size-5 text-muted-foreground hover:text-foreground"
          }
        )
      ]
    }
  );
}

export {
  Digest
};
