// @docker/trident/dist/chunk-FWSKXSDB.js
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/step-indicator.tsx
import { ChevronLeft, ChevronRight, CircleCheck } from "lucide-react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var StepIndicator = ({
  steps,
  currentStep,
  variant = "default",
  showNavigation = false,
  onPrevious,
  onNext,
  className
}) => {
  if (variant === "dots") {
    return /* @__PURE__ */ jsxs(
      "div",
      {
        "data-slot": "step-indicator",
        "data-variant": "dots",
        className: cn("flex items-center gap-3", className),
        children: [
          showNavigation && /* @__PURE__ */ jsx(
            Button,
            {
              variant: "outline",
              size: "icon-sm",
              onClick: onPrevious,
              disabled: currentStep <= 1,
              "aria-label": "Previous step",
              children: /* @__PURE__ */ jsx(ChevronLeft, {})
            }
          ),
          /* @__PURE__ */ jsx(
            "fieldset",
            {
              className: "m-0 flex min-w-0 items-center gap-1.5 border-0 p-0",
              "aria-label": `Step ${currentStep} of ${steps}`,
              children: Array.from({ length: steps }, (_, i) => {
                const step = i + 1;
                const isActive = step === currentStep;
                const isComplete = step < currentStep;
                return /* @__PURE__ */ jsx(
                  "span",
                  {
                    "aria-hidden": true,
                    className: cn(
                      "size-2 rounded-full transition-colors",
                      isActive && "bg-primary",
                      isComplete && "bg-accent-track-fill",
                      !isActive && !isComplete && "bg-border"
                    )
                  },
                  step
                );
              })
            }
          ),
          showNavigation && /* @__PURE__ */ jsx(
            Button,
            {
              variant: "outline",
              size: "icon-sm",
              onClick: onNext,
              disabled: currentStep >= steps,
              "aria-label": "Next step",
              children: /* @__PURE__ */ jsx(ChevronRight, {})
            }
          )
        ]
      }
    );
  }
  return /* @__PURE__ */ jsx(
    "nav",
    {
      "data-slot": "step-indicator",
      "data-variant": "default",
      "aria-label": "Progress",
      className: cn("flex items-center gap-2", className),
      children: /* @__PURE__ */ jsx("ol", { className: "flex items-center gap-2", role: "list", children: Array.from({ length: steps }, (_, i) => {
        const step = i + 1;
        const isActive = step === currentStep;
        const isComplete = step < currentStep;
        const stepLabel = `Step ${step} of ${steps}${isComplete ? ", completed" : isActive ? ", current" : ""}`;
        return /* @__PURE__ */ jsxs("li", { className: "flex items-center gap-2", role: "listitem", children: [
          /* @__PURE__ */ jsx(
            "span",
            {
              role: "img",
              "aria-label": stepLabel,
              "aria-current": isActive ? "step" : void 0,
              className: cn(
                "flex size-6 items-center justify-center rounded-full text-xs font-label transition-colors",
                isActive && "bg-primary text-primary-foreground",
                isComplete && "bg-accent-track text-primary",
                !isActive && !isComplete && "bg-muted text-muted-foreground"
              ),
              children: isComplete ? /* @__PURE__ */ jsxs(Fragment, { children: [
                /* @__PURE__ */ jsx(CircleCheck, { className: "size-3.5", "aria-hidden": "true" }),
                /* @__PURE__ */ jsx("span", { className: "sr-only", children: "Completed" })
              ] }) : /* @__PURE__ */ jsx("span", { "aria-hidden": "true", children: step })
            }
          ),
          i < steps - 1 && /* @__PURE__ */ jsx(
            "div",
            {
              "aria-hidden": "true",
              className: cn(
                "h-px w-8",
                step < currentStep ? "bg-accent-track-fill" : "bg-border"
              )
            }
          )
        ] }, step);
      }) })
    }
  );
};

export {
  StepIndicator
};
