// @docker/trident/dist/chunk-Z24NXZ3W.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/blockquote.tsx
import { cva } from "class-variance-authority";
import { jsx, jsxs } from "react/jsx-runtime";
var blockquoteVariants = cva("my-2 border-l-4 px-4 py-3 text-sm", {
  variants: {
    variant: {
      note: "border-l-blockquote-note bg-blockquote-note-bg",
      tip: "border-l-blockquote-tip bg-blockquote-tip-bg",
      important: "border-l-blockquote-important bg-blockquote-important-bg",
      warning: "border-l-blockquote-warning bg-blockquote-warning-bg",
      caution: "border-l-blockquote-caution bg-blockquote-caution-bg"
    }
  },
  defaultVariants: {
    variant: "note"
  }
});
var titleColorMap = {
  note: "text-blockquote-note-title",
  tip: "text-blockquote-tip-title",
  important: "text-blockquote-important-title",
  warning: "text-blockquote-warning-title",
  caution: "text-blockquote-caution-title"
};
function Blockquote({
  variant = "note",
  title,
  className,
  children,
  ...props
}) {
  return /* @__PURE__ */ jsxs(
    "blockquote",
    {
      "data-slot": "blockquote",
      "data-variant": variant,
      className: cn(blockquoteVariants({ variant }), className),
      ...props,
      children: [
        title && /* @__PURE__ */ jsx(
          "p",
          {
            className: cn(
              "mb-1 text-sm font-heading",
              titleColorMap[variant ?? "note"]
            ),
            children: title
          }
        ),
        /* @__PURE__ */ jsx("div", { className: "text-sm text-foreground", children })
      ]
    }
  );
}

export {
  blockquoteVariants,
  Blockquote
};
