// @docker/trident/dist/chunk-2R34D44J.js
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/tag.tsx
import { cva } from "class-variance-authority";
import { X } from "lucide-react";
import { Slot } from "radix-ui";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var tagVariants = cva(
  "inline-flex items-center gap-1 rounded border select-none data-disabled:pointer-events-none data-disabled:opacity-60",
  {
    variants: {
      variant: {
        default: "border-border bg-muted text-foreground",
        success: "border-transparent bg-tag-success-bg text-tag-success-fg",
        warning: "border-transparent bg-tag-warning-bg text-tag-warning-fg",
        destructive: "border-transparent bg-tag-destructive-bg text-tag-destructive-fg",
        info: "border-transparent bg-tag-info-bg text-tag-info-fg"
      },
      size: {
        sm: "h-5 px-1.5 text-2xs font-label",
        md: "h-6 px-2 text-xs font-label",
        lg: "h-7 px-2.5 text-sm font-label"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "md"
    }
  }
);
var Tag = ({
  className,
  children,
  variant = "default",
  size = "md",
  onRemove,
  disabled,
  asChild = false,
  ...props
}) => {
  const sharedProps = {
    "data-slot": "tag",
    "data-disabled": disabled || void 0,
    className: cn(tagVariants({ variant, size }), className),
    ...props
  };
  const removeButton = onRemove ? /* @__PURE__ */ jsx(
    Button,
    {
      variant: "ghost",
      size: "icon-xs",
      onClick: onRemove,
      disabled,
      className: "-mr-1 size-4 rounded opacity-60 hover:opacity-100",
      "aria-label": `Remove ${typeof children === "string" ? children : "tag"}`,
      children: /* @__PURE__ */ jsx(X, { className: "size-2.5", "aria-hidden": "true" })
    }
  ) : null;
  if (asChild) {
    const child = React.Children.only(children);
    if (!React.isValidElement(child)) return null;
    return /* @__PURE__ */ jsx(Slot.Root, { ...sharedProps, children: React.cloneElement(child, {}, child.props.children, removeButton) });
  }
  return /* @__PURE__ */ jsxs("span", { ...sharedProps, children: [
    children,
    removeButton
  ] });
};

export {
  tagVariants,
  Tag
};
