// @docker/trident/dist/chunk-OP3FHSXA.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/provenance-indicator.tsx
import { Check, HelpCircle, X } from "lucide-react";
import { jsx, jsxs } from "react/jsx-runtime";
var STATUS_CONFIG = {
  signed: {
    Icon: Check,
    label: "Provenance signed",
    colorClass: "text-success"
  },
  unsigned: {
    Icon: X,
    label: "Provenance unsigned",
    colorClass: "text-muted-foreground"
  },
  unknown: {
    Icon: HelpCircle,
    label: "Provenance unknown",
    colorClass: "text-muted-foreground"
  }
};
function ProvenanceIndicator({
  status,
  className,
  ...props
}) {
  const { Icon, label, colorClass } = STATUS_CONFIG[status];
  return /* @__PURE__ */ jsxs(
    "span",
    {
      "data-slot": "provenance-indicator",
      role: "img",
      "aria-label": label,
      className: cn(
        "inline-flex items-center gap-1 font-mono text-xs",
        colorClass,
        className
      ),
      ...props,
      children: [
        /* @__PURE__ */ jsx(Icon, { className: "size-3", "aria-hidden": true }),
        "Provenance"
      ]
    }
  );
}

export {
  ProvenanceIndicator
};
