// @docker/trident/dist/chunk-QT72EBV6.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/fab.tsx
import { cva } from "class-variance-authority";
import { jsx } from "react/jsx-runtime";
var fabVariants = cva(
  "inline-flex shrink-0 items-center justify-center rounded-full font-label shadow-floating ring-1 ring-border transition-all ease-default outline-none select-none focus-visible:ring-3 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0",
  {
    variants: {
      variant: {
        default: "bg-primary text-primary-foreground hover:bg-accent-hover",
        secondary: "bg-secondary text-secondary-foreground hover:bg-secondary-hover",
        destructive: "bg-destructive text-destructive-foreground hover:bg-destructive-hover focus-visible:ring-destructive/50",
        outline: "border border-border bg-card hover:bg-muted hover:text-foreground"
      },
      size: {
        sm: "size-10 [&_svg:not([class*='size-'])]:size-4",
        default: "size-14 [&_svg:not([class*='size-'])]:size-6",
        lg: "size-16 [&_svg:not([class*='size-'])]:size-7"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
);
function Fab({ className, variant, size, ...props }) {
  return /* @__PURE__ */ jsx(
    "button",
    {
      "data-slot": "fab",
      className: cn(fabVariants({ variant, size, className })),
      ...props
    }
  );
}

export {
  fabVariants,
  Fab
};
