// @docker/trident/dist/chunk-KPBOQNC6.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/timeline.tsx
import { cva } from "class-variance-authority";
import { jsx } from "react/jsx-runtime";
function Timeline({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "ol",
    {
      "data-slot": "timeline",
      className: cn("flex flex-col", className),
      ...props
    }
  );
}
function TimelineItem({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "li",
    {
      "data-slot": "timeline-item",
      className: cn("flex gap-3", className),
      ...props
    }
  );
}
function TimelineSeparator({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "timeline-separator",
      className: cn(
        "flex flex-col items-center",
        'before:mb-1 before:w-px before:flex-1 before:bg-border before:content-[""]',
        'after:mt-1 after:w-px after:flex-1 after:bg-border after:content-[""]',
        "[li:first-child_&]:before:invisible [li:first-child_&]:before:mb-0",
        "[li:last-child_&]:after:invisible [li:last-child_&]:after:mt-0",
        className
      ),
      ...props
    }
  );
}
var timelineDotVariants = cva(
  "flex shrink-0 items-center justify-center rounded-full [&>svg]:size-3.5",
  {
    variants: {
      variant: {
        default: "size-3 bg-border",
        outline: "size-3 border-2 border-border bg-transparent",
        filled: "size-3 bg-primary"
      },
      size: {
        default: "",
        icon: "size-7"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
);
function TimelineDot({
  className,
  variant = "default",
  children,
  ...props
}) {
  const hasChildren = children != null;
  return /* @__PURE__ */ jsx(
    "span",
    {
      "data-slot": "timeline-dot",
      "data-variant": variant,
      className: cn(
        timelineDotVariants({
          variant,
          size: hasChildren ? "icon" : "default"
        }),
        className
      ),
      ...props,
      children
    }
  );
}
function TimelineContent({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "timeline-content",
      className: cn(
        "py-1 text-sm [li:first-child_&]:pt-0 [li:last-child_&]:pb-0",
        className
      ),
      ...props
    }
  );
}

export {
  Timeline,
  TimelineItem,
  TimelineSeparator,
  timelineDotVariants,
  TimelineDot,
  TimelineContent
};
