// @docker/trident/dist/chunk-3FGH5OXJ.js
// src/components/data-grid-tree.ts
function buildForest(rows, getTreeDataPath) {
  const nodeMap = /* @__PURE__ */ new Map();
  for (const row of rows) {
    const segments = getTreeDataPath(row);
    const pathKey = segments.join("/");
    nodeMap.set(pathKey, { pathKey, segments, row, children: [] });
  }
  const roots = [];
  for (const row of rows) {
    const segments = getTreeDataPath(row);
    const pathKey = segments.join("/");
    const node = nodeMap.get(pathKey);
    if (!node) continue;
    if (segments.length <= 1) {
      roots.push(node);
    } else {
      const parentKey = segments.slice(0, -1).join("/");
      const parent = nodeMap.get(parentKey);
      if (parent) {
        parent.children.push(node);
      } else {
        roots.push(node);
      }
    }
  }
  return roots;
}
function flattenForest(nodes, expandedPaths, indent) {
  const result = [];
  for (const node of nodes) {
    const hasChildren = node.children.length > 0;
    const isExpanded = hasChildren && expandedPaths.has(node.pathKey);
    result.push({
      row: node.row,
      meta: {
        indent,
        hasChildren,
        isExpanded,
        pathKey: node.pathKey
      }
    });
    if (isExpanded) {
      const children = flattenForest(node.children, expandedPaths, indent + 1);
      result.push(...children);
    }
  }
  return result;
}
function buildTreeDisplayList(rows, getTreeDataPath, expandedPaths) {
  const forest = buildForest(rows, getTreeDataPath);
  return flattenForest(forest, expandedPaths, 0);
}
function toggleTreePath(expandedPaths, pathKey) {
  const next = new Set(expandedPaths);
  if (next.has(pathKey)) {
    next.delete(pathKey);
  } else {
    next.add(pathKey);
  }
  return next;
}

export {
  buildTreeDisplayList,
  toggleTreePath
};
