// @docker/trident/dist/chunk-BJJFXHWE.js
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue
} from "./chunk-IJMN6FVR.js";
import {
  Badge
} from "./chunk-DBSTRLSQ.js";
import {
  Popover,
  PopoverContent,
  PopoverTrigger
} from "./chunk-R6CWLGJU.js";
import {
  Input
} from "./chunk-5CQB3X63.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/data-grid-filter.tsx
import { FilterIcon } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var OPERATOR_LABELS = {
  contains: "Contains",
  equals: "Equals",
  startsWith: "Starts with",
  endsWith: "Ends with",
  greaterThan: "Greater than",
  lessThan: "Less than",
  isEmpty: "Is empty",
  isNotEmpty: "Is not empty"
};
var STRING_OPERATORS = [
  "contains",
  "equals",
  "startsWith",
  "endsWith",
  "isEmpty",
  "isNotEmpty"
];
var NUMBER_OPERATORS = [
  "equals",
  "greaterThan",
  "lessThan",
  "isEmpty",
  "isNotEmpty"
];
var NO_VALUE_OPERATORS = /* @__PURE__ */ new Set(["isEmpty", "isNotEmpty"]);
var ALL_OPERATORS = /* @__PURE__ */ new Set([
  "contains",
  "equals",
  "startsWith",
  "endsWith",
  "greaterThan",
  "lessThan",
  "isEmpty",
  "isNotEmpty"
]);
function isFilterOperator(v) {
  return ALL_OPERATORS.has(v);
}
function applyFilters(rows, filters, columns) {
  if (filters.length === 0) return rows;
  return rows.filter((row) => {
    for (const filter of filters) {
      const col = columns.find((c) => c.field === filter.columnId);
      if (!col) continue;
      const raw = col.valueGetter ? col.valueGetter(row) : Reflect.get(row, filter.columnId);
      if (!evaluateFilter(raw, filter)) return false;
    }
    return true;
  });
}
function toStr(val) {
  if (val == null) return "";
  if (typeof val === "object") return JSON.stringify(val);
  if (typeof val === "string" || typeof val === "number" || typeof val === "boolean") {
    return String(val);
  }
  return JSON.stringify(val);
}
function evaluateFilter(raw, filter) {
  const { operator, value } = filter;
  if (operator === "isEmpty") {
    return raw == null || toStr(raw).trim() === "";
  }
  if (operator === "isNotEmpty") {
    return raw != null && toStr(raw).trim() !== "";
  }
  if (value === void 0 || value === "") return true;
  const cellStr = toStr(raw ?? "").toLowerCase();
  const filterStr = value.toLowerCase();
  switch (operator) {
    case "contains":
      return cellStr.includes(filterStr);
    case "equals":
      return cellStr === filterStr;
    case "startsWith":
      return cellStr.startsWith(filterStr);
    case "endsWith":
      return cellStr.endsWith(filterStr);
    case "greaterThan": {
      const cellNum = Number(raw);
      const filterNum = Number(value);
      return !Number.isNaN(cellNum) && !Number.isNaN(filterNum) && cellNum > filterNum;
    }
    case "lessThan": {
      const cellNum = Number(raw);
      const filterNum = Number(value);
      return !Number.isNaN(cellNum) && !Number.isNaN(filterNum) && cellNum < filterNum;
    }
    default:
      return true;
  }
}
var DataGridFilterButton = ({
  columnId,
  filterModel,
  onFilterChange,
  operators = STRING_OPERATORS
}) => {
  const activeFilter = filterModel.find((f) => f.columnId === columnId);
  const [open, setOpen] = React.useState(false);
  const [operator, setOperator] = React.useState(
    activeFilter?.operator ?? operators[0] ?? "contains"
  );
  const [value, setValue] = React.useState(activeFilter?.value ?? "");
  React.useEffect(() => {
    if (activeFilter) {
      setOperator(activeFilter.operator);
      setValue(activeFilter.value ?? "");
    } else {
      setOperator(operators[0] ?? "contains");
      setValue("");
    }
  }, [activeFilter, operators]);
  const isActive = activeFilter !== void 0;
  const needsValue = !NO_VALUE_OPERATORS.has(operator);
  const handleApply = () => {
    onFilterChange({
      columnId,
      operator,
      value: needsValue ? value : void 0
    });
    setOpen(false);
  };
  const handleClear = () => {
    onFilterChange(null);
    setValue("");
    setOperator(operators[0] ?? "contains");
    setOpen(false);
  };
  const handleKeyDown = (e) => {
    if (e.key === "Enter") handleApply();
  };
  return /* @__PURE__ */ jsxs(Popover, { open, onOpenChange: setOpen, children: [
    /* @__PURE__ */ jsx(PopoverTrigger, { asChild: true, children: /* @__PURE__ */ jsxs(
      "button",
      {
        type: "button",
        "data-slot": "data-grid-filter-button",
        "aria-label": `Filter ${columnId} column`,
        "aria-pressed": isActive,
        onClick: (e) => e.stopPropagation(),
        className: cn(
          "relative inline-flex items-center justify-center rounded p-0.5 transition-colors",
          "text-muted-foreground hover:bg-muted hover:text-foreground",
          "focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none",
          isActive && "text-primary"
        ),
        children: [
          /* @__PURE__ */ jsx(
            FilterIcon,
            {
              className: cn(
                "size-3.5",
                isActive && "fill-primary stroke-primary"
              ),
              "aria-hidden": "true"
            }
          ),
          isActive && /* @__PURE__ */ jsx(
            Badge,
            {
              variant: "default",
              size: "sm",
              className: "absolute -top-1.5 -right-1.5 size-3.5 min-w-0 px-0 text-2xs leading-none",
              "aria-hidden": "true",
              children: "1"
            }
          )
        ]
      }
    ) }),
    /* @__PURE__ */ jsx(
      PopoverContent,
      {
        align: "start",
        sideOffset: 6,
        className: "w-64",
        "aria-label": "Column filter",
        onClick: (e) => e.stopPropagation(),
        children: /* @__PURE__ */ jsxs("div", { className: "flex flex-col gap-dds-200", children: [
          /* @__PURE__ */ jsx("p", { className: "text-xs font-label tracking-wide text-muted-foreground uppercase", children: "Filter" }),
          /* @__PURE__ */ jsxs(
            Select,
            {
              value: operator,
              onValueChange: (v) => {
                if (isFilterOperator(v)) setOperator(v);
              },
              children: [
                /* @__PURE__ */ jsx(
                  SelectTrigger,
                  {
                    size: "sm",
                    className: "w-full",
                    "aria-label": "Filter operator",
                    children: /* @__PURE__ */ jsx(SelectValue, {})
                  }
                ),
                /* @__PURE__ */ jsx(SelectContent, { position: "popper", sideOffset: 4, children: operators.map((op) => /* @__PURE__ */ jsx(SelectItem, { value: op, children: OPERATOR_LABELS[op] }, op)) })
              ]
            }
          ),
          needsValue && /* @__PURE__ */ jsx(
            Input,
            {
              size: "sm",
              placeholder: "Filter value\u2026",
              value,
              onChange: (e) => setValue(e.target.value),
              onKeyDown: handleKeyDown,
              "aria-label": "Filter value"
            }
          ),
          /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-end gap-dds-100", children: [
            /* @__PURE__ */ jsx(
              Button,
              {
                variant: "ghost",
                size: "sm",
                onClick: handleClear,
                disabled: !isActive,
                children: "Clear"
              }
            ),
            /* @__PURE__ */ jsx(Button, { size: "sm", onClick: handleApply, children: "Apply" })
          ] })
        ] })
      }
    )
  ] });
};

export {
  OPERATOR_LABELS,
  STRING_OPERATORS,
  NUMBER_OPERATORS,
  applyFilters,
  DataGridFilterButton
};
