// @docker/trident/dist/chunk-YZWH6CFD.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/field-counter.tsx
import { useCallback, useState } from "react";
import { jsxs } from "react/jsx-runtime";
var FieldCounter = ({
  value,
  max,
  warnAt = 80,
  errorAt = 100,
  className,
  id
}) => {
  const count = typeof value === "string" ? value.length : value ?? 0;
  const pct = max > 0 ? count / max * 100 : 0;
  const isError = pct >= errorAt || count >= max;
  const isWarn = !isError && pct >= warnAt;
  const colorClass = isError ? "text-destructive high-contrast:text-tag-destructive-fg" : isWarn ? "text-warning" : "text-muted-foreground high-contrast:text-foreground";
  return /* @__PURE__ */ jsxs(
    "output",
    {
      id,
      "aria-live": "polite",
      "data-slot": "field-counter",
      className: cn(
        "inline-flex items-center justify-end text-sm tabular-nums",
        colorClass,
        className
      ),
      children: [
        /* @__PURE__ */ jsxs("span", { "aria-hidden": "true", children: [
          count,
          " / ",
          max
        ] }),
        /* @__PURE__ */ jsxs("span", { className: "sr-only", children: [
          count,
          " of ",
          max,
          " characters used"
        ] })
      ]
    }
  );
};
var useFieldCounter = (max) => {
  const [value, setValue] = useState("");
  const onChange = useCallback((e) => {
    setValue(e.target.value);
  }, []);
  return {
    value,
    count: value.length,
    inputProps: {
      value,
      onChange,
      maxLength: max
    },
    counterProps: {
      value: value.length,
      max
    }
  };
};

export {
  FieldCounter,
  useFieldCounter
};
