// @docker/trident/dist/chunk-XM3VHVK2.js
import {
  Sidebar,
  SidebarContent,
  SidebarFooter,
  SidebarHeader,
  SidebarMenuButton,
  SidebarMenuItem,
  SidebarMenuSub,
  SidebarMenuSubButton,
  SidebarMenuSubItem,
  SidebarProvider,
  SidebarRail,
  useSidebar
} from "./chunk-BZTV2IDP.js";
import {
  Collapsible,
  CollapsibleContent,
  CollapsibleTrigger
} from "./chunk-TCECHOOI.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/docker-sidebar.tsx
import { ChevronRight } from "lucide-react";
import * as React from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var COOKIE_NAME = "sidebar:state";
var COOKIE_MAX_AGE = 60 * 60 * 24 * 7;
function readCookie(name) {
  if (typeof document === "undefined") return null;
  const match = document.cookie.split("; ").find((row) => row.startsWith(`${name}=`));
  return match ? decodeURIComponent(match.split("=")[1] ?? "") : null;
}
function writeCookie(name, value, maxAge) {
  if (typeof document === "undefined") return;
  document.cookie = `${name}=${encodeURIComponent(value)}; path=/; max-age=${maxAge}; SameSite=Lax`;
}
var DockerSidebarInner = ({
  sidebar,
  children,
  keyboardShortcut,
  persistState
}) => {
  const { open, setOpen } = useSidebar();
  React.useEffect(() => {
    if (!persistState) return;
    writeCookie(COOKIE_NAME, String(open), COOKIE_MAX_AGE);
  }, [open, persistState]);
  React.useEffect(() => {
    if (!keyboardShortcut) return;
    const handleKeyDown = (event) => {
      const isModifier = event.metaKey || event.ctrlKey;
      if (isModifier && event.key.toLowerCase() === keyboardShortcut.toLowerCase()) {
        event.preventDefault();
        setOpen((prev) => !prev);
      }
    };
    document.addEventListener("keydown", handleKeyDown);
    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [keyboardShortcut, setOpen]);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    sidebar,
    children
  ] });
};
var DockerSidebar = ({
  defaultOpen = true,
  keyboardShortcut = "b",
  persistState = true,
  sidebar,
  children
}) => {
  const resolvedDefaultOpen = React.useMemo(() => {
    if (!persistState) return defaultOpen;
    const cookie = readCookie(COOKIE_NAME);
    if (cookie === "true") return true;
    if (cookie === "false") return false;
    return defaultOpen;
  }, [defaultOpen, persistState]);
  return /* @__PURE__ */ jsx(SidebarProvider, { defaultOpen: resolvedDefaultOpen, children: /* @__PURE__ */ jsx(
    DockerSidebarInner,
    {
      sidebar,
      keyboardShortcut,
      persistState,
      children
    }
  ) });
};
var AppSidebar = ({
  children,
  header,
  footer,
  className,
  variant
}) => {
  return /* @__PURE__ */ jsxs(Sidebar, { collapsible: "icon", variant, className, children: [
    header && /* @__PURE__ */ jsx(SidebarHeader, { children: header }),
    /* @__PURE__ */ jsx(SidebarContent, { children }),
    footer && /* @__PURE__ */ jsx(SidebarFooter, { children: footer }),
    /* @__PURE__ */ jsx(SidebarRail, {})
  ] });
};
var AppSidebarItem = ({
  icon,
  label,
  detail,
  href,
  isActive,
  children,
  className
}) => {
  const hasChildren = Boolean(children);
  const buttonSize = detail ? "lg" : "default";
  const itemContent = /* @__PURE__ */ jsxs(Fragment, { children: [
    icon,
    detail ? /* @__PURE__ */ jsxs("div", { className: "grid min-w-0 gap-dds-050 leading-tight group-data-[collapsible=icon]:hidden", children: [
      /* @__PURE__ */ jsx("span", { className: "truncate text-label leading-label font-label tracking-label", children: label }),
      /* @__PURE__ */ jsx("span", { className: "truncate text-micro leading-snug tracking-caption text-sidebar-foreground-muted", children: detail })
    ] }) : label && /* @__PURE__ */ jsx("span", { children: label })
  ] });
  if (hasChildren) {
    return /* @__PURE__ */ jsx(Collapsible, { asChild: true, defaultOpen: isActive, className: "group/collapsible", children: /* @__PURE__ */ jsxs(SidebarMenuItem, { children: [
      /* @__PURE__ */ jsx(CollapsibleTrigger, { asChild: true, children: /* @__PURE__ */ jsxs(
        SidebarMenuButton,
        {
          tooltip: label,
          isActive,
          size: buttonSize,
          className: cn(className),
          children: [
            itemContent,
            /* @__PURE__ */ jsx(
              ChevronRight,
              {
                className: cn(
                  "ml-auto h-4 w-4 transition-transform duration-200",
                  "group-data-[state=open]/collapsible:rotate-90"
                )
              }
            )
          ]
        }
      ) }),
      /* @__PURE__ */ jsx(CollapsibleContent, { children: /* @__PURE__ */ jsx(SidebarMenuSub, { children }) })
    ] }) });
  }
  return /* @__PURE__ */ jsx(SidebarMenuItem, { children: /* @__PURE__ */ jsx(
    SidebarMenuButton,
    {
      asChild: Boolean(href),
      tooltip: label,
      isActive,
      size: buttonSize,
      className: cn(className),
      "data-active": isActive ? "true" : void 0,
      children: href ? /* @__PURE__ */ jsx("a", { href, "aria-label": label, children: itemContent }) : itemContent
    }
  ) });
};
var AppSidebarSubItem = ({
  label,
  href,
  isActive,
  className
}) => {
  return /* @__PURE__ */ jsx(SidebarMenuSubItem, { children: /* @__PURE__ */ jsx(
    SidebarMenuSubButton,
    {
      asChild: Boolean(href),
      isActive,
      className: cn(className),
      "data-active": isActive ? "true" : void 0,
      children: href ? /* @__PURE__ */ jsx("a", { href, children: label }) : /* @__PURE__ */ jsx("span", { children: label })
    }
  ) });
};

export {
  DockerSidebar,
  AppSidebar,
  AppSidebarItem,
  AppSidebarSubItem
};
