// @docker/trident/dist/chunk-M27MAPTK.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/sbom-indicator.tsx
import { Check, Minus, X } from "lucide-react";
import { jsx, jsxs } from "react/jsx-runtime";
var STATUS_CONFIG = {
  present: { Icon: Check, label: "SBOM present", colorClass: "text-success" },
  absent: {
    Icon: X,
    label: "SBOM absent",
    colorClass: "text-muted-foreground"
  },
  partial: { Icon: Minus, label: "SBOM partial", colorClass: "text-warning" }
};
function SbomIndicator({ status, className, ...props }) {
  const { Icon, label, colorClass } = STATUS_CONFIG[status];
  return /* @__PURE__ */ jsxs(
    "span",
    {
      "data-slot": "sbom-indicator",
      role: "img",
      "aria-label": label,
      className: cn(
        "inline-flex items-center gap-1 font-mono text-xs",
        colorClass,
        className
      ),
      ...props,
      children: [
        /* @__PURE__ */ jsx(Icon, { className: "size-3", "aria-hidden": true }),
        "SBOM"
      ]
    }
  );
}

export {
  SbomIndicator
};
