// @docker/trident/dist/chunk-K622JLZ6.js
import {
  Separator
} from "./chunk-LWARFSUM.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/app-footer.tsx
import { Rss } from "lucide-react";
import { useId } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var AppFooter = ({
  children,
  showLinks = true,
  enableCookieButton = false,
  className
}) => {
  const year = new Intl.DateTimeFormat("en-US", { year: "numeric" }).format(
    Date.now()
  );
  return /* @__PURE__ */ jsx(
    "footer",
    {
      className: cn(
        "shrink-0 bg-sidebar py-dds-600 text-sidebar-foreground",
        className
      ),
      children: /* @__PURE__ */ jsxs("div", { className: "mx-auto max-w-screen-2xl px-dds-600", children: [
        showLinks && /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsxs("div", { className: "mb-dds-600 flex flex-wrap items-start gap-dds-800", children: [
            /* @__PURE__ */ jsx("span", { className: "text-lg font-bold text-foreground", children: "Docker" }),
            children,
            /* @__PURE__ */ jsx(FooterCompanyColumn, {})
          ] }),
          /* @__PURE__ */ jsx(Separator, { className: "mb-dds-600 bg-sidebar-border" })
        ] }),
        /* @__PURE__ */ jsxs("div", { className: "flex flex-col items-start justify-between gap-dds-400 md:flex-row md:items-center", children: [
          /* @__PURE__ */ jsxs("div", { className: "flex flex-wrap gap-x-dds-400 gap-y-dds-100 text-sm text-muted-foreground", children: [
            /* @__PURE__ */ jsxs("span", { className: "whitespace-nowrap", children: [
              "\xA9 ",
              year,
              " Docker, Inc. All rights reserved."
            ] }),
            /* @__PURE__ */ jsx("span", { "aria-hidden": "true", children: "|" }),
            /* @__PURE__ */ jsx(
              "a",
              {
                href: "https://www.docker.com/legal/docker-terms-service",
                target: "_blank",
                rel: "noopener noreferrer",
                className: "underline underline-offset-4 hover:text-foreground",
                children: "Terms of Service"
              }
            ),
            /* @__PURE__ */ jsx("span", { "aria-hidden": "true", children: "|" }),
            /* @__PURE__ */ jsx(
              "a",
              {
                href: "https://www.docker.com/legal/docker-subscription-service-agreement",
                target: "_blank",
                rel: "noopener noreferrer",
                className: "underline underline-offset-4 hover:text-foreground",
                children: "Subscription Service Agreement"
              }
            ),
            /* @__PURE__ */ jsx("span", { "aria-hidden": "true", children: "|" }),
            /* @__PURE__ */ jsx(
              "a",
              {
                href: "https://www.docker.com/legal/privacy",
                target: "_blank",
                rel: "noopener noreferrer",
                className: "underline underline-offset-4 hover:text-foreground",
                children: "Privacy"
              }
            ),
            /* @__PURE__ */ jsx("span", { "aria-hidden": "true", children: "|" }),
            /* @__PURE__ */ jsx(
              "a",
              {
                href: "https://www.docker.com/legal",
                target: "_blank",
                rel: "noopener noreferrer",
                className: "underline underline-offset-4 hover:text-foreground",
                children: "Legal"
              }
            )
          ] }),
          /* @__PURE__ */ jsx(SocialLinks, {})
        ] }),
        enableCookieButton && /* @__PURE__ */ jsx(
          "button",
          {
            type: "button",
            id: "ot-sdk-btn",
            className: "chromatic-ignore ot-sdk-show-settings mt-dds-400 text-sm text-muted-foreground underline underline-offset-4",
            children: "Cookies Settings"
          }
        )
      ] })
    }
  );
};
var FooterLinkColumn = ({
  heading,
  children,
  className
}) => {
  const uid = useId();
  const id = `footer-${uid}-heading`;
  return /* @__PURE__ */ jsxs(
    "section",
    {
      "aria-labelledby": id,
      className: cn("flex min-w-[8rem] flex-col gap-dds-200", className),
      children: [
        /* @__PURE__ */ jsx("h2", { id, className: "text-sm font-heading text-foreground", children: heading }),
        /* @__PURE__ */ jsx("div", { className: "flex flex-col gap-dds-100 text-sm text-muted-foreground", children })
      ]
    }
  );
};
var COMPANY_LINKS = [
  { label: "About Us", href: "https://www.docker.com/company" },
  { label: "Blog", href: "https://www.docker.com/blog/" },
  { label: "Careers", href: "https://www.docker.com/careers" },
  { label: "Contact Us", href: "https://www.docker.com/company/contact" },
  { label: "Newsroom", href: "https://www.docker.com/company/newsroom" },
  { label: "System Status", href: "https://www.dockerstatus.com" }
];
var FooterCompanyColumn = () => /* @__PURE__ */ jsx(FooterLinkColumn, { heading: "Company", children: COMPANY_LINKS.map(({ label, href }) => /* @__PURE__ */ jsx(
  "a",
  {
    href,
    target: "_blank",
    rel: "noopener noreferrer",
    className: "hover:text-foreground",
    children: label
  },
  href
)) });
var LinkedInIcon = () => /* @__PURE__ */ jsxs(
  "svg",
  {
    "aria-hidden": "true",
    viewBox: "0 0 24 24",
    fill: "currentColor",
    className: "size-4",
    children: [
      /* @__PURE__ */ jsx("path", { d: "M16 8a6 6 0 0 1 6 6v7h-4v-7a2 2 0 0 0-2-2 2 2 0 0 0-2 2v7h-4v-7a6 6 0 0 1 6-6z" }),
      /* @__PURE__ */ jsx("rect", { width: "4", height: "12", x: "2", y: "9" }),
      /* @__PURE__ */ jsx("circle", { cx: "4", cy: "4", r: "2" })
    ]
  }
);
var XIcon = () => /* @__PURE__ */ jsx(
  "svg",
  {
    "aria-hidden": "true",
    viewBox: "0 0 24 24",
    fill: "currentColor",
    className: "size-4",
    children: /* @__PURE__ */ jsx("path", { d: "M18.244 2.25h3.308l-7.227 8.26 8.502 11.24H16.17l-4.714-6.231-5.401 6.231H2.744l7.73-8.835L1.254 2.25H8.08l4.713 6.231zm-1.161 17.52h1.833L7.084 4.126H5.117z" })
  }
);
var YouTubeIcon = () => /* @__PURE__ */ jsxs(
  "svg",
  {
    "aria-hidden": "true",
    viewBox: "0 0 24 24",
    fill: "currentColor",
    className: "size-4",
    children: [
      /* @__PURE__ */ jsx("path", { d: "M22.54 6.42a2.78 2.78 0 0 0-1.95-1.96C18.88 4 12 4 12 4s-6.88 0-8.59.46a2.78 2.78 0 0 0-1.95 1.96A29 29 0 0 0 1 12a29 29 0 0 0 .46 5.58 2.78 2.78 0 0 0 1.95 1.95C5.12 20 12 20 12 20s6.88 0 8.59-.47a2.78 2.78 0 0 0 1.95-1.95A29 29 0 0 0 23 12a29 29 0 0 0-.46-5.58z" }),
      /* @__PURE__ */ jsx("polygon", { points: "9.75 15.02 15.5 12 9.75 8.98 9.75 15.02" })
    ]
  }
);
var SOCIAL_LINKS = [
  {
    label: "Visit our LinkedIn page",
    href: "https://www.linkedin.com/company/docker",
    Icon: LinkedInIcon
  },
  { label: "Visit our X page", href: "https://x.com/docker", Icon: XIcon },
  {
    label: "Visit our YouTube page",
    href: "https://www.youtube.com/user/dockerrun",
    Icon: YouTubeIcon
  },
  {
    label: "Visit our RSS feed",
    href: "https://www.docker.com/feed/",
    Icon: () => /* @__PURE__ */ jsx(Rss, { "aria-hidden": "true", className: "size-4" })
  }
];
var SocialLinks = () => /* @__PURE__ */ jsx("div", { className: "flex gap-dds-100", children: SOCIAL_LINKS.map(({ label, href, Icon }) => /* @__PURE__ */ jsx(
  "a",
  {
    href,
    target: "_blank",
    rel: "noopener noreferrer",
    "aria-label": label,
    className: "rounded p-dds-200 text-muted-foreground transition-colors hover:text-foreground",
    children: /* @__PURE__ */ jsx(Icon, {})
  },
  href
)) });

export {
  AppFooter,
  FooterLinkColumn
};
