// @docker/trident/dist/chunk-R2QGPDNO.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/inline-segmented-toggle.tsx
import { jsx } from "react/jsx-runtime";
function InlineSegmentedToggle({
  options,
  value,
  onChange,
  className
}) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      role: "radiogroup",
      className: cn(
        "flex items-center gap-px rounded-[3px] border border-border-subtle bg-hover-overlay p-[2px]",
        className
      ),
      children: options.map((opt) => /* @__PURE__ */ jsx(
        InlineSegmentedToggleItem,
        {
          active: opt.value === value,
          onClick: () => onChange(opt.value),
          children: opt.label
        },
        opt.value
      ))
    }
  );
}
function InlineSegmentedToggleItem({
  active,
  onClick,
  children
}) {
  return /* @__PURE__ */ jsx(
    "button",
    {
      type: "button",
      role: "radio",
      "aria-checked": active,
      onClick: (e) => {
        e.stopPropagation();
        onClick();
      },
      className: cn(
        "rounded-[2px] px-1.5 py-[1px] text-[10px] font-label tracking-[0.04em] uppercase transition-colors",
        active ? "bg-primary text-primary-foreground" : "text-muted-foreground hover:bg-hover-overlay hover:text-foreground"
      ),
      children
    }
  );
}

export {
  InlineSegmentedToggle
};
