// @docker/trident/dist/chunk-EYCO2MID.js
import {
  Input
} from "./chunk-5CQB3X63.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/secret-input.tsx
import { Eye, EyeOff } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function SecretInput({ className, ...props }) {
  const [visible, setVisible] = React.useState(false);
  return /* @__PURE__ */ jsxs("div", { className: cn("relative flex items-center", className), children: [
    /* @__PURE__ */ jsx(
      Input,
      {
        type: visible ? "text" : "password",
        className: "pr-10 font-mono",
        ...props
      }
    ),
    /* @__PURE__ */ jsx(
      Button,
      {
        type: "button",
        variant: "ghost",
        size: "icon",
        "aria-label": visible ? "Hide secret" : "Show secret",
        className: "absolute right-1 h-7 w-7 text-muted-foreground hover:text-foreground",
        onClick: () => setVisible((v) => !v),
        children: visible ? /* @__PURE__ */ jsx(EyeOff, { className: "h-4 w-4" }) : /* @__PURE__ */ jsx(Eye, { className: "h-4 w-4" })
      }
    )
  ] });
}

export {
  SecretInput
};
