// @docker/trident/dist/chunk-ND3FFSQH.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/status-indicator.tsx
import { cva } from "class-variance-authority";
import {
  AlertTriangle,
  Circle,
  CircleCheck,
  Info,
  Minus,
  OctagonX,
  X
} from "lucide-react";
import { Slot } from "radix-ui";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function CreatedIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    Circle,
    {
      className: cn("size-4 text-muted-foreground opacity-40", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function RunningIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    CircleCheck,
    {
      className: cn("size-4 text-success", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function PausedIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    Minus,
    {
      className: cn("size-4 text-muted-foreground", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function ExitedIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    Circle,
    {
      className: cn("size-4 text-muted-foreground opacity-20", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function DeadIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    X,
    {
      className: cn("size-4 text-muted-foreground", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function CompletedIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    CircleCheck,
    {
      className: cn("size-4 text-success", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function MixedIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    Info,
    {
      className: cn("size-4 text-primary", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function ErrorIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    OctagonX,
    {
      className: cn("size-4 text-destructive", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function WarningIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    AlertTriangle,
    {
      className: cn("size-4 text-warning", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function OtherIcon({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    AlertTriangle,
    {
      className: cn("size-4 text-warning", className),
      "aria-hidden": !props["aria-label"],
      ...props
    }
  );
}
function RestartingIcon({ className, ...props }) {
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      viewBox: "0 0 24 24",
      fill: "none",
      stroke: "currentColor",
      strokeWidth: "2",
      strokeLinecap: "round",
      strokeLinejoin: "round",
      className: cn("size-4 animate-spin text-primary", className),
      "aria-hidden": !props["aria-label"],
      ...props,
      children: [
        !props["aria-label"] && /* @__PURE__ */ jsx("title", { children: "Restarting" }),
        /* @__PURE__ */ jsx("path", { d: "M21 12a9 9 0 1 1-6.219-8.56" })
      ]
    }
  );
}
function RemovingIcon({ className, ...props }) {
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      viewBox: "0 0 24 24",
      fill: "none",
      stroke: "currentColor",
      strokeWidth: "2",
      strokeLinecap: "round",
      strokeLinejoin: "round",
      className: cn("size-4 animate-spin text-destructive", className),
      "aria-hidden": !props["aria-label"],
      ...props,
      children: [
        !props["aria-label"] && /* @__PURE__ */ jsx("title", { children: "Removing" }),
        /* @__PURE__ */ jsx("path", { d: "M21 12a9 9 0 1 1-6.219-8.56" })
      ]
    }
  );
}
var CreatingIcon = RestartingIcon;
var StartingIcon = RestartingIcon;
var STATUS_LABELS = {
  running: "Running",
  stopped: "Stopped",
  warning: "Warning",
  error: "Error",
  starting: "Starting"
};
var dotVariants = cva("shrink-0 rounded-full", {
  variants: {
    status: {
      running: "bg-status-running",
      stopped: "bg-status-stopped",
      warning: "bg-status-paused",
      error: "bg-status-error",
      starting: "animate-pulse bg-primary"
    },
    size: {
      sm: "size-2",
      md: "size-2.5",
      lg: "size-3"
    }
  },
  defaultVariants: {
    status: "stopped",
    size: "md"
  }
});
var GLOW_STATES = /* @__PURE__ */ new Set(["running", "warning", "error"]);
function StatusIndicator({
  status,
  size,
  showLabel = false,
  label,
  glow = true,
  className,
  asChild = false,
  children,
  ...props
}) {
  const resolvedLabel = label ?? STATUS_LABELS[status];
  const ariaLabel = `Status: ${resolvedLabel.toLowerCase()}`;
  const shouldGlow = glow && GLOW_STATES.has(status);
  const sharedProps = {
    "data-slot": "status-indicator",
    "data-status": status,
    className: cn("inline-flex items-center gap-1.5", className),
    ...props
  };
  const dot = /* @__PURE__ */ jsx(
    "span",
    {
      role: "img",
      "aria-label": ariaLabel,
      "data-glow": shouldGlow ? "" : void 0,
      className: cn(
        dotVariants({ status, size }),
        shouldGlow && status === "running" && "dark:shadow-[var(--dds-shadow-glow-running)]",
        shouldGlow && status === "warning" && "dark:shadow-[var(--dds-shadow-glow-warning)]",
        shouldGlow && status === "error" && "dark:shadow-[var(--dds-shadow-glow-error)]",
        shouldGlow && "high-contrast:shadow-none high-contrast:outline-2 high-contrast:outline-offset-1 high-contrast:outline-current"
      )
    }
  );
  const labelEl = showLabel ? /* @__PURE__ */ jsx("span", { "aria-hidden": "true", className: "text-sm leading-none text-foreground", children: resolvedLabel }) : null;
  if (asChild) {
    const child = React.Children.only(children);
    if (!React.isValidElement(child)) return null;
    return /* @__PURE__ */ jsx(Slot.Root, { ...sharedProps, children: React.cloneElement(child, {}, dot, labelEl) });
  }
  return /* @__PURE__ */ jsxs("span", { ...sharedProps, children: [
    dot,
    labelEl
  ] });
}

export {
  CreatedIcon,
  RunningIcon,
  PausedIcon,
  ExitedIcon,
  DeadIcon,
  CompletedIcon,
  MixedIcon,
  ErrorIcon,
  WarningIcon,
  OtherIcon,
  RestartingIcon,
  RemovingIcon,
  CreatingIcon,
  StartingIcon,
  dotVariants,
  StatusIndicator
};
