// @docker/trident/dist/chunk-R5XEJKHJ.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/checkbox.tsx
import { cva } from "class-variance-authority";
import { Check, Minus } from "lucide-react";
import { Checkbox as CheckboxPrimitive } from "radix-ui";
import { jsx, jsxs } from "react/jsx-runtime";
var checkboxVariants = cva(
  // Note: `data-checked:` only matches `data-state="checked"`. The indeterminate
  // state is `data-state="indeterminate"` and must be styled explicitly so the
  // tri-state checkbox stays visible (otherwise the indicator paints
  // primary-foreground on an unstyled bg → white-on-white in HC mode).
  "peer relative flex shrink-0 items-center justify-center rounded-sm border border-input transition-shadow ease-default outline-none group-has-disabled/field:opacity-50 focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:border-destructive aria-invalid:ring-3 aria-invalid:ring-destructive-subtle aria-invalid:aria-checked:border-primary data-[state=indeterminate]:border-primary data-[state=indeterminate]:bg-primary data-[state=indeterminate]:text-primary-foreground dark:bg-input-bg dark:data-[state=indeterminate]:bg-primary data-checked:border-primary data-checked:bg-primary data-checked:text-primary-foreground dark:data-checked:bg-primary",
  {
    variants: {
      size: {
        sm: "size-3.5 after:absolute after:-inset-x-3.5 after:-inset-y-2.5",
        default: "size-4 after:absolute after:-inset-x-3 after:-inset-y-2",
        lg: "size-5 after:absolute after:-inset-x-2.5 after:-inset-y-1.5"
      }
    },
    defaultVariants: {
      size: "default"
    }
  }
);
var indicatorSizeClasses = {
  sm: "[&>svg]:size-3",
  default: "[&>svg]:size-3.5",
  lg: "[&>svg]:size-4"
};
function Checkbox({ className, size = "default", ...props }) {
  return /* @__PURE__ */ jsx(
    CheckboxPrimitive.Root,
    {
      "data-slot": "checkbox",
      className: cn(checkboxVariants({ size }), className),
      ...props,
      children: /* @__PURE__ */ jsxs(
        CheckboxPrimitive.Indicator,
        {
          "data-slot": "checkbox-indicator",
          className: cn(
            "group/checkbox-indicator grid place-content-center text-current transition-none",
            indicatorSizeClasses[size ?? "default"]
          ),
          children: [
            /* @__PURE__ */ jsx(Check, { className: "hidden group-data-[state=checked]/checkbox-indicator:block" }),
            /* @__PURE__ */ jsx(Minus, { className: "hidden group-data-[state=indeterminate]/checkbox-indicator:block" })
          ]
        }
      )
    }
  );
}

export {
  checkboxVariants,
  Checkbox
};
