// @docker/trident/dist/chunk-7NIC64DL.js
import {
  TransferListPanel
} from "./chunk-JY5DMKQ2.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/transfer-list.tsx
import {
  ChevronLeft,
  ChevronRight,
  ChevronsLeft,
  ChevronsRight
} from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var TransferList = ({
  available,
  selected,
  onChange,
  availableLabel = "Available",
  selectedLabel = "Selected",
  availableDescription,
  selectedDescription,
  searchable = true,
  searchPlaceholder = "Search\u2026",
  showMoveAll = true,
  renderItem,
  disabled = false,
  className
}) => {
  const [availableChecked, setAvailableChecked] = React.useState(
    /* @__PURE__ */ new Set()
  );
  const [selectedChecked, setSelectedChecked] = React.useState(
    /* @__PURE__ */ new Set()
  );
  const [availableSearch, setAvailableSearch] = React.useState("");
  const [selectedSearch, setSelectedSearch] = React.useState("");
  const filteredAvailable = React.useMemo(
    () => available.filter(
      (i) => i.label.toLowerCase().includes(availableSearch.toLowerCase())
    ),
    [available, availableSearch]
  );
  const filteredSelected = React.useMemo(
    () => selected.filter(
      (i) => i.label.toLowerCase().includes(selectedSearch.toLowerCase())
    ),
    [selected, selectedSearch]
  );
  const availableCheckedCount = filteredAvailable.filter(
    (i) => availableChecked.has(i.id)
  ).length;
  const selectedCheckedCount = filteredSelected.filter(
    (i) => selectedChecked.has(i.id)
  ).length;
  const selectedIds = React.useMemo(
    () => new Set(selected.map((i) => i.id)),
    [selected]
  );
  const moveCheckedToSelected = () => {
    const toMove = filteredAvailable.filter(
      (i) => availableChecked.has(i.id) && !i.disabled
    );
    if (toMove.length === 0) return;
    const toMoveIds = new Set(toMove.map((i) => i.id));
    onChange([...selected, ...toMove]);
    setAvailableChecked(
      new Set([...availableChecked].filter((id) => !toMoveIds.has(id)))
    );
  };
  const moveAllToSelected = () => {
    const toMove = filteredAvailable.filter(
      (i) => !i.disabled && !selectedIds.has(i.id)
    );
    if (toMove.length === 0) return;
    const toMoveIds = new Set(toMove.map((i) => i.id));
    onChange([...selected, ...toMove]);
    setAvailableChecked(
      new Set([...availableChecked].filter((id) => !toMoveIds.has(id)))
    );
  };
  const moveCheckedToAvailable = () => {
    const toRemoveIds = new Set(
      filteredSelected.filter((i) => selectedChecked.has(i.id) && !i.disabled).map((i) => i.id)
    );
    if (toRemoveIds.size === 0) return;
    onChange(selected.filter((i) => !toRemoveIds.has(i.id)));
    setSelectedChecked(
      new Set([...selectedChecked].filter((id) => !toRemoveIds.has(id)))
    );
  };
  const moveAllToAvailable = () => {
    const toRemoveIds = new Set(
      filteredSelected.filter((i) => !i.disabled).map((i) => i.id)
    );
    if (toRemoveIds.size === 0) return;
    onChange(selected.filter((i) => !toRemoveIds.has(i.id)));
    setSelectedChecked(
      new Set([...selectedChecked].filter((id) => !toRemoveIds.has(id)))
    );
  };
  const handleAvailableSelectAll = (checked) => {
    setAvailableChecked(
      checked ? new Set(filteredAvailable.filter((i) => !i.disabled).map((i) => i.id)) : /* @__PURE__ */ new Set()
    );
  };
  const handleSelectedSelectAll = (checked) => {
    setSelectedChecked(
      checked ? new Set(filteredSelected.filter((i) => !i.disabled).map((i) => i.id)) : /* @__PURE__ */ new Set()
    );
  };
  const handleAvailableCheck = (id, checked) => {
    setAvailableChecked((prev) => {
      const next = new Set(prev);
      if (checked) next.add(id);
      else next.delete(id);
      return next;
    });
  };
  const handleSelectedCheck = (id, checked) => {
    setSelectedChecked((prev) => {
      const next = new Set(prev);
      if (checked) next.add(id);
      else next.delete(id);
      return next;
    });
  };
  const moveRightDisabled = disabled || availableCheckedCount === 0;
  const moveAllRightDisabled = disabled || filteredAvailable.filter((i) => !i.disabled && !selectedIds.has(i.id)).length === 0;
  const moveLeftDisabled = disabled || selectedCheckedCount === 0;
  const moveAllLeftDisabled = disabled || filteredSelected.filter((i) => !i.disabled).length === 0;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "transfer-list",
      className: cn("flex items-stretch gap-3", className),
      children: [
        /* @__PURE__ */ jsx(
          TransferListPanel,
          {
            label: availableLabel,
            description: availableDescription,
            items: available,
            checkedIds: availableChecked,
            onCheckedChange: handleAvailableCheck,
            onSelectAll: handleAvailableSelectAll,
            searchable,
            searchPlaceholder,
            searchQuery: availableSearch,
            onSearchChange: setAvailableSearch,
            renderItem,
            side: "available",
            disabled,
            filteredItems: filteredAvailable
          }
        ),
        /* @__PURE__ */ jsxs("div", { className: "flex shrink-0 flex-col items-center justify-center gap-1.5", children: [
          showMoveAll && /* @__PURE__ */ jsx(
            Button,
            {
              variant: "outline",
              size: "icon-sm",
              onClick: moveAllToSelected,
              disabled: moveAllRightDisabled,
              "aria-label": `Move all ${availableLabel} items to ${selectedLabel}`,
              title: "Move all \u2192",
              children: /* @__PURE__ */ jsx(ChevronsRight, { className: "size-4", "aria-hidden": "true" })
            }
          ),
          /* @__PURE__ */ jsx(
            Button,
            {
              variant: "outline",
              size: "icon-sm",
              onClick: moveCheckedToSelected,
              disabled: moveRightDisabled,
              "aria-label": `Move checked ${availableLabel} items to ${selectedLabel}`,
              title: "Move selected \u2192",
              onKeyDown: (e) => {
                if (e.key === "Enter") moveCheckedToSelected();
              },
              children: /* @__PURE__ */ jsx(ChevronRight, { className: "size-4", "aria-hidden": "true" })
            }
          ),
          /* @__PURE__ */ jsx(
            Button,
            {
              variant: "outline",
              size: "icon-sm",
              onClick: moveCheckedToAvailable,
              disabled: moveLeftDisabled,
              "aria-label": `Move checked ${selectedLabel} items back to ${availableLabel}`,
              title: "\u2190 Move selected",
              onKeyDown: (e) => {
                if (e.key === "Enter") moveCheckedToAvailable();
              },
              children: /* @__PURE__ */ jsx(ChevronLeft, { className: "size-4", "aria-hidden": "true" })
            }
          ),
          showMoveAll && /* @__PURE__ */ jsx(
            Button,
            {
              variant: "outline",
              size: "icon-sm",
              onClick: moveAllToAvailable,
              disabled: moveAllLeftDisabled,
              "aria-label": `Move all ${selectedLabel} items back to ${availableLabel}`,
              title: "\u2190 Move all",
              children: /* @__PURE__ */ jsx(ChevronsLeft, { className: "size-4", "aria-hidden": "true" })
            }
          )
        ] }),
        /* @__PURE__ */ jsx(
          TransferListPanel,
          {
            label: selectedLabel,
            description: selectedDescription,
            items: selected,
            checkedIds: selectedChecked,
            onCheckedChange: handleSelectedCheck,
            onSelectAll: handleSelectedSelectAll,
            searchable,
            searchPlaceholder,
            searchQuery: selectedSearch,
            onSearchChange: setSelectedSearch,
            renderItem,
            side: "selected",
            disabled,
            filteredItems: filteredSelected
          }
        )
      ]
    }
  );
};

export {
  TransferList
};
