// @docker/trident/dist/chunk-MPTXQIBL.js
import {
  ButtonGroup
} from "./chunk-VGLY7RTA.js";
import {
  formatLineNumber,
  gutterChar,
  gutterCharClasses,
  lineClasses,
  parseDiff,
  toSplitRows
} from "./chunk-YU5N53H7.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/code-diff-viewer.tsx
import { AlignLeftIcon, ChevronDownIcon, Columns2Icon } from "lucide-react";
import { useCallback, useState } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var LineNumberCell = ({
  value,
  show
}) => {
  if (!show) return null;
  return /* @__PURE__ */ jsx("span", { className: "min-w-[3ch] pr-3 text-right font-mono text-xs text-muted-foreground select-none", children: value ?? "" });
};
var UnifiedRow = ({
  line,
  showLineNumbers
}) => {
  const char = gutterChar[line.type] ?? " ";
  const charCls = gutterCharClasses[line.type] ?? "";
  return /* @__PURE__ */ jsxs("div", { className: cn("flex min-w-0", lineClasses[line.type] ?? ""), children: [
    /* @__PURE__ */ jsxs("div", { className: "flex shrink-0 items-center gap-1 border-r border-border px-2 select-none", children: [
      /* @__PURE__ */ jsx(
        LineNumberCell,
        {
          value: formatLineNumber(line.oldLineNumber) || void 0,
          show: showLineNumbers
        }
      ),
      /* @__PURE__ */ jsx(
        LineNumberCell,
        {
          value: formatLineNumber(line.newLineNumber) || void 0,
          show: showLineNumbers
        }
      ),
      /* @__PURE__ */ jsx("span", { className: cn("w-3 font-mono text-xs", charCls), children: char })
    ] }),
    /* @__PURE__ */ jsx("pre", { className: "min-w-0 flex-1 overflow-x-auto px-3 py-0 font-mono text-sm leading-5", children: /* @__PURE__ */ jsx("code", { children: line.content }) })
  ] });
};
var SplitRowPair = ({
  row,
  showLineNumbers
}) => {
  const renderSide = (line, side) => {
    const type = line?.type ?? "context";
    const char = line ? gutterChar[type] ?? " " : " ";
    const charCls = line ? gutterCharClasses[type] ?? "" : "";
    const lineNumValue = side === "left" ? formatLineNumber(line?.oldLineNumber) || void 0 : formatLineNumber(line?.newLineNumber) || void 0;
    return (
      // tabIndex={0} is required: WCAG scrollable-region-focusable mandates
      // keyboard access for overflow-x-auto regions
      /* @__PURE__ */ jsxs(
        "div",
        {
          tabIndex: 0,
          className: cn(
            "flex min-w-0 flex-1 overflow-x-auto",
            line ? lineClasses[type] ?? "" : "bg-background"
          ),
          children: [
            /* @__PURE__ */ jsxs("div", { className: "flex shrink-0 items-center gap-1 border-r border-border px-2 select-none", children: [
              /* @__PURE__ */ jsx(LineNumberCell, { value: lineNumValue, show: showLineNumbers }),
              /* @__PURE__ */ jsx("span", { className: cn("w-3 font-mono text-xs", charCls), children: char })
            ] }),
            /* @__PURE__ */ jsx("pre", { className: "min-w-0 flex-1 px-3 py-0 font-mono text-sm leading-5", children: /* @__PURE__ */ jsx("code", { children: line?.content ?? "" }) })
          ]
        }
      )
    );
  };
  return /* @__PURE__ */ jsxs("div", { className: "flex divide-x divide-border", children: [
    renderSide(row.left, "left"),
    renderSide(row.right, "right")
  ] });
};
var ExpanderRow = ({
  hiddenCount,
  onExpand
}) => /* @__PURE__ */ jsx("div", { className: "flex items-center justify-center border-y border-border bg-diff-hunk-bg py-0.5", children: /* @__PURE__ */ jsxs(
  Button,
  {
    variant: "ghost",
    size: "xs",
    className: "gap-1 text-muted-foreground hover:text-foreground",
    onClick: onExpand,
    children: [
      /* @__PURE__ */ jsx(ChevronDownIcon, { className: "size-3" }),
      "Show ",
      hiddenCount,
      " more line",
      hiddenCount !== 1 ? "s" : ""
    ]
  }
) });
var HunkBlock = ({
  hunk,
  mode,
  showLineNumbers,
  expandContext
}) => {
  const [expandedRuns, setExpandedRuns] = useState(/* @__PURE__ */ new Set());
  const segments = [];
  let runIndex = 0;
  let i = 0;
  const lines = hunk.lines;
  while (i < lines.length) {
    if (lines[i].type === "context") {
      const start = i;
      while (i < lines.length && lines[i].type === "context") i++;
      segments.push({
        kind: "context-run",
        lines: lines.slice(start, i),
        runIndex: runIndex++
      });
    } else {
      const start = i;
      while (i < lines.length && lines[i].type !== "context") i++;
      segments.push({ kind: "change", lines: lines.slice(start, i) });
    }
  }
  const handleExpand = useCallback((ri) => {
    setExpandedRuns((prev) => {
      const next = new Set(prev);
      next.add(ri);
      return next;
    });
  }, []);
  const renderLines = (diffLines) => {
    if (mode === "unified") {
      return diffLines.map((line) => {
        const key = `u-${line.type}-${line.oldLineNumber ?? "x"}-${line.newLineNumber ?? "x"}`;
        return /* @__PURE__ */ jsx(UnifiedRow, { line, showLineNumbers }, key);
      });
    }
    const rows = toSplitRows(diffLines);
    return rows.map((row) => {
      const leftKey = row.left ? `${row.left.type}-${row.left.oldLineNumber}` : "null";
      const rightKey = row.right ? `${row.right.type}-${row.right.newLineNumber}` : "null";
      const key = `s-${leftKey}-${rightKey}`;
      return /* @__PURE__ */ jsx(SplitRowPair, { row, showLineNumbers }, key);
    });
  };
  return /* @__PURE__ */ jsx(Fragment, { children: segments.map((seg, si) => {
    const segKey = seg.kind === "change" ? `change-${si}` : `ctx-${seg.runIndex}`;
    if (seg.kind === "change") {
      return /* @__PURE__ */ jsx("div", { "data-testid": "diff-change-block", children: renderLines(seg.lines) }, segKey);
    }
    const isFirst = si === 0;
    const isLast = si === segments.length - 1;
    const isInterior = !isFirst && !isLast;
    const runLines = seg.lines;
    const isExpanded = expandedRuns.has(seg.runIndex);
    if (isInterior && runLines.length > expandContext * 2 && !isExpanded) {
      const head = runLines.slice(0, expandContext);
      const tail = runLines.slice(runLines.length - expandContext);
      const hidden = runLines.length - expandContext * 2;
      return /* @__PURE__ */ jsxs("div", { "data-testid": "diff-context-collapsed", children: [
        renderLines(head),
        /* @__PURE__ */ jsx(
          ExpanderRow,
          {
            hiddenCount: hidden,
            onExpand: () => handleExpand(seg.runIndex)
          }
        ),
        renderLines(tail)
      ] }, segKey);
    }
    if ((isFirst || isLast) && runLines.length > expandContext && !isExpanded) {
      const visible = isFirst ? runLines.slice(runLines.length - expandContext) : runLines.slice(0, expandContext);
      const hidden = runLines.length - expandContext;
      return /* @__PURE__ */ jsxs("div", { "data-testid": "diff-context-edge", children: [
        isFirst && /* @__PURE__ */ jsx(
          ExpanderRow,
          {
            hiddenCount: hidden,
            onExpand: () => handleExpand(seg.runIndex)
          }
        ),
        renderLines(visible),
        isLast && /* @__PURE__ */ jsx(
          ExpanderRow,
          {
            hiddenCount: hidden,
            onExpand: () => handleExpand(seg.runIndex)
          }
        )
      ] }, segKey);
    }
    return /* @__PURE__ */ jsx("div", { "data-testid": "diff-context-run", children: renderLines(runLines) }, segKey);
  }) });
};
var CodeDiffViewer = ({
  original,
  modified,
  language,
  originalLabel = "Before",
  modifiedLabel = "After",
  mode: modeProp = "split",
  showLineNumbers = true,
  expandContext = 3,
  filename,
  className
}) => {
  const [mode, setMode] = useState(modeProp);
  const parsed = parseDiff(original, modified);
  if (!parsed.hasChanges) {
    return /* @__PURE__ */ jsx(
      "div",
      {
        "data-slot": "code-diff-viewer",
        className: cn(
          "overflow-hidden rounded border border-border bg-muted font-mono text-sm",
          className
        ),
        children: /* @__PURE__ */ jsx("div", { className: "flex items-center justify-center py-8 text-muted-foreground", children: "No changes" })
      }
    );
  }
  const headerLeft = filename ? /* @__PURE__ */ jsx("span", { className: "truncate font-mono text-xs text-foreground", children: filename }) : /* @__PURE__ */ jsxs("span", { className: "flex items-center gap-2 text-xs text-muted-foreground", children: [
    /* @__PURE__ */ jsx("span", { children: originalLabel }),
    /* @__PURE__ */ jsx("span", { "aria-hidden": true, children: "\u2192" }),
    /* @__PURE__ */ jsx("span", { children: modifiedLabel }),
    language && /* @__PURE__ */ jsx("span", { className: "rounded bg-muted px-1.5 py-0.5 text-xs capitalize", children: language })
  ] });
  const splitColumnHeaders = mode === "split" ? /* @__PURE__ */ jsxs("div", { className: "flex divide-x divide-border border-b border-border bg-muted text-xs text-muted-foreground", children: [
    /* @__PURE__ */ jsx("div", { className: "flex-1 px-3 py-1", children: originalLabel }),
    /* @__PURE__ */ jsx("div", { className: "flex-1 px-3 py-1", children: modifiedLabel })
  ] }) : null;
  return /* @__PURE__ */ jsxs(
    "section",
    {
      "data-slot": "code-diff-viewer",
      "aria-label": filename ? `Code diff: ${filename}` : "Code diff",
      className: cn(
        "overflow-hidden rounded border border-border bg-background font-mono text-sm",
        className
      ),
      children: [
        /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-between gap-2 border-b border-border bg-muted px-3 py-2", children: [
          headerLeft,
          /* @__PURE__ */ jsxs(ButtonGroup, { children: [
            /* @__PURE__ */ jsxs(
              Button,
              {
                variant: "outline",
                size: "xs",
                "aria-pressed": mode === "split",
                "data-testid": "mode-split",
                onClick: () => setMode("split"),
                className: cn(mode === "split" && "bg-background shadow-none"),
                children: [
                  /* @__PURE__ */ jsx(Columns2Icon, {}),
                  "Split"
                ]
              }
            ),
            /* @__PURE__ */ jsxs(
              Button,
              {
                variant: "outline",
                size: "xs",
                "aria-pressed": mode === "unified",
                "data-testid": "mode-unified",
                onClick: () => setMode("unified"),
                className: cn(mode === "unified" && "bg-background shadow-none"),
                children: [
                  /* @__PURE__ */ jsx(AlignLeftIcon, {}),
                  "Unified"
                ]
              }
            )
          ] })
        ] }),
        splitColumnHeaders,
        /* @__PURE__ */ jsx("div", { className: "divide-y divide-border", "data-testid": "diff-body", children: parsed.hunks.map((hunk) => /* @__PURE__ */ jsx(
          HunkBlock,
          {
            hunk,
            mode,
            showLineNumbers,
            expandContext
          },
          hunk.header
        )) })
      ]
    }
  );
};

export {
  CodeDiffViewer
};
