// @docker/trident/dist/chunk-RC6ZLLPR.js
import {
  fileMatchesAccept,
  formatBytes,
  isImageFile
} from "./chunk-GPVFILLR.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/file-upload.tsx
import { cva } from "class-variance-authority";
import {
  FileIcon,
  ImageIcon,
  Loader2Icon,
  UploadCloudIcon,
  XIcon
} from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var dropZoneVariants = cva(
  [
    "relative flex w-full cursor-pointer flex-col items-center justify-center rounded border-2 border-dashed",
    "transition-colors duration-150",
    "focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none",
    "aria-disabled:pointer-events-none aria-disabled:cursor-not-allowed aria-disabled:opacity-50"
  ],
  {
    variants: {
      variant: {
        default: "border-border bg-hover-overlay hover:bg-hover-overlay",
        compact: "border-border bg-hover-overlay hover:bg-hover-overlay"
      },
      size: {
        sm: "gap-1 px-4 py-4 text-sm",
        default: "gap-2 px-6 py-8 text-sm",
        lg: "gap-3 px-8 py-12 text-base"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
);
var FilePreviewItem = ({
  file,
  showThumbnail,
  onRemove,
  disabled
}) => {
  const [objectUrl, setObjectUrl] = React.useState(null);
  React.useEffect(() => {
    if (!showThumbnail || !isImageFile(file)) return;
    const url = URL.createObjectURL(file);
    setObjectUrl(url);
    return () => URL.revokeObjectURL(url);
  }, [file, showThumbnail]);
  return /* @__PURE__ */ jsxs(
    "li",
    {
      "data-slot": "file-upload-item",
      className: "flex items-center gap-3 rounded border border-border bg-background px-3 py-2 text-sm",
      children: [
        showThumbnail && objectUrl ? /* @__PURE__ */ jsx(
          "img",
          {
            src: objectUrl,
            alt: file.name,
            className: "size-8 shrink-0 rounded object-cover"
          }
        ) : /* @__PURE__ */ jsx("span", { className: "flex size-8 shrink-0 items-center justify-center rounded bg-muted text-muted-foreground", children: isImageFile(file) ? /* @__PURE__ */ jsx(ImageIcon, { className: "size-4" }) : /* @__PURE__ */ jsx(FileIcon, { className: "size-4" }) }),
        /* @__PURE__ */ jsxs("div", { className: "min-w-0 flex-1", children: [
          /* @__PURE__ */ jsx("p", { className: "truncate font-label text-foreground", children: file.name }),
          /* @__PURE__ */ jsx("p", { className: "text-xs text-muted-foreground", children: formatBytes(file.size) })
        ] }),
        /* @__PURE__ */ jsx(
          "button",
          {
            type: "button",
            onClick: onRemove,
            disabled,
            "aria-label": `Remove ${file.name}`,
            className: cn(
              "flex size-6 shrink-0 items-center justify-center rounded text-muted-foreground",
              "hover:bg-muted hover:text-foreground",
              "focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none",
              "disabled:pointer-events-none disabled:opacity-50",
              "transition-colors"
            ),
            children: /* @__PURE__ */ jsx(XIcon, { className: "size-3.5" })
          }
        )
      ]
    }
  );
};
var FileUpload = ({
  accept,
  maxSize = 5 * 1024 * 1024,
  maxFiles = 1,
  value,
  onFilesChange,
  onError,
  uploading = false,
  disabled = false,
  className,
  label,
  description,
  variant = "default",
  size = "default"
}) => {
  const multiple = maxFiles > 1;
  const inputRef = React.useRef(null);
  const [dragOver, setDragOver] = React.useState(false);
  const [internalFiles, setInternalFiles] = React.useState([]);
  const [error, setError] = React.useState(null);
  const files = value ?? internalFiles;
  const showThumbnails = Boolean(accept?.includes("image"));
  const validateAndAdd = React.useCallback(
    (incoming) => {
      const emitError = (msg) => {
        setError(msg);
        onError?.(msg);
      };
      if (accept) {
        const invalid = incoming.find((f) => !fileMatchesAccept(f, accept));
        if (invalid) {
          emitError(`Invalid file type: ${invalid.name}`);
          return;
        }
      }
      const tooBig = incoming.find((f) => f.size > maxSize);
      if (tooBig) {
        emitError(
          `File too large: ${tooBig.name} (${formatBytes(tooBig.size)} \u2014 max ${formatBytes(maxSize)})`
        );
        return;
      }
      const combined = [...files, ...incoming].slice(0, maxFiles);
      if (files.length + incoming.length > maxFiles) {
        emitError(`Too many files \u2014 max ${maxFiles}`);
      } else {
        setError(null);
      }
      if (!value) setInternalFiles(combined);
      onFilesChange?.(combined);
    },
    [accept, maxSize, maxFiles, files, value, onFilesChange, onError]
  );
  const handleInputChange = (e) => {
    const picked = Array.from(e.target.files ?? []);
    if (picked.length > 0) validateAndAdd(picked);
    e.target.value = "";
  };
  const openPicker = () => {
    if (disabled || uploading) return;
    inputRef.current?.click();
  };
  const handleKeyDown = (e) => {
    if (e.key === "Enter" || e.key === " ") {
      e.preventDefault();
      openPicker();
    }
  };
  const handleDragOver = (e) => {
    e.preventDefault();
    if (!disabled && !uploading) setDragOver(true);
  };
  const handleDragLeave = (e) => {
    e.preventDefault();
    setDragOver(false);
  };
  const handleDrop = (e) => {
    e.preventDefault();
    setDragOver(false);
    if (disabled || uploading) return;
    const dropped = Array.from(e.dataTransfer.files);
    if (dropped.length > 0) validateAndAdd(dropped);
  };
  const removeFile = (index) => {
    const next = files.filter((_, i) => i !== index);
    if (!value) setInternalFiles(next);
    onFilesChange?.(next);
    if (next.length === 0) setError(null);
  };
  const defaultLabel = variant === "compact" ? "Click or drag to upload" : "Drop files here or click to upload";
  const defaultDescription = (() => {
    const parts = [];
    if (accept) parts.push(accept.replace(/,/g, ", "));
    if (maxSize) parts.push(`max ${formatBytes(maxSize)}`);
    if (maxFiles > 1) parts.push(`up to ${maxFiles} files`);
    return parts.join(" \xB7 ");
  })();
  const isDisabled = disabled || uploading;
  const iconSize = size === "sm" ? "size-6" : size === "lg" ? "size-10" : "size-8";
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "file-upload",
      className: cn("flex flex-col gap-3", className),
      children: [
        /* @__PURE__ */ jsx(
          "input",
          {
            ref: inputRef,
            type: "file",
            accept,
            multiple,
            disabled: isDisabled,
            onChange: handleInputChange,
            "aria-hidden": "true",
            "aria-label": "File upload",
            tabIndex: -1,
            className: "sr-only",
            "data-slot": "file-upload-input"
          }
        ),
        /* @__PURE__ */ jsxs(
          "button",
          {
            type: "button",
            "aria-label": "File upload drop zone",
            "aria-disabled": isDisabled,
            disabled: isDisabled,
            onKeyDown: handleKeyDown,
            onClick: openPicker,
            onDragOver: handleDragOver,
            onDragLeave: handleDragLeave,
            onDrop: handleDrop,
            className: cn(
              dropZoneVariants({ variant, size }),
              dragOver && "border-primary bg-accent-subtle",
              error && "border-destructive bg-destructive-wash focus-visible:ring-destructive"
            ),
            children: [
              uploading && /* @__PURE__ */ jsx(
                "output",
                {
                  "aria-label": "Uploading",
                  className: "absolute inset-0 flex items-center justify-center rounded bg-overlay-frosted backdrop-blur-sm",
                  children: /* @__PURE__ */ jsx(Loader2Icon, { className: "size-6 animate-spin text-muted-foreground" })
                }
              ),
              /* @__PURE__ */ jsx(
                UploadCloudIcon,
                {
                  className: cn(iconSize, "text-muted-foreground"),
                  "aria-hidden": "true"
                }
              ),
              /* @__PURE__ */ jsx("p", { className: "font-label text-foreground", children: label ?? defaultLabel }),
              description ?? defaultDescription ? /* @__PURE__ */ jsx("p", { className: "text-center text-xs text-muted-foreground", children: description ?? defaultDescription }) : null
            ]
          }
        ),
        error && /* @__PURE__ */ jsx(
          "p",
          {
            role: "alert",
            "data-slot": "file-upload-error",
            className: "text-sm text-destructive",
            children: error
          }
        ),
        files.length > 0 && /* @__PURE__ */ jsx(
          "ul",
          {
            "data-slot": "file-upload-list",
            "aria-label": "Selected files",
            className: "flex flex-col gap-2",
            children: files.map((file, i) => /* @__PURE__ */ jsx(
              FilePreviewItem,
              {
                file,
                showThumbnail: showThumbnails,
                onRemove: () => removeFile(i),
                disabled: isDisabled
              },
              `${file.name}-${file.size}`
            ))
          }
        )
      ]
    }
  );
};

export {
  dropZoneVariants,
  FileUpload
};
