// @docker/trident/dist/chunk-T4HZBNQU.js
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow
} from "./chunk-SKOSYJCU.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/data-table.tsx
import {
  flexRender,
  getCoreRowModel,
  getFilteredRowModel,
  getPaginationRowModel,
  getSortedRowModel,
  useReactTable
} from "@tanstack/react-table";
import {
  ArrowDown,
  ArrowUp,
  ChevronLeft,
  ChevronRight,
  ChevronsUpDown
} from "lucide-react";
import { useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function DataTableColumnHeader({
  header,
  title,
  className
}) {
  if (!header.column.getCanSort()) {
    return /* @__PURE__ */ jsx("span", { className: cn(className), children: title });
  }
  const sorted = header.column.getIsSorted();
  return /* @__PURE__ */ jsxs(
    Button,
    {
      variant: "ghost",
      size: "sm",
      className: cn("-ml-2 h-8 gap-1 font-label", className),
      onClick: () => header.column.toggleSorting(),
      children: [
        title,
        sorted === "asc" ? /* @__PURE__ */ jsx(ArrowUp, { className: "size-3.5" }) : sorted === "desc" ? /* @__PURE__ */ jsx(ArrowDown, { className: "size-3.5" }) : /* @__PURE__ */ jsx(ChevronsUpDown, { className: "size-3.5 opacity-50" })
      ]
    }
  );
}
function DataTablePagination({
  table,
  className
}) {
  const pageIndex = table.getState().pagination.pageIndex;
  const pageSize = table.getState().pagination.pageSize;
  const totalRows = table.getFilteredRowModel().rows.length;
  const start = pageIndex * pageSize + 1;
  const end = Math.min((pageIndex + 1) * pageSize, totalRows);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "data-table-pagination",
      className: cn("flex items-center justify-between px-2 py-4", className),
      children: [
        /* @__PURE__ */ jsx("p", { className: "text-sm text-muted-foreground", children: totalRows > 0 ? `${start}\u2013${end} of ${totalRows} row${totalRows === 1 ? "" : "s"}` : "No results" }),
        /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-2", children: [
          /* @__PURE__ */ jsxs(
            Button,
            {
              variant: "outline",
              size: "sm",
              onClick: () => table.previousPage(),
              disabled: !table.getCanPreviousPage(),
              children: [
                /* @__PURE__ */ jsx(ChevronLeft, { "data-icon": "inline-start" }),
                "Previous"
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            Button,
            {
              variant: "outline",
              size: "sm",
              onClick: () => table.nextPage(),
              disabled: !table.getCanNextPage(),
              children: [
                "Next",
                /* @__PURE__ */ jsx(ChevronRight, { "data-icon": "inline-end" })
              ]
            }
          )
        ] })
      ]
    }
  );
}
function DataTable({
  columns,
  data,
  pageSize = 10,
  noResultsMessage = "No results",
  tableOptions,
  className
}) {
  const [sorting, setSorting] = useState([]);
  const [columnFilters, setColumnFilters] = useState([]);
  const [columnVisibility, setColumnVisibility] = useState({});
  const table = useReactTable({
    data,
    columns,
    getCoreRowModel: getCoreRowModel(),
    ...pageSize > 0 && { getPaginationRowModel: getPaginationRowModel() },
    getSortedRowModel: getSortedRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    onSortingChange: setSorting,
    onColumnFiltersChange: setColumnFilters,
    onColumnVisibilityChange: setColumnVisibility,
    state: {
      sorting,
      columnFilters,
      columnVisibility,
      ...pageSize > 0 && { pagination: { pageIndex: 0, pageSize } },
      ...tableOptions?.state
    },
    ...tableOptions
  });
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "data-table",
      className: cn("@container/data-table", className),
      children: [
        /* @__PURE__ */ jsx("div", { className: "@max-[400px]/data-table:overflow-x-auto", children: /* @__PURE__ */ jsxs(Table, { children: [
          /* @__PURE__ */ jsx(TableHeader, { children: table.getHeaderGroups().map((headerGroup) => /* @__PURE__ */ jsx(TableRow, { children: headerGroup.headers.map((header) => /* @__PURE__ */ jsx(
            TableHead,
            {
              "aria-sort": header.column.getIsSorted() === "asc" ? "ascending" : header.column.getIsSorted() === "desc" ? "descending" : header.column.getCanSort() ? "none" : void 0,
              className: "@max-[400px]/data-table:px-2",
              children: header.isPlaceholder ? null : flexRender(
                header.column.columnDef.header,
                header.getContext()
              )
            },
            header.id
          )) }, headerGroup.id)) }),
          /* @__PURE__ */ jsx(TableBody, { children: table.getRowModel().rows.length > 0 ? table.getRowModel().rows.map((row) => /* @__PURE__ */ jsx(
            TableRow,
            {
              "data-state": row.getIsSelected() ? "selected" : void 0,
              children: row.getVisibleCells().map((cell) => /* @__PURE__ */ jsx(
                TableCell,
                {
                  className: "@max-[400px]/data-table:px-2 @max-[400px]/data-table:py-2 @max-[400px]/data-table:text-xs",
                  children: flexRender(
                    cell.column.columnDef.cell,
                    cell.getContext()
                  )
                },
                cell.id
              ))
            },
            row.id
          )) : /* @__PURE__ */ jsx(TableRow, { children: /* @__PURE__ */ jsx(
            TableCell,
            {
              colSpan: columns.length,
              className: "h-24 text-center text-muted-foreground",
              children: noResultsMessage
            }
          ) }) })
        ] }) }),
        pageSize > 0 && /* @__PURE__ */ jsx(DataTablePagination, { table })
      ]
    }
  );
}

export {
  DataTableColumnHeader,
  DataTablePagination,
  DataTable
};
