// @docker/trident/dist/chunk-DIR227VQ.js
import {
  StepIndicator
} from "./chunk-FWSKXSDB.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/stepper.tsx
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var StepperContext = React.createContext(null);
var useStepper = () => {
  const ctx = React.useContext(StepperContext);
  if (!ctx) {
    throw new Error("useStepper must be used inside a <Stepper> component.");
  }
  return ctx;
};
var Stepper = ({
  steps,
  currentStep: controlledStep,
  onStepChange,
  onComplete,
  className,
  children
}) => {
  const isControlled = controlledStep !== void 0;
  const [internalStep, setInternalStep] = React.useState(0);
  const [validationError, setValidationError] = React.useState(
    null
  );
  const [isValidating, setIsValidating] = React.useState(false);
  const currentStep = isControlled ? controlledStep : internalStep;
  const totalSteps = steps.length;
  const isFirst = currentStep === 0;
  const isLast = currentStep === totalSteps - 1;
  const setStep = React.useCallback(
    (next) => {
      if (isControlled) {
        onStepChange?.(next);
      } else {
        setInternalStep(next);
      }
    },
    [isControlled, onStepChange]
  );
  const goNext = React.useCallback(async () => {
    const stepConfig = steps[currentStep];
    setValidationError(null);
    if (stepConfig?.validate) {
      setIsValidating(true);
      let result;
      try {
        result = await stepConfig.validate();
      } finally {
        setIsValidating(false);
      }
      if (result === false) {
        return;
      }
      if (typeof result === "string") {
        setValidationError(result);
        return;
      }
    }
    if (isLast) {
      onComplete?.();
      return;
    }
    setStep(currentStep + 1);
  }, [currentStep, isLast, onComplete, setStep, steps]);
  const goBack = React.useCallback(() => {
    if (isFirst) return;
    setValidationError(null);
    setStep(currentStep - 1);
  }, [currentStep, isFirst, setStep]);
  const goTo = React.useCallback(
    (step) => {
      if (step < 0 || step >= totalSteps) return;
      setValidationError(null);
      setStep(step);
    },
    [setStep, totalSteps]
  );
  const contextValue = React.useMemo(
    () => ({
      steps,
      currentStep,
      totalSteps,
      isFirst,
      isLast,
      canGoBack: !isFirst,
      validationError,
      isValidating,
      goNext,
      goBack,
      goTo
    }),
    [
      steps,
      currentStep,
      totalSteps,
      isFirst,
      isLast,
      validationError,
      isValidating,
      goNext,
      goBack,
      goTo
    ]
  );
  return /* @__PURE__ */ jsx(StepperContext.Provider, { value: contextValue, children: /* @__PURE__ */ jsxs("div", { "data-slot": "stepper", className: cn("flex flex-col gap-6", className), children: [
    /* @__PURE__ */ jsx(
      StepIndicator,
      {
        steps: totalSteps,
        currentStep: currentStep + 1,
        "data-slot": "stepper-indicator"
      }
    ),
    /* @__PURE__ */ jsx(StepperHeader, {}),
    /* @__PURE__ */ jsx(
      "div",
      {
        "aria-live": "polite",
        "aria-atomic": "true",
        className: "sr-only",
        "data-slot": "stepper-announcer",
        children: `Step ${currentStep + 1} of ${totalSteps}: ${steps[currentStep]?.title ?? ""}`
      }
    ),
    children
  ] }) });
};
var StepperHeader = () => {
  const { steps, currentStep } = useStepper();
  const step = steps[currentStep];
  if (!step) return null;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "stepper-header",
      "aria-current": "step",
      className: "flex flex-col gap-1",
      children: [
        /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-2", children: [
          /* @__PURE__ */ jsx("h2", { className: "text-base font-heading text-foreground", children: step.title }),
          step.optional && /* @__PURE__ */ jsx("span", { className: "text-xs text-muted-foreground", children: "(Optional)" })
        ] }),
        step.description && /* @__PURE__ */ jsx("p", { className: "text-sm text-muted-foreground", children: step.description })
      ]
    }
  );
};
var StepperContent = ({ step, children, className }) => {
  const { currentStep } = useStepper();
  const contentRef = React.useRef(null);
  React.useEffect(() => {
    if (currentStep === step && contentRef.current) {
      contentRef.current.focus({ preventScroll: false });
    }
  }, [currentStep, step]);
  if (currentStep !== step) return null;
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: contentRef,
      "data-slot": "stepper-content",
      "data-step": step,
      tabIndex: -1,
      className: cn("outline-none", className),
      children
    }
  );
};
var StepperNavigation = ({
  nextLabel = "Next",
  backLabel = "Back",
  completeLabel = "Complete",
  nextLoading,
  className
}) => {
  const {
    isFirst,
    isLast,
    canGoBack,
    goNext,
    goBack,
    isValidating,
    validationError
  } = useStepper();
  const loading = nextLoading ?? isValidating;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "stepper-navigation",
      className: cn("flex flex-col gap-2", className),
      children: [
        /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-2", children: [
          canGoBack && /* @__PURE__ */ jsx(
            Button,
            {
              variant: "outline",
              onClick: goBack,
              disabled: isFirst,
              "data-slot": "stepper-back",
              children: backLabel
            }
          ),
          /* @__PURE__ */ jsx(
            Button,
            {
              onClick: () => void goNext(),
              loading,
              "data-slot": "stepper-next",
              children: isLast ? completeLabel : nextLabel
            }
          )
        ] }),
        validationError && /* @__PURE__ */ jsx(
          "p",
          {
            role: "alert",
            "data-slot": "stepper-error",
            className: "text-sm text-destructive",
            children: validationError
          }
        )
      ]
    }
  );
};

export {
  useStepper,
  Stepper,
  StepperContent,
  StepperNavigation
};
