// @docker/trident/dist/chunk-OVA56T54.js
import {
  Calendar
} from "./chunk-SL6PT4EM.js";
import {
  Label
} from "./chunk-7266IQJX.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/date-time-picker.tsx
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function TimeInput({ value, onChange, label, className }) {
  const hoursStr = String(value.hours).padStart(2, "0");
  const minutesStr = String(value.minutes).padStart(2, "0");
  const handleHoursChange = (e) => {
    const h = Math.min(
      23,
      Math.max(0, Number.parseInt(e.target.value, 10) || 0)
    );
    onChange({ ...value, hours: h });
  };
  const handleMinutesChange = (e) => {
    const m = Math.min(
      59,
      Math.max(0, Number.parseInt(e.target.value, 10) || 0)
    );
    onChange({ ...value, minutes: m });
  };
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "time-input",
      className: cn("flex items-center gap-1.5", className),
      children: [
        label && /* @__PURE__ */ jsx(Label, { className: "text-xs text-muted-foreground", children: label }),
        /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-0.5", children: [
          /* @__PURE__ */ jsx(
            "input",
            {
              type: "number",
              min: 0,
              max: 23,
              value: hoursStr,
              onChange: handleHoursChange,
              "aria-label": label ? `${label} hours` : "Hours",
              className: "h-8 w-10 rounded border border-input bg-background px-1 text-center text-sm tabular-nums outline-none focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring"
            }
          ),
          /* @__PURE__ */ jsx("span", { className: "text-sm text-muted-foreground", children: ":" }),
          /* @__PURE__ */ jsx(
            "input",
            {
              type: "number",
              min: 0,
              max: 59,
              value: minutesStr,
              onChange: handleMinutesChange,
              "aria-label": label ? `${label} minutes` : "Minutes",
              className: "h-8 w-10 rounded border border-input bg-background px-1 text-center text-sm tabular-nums outline-none focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring"
            }
          )
        ] })
      ]
    }
  );
}
function DateTimePicker({
  value,
  onChange,
  minDate,
  maxDate,
  className
}) {
  const time = React.useMemo(
    () => ({
      hours: value?.getHours() ?? 0,
      minutes: value?.getMinutes() ?? 0
    }),
    [value]
  );
  const handleDateSelect = (date) => {
    if (!date) {
      onChange?.(void 0);
      return;
    }
    const merged = new Date(date);
    merged.setHours(time.hours, time.minutes, 0, 0);
    onChange?.(merged);
  };
  const handleTimeChange = (newTime) => {
    const base = value ?? /* @__PURE__ */ new Date();
    const merged = new Date(base);
    merged.setHours(newTime.hours, newTime.minutes, 0, 0);
    onChange?.(merged);
  };
  return /* @__PURE__ */ jsxs("div", { "data-slot": "date-time-picker", className: cn("space-y-3", className), children: [
    /* @__PURE__ */ jsx(
      Calendar,
      {
        mode: "single",
        selected: value,
        onSelect: handleDateSelect,
        disabled: minDate || maxDate ? (date) => {
          if (minDate && date < minDate) return true;
          if (maxDate && date > maxDate) return true;
          return false;
        } : void 0
      }
    ),
    /* @__PURE__ */ jsx("div", { className: "flex justify-center border-t border-border px-3 pt-3", children: /* @__PURE__ */ jsx(TimeInput, { value: time, onChange: handleTimeChange, label: "Time" }) })
  ] });
}
function DateTimeRangePicker({
  value,
  onChange,
  className
}) {
  const fromTime = React.useMemo(
    () => ({
      hours: value?.from?.getHours() ?? 0,
      minutes: value?.from?.getMinutes() ?? 0
    }),
    [value?.from]
  );
  const toTime = React.useMemo(
    () => ({
      hours: value?.to?.getHours() ?? 23,
      minutes: value?.to?.getMinutes() ?? 59
    }),
    [value?.to]
  );
  const handleRangeSelect = (range) => {
    if (!range) {
      onChange?.({});
      return;
    }
    const from = range.from ? new Date(range.from) : void 0;
    const to = range.to ? new Date(range.to) : void 0;
    if (from) from.setHours(fromTime.hours, fromTime.minutes, 0, 0);
    if (to) to.setHours(toTime.hours, toTime.minutes, 0, 0);
    onChange?.({ from, to });
  };
  const handleFromTimeChange = (time) => {
    const from = value?.from ? new Date(value.from) : /* @__PURE__ */ new Date();
    from.setHours(time.hours, time.minutes, 0, 0);
    onChange?.({ ...value, from });
  };
  const handleToTimeChange = (time) => {
    const to = value?.to ? new Date(value.to) : /* @__PURE__ */ new Date();
    to.setHours(time.hours, time.minutes, 0, 0);
    onChange?.({ ...value, to });
  };
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "date-time-range-picker",
      className: cn("space-y-3", className),
      children: [
        /* @__PURE__ */ jsx(
          Calendar,
          {
            mode: "range",
            selected: value?.from && value?.to ? { from: value.from, to: value.to } : void 0,
            onSelect: handleRangeSelect,
            numberOfMonths: 2
          }
        ),
        /* @__PURE__ */ jsxs("div", { className: "flex justify-center gap-6 border-t border-border px-3 pt-3", children: [
          /* @__PURE__ */ jsx(
            TimeInput,
            {
              value: fromTime,
              onChange: handleFromTimeChange,
              label: "Start"
            }
          ),
          /* @__PURE__ */ jsx(TimeInput, { value: toTime, onChange: handleToTimeChange, label: "End" })
        ] })
      ]
    }
  );
}

export {
  TimeInput,
  DateTimePicker,
  DateTimeRangePicker
};
