// @docker/trident/dist/chunk-U7OSE5YZ.js
import {
  Label
} from "./chunk-7266IQJX.js";
import {
  Input
} from "./chunk-5CQB3X63.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/copyable-field.tsx
import { Check, Copy } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function CopyableField({
  value,
  label,
  hideLabel = false,
  onCopy,
  className
}) {
  const [copied, setCopied] = React.useState(false);
  const id = React.useId();
  React.useEffect(() => {
    if (!copied) return;
    const timer = setTimeout(() => setCopied(false), 2e3);
    return () => clearTimeout(timer);
  }, [copied]);
  const handleCopy = React.useCallback(async () => {
    try {
      const writePromise = navigator.clipboard.writeText(value);
      setCopied(true);
      await writePromise;
      onCopy?.();
    } catch {
      setCopied(false);
      console.error("[CopyableField] Failed to copy to clipboard");
    }
  }, [value, onCopy]);
  return /* @__PURE__ */ jsxs("div", { className: cn("flex flex-col gap-1.5", className), children: [
    /* @__PURE__ */ jsx(Label, { htmlFor: id, className: cn(hideLabel ? "sr-only" : void 0), children: label }),
    /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-2", children: [
      /* @__PURE__ */ jsx(
        Input,
        {
          id,
          readOnly: true,
          value,
          className: "font-mono text-sm",
          "aria-label": label
        }
      ),
      /* @__PURE__ */ jsx(
        Button,
        {
          type: "button",
          variant: "outline",
          size: "icon",
          "aria-label": copied ? "Copied!" : `Copy ${label}`,
          onClick: handleCopy,
          className: "shrink-0",
          children: copied ? /* @__PURE__ */ jsx(Check, { className: "h-4 w-4 text-success" }) : /* @__PURE__ */ jsx(Copy, { className: "h-4 w-4" })
        }
      )
    ] })
  ] });
}

export {
  CopyableField
};
