// @docker/trident/dist/chunk-TLIJZ6UL.js
import {
  Spinner
} from "./chunk-K3R22AX2.js";
import {
  Input
} from "./chunk-5CQB3X63.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/search.tsx
import { Search as SearchIcon, X } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function Search({
  className,
  placeholder = "Search\u2026",
  value,
  defaultValue,
  disabled,
  loading,
  onChange,
  onClear,
  ...props
}) {
  const [internalValue, setInternalValue] = React.useState(
    typeof defaultValue === "string" ? defaultValue : ""
  );
  const isControlled = value !== void 0;
  const currentValue = isControlled ? value : internalValue;
  const showClear = !!currentValue && !loading;
  const handleChange = (e) => {
    const val = e.target.value;
    if (!isControlled) setInternalValue(val);
    onChange?.(val);
  };
  const handleClear = () => {
    if (!isControlled) setInternalValue("");
    onChange?.("");
    onClear?.();
  };
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "search",
      className: cn("relative flex items-center", className),
      children: [
        /* @__PURE__ */ jsx(
          SearchIcon,
          {
            className: "pointer-events-none absolute left-2.5 size-4 shrink-0 text-muted-foreground",
            "aria-hidden": "true"
          }
        ),
        /* @__PURE__ */ jsx(
          Input,
          {
            type: "search",
            placeholder,
            value: currentValue,
            disabled,
            onChange: handleChange,
            className: "pl-8 [&::-webkit-search-cancel-button]:hidden",
            ...props
          }
        ),
        loading && /* @__PURE__ */ jsx(
          Spinner,
          {
            className: "absolute right-2.5 size-4 text-muted-foreground",
            "aria-label": "Searching"
          }
        ),
        showClear && /* @__PURE__ */ jsx(
          Button,
          {
            variant: "ghost",
            size: "icon-xs",
            onClick: handleClear,
            disabled,
            className: "absolute right-1.5 size-6 text-muted-foreground hover:text-foreground",
            "aria-label": "Clear search",
            children: /* @__PURE__ */ jsx(X, { className: "size-3.5", "aria-hidden": "true" })
          }
        )
      ]
    }
  );
}

export {
  Search
};
