// @docker/trident/dist/chunk-MICCEESH.js
// src/components/data-grid-utils.ts
function dedup(arr) {
  return arr.filter((v, i, a) => a.indexOf(v) === i);
}
function buildColumnPinning(columns, pinnedColumns) {
  const left = dedup([
    ...columns.filter((c) => c.pin === "left").map((c) => c.field),
    ...pinnedColumns?.left ?? []
  ]);
  const right = dedup([
    ...columns.filter((c) => c.pin === "right").map((c) => c.field),
    ...pinnedColumns?.right ?? []
  ]);
  return {
    left: left.length ? left : void 0,
    right: right.length ? right : void 0
  };
}
function unknownToStr(v) {
  if (v == null) return "";
  if (typeof v === "string") return v;
  if (typeof v === "number" || typeof v === "boolean" || typeof v === "bigint")
    return String(v);
  return JSON.stringify(v);
}
function toColumnDef(col) {
  return {
    id: col.field,
    accessorFn: col.valueGetter ? (row) => col.valueGetter?.(row) : (row) => Reflect.get(row, col.field),
    header: col.renderHeader ? () => col.renderHeader?.() : col.headerName ?? col.field,
    cell: col.renderCell ? ({ row, getValue }) => col.renderCell?.({ row: row.original, value: getValue() }) : col.valueFormatter ? ({ getValue }) => unknownToStr(col.valueFormatter?.(getValue())) : ({ getValue }) => unknownToStr(getValue()),
    enableSorting: col.sortable !== false,
    sortDescFirst: false,
    // MUI parity: first click always sorts ascending
    size: col.width,
    minSize: col.minWidth,
    maxSize: col.maxWidth,
    meta: {
      align: col.align,
      headerAlign: col.headerAlign ?? col.align,
      disableColumnMenu: col.disableColumnMenu,
      resizable: col.resizable,
      pin: col.pin,
      filterable: col.filterable,
      filterOperators: col.filterOperators,
      editable: col.editable
    }
  };
}
function exportToCsv(columns, rows, options = {}) {
  const { fileName = "export", delimiter = ",", utf8WithBom = false } = options;
  const visible = columns.filter((c) => !c.hide);
  const header = visible.map((c) => c.headerName ?? c.field);
  const data = rows.map(
    (row) => visible.map((col) => {
      const raw = col.valueGetter ? col.valueGetter(row) : Reflect.get(row, col.field);
      const formatted = col.valueFormatter ? col.valueFormatter(raw) : raw;
      return unknownToStr(formatted);
    })
  );
  const csv = [header, ...data].map(
    (line) => line.map((v) => `"${String(v).replace(/"/g, '""')}"`).join(delimiter)
  ).join("\n");
  const bom = utf8WithBom ? "\uFEFF" : "";
  const blob = new Blob([bom + csv], { type: "text/csv;charset=utf-8;" });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `${fileName}.csv`;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

export {
  buildColumnPinning,
  unknownToStr,
  toColumnDef,
  exportToCsv
};
