// @docker/trident/dist/chunk-ZYUYEPF7.js
import {
  Search
} from "./chunk-TLIJZ6UL.js";
import {
  parseAnsi
} from "./chunk-6AWMMDHE.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/log-viewer.tsx
import { useVirtualizer } from "@tanstack/react-virtual";
import { ArrowDownIcon, PauseIcon, PlayIcon } from "lucide-react";
import { useCallback, useEffect, useMemo, useRef, useState } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var LEVEL_CLASS = {
  error: "text-destructive",
  warn: "text-warning",
  debug: "text-muted-foreground",
  info: "text-foreground"
};
var AT_BOTTOM_THRESHOLD = 8;
var ESTIMATED_ROW_HEIGHT = 24;
var OVERSCAN = 10;
var defaultFilterFn = (entry, search) => entry.message.toLowerCase().includes(search.toLowerCase());
function formatTimestamp(ts) {
  const date = ts instanceof Date ? ts : new Date(ts);
  return date.toISOString();
}
function splitBySearch(text, lowerSearch) {
  if (!lowerSearch) return [{ text, match: false }];
  const parts = [];
  const lowerText = text.toLowerCase();
  let cursor = 0;
  while (cursor < text.length) {
    const idx = lowerText.indexOf(lowerSearch, cursor);
    if (idx === -1) {
      parts.push({ text: text.slice(cursor), match: false });
      break;
    }
    if (idx > cursor) {
      parts.push({ text: text.slice(cursor, idx), match: false });
    }
    parts.push({
      text: text.slice(idx, idx + lowerSearch.length),
      match: true
    });
    cursor = idx + lowerSearch.length;
  }
  return parts;
}
var LogMessage = ({ message, search }) => {
  const segments = parseAnsi(message);
  const lowerSearch = search.toLowerCase();
  return /* @__PURE__ */ jsx(Fragment, { children: segments.map((seg, segIdx) => {
    if (!search) {
      return /* @__PURE__ */ jsx("span", { className: seg.className, children: seg.text }, segIdx);
    }
    const parts = splitBySearch(seg.text, lowerSearch);
    return /* @__PURE__ */ jsx("span", { className: seg.className, children: parts.map(
      (part, partIdx) => part.match ? /* @__PURE__ */ jsx(
        "mark",
        {
          className: "rounded-xs bg-log-warn-bg text-alert-warning-fg",
          children: part.text
        },
        partIdx
      ) : /* @__PURE__ */ jsx("span", { children: part.text }, partIdx)
    ) }, segIdx);
  }) });
};
var LogRow = ({
  entry,
  index,
  showTimestamps,
  showLineNumbers,
  searchValue
}) => {
  const isStderr = entry.stream === "stderr";
  const levelClass = entry.level ? LEVEL_CLASS[entry.level] : "text-foreground";
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "log-row",
      className: cn(
        "flex min-h-(--dds-space-600) w-full items-baseline gap-2 px-3 py-0.5",
        "font-mono text-xs",
        isStderr && "bg-destructive-wash"
      ),
      children: [
        showLineNumbers && /* @__PURE__ */ jsx(
          "span",
          {
            "aria-hidden": "true",
            className: "w-10 shrink-0 text-right text-muted-foreground tabular-nums select-none",
            children: index + 1
          }
        ),
        showTimestamps && entry.timestamp && /* @__PURE__ */ jsx("span", { className: "shrink-0 text-muted-foreground tabular-nums", children: /* @__PURE__ */ jsx("time", { dateTime: formatTimestamp(entry.timestamp), children: formatTimestamp(entry.timestamp) }) }),
        entry.level && /* @__PURE__ */ jsx("span", { className: cn("w-12 shrink-0 font-bold uppercase", levelClass), children: entry.level }),
        /* @__PURE__ */ jsx("span", { className: cn("flex-1 break-all whitespace-pre-wrap", levelClass), children: /* @__PURE__ */ jsx(LogMessage, { message: entry.message, search: searchValue }) })
      ]
    }
  );
};
function LogViewer({
  entries,
  height = 400,
  follow = false,
  onFollowChange,
  showTimestamps = true,
  showLineNumbers = false,
  searchValue,
  onSearchChange,
  filterFn = defaultFilterFn,
  className,
  fontFamily
}) {
  const scrollRef = useRef(null);
  const [internalSearch, setInternalSearch] = useState("");
  const isSearchControlled = searchValue !== void 0;
  const activeSearch = isSearchControlled ? searchValue : internalSearch;
  const handleSearchChange = useCallback(
    (next) => {
      if (!isSearchControlled) setInternalSearch(next);
      onSearchChange?.(next);
    },
    [isSearchControlled, onSearchChange]
  );
  const filteredEntries = useMemo(() => {
    if (activeSearch === "") {
      return entries.map((entry, originalIndex) => ({ entry, originalIndex }));
    }
    const result = [];
    for (let i = 0; i < entries.length; i += 1) {
      const entry = entries[i];
      if (filterFn(entry, activeSearch)) {
        result.push({ entry, originalIndex: i });
      }
    }
    return result;
  }, [entries, activeSearch, filterFn]);
  const visibleCount = filteredEntries.length;
  const prevCountRef = useRef(visibleCount);
  const virtualizer = useVirtualizer({
    count: visibleCount,
    getScrollElement: () => scrollRef.current,
    estimateSize: () => ESTIMATED_ROW_HEIGHT,
    overscan: OVERSCAN,
    getItemKey: (index) => filteredEntries[index]?.entry.id ?? index
  });
  useEffect(() => {
    if (!follow) return;
    if (visibleCount === prevCountRef.current) return;
    prevCountRef.current = visibleCount;
    if (visibleCount > 0) {
      virtualizer.scrollToIndex(visibleCount - 1, { align: "end" });
    }
  }, [follow, visibleCount, virtualizer]);
  useEffect(() => {
    if (follow) {
      prevCountRef.current = visibleCount;
      if (visibleCount > 0) {
        virtualizer.scrollToIndex(visibleCount - 1, { align: "end" });
      }
    }
  }, [follow, visibleCount, virtualizer]);
  const handleScroll = useCallback(
    (e) => {
      if (!follow) return;
      const el = e.currentTarget;
      const atBottom = Math.abs(el.scrollHeight - el.scrollTop - el.clientHeight) < AT_BOTTOM_THRESHOLD;
      if (!atBottom) {
        onFollowChange?.(false);
      }
    },
    [follow, onFollowChange]
  );
  const handleFollowToggle = useCallback(() => {
    const next = !follow;
    onFollowChange?.(next);
    if (next && visibleCount > 0) {
      virtualizer.scrollToIndex(visibleCount - 1, { align: "end" });
    }
  }, [follow, onFollowChange, visibleCount, virtualizer]);
  const outerStyle = {
    height: typeof height === "number" ? `${height}px` : height,
    ...fontFamily ? { fontFamily } : {}
  };
  const totalSize = virtualizer.getTotalSize();
  const virtualItems = virtualizer.getVirtualItems();
  const isEmpty = entries.length === 0;
  const hasNoMatches = !isEmpty && visibleCount === 0;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "log-viewer",
      className: cn(
        "relative flex flex-col overflow-hidden rounded border border-border bg-background",
        className
      ),
      style: outerStyle,
      children: [
        /* @__PURE__ */ jsxs(
          "div",
          {
            role: "toolbar",
            "aria-label": "Log viewer controls",
            "data-slot": "log-viewer-toolbar",
            className: "flex shrink-0 items-center gap-2 border-b border-border bg-hover-overlay px-2 py-1.5",
            children: [
              /* @__PURE__ */ jsx(
                Search,
                {
                  className: "h-7 flex-1 text-xs",
                  placeholder: "Search logs...",
                  value: activeSearch,
                  onChange: handleSearchChange,
                  "aria-label": "Search log entries"
                }
              ),
              /* @__PURE__ */ jsx(
                Button,
                {
                  variant: "ghost",
                  size: "icon-sm",
                  onClick: handleFollowToggle,
                  "aria-label": follow ? "Pause log follow" : "Follow latest logs",
                  "aria-pressed": follow,
                  title: follow ? "Pause follow" : "Follow latest logs",
                  children: follow ? /* @__PURE__ */ jsx(PauseIcon, { className: "size-3.5" }) : /* @__PURE__ */ jsx(PlayIcon, { className: "size-3.5" })
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ jsxs("div", { className: "relative min-h-0 flex-1", children: [
          isEmpty ? /* @__PURE__ */ jsx(
            "div",
            {
              role: "status",
              "aria-live": "polite",
              className: "flex h-full items-center justify-center font-mono text-xs text-muted-foreground",
              children: "No log entries"
            }
          ) : hasNoMatches ? /* @__PURE__ */ jsxs(
            "div",
            {
              role: "status",
              "aria-live": "polite",
              className: "flex h-full items-center justify-center font-mono text-xs text-muted-foreground",
              children: [
                "No log entries match \u201C",
                activeSearch,
                "\u201D"
              ]
            }
          ) : /* @__PURE__ */ jsx(
            "div",
            {
              role: "region",
              "aria-label": "Log output",
              "aria-live": "polite",
              "aria-relevant": "additions",
              ref: scrollRef,
              className: "size-full overflow-y-auto",
              onScroll: handleScroll,
              tabIndex: 0,
              children: /* @__PURE__ */ jsx(
                "ol",
                {
                  "aria-label": "Log entries",
                  style: { position: "relative", height: `${totalSize}px` },
                  children: virtualItems.map((virtualRow) => {
                    const visible = filteredEntries[virtualRow.index];
                    if (!visible) return null;
                    const { entry, originalIndex } = visible;
                    return /* @__PURE__ */ jsx(
                      "li",
                      {
                        "data-index": virtualRow.index,
                        ref: virtualizer.measureElement,
                        style: {
                          position: "absolute",
                          top: 0,
                          left: 0,
                          width: "100%",
                          transform: `translateY(${virtualRow.start}px)`
                        },
                        children: /* @__PURE__ */ jsx(
                          LogRow,
                          {
                            entry,
                            index: originalIndex,
                            showTimestamps,
                            showLineNumbers,
                            searchValue: activeSearch
                          }
                        )
                      },
                      virtualRow.key
                    );
                  })
                }
              )
            }
          ),
          !follow && !isEmpty && !hasNoMatches && /* @__PURE__ */ jsxs(
            "button",
            {
              type: "button",
              onClick: handleFollowToggle,
              className: cn(
                "absolute right-5 bottom-3 z-raised flex items-center gap-1.5",
                "rounded-full border border-border bg-background px-3 py-1",
                "text-xs font-label shadow-raised transition-colors hover:bg-muted"
              ),
              "aria-label": "Resume scrolling to latest logs",
              children: [
                /* @__PURE__ */ jsx(ArrowDownIcon, { className: "size-3", "aria-hidden": "true" }),
                "Resume"
              ]
            }
          )
        ] })
      ]
    }
  );
}

export {
  LogRow,
  LogViewer
};
