// @docker/trident/dist/chunk-5CQB3X63.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/input.tsx
import { cva } from "class-variance-authority";
import { jsx } from "react/jsx-runtime";
var inputVariants = cva(
  "w-full min-w-0 rounded border border-input bg-input-bg text-base text-foreground transition-[color,box-shadow] ease-default outline-none file:inline-flex file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:border-destructive aria-invalid:ring-3 aria-invalid:ring-destructive-subtle md:text-sm high-contrast:placeholder:text-foreground",
  {
    variants: {
      size: {
        sm: "h-8 px-2 py-0.5 text-sm file:h-6",
        default: "h-9 px-2.5 py-1 file:h-7",
        lg: "h-10 px-3 py-1.5 file:h-8"
      }
    },
    defaultVariants: {
      size: "default"
    }
  }
);
function Input({ className, type, size, ref, ...props }) {
  return /* @__PURE__ */ jsx(
    "input",
    {
      ref,
      type,
      "data-slot": "input",
      className: cn(inputVariants({ size }), className),
      ...props
    }
  );
}

export {
  inputVariants,
  Input
};
