// @docker/trident/dist/chunk-X4ZQ4UF6.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/link.tsx
import { cva } from "class-variance-authority";
import { Slot } from "radix-ui";
import { jsx } from "react/jsx-runtime";
var linkVariants = cva(
  "underline underline-offset-4 transition-colors ease-default outline-none focus-visible:ring-3 focus-visible:ring-ring high-contrast:focus-visible:ring-ring",
  {
    variants: {
      variant: {
        default: "text-primary hover:text-link-hover high-contrast:text-link-hover high-contrast:hover:text-white",
        muted: "text-muted-foreground hover:text-foreground high-contrast:text-foreground",
        destructive: "text-destructive hover:text-destructive-link-hover high-contrast:text-text-error high-contrast:hover:text-white"
      }
    },
    defaultVariants: { variant: "default" }
  }
);
function Link({ className, variant, asChild = false, ...props }) {
  const Comp = asChild ? Slot.Root : "a";
  return /* @__PURE__ */ jsx(
    Comp,
    {
      "data-slot": "link",
      className: cn(linkVariants({ variant }), className),
      ...props
    }
  );
}

export {
  linkVariants,
  Link
};
