// @docker/trident/dist/chunk-EV3VZWOF.js
import {
  Badge
} from "./chunk-DBSTRLSQ.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/selection-card.tsx
import { CircleCheck } from "lucide-react";
import { jsx, jsxs } from "react/jsx-runtime";
var SelectionCard = ({
  selected,
  onSelect,
  title,
  description,
  badge,
  disabled = false,
  className
}) => /* @__PURE__ */ jsxs(
  "label",
  {
    className: cn(
      "group relative flex w-full cursor-pointer flex-col gap-1 rounded border p-4 text-left transition-all has-[:focus-visible]:ring-2 has-[:focus-visible]:ring-ring has-[:focus-visible]:ring-offset-2",
      selected ? "border-primary bg-accent-wash ring-2 ring-accent-subtle" : "border-border hover:border-primary hover:bg-hover-overlay",
      disabled && "pointer-events-none opacity-50",
      className
    ),
    children: [
      /* @__PURE__ */ jsx(
        "input",
        {
          type: "radio",
          checked: selected,
          onChange: onSelect,
          onKeyDown: (e) => {
            if (e.key === "Enter" && !disabled) {
              e.preventDefault();
              onSelect();
            }
          },
          disabled,
          className: "sr-only"
        }
      ),
      /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-2", children: [
        /* @__PURE__ */ jsx(
          "span",
          {
            className: cn(
              "flex size-4 shrink-0 items-center justify-center rounded-full border",
              selected ? "border-primary bg-primary text-primary-foreground" : "border-input"
            ),
            children: selected && /* @__PURE__ */ jsx(CircleCheck, { className: "size-3", "aria-hidden": true })
          }
        ),
        /* @__PURE__ */ jsx("span", { className: "text-sm font-label", children: title }),
        badge && /* @__PURE__ */ jsx(Badge, { variant: "secondary", size: "sm", children: badge })
      ] }),
      description && /* @__PURE__ */ jsx("p", { className: "pl-6 text-xs text-muted-foreground", children: description })
    ]
  }
);

export {
  SelectionCard
};
