// @docker/trident/dist/chunk-7AWKNMM6.js
import {
  TrustBadge
} from "./chunk-6D3M5CWS.js";
import {
  HealthGrade
} from "./chunk-HIEPEM4N.js";
import {
  Badge
} from "./chunk-DBSTRLSQ.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/repo-card.tsx
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function RepoCard({
  namespace,
  name,
  description,
  tier = "community",
  pullCount,
  healthGrade,
  securitySlot,
  onClick,
  className,
  imageUrl,
  imageAlt: _imageAlt,
  starCount,
  variant
}) {
  const baseClassName = cn(
    "flex w-full items-start gap-3 rounded border border-border bg-card p-4",
    className
  );
  const body = /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      "div",
      {
        "data-slot": "repo-card-icon",
        "aria-hidden": "true",
        className: "flex h-10 w-10 shrink-0 items-center justify-center overflow-hidden rounded border border-border bg-accent text-sm font-bold text-accent-foreground uppercase",
        children: imageUrl ? /* @__PURE__ */ jsx("img", { src: imageUrl, alt: "", className: "h-full w-full object-cover" }) : name.slice(0, 2)
      }
    ),
    /* @__PURE__ */ jsxs("div", { className: "min-w-0 flex-1", children: [
      /* @__PURE__ */ jsxs("div", { className: "flex flex-wrap items-center gap-2", children: [
        /* @__PURE__ */ jsxs(
          "span",
          {
            "data-slot": "repo-card-name",
            className: "text-sm leading-tight font-heading text-foreground",
            children: [
              namespace,
              "/",
              name
            ]
          }
        ),
        /* @__PURE__ */ jsx(TrustBadge, { tier })
      ] }),
      description ? /* @__PURE__ */ jsx(
        "p",
        {
          "data-slot": "repo-card-description",
          className: "mt-0.5 line-clamp-2 text-xs leading-relaxed text-muted-foreground high-contrast:text-foreground",
          children: description
        }
      ) : null,
      variant ?? starCount ? /* @__PURE__ */ jsxs("div", { className: "mt-1.5 flex items-center gap-2", children: [
        variant ? /* @__PURE__ */ jsx(Badge, { variant: "secondary", size: "sm", children: variant === "hardened" ? "Hardened" : "Marketplace" }) : null,
        starCount ? /* @__PURE__ */ jsxs(
          "span",
          {
            role: "img",
            "aria-label": `${starCount} stars`,
            className: "flex items-center gap-1 text-xs text-muted-foreground high-contrast:text-foreground",
            children: [
              /* @__PURE__ */ jsx(
                "svg",
                {
                  "aria-hidden": "true",
                  className: "h-3 w-3 fill-current",
                  viewBox: "0 0 16 16",
                  children: /* @__PURE__ */ jsx("path", { d: "M8 .25a.75.75 0 0 1 .673.418l1.882 3.815 4.21.612a.75.75 0 0 1 .416 1.279l-3.046 2.97.719 4.192a.75.75 0 0 1-1.088.791L8 12.347l-3.766 1.98a.75.75 0 0 1-1.088-.79l.72-4.194L.818 6.374a.75.75 0 0 1 .416-1.28l4.21-.611L7.327.668A.75.75 0 0 1 8 .25Z" })
                }
              ),
              /* @__PURE__ */ jsx("span", { "aria-hidden": "true", children: starCount })
            ]
          }
        ) : null
      ] }) : null,
      /* @__PURE__ */ jsxs(
        "div",
        {
          "data-slot": "repo-card-signals",
          className: "mt-2 flex flex-wrap items-center gap-3",
          children: [
            pullCount ? /* @__PURE__ */ jsx("span", { className: "text-xs text-muted-foreground tabular-nums high-contrast:text-foreground", children: pullCount }) : null,
            healthGrade ? /* @__PURE__ */ jsx(HealthGrade, { grade: healthGrade, className: "h-6 w-6 text-xs" }) : null,
            securitySlot
          ]
        }
      )
    ] })
  ] });
  if (onClick) {
    return /* @__PURE__ */ jsx(
      "button",
      {
        type: "button",
        "data-slot": "repo-card",
        "aria-label": `${namespace}/${name} repository`,
        className: cn(
          baseClassName,
          "cursor-pointer text-left transition-all ease-default hover:border-dds-blue-outlined-border hover:shadow-raised"
        ),
        onClick,
        children: body
      }
    );
  }
  return /* @__PURE__ */ jsx("div", { "data-slot": "repo-card", className: baseClassName, children: body });
}

export {
  RepoCard
};
