// @docker/trident/dist/chunk-632YNUL5.js
import {
  SpeedDialContext,
  useSpeedDial
} from "./chunk-Z2GVHZC5.js";
import {
  Tooltip,
  TooltipContent,
  TooltipTrigger
} from "./chunk-HVYJXDDW.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/speed-dial.tsx
import { Plus } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function SpeedDial({
  className,
  position = "bottom-right",
  direction = "up",
  open: controlledOpen,
  onOpenChange,
  children,
  ...props
}) {
  const [internalOpen, setInternalOpen] = React.useState(false);
  const isControlled = controlledOpen !== void 0;
  const open = isControlled ? controlledOpen : internalOpen;
  const setOpen = React.useCallback(
    (value) => {
      const next = typeof value === "function" ? value(open) : value;
      if (!isControlled) {
        setInternalOpen(next);
      }
      onOpenChange?.(next);
    },
    [isControlled, open, onOpenChange]
  );
  const [actionCount, setActionCount] = React.useState(0);
  const [focusedIndex, setFocusedIndex] = React.useState(-1);
  const actionRefs = React.useRef([]);
  const triggerRef = React.useRef(null);
  const contextValue = React.useMemo(
    () => ({
      open,
      setOpen,
      position,
      direction,
      actionCount,
      setActionCount,
      focusedIndex,
      setFocusedIndex,
      actionRefs,
      triggerRef
    }),
    [open, setOpen, position, direction, actionCount, focusedIndex]
  );
  const containerRef = React.useRef(null);
  React.useEffect(() => {
    if (!open) return;
    const handlePointerDown = (e) => {
      if (containerRef.current && e.target instanceof Node && !containerRef.current.contains(e.target)) {
        setOpen(false);
      }
    };
    document.addEventListener("pointerdown", handlePointerDown);
    return () => document.removeEventListener("pointerdown", handlePointerDown);
  }, [open, setOpen]);
  React.useEffect(() => {
    if (!open) return;
    const handleKeyDown = (e) => {
      if (e.key === "Escape") {
        setOpen(false);
        triggerRef.current?.focus();
      }
    };
    document.addEventListener("keydown", handleKeyDown);
    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [open, setOpen]);
  const positionClasses = {
    "bottom-right": "fixed right-6 bottom-6",
    "bottom-left": "fixed bottom-6 left-6",
    "top-right": "fixed top-6 right-6",
    "top-left": "fixed top-6 left-6"
  };
  const directionClasses = {
    up: "flex-col-reverse",
    down: "flex-col"
  };
  return /* @__PURE__ */ jsx(SpeedDialContext.Provider, { value: contextValue, children: /* @__PURE__ */ jsx(
    "div",
    {
      ref: containerRef,
      "data-slot": "speed-dial",
      "data-state": open ? "open" : "closed",
      "data-position": position,
      "data-direction": direction,
      role: "group",
      className: cn(
        "z-toast flex items-center gap-3",
        positionClasses[position],
        directionClasses[direction],
        className
      ),
      ...props,
      children
    }
  ) });
}
function SpeedDialTrigger({ className, icon, ...props }) {
  const { open, setOpen, actionRefs, setFocusedIndex, actionCount, triggerRef } = useSpeedDial();
  const handleClick = () => {
    setOpen((prev) => !prev);
    setFocusedIndex(-1);
  };
  const handleKeyDown = (e) => {
    if (e.key === "ArrowUp" || e.key === "ArrowDown") {
      e.preventDefault();
      if (!open) {
        setOpen(true);
      }
      const nextIndex = e.key === "ArrowUp" ? actionCount - 1 : 0;
      setFocusedIndex(nextIndex);
      actionRefs.current[nextIndex]?.focus();
    }
  };
  return /* @__PURE__ */ jsx(
    Button,
    {
      ref: triggerRef,
      "data-slot": "speed-dial-trigger",
      variant: "default",
      size: "icon-lg",
      "aria-expanded": open,
      "aria-haspopup": "true",
      "aria-label": open ? "Close actions" : "Open actions",
      onClick: handleClick,
      onKeyDown: handleKeyDown,
      className: cn(
        "size-14 rounded-full shadow-floating ring-1 ring-border transition-transform duration-normal ease-default focus-visible:ring-3 focus-visible:ring-ring",
        open && "rotate-45",
        className
      ),
      ...props,
      children: icon ?? /* @__PURE__ */ jsx(Plus, { className: "size-6" })
    }
  );
}
function SpeedDialAction({
  className,
  icon,
  label,
  index: indexProp,
  ...props
}) {
  const {
    open,
    setOpen,
    position,
    direction,
    actionRefs,
    focusedIndex,
    setFocusedIndex,
    actionCount,
    setActionCount,
    triggerRef
  } = useSpeedDial();
  const indexRef = React.useRef(-1);
  React.useEffect(() => {
    if (indexProp !== void 0) {
      indexRef.current = indexProp;
      setActionCount((prev) => Math.max(prev, indexProp + 1));
    } else {
      setActionCount((prev) => {
        indexRef.current = prev;
        return prev + 1;
      });
    }
    return () => {
      if (indexProp === void 0) {
        setActionCount((prev) => Math.max(0, prev - 1));
      }
    };
  }, [indexProp, setActionCount]);
  const currentIndex = indexProp ?? indexRef.current;
  const registerRef = React.useCallback(
    (el) => {
      actionRefs.current[currentIndex] = el;
    },
    [currentIndex, actionRefs]
  );
  const handleKeyDown = (e) => {
    if (e.key === "ArrowUp") {
      e.preventDefault();
      const prev = focusedIndex - 1;
      if (prev >= 0) {
        setFocusedIndex(prev);
        actionRefs.current[prev]?.focus();
      } else {
        triggerRef.current?.focus();
        setFocusedIndex(-1);
      }
    } else if (e.key === "ArrowDown") {
      e.preventDefault();
      const next = focusedIndex + 1;
      if (next < actionCount) {
        setFocusedIndex(next);
        actionRefs.current[next]?.focus();
      } else {
        triggerRef.current?.focus();
        setFocusedIndex(-1);
      }
    } else if (e.key === "Enter" || e.key === " ") {
      requestAnimationFrame(() => {
        setOpen(false);
        triggerRef.current?.focus();
      });
    }
  };
  const tooltipSide = position.includes("right") ? "left" : "right";
  const staggerDelay = currentIndex * 50;
  const slideFrom = direction === "up" ? "translate-y-2 opacity-0" : "-translate-y-2 opacity-0";
  const actionButton = /* @__PURE__ */ jsx(
    Button,
    {
      ref: registerRef,
      "data-slot": "speed-dial-action",
      variant: "outline",
      size: "icon",
      tabIndex: open ? 0 : -1,
      "aria-label": label,
      onClick: (e) => {
        props.onClick?.(e);
        setOpen(false);
        triggerRef.current?.focus();
      },
      onKeyDown: handleKeyDown,
      className: cn(
        "size-11 rounded-full bg-popover shadow-floating ring-1 ring-border transition-all ease-out",
        open ? "translate-y-0 scale-100 opacity-100" : `pointer-events-none ${slideFrom} scale-75`,
        className
      ),
      style: {
        transitionDelay: open ? `${staggerDelay}ms` : "0ms",
        transitionDuration: "var(--dds-duration-normal)"
      },
      ...props,
      children: icon
    }
  );
  if (!label) {
    return actionButton;
  }
  return /* @__PURE__ */ jsxs(Tooltip, { open: open && focusedIndex === currentIndex ? true : void 0, children: [
    /* @__PURE__ */ jsx(TooltipTrigger, { asChild: true, children: actionButton }),
    /* @__PURE__ */ jsx(TooltipContent, { side: tooltipSide, sideOffset: 8, children: label })
  ] });
}

export {
  SpeedDial,
  SpeedDialTrigger,
  SpeedDialAction
};
