// @docker/trident/dist/chunk-G6JSWBM7.js
import {
  cssVars
} from "./chunk-UEVWMCT6.js";

// src/components/ui/sonner.tsx
import {
  AlertTriangle,
  CircleCheck,
  Info,
  Loader2Icon,
  OctagonX
} from "lucide-react";
import { useTheme } from "next-themes";
import { useEffect, useState } from "react";
import { Toaster as Sonner } from "sonner";
import { toast } from "sonner";
import { jsx } from "react/jsx-runtime";
var validThemes = ["dark", "light", "system"];
var errorListeners = /* @__PURE__ */ new Set();
function announceError(message) {
  for (const fn of errorListeners) fn(message);
}
function ToastAnnouncer() {
  const [message, setMessage] = useState("");
  useEffect(() => {
    const handler = (msg) => {
      setMessage("");
      requestAnimationFrame(() => setMessage(msg));
    };
    errorListeners.add(handler);
    return () => {
      errorListeners.delete(handler);
    };
  }, []);
  return /* @__PURE__ */ jsx(
    "div",
    {
      role: "alert",
      "aria-live": "assertive",
      "aria-atomic": "true",
      "data-slot": "toast-announcer",
      className: "sr-only",
      children: message
    }
  );
}
var Toaster = ({ ...props }) => {
  const { theme = "system" } = useTheme();
  const resolvedTheme = validThemes.find((t) => t === theme) ?? "system";
  return /* @__PURE__ */ jsx(
    Sonner,
    {
      theme: resolvedTheme,
      className: "toaster group",
      icons: {
        success: /* @__PURE__ */ jsx(CircleCheck, { className: "size-4" }),
        info: /* @__PURE__ */ jsx(Info, { className: "size-4" }),
        warning: /* @__PURE__ */ jsx(AlertTriangle, { className: "size-4" }),
        error: /* @__PURE__ */ jsx(OctagonX, { className: "size-4" }),
        loading: /* @__PURE__ */ jsx(Loader2Icon, { className: "size-4 animate-spin" })
      },
      style: cssVars({
        "--normal-bg": "var(--popover)",
        "--normal-text": "var(--popover-foreground)",
        "--normal-border": "var(--border)",
        "--border-radius": "var(--radius)"
      }),
      toastOptions: {
        classNames: {
          toast: "cn-toast"
        }
      },
      ...props
    }
  );
};

export {
  announceError,
  ToastAnnouncer,
  Toaster,
  toast
};
