// @docker/trident/dist/chunk-HQNWTFFV.js
import {
  computeMultiSelection,
  getVisibleIds
} from "./chunk-AJJ5P6IF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/tree-view.tsx
import { ChevronRight, File, Folder } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var TreeViewNode = ({
  node,
  depth,
  expandedIds,
  selectedIds,
  focusedId,
  onToggle,
  onSelect,
  onKeyDown,
  renderNode,
  setRef
}) => {
  const hasChildren = Boolean(node.children && node.children.length > 0);
  const isExpanded = expandedIds.has(node.id);
  const isSelected = selectedIds.has(node.id);
  const isFocused = focusedId === node.id;
  const handleClick = (e) => {
    e.stopPropagation();
    if (hasChildren) {
      onToggle(node.id);
    }
    onSelect(node.id, node, {
      ctrlKey: e.ctrlKey || e.metaKey,
      shiftKey: e.shiftKey
    });
  };
  const handleKeyDown = (e) => {
    e.stopPropagation();
    onKeyDown(e, node.id, node, hasChildren, isExpanded);
  };
  const indentStyle = depth > 0 ? { paddingInlineStart: `calc(${depth} * var(--dds-space-500))` } : void 0;
  return /* @__PURE__ */ jsxs(
    "li",
    {
      role: "treeitem",
      "aria-expanded": hasChildren ? isExpanded : void 0,
      "aria-selected": isSelected,
      "aria-level": depth + 1,
      ref: (el) => setRef(node.id, el),
      tabIndex: isFocused ? 0 : -1,
      onClick: handleClick,
      onKeyDown: handleKeyDown,
      "data-slot": "tree-view-node",
      "data-node-id": node.id,
      "data-selected": isSelected ? "true" : void 0,
      className: "group outline-none",
      children: [
        /* @__PURE__ */ jsxs(
          "div",
          {
            style: indentStyle,
            className: cn(
              "flex cursor-pointer items-center gap-1.5 rounded px-2 py-1 text-sm transition-colors select-none",
              "hover:bg-hover-overlay",
              "group-focus-visible:ring-2 group-focus-visible:ring-ring group-focus-visible:ring-offset-1",
              isSelected && "bg-muted text-foreground",
              !isSelected && "text-foreground"
            ),
            children: [
              /* @__PURE__ */ jsx(
                "span",
                {
                  "aria-hidden": "true",
                  className: cn(
                    "flex size-4 shrink-0 items-center justify-center text-muted-foreground transition-transform duration-150 ease-in-out",
                    hasChildren && isExpanded && "rotate-90",
                    !hasChildren && "pointer-events-none opacity-0"
                  ),
                  children: /* @__PURE__ */ jsx(ChevronRight, { className: "size-3.5" })
                }
              ),
              /* @__PURE__ */ jsx(
                "span",
                {
                  "aria-hidden": "true",
                  className: "flex shrink-0 items-center text-muted-foreground",
                  children: hasChildren ? /* @__PURE__ */ jsx(Folder, { className: "size-4" }) : /* @__PURE__ */ jsx(File, { className: "size-4" })
                }
              ),
              renderNode ? renderNode(node, depth) : /* @__PURE__ */ jsx("span", { className: "truncate", children: node.label })
            ]
          }
        ),
        hasChildren && isExpanded && /* @__PURE__ */ jsx("ul", { role: "group", children: (node.children ?? []).map((child) => /* @__PURE__ */ jsx(
          TreeViewNode,
          {
            node: child,
            depth: depth + 1,
            expandedIds,
            selectedIds,
            focusedId,
            onToggle,
            onSelect,
            onKeyDown,
            renderNode,
            setRef
          },
          child.id
        )) })
      ]
    }
  );
};
var TreeView = (props) => {
  const {
    nodes,
    renderNode,
    defaultExpanded = [],
    expanded: controlledExpanded,
    onExpandedChange,
    className
  } = props;
  const isMultiSelect = props.multiSelect === true;
  const isControlledExpand = controlledExpanded !== void 0;
  const [internalExpanded, setInternalExpanded] = React.useState(
    () => new Set(defaultExpanded)
  );
  const expandedIds = React.useMemo(
    () => isControlledExpand ? new Set(controlledExpanded) : internalExpanded,
    [isControlledExpand, controlledExpanded, internalExpanded]
  );
  const toggleExpanded = React.useCallback(
    (id) => {
      const next = new Set(expandedIds);
      if (next.has(id)) {
        next.delete(id);
      } else {
        next.add(id);
      }
      if (!isControlledExpand) {
        setInternalExpanded(next);
      }
      onExpandedChange?.(Array.from(next));
    },
    [expandedIds, isControlledExpand, onExpandedChange]
  );
  const controlledSingleProp = !isMultiSelect ? props.selected : void 0;
  const isControlledSingle = controlledSingleProp !== void 0;
  const [internalSingleSelected, setInternalSingleSelected] = React.useState(void 0);
  const controlledMultiProp = isMultiSelect ? props.selected : void 0;
  const isControlledMulti = controlledMultiProp !== void 0;
  const [internalMultiSelected, setInternalMultiSelected] = React.useState(/* @__PURE__ */ new Set());
  const lastSelectedRef = React.useRef(void 0);
  const selectedIds = React.useMemo(() => {
    if (isMultiSelect) {
      return isControlledMulti && controlledMultiProp ? new Set(controlledMultiProp) : internalMultiSelected;
    }
    const single = isControlledSingle ? controlledSingleProp : internalSingleSelected;
    return single ? /* @__PURE__ */ new Set([single]) : /* @__PURE__ */ new Set();
  }, [
    isMultiSelect,
    isControlledMulti,
    isControlledSingle,
    controlledMultiProp,
    controlledSingleProp,
    internalMultiSelected,
    internalSingleSelected
  ]);
  const [focusedId, setFocusedId] = React.useState(
    () => nodes[0]?.id
  );
  const nodeRefs = React.useRef(/* @__PURE__ */ new Map());
  const setRef = React.useCallback((id, el) => {
    if (el) {
      nodeRefs.current.set(id, el);
    } else {
      nodeRefs.current.delete(id);
    }
  }, []);
  const handleSelect = React.useCallback(
    (id, node, meta) => {
      setFocusedId(id);
      if (isMultiSelect) {
        const visibleIds = getVisibleIds(nodes, expandedIds);
        const next = computeMultiSelection({
          clickedId: id,
          currentSelection: selectedIds,
          lastSelectedId: lastSelectedRef.current,
          visibleIds,
          ctrlKey: meta.ctrlKey,
          shiftKey: meta.shiftKey
        });
        lastSelectedRef.current = id;
        if (!isControlledMulti) {
          setInternalMultiSelected(next);
        }
        props.onSelect?.(
          Array.from(next)
        );
      } else {
        lastSelectedRef.current = id;
        if (!isControlledSingle) {
          setInternalSingleSelected(id);
        }
        props.onSelect?.(id, node);
      }
    },
    [
      isMultiSelect,
      isControlledMulti,
      isControlledSingle,
      nodes,
      expandedIds,
      selectedIds,
      props
    ]
  );
  const handleKeyDown = React.useCallback(
    (e, id, node, hasChildren, isExpanded) => {
      const visibleIds = getVisibleIds(nodes, expandedIds);
      const currentIndex = visibleIds.indexOf(id);
      switch (e.key) {
        case "ArrowDown": {
          e.preventDefault();
          const nextId = visibleIds[currentIndex + 1];
          if (nextId) {
            setFocusedId(nextId);
            nodeRefs.current.get(nextId)?.focus();
          }
          break;
        }
        case "ArrowUp": {
          e.preventDefault();
          const prevId = visibleIds[currentIndex - 1];
          if (prevId) {
            setFocusedId(prevId);
            nodeRefs.current.get(prevId)?.focus();
          }
          break;
        }
        case "ArrowRight": {
          e.preventDefault();
          if (hasChildren && !isExpanded) {
            toggleExpanded(id);
          } else if (hasChildren && isExpanded) {
            const firstChildId = visibleIds[currentIndex + 1];
            if (firstChildId) {
              setFocusedId(firstChildId);
              nodeRefs.current.get(firstChildId)?.focus();
            }
          }
          break;
        }
        case "ArrowLeft": {
          e.preventDefault();
          if (hasChildren && isExpanded) {
            toggleExpanded(id);
          } else {
            const currentEl = nodeRefs.current.get(id);
            if (currentEl) {
              const currentLevel = parseInt(
                currentEl.closest("[role=treeitem]")?.getAttribute("aria-level") ?? "1",
                10
              );
              for (let i = currentIndex - 1; i >= 0; i--) {
                const candidateId = visibleIds[i];
                const candidateEl = nodeRefs.current.get(candidateId);
                const candidateLevel = parseInt(
                  candidateEl?.closest("[role=treeitem]")?.getAttribute("aria-level") ?? "1",
                  10
                );
                if (candidateLevel < currentLevel) {
                  setFocusedId(candidateId);
                  candidateEl?.focus();
                  break;
                }
              }
            }
          }
          break;
        }
        case "Enter":
        case " ": {
          e.preventDefault();
          handleSelect(id, node, { ctrlKey: false, shiftKey: false });
          break;
        }
        case "Home": {
          e.preventDefault();
          const firstId = visibleIds[0];
          if (firstId) {
            setFocusedId(firstId);
            nodeRefs.current.get(firstId)?.focus();
          }
          break;
        }
        case "End": {
          e.preventDefault();
          const lastId = visibleIds[visibleIds.length - 1];
          if (lastId) {
            setFocusedId(lastId);
            nodeRefs.current.get(lastId)?.focus();
          }
          break;
        }
      }
    },
    [nodes, expandedIds, toggleExpanded, handleSelect]
  );
  React.useEffect(() => {
    if (!focusedId) {
      setFocusedId(nodes[0]?.id);
    }
  }, [nodes, focusedId]);
  return /* @__PURE__ */ jsx(
    "ul",
    {
      role: "tree",
      "data-slot": "tree-view",
      "aria-multiselectable": isMultiSelect ? "true" : void 0,
      className: cn("flex flex-col gap-0.5 p-1", className),
      children: nodes.map((node) => /* @__PURE__ */ jsx(
        TreeViewNode,
        {
          node,
          depth: 0,
          expandedIds,
          selectedIds,
          focusedId,
          onToggle: toggleExpanded,
          onSelect: handleSelect,
          onKeyDown: handleKeyDown,
          renderNode,
          setRef
        },
        node.id
      ))
    }
  );
};

export {
  TreeView
};
