// @docker/trident/dist/chunk-YU5N53H7.js
// src/components/code-diff-viewer-utils.ts
import { structuredPatch } from "diff";
function parseDiff(original, modified) {
  const patch = structuredPatch(
    "original",
    "modified",
    original,
    modified,
    "",
    "",
    { context: Number.MAX_SAFE_INTEGER }
  );
  if (patch.hunks.length === 0) {
    return { hunks: [], hasChanges: false };
  }
  const hasChanges = patch.hunks.some(
    (h) => h.lines.some((l) => l[0] === "+" || l[0] === "-")
  );
  if (!hasChanges) {
    return { hunks: [], hasChanges: false };
  }
  const hunks = patch.hunks.map((hunk) => {
    const header = `@@ -${hunk.oldStart},${hunk.oldLines} +${hunk.newStart},${hunk.newLines} @@`;
    const lines = [];
    let oldLineNum = hunk.oldStart;
    let newLineNum = hunk.newStart;
    for (const rawLine of hunk.lines) {
      if (rawLine.startsWith("\\ ")) continue;
      const prefix = rawLine[0];
      const content = rawLine.slice(1);
      if (prefix === "+") {
        lines.push({ type: "added", content, newLineNumber: newLineNum++ });
      } else if (prefix === "-") {
        lines.push({ type: "removed", content, oldLineNumber: oldLineNum++ });
      } else {
        lines.push({
          type: "context",
          content,
          oldLineNumber: oldLineNum++,
          newLineNumber: newLineNum++
        });
      }
    }
    return { header, lines };
  });
  return { hunks, hasChanges: true };
}
function toSplitRows(lines) {
  const rows = [];
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    if (line.type === "context") {
      rows.push({ left: line, right: line });
      i++;
      continue;
    }
    const removed = [];
    const added = [];
    while (i < lines.length && lines[i].type === "removed") {
      removed.push(lines[i++]);
    }
    while (i < lines.length && lines[i].type === "added") {
      added.push(lines[i++]);
    }
    const maxLen = Math.max(removed.length, added.length);
    for (let j = 0; j < maxLen; j++) {
      rows.push({
        left: removed[j] ?? null,
        right: added[j] ?? null
      });
    }
  }
  return rows;
}
function formatLineNumber(n) {
  return n !== void 0 ? String(n) : "";
}
var lineClasses = {
  added: "bg-diff-added-bg",
  removed: "bg-diff-removed-bg",
  context: "bg-background",
  "hunk-header": "bg-diff-hunk-bg"
};
var gutterCharClasses = {
  added: "text-diff-added-text",
  removed: "text-diff-removed-text",
  context: "text-transparent select-none"
};
var gutterChar = {
  added: "+",
  removed: "-",
  context: " "
};

export {
  parseDiff,
  toSplitRows,
  formatLineNumber,
  lineClasses,
  gutterCharClasses,
  gutterChar
};
