// @docker/trident/dist/chunk-6M5OUUJQ.js
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/notification-snackbar.tsx
import { X } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var AUTO_DISMISS_MS = 1e4;
function NotificationSnackbar({
  open,
  icon,
  message,
  messageTitle,
  persist = false,
  hideDismissAction = false,
  action,
  onClose,
  urgency = "assertive",
  className
}) {
  React.useEffect(() => {
    if (!open || persist) return;
    const timer = setTimeout(onClose, AUTO_DISMISS_MS);
    return () => clearTimeout(timer);
  }, [open, persist, onClose]);
  if (!open) return null;
  const role = urgency === "polite" ? "status" : "alert";
  const ariaLive = urgency === "polite" ? "polite" : "assertive";
  return /* @__PURE__ */ jsxs(
    "div",
    {
      role,
      "aria-live": ariaLive,
      "aria-atomic": "true",
      "data-slot": "notification-snackbar",
      className: cn(
        "flex w-full max-w-sm items-start gap-3 rounded border border-border bg-popover p-4 shadow-floating",
        className
      ),
      children: [
        /* @__PURE__ */ jsx("div", { "aria-hidden": "true", className: "mt-0.5 shrink-0 text-foreground", children: icon }),
        /* @__PURE__ */ jsxs("div", { className: "min-w-0 flex-1", children: [
          messageTitle ? /* @__PURE__ */ jsx("p", { className: "line-clamp-1 text-sm font-heading text-foreground", children: messageTitle }) : null,
          /* @__PURE__ */ jsx("p", { className: "line-clamp-3 text-sm text-muted-foreground", children: message }),
          action ? /* @__PURE__ */ jsx(
            Button,
            {
              variant: "link",
              size: "sm",
              "aria-label": action.label ?? action.text,
              className: "mt-1 h-auto p-0 text-sm",
              onClick: action.fn,
              children: action.text
            }
          ) : null
        ] }),
        hideDismissAction ? null : /* @__PURE__ */ jsx(
          Button,
          {
            type: "button",
            variant: "ghost",
            size: "icon",
            "aria-label": "Dismiss notification",
            className: "h-6 w-6 shrink-0",
            onClick: onClose,
            children: /* @__PURE__ */ jsx(X, { "aria-hidden": "true", className: "h-3.5 w-3.5" })
          }
        )
      ]
    }
  );
}

export {
  NotificationSnackbar
};
