// @docker/trident/dist/chunk-UBN6BSAX.js
import {
  Link
} from "./chunk-X4ZQ4UF6.js";
import {
  Tooltip,
  TooltipContent,
  TooltipTrigger
} from "./chunk-HVYJXDDW.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/external-link.tsx
import { jsx, jsxs } from "react/jsx-runtime";
function ExternalIcon({ className }) {
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      viewBox: "0 0 24 24",
      fill: "none",
      stroke: "currentColor",
      strokeWidth: "2",
      strokeLinecap: "round",
      strokeLinejoin: "round",
      className: cn("size-3", className),
      "aria-hidden": "true",
      children: [
        /* @__PURE__ */ jsx("path", { d: "M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6" }),
        /* @__PURE__ */ jsx("polyline", { points: "15 3 21 3 21 9" }),
        /* @__PURE__ */ jsx("line", { x1: "10", y1: "14", x2: "21", y2: "3" })
      ]
    }
  );
}
function ExternalLink({
  className,
  children,
  showIcon = true,
  tooltip,
  title,
  tooltipPlacement = "top",
  href,
  onAuxClick,
  onClick,
  ...props
}) {
  const tooltipContent = tooltip || title || (href ?? void 0);
  const link = /* @__PURE__ */ jsxs(
    Link,
    {
      "data-slot": "external-link",
      href,
      target: "_blank",
      rel: "noopener noreferrer",
      className: cn("inline-flex items-baseline gap-0.5", className),
      onAuxClick: (event) => {
        if (onAuxClick) {
          onAuxClick(event);
          return;
        }
        if (event.button === 1) {
          onClick?.(event);
        }
      },
      onClick,
      ...props,
      children: [
        children,
        /* @__PURE__ */ jsx("span", { className: "sr-only", children: "(opens in new tab)" }),
        (showIcon || !children) && /* @__PURE__ */ jsx(ExternalIcon, { className: "relative top-px shrink-0" })
      ]
    }
  );
  if (!tooltipContent) return link;
  return /* @__PURE__ */ jsxs(Tooltip, { children: [
    /* @__PURE__ */ jsx(TooltipTrigger, { asChild: true, children: link }),
    /* @__PURE__ */ jsx(TooltipContent, { side: tooltipPlacement, children: tooltipContent })
  ] });
}

export {
  ExternalLink
};
