// @docker/trident/dist/chunk-Q6QXRLTV.js
import {
  EditableCell
} from "./chunk-DK3EKQAE.js";
import {
  unknownToStr
} from "./chunk-MICCEESH.js";
import {
  TableCell,
  TableRow
} from "./chunk-SKOSYJCU.js";
import {
  TEXT_ALIGN_CLASS
} from "./chunk-BBDVHE6E.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/data-grid-tree-row.tsx
import { flexRender } from "@tanstack/react-table";
import { ChevronDown, ChevronRight } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var INDENT_PX = 20;
var TreeCell = ({
  cell,
  cClass,
  isFirstDataCell,
  treeMeta,
  rowId,
  editingCell,
  onStartEdit,
  onCommit,
  onCancel,
  onToggle
}) => {
  const cMeta = cell.column.columnDef.meta;
  const cAlign = cMeta?.align ?? "left";
  const isPinned = cell.column.getIsPinned();
  const isEditable = cMeta?.editable === true;
  const isEditing = isEditable && editingCell?.rowId === rowId && editingCell?.field === cell.column.id;
  const handleKeyDown = React.useCallback(
    (e) => {
      if (e.key === "Enter" || e.key === " ") {
        e.preventDefault();
        onToggle(treeMeta.pathKey);
      }
    },
    [onToggle, treeMeta.pathKey]
  );
  const cellContent = isEditing ? /* @__PURE__ */ jsx(
    EditableCell,
    {
      value: unknownToStr(cell.getValue()),
      onCommit,
      onCancel
    }
  ) : flexRender(cell.column.columnDef.cell, cell.getContext());
  return /* @__PURE__ */ jsx(
    TableCell,
    {
      "data-align": cAlign,
      className: cn(
        cClass,
        TEXT_ALIGN_CLASS[cAlign],
        isPinned && "bg-background",
        isEditable && !isEditing && "cursor-text"
      ),
      style: {
        position: isPinned ? "sticky" : void 0,
        left: isPinned === "left" ? `${cell.column.getStart("left")}px` : void 0,
        right: isPinned === "right" ? `${cell.column.getAfter("right")}px` : void 0
      },
      onClick: isEditable && !isEditing ? (e) => {
        e.stopPropagation();
        onStartEdit({ rowId, field: cell.column.id });
      } : void 0,
      children: isFirstDataCell ? /* @__PURE__ */ jsxs(
        "span",
        {
          className: "flex items-center",
          style: { paddingLeft: `${treeMeta.indent * INDENT_PX}px` },
          children: [
            treeMeta.hasChildren ? /* @__PURE__ */ jsx(
              "button",
              {
                type: "button",
                "data-slot": "tree-expand",
                "aria-expanded": treeMeta.isExpanded,
                "aria-label": treeMeta.isExpanded ? "Collapse row" : "Expand row",
                onClick: (e) => {
                  e.stopPropagation();
                  onToggle(treeMeta.pathKey);
                },
                onKeyDown: handleKeyDown,
                className: "mr-1 flex shrink-0 items-center rounded text-muted-foreground hover:text-foreground focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none",
                children: treeMeta.isExpanded ? /* @__PURE__ */ jsx(ChevronDown, { className: "size-4" }) : /* @__PURE__ */ jsx(ChevronRight, { className: "size-4" })
              }
            ) : /* @__PURE__ */ jsx(
              "span",
              {
                className: "mr-1 inline-block size-4 shrink-0",
                "aria-hidden": "true"
              }
            ),
            cellContent
          ]
        }
      ) : cellContent
    }
  );
};
var DataGridTreeRow = ({
  row,
  treeMeta,
  cClass,
  editingCell,
  onStartEdit,
  onCommitEdit,
  onCancelEdit,
  onToggle,
  onRowClick,
  disableRowSelectionOnClick
}) => {
  const cells = row.getVisibleCells();
  const firstDataCellIndex = cells.findIndex((c) => c.column.id !== "__select__");
  const rowInteractive = Boolean(onRowClick);
  const handleRowActivate = onRowClick ? () => onRowClick({ row: row.original, id: row.id }) : void 0;
  return /* @__PURE__ */ jsx(
    TableRow,
    {
      "data-slot": "data-grid-row",
      "data-state": row.getIsSelected() ? "selected" : void 0,
      "aria-level": treeMeta.indent + 1,
      "aria-expanded": treeMeta.hasChildren ? treeMeta.isExpanded : void 0,
      tabIndex: rowInteractive ? 0 : void 0,
      "aria-label": rowInteractive ? `Row ${row.id}` : void 0,
      className: cn(
        rowInteractive && !disableRowSelectionOnClick && "cursor-pointer",
        rowInteractive && "focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none"
      ),
      onClick: handleRowActivate,
      onKeyDown: rowInteractive && handleRowActivate ? (e) => {
        if (e.target !== e.currentTarget) return;
        if (e.key === "Enter") {
          e.preventDefault();
          handleRowActivate();
        } else if (e.key === " ") {
          e.preventDefault();
          handleRowActivate();
        }
      } : void 0,
      children: cells.map((cell, cellIndex) => /* @__PURE__ */ jsx(
        TreeCell,
        {
          cell,
          cClass,
          isFirstDataCell: cellIndex === firstDataCellIndex,
          treeMeta,
          rowId: row.id,
          editingCell,
          onStartEdit,
          onCommit: (v) => onCommitEdit(row.original, cell.column.id, v),
          onCancel: onCancelEdit,
          onToggle
        },
        cell.id
      ))
    }
  );
};

export {
  DataGridTreeRow
};
