// @docker/trident/dist/chunk-3LIKYSE2.js
import {
  Calendar
} from "./chunk-SL6PT4EM.js";
import {
  Popover,
  PopoverAnchor,
  PopoverContent,
  PopoverTrigger
} from "./chunk-R6CWLGJU.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/date-range-picker.tsx
import {
  endOfMonth,
  format,
  isValid,
  parse,
  startOfMonth,
  startOfToday,
  subDays
} from "date-fns";
import { CalendarIcon } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var INPUT_FORMAT = "MM/dd/yyyy";
var INPUT_PLACEHOLDER = "MM/DD/YYYY";
var PARSE_FORMATS = [
  "MM/dd/yyyy",
  "M/d/yyyy",
  "MMM d, yyyy",
  "yyyy-MM-dd",
  "MM/dd/yy"
];
function tryParseDate(str) {
  const trimmed = str.trim();
  if (!trimmed) return null;
  const ref = /* @__PURE__ */ new Date();
  for (const fmt of PARSE_FORMATS) {
    const parsedDate = parse(trimmed, fmt, ref);
    if (isValid(parsedDate)) return parsedDate;
  }
  return null;
}
var getDefaultPresets = () => {
  const now = /* @__PURE__ */ new Date();
  return [
    { label: "Last 7 days", range: { from: subDays(now, 6), to: now } },
    { label: "Last 30 days", range: { from: subDays(now, 29), to: now } },
    { label: "Last 90 days", range: { from: subDays(now, 89), to: now } },
    {
      label: "This month",
      range: { from: startOfMonth(now), to: endOfMonth(now) }
    }
  ];
};
var DEFAULT_PRESETS = getDefaultPresets();
function DateRangePicker({
  value,
  onValueChange,
  presets = DEFAULT_PRESETS,
  placeholder = "Pick a date range",
  disabled = false,
  minDate,
  maxDate,
  disableFuture = false,
  disablePast = false,
  shouldDisableDate,
  numberOfMonths = 2,
  open: controlledOpen,
  onOpen,
  onClose,
  variant = "button",
  className
}) {
  const [internalOpen, setInternalOpen] = React.useState(false);
  const isControlled = controlledOpen !== void 0;
  const open = isControlled ? controlledOpen : internalOpen;
  const [draft, setDraft] = React.useState(value);
  const [fromStr, setFromStr] = React.useState(
    value?.from ? format(value.from, INPUT_FORMAT) : ""
  );
  const [toStr, setToStr] = React.useState(
    value?.to ? format(value.to, INPUT_FORMAT) : ""
  );
  React.useEffect(() => {
    if (open) setDraft(value);
    else {
      setFromStr(value?.from ? format(value.from, INPUT_FORMAT) : "");
      setToStr(value?.to ? format(value.to, INPUT_FORMAT) : "");
    }
  }, [open, value]);
  const commitClose = (next, nextFrom, nextTo) => {
    onValueChange?.(next);
    setFromStr(nextFrom);
    setToStr(nextTo);
    if (!isControlled) setInternalOpen(false);
    onClose?.();
  };
  const handleOpenChange = React.useCallback(
    (next) => {
      if (!isControlled) setInternalOpen(next);
      if (next && !open) onOpen?.();
      if (!next && open) onClose?.();
    },
    [isControlled, open, onOpen, onClose]
  );
  const handleApply = () => commitClose(
    draft,
    draft?.from ? format(draft.from, INPUT_FORMAT) : "",
    draft?.to ? format(draft.to, INPUT_FORMAT) : ""
  );
  const handlePreset = (range) => commitClose(
    range,
    range.from ? format(range.from, INPUT_FORMAT) : "",
    range.to ? format(range.to, INPUT_FORMAT) : ""
  );
  const handleClear = () => commitClose(void 0, "", "");
  const handleRangeSelect = (range) => {
    setDraft(range);
    setFromStr(range?.from ? format(range.from, INPUT_FORMAT) : "");
    setToStr(range?.to ? format(range.to, INPUT_FORMAT) : "");
  };
  const handleFromBlur = () => {
    const parsed = tryParseDate(fromStr);
    if (parsed) {
      setDraft((prev) => ({ ...prev, from: parsed }));
    } else {
      setFromStr(draft?.from ? format(draft.from, INPUT_FORMAT) : "");
    }
  };
  const handleToBlur = () => {
    const parsed = tryParseDate(toStr);
    if (parsed) {
      setDraft((prev) => ({ from: prev?.from, to: parsed }));
    } else {
      setToStr(draft?.to ? format(draft.to, INPUT_FORMAT) : "");
    }
  };
  const triggerLabel = value?.from ? value.to ? `${format(value.from, "MMM d, yyyy")} \u2013 ${format(value.to, "MMM d, yyyy")}` : format(value.from, "MMM d, yyyy") : placeholder;
  const calendarDisabled = React.useCallback(
    (date) => {
      if (minDate && date < minDate) return true;
      if (maxDate && date > maxDate) return true;
      if (disableFuture && date > /* @__PURE__ */ new Date()) return true;
      if (disablePast && date < startOfToday()) return true;
      if (shouldDisableDate?.(date)) return true;
      return false;
    },
    [minDate, maxDate, disableFuture, disablePast, shouldDisableDate]
  );
  const popoverContent = /* @__PURE__ */ jsx(
    PopoverContent,
    {
      align: "start",
      className: "w-auto p-0",
      "aria-label": "Date range picker",
      children: /* @__PURE__ */ jsxs("div", { "data-slot": "date-range-picker", className: "flex flex-col md:flex-row", children: [
        presets.length > 0 && /* @__PURE__ */ jsx(
          "div",
          {
            "data-slot": "date-range-presets",
            className: "flex flex-row flex-wrap gap-1 border-b border-border p-3 md:flex-col md:border-r md:border-b-0",
            children: presets.map((preset) => /* @__PURE__ */ jsx(
              Button,
              {
                variant: "ghost",
                size: "sm",
                onClick: () => handlePreset(preset.range),
                className: "justify-start text-sm",
                children: preset.label
              },
              preset.label
            ))
          }
        ),
        /* @__PURE__ */ jsxs("div", { "data-slot": "date-range-calendars", className: "p-3", children: [
          /* @__PURE__ */ jsx(
            Calendar,
            {
              mode: "range",
              selected: draft,
              onSelect: handleRangeSelect,
              numberOfMonths,
              disabled: calendarDisabled
            }
          ),
          /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-end gap-2 border-t border-border pt-3", children: [
            /* @__PURE__ */ jsx(Button, { variant: "ghost", size: "sm", onClick: handleClear, children: "Clear" }),
            /* @__PURE__ */ jsx(Button, { size: "sm", onClick: handleApply, children: "Apply" })
          ] })
        ] })
      ] })
    }
  );
  if (variant === "input") {
    return /* @__PURE__ */ jsxs(Popover, { open, onOpenChange: handleOpenChange, children: [
      /* @__PURE__ */ jsx(PopoverAnchor, { asChild: true, children: /* @__PURE__ */ jsxs(
        "div",
        {
          "data-slot": "date-range-input",
          className: cn(
            "flex h-9 items-center gap-1 rounded border border-input bg-transparent px-2.5 text-sm shadow-raised transition-[border-color,box-shadow] focus-within:border-ring focus-within:ring-3 focus-within:ring-ring",
            disabled && "pointer-events-none cursor-not-allowed opacity-50",
            className
          ),
          children: [
            /* @__PURE__ */ jsx(
              CalendarIcon,
              {
                className: "size-4 shrink-0 text-muted-foreground",
                "aria-hidden": "true"
              }
            ),
            /* @__PURE__ */ jsx(
              "input",
              {
                type: "text",
                className: "w-[5.5rem] min-w-0 bg-transparent outline-none placeholder:text-muted-foreground",
                value: fromStr,
                onChange: (e) => setFromStr(e.target.value),
                onFocus: () => !open && handleOpenChange(true),
                onBlur: handleFromBlur,
                placeholder: INPUT_PLACEHOLDER,
                "aria-label": "Start date",
                disabled
              }
            ),
            /* @__PURE__ */ jsx("span", { className: "text-muted-foreground", "aria-hidden": "true", children: "\u2013" }),
            /* @__PURE__ */ jsx(
              "input",
              {
                type: "text",
                className: "w-[5.5rem] min-w-0 bg-transparent outline-none placeholder:text-muted-foreground",
                value: toStr,
                onChange: (e) => setToStr(e.target.value),
                onFocus: () => !open && handleOpenChange(true),
                onBlur: handleToBlur,
                placeholder: INPUT_PLACEHOLDER,
                "aria-label": "End date",
                disabled
              }
            )
          ]
        }
      ) }),
      popoverContent
    ] });
  }
  return /* @__PURE__ */ jsxs(Popover, { open, onOpenChange: handleOpenChange, children: [
    /* @__PURE__ */ jsx(PopoverTrigger, { asChild: true, children: /* @__PURE__ */ jsxs(
      Button,
      {
        variant: "outline",
        disabled,
        "aria-label": triggerLabel,
        className: cn(
          "justify-start gap-2 font-normal",
          !value?.from && "text-muted-foreground",
          className
        ),
        children: [
          /* @__PURE__ */ jsx(CalendarIcon, { className: "size-4 shrink-0", "aria-hidden": "true" }),
          /* @__PURE__ */ jsx("span", { children: triggerLabel })
        ]
      }
    ) }),
    popoverContent
  ] });
}

export {
  getDefaultPresets,
  DEFAULT_PRESETS,
  DateRangePicker
};
