// @docker/trident/dist/chunk-6D3M5CWS.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/trust-badge.tsx
import { cva } from "class-variance-authority";
import { Slot } from "radix-ui";
import { jsx } from "react/jsx-runtime";
var trustBadgeVariants = cva(
  "inline-flex items-center rounded border px-1.5 py-0.5 text-xs leading-none font-label select-none",
  {
    variants: {
      tier: {
        /** Docker Official Image — highest trust, Docker-maintained */
        official: "border-trust-official-border bg-trust-official-bg text-trust-official-text",
        /** Docker Hardened Image — distroless, security-hardened */
        hardened: "border-trust-hardened-border bg-trust-hardened-bg text-trust-hardened-text",
        /** Verified Publisher — vetted commercial/OSS publisher */
        verified: "border-trust-verified-border bg-trust-verified-bg text-trust-verified-text",
        /** Community — unvetted, public */
        community: "border-trust-community-border bg-trust-community-bg text-trust-community-text"
      }
    },
    defaultVariants: {
      tier: "community"
    }
  }
);
var TIER_LABELS = {
  official: "Docker Official",
  hardened: "Docker Hardened",
  verified: "Verified Publisher",
  community: "Community"
};
function TrustBadge({
  tier,
  className,
  asChild = false,
  children,
  ...props
}) {
  const Comp = asChild ? Slot.Root : "span";
  return /* @__PURE__ */ jsx(
    Comp,
    {
      "data-slot": "trust-badge",
      className: cn(trustBadgeVariants({ tier }), className),
      ...props,
      children: asChild ? children : TIER_LABELS[tier ?? "community"]
    }
  );
}

export {
  trustBadgeVariants,
  TrustBadge
};
