// @docker/trident/dist/chunk-LJSD3DSH.js
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/copy-to-clipboard-button.tsx
import { Check, ClipboardCopy } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function CopyToClipboardButton({
  text,
  showLabel = false,
  onCopy,
  className,
  size,
  ...props
}) {
  const [copied, setCopied] = React.useState(false);
  React.useEffect(() => {
    if (!copied) return;
    const timer = setTimeout(() => setCopied(false), 2e3);
    return () => clearTimeout(timer);
  }, [copied]);
  const handleCopy = React.useCallback(async () => {
    try {
      await navigator.clipboard.writeText(text);
      setCopied(true);
      onCopy?.();
    } catch {
      console.error("[CopyToClipboardButton] Failed to copy to clipboard");
    }
  }, [text, onCopy]);
  return /* @__PURE__ */ jsxs(
    Button,
    {
      "data-slot": "copy-to-clipboard-button",
      variant: "ghost",
      size: size ?? (showLabel ? "sm" : "icon-sm"),
      onClick: handleCopy,
      "aria-label": copied ? "Copied!" : "Copy to clipboard",
      className: cn(className),
      ...props,
      children: [
        copied ? /* @__PURE__ */ jsx(Check, { className: "size-3.5", "aria-hidden": "true" }) : /* @__PURE__ */ jsx(ClipboardCopy, { className: "size-3.5", "aria-hidden": "true" }),
        showLabel && /* @__PURE__ */ jsx("span", { children: copied ? "Copied!" : "Copy" })
      ]
    }
  );
}

export {
  CopyToClipboardButton
};
