// @docker/trident/dist/chunk-7GDHTOZE.js
import {
  Skeleton
} from "./chunk-N2BJEDAO.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/virtual-list.tsx
import { useVirtualizer } from "@tanstack/react-virtual";
import { useRef } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var DEFAULT_ESTIMATE_SIZE = 48;
var DEFAULT_OVERSCAN = 5;
var DEFAULT_HEIGHT = 400;
var DEFAULT_LOADING_ROWS = 6;
function VirtualList({
  items,
  estimateSize = DEFAULT_ESTIMATE_SIZE,
  overscan = DEFAULT_OVERSCAN,
  height = DEFAULT_HEIGHT,
  renderItem,
  getItemKey,
  className,
  emptyState,
  header,
  footer,
  loading = false,
  loadingRows = DEFAULT_LOADING_ROWS,
  style,
  ...divProps
}) {
  const scrollRef = useRef(null);
  const virtualizer = useVirtualizer({
    count: items.length,
    getScrollElement: () => scrollRef.current,
    estimateSize: () => estimateSize,
    overscan,
    getItemKey: getItemKey ? (index) => getItemKey(items[index], index) : void 0
  });
  const containerStyle = {
    height: typeof height === "number" ? `${height}px` : height,
    ...style
  };
  if (loading) {
    return /* @__PURE__ */ jsxs(
      "div",
      {
        "data-slot": "virtual-list",
        tabIndex: 0,
        ...divProps,
        className: cn("overflow-y-auto", className),
        style: containerStyle,
        children: [
          header,
          /* @__PURE__ */ jsx("div", { className: "flex flex-col gap-2 p-2", children: Array.from({ length: loadingRows }).map((_, i) => /* @__PURE__ */ jsx(Skeleton, { className: "h-10 w-full rounded" }, i)) }),
          footer
        ]
      }
    );
  }
  if (items.length === 0) {
    return /* @__PURE__ */ jsxs(
      "div",
      {
        "data-slot": "virtual-list",
        tabIndex: 0,
        ...divProps,
        className: cn("overflow-y-auto", className),
        style: containerStyle,
        children: [
          header,
          /* @__PURE__ */ jsx("div", { className: "flex h-full items-center justify-center text-sm text-muted-foreground", children: emptyState ?? "No items" }),
          footer
        ]
      }
    );
  }
  const virtualItems = virtualizer.getVirtualItems();
  const totalSize = virtualizer.getTotalSize();
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "virtual-list",
      tabIndex: 0,
      ...divProps,
      ref: scrollRef,
      className: cn("overflow-y-auto", className),
      style: containerStyle,
      children: [
        header,
        /* @__PURE__ */ jsx("div", { style: { position: "relative", height: `${totalSize}px` }, children: virtualItems.map((virtualRow) => /* @__PURE__ */ jsx(
          "div",
          {
            "data-index": virtualRow.index,
            ref: virtualizer.measureElement,
            style: {
              position: "absolute",
              top: 0,
              left: 0,
              width: "100%",
              transform: `translateY(${virtualRow.start}px)`
            },
            children: renderItem(items[virtualRow.index], virtualRow.index)
          },
          virtualRow.key
        )) }),
        footer
      ]
    }
  );
}

export {
  VirtualList
};
