// @docker/trident/dist/chunk-DK3EKQAE.js
import {
  unknownToStr
} from "./chunk-MICCEESH.js";
import {
  TableCell
} from "./chunk-SKOSYJCU.js";
import {
  TEXT_ALIGN_CLASS
} from "./chunk-BBDVHE6E.js";
import {
  Input
} from "./chunk-5CQB3X63.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/data-grid-edit.tsx
import { flexRender } from "@tanstack/react-table";
import * as React from "react";
import { jsx } from "react/jsx-runtime";
var EditableCell = ({ value, onCommit, onCancel }) => {
  const [draft, setDraft] = React.useState(value);
  const committed = React.useRef(false);
  const commit = React.useCallback(
    (val) => {
      if (committed.current) return;
      committed.current = true;
      onCommit(val);
    },
    [onCommit]
  );
  const handleKeyDown = React.useCallback(
    (e) => {
      if (e.key === "Enter") {
        e.preventDefault();
        commit(draft);
      } else if (e.key === "Escape") {
        e.preventDefault();
        committed.current = true;
        onCancel();
      }
    },
    [draft, commit, onCancel]
  );
  const inputRef = React.useRef(null);
  React.useEffect(() => {
    inputRef.current?.focus();
  }, []);
  return /* @__PURE__ */ jsx(
    Input,
    {
      ref: inputRef,
      size: "sm",
      value: draft,
      "aria-label": "Edit cell value",
      onChange: (e) => setDraft(e.target.value),
      onKeyDown: handleKeyDown,
      onBlur: () => commit(draft),
      onClick: (e) => e.stopPropagation()
    }
  );
};
var DataGridCell = ({
  cell,
  cClass,
  rowId,
  editingCell,
  onStartEdit,
  onCommit,
  onCancel
}) => {
  const cMeta = cell.column.columnDef.meta;
  const cAlign = cMeta?.align ?? "left";
  const isPinned = cell.column.getIsPinned();
  const isEditable = cMeta?.editable === true;
  const isEditing = isEditable && editingCell?.rowId === rowId && editingCell?.field === cell.column.id;
  const canStartEdit = isEditable && !isEditing;
  const handleKeyDown = canStartEdit ? (e) => {
    if (e.key === "Enter" || e.key === "F2" || e.key === " ") {
      e.preventDefault();
      e.stopPropagation();
      onStartEdit({ rowId, field: cell.column.id });
    }
  } : void 0;
  return /* @__PURE__ */ jsx(
    TableCell,
    {
      role: "gridcell",
      tabIndex: canStartEdit ? 0 : -1,
      "data-align": cAlign,
      className: cn(
        cClass,
        TEXT_ALIGN_CLASS[cAlign],
        isPinned && "bg-background",
        canStartEdit && "cursor-text focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none focus-visible:ring-inset"
      ),
      style: {
        position: isPinned ? "sticky" : void 0,
        left: isPinned === "left" ? `${cell.column.getStart("left")}px` : void 0,
        right: isPinned === "right" ? `${cell.column.getAfter("right")}px` : void 0
      },
      onClick: canStartEdit ? (e) => {
        e.stopPropagation();
        onStartEdit({ rowId, field: cell.column.id });
      } : void 0,
      onKeyDown: handleKeyDown,
      children: isEditing ? /* @__PURE__ */ jsx(
        EditableCell,
        {
          value: unknownToStr(cell.getValue()),
          onCommit,
          onCancel
        }
      ) : flexRender(cell.column.columnDef.cell, cell.getContext())
    },
    cell.id
  );
};

export {
  EditableCell,
  DataGridCell
};
