// @docker/trident/dist/chunk-AEEGWDL3.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/alert.tsx
import { cva } from "class-variance-authority";
import { jsx } from "react/jsx-runtime";
var alertVariants = cva(
  "group/alert relative grid w-full gap-0.5 rounded border px-4 py-3 text-left text-sm has-data-[slot=alert-action]:relative has-data-[slot=alert-action]:pr-18 has-[>svg]:grid-cols-[auto_1fr] has-[>svg]:gap-x-2.5 *:[svg]:row-span-2 *:[svg]:translate-y-0.5 *:[svg]:text-current *:[svg:not([class*='size-'])]:size-4",
  {
    variants: {
      variant: {
        default: "bg-card text-card-foreground",
        destructive: "bg-alert-error-bg text-alert-error-fg *:data-[slot=alert-description]:text-alert-error-fg *:[svg]:text-alert-error-icon",
        success: "bg-alert-success-bg text-alert-success-fg *:data-[slot=alert-description]:text-alert-success-fg *:[svg]:text-alert-success-icon",
        warning: "bg-alert-warning-bg text-alert-warning-fg *:data-[slot=alert-description]:text-alert-warning-fg *:[svg]:text-alert-warning-icon",
        info: "bg-alert-info-bg text-alert-info-fg *:data-[slot=alert-description]:text-alert-info-fg *:[svg]:text-alert-info-icon"
      }
    },
    defaultVariants: {
      variant: "default"
    }
  }
);
function Alert({
  className,
  variant,
  ...props
}) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "alert",
      role: "alert",
      className: cn(alertVariants({ variant }), className),
      ...props
    }
  );
}
function AlertTitle({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "alert-title",
      className: cn(
        "leading-heading font-heading tracking-heading group-has-[>svg]/alert:col-start-2 [&_a]:underline [&_a]:underline-offset-3 [&_a]:hover:text-foreground",
        className
      ),
      ...props
    }
  );
}
function AlertDescription({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "alert-description",
      className: cn(
        "text-sm text-balance text-muted-foreground md:text-pretty [&_a]:underline [&_a]:underline-offset-3 [&_a]:hover:text-foreground [&_p:not(:last-child)]:mb-4",
        className
      ),
      ...props
    }
  );
}
function AlertAction({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "alert-action",
      className: cn("absolute top-2.5 right-3", className),
      ...props
    }
  );
}

export {
  Alert,
  AlertTitle,
  AlertDescription,
  AlertAction
};
