// @docker/trident/dist/chunk-DMKQJJJ4.js
import {
  Calendar
} from "./chunk-SL6PT4EM.js";
import {
  Popover,
  PopoverContent,
  PopoverTrigger
} from "./chunk-R6CWLGJU.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/date-picker.tsx
import { format } from "date-fns";
import { CalendarIcon } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function DatePicker({
  value,
  onChange,
  placeholder = "Pick a date",
  disabled,
  className,
  calendarProps
}) {
  return /* @__PURE__ */ jsxs(Popover, { children: [
    /* @__PURE__ */ jsx(PopoverTrigger, { asChild: true, children: /* @__PURE__ */ jsxs(
      Button,
      {
        variant: "outline",
        "data-empty": !value,
        disabled,
        className: cn(
          "w-[var(--dds-width-date-picker-sm)] justify-start text-left font-normal data-[empty=true]:text-muted-foreground",
          className
        ),
        children: [
          /* @__PURE__ */ jsx(CalendarIcon, {}),
          value ? format(value, "PPP") : /* @__PURE__ */ jsx("span", { children: placeholder })
        ]
      }
    ) }),
    /* @__PURE__ */ jsx(PopoverContent, { className: "w-auto p-0", align: "start", children: /* @__PURE__ */ jsx(
      Calendar,
      {
        ...calendarProps,
        mode: "single",
        selected: value,
        onSelect: onChange,
        initialFocus: calendarProps?.initialFocus ?? true
      }
    ) })
  ] });
}
function DateRangePicker({
  value,
  onChange,
  placeholder = "Pick a date range",
  disabled,
  numberOfMonths = 2,
  className,
  calendarProps
}) {
  const label = React.useMemo(() => {
    if (!value?.from) return placeholder;
    if (!value.to) return format(value.from, "LLL dd, y");
    return `${format(value.from, "LLL dd, y")} - ${format(value.to, "LLL dd, y")}`;
  }, [value, placeholder]);
  return /* @__PURE__ */ jsxs(Popover, { children: [
    /* @__PURE__ */ jsx(PopoverTrigger, { asChild: true, children: /* @__PURE__ */ jsxs(
      Button,
      {
        variant: "outline",
        "data-empty": !value?.from,
        disabled,
        className: cn(
          "w-[var(--dds-width-date-picker-md)] justify-start text-left font-normal data-[empty=true]:text-muted-foreground",
          className
        ),
        children: [
          /* @__PURE__ */ jsx(CalendarIcon, {}),
          /* @__PURE__ */ jsx("span", { children: label })
        ]
      }
    ) }),
    /* @__PURE__ */ jsx(PopoverContent, { className: "w-auto p-0", align: "start", children: /* @__PURE__ */ jsx(
      Calendar,
      {
        ...calendarProps,
        mode: "range",
        selected: value,
        onSelect: onChange,
        numberOfMonths: calendarProps?.numberOfMonths ?? numberOfMonths,
        initialFocus: calendarProps?.initialFocus ?? true
      }
    ) })
  ] });
}

export {
  DatePicker,
  DateRangePicker
};
