// @docker/trident/dist/chunk-EI4H22EW.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/relative-time.tsx
import { jsx } from "react/jsx-runtime";
function formatRelative(date) {
  const now = Date.now();
  const diffMs = now - date.getTime();
  const diffSecs = Math.round(diffMs / 1e3);
  if (diffSecs < 60) return "just now";
  const diffMins = Math.round(diffSecs / 60);
  if (diffMins < 60)
    return `${diffMins} minute${diffMins !== 1 ? "s" : ""} ago`;
  const diffHours = Math.round(diffMins / 60);
  if (diffHours < 24)
    return `${diffHours} hour${diffHours !== 1 ? "s" : ""} ago`;
  const diffDays = Math.round(diffHours / 24);
  if (diffDays < 30) return `${diffDays} day${diffDays !== 1 ? "s" : ""} ago`;
  const diffMonths = Math.round(diffDays / 30);
  if (diffMonths < 12)
    return `${diffMonths} month${diffMonths !== 1 ? "s" : ""} ago`;
  const diffYears = Math.round(diffMonths / 12);
  return `${diffYears} year${diffYears !== 1 ? "s" : ""} ago`;
}
function RelativeTime({ date, className }) {
  const dateObj = date instanceof Date ? date : new Date(date);
  const isoString = dateObj.toISOString();
  const label = formatRelative(dateObj);
  return /* @__PURE__ */ jsx(
    "time",
    {
      "data-slot": "relative-time",
      dateTime: isoString,
      title: dateObj.toLocaleString(),
      className: cn("text-sm text-muted-foreground", className),
      children: label
    }
  );
}

export {
  RelativeTime
};
