// @docker/trident/dist/chunk-IJW7EK26.js
import {
  Checkbox
} from "./chunk-R5XEJKHJ.js";
import {
  DataGridFooter
} from "./chunk-HU6TU2L2.js";
import {
  DataGridGroupRow,
  DataGridHeader
} from "./chunk-KXARZUZR.js";
import {
  applyFilters
} from "./chunk-BJJFXHWE.js";
import {
  DataGridTreeRow
} from "./chunk-Q6QXRLTV.js";
import {
  DataGridCell
} from "./chunk-DK3EKQAE.js";
import {
  buildColumnPinning,
  exportToCsv,
  toColumnDef,
  unknownToStr
} from "./chunk-MICCEESH.js";
import {
  Table,
  TableBody,
  TableCell,
  TableRow
} from "./chunk-SKOSYJCU.js";
import {
  buildTreeDisplayList,
  toggleTreePath
} from "./chunk-3FGH5OXJ.js";
import {
  CELL_CLASS,
  HEADER_CLASS
} from "./chunk-BBDVHE6E.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/data-grid.tsx
import {
  getCoreRowModel,
  getExpandedRowModel,
  getFilteredRowModel,
  getGroupedRowModel,
  getPaginationRowModel,
  getSortedRowModel,
  useReactTable
} from "@tanstack/react-table";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function DataGrid({
  rows,
  columns,
  loading = false,
  pagination = false,
  pageSize = 25,
  rowsPerPageOptions,
  rowCount,
  onRowClick,
  checkboxSelection = false,
  disableRowSelectionOnClick = false,
  sortingMode = "client",
  onSortModelChange,
  getRowId,
  slots,
  slotProps,
  density = "default",
  disableColumnMenu = false,
  disableColumnResize = false,
  disableColumnReorder = false,
  pinnedColumns,
  rowGroupingModel,
  onRowGroupingModelChange,
  columnVisibilityModel,
  onColumnVisibilityModelChange,
  processRowUpdate,
  onProcessRowUpdateError,
  apiRef,
  hideFooter = false,
  autoHeight = false,
  className,
  filterModel: filterModelProp,
  onFilterModelChange,
  disableColumnFilter = false,
  treeData = false,
  getTreeDataPath,
  "aria-label": ariaLabel,
  "aria-labelledby": ariaLabelledby,
  onRowSelectionChange
}) {
  const [editingCell, setEditingCell] = React.useState(null);
  const [sorting, setSorting] = React.useState([]);
  const [paginationState, setPaginationState] = React.useState(
    { pageIndex: 0, pageSize }
  );
  const [rowSelection, setRowSelection] = React.useState({});
  const [columnSizing, setColumnSizing] = React.useState({});
  const [grouping, setGrouping] = React.useState(
    rowGroupingModel ?? []
  );
  const [expanded, setExpanded] = React.useState({});
  const [treeExpandedPaths, setTreeExpandedPaths] = React.useState(
    () => /* @__PURE__ */ new Set()
  );
  const [internalFilterModel, setInternalFilterModel] = React.useState([]);
  const filterModel = filterModelProp ?? internalFilterModel;
  const handleFilterChange = React.useCallback(
    (filter, columnId) => {
      const next = filter ? [...filterModel.filter((f) => f.columnId !== columnId), filter] : filterModel.filter((f) => f.columnId !== columnId);
      if (filterModelProp === void 0) setInternalFilterModel(next);
      onFilterModelChange?.(next);
    },
    [filterModel, filterModelProp, onFilterModelChange]
  );
  const [columnVisibility, setColumnVisibility] = React.useState(() => {
    if (columnVisibilityModel !== void 0) return columnVisibilityModel;
    return Object.fromEntries(
      columns.filter((c) => c.hide).map((c) => [c.field, false])
    );
  });
  React.useEffect(() => {
    setPaginationState((prev) => ({ ...prev, pageSize }));
  }, [pageSize]);
  React.useEffect(() => {
    if (rowGroupingModel !== void 0) setGrouping(rowGroupingModel);
  }, [rowGroupingModel]);
  const [columnOrder, setColumnOrder] = React.useState(() => {
    const base = columns.map((c) => c.field);
    return checkboxSelection ? ["__select__", ...base] : base;
  });
  const columnPinning = React.useMemo(
    () => buildColumnPinning(columns, pinnedColumns),
    [columns, pinnedColumns]
  );
  const dragId = React.useRef(null);
  const columnDefs = React.useMemo(() => {
    const selectCol = {
      id: "__select__",
      header: ({ table: table2 }) => /* @__PURE__ */ jsx(
        Checkbox,
        {
          checked: table2.getIsAllRowsSelected() ? true : table2.getIsSomeRowsSelected() ? "indeterminate" : false,
          onCheckedChange: (v) => table2.toggleAllRowsSelected(v === true),
          "aria-label": "Select all rows"
        }
      ),
      cell: ({ row }) => /* @__PURE__ */ jsx(
        Checkbox,
        {
          checked: row.getIsSelected(),
          onCheckedChange: (v) => row.toggleSelected(v === true),
          "aria-label": `Select row ${row.id}`,
          onClick: (e) => e.stopPropagation()
        }
      ),
      enableSorting: false,
      size: 40
    };
    const dataCols = columns.map(toColumnDef);
    return checkboxSelection ? [selectCol, ...dataCols] : dataCols;
  }, [columns, checkboxSelection]);
  const rowIdFn = React.useCallback(
    (row, index) => {
      if (getRowId) return getRowId(row);
      const id = unknownToStr(Reflect.get(row, "id"));
      return id !== "" ? id : String(index);
    },
    [getRowId]
  );
  const filteredRows = React.useMemo(
    () => applyFilters(rows, filterModel, columns),
    [rows, filterModel, columns]
  );
  const treeDisplayList = React.useMemo(
    () => treeData && getTreeDataPath ? buildTreeDisplayList(filteredRows, getTreeDataPath, treeExpandedPaths) : null,
    [treeData, getTreeDataPath, filteredRows, treeExpandedPaths]
  );
  const tableRows = treeDisplayList ? treeDisplayList.map((item) => item.row) : filteredRows;
  const table = useReactTable({
    data: tableRows,
    columns: columnDefs,
    state: {
      sorting,
      pagination: paginationState,
      rowSelection,
      columnSizing,
      columnOrder,
      columnPinning,
      columnVisibility: columnVisibilityModel ?? columnVisibility,
      grouping,
      expanded
    },
    columnResizeMode: "onChange",
    onSortingChange: (updater) => {
      const next = typeof updater === "function" ? updater(sorting) : updater;
      setSorting(next);
      onSortModelChange?.(
        next.map((s) => ({ field: s.id, sort: s.desc ? "desc" : "asc" }))
      );
    },
    onColumnVisibilityChange: (updater) => {
      const current = columnVisibilityModel ?? columnVisibility;
      const next = typeof updater === "function" ? updater(current) : updater;
      setColumnVisibility(next);
      onColumnVisibilityModelChange?.(next);
    },
    onPaginationChange: setPaginationState,
    onRowSelectionChange: setRowSelection,
    onColumnSizingChange: setColumnSizing,
    onColumnOrderChange: setColumnOrder,
    onGroupingChange: (updater) => {
      const next = typeof updater === "function" ? updater(grouping) : updater;
      setGrouping(next);
      onRowGroupingModelChange?.(next);
    },
    onExpandedChange: setExpanded,
    getCoreRowModel: getCoreRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    getGroupedRowModel: grouping.length ? getGroupedRowModel() : void 0,
    getSortedRowModel: getSortedRowModel(),
    getExpandedRowModel: grouping.length ? getExpandedRowModel() : void 0,
    ...pagination && { getPaginationRowModel: getPaginationRowModel() },
    getRowId: rowIdFn,
    enableRowSelection: checkboxSelection,
    manualSorting: sortingMode === "server",
    groupedColumnMode: false
  });
  const isFirstRender = React.useRef(true);
  React.useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
      return;
    }
    if (!onRowSelectionChange) return;
    const selected = rows.filter(
      (row, i) => Boolean(rowSelection[rowIdFn(row, i)])
    );
    onRowSelectionChange(selected);
  }, [onRowSelectionChange, rowIdFn, rowSelection, rows]);
  React.useEffect(() => {
    if (!ariaLabel && !ariaLabelledby) {
      console.warn(
        '[DataGrid] Missing accessible name. Add aria-label="Description of this grid" or aria-labelledby="heading-element-id" to comply with WCAG 4.1.2 (role="region" must have an accessible name).'
      );
    }
  }, [ariaLabel, ariaLabelledby]);
  React.useImperativeHandle(
    apiRef,
    () => ({
      exportDataAsCsv: (opts) => exportToCsv(columns, rows, opts ?? {})
    }),
    [columns, rows]
  );
  const commitEdit = React.useCallback(
    async (rowOriginal, field, value) => {
      setEditingCell(null);
      if (processRowUpdate) {
        try {
          await processRowUpdate(
            { ...rowOriginal, [field]: value },
            rowOriginal
          );
        } catch (err) {
          onProcessRowUpdateError?.(err);
        }
      }
    },
    [processRowUpdate, onProcessRowUpdateError]
  );
  const treeMetaMap = React.useMemo(() => {
    if (!treeDisplayList) return null;
    const map = /* @__PURE__ */ new Map();
    for (const item of treeDisplayList) map.set(item.row, item.meta);
    return map;
  }, [treeDisplayList]);
  const handleTreeToggle = React.useCallback((pathKey) => {
    setTreeExpandedPaths((prev) => toggleTreePath(prev, pathKey));
  }, []);
  const visibleRows = table.getRowModel().rows;
  const hClass = HEADER_CLASS[density];
  const cClass = CELL_CLASS[density];
  const Toolbar = slots?.toolbar;
  const NoRowsOverlay = slots?.noRowsOverlay;
  const LoadingOverlay = slots?.loadingOverlay;
  const { pageIndex, pageSize: currentPageSize } = table.getState().pagination;
  const totalRows = rowCount ?? rows.length;
  const firstRow = totalRows === 0 ? 0 : pageIndex * currentPageSize + 1;
  const lastRow = Math.min((pageIndex + 1) * currentPageSize, totalRows);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "data-grid",
      "data-density": density,
      role: "region",
      "aria-label": ariaLabelledby ? void 0 : ariaLabel ?? "Data grid",
      "aria-labelledby": ariaLabelledby,
      "aria-busy": loading || void 0,
      className: cn(
        "relative flex flex-col overflow-hidden rounded border border-border bg-background text-foreground",
        autoHeight ? "h-auto" : "h-full",
        className
      ),
      children: [
        Toolbar && /* @__PURE__ */ jsx(Toolbar, { ...slotProps?.toolbar ?? {} }),
        loading && /* @__PURE__ */ jsx(
          "div",
          {
            "data-slot": "data-grid-loading",
            className: "absolute inset-x-0 top-0 z-raised h-dds-050 overflow-hidden bg-muted",
            children: LoadingOverlay ? /* @__PURE__ */ jsx(LoadingOverlay, { ...slotProps?.loadingOverlay ?? {} }) : /* @__PURE__ */ jsx("div", { className: "h-full w-1/3 animate-pulse bg-primary" })
          }
        ),
        /* @__PURE__ */ jsx("div", { className: "flex-1 overflow-auto", children: /* @__PURE__ */ jsxs(Table, { ...treeData ? { role: "treegrid" } : {}, children: [
          /* @__PURE__ */ jsx(
            DataGridHeader,
            {
              table,
              hClass,
              disableColumnMenu,
              disableColumnResize,
              disableColumnReorder,
              disableColumnFilter,
              dragId,
              setColumnOrder,
              filterModel,
              onFilterChange: handleFilterChange
            }
          ),
          /* @__PURE__ */ jsx(TableBody, { children: visibleRows.length > 0 ? visibleRows.map((row) => {
            if (row.getIsGrouped()) {
              return /* @__PURE__ */ jsx(
                DataGridGroupRow,
                {
                  row,
                  cClass
                },
                row.id
              );
            }
            const treeMeta = treeMetaMap?.get(row.original);
            if (treeMeta) {
              return /* @__PURE__ */ jsx(
                DataGridTreeRow,
                {
                  row,
                  treeMeta,
                  cClass,
                  editingCell,
                  onStartEdit: setEditingCell,
                  onCommitEdit: commitEdit,
                  onCancelEdit: () => setEditingCell(null),
                  onToggle: handleTreeToggle,
                  onRowClick,
                  disableRowSelectionOnClick
                },
                row.id
              );
            }
            const rowInteractive = Boolean(onRowClick);
            const handleRowActivate = onRowClick ? () => onRowClick({ row: row.original, id: row.id }) : void 0;
            return /* @__PURE__ */ jsx(
              TableRow,
              {
                "data-slot": "data-grid-row",
                "data-state": row.getIsSelected() ? "selected" : void 0,
                tabIndex: rowInteractive ? 0 : void 0,
                "aria-label": rowInteractive ? `Row ${row.id}` : void 0,
                className: cn(
                  rowInteractive && !disableRowSelectionOnClick && "cursor-pointer",
                  rowInteractive && "focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none"
                ),
                onClick: handleRowActivate,
                onKeyDown: rowInteractive && handleRowActivate ? (e) => {
                  if (e.target !== e.currentTarget) return;
                  if (e.key === "Enter") {
                    e.preventDefault();
                    handleRowActivate();
                  } else if (e.key === " ") {
                    e.preventDefault();
                    handleRowActivate();
                  }
                } : void 0,
                children: row.getVisibleCells().map((cell) => /* @__PURE__ */ jsx(
                  DataGridCell,
                  {
                    cell,
                    cClass,
                    rowId: row.id,
                    editingCell,
                    onStartEdit: setEditingCell,
                    onCommit: (v) => commitEdit(row.original, cell.column.id, v),
                    onCancel: () => setEditingCell(null)
                  },
                  cell.id
                ))
              },
              row.id
            );
          }) : !loading && /* @__PURE__ */ jsx(TableRow, { children: /* @__PURE__ */ jsx(
            TableCell,
            {
              colSpan: columnDefs.length || 1,
              className: "h-dds-2400 text-center",
              children: /* @__PURE__ */ jsx("div", { "data-slot": "data-grid-no-rows", children: NoRowsOverlay ? /* @__PURE__ */ jsx(
                NoRowsOverlay,
                {
                  ...slotProps?.noRowsOverlay ?? {}
                }
              ) : /* @__PURE__ */ jsx("span", { className: "text-sm text-muted-foreground", children: "No rows" }) })
            }
          ) }) })
        ] }) }),
        !hideFooter && /* @__PURE__ */ jsx(
          DataGridFooter,
          {
            table,
            pagination,
            rowsPerPageOptions,
            totalRows,
            firstRow,
            lastRow,
            currentPageSize,
            setPaginationState
          }
        )
      ]
    }
  );
}

export {
  DataGrid
};
