// @docker/trident/dist/chunk-LO5GFZDE.js
import {
  useNextTheme
} from "./chunk-Z4BVHI6U.js";
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger
} from "./chunk-X3GAIHTA.js";
import {
  Button
} from "./chunk-2MKIJ7TF.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/theme-switcher.tsx
import { Monitor, Moon, Sun } from "lucide-react";
import { jsx, jsxs } from "react/jsx-runtime";
var THEME_ICONS = {
  light: /* @__PURE__ */ jsx(
    Sun,
    {
      "data-testid": "theme-switcher-icon",
      "data-theme": "light",
      className: "h-4 w-4"
    }
  ),
  dark: /* @__PURE__ */ jsx(
    Moon,
    {
      "data-testid": "theme-switcher-icon",
      "data-theme": "dark",
      className: "h-4 w-4"
    }
  ),
  system: /* @__PURE__ */ jsx(
    Monitor,
    {
      "data-testid": "theme-switcher-icon",
      "data-theme": "system",
      className: "h-4 w-4"
    }
  )
};
var THEME_LABELS = {
  light: "Light",
  dark: "Dark",
  system: "System"
};
var ThemeSwitcher = (props) => {
  if (props.systemPreference === true) {
    return /* @__PURE__ */ jsx(ThemeSwitcherSystem, { className: props.className });
  }
  const { theme, onThemeChange, className } = props;
  const isDark = theme === "dark";
  return /* @__PURE__ */ jsx(
    Button,
    {
      variant: "ghost",
      size: "icon",
      "aria-label": isDark ? "Switch to light mode" : "Switch to dark mode",
      onClick: () => onThemeChange(isDark ? "light" : "dark"),
      className: cn(className),
      children: isDark ? /* @__PURE__ */ jsx(
        Sun,
        {
          "data-testid": "theme-switcher-icon",
          "data-theme": "dark",
          className: "h-4 w-4"
        }
      ) : /* @__PURE__ */ jsx(
        Moon,
        {
          "data-testid": "theme-switcher-icon",
          "data-theme": "light",
          className: "h-4 w-4"
        }
      )
    }
  );
};
var ThemeSwitcherSystem = ({ className }) => {
  const { theme, setTheme } = useNextTheme();
  const currentTheme = theme === "light" || theme === "dark" || theme === "system" ? theme : "system";
  return /* @__PURE__ */ jsxs(DropdownMenu, { children: [
    /* @__PURE__ */ jsx(DropdownMenuTrigger, { asChild: true, children: /* @__PURE__ */ jsx(
      Button,
      {
        variant: "ghost",
        size: "icon",
        "aria-label": `Theme: ${THEME_LABELS[currentTheme]}. Click to change`,
        className: cn(className),
        children: THEME_ICONS[currentTheme]
      }
    ) }),
    /* @__PURE__ */ jsx(DropdownMenuContent, { align: "end", children: ["light", "dark", "system"].map((mode) => /* @__PURE__ */ jsxs(
      DropdownMenuItem,
      {
        onSelect: () => setTheme(mode),
        "aria-current": currentTheme === mode ? "true" : void 0,
        className: cn(
          "flex items-center gap-2",
          currentTheme === mode && "font-label"
        ),
        children: [
          THEME_ICONS[mode],
          THEME_LABELS[mode]
        ]
      },
      mode
    )) })
  ] });
};

export {
  ThemeSwitcher
};
