// @docker/trident/dist/chunk-JUGBJ42Q.js
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/ui/toggle.tsx
import { cva } from "class-variance-authority";
import { Toggle as TogglePrimitive } from "radix-ui";
import { jsx } from "react/jsx-runtime";
var toggleVariants = cva(
  "group/toggle inline-flex items-center justify-center gap-1 rounded text-sm font-label whitespace-nowrap transition-[color,box-shadow] ease-default outline-none hover:bg-muted hover:text-foreground focus-visible:border-ring focus-visible:ring-3 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 aria-invalid:border-destructive aria-invalid:ring-destructive/20 aria-pressed:bg-muted dark:aria-invalid:ring-destructive/40 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
  {
    variants: {
      variant: {
        default: "bg-transparent",
        outline: "border border-input bg-transparent shadow-raised hover:bg-muted"
      },
      size: {
        default: "h-9 min-w-9 px-2",
        sm: "h-8 min-w-8 px-1.5",
        lg: "h-10 min-w-10 px-2.5"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
);
function Toggle({
  className,
  variant = "default",
  size = "default",
  ...props
}) {
  return /* @__PURE__ */ jsx(
    TogglePrimitive.Root,
    {
      "data-slot": "toggle",
      className: cn(toggleVariants({ variant, size, className })),
      ...props
    }
  );
}

export {
  toggleVariants,
  Toggle
};
