// @docker/trident/dist/chunk-HU6TU2L2.js
import {
  Button
} from "./chunk-2MKIJ7TF.js";

// src/components/data-grid-footer.tsx
import { ChevronLeft, ChevronRight } from "lucide-react";
import { jsx, jsxs } from "react/jsx-runtime";
function DataGridFooter({
  table,
  pagination,
  rowsPerPageOptions,
  totalRows,
  firstRow,
  lastRow,
  currentPageSize,
  setPaginationState
}) {
  return /* @__PURE__ */ jsxs(
    "nav",
    {
      "data-slot": "data-grid-footer",
      "aria-label": "Table pagination",
      className: "flex items-center justify-between border-t border-border px-dds-400 py-dds-200 text-sm text-muted-foreground",
      children: [
        /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-dds-300", children: [
          /* @__PURE__ */ jsx("span", { "aria-live": "polite", "aria-atomic": "true", children: totalRows === 0 ? "No rows" : pagination ? `${firstRow}\u2013${lastRow} of ${totalRows}` : `${totalRows} row${totalRows !== 1 ? "s" : ""}` }),
          pagination && rowsPerPageOptions && rowsPerPageOptions.length > 0 && /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-dds-150", children: [
            /* @__PURE__ */ jsx(
              "label",
              {
                htmlFor: "data-grid-page-size",
                className: "text-xs text-muted-foreground",
                children: "Rows per page:"
              }
            ),
            /* @__PURE__ */ jsx(
              "select",
              {
                id: "data-grid-page-size",
                value: currentPageSize,
                onChange: (e) => setPaginationState({
                  pageIndex: 0,
                  pageSize: Number(e.target.value)
                }),
                className: "h-7 rounded border border-input bg-transparent px-dds-100 text-sm outline-none focus:ring-1 focus:ring-ring",
                children: rowsPerPageOptions.map((s) => /* @__PURE__ */ jsx("option", { value: s, children: s }, s))
              }
            )
          ] })
        ] }),
        pagination && /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-dds-100", children: [
          /* @__PURE__ */ jsxs(
            Button,
            {
              variant: "ghost",
              size: "sm",
              "aria-label": "Previous page",
              disabled: !table.getCanPreviousPage(),
              onClick: () => table.previousPage(),
              children: [
                /* @__PURE__ */ jsx(ChevronLeft, { "aria-hidden": "true", className: "size-4" }),
                /* @__PURE__ */ jsx("span", { className: "sr-only", children: "Prev" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            Button,
            {
              variant: "ghost",
              size: "sm",
              "aria-label": "Next page",
              disabled: !table.getCanNextPage(),
              onClick: () => table.nextPage(),
              children: [
                /* @__PURE__ */ jsx(ChevronRight, { "aria-hidden": "true", className: "size-4" }),
                /* @__PURE__ */ jsx("span", { className: "sr-only", children: "Next" })
              ]
            }
          )
        ] })
      ]
    }
  );
}

export {
  DataGridFooter
};
