// @docker/trident/dist/chunk-PSIYMFPI.js
import {
  SidebarTrigger
} from "./chunk-BZTV2IDP.js";
import {
  Input
} from "./chunk-5CQB3X63.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/app-header.tsx
import { Slot } from "radix-ui";
import { jsx, jsxs } from "react/jsx-runtime";
var APP_HEADER_INTERACTIVE = "outline-hidden transition-colors focus-visible:ring-3 focus-visible:ring-ring";
var AppHeaderLeftNav = ({ children, className }) => {
  return /* @__PURE__ */ jsx(
    "nav",
    {
      "aria-label": "Primary navigation",
      className: cn("flex items-center gap-dds-150", className),
      children
    }
  );
};
var AppHeaderLink = ({
  asChild = false,
  isActive = false,
  className,
  ...props
}) => {
  const Comp = asChild ? Slot.Root : "a";
  return /* @__PURE__ */ jsx(
    Comp,
    {
      className: cn(
        APP_HEADER_INTERACTIVE,
        "inline-flex items-center rounded-[calc(var(--radius)+1px)] px-dds-200 py-dds-100 whitespace-nowrap",
        "text-label leading-label font-label tracking-label text-header-foreground-muted",
        "hover:bg-primary-foreground-wash hover:text-header-foreground",
        isActive && "bg-primary-foreground-wash text-header-foreground",
        className
      ),
      ...props
    }
  );
};
var AppHeaderPill = ({
  asChild = false,
  className,
  children,
  ...props
}) => {
  const Comp = asChild ? Slot.Root : "span";
  return /* @__PURE__ */ jsx(
    Comp,
    {
      className: cn(
        APP_HEADER_INTERACTIVE,
        "inline-flex items-center gap-dds-150 rounded-full bg-primary-foreground-wash px-dds-200 py-dds-100 whitespace-nowrap text-header-foreground",
        className
      ),
      ...props,
      children
    }
  );
};
var AppHeaderMeta = ({ className, ...props }) => {
  return /* @__PURE__ */ jsx(
    "span",
    {
      className: cn(
        "whitespace-nowrap text-header-foreground-muted",
        className
      ),
      ...props
    }
  );
};
var AppHeaderBrand = ({
  icon,
  title,
  meta,
  className
}) => {
  return /* @__PURE__ */ jsxs("div", { className: cn("flex min-w-0 items-center gap-dds-200", className), children: [
    icon ? /* @__PURE__ */ jsx("span", { className: "flex shrink-0 items-center justify-center text-header-foreground", children: icon }) : null,
    /* @__PURE__ */ jsxs("div", { className: "grid min-w-0 gap-[2px] leading-none", children: [
      /* @__PURE__ */ jsx("span", { className: "truncate text-label leading-label font-label tracking-label text-header-foreground", children: title }),
      meta ? /* @__PURE__ */ jsx(AppHeaderMeta, { className: "truncate text-micro leading-snug tracking-caption", children: meta }) : null
    ] })
  ] });
};
var AppHeaderSearchField = ({
  className,
  ...props
}) => {
  return /* @__PURE__ */ jsx(
    Input,
    {
      className: cn("h-8 w-48 shadow-none lg:w-56", className),
      ...props
    }
  );
};
var AppHeader = ({
  logo,
  leftNav,
  rightNav,
  search,
  className
}) => {
  return /* @__PURE__ */ jsxs(
    "header",
    {
      className: cn(
        "sticky top-0 h-[var(--app-header-height)]",
        "z-sticky",
        "bg-gradient-to-r from-header-from to-header-to text-header-foreground",
        "flex items-center gap-dds-300 px-dds-400",
        "border-b border-accent",
        className
      ),
      children: [
        /* @__PURE__ */ jsxs("div", { className: "flex min-w-0 flex-1 items-center gap-dds-200", children: [
          /* @__PURE__ */ jsx(SidebarTrigger, { className: "shrink-0 text-header-foreground hover:bg-primary-foreground-wash" }),
          /* @__PURE__ */ jsx("div", { className: "shrink-0", children: logo }),
          leftNav && /* @__PURE__ */ jsx("div", { className: "flex min-w-0 items-center gap-dds-200", children: leftNav })
        ] }),
        /* @__PURE__ */ jsxs("div", { className: "flex min-w-0 items-center gap-dds-200", children: [
          search && /* @__PURE__ */ jsx("div", { className: "min-w-0", children: search }),
          /* @__PURE__ */ jsx("div", { className: "shrink-0", children: rightNav })
        ] })
      ]
    }
  );
};

export {
  AppHeaderLeftNav,
  AppHeaderLink,
  AppHeaderPill,
  AppHeaderMeta,
  AppHeaderBrand,
  AppHeaderSearchField,
  AppHeader
};
