// @docker/trident/dist/chunk-SN5QYS5I.js
import {
  RelativeTime
} from "./chunk-EI4H22EW.js";
import {
  cn
} from "./chunk-UEVWMCT6.js";

// src/components/freshness-indicator.tsx
import { jsx } from "react/jsx-runtime";
function inferBand(date) {
  const d = date instanceof Date ? date : new Date(date);
  const diffDays = (Date.now() - d.getTime()) / (1e3 * 60 * 60 * 24);
  if (diffDays < 7) return "fresh";
  if (diffDays < 30) return "aging";
  return "stale";
}
var BAND_COLOR = {
  fresh: "text-success",
  aging: "text-warning",
  stale: "text-destructive"
};
function FreshnessIndicator({
  date,
  band,
  className,
  ...props
}) {
  const resolvedBand = band ?? inferBand(date);
  return /* @__PURE__ */ jsx(
    "span",
    {
      "data-slot": "freshness-indicator",
      className: cn(
        "inline-flex items-center gap-1 text-xs",
        BAND_COLOR[resolvedBand],
        className
      ),
      ...props,
      children: /* @__PURE__ */ jsx(RelativeTime, { date, className: "text-xs text-inherit" })
    }
  );
}

export {
  inferBand,
  FreshnessIndicator
};
