// @docker/trident/dist/components/agent-approval-card.js
import {
  Button
} from "../chunk-2MKIJ7TF.js";
import {
  cn
} from "../chunk-UEVWMCT6.js";

// src/components/agent-approval-card.tsx
import { ShieldAlertIcon } from "lucide-react";
import "react";
import { jsx, jsxs } from "react/jsx-runtime";
function AgentApprovalCard({
  toolName,
  argsPreview,
  destructive = false,
  onApprove,
  onReject,
  children,
  className,
  ...props
}) {
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "approval-card",
      "data-destructive": destructive || void 0,
      className: cn(
        "rounded-dds-radius-sm flex flex-col overflow-hidden border bg-warning-wash",
        destructive ? "border-destructive/40" : "border-warning-subtle",
        className
      ),
      ...props,
      children: [
        /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-1.5 px-3 py-2", children: [
          /* @__PURE__ */ jsx(
            ShieldAlertIcon,
            {
              "aria-hidden": "true",
              className: cn(
                "size-3.5 shrink-0",
                destructive ? "text-destructive" : "text-warning"
              )
            }
          ),
          /* @__PURE__ */ jsx("span", { className: "min-w-0 flex-1 truncate font-mono text-xs font-medium text-foreground", children: toolName }),
          /* @__PURE__ */ jsx("span", { className: "shrink-0 text-[11px] text-muted-foreground", children: "Needs approval" })
        ] }),
        argsPreview && /* @__PURE__ */ jsx("div", { className: "px-3 pb-2", children: /* @__PURE__ */ jsx("p", { className: "font-mono text-[11px] text-muted-foreground", children: argsPreview }) }),
        children && /* @__PURE__ */ jsx("div", { className: "px-3 pb-2 text-xs text-foreground", children }),
        /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-between border-t border-border px-3 py-1.5", children: [
          /* @__PURE__ */ jsx(
            Button,
            {
              variant: "ghost",
              size: "xs",
              onClick: onReject,
              className: "text-muted-foreground hover:text-foreground",
              children: "Reject"
            }
          ),
          /* @__PURE__ */ jsx(Button, { variant: "default", size: "xs", onClick: onApprove, children: "Approve" })
        ] })
      ]
    }
  );
}
export {
  AgentApprovalCard
};
