// @docker/trident/dist/components/agent-mark.js
import {
  cn
} from "../chunk-UEVWMCT6.js";

// src/components/agent-mark.tsx
import { jsx, jsxs } from "react/jsx-runtime";
var CIRCLE_RADIUS = 9.5;
var CUT_HEIGHT = 2.5;
var CUT_Y_POSITIONS = [3.6, 8.7, 13.8];
var STATE_CLASSES = {
  default: "text-primary",
  active: "text-primary animate-agent-breathing",
  resolved: "text-success",
  muted: "text-muted-foreground"
};
var MASK_ID = "dds-agent-mark-mask";
function AgentMark({
  state = "default",
  size = 20,
  className,
  ...props
}) {
  return /* @__PURE__ */ jsx(
    "span",
    {
      "data-slot": "agent-mark",
      "data-state": state,
      className: cn("inline-flex shrink-0", STATE_CLASSES[state], className),
      ...props,
      children: /* @__PURE__ */ jsxs(
        "svg",
        {
          width: size,
          height: size,
          viewBox: "0 0 20 20",
          fill: "none",
          "aria-label": "Agent mark",
          role: "img",
          children: [
            /* @__PURE__ */ jsx("defs", { children: /* @__PURE__ */ jsxs("mask", { id: MASK_ID, children: [
              /* @__PURE__ */ jsx("circle", { cx: "10", cy: "10", r: CIRCLE_RADIUS, fill: "white" }),
              CUT_Y_POSITIONS.map((y) => /* @__PURE__ */ jsx(
                "rect",
                {
                  x: "0",
                  y,
                  width: "20",
                  height: CUT_HEIGHT,
                  fill: "black"
                },
                y
              ))
            ] }) }),
            /* @__PURE__ */ jsx(
              "circle",
              {
                cx: "10",
                cy: "10",
                r: CIRCLE_RADIUS,
                fill: "currentColor",
                mask: `url(#${MASK_ID})`
              }
            )
          ]
        }
      )
    }
  );
}
export {
  AgentMark
};
