// @docker/trident/dist/components/agent-activity-line.js
"use client";
import {
  cn
} from "../chunk-UEVWMCT6.js";

// src/components/agent-activity-line.tsx
import { ChevronRightIcon } from "lucide-react";
import * as React from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function formatDuration(ms) {
  if (ms >= 6e4) {
    return `${Math.floor(ms / 6e4)}m ${Math.round(ms % 6e4 / 1e3)}s`;
  }
  if (ms >= 1e3) {
    return `${(ms / 1e3).toFixed(1)}s`;
  }
  return `${Math.round(ms)}ms`;
}
var STATE_ICON_CLASSES = {
  running: "text-muted-foreground",
  done: "text-muted-foreground",
  error: "text-destructive"
};
function AgentActivityLine({
  verb,
  target,
  result,
  isError = false,
  icon,
  state = "done",
  duration,
  expandable: expandableProp,
  children,
  className,
  ...props
}) {
  const [open, setOpen] = React.useState(false);
  const isExpandable = expandableProp ?? Boolean(children);
  const resolvedState = isError ? "error" : state;
  const iconColorClass = STATE_ICON_CLASSES[resolvedState];
  const resultColorClass = resolvedState === "error" ? "text-destructive" : "text-success";
  const rowContent = /* @__PURE__ */ jsxs(Fragment, { children: [
    icon !== void 0 && /* @__PURE__ */ jsx(
      "span",
      {
        "data-slot": "activity-icon",
        "aria-hidden": "true",
        className: cn(
          "flex size-4 shrink-0 items-center justify-center",
          iconColorClass
        ),
        children: icon
      }
    ),
    /* @__PURE__ */ jsx("span", { className: "shrink-0 font-[520] text-muted-foreground", children: verb }),
    /* @__PURE__ */ jsx("span", { className: "min-w-0 flex-1 truncate font-mono text-foreground", children: target }),
    result !== void 0 && /* @__PURE__ */ jsx(
      "span",
      {
        "data-slot": "activity-result",
        className: cn("shrink-0 font-mono", resultColorClass),
        children: result
      }
    ),
    resolvedState === "done" && duration !== void 0 && /* @__PURE__ */ jsx(
      "span",
      {
        "data-slot": "activity-duration",
        className: "shrink-0 font-mono text-[11px] text-muted-foreground",
        children: formatDuration(duration)
      }
    ),
    isExpandable && /* @__PURE__ */ jsx(
      ChevronRightIcon,
      {
        "aria-hidden": "true",
        className: cn(
          "size-3 shrink-0 text-muted-foreground transition-transform duration-150",
          open && "rotate-90"
        )
      }
    )
  ] });
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "activity-line",
      "data-state": resolvedState,
      className: cn("flex flex-col", className),
      ...props,
      children: [
        isExpandable ? /* @__PURE__ */ jsx(
          "button",
          {
            type: "button",
            "aria-expanded": open,
            onClick: () => setOpen((v) => !v),
            className: "flex h-7 w-full items-center gap-2 rounded-sm text-left text-xs hover:bg-hover-overlay focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none",
            children: rowContent
          }
        ) : /* @__PURE__ */ jsx("div", { className: "flex h-7 items-center gap-2 text-xs", children: rowContent }),
        isExpandable && open && children && /* @__PURE__ */ jsxs(
          "div",
          {
            "data-slot": "activity-expanded",
            className: cn("flex gap-2", icon !== void 0 ? "pl-6" : "pl-0"),
            children: [
              icon !== void 0 && /* @__PURE__ */ jsx(
                "div",
                {
                  "aria-hidden": "true",
                  className: "mr-1 ml-[-1.25rem] w-px shrink-0 self-stretch bg-border"
                }
              ),
              /* @__PURE__ */ jsx("div", { className: "min-w-0 flex-1 py-1 pb-2 text-xs text-muted-foreground", children })
            ]
          }
        )
      ]
    }
  );
}
export {
  AgentActivityLine
};
