// @docker/trident/dist/components/agent-composer.js
"use client";
import {
  cn
} from "../chunk-UEVWMCT6.js";

// src/components/agent-composer.tsx
import { ArrowUpIcon } from "lucide-react";
import * as React from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function AgentComposer({
  placeholder = "Message\u2026",
  disabled = false,
  onSubmit,
  children,
  className,
  ...props
}) {
  const [value, setValue] = React.useState("");
  const textareaRef = React.useRef(null);
  const hasContent = value.trim().length > 0;
  function handleKeyDown(e) {
    if (e.key === "Enter" && !e.shiftKey) {
      e.preventDefault();
      submit();
    }
  }
  function submit() {
    const trimmed = value.trim();
    if (!trimmed || disabled) return;
    onSubmit(trimmed);
    setValue("");
    if (textareaRef.current) {
      textareaRef.current.style.height = "auto";
    }
  }
  function handleInput(e) {
    setValue(e.target.value);
    const el = e.target;
    el.style.height = "auto";
    el.style.height = `${Math.min(el.scrollHeight, 160)}px`;
  }
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "agent-composer",
      className: cn(
        "flex flex-col rounded-xl border border-border bg-popover",
        disabled && "opacity-50",
        className
      ),
      ...props,
      children: [
        /* @__PURE__ */ jsx(
          "textarea",
          {
            ref: textareaRef,
            "data-slot": "composer-textarea",
            value,
            onChange: handleInput,
            onKeyDown: handleKeyDown,
            placeholder,
            disabled,
            rows: 1,
            "aria-label": placeholder,
            className: cn(
              "w-full resize-none bg-transparent px-3 pt-2.5 pb-1 text-sm text-foreground placeholder:text-muted-foreground disabled:cursor-not-allowed",
              // min-h for single line, max-h enforced via JS
              "min-h-[36px]"
            ),
            style: {
              outline: "none",
              overflowY: "auto"
            }
          }
        ),
        /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-between px-2 pb-2", children: [
          /* @__PURE__ */ jsx("div", { className: "flex items-center gap-1", children }),
          /* @__PURE__ */ jsx(
            "button",
            {
              type: "button",
              "data-slot": "composer-send",
              "aria-label": "Send message",
              disabled: disabled || !hasContent,
              onClick: submit,
              className: cn(
                "flex size-7 shrink-0 items-center justify-center rounded-full bg-primary text-primary-foreground transition-opacity duration-150",
                "focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none",
                "disabled:pointer-events-none",
                hasContent ? "opacity-100" : "pointer-events-none opacity-0"
              ),
              children: /* @__PURE__ */ jsx(ArrowUpIcon, { "aria-hidden": "true", className: "size-3.5" })
            }
          )
        ] })
      ]
    }
  );
}
export {
  AgentComposer
};
