// @docker/trident/dist/components/conversation-layout.js
import {
  cn
} from "../chunk-UEVWMCT6.js";

// src/components/conversation-layout.tsx
import { jsx } from "react/jsx-runtime";
function ConversationLayout({
  width = 380,
  className,
  children
}) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "conversation-layout",
      style: { width },
      className: cn(
        "flex shrink-0 flex-col overflow-hidden border-r border-border bg-card",
        className
      ),
      children
    }
  );
}
function ConversationLayoutHeader({
  className,
  ...props
}) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "conversation-layout-header",
      className: cn(
        "flex h-10 shrink-0 items-center border-b border-border-subtle px-4",
        className
      ),
      ...props
    }
  );
}
function ConversationLayoutMessages({
  className,
  ...props
}) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "conversation-layout-messages",
      className: cn("flex min-h-0 flex-1 flex-col overflow-y-auto", className),
      ...props
    }
  );
}
function ConversationLayoutInput({
  className,
  ...props
}) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "conversation-layout-input",
      className: cn("shrink-0 border-t border-border-subtle", className),
      ...props
    }
  );
}
var ConversationLayoutCompound = Object.assign(ConversationLayout, {
  Header: ConversationLayoutHeader,
  Messages: ConversationLayoutMessages,
  Input: ConversationLayoutInput
});
export {
  ConversationLayoutCompound as ConversationLayout
};
