// @docker/trident/dist/components/inspector-panel.js
import {
  cn
} from "../chunk-UEVWMCT6.js";

// src/components/inspector-panel.tsx
import { X } from "lucide-react";
import { jsx, jsxs } from "react/jsx-runtime";
function InspectorPanel({
  open,
  onClose,
  title,
  width = 320,
  children,
  className
}) {
  return /* @__PURE__ */ jsxs(
    "aside",
    {
      "data-slot": "inspector-panel",
      className: cn(
        "absolute top-0 right-0 z-overlay flex h-full flex-col overflow-hidden",
        "bg-[var(--card)] transition-transform duration-normal ease-[var(--dds-ease-out)]",
        open ? "translate-x-0" : "pointer-events-none translate-x-full",
        className
      ),
      style: {
        width,
        boxShadow: open ? "-8px 0 24px oklch(0 0 0 / 0.25), -1px 0 0 oklch(0 0 0 / 0.1)" : "none"
      },
      "aria-hidden": !open,
      inert: !open,
      children: [
        title !== void 0 && /* @__PURE__ */ jsxs(
          "header",
          {
            "data-slot": "inspector-panel-header",
            className: "flex h-9 shrink-0 items-center justify-between border-b border-[var(--border)] px-3.5",
            children: [
              /* @__PURE__ */ jsx("span", { className: "min-w-0 truncate font-mono text-[13px] font-semibold tracking-[-0.011em] text-[var(--foreground)]", children: title }),
              /* @__PURE__ */ jsx(
                "button",
                {
                  type: "button",
                  onClick: onClose,
                  "aria-label": "Close",
                  className: "flex size-5 shrink-0 items-center justify-center rounded-[var(--dds-radius-sm)] text-muted-foreground transition-colors hover:bg-muted hover:text-foreground focus-visible:ring-2 focus-visible:ring-ring focus-visible:outline-none",
                  children: /* @__PURE__ */ jsx(X, { className: "size-3" })
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ jsx(
          "div",
          {
            "data-slot": "inspector-panel-content",
            className: "flex-1 overflow-y-auto",
            children
          }
        )
      ]
    }
  );
}
export {
  InspectorPanel
};
