// @docker/trident/dist/components/node-popover.js
import {
  cn
} from "../chunk-UEVWMCT6.js";

// src/components/node-popover.tsx
import { createPortal } from "react-dom";
import { jsx, jsxs } from "react/jsx-runtime";
var NODE_POPOVER_WIDTH = 220;
var NODE_POPOVER_GAP = 8;
function NodePopover({
  anchorRect,
  onClose,
  children,
  "aria-label": ariaLabel
}) {
  const viewportWidth = typeof window !== "undefined" ? window.innerWidth : 1440;
  const fitsRight = anchorRect.left + anchorRect.width + NODE_POPOVER_GAP + NODE_POPOVER_WIDTH < viewportWidth;
  const style = fitsRight ? {
    position: "fixed",
    top: anchorRect.top,
    left: anchorRect.left + anchorRect.width + NODE_POPOVER_GAP,
    width: NODE_POPOVER_WIDTH,
    zIndex: 50
  } : {
    position: "fixed",
    top: anchorRect.top,
    left: anchorRect.left - NODE_POPOVER_WIDTH - NODE_POPOVER_GAP,
    width: NODE_POPOVER_WIDTH,
    zIndex: 50
  };
  return createPortal(
    /* @__PURE__ */ jsx(
      "div",
      {
        role: "dialog",
        "aria-modal": "false",
        "aria-label": ariaLabel ?? "Node details",
        style: { ...style, boxShadow: "var(--dds-material-floating-shadow)" },
        className: "rounded-[var(--dds-radius-default)] border border-border bg-popover p-3",
        onClick: (e) => e.stopPropagation(),
        onKeyDown: (e) => {
          if (e.key === "Escape") onClose?.();
        },
        children
      }
    ),
    document.body
  );
}
function NodePopoverHeader({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "node-popover-header",
      className: cn("flex items-center gap-2", className),
      ...props
    }
  );
}
function NodePopoverTitle({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "span",
    {
      "data-slot": "node-popover-title",
      className: cn(
        "text-[13px] font-label tracking-[-0.008em] text-foreground",
        className
      ),
      ...props
    }
  );
}
var BADGE_TONE_CLASS = {
  ok: "text-success",
  err: "text-destructive",
  warn: "text-warning",
  loading: "text-primary"
};
function NodePopoverBadge({
  tone = "loading",
  className,
  ...props
}) {
  return /* @__PURE__ */ jsx(
    "span",
    {
      "data-slot": "node-popover-badge",
      className: cn("font-mono text-[12px]", BADGE_TONE_CLASS[tone], className),
      ...props
    }
  );
}
function NodePopoverSection({
  label,
  className,
  children,
  ...props
}) {
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-slot": "node-popover-section",
      className: cn("mt-2.5 border-t border-border-subtle pt-2", className),
      ...props,
      children: [
        label && /* @__PURE__ */ jsx("span", { className: "text-[10px] font-label tracking-[0.06em] text-muted-foreground uppercase", children: label }),
        /* @__PURE__ */ jsx("div", { className: cn("flex flex-col gap-0.5", label && "mt-1"), children })
      ]
    }
  );
}
function NodePopoverMetric({ label, value }) {
  return /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-3 font-mono text-[12px] text-muted-foreground", children: [
    /* @__PURE__ */ jsx("span", { className: "text-muted-foreground", children: label }),
    /* @__PURE__ */ jsx("span", { children: value })
  ] });
}
function NodePopoverLink({ className, ...props }) {
  return /* @__PURE__ */ jsx(
    "span",
    {
      "data-slot": "node-popover-link",
      className: cn("font-mono text-[12px] text-muted-foreground", className),
      ...props
    }
  );
}
function NodePopoverActions({
  className,
  ...props
}) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-slot": "node-popover-actions",
      className: cn(
        "mt-2.5 flex items-center gap-1.5 border-t border-border-subtle pt-2",
        className
      ),
      ...props
    }
  );
}
var NodePopoverCompound = Object.assign(NodePopover, {
  Header: NodePopoverHeader,
  Title: NodePopoverTitle,
  Badge: NodePopoverBadge,
  Section: NodePopoverSection,
  Metric: NodePopoverMetric,
  Link: NodePopoverLink,
  Actions: NodePopoverActions
});
export {
  NodePopoverCompound as NodePopover
};
