// @docker/trident/dist/components/session-picker.js
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList
} from "../chunk-MP6BNYMD.js";
import "../chunk-HYW6PLXL.js";
import "../chunk-S3W6RB5E.js";
import "../chunk-H3A5BH2J.js";
import {
  Popover,
  PopoverContent,
  PopoverTrigger
} from "../chunk-R6CWLGJU.js";
import "../chunk-5CQB3X63.js";
import "../chunk-2MKIJ7TF.js";
import {
  cn
} from "../chunk-UEVWMCT6.js";

// src/components/session-picker.tsx
import { Check, ChevronDown, Plus } from "lucide-react";
import { useState } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function SessionPicker({
  sessions,
  activeId,
  onSelect,
  onNew,
  className
}) {
  const [open, setOpen] = useState(false);
  const activeSession = sessions.find((s) => s.id === activeId);
  const title = activeSession?.title?.trim() || "New conversation";
  return /* @__PURE__ */ jsxs(Popover, { open, onOpenChange: setOpen, children: [
    /* @__PURE__ */ jsx(PopoverTrigger, { asChild: true, children: /* @__PURE__ */ jsxs(
      "button",
      {
        type: "button",
        "aria-expanded": open,
        "aria-haspopup": "listbox",
        "aria-label": `Current session: ${title}. Click to switch.`,
        className: cn(
          "flex min-w-0 items-center gap-2 rounded-[var(--dds-radius-sm)] px-1 py-1 transition-colors",
          "hover:bg-hover-overlay",
          "focus-visible:ring-2 focus-visible:ring-primary-subtle focus-visible:outline-none",
          className
        ),
        children: [
          /* @__PURE__ */ jsx("span", { className: "truncate text-[13px] font-label tracking-[-0.008em] text-foreground", children: title }),
          /* @__PURE__ */ jsx(
            ChevronDown,
            {
              className: "size-3 shrink-0 text-muted-foreground",
              "aria-hidden": true
            }
          )
        ]
      }
    ) }),
    /* @__PURE__ */ jsx(
      PopoverContent,
      {
        className: "w-72 p-0",
        align: "start",
        sideOffset: 4,
        "aria-label": "Sessions",
        children: /* @__PURE__ */ jsxs(Command, { children: [
          /* @__PURE__ */ jsx(CommandInput, { placeholder: "Search sessions..." }),
          /* @__PURE__ */ jsxs(CommandList, { children: [
            /* @__PURE__ */ jsx(CommandEmpty, { className: "py-4 text-center text-[13px] text-muted-foreground", children: "No sessions." }),
            /* @__PURE__ */ jsx(CommandGroup, { children: sessions.map((session) => /* @__PURE__ */ jsxs(
              CommandItem,
              {
                value: session.id,
                onSelect: () => {
                  onSelect(session.id);
                  setOpen(false);
                },
                className: "flex items-center gap-2",
                children: [
                  /* @__PURE__ */ jsx("span", { className: "min-w-0 flex-1 truncate text-[13px]", children: session.title?.trim() || "New conversation" }),
                  session.updatedAt && /* @__PURE__ */ jsx("span", { className: "shrink-0 font-mono text-[11px] text-muted-foreground", children: session.updatedAt }),
                  session.id === activeId && /* @__PURE__ */ jsx(
                    Check,
                    {
                      className: "size-3 shrink-0 text-primary",
                      "aria-hidden": true
                    }
                  )
                ]
              },
              session.id
            )) }),
            onNew && /* @__PURE__ */ jsxs(Fragment, { children: [
              /* @__PURE__ */ jsx("div", { className: "mx-2 border-t border-border-subtle" }),
              /* @__PURE__ */ jsx(CommandGroup, { children: /* @__PURE__ */ jsxs(
                CommandItem,
                {
                  onSelect: () => {
                    onNew();
                    setOpen(false);
                  },
                  className: "gap-2 text-primary",
                  children: [
                    /* @__PURE__ */ jsx(Plus, { className: "size-3.5 shrink-0", "aria-hidden": true }),
                    /* @__PURE__ */ jsx("span", { className: "text-[13px]", children: "New conversation" })
                  ]
                }
              ) })
            ] })
          ] })
        ] })
      }
    )
  ] });
}
export {
  SessionPicker
};
