// @docker/trident/dist/components/topology-node.js
import {
  InlineSegmentedToggle
} from "../chunk-R2QGPDNO.js";
import {
  cn
} from "../chunk-UEVWMCT6.js";

// src/components/topology-node.tsx
import { ChevronRight, Cloud, Home, KeyRound } from "lucide-react";
import { jsx, jsxs } from "react/jsx-runtime";
var TOPOLOGY_NODE_W = 140;
var TOPOLOGY_NODE_H = 88;
var MODE_TOGGLE_OPTIONS = [
  { value: "local", label: "local" },
  { value: "remote", label: "remote" }
];
var BADGE_DOT_CLASS = {
  ok: "bg-success",
  err: "bg-destructive",
  warn: "bg-warning",
  loading: "bg-primary animate-pulse"
};
var BADGE_TEXT_CLASS = {
  ok: "text-success",
  err: "font-semibold text-destructive",
  warn: "text-warning",
  loading: "text-primary"
};
function formatRate(n) {
  if (n >= 1e3) return `${(n / 1e3).toFixed(n >= 1e4 ? 0 : 1)}k`;
  return String(n);
}
function TopologyNode({
  label,
  mode,
  endpoint,
  badge,
  badgeTone = "loading",
  selected = false,
  requestsPerSec,
  latencyMs,
  secretCount,
  hasDrilldown = false,
  extra,
  onClick,
  onDoubleClick,
  onToggleMode,
  className
}) {
  const isRemote = mode === "remote";
  const trafficParts = [];
  if (typeof requestsPerSec === "number")
    trafficParts.push(`${formatRate(requestsPerSec)} req/s`);
  if (isRemote && typeof latencyMs === "number")
    trafficParts.push(`${latencyMs}ms`);
  const trafficLine = trafficParts.join(" \xB7 ");
  const decorativeShadow = selected ? "0 0 12px color-mix(in oklch, var(--primary) 15%, transparent), inset 0 1px 0 var(--highlight)" : "inset 0 1px 0 var(--highlight)";
  return /* @__PURE__ */ jsxs("div", { className: cn("flex w-full flex-col", className), children: [
    /* @__PURE__ */ jsxs(
      "button",
      {
        type: "button",
        onClick,
        onDoubleClick,
        "aria-pressed": selected,
        style: {
          height: onToggleMode ? void 0 : TOPOLOGY_NODE_H,
          "--topology-node-shadow": decorativeShadow
        },
        className: cn(
          "relative flex w-full cursor-pointer flex-col border border-border bg-popover pt-1.5 pr-2.5 pb-1.5 pl-2.5 text-left transition-all duration-100",
          // Decorative inset highlight + selected glow on a pseudo-layer so
          // focus-visible ring (compiled to box-shadow by Tailwind) is not
          // clobbered.
          'before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:shadow-[var(--topology-node-shadow)] before:content-[""]',
          onToggleMode ? "rounded-t-[var(--dds-radius-default)]" : "rounded-[var(--dds-radius-default)]",
          "hover:-translate-y-px hover:border-muted-foreground",
          "focus-visible:ring-3 focus-visible:ring-ring focus-visible:outline-none",
          selected && "border-primary"
        ),
        children: [
          /* @__PURE__ */ jsxs("div", { className: "flex items-start gap-1", children: [
            isRemote ? /* @__PURE__ */ jsx(
              Cloud,
              {
                "aria-label": "remote",
                className: "mt-px mr-1 size-3 shrink-0 text-muted-foreground"
              }
            ) : /* @__PURE__ */ jsx(
              Home,
              {
                "aria-label": "local",
                className: "mt-px mr-1 size-3 shrink-0 text-primary"
              }
            ),
            /* @__PURE__ */ jsx("span", { className: "min-w-0 flex-1 truncate text-[14px] leading-[1.3] font-label tracking-[-0.011em] text-foreground", children: label }),
            secretCount !== null && secretCount !== void 0 && secretCount > 0 && /* @__PURE__ */ jsx(
              "span",
              {
                className: "mt-px flex size-3 shrink-0 items-center justify-center text-muted-foreground",
                title: `${secretCount} secret${secretCount === 1 ? "" : "s"}`,
                children: /* @__PURE__ */ jsx(
                  KeyRound,
                  {
                    "aria-label": `${secretCount} secret${secretCount === 1 ? "" : "s"}`,
                    className: "size-3"
                  }
                )
              }
            ),
            hasDrilldown && /* @__PURE__ */ jsx(
              ChevronRight,
              {
                "aria-label": "drill down",
                className: "mt-px size-3 shrink-0 text-primary"
              }
            )
          ] }),
          endpoint && /* @__PURE__ */ jsx("span", { className: "block truncate font-mono text-[12px] leading-[1.4] text-muted-foreground", children: endpoint }),
          trafficLine && /* @__PURE__ */ jsx("span", { className: "block truncate font-mono text-[11px] leading-[1.35] text-muted-foreground", children: trafficLine }),
          badge && /* @__PURE__ */ jsxs("div", { className: "mt-1 flex items-center gap-1 font-mono text-[12px]", children: [
            /* @__PURE__ */ jsx(
              "span",
              {
                "aria-hidden": true,
                className: cn(
                  "size-[6px] shrink-0 rounded-full",
                  BADGE_DOT_CLASS[badgeTone]
                )
              }
            ),
            /* @__PURE__ */ jsx("span", { className: cn("truncate", BADGE_TEXT_CLASS[badgeTone]), children: badge })
          ] }),
          extra && /* @__PURE__ */ jsx("span", { className: "mt-0.5 block font-mono text-[10px] leading-[1.3] text-primary", children: extra })
        ]
      }
    ),
    onToggleMode && /* @__PURE__ */ jsx("div", { className: "rounded-b-[var(--dds-radius-default)] border border-t-0 border-border bg-popover px-2.5 pb-1.5", children: /* @__PURE__ */ jsx(
      InlineSegmentedToggle,
      {
        options: MODE_TOGGLE_OPTIONS,
        value: mode,
        onChange: onToggleMode
      }
    ) })
  ] });
}
export {
  TOPOLOGY_NODE_H,
  TOPOLOGY_NODE_W,
  TopologyNode
};
