// @docker/trident/src/components/code-block/code-block-server.tsx
/**
 * React Server Component (RSC) variant of `CodeBlockContent`.
 *
 * Why this file exists alongside `index.tsx`:
 *   - `index.tsx` is a `'use client'` component. Client components cannot be
 *     `async`, so it runs shiki's `codeToHtml` inside a `useEffect`, shows a
 *     plain `<pre>` fallback first, and then swaps in the highlighted HTML
 *     once shiki's WASM has loaded in the browser. Highlighting always
 *     happens on the client.
 *   - This file has no `'use client'` directive and the component itself is
 *     `async`, which only works as a React Server Component. `codeToHtml`
 *     runs on the server during render and the response contains the
 *     already-highlighted HTML — no shiki JS or WASM is shipped to the
 *     client for this piece.
 *
 * When to use which:
 *   - In an RSC-capable framework (e.g. Next.js App Router), import
 *     `CodeBlockContent` from `code-block/server` to get true server-rendered
 *     syntax highlighting with zero client-side highlighting cost.
 *   - In a non-RSC app (plain SPA, or traditional SSR like Vite SSR without
 *     RSC support), this file is unusable — there is no RSC runtime to
 *     execute an `async` component. Use the client `CodeBlockContent` from
 *     `index.tsx` instead.
 *
 * Selection is manual: you must explicitly import from `code-block/server`.
 * Per the kibo-ui docs:
 *   "You can use the RSC version of the Code Block component by importing
 *    the `CodeBlockContent` component from `code-block/server`"
 */

import type {HTMLAttributes} from 'react';

import {
  transformerNotationDiff,
  transformerNotationErrorLevel,
  transformerNotationFocus,
  transformerNotationHighlight,
  transformerNotationWordHighlight,
} from '@shikijs/transformers';
import {
  type BundledLanguage,
  type CodeOptionsMultipleThemes,
  codeToHtml,
} from 'shiki';

export type CodeBlockContentProps = HTMLAttributes<HTMLDivElement> & {
  themes?: CodeOptionsMultipleThemes['themes'];
  language?: BundledLanguage;
  children: string;
  syntaxHighlighting?: boolean;
};

export const CodeBlockContent = async ({
  children,
  themes,
  language,
  syntaxHighlighting = true,
  ...props
}: CodeBlockContentProps) => {
  if (!syntaxHighlighting) {
    return (
      <div {...props}>
        <pre>
          <code>{children}</code>
        </pre>
      </div>
    );
  }

  const html = await codeToHtml(children, {
    lang: language ?? 'typescript',
    themes: themes ?? {
      light: 'github-light',
      dark: 'github-dark-default',
    },
    transformers: [
      transformerNotationDiff({
        matchAlgorithm: 'v3',
      }),
      transformerNotationHighlight({
        matchAlgorithm: 'v3',
      }),
      transformerNotationWordHighlight({
        matchAlgorithm: 'v3',
      }),
      transformerNotationFocus({
        matchAlgorithm: 'v3',
      }),
      transformerNotationErrorLevel({
        matchAlgorithm: 'v3',
      }),
    ],
  });

  return <div dangerouslySetInnerHTML={{__html: html}} {...props} />;
};
